{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "327e58cd-d270-4af2-b3be-4199b82e5a5e",
   "metadata": {},
   "source": [
    "# DNA Clustering and Classification\n",
    "\n",
    "This notebook explores whether we can use mutational and copy number information to identify clusters of GLASS samples that overlap with Verhaak transcriptional subtype.\n",
    "\n",
    "The actual consensus clustering is done by the NMFConsensus script from GenePattern here: https://github.com/genepattern/NMFConsensus/blob/develop/nmfconsensus.R, and the clustering parameters are as listed here: gs://fc-secure-3f0f305e-f110-432b-bc9a-c8d96824cea2/jweiss/local_files/results/2024-07-22_GLASS-Verhaak-classifier/GLASS_sample_clustering.params.txt\n",
    "\n",
    "The notebook also trains a classifier to predict membership in these DNA-based clusters and then applies this classifier to the ICB OncoPanel data.\n",
    "\n",
    "This notebook generates the following figures:\n",
    "* The clustering heatmap shown in Fig. 1c\n",
    "* The sankey plot shown in Extended Data Fig. 3d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1ad4df9b-047d-408a-acbf-7fddc296f9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn import metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9527aa9a-803b-4a97-91a1-1b63e67486e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "45b03d54-e19d-47d6-ae00-a6b78cf90767",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dir = '../results'\n",
    "if not os.path.exists(results_dir):\n",
    "    os.mkdir(results_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e690a97a-9ae1-4557-bbb3-1581b6574563",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load and preprocess GLASS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e68c7b-6a7d-424d-b669-e3b902ea8d4f",
   "metadata": {},
   "source": [
    "## Clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "82324a6c-798f-4b2a-bad9-4fcd54acf3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_surgeries_df = pd.read_csv('../data/glass_clinical_surgeries_table.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a62849fe-9475-4602-96a8-420cbf344afb",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_adjusted_os = pd.read_csv(\"../data/sample_clinical_data_GLASS_removenonGBM_removeIDHmut_flagICB.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ab2684b5-4846-46ef-9b80-8f3be5be8207",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_surgeries_df = (glass_surgeries_df[(glass_surgeries_df[\"sample_barcode\"].notnull()) & \\\n",
    "                                        (glass_surgeries_df[\"idh_status\"] != \"IDHmut\") & \\\n",
    "                                        (glass_surgeries_df[\"histology\"].str.match(\"Glioblastoma|Gliosarcoma\").fillna(False).astype(bool))]#[\"case_barcode\"].nunique()\n",
    "                     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3174a055-06b1-4b25-b46a-eb961175ef37",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_samples_df = (glass_surgeries_df.merge(glass_adjusted_os[[\"sample_barcode\", \"sample_overall_survival_days\"]], \n",
    "                                             on=\"sample_barcode\", how=\"left\")\n",
    "                    .set_index(\"sample_barcode\")\n",
    "                   )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bc56e85-de94-481a-8728-29c9b8872aa7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## WES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaec0811-e0f2-4d4d-9384-6d89c83f7aad",
   "metadata": {},
   "source": [
    "### Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f04862e1-7178-4640-82cd-da64c18b59b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "glass_variants_passgeno_df = pd.read_csv('../data/glass_variants_passgeno.csv', sep = ',', low_memory = False)\n",
    "glass_variants_anno_df = pd.read_csv('../data/glass_variants_anno.csv', sep = ',', low_memory = False)\n",
    "oncopanel_genes = pd.read_csv(os.path.join('../data/oncopanel_exonic_v123.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6e5ceaec-7a11-4f0a-84b6-8ff79f50053c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create and save \"sample\" column\n",
    "glass_variants_passgeno_df[\"sample\"] = glass_variants_passgeno_df[\"aliquot_barcode\"].str[:15]\n",
    "wes_mut_samples = set(glass_variants_passgeno_df[\"sample\"])\n",
    "\n",
    "# Keep only the variants detected by single-sample MuTect2\n",
    "glass_variants_passgeno_df = glass_variants_passgeno_df[glass_variants_passgeno_df[\"ssm2_pass_call\"] == \"t\"]\n",
    "\n",
    "# Keep only variants above the VAF threshold (0.1)\n",
    "VAF_THRESHOLD = 0.1\n",
    "glass_variants_passgeno_df = glass_variants_passgeno_df[glass_variants_passgeno_df['af'] >= VAF_THRESHOLD]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b6cfe8d-3f95-4f18-b310-cd2a764c8977",
   "metadata": {},
   "source": [
    "Join the variant annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3d8aa671-2b32-4491-85df-482e8d00bbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def append_variant_id(df, colnames=[\"chrom\",\"start\",\"end\",\"alt\"]):\n",
    "    df[\"variant_id\"] = df[colnames[0]].astype(str)\n",
    "    for c in colnames[1:]:\n",
    "        df[\"variant_id\"] = df[\"variant_id\"] + \":\" + df[c].astype(str)\n",
    "\n",
    "def chrom_float_to_int(chrom_float):\n",
    "    if np.isnan(chrom_float):\n",
    "        return -1\n",
    "    else:\n",
    "        return int(chrom_float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "85dc6542-ec07-4c3d-a956-4085ae9cb74a",
   "metadata": {},
   "outputs": [],
   "source": [
    "append_variant_id(glass_variants_passgeno_df)\n",
    "append_variant_id(glass_variants_anno_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d0fb47b8-e924-4f11-9576-6b90699861e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_variants_anno_df.set_index(\"variant_id\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c9aba2a5-8ea1-4eca-ab21-00665149208f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3014019, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:168513274:168513274:G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168513274.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>XCL2</td>\n",
       "      <td>FIVE_PRIME_FLANK</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:168545864:168545864:G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168545864.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>XCL1</td>\n",
       "      <td>FIVE_PRIME_UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:197398786:197398786:A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>197398786.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CRB1</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:211652463:211652463:C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>211652463.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>RD3</td>\n",
       "      <td>MISSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.T168S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:229771597:229771597:T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>229771597.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>URB2</td>\n",
       "      <td>NONSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.K413*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample               variant_id  chrom        start ref alt  \\\n",
       "583  GLSS-19-0266-R1  1:168513274:168513274:G    1.0  168513274.0   A   G   \n",
       "584  GLSS-19-0266-R1  1:168545864:168545864:G    1.0  168545864.0   A   G   \n",
       "593  GLSS-19-0266-R1  1:197398786:197398786:A    1.0  197398786.0   G   A   \n",
       "598  GLSS-19-0266-R1  1:211652463:211652463:C    1.0  211652463.0   G   C   \n",
       "601  GLSS-19-0266-R1  1:229771597:229771597:T    1.0  229771597.0   A   T   \n",
       "\n",
       "    gene_symbol variant_classification variant_type protein_change  \n",
       "583        XCL2       FIVE_PRIME_FLANK          SNP            NaN  \n",
       "584        XCL1         FIVE_PRIME_UTR          SNP            NaN  \n",
       "593        CRB1                 INTRON          SNP            NaN  \n",
       "598         RD3               MISSENSE          SNP        p.T168S  \n",
       "601        URB2               NONSENSE          SNP        p.K413*  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_passgeno_cols = [\"sample\", \"variant_id\"] \n",
    "kept_anno_cols = [\"chrom\", \"start\", \"ref\", \"alt\", \"gene_symbol\", \"variant_classification\", \"variant_type\", \"protein_change\"]\n",
    "\n",
    "glass_variants_df = glass_variants_passgeno_df[kept_passgeno_cols].join(glass_variants_anno_df[kept_anno_cols], on=\"variant_id\", how=\"left\")\n",
    "print(glass_variants_df.shape)\n",
    "glass_variants_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04fe63b0-45c8-4ee8-8070-daf63fd2812a",
   "metadata": {},
   "source": [
    "Reorder/filter the columns to a more standard format.\n",
    "\n",
    "Keep only genes with exons covered by all versions (v1, v2, v3) of OncoPanel.\n",
    "\n",
    "Keep only variants on valid chromosomes (1-X).\n",
    "\n",
    "Make sure variant positions are stored as integers.\n",
    "\n",
    "Remove duplicate rows (not sure why these exist)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "556aab90-65e1-4d75-8eee-3c12f880ad57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29772, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>3</td>\n",
       "      <td>70005722</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MITF</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>13</td>\n",
       "      <td>28913225</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>FLT1</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>X</td>\n",
       "      <td>123220608</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>STAG2</td>\n",
       "      <td>NONSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.Q1089*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1267</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>69990330</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>MITF</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1422</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>7</td>\n",
       "      <td>140481402</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>MISSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.G469A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample chrom      start ref alt gene_symbol  \\\n",
       "687   GLSS-19-0266-R1     3   70005722   G   A        MITF   \n",
       "970   GLSS-19-0266-R1    13   28913225   T   G        FLT1   \n",
       "1160  GLSS-19-0266-R1     X  123220608   C   T       STAG2   \n",
       "1267  GLSS-19-0266-TP     3   69990330   G   T        MITF   \n",
       "1422  GLSS-19-0266-TP     7  140481402   C   G        BRAF   \n",
       "\n",
       "     variant_classification variant_type protein_change  \n",
       "687                  INTRON          SNP            NaN  \n",
       "970                  INTRON          SNP            NaN  \n",
       "1160               NONSENSE          SNP       p.Q1089*  \n",
       "1267                 INTRON          SNP            NaN  \n",
       "1422               MISSENSE          SNP        p.G469A  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_glass_cols = [\"sample\", \"chrom\", \"start\", \"ref\", \"alt\", \"gene_symbol\", \"variant_classification\", \"variant_type\", \"protein_change\"]\n",
    "glass_variants_df = glass_variants_df.loc[:, kept_glass_cols]\n",
    "\n",
    "glass_variants_df = glass_variants_df[glass_variants_df['gene_symbol'].isin(oncopanel_genes['Gene'])]\n",
    "\n",
    "glass_variants_df[\"chrom\"] = glass_variants_df[\"chrom\"].map(chrom_float_to_int).astype(int)\n",
    "glass_variants_df = glass_variants_df[glass_variants_df[\"chrom\"] > 0] # Missing chromosome labels were mapped to -1\n",
    "glass_variants_df[\"chrom\"] = np.where(glass_variants_df[\"chrom\"] == 23, \"X\", glass_variants_df[\"chrom\"])\n",
    "glass_variants_df[\"start\"] = glass_variants_df[\"start\"].astype(int)\n",
    "\n",
    "glass_variants_df = glass_variants_df.drop_duplicates()\n",
    "print(glass_variants_df.shape)\n",
    "glass_variants_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4176a6db-26ad-46a0-9334-608a48c5fac2",
   "metadata": {},
   "source": [
    "Keep only non-synonymous variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0c43605d-182c-4644-9148-8022bd5bf297",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "INTRON                     23579\n",
       "MISSENSE                    2521\n",
       "FIVE_PRIME_FLANK            1102\n",
       "SILENT                       799\n",
       "THREE_PRIME_UTR              792\n",
       "SPLICE_SITE                  253\n",
       "FRAME_SHIFT_DEL              230\n",
       "NONSENSE                     213\n",
       "FIVE_PRIME_UTR               180\n",
       "FRAME_SHIFT_INS               49\n",
       "IN_FRAME_DEL                  40\n",
       "IN_FRAME_INS                   4\n",
       "START_CODON_SNP                3\n",
       "DE_NOVO_START_IN_FRAME         3\n",
       "DE_NOVO_START_OUT_FRAME        3\n",
       "NONSTOP                        1\n",
       "Name: variant_classification, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_variants_df['variant_classification'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "14c9d1b5-8366-446c-90d4-e20d41c58c93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MISSENSE                   2521\n",
       "SPLICE_SITE                 253\n",
       "FRAME_SHIFT_DEL             230\n",
       "NONSENSE                    213\n",
       "FRAME_SHIFT_INS              49\n",
       "IN_FRAME_DEL                 40\n",
       "IN_FRAME_INS                  4\n",
       "START_CODON_SNP               3\n",
       "DE_NOVO_START_IN_FRAME        3\n",
       "DE_NOVO_START_OUT_FRAME       3\n",
       "NONSTOP                       1\n",
       "Name: variant_classification, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonsyn_classifications = ['MISSENSE',\n",
    "                          'SPLICE_SITE',\n",
    "                          'FRAME_SHIFT_DEL',\n",
    "                          'NONSENSE',\n",
    "                          'FRAME_SHIFT_INS',\n",
    "                          'IN_FRAME_DEL',\n",
    "                          'IN_FRAME_INS',\n",
    "                          'START_CODON_SNP',\n",
    "                          'DE_NOVO_START_IN_FRAME',\n",
    "                          'DE_NOVO_START_OUT_FRAME',\n",
    "                          'NONSTOP']\n",
    "\n",
    "glass_variants_df = glass_variants_df[glass_variants_df['variant_classification'].isin(nonsyn_classifications)]\n",
    "glass_variants_df['variant_classification'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cb43ab0-c6e3-4583-8067-2e8be6e2690f",
   "metadata": {},
   "source": [
    "### CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7841966b-e6a6-430d-aaa4-8915c1d6962e",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_cnvs_df = pd.read_csv('../data/glass_cnvs.csv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bde655e2-960d-4910-af81-3e2e55719666",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35563, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliquot_barcode</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>hlvl_call</th>\n",
       "      <th>wcr</th>\n",
       "      <th>sample</th>\n",
       "      <th>cnv</th>\n",
       "      <th>hlvl_call_abs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2228044</th>\n",
       "      <td>GLSS-AT-GP01-TP-02D-WXS-DPVLWZ</td>\n",
       "      <td>PDCD1LG2</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.807710</td>\n",
       "      <td>GLSS-AT-GP01-TP</td>\n",
       "      <td>del</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156480</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.141363</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156491</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>EP300</td>\n",
       "      <td>1</td>\n",
       "      <td>1.100724</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156625</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>MCL1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.145647</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156707</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>CREBBP</td>\n",
       "      <td>1</td>\n",
       "      <td>1.109192</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         aliquot_barcode gene_symbol  hlvl_call       wcr  \\\n",
       "2228044   GLSS-AT-GP01-TP-02D-WXS-DPVLWZ    PDCD1LG2         -1  0.807710   \n",
       "29156480  GLSS-MD-0044-R1-01D-WGS-VG0QP6        ABL1          1  1.141363   \n",
       "29156491  GLSS-MD-0044-R1-01D-WGS-VG0QP6       EP300          1  1.100724   \n",
       "29156625  GLSS-MD-0044-R1-01D-WGS-VG0QP6        MCL1          1  1.145647   \n",
       "29156707  GLSS-MD-0044-R1-01D-WGS-VG0QP6      CREBBP          1  1.109192   \n",
       "\n",
       "                   sample  cnv  hlvl_call_abs  \n",
       "2228044   GLSS-AT-GP01-TP  del              1  \n",
       "29156480  GLSS-MD-0044-R1  amp              1  \n",
       "29156491  GLSS-MD-0044-R1  amp              1  \n",
       "29156625  GLSS-MD-0044-R1  amp              1  \n",
       "29156707  GLSS-MD-0044-R1  amp              1  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create and save \"sample\" column\n",
    "glass_cnvs_df[\"sample\"] = glass_cnvs_df[\"aliquot_barcode\"].str[:15]\n",
    "wes_cnv_samples = set(glass_cnvs_df[\"sample\"])\n",
    "\n",
    "# Keep only CNVs (not copy-neutral segments)\n",
    "glass_cnvs_df = glass_cnvs_df[glass_cnvs_df['hlvl_call'] != 0]\n",
    "\n",
    "# Keep only genes with exons covered by all versions (v1, v2, v3) of OncoPanel.\n",
    "glass_cnvs_df = glass_cnvs_df[glass_cnvs_df[\"gene_symbol\"].isin(oncopanel_genes[\"Gene\"])]\n",
    "\n",
    "# Simplify CNV calls to just amp or del\n",
    "glass_cnvs_df['cnv'] = glass_cnvs_df['hlvl_call'].map({2: 'amp', 1: 'amp', -1: 'del', -2: 'del'})\n",
    "\n",
    "# Scoring\n",
    "glass_cnvs_df['hlvl_call_abs'] = np.abs(glass_cnvs_df['hlvl_call'])\n",
    "\n",
    "# Deduplicate, keeping strongest event\n",
    "glass_cnvs_df.sort_values('hlvl_call_abs', inplace = True)\n",
    "glass_cnvs_df = glass_cnvs_df.drop_duplicates(['sample', 'gene_symbol', 'cnv'], keep = 'first')\n",
    "\n",
    "print(glass_cnvs_df.shape)\n",
    "glass_cnvs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc7a0373-09de-47f1-b46e-998531885a4c",
   "metadata": {},
   "source": [
    "### Intersect sample lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "48113151-34b6-40ac-a71e-0da8f533005c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(478, 30)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_samples_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f526a132-40c4-4fa0-813e-a6066aa2e50b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "389"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_intersection = (set(glass_samples_df.index).\\\n",
    "    intersection(wes_mut_samples).\\\n",
    "    intersection(wes_cnv_samples))\n",
    "\n",
    "len(sample_intersection)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c9be5e-17b8-4a8f-9fb7-2cf4bdbd6bc6",
   "metadata": {},
   "source": [
    "### Subset dataframes to these samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d34953d1-af8f-49f7-a359-f0778de5bac7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "330 361\n"
     ]
    }
   ],
   "source": [
    "variants_subset = glass_variants_df[glass_variants_df['sample'].isin(sample_intersection)]\n",
    "\n",
    "cnvs_subset = glass_cnvs_df[glass_cnvs_df['sample'].isin(sample_intersection)]\n",
    "\n",
    "print(variants_subset['sample'].nunique(),\n",
    "      cnvs_subset['sample'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4e8315f1-e262-4f01-9609-9047b2979f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_variant_median_rates(variant_df, pt_col=\"sample\", variant_col=\"hlvl_call\"):\n",
    "    counts = variant_df[variant_col].value_counts()\n",
    "    per_pt_median_rates = variant_df.groupby([variant_col, pt_col]).size().reset_index().pivot_table(index=pt_col, columns=variant_col, fill_value=0.0).droplevel(0,axis=1).median()\n",
    "    plt.bar(per_pt_median_rates.index, per_pt_median_rates.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dc0b2c55-e873-4444-822a-1c1a5e4617b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGxCAYAAADCo9TSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAlO0lEQVR4nO3deVTVdf7H8deN5YKCFKFsEhBZai4zLuNSCmZxpDLTMpcybHCOldY4ztTPZRxQS0odxxpzaSM7uWWlLXpKKkXnqKWenMqW0QlHDMmlBLcw8PP7o8M9Xi8I6OVzAZ+Pc+45c7/3e+99f799B55+74LDGGMEAABgyWW+HgAAAFxaiA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA/U2iuvvCKHwyGHw6ENGzZ43G6M0TXXXCOHw6GUlBSvPndCQoJGjhzpur5hw4Yq57Bl06ZNuueeexQbG6vAwECFhYWpZ8+eWrBggU6cOOFaLyEhQQ6HQw8++KDHY1RsxxtvvCFJGjhwoIKDg3X06NEqn/fee+9VQECAfvjhB69vky1r165VVlaWr8do0AoLC5WVlaWdO3d63JaVlSWHw1Hrxxw5cqQSEhIufjgv4ThpfIgPXLDQ0FC99NJLHsvz8vL03//+V6GhoXU+Q6dOnbRlyxZ16tSpzp+rMpmZmerdu7e+//57TZ8+Xbm5uVq+fLn69u2rrKws/fWvf/W4z0svvaRvv/32vI+bkZGhn3/+WUuXLq309uLiYq1atUq33367IiMjvbItvrB27VpNnTrV12M0aIWFhZo6dWql8TFq1Cht2bKl1o85ZcoUrVq1ygvTeQfHSeNDfOCCDRkyRG+++aZKSkrclr/00kvq0aOHrrrqqjqfoVmzZurevbuaNWtW5891rpUrV2ratGnKyMjQv/71L/3+979XcnKy0tLSNH36dO3Zs0dpaWlu9+nRo4eaNm2qSZMmnfex09LSFBMTo5dffrnS25ctW6ZTp04pIyPDa9uDmisvL1dpaWm9n6Fly5bq3r17rR87KSlJv/3tby90NKBaxAcu2LBhwyT9+ouwQnFxsd588039/ve/r/Q+p0+f1hNPPKHWrVvL6XSqefPmeuCBB3To0CG39X755Rc9/vjjioqKUpMmTXTjjTfq008/9Xi8yl522b59u4YOHaqEhAQFBwcrISFBw4YN0//+9z+3+1a8fLR+/Xo99NBDioiI0JVXXqlBgwapsLCw2u2fNm2arrjiCj377LOVntoODQ1Vamqq27Lw8HBNmDBBb731lrZu3VrlY/v5+Sk9PV07duzQF1984XF7Tk6OoqOjPeKmMkuXLlWPHj0UEhKikJAQ/eY3v/E4Y/Xyyy+rY8eOCgoKUnh4uAYOHKivv/7abZ2UlJRKX0Y79xT93r175XA4NHv2bM2ZM0eJiYkKCQlRjx493LZ55MiReu655yTJ9TKew+HQ3r17Jf0ad926dVNYWJiaNGmiq6++usrj6mwOh0Njx47VokWLdO2118rpdKpt27Zavny5x7pFRUUaPXq0WrZsqcDAQCUmJmrq1KkqKyvz2J6ZM2fqiSeeUGJiopxOp9avX3/RMxw6dEgPP/yw2rZtq5CQELVo0UI33XSTNm3a5Lbe+Wbo2rWrJOmBBx5w7cOKlyiqetmlumOispdd6mqbLvY4QQNlgFrKyckxksy2bdvMiBEjzO9+9zvXbQsWLDBNmzY1JSUl5vrrrzfJycmu28rLy02/fv1M06ZNzdSpU01ubq558cUXTWxsrGnbtq05efKka9309HTjcDjMY489ZtatW2fmzJljYmNjTbNmzUx6erprvfXr1xtJZv369a5lK1euNH/729/MqlWrTF5enlm+fLlJTk42zZs3N4cOHfLYjquvvto88sgj5oMPPjAvvviiueKKK0yfPn3Ouw8KCwuNJDNkyJAa77f4+Hhz2223mZMnT5rY2FjTq1cvj+1YuXKla9nu3buNw+Ew48aNc3ucXbt2GUlmwoQJ1T7nlClTjCQzaNAgs3LlSte+nDJlimudGTNmGElm2LBhZs2aNebVV181V199tQkLCzP/+c9/XOslJye7/feskJ6ebuLj413X8/PzjSSTkJBg+vXrZ1avXm1Wr15t2rdvb6644gpz9OhRY4wxe/bsMXfffbeRZLZs2eK6/Pzzz2bz5s3G4XCYoUOHmrVr15qPP/7Y5OTkmBEjRlS7zZJMXFycadu2rVm2bJl55513TL9+/Tz274EDB0xcXJyJj483ixYtMh9++KGZPn26cTqdZuTIkR7bExsba/r06WPeeOMNs27dOpOfn3/RM3zzzTfmoYceMsuXLzcbNmww7733nsnIyDCXXXaZ2zFd1Qz//ve/XcfxX//6V9c+LCgoMMYYk5mZac79MV+TY+Lc/6Z1uU0Xc5yg4SI+UGtnx0fFL80vv/zSGGNM165dXT+4z42PZcuWGUnmzTffdHu8bdu2GUlm/vz5xhhjvv76ayPJ/OlPf3Jbb8mSJUZStfFxrrKyMnP8+HHTtGlT88wzz3hsx8MPP+y2/syZM40kc+DAgSofc+vWrTUOgAoV8WGMMS+88IKRZN5991237Tj7h7gxv/7Cj4iIMKdPn3Yt+/Of/2wkuYVBZb777jvj5+dn7r333irX+emnn0xwcLC59dZb3Zbv27fPOJ1OM3z4cLdZahMf7du3N2VlZa7ln376qZFkli1b5lo2ZswYj1+Oxhgze/ZsI8n1C6g2JJng4GBTVFTkWlZWVmZat25trrnmGtey0aNHm5CQEPO///2v0ufetWuX2/YkJSW5/XfwxgznKisrM7/88ovp27evGThwoGv5+Wao+P9PTk6Ox+OdGx81OSaMqTo+6mKbLuY4QcPFyy64KMnJyUpKStLLL7+sL774Qtu2bavy1Ph7772nyy+/XP3791dZWZnr8pvf/EZRUVGul04qTmffe++9bve/55575O/vX+1Mx48f1//93//pmmuukb+/v/z9/RUSEqITJ054vJQgSXfccYfb9Q4dOkiSx8s03vTAAw+obdu2mjBhgs6cOVPlehkZGTp8+LDeeecdSVJZWZlee+019erVS61atTrvc+Tm5qq8vFxjxoypcp0tW7bo1KlTbp8gkqS4uDjddNNN+uijj2q+Uee47bbb5Ofn57pem/1a8VLCPffco9dff13ff/99rZ67b9++bm/E9fPz05AhQ7Rnzx7t379f0q/HY58+fRQTE+N2PFa8lJWXl+f2mHfccYcCAgK8OoMkLVy4UJ06dVJQUJD8/f0VEBCgjz76qMpjtTYznKsmx8T51MU2XcxxgoaL+MBFcTgceuCBB/Taa69p4cKFuvbaa9WrV69K1/3hhx909OhRBQYGKiAgwO1SVFSkw4cPS5KOHDkiSYqKinK7v7+/v6688spqZxo+fLjmzZunUaNG6YMPPtCnn36qbdu2qXnz5jp16pTH+uc+ptPplKRK161Q8Wba/Pz8auepjJ+fn2bMmKFdu3Zp8eLFVa539913KywsTDk5OZJ+fdf/Dz/8UKM3mla8j6Zly5ZVrlOxr6Ojoz1ui4mJcd1+IS5kv1bo3bu3Vq9erbKyMt1///1q2bKl2rVr5/b+ovM599g5e1nFNv3www969913PY7F66+/XpJcx2OFyvbRxc4wZ84cPfTQQ+rWrZvefPNNbd26Vdu2bVO/fv0q3U+1neFcNTkmzqcutulijhM0XNX/MxKoxsiRI/W3v/1NCxcu1JNPPlnlehVv6Hz//fcrvb3io7kVP4yKiooUGxvrur2srKzaX4bFxcV67733lJmZqQkTJriWl5aW6scff6zxNlUnOjpa7du317p163Ty5Ek1adKk1o8xYMAA3XDDDcrMzNTzzz9f6TrBwcEaNmyYXnjhBR04cEAvv/yyQkNDNXjw4Gofv3nz5pKk/fv3Ky4urtJ1Kvb1gQMHPG4rLCxURESE63pQUJCKi4s91jv3l7S3DBgwQAMGDFBpaam2bt2q7OxsDR8+XAkJCerRo8d571tUVFTlsoptjoiIUIcOHao8ZmNiYtyu1/b7Mmoyw2uvvaaUlBQtWLDAbb1jx45V+pgX8p0dZ6vJMXE+dbFNuDRx5gMXLTY2Vo899pj69++v9PT0Kte7/fbbdeTIEZWXl6tLly4el+uuu06SXJ+oWLJkidv9X3/9dbdPIVTG4XDIGOP611OFF198UeXl5RewdVWbMmWKfvrpJz366KMyxnjcfvz4ca1bt+68j/H000+roKBAzz77bJXrZGRkqLy8XLNmzdLatWs1dOjQGsVOamqq/Pz8PH4JnK1Hjx4KDg7Wa6+95rZ8//79+vjjj9W3b1/XsoSEBP3nP/9x+3jnkSNHtHnz5mpnqUpN/pXrdDqVnJysp59+WpL02WefVfu4H330kduXr5WXl2vFihVKSkpy/av/9ttv15dffqmkpKRKj8dz46O2ajKDw+HwOFY///zzWn03R23OFNTkmDgfW9t0Ls6GND6c+YBXPPXUU9WuM3ToUC1ZskS33nqr/vjHP+p3v/udAgICtH//fq1fv14DBgzQwIED1aZNG913332aO3euAgICdPPNN+vLL7/U7Nmzq/0+j2bNmql3796aNWuWIiIilJCQoLy8PL300ku6/PLLvbS1vxo8eLCmTJmi6dOn65tvvlFGRoaSkpJ08uRJffLJJ1q0aJGGDBni8XHbs91www0aMGCA3n777SrX6dKlizp06KC5c+fKGFPj7/ZISEjQpEmTNH36dJ06dUrDhg1TWFiYvvrqKx0+fFhTp07V5ZdfrilTpmjSpEm6//77NWzYMB05ckRTp05VUFCQMjMzXY83YsQILVq0SPfdd5/+8Ic/6MiRI5o5c+ZFfcdK+/btJf0aYWlpafLz81OHDh30xBNPaP/+/erbt69atmypo0eP6plnnlFAQICSk5OrfdyIiAjddNNNmjJlipo2bar58+frm2++cftY6LRp05Sbm6uePXvq0Ucf1XXXXaeff/5Ze/fu1dq1a7Vw4cILfnmipjPcfvvtmj59ujIzM5WcnKxvv/1W06ZNU2JiYrWhXSEpKUnBwcFasmSJ2rRpo5CQEMXExFQaTzU5JurDNp2rquMkMDDwgh4P9YCP3/CKBujsT7ucz7mfdjHGmF9++cXMnj3bdOzY0QQFBZmQkBDTunVrM3r0aLN7927XeqWlpebPf/6zadGihQkKCjLdu3c3W7ZsMfHx8dV+2mX//v3mrrvuMldccYUJDQ01/fr1M19++aXHfavajpp8guZseXl55u677zbR0dEmICDANGvWzPTo0cPMmjXLlJSUuNY7+9MuZ/vqq6+Mn59fpZ92qfDMM88YSaZt27Y1mulsr776qunatatrf//2t7/1+GTEiy++aDp06GACAwNNWFiYGTBggOvTHmdbvHixadOmjQkKCjJt27Y1K1asqPLTLrNmzfK4vySTmZnpul5aWmpGjRplmjdvbhwOh5Fk8vPzzXvvvWfS0tJMbGysCQwMNC1atDC33nqr2bRpU7XbK8mMGTPGzJ8/3yQlJZmAgADTunVrs2TJEo91Dx06ZB599FGTmJhoAgICTHh4uOncubOZPHmyOX78eLXbc7EzlJaWmr/85S8mNjbWBAUFmU6dOpnVq1fXap8a8+snyVq3bm0CAgLc9nFlH7U1pvpjoqpPu9jappoeJ2i4HMZUcr4YABooh8OhMWPGaN68eZf0DN7WGLcJvsN7PgAAgFXEBwAAsIqXXQAAgFWc+QAAAFYRHwAAwCriAwAAWFXvvmTszJkzKiwsVGho6EV/lTAAALDDGKNjx44pJiZGl112/nMb9S4+CgsLL+hvDgAAAN8rKCio9tuB6118VPxxsYKCgov62mYAAGBPSUmJ4uLiXL/Hz6fexUfFSy3NmjUjPgAAaGBq8pYJ3nAKAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWOXv6wFwaUiYsMbXIzQYe5+6zdcjAECd4swHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYJW/rwcAgMYmYcIaX4/QYOx96jZfjwAf4MwHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABW1So+srOz1bVrV4WGhqpFixa688479e2337qtY4xRVlaWYmJiFBwcrJSUFO3atcurQwMAgIarVvGRl5enMWPGaOvWrcrNzVVZWZlSU1N14sQJ1zozZ87UnDlzNG/ePG3btk1RUVG65ZZbdOzYMa8PDwAAGp5a/WG5999/3+16Tk6OWrRooR07dqh3794yxmju3LmaPHmyBg0aJElavHixIiMjtXTpUo0ePdp7kwMAgAbpot7zUVxcLEkKDw+XJOXn56uoqEipqamudZxOp5KTk7V58+ZKH6O0tFQlJSVuFwAA0HhdcHwYYzR+/HjdeOONateunSSpqKhIkhQZGem2bmRkpOu2c2VnZyssLMx1iYuLu9CRAABAA3DB8TF27Fh9/vnnWrZsmcdtDofD7boxxmNZhYkTJ6q4uNh1KSgouNCRAABAA1Cr93xUeOSRR/TOO+9o48aNatmypWt5VFSUpF/PgERHR7uWHzx40ONsSAWn0ymn03khYwAAgAaoVmc+jDEaO3as3nrrLX388cdKTEx0uz0xMVFRUVHKzc11LTt9+rTy8vLUs2dP70wMAAAatFqd+RgzZoyWLl2qt99+W6Ghoa73cYSFhSk4OFgOh0Pjxo3TjBkz1KpVK7Vq1UozZsxQkyZNNHz48DrZAAAA0LDUKj4WLFggSUpJSXFbnpOTo5EjR0qSHn/8cZ06dUoPP/ywfvrpJ3Xr1k3r1q1TaGioVwYGAAANW63iwxhT7ToOh0NZWVnKysq60JkAAEAjxt92AQAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAqlrHx8aNG9W/f3/FxMTI4XBo9erVbrePHDlSDofD7dK9e3dvzQsAABq4WsfHiRMn1LFjR82bN6/Kdfr166cDBw64LmvXrr2oIQEAQOPhX9s7pKWlKS0t7bzrOJ1ORUVFXfBQAACg8aqT93xs2LBBLVq00LXXXqs//OEPOnjwYJXrlpaWqqSkxO0CAAAaL6/HR1pampYsWaKPP/5Yf//737Vt2zbddNNNKi0trXT97OxshYWFuS5xcXHeHgkAANQjtX7ZpTpDhgxx/e927dqpS5cuio+P15o1azRo0CCP9SdOnKjx48e7rpeUlBAgAAA0Yl6Pj3NFR0crPj5eu3fvrvR2p9Mpp9NZ12MAAIB6os6/5+PIkSMqKChQdHR0XT8VAABoAGp95uP48ePas2eP63p+fr527typ8PBwhYeHKysrS3fddZeio6O1d+9eTZo0SRERERo4cKBXBwcAAA1TreNj+/bt6tOnj+t6xfs10tPTtWDBAn3xxRd69dVXdfToUUVHR6tPnz5asWKFQkNDvTc1AABosGodHykpKTLGVHn7Bx98cFEDAQCAxo2/7QIAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKyqdXxs3LhR/fv3V0xMjBwOh1avXu12uzFGWVlZiomJUXBwsFJSUrRr1y5vzQsAABq4WsfHiRMn1LFjR82bN6/S22fOnKk5c+Zo3rx52rZtm6KionTLLbfo2LFjFz0sAABo+Pxre4e0tDSlpaVVepsxRnPnztXkyZM1aNAgSdLixYsVGRmppUuXavTo0Rc3LQAAaPC8+p6P/Px8FRUVKTU11bXM6XQqOTlZmzdvrvQ+paWlKikpcbsAAIDGy6vxUVRUJEmKjIx0Wx4ZGem67VzZ2dkKCwtzXeLi4rw5EgAAqGfq5NMuDofD7boxxmNZhYkTJ6q4uNh1KSgoqIuRAABAPVHr93ycT1RUlKRfz4BER0e7lh88eNDjbEgFp9Mpp9PpzTEAAEA95tUzH4mJiYqKilJubq5r2enTp5WXl6eePXt686kAAEADVeszH8ePH9eePXtc1/Pz87Vz506Fh4frqquu0rhx4zRjxgy1atVKrVq10owZM9SkSRMNHz7cq4MDAICGqdbxsX37dvXp08d1ffz48ZKk9PR0vfLKK3r88cd16tQpPfzww/rpp5/UrVs3rVu3TqGhod6bGgAANFi1jo+UlBQZY6q83eFwKCsrS1lZWRczFwAAaKT42y4AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFjl9fjIysqSw+Fwu0RFRXn7aQAAQAPlXxcPev311+vDDz90Xffz86uLpwEAAA1QncSHv78/ZzsAAECl6uQ9H7t371ZMTIwSExM1dOhQfffdd1WuW1paqpKSErcLAABovLweH926ddOrr76qDz74QC+88IKKiorUs2dPHTlypNL1s7OzFRYW5rrExcV5eyQAAFCPeD0+0tLSdNddd6l9+/a6+eabtWbNGknS4sWLK11/4sSJKi4udl0KCgq8PRIAAKhH6uQ9H2dr2rSp2rdvr927d1d6u9PplNPprOsxAABAPVHn3/NRWlqqr7/+WtHR0XX9VAAAoAHwenz85S9/UV5envLz8/XJJ5/o7rvvVklJidLT0739VAAAoAHy+ssu+/fv17Bhw3T48GE1b95c3bt319atWxUfH+/tpwIAAA2Q1+Nj+fLl3n5IAADQiPC3XQAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGCVv68HsC1hwhpfj9Bg7H3qNl+PAABohDjzAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFjl7+sBAADwhoQJa3w9QoOx96nbfPr8nPkAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMCqOouP+fPnKzExUUFBQercubM2bdpUV08FAAAakDqJjxUrVmjcuHGaPHmyPvvsM/Xq1UtpaWnat29fXTwdAABoQOokPubMmaOMjAyNGjVKbdq00dy5cxUXF6cFCxbUxdMBAIAGxOt/1fb06dPasWOHJkyY4LY8NTVVmzdv9li/tLRUpaWlruvFxcWSpJKSEm+PJkk6U3qyTh63MfLmfwP2e83V1bEPezjea46fM75RFz9nKh7TGFPtul6Pj8OHD6u8vFyRkZFuyyMjI1VUVOSxfnZ2tqZOneqxPC4uztujoZbC5vp6gksT+x2XEo5336jL/X7s2DGFhYWddx2vx0cFh8Phdt0Y47FMkiZOnKjx48e7rp85c0Y//vijrrzyykrXb4xKSkoUFxengoICNWvWzNfjXDLY7/axz32D/e4bl9p+N8bo2LFjiomJqXZdr8dHRESE/Pz8PM5yHDx40ONsiCQ5nU45nU63ZZdffrm3x2oQmjVrdkkcoPUN+90+9rlvsN9941La79Wd8ajg9TecBgYGqnPnzsrNzXVbnpubq549e3r76QAAQANTJy+7jB8/XiNGjFCXLl3Uo0cPPf/889q3b58efPDBung6AADQgNRJfAwZMkRHjhzRtGnTdODAAbVr105r165VfHx8XTxdg+d0OpWZmenx8hPqFvvdPva5b7DffYP9XjWHqclnYgAAALyEv+0CAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriox7Zu3evMjIylJiYqODgYCUlJSkzM1OnT5/29WiN3pNPPqmePXuqSZMml+w37Nowf/58JSYmKigoSJ07d9amTZt8PVKjt3HjRvXv318xMTFyOBxavXq1r0dq9LKzs9W1a1eFhoaqRYsWuvPOO/Xtt9/6eqx6hfioR7755hudOXNGixYt0q5du/SPf/xDCxcu1KRJk3w9WqN3+vRpDR48WA899JCvR2m0VqxYoXHjxmny5Mn67LPP1KtXL6WlpWnfvn2+Hq1RO3HihDp27Kh58+b5epRLRl5ensaMGaOtW7cqNzdXZWVlSk1N1YkTJ3w9Wr3B93zUc7NmzdKCBQv03Xff+XqUS8Irr7yicePG6ejRo74epdHp1q2bOnXqpAULFriWtWnTRnfeeaeys7N9ONmlw+FwaNWqVbrzzjt9Pcol5dChQ2rRooXy8vLUu3dvX49TL3Dmo54rLi5WeHi4r8cALsrp06e1Y8cOpaamui1PTU3V5s2bfTQVYEdxcbEk8bP8LMRHPfbf//5X//znP/mbOGjwDh8+rPLyco+/bB0ZGenxF7CBxsQYo/Hjx+vGG29Uu3btfD1OvUF8WJCVlSWHw3Hey/bt293uU1hYqH79+mnw4MEaNWqUjyZv2C5kv6NuORwOt+vGGI9lQGMyduxYff7551q2bJmvR6lX6uQPy8Hd2LFjNXTo0POuk5CQ4PrfhYWF6tOnj+svAuPC1Ha/o+5ERETIz8/P4yzHwYMHPc6GAI3FI488onfeeUcbN25Uy5YtfT1OvUJ8WBAREaGIiIgarfv999+rT58+6ty5s3JycnTZZZyculC12e+oW4GBgercubNyc3M1cOBA1/Lc3FwNGDDAh5MB3meM0SOPPKJVq1Zpw4YNSkxM9PVI9Q7xUY8UFhYqJSVFV111lWbPnq1Dhw65bouKivLhZI3fvn379OOPP2rfvn0qLy/Xzp07JUnXXHONQkJCfDtcIzF+/HiNGDFCXbp0cZ3V27dvH+9pqmPHjx/Xnj17XNfz8/O1c+dOhYeH66qrrvLhZI3XmDFjtHTpUr399tsKDQ11nfELCwtTcHCwj6erJwzqjZycHCOp0gvqVnp6eqX7ff369b4erVF57rnnTHx8vAkMDDSdOnUyeXl5vh6p0Vu/fn2lx3Z6erqvR2u0qvo5npOT4+vR6g2+5wMAAFjFGwoAAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFb9P6ANf78tcSG3AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_variant_median_rates(cnvs_subset, pt_col=\"sample\", variant_col=\"hlvl_call\")\n",
    "plt.title(\"Median CNV counts per participant\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00682ef6-0bd3-40de-8014-0bada5fd0d49",
   "metadata": {},
   "source": [
    "### Create a combined mutation + CNV + subtype matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "865ddd1e-cef4-4f2f-b8ba-fd84270b056c",
   "metadata": {},
   "source": [
    "#### Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "970b4eea-37c0-40ba-acea-862b73f2aeca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_21339/389233590.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  variants_subset['gene_mut'] = variants_subset['gene_symbol'] + '_mut'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_mut</th>\n",
       "      <th>ABL1_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>AKT2_mut</th>\n",
       "      <th>AKT3_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>TSC1_mut</th>\n",
       "      <th>TSC2_mut</th>\n",
       "      <th>U2AF1_mut</th>\n",
       "      <th>WRN_mut</th>\n",
       "      <th>WT1_mut</th>\n",
       "      <th>XPA_mut</th>\n",
       "      <th>XPC_mut</th>\n",
       "      <th>XPO1_mut</th>\n",
       "      <th>ZNF217_mut</th>\n",
       "      <th>ZRSR2_mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 211 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_mut         ABL1_mut  AKT1_mut  AKT2_mut  AKT3_mut  ALK_mut  APC_mut  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1         0         0         0         0        0        0   \n",
       "GLSS-19-0266-TP         0         0         0         0        0        0   \n",
       "GLSS-19-0267-R1         0         0         0         0        0        0   \n",
       "GLSS-19-0267-TP         0         0         0         0        0        0   \n",
       "GLSS-19-0268-R1         0         0         0         0        0        0   \n",
       "\n",
       "gene_mut         ARAF_mut  ARID1A_mut  ARID1B_mut  ARID2_mut  ...  TSC1_mut  \\\n",
       "sample                                                        ...             \n",
       "GLSS-19-0266-R1         0           0           0          0  ...         0   \n",
       "GLSS-19-0266-TP         0           0           0          0  ...         0   \n",
       "GLSS-19-0267-R1         0           0           0          0  ...         0   \n",
       "GLSS-19-0267-TP         0           0           0          0  ...         0   \n",
       "GLSS-19-0268-R1         0           0           0          0  ...         0   \n",
       "\n",
       "gene_mut         TSC2_mut  U2AF1_mut  WRN_mut  WT1_mut  XPA_mut  XPC_mut  \\\n",
       "sample                                                                     \n",
       "GLSS-19-0266-R1         0          0        0        0        0        0   \n",
       "GLSS-19-0266-TP         0          0        0        0        0        0   \n",
       "GLSS-19-0267-R1         0          0        0        0        0        0   \n",
       "GLSS-19-0267-TP         0          0        0        0        0        0   \n",
       "GLSS-19-0268-R1         0          0        0        0        0        0   \n",
       "\n",
       "gene_mut         XPO1_mut  ZNF217_mut  ZRSR2_mut  \n",
       "sample                                            \n",
       "GLSS-19-0266-R1         0           0          0  \n",
       "GLSS-19-0266-TP         0           0          0  \n",
       "GLSS-19-0267-R1         0           0          0  \n",
       "GLSS-19-0267-TP         0           0          0  \n",
       "GLSS-19-0268-R1         0           0          0  \n",
       "\n",
       "[5 rows x 211 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_subset['gene_mut'] = variants_subset['gene_symbol'] + '_mut'\n",
    "variants_matrix = variants_subset.pivot_table(index = 'sample', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# Scoring: point mutations = 2\n",
    "variants_matrix = variants_matrix * 2\n",
    "\n",
    "variants_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c255550-1889-4b0a-b22b-c3a8fd51b6a3",
   "metadata": {},
   "source": [
    "#### CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "242fad5a-5ae5-4015-82a9-e7edbb9cbd12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_21339/4001559487.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cnvs_subset['gene_cnv'] = cnvs_subset['gene_symbol'] + '_' + cnvs_subset['cnv']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_cnv</th>\n",
       "      <th>ABL1_amp</th>\n",
       "      <th>ABL1_del</th>\n",
       "      <th>AKT1_amp</th>\n",
       "      <th>AKT1_del</th>\n",
       "      <th>AKT2_amp</th>\n",
       "      <th>AKT2_del</th>\n",
       "      <th>AKT3_amp</th>\n",
       "      <th>AKT3_del</th>\n",
       "      <th>ALK_amp</th>\n",
       "      <th>ALK_del</th>\n",
       "      <th>...</th>\n",
       "      <th>XPA_amp</th>\n",
       "      <th>XPA_del</th>\n",
       "      <th>XPC_amp</th>\n",
       "      <th>XPC_del</th>\n",
       "      <th>XPO1_amp</th>\n",
       "      <th>XPO1_del</th>\n",
       "      <th>ZNF217_amp</th>\n",
       "      <th>ZNF217_del</th>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <th>ZRSR2_del</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 475 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_cnv         ABL1_amp  ABL1_del  AKT1_amp  AKT1_del  AKT2_amp  AKT2_del  \\\n",
       "sample                                                                        \n",
       "GLSS-19-0266-R1         0         0         0         0         0         0   \n",
       "GLSS-19-0266-TP         0         0         0         0         1         0   \n",
       "GLSS-19-0267-R1         0         0         0         0         0         0   \n",
       "GLSS-19-0267-TP         0         1         0         0         0         0   \n",
       "GLSS-19-0268-TP         0         0         0         0         0         0   \n",
       "\n",
       "gene_cnv         AKT3_amp  AKT3_del  ALK_amp  ALK_del  ...  XPA_amp  XPA_del  \\\n",
       "sample                                                 ...                     \n",
       "GLSS-19-0266-R1         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0266-TP         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0267-R1         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0267-TP         0         0        0        0  ...        0        1   \n",
       "GLSS-19-0268-TP         0         0        0        0  ...        0        1   \n",
       "\n",
       "gene_cnv         XPC_amp  XPC_del  XPO1_amp  XPO1_del  ZNF217_amp  ZNF217_del  \\\n",
       "sample                                                                          \n",
       "GLSS-19-0266-R1        0        0         0         0           1           0   \n",
       "GLSS-19-0266-TP        0        0         0         0           1           0   \n",
       "GLSS-19-0267-R1        0        0         0         0           0           0   \n",
       "GLSS-19-0267-TP        0        0         0         0           0           0   \n",
       "GLSS-19-0268-TP        1        0         0         0           0           1   \n",
       "\n",
       "gene_cnv         ZRSR2_amp  ZRSR2_del  \n",
       "sample                                 \n",
       "GLSS-19-0266-R1          0          0  \n",
       "GLSS-19-0266-TP          0          0  \n",
       "GLSS-19-0267-R1          0          0  \n",
       "GLSS-19-0267-TP          0          0  \n",
       "GLSS-19-0268-TP          0          0  \n",
       "\n",
       "[5 rows x 475 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnvs_subset['gene_cnv'] = cnvs_subset['gene_symbol'] + '_' + cnvs_subset['cnv']\n",
    "cnvs_matrix = cnvs_subset.pivot_table(index = 'sample', columns = 'gene_cnv', values = 'hlvl_call_abs', aggfunc = 'sum', fill_value = 0)\n",
    "\n",
    "cnvs_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e8107a5-914f-40f8-bd38-6a61296bebc8",
   "metadata": {},
   "source": [
    "#### Combined\n",
    "Keep only variant types with 2 or more samples with an alteration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "005bb477-f614-4df7-93b4-87e22d4eaed6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(652, 365)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix = variants_matrix.join(cnvs_matrix, how = 'outer').fillna(0)\n",
    "matrix = matrix.loc[:, matrix.astype(bool).astype(int).sum() >= 2].T\n",
    "\n",
    "matrix.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8dde6a-0e03-4e3b-84c5-9f413863bf77",
   "metadata": {},
   "source": [
    "#### Write to file\n",
    "Will need to convert this to .gct format manually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "514f448e-4d43-4318-abd2-974d729f05db",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.to_csv('../data/gene_sample_matrix.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c13384b8-67c9-465d-b533-fde8c490ea39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <th>GLSS-19-0269-R1</th>\n",
       "      <th>GLSS-19-0269-TP</th>\n",
       "      <th>GLSS-19-0270-R1</th>\n",
       "      <th>GLSS-19-0270-TP</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-14-1034-TP</th>\n",
       "      <th>TCGA-14-1402-R1</th>\n",
       "      <th>TCGA-14-1402-TP</th>\n",
       "      <th>TCGA-19-0957-R1</th>\n",
       "      <th>TCGA-19-0957-TP</th>\n",
       "      <th>TCGA-19-1389-R1</th>\n",
       "      <th>TCGA-19-1389-TP</th>\n",
       "      <th>TCGA-19-4065-R1</th>\n",
       "      <th>TCGA-19-4065-TP</th>\n",
       "      <th>TCGA-FG-5963-R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT2_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT3_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALK_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPO1_del</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF217_amp</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF217_del</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRSR2_del</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>652 rows × 365 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample      GLSS-19-0266-R1  GLSS-19-0266-TP  GLSS-19-0267-R1  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              1.0              1.0              0.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      GLSS-19-0267-TP  GLSS-19-0268-R1  GLSS-19-0268-TP  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              0.0              0.0              0.0   \n",
       "ZNF217_del              0.0              0.0              1.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      GLSS-19-0269-R1  GLSS-19-0269-TP  GLSS-19-0270-R1  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              0.0              0.0              1.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      GLSS-19-0270-TP  ...  TCGA-14-1034-TP  TCGA-14-1402-R1  \\\n",
       "ABL1_mut                0.0  ...              0.0              2.0   \n",
       "AKT1_mut                0.0  ...              0.0              6.0   \n",
       "AKT2_mut                0.0  ...              0.0              0.0   \n",
       "AKT3_mut                0.0  ...              0.0              2.0   \n",
       "ALK_mut                 0.0  ...              0.0              6.0   \n",
       "...                     ...  ...              ...              ...   \n",
       "XPO1_del                0.0  ...              0.0              0.0   \n",
       "ZNF217_amp              1.0  ...              0.0              1.0   \n",
       "ZNF217_del              0.0  ...              0.0              0.0   \n",
       "ZRSR2_amp               0.0  ...              0.0              0.0   \n",
       "ZRSR2_del               0.0  ...              0.0              0.0   \n",
       "\n",
       "sample      TCGA-14-1402-TP  TCGA-19-0957-R1  TCGA-19-0957-TP  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              1.0              0.0              1.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      TCGA-19-1389-R1  TCGA-19-1389-TP  TCGA-19-4065-R1  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              0.0              1.0              0.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               2.0              1.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      TCGA-19-4065-TP  TCGA-FG-5963-R1  \n",
       "ABL1_mut                0.0              0.0  \n",
       "AKT1_mut                0.0              0.0  \n",
       "AKT2_mut                0.0              0.0  \n",
       "AKT3_mut                0.0              0.0  \n",
       "ALK_mut                 0.0              0.0  \n",
       "...                     ...              ...  \n",
       "XPO1_del                0.0              0.0  \n",
       "ZNF217_amp              0.0              0.0  \n",
       "ZNF217_del              0.0              0.0  \n",
       "ZRSR2_amp               2.0              0.0  \n",
       "ZRSR2_del               0.0              0.0  \n",
       "\n",
       "[652 rows x 365 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "516ee05d-aaee-4c61-ba27-cd1a2bb9aa02",
   "metadata": {},
   "source": [
    "Clustering on GLASS DNA, apply those to our data, assess overlap with Verhaak subtype, use those in our regressions\n",
    "- Richter's and DLBCL papers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10dd3b58-6ad9-47fa-b85c-57cba2a500e8",
   "metadata": {},
   "source": [
    "## Run clustering with external script\n",
    "\n",
    "As noted above:\n",
    "> The actual consensus clustering is done by the NMFConsensus script from GenePattern here: https://github.com/genepattern/NMFConsensus/blob/develop/nmfconsensus.R, and the clustering parameters are as listed here: gs://fc-secure-3f0f305e-f110-432b-bc9a-c8d96824cea2/jweiss/local_files/results/2024-07-22_GLASS-Verhaak-classifier/GLASS_sample_clustering.params.txt\n",
    "\n",
    "The outputs appear to be located at `os.path.join(results_dir, 'GLASS_sample_clustering_v2.membership.gct')`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c03f0ec-69bd-4c1b-a243-17fa17ef92bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Assess clustering overlap with Verhaak subtype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3897da0-3ee4-412a-b82e-651facf565b3",
   "metadata": {},
   "source": [
    "## Load clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "03204775-058e-4831-bf2d-fc6ff4baf7ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>membership</th>\n",
       "      <th>membership.1</th>\n",
       "      <th>membership.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0266.R1</th>\n",
       "      <td>GLSS.19.0266.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0266.TP</th>\n",
       "      <td>GLSS.19.0266.TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0267.R1</th>\n",
       "      <td>GLSS.19.0267.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0267.TP</th>\n",
       "      <td>GLSS.19.0267.TP</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0268.R1</th>\n",
       "      <td>GLSS.19.0268.R1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.1389.R1</th>\n",
       "      <td>TCGA.19.1389.R1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.1389.TP</th>\n",
       "      <td>TCGA.19.1389.TP</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.4065.R1</th>\n",
       "      <td>TCGA.19.4065.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.4065.TP</th>\n",
       "      <td>TCGA.19.4065.TP</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.FG.5963.R1</th>\n",
       "      <td>TCGA.FG.5963.R1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Description  membership  membership.1  membership.2\n",
       "Name                                                                    \n",
       "GLSS.19.0266.R1  GLSS.19.0266.R1           1             1             1\n",
       "GLSS.19.0266.TP  GLSS.19.0266.TP           2             2             2\n",
       "GLSS.19.0267.R1  GLSS.19.0267.R1           1             1             1\n",
       "GLSS.19.0267.TP  GLSS.19.0267.TP           3             3             3\n",
       "GLSS.19.0268.R1  GLSS.19.0268.R1           4             3             3\n",
       "...                          ...         ...           ...           ...\n",
       "TCGA.19.1389.R1  TCGA.19.1389.R1           3             3             3\n",
       "TCGA.19.1389.TP  TCGA.19.1389.TP           6             5             4\n",
       "TCGA.19.4065.R1  TCGA.19.4065.R1           1             5             4\n",
       "TCGA.19.4065.TP  TCGA.19.4065.TP           1             5             4\n",
       "TCGA.FG.5963.R1  TCGA.FG.5963.R1           3             3             3\n",
       "\n",
       "[365 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_68 = pd.read_csv(\"../data/GLASS_sample_clustering_revision.membership.gct\", sep=\"\\t\", skiprows=2, index_col=0)\n",
    "memberships_25 = pd.read_csv(\"../data/GLASS_sample_clustering_v2.membership.gct\", sep=\"\\t\", skiprows=2, index_col=0)\n",
    "memberships_68"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0736b97c-5ea9-457f-bec4-96f576083745",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_21339/4179820838.py:7: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  memberships.index = memberships.index.str.replace('.', '-')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8\n",
       "Name                                                            \n",
       "GLSS-19-0266-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0266-TP      2      2      2      2      2      2      2\n",
       "GLSS-19-0267-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0267-TP      2      2      2      3      3      3      3\n",
       "GLSS-19-0268-R1      2      3      3      4      4      3      3"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_25.rename(columns = {**{'membership': 'k = 2'}, **{f'membership.{i}': f'k = {i+2}' for i in range(1,8)}}, inplace = True)\n",
    "memberships_68.rename(columns = {**{'membership': 'k = 6'}, **{f'membership.{i}': f'k = {i+6}' for i in range(1,8)}}, inplace = True)\n",
    "\n",
    "memberships = memberships_25.join(memberships_68[[c for c in memberships_68.columns if c.startswith(\"k\")]])\n",
    "\n",
    "memberships.drop(columns = 'Description', inplace = True)\n",
    "memberships.index = memberships.index.str.replace('.', '-')\n",
    "memberships.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc7c6fc-3237-410b-93b3-8c3160cc2205",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load ssGSEA\n",
    "(Verhaak subtype scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a6ed5f47-3e3a-49d3-be5f-9b4425301066",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(425, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_CLASSICAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_NEURAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_PRONEURAL</th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "      <th>surgery_number</th>\n",
       "      <th>surgical_interval_mo</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>4499.507042</td>\n",
       "      <td>3725.238591</td>\n",
       "      <td>1945.457473</td>\n",
       "      <td>2028.442349</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>3866.366259</td>\n",
       "      <td>3899.952531</td>\n",
       "      <td>3055.641397</td>\n",
       "      <td>1970.036575</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</td>\n",
       "      <td>2</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3998.177157</td>\n",
       "      <td>2373.110794</td>\n",
       "      <td>2792.203914</td>\n",
       "      <td>3845.965090</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3403.500159</td>\n",
       "      <td>1489.313078</td>\n",
       "      <td>4904.914629</td>\n",
       "      <td>3993.835349</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_NEURAL</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <td>GLSS-19-0268</td>\n",
       "      <td>4367.052085</td>\n",
       "      <td>2494.211618</td>\n",
       "      <td>813.639386</td>\n",
       "      <td>3709.522026</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 case_barcode  VERHAAK_GLIOBLASTOMA_CLASSICAL  \\\n",
       "sample_barcode                                                  \n",
       "GLSS-19-0266-TP  GLSS-19-0266                     4499.507042   \n",
       "GLSS-19-0266-R1  GLSS-19-0266                     3866.366259   \n",
       "GLSS-19-0267-TP  GLSS-19-0267                     3998.177157   \n",
       "GLSS-19-0267-R1  GLSS-19-0267                     3403.500159   \n",
       "GLSS-19-0268-TP  GLSS-19-0268                     4367.052085   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_MESENCHYMAL  \\\n",
       "sample_barcode                                      \n",
       "GLSS-19-0266-TP                       3725.238591   \n",
       "GLSS-19-0266-R1                       3899.952531   \n",
       "GLSS-19-0267-TP                       2373.110794   \n",
       "GLSS-19-0267-R1                       1489.313078   \n",
       "GLSS-19-0268-TP                       2494.211618   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_NEURAL  VERHAAK_GLIOBLASTOMA_PRONEURAL  \\\n",
       "sample_barcode                                                                 \n",
       "GLSS-19-0266-TP                  1945.457473                     2028.442349   \n",
       "GLSS-19-0266-R1                  3055.641397                     1970.036575   \n",
       "GLSS-19-0267-TP                  2792.203914                     3845.965090   \n",
       "GLSS-19-0267-R1                  4904.914629                     3993.835349   \n",
       "GLSS-19-0268-TP                   813.639386                     3709.522026   \n",
       "\n",
       "                                  Verhaak_subtype  surgery_number  \\\n",
       "sample_barcode                                                      \n",
       "GLSS-19-0266-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "GLSS-19-0266-R1  VERHAAK_GLIOBLASTOMA_MESENCHYMAL               2   \n",
       "GLSS-19-0267-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "GLSS-19-0267-R1       VERHAAK_GLIOBLASTOMA_NEURAL               2   \n",
       "GLSS-19-0268-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "\n",
       "                 surgical_interval_mo  \n",
       "sample_barcode                         \n",
       "GLSS-19-0266-TP                   0.0  \n",
       "GLSS-19-0266-R1                  18.0  \n",
       "GLSS-19-0267-TP                   0.0  \n",
       "GLSS-19-0267-R1                  10.0  \n",
       "GLSS-19-0268-TP                   0.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea = pd.read_csv(os.path.join(\"..\", \"data\", 'glass_verhaak_scores.tsv'), sep = '\\t', index_col = 1)\n",
    "print(ssgsea.shape)\n",
    "ssgsea.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e2be0ae-3832-49fb-a5db-5c4cff83a80a",
   "metadata": {},
   "source": [
    "Combine duplicated sample barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "11bb92a4-3e47-4fd0-8eb2-5cef6fe8c58a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(419, 6)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea = ssgsea.groupby('sample_barcode').agg({'case_barcode': 'min',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_CLASSICAL' : 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_MESENCHYMAL': 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_PRONEURAL' : 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_NEURAL' : 'mean'})\n",
    "\n",
    "ssgsea['Verhaak_subtype'] = ssgsea.iloc[:, 1:].idxmax(axis=1)\n",
    "ssgsea.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "380dcc00-8286-45a5-a143-1da7b9cd0e0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_CLASSICAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_PRONEURAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_NEURAL</th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>3866.366259</td>\n",
       "      <td>3899.952531</td>\n",
       "      <td>1970.036575</td>\n",
       "      <td>3055.641397</td>\n",
       "      <td>MESENCHYMAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>4499.507042</td>\n",
       "      <td>3725.238591</td>\n",
       "      <td>2028.442349</td>\n",
       "      <td>1945.457473</td>\n",
       "      <td>CLASSICAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3403.500159</td>\n",
       "      <td>1489.313078</td>\n",
       "      <td>3993.835349</td>\n",
       "      <td>4904.914629</td>\n",
       "      <td>NEURAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3998.177157</td>\n",
       "      <td>2373.110794</td>\n",
       "      <td>3845.965090</td>\n",
       "      <td>2792.203914</td>\n",
       "      <td>CLASSICAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>GLSS-19-0268</td>\n",
       "      <td>3271.104740</td>\n",
       "      <td>3855.965091</td>\n",
       "      <td>2225.823185</td>\n",
       "      <td>3986.308979</td>\n",
       "      <td>NEURAL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 case_barcode  VERHAAK_GLIOBLASTOMA_CLASSICAL  \\\n",
       "sample_barcode                                                  \n",
       "GLSS-19-0266-R1  GLSS-19-0266                     3866.366259   \n",
       "GLSS-19-0266-TP  GLSS-19-0266                     4499.507042   \n",
       "GLSS-19-0267-R1  GLSS-19-0267                     3403.500159   \n",
       "GLSS-19-0267-TP  GLSS-19-0267                     3998.177157   \n",
       "GLSS-19-0268-R1  GLSS-19-0268                     3271.104740   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_MESENCHYMAL  \\\n",
       "sample_barcode                                      \n",
       "GLSS-19-0266-R1                       3899.952531   \n",
       "GLSS-19-0266-TP                       3725.238591   \n",
       "GLSS-19-0267-R1                       1489.313078   \n",
       "GLSS-19-0267-TP                       2373.110794   \n",
       "GLSS-19-0268-R1                       3855.965091   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_PRONEURAL  VERHAAK_GLIOBLASTOMA_NEURAL  \\\n",
       "sample_barcode                                                                 \n",
       "GLSS-19-0266-R1                     1970.036575                  3055.641397   \n",
       "GLSS-19-0266-TP                     2028.442349                  1945.457473   \n",
       "GLSS-19-0267-R1                     3993.835349                  4904.914629   \n",
       "GLSS-19-0267-TP                     3845.965090                  2792.203914   \n",
       "GLSS-19-0268-R1                     2225.823185                  3986.308979   \n",
       "\n",
       "                Verhaak_subtype  \n",
       "sample_barcode                   \n",
       "GLSS-19-0266-R1     MESENCHYMAL  \n",
       "GLSS-19-0266-TP       CLASSICAL  \n",
       "GLSS-19-0267-R1          NEURAL  \n",
       "GLSS-19-0267-TP       CLASSICAL  \n",
       "GLSS-19-0268-R1          NEURAL  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea['Verhaak_subtype'] = ssgsea['Verhaak_subtype'].apply(lambda x: x.split('_')[-1])\n",
    "ssgsea.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e84501f-0905-4451-9220-2093fb09ad68",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b24d5c90-508e-4eb9-8c35-29948f196b65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "365 419 246\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CLASSICAL      80\n",
       "MESENCHYMAL    79\n",
       "NEURAL         58\n",
       "PRONEURAL      29\n",
       "Name: Verhaak_subtype, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap = ssgsea.join(memberships, how = 'inner').reset_index()\n",
    "# overlap = overlap[overlap['Verhaak_subtype'].isin(['PRONEURAL', 'MESENCHYMAL'])]\n",
    "print(len(memberships), len(ssgsea), len(overlap))\n",
    "overlap['Verhaak_subtype'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b577a3-1f1d-44cc-9950-1d7d892391e8",
   "metadata": {},
   "source": [
    "## Calculate clustering overlap metrics\n",
    "- Adjusted Rand Index\n",
    "- Adjusted Mutual Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6eb1a6c3-f311-4697-a481-b2a4f81249b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x44a8948b0>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAG2CAYAAABcYt1RAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAA9hAAAPYQGoP6dpAABgkklEQVR4nO3deZzN9eLH8deZHbMxmM0Yg8IYSzNKY2sTUpauorpI1L2ksrQgibRIpdSvcCtLboVupSTJErJMyZplLNlGzGAsM7bZzjm/P77mcMwMM4M53xnv5+NxHuZ8zuf7nc/3GM57PtvXYrfb7YiIiIiYmJurGyAiIiJyOQosIiIiYnoKLCIiImJ6CiwiIiJiegosIiIiYnoKLCIiImJ6CiwiIiJiegosIiIiYnoKLCIiImJ6CiwiIiJiesUKLBMmTCAqKgofHx/i4uJYvnx5gXWTk5N55JFHqFOnDm5ubgwcODDfet988w3R0dF4e3sTHR3N7Nmzi9M0ERERKYOKHFhmzZrFwIEDGT58OOvXr6dly5bcc889JCUl5Vs/MzOTKlWqMHz4cBo1apRvnYSEBLp160aPHj3YuHEjPXr0oGvXrvz+++9FbZ6IiIiUQZai3vywadOmxMbGMnHiREdZvXr16Ny5M2PGjLnksbfffjuNGzdm/PjxTuXdunUjPT2dn376yVHWrl07KlasyIwZM4rSPBERESmDPIpSOSsri7Vr1zJ06FCn8jZt2rBq1apiNyIhIYFBgwY5lbVt2zZPsLlQZmYmmZmZjuc2m41jx44RFBSExWIpdltERESk5Njtdk6ePElYWBhubgUP/BQpsKSmpmK1WgkODnYqDw4OJiUlpXgtBVJSUop8zjFjxvDKK68U+3uKiIiIeezfv59q1aoV+HqRAkuui3sw7Hb7FfdqFPWcw4YNY/DgwY7naWlpVK9enf379+Pv739FbREREZGSkZ6eTkREBH5+fpesV6TAUrlyZdzd3fP0fBw+fDhPD0lRhISEFPmc3t7eeHt75yn39/dXYBERESllLtfxUaRVQl5eXsTFxbFw4UKn8oULF9KsWbOit+6c+Pj4POdcsGDBFZ1TREREyo4iDwkNHjyYHj160KRJE+Lj4/n4449JSkqib9++gDFUc+DAAaZPn+44ZsOGDQCcOnWKI0eOsGHDBry8vIiOjgZgwIABtGrVirFjx9KpUye+//57Fi1axIoVK67CJYqIiEhpV+TA0q1bN44ePcro0aNJTk4mJiaGefPmERkZCRgbxV28J8tNN93k+Hrt2rV8+eWXREZGsnfvXgCaNWvGzJkzeemllxgxYgS1atVi1qxZNG3a9AouTURERMqKIu/DYlbp6ekEBASQlpamOSwipZzVaiU7O9vVzRCRq8DT0xN3d/cCXy/s53exVgmJiFwLdrudlJQUTpw44eqmiMhVFBgYSEhIyBWtKFZgERHTyA0rVatWpXz58toEUqSUs9vtnDlzhsOHDwMQGhpa7HMpsIiIKVitVkdYCQoKcnVzROQqKVeuHGBsV1K1atVLDg9dSrHu1iwicrXlzlkpX768i1siIldb7r/rK5mbpsAiIqaiYSCRsudq/LtWYBERERHTU2AREXGRUaNG0bhx4wKfm8ntt9/OwIEDXd2MAvXq1YvOnTuXme8jeSmwiEiZY7XZSdh1lO83HCBh11GstpLZbmrVqlW4u7vTrl27Yh3/3HPPsXjx4qvWnpIMGdOmTcNisTgewcHBdOjQgS1btpTI97+cpUuXYrFYtGS+FNMqIREpU+ZvTuaVH7aSnJbhKAsN8GFkh2jaxRR/SWVhTJkyhaeffppPP/2UpKQkqlevXqTjfX198fX1vUatu/b8/f3Zvn07drudAwcO8MILL3DvvfeyY8cOvLy8XN08KeXUwyIiZcb8zcn0+3ydU1gBSEnLoN/n65i/Ofmafe/Tp0/z1Vdf0a9fP+677z6mTZuWp86bb75JcHAwfn5+9OnTh4wM53ZePCSUXw9J586d6dWrl+P5hAkTuOGGG/Dx8SE4OJgHHngAMIYuli1bxvvvv+/o9ci9HcrWrVtp3749vr6+BAcH06NHD1JTU52upWfPnvj6+hIaGsq4ceMK9R5YLBZCQkIIDQ2lSZMmDBo0iH379rF9+3ZHnXfffZcGDRpQoUIFIiIiePLJJzl16pTj9WnTphEYGMjPP/9MvXr18PX1pV27diQnn/+7s1qtDB48mMDAQIKCgnjhhRco6qbtV+v72O123nrrLWrWrEm5cuVo1KgRX3/9teO11q1b065dO8dxJ06coHr16gwfPrxI7RUFFhEpBc5k5RT4yMi2AsYw0Cs/bCW/j63cslE/bHUaHsrvfMU1a9Ys6tSpQ506dejevTtTp051+nD76quvGDlyJK+//jpr1qwhNDSUCRMmFPv7AaxZs4ZnnnmG0aNHs337dubPn0+rVq0AeP/994mPj+eJJ54gOTmZ5ORkIiIiSE5O5rbbbqNx48asWbOG+fPnc+jQIbp27eo47/PPP8+SJUuYPXs2CxYsYOnSpaxdu7ZIbTtx4gRffvklYGzNnsvNzY0PPviAzZs389lnn/HLL7/wwgsvOB175swZ3nnnHf773//y66+/kpSUxHPPPed4fdy4cUyZMoXJkyezYsUKjh07xuzZs4v8/l2N7/PSSy8xdepUJk6cyJYtWxg0aBDdu3dn2bJlWCwWPvvsM1avXs0HH3wAQN++fQkODmbUqFFFbu/1TkNCImJ60S//XOBrd9SpwtTHbmH1nmN5elYuZMfoaVm95xjxtYyN6VqMXcKx01lO9fa+eW+x2jh58mS6d+8OQLt27Th16hSLFy+mdevWAIwfP57evXvz+OOPA/Daa6+xaNGiPL0sRZGUlESFChW477778PPzIzIy0nGz2YCAALy8vChfvjwhISGOYyZOnEhsbCxvvPGGo2zKlClERESwY8cOwsLCmDx5MtOnT+fuu+8G4LPPPqNatWqXbU9aWhq+vr6O3U0BOnbsSN26dR11LuwxioqK4tVXX6Vfv35O4S07O5tJkyZRq1YtAJ566ilGjx7teH38+PEMGzaMLl26ADBp0iR+/rngn5GCXOn3OX36NO+++y6//PIL8fHxANSsWZMVK1bwn//8h9tuu43w8HD+85//0KNHDw4dOsQPP/zA+vXrnUKcFI4Ci4iUCYdPFu6Dv7D1imL79u2sXr2ab7/9FgAPDw+6devGlClTHIElMTGRvn37Oh0XHx/PkiVLiv197777biIjI6lZsybt2rWjXbt23H///ZfcfG/t2rUsWbIk37kyu3bt4uzZs2RlZTk+gAEqVapEnTp1LtsePz8/1q1bR05ODsuWLePtt99m0qRJTnWWLFnCG2+8wdatW0lPTycnJ4eMjAxOnz5NhQoVAGOTsdwQAcZ27rlbu6elpZGcnOzUPg8PD5o0aVLkYaEr/T5bt24lIyPDEexyZWVlOYIjwIMPPsjs2bMZM2YMEydO5MYbbyxSO8WgwCIiprd1dNsCX3M7tyFVVT+fQp3rwnorhtxxZQ07Z/LkyeTk5BAeHu4os9vteHp6cvz4cSpWrFis87q5ueX5EL5wp9DcgLB06VIWLFjAyy+/zKhRo/jjjz8IDAzM95w2m40OHTowduzYPK+Fhoayc+fOYrU1t721a9cGoG7duqSkpNCtWzd+/fVXAPbt20f79u3p27cvr776KpUqVWLFihX06dPH6bou7n2wWCxFDiOFcaXfx2azAfDjjz86/d0DeHt7O74+c+YMa9euxd3d/Yre3+ud5rCIiOmV9/Io8OHjadyX5JaoSoQG+FDQfpoWjNVCt0RVuuR5iyonJ4fp06czbtw4NmzY4Hhs3LiRyMhIvvjiCwDq1avHb7/95nTsxc8vVqVKlTyTQDdv3uxUx8PDg9atW/PWW2/x559/snfvXn755RcAvLy8sFqtTvVjY2PZsmULNWrUoHbt2k6PChUqULt2bTw9PZ3advz4cXbs2FHk92bQoEFs3LjRMe9jzZo15OTkMG7cOG699VZuvPFGDh48WKRzBgQEEBoa6tS+nJycIs+xuRrfJzo6Gm9vb5KSkvK8lxEREY56zz77LG5ubvz000988MEHjr8fKRr1sIhImeDuZmFkh2j6fb4OCzhNvs0NMSM7ROPudnW3/p87dy7Hjx+nT58+BAQEOL32wAMPMHnyZJ566ikGDBjAo48+SpMmTWjRogVffPEFW7ZsoWbNmgWe+84772Tw4MH8+OOP1KpVi/fee89pH5G5c+eye/duWrVqRcWKFZk3bx42m80xfFOjRg1+//139u7di6+vL5UqVaJ///588sknPPzwwzz//PNUrlyZv/76i5kzZ/LJJ5/g6+tLnz59eP755wkKCiI4OJjhw4fj5lb032/9/f15/PHHGTlyJJ07d6ZWrVrk5OTwf//3f3To0IGVK1fmGTIqjAEDBvDmm29yww03UK9ePd59991rsr/K5b6Pn58fzz33HIMGDcJms9GiRQvS09NZtWoVvr6+PProo/z4449MmTKFhIQEYmNjGTp0KI8++ih//vlnsXverlfqYRGRMqNdTCgTu8cSEuA8PBQS4MPE7rHXZB+WyZMn07p16zxhBaBLly5s2LCBdevW0a1bN15++WWGDBlCXFwc+/bto1+/fpc8d+/evXn00Ufp2bMnt912G1FRUdxxx/lhrMDAQL799lvuvPNO6tWrx6RJk5gxYwb169cHjI3o3N3diY6OpkqVKiQlJREWFsbKlSuxWq20bduWmJgYBgwYQEBAgCOUvP3227Rq1YqOHTvSunVrWrRoQVxcXLHenwEDBpCYmMj//vc/GjduzLvvvsvYsWOJiYnhiy++YMyYMUU+57PPPkvPnj3p1asX8fHx+Pn5cf/99xerfVf6fV599VVefvllxowZQ7169Wjbti0//PADUVFRHDlyhD59+jBq1ChiY2MBGDlyJGFhYXnmM8nlWezXYmDQBdLT0wkICCAtLQ1/f39XN0dEiigjI4M9e/YQFRWFj0/h5qMUxGqzs3rPMQ6fzKCqnzEMdLV7Vq6FYcOGsXz5clasWOHqpohcVZf6913Yz28NCYlImePuZnEsXS4N7HY7u3fvZvHixU6rS0TkPA0JiYi4WFpaGtHR0Xh5efHiiy+6ujkipqQeFhERFwsMDCQzM9PVzRAxNfWwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIi4yKhRo2jcuHGBz83k9ttvZ+DAga5uxlVTo0YNxo8ff1XPeebMGbp06YK/vz8Wi+Wa3N/oWli6dGmpaK8Ci4iUPTYr7FkOm742/rRZL3/MVbBq1Src3d1p165dsY5/7rnnWLx48VVrT0mGjGnTpmGxWKhXr16e17766issFgs1atQo0jldGZKKEx4/++wzli9fzqpVq0hOTs73/lKult972qxZM9O290LaOE5Eypatc2D+EEg/eL7MPwzajYXojtf0W0+ZMoWnn36aTz/9lKSkJKpXr16k4319ffH19b1Grbv2KlSowOHDh0lISCA+Pt5RPmXKlCK/F6XRrl27qFevHjExMcU+h9VqxWKxFOvu2MXl5eVFSEhIiX2/4lIPi4iUHVvnwFc9ncMKQHqyUb51zjX71qdPn+arr76iX79+3HfffUybNi1PnTfffJPg4GD8/Pzo06cPGRkZTq9f/Ft9fr8Nd+7cmV69ejmeT5gwgRtuuAEfHx+Cg4N54IEHAOjVqxfLli3j/fffx2KxYLFY2Lt3LwBbt26lffv2+Pr6EhwcTI8ePUhNTXW6lp49e+Lr60toaCjjxo0r1Hvg4eHBI488wpQpUxxlf//9N0uXLuWRRx5xqturVy86d+7sVDZw4EBuv/32S7Z/2rRpBAYGOh333XffYbGcv7nlrl276NSpE8HBwfj6+nLzzTezaNGiQl1DQXLb+8477xAaGkpQUBD9+/cnOzsbMP6uxo0bx6+//orFYnFcx/Hjx+nZsycVK1akfPny3HPPPezcudNx3tzrmTt3LtHR0Xh7e7Nv3z5q1KjBa6+95vh7iIyM5Pvvv+fIkSN06tQJX19fGjRowJo1axznOnr0KA8//DDVqlWjfPnyNGjQgBkzZjhdQ37vaX5DQt988w3169fH29ubGjVq5PkZqFGjBm+88Qa9e/fGz8+P6tWr8/HHH1/Re3w5CiwiYn5Zpwt+ZJ/70LdZjZ4V8rsB/bmy+UOch4fyO18xzZo1izp16lCnTh26d+/O1KlTsdvPt+Wrr75i5MiRvP7666xZs4bQ0FAmTJhQ7O8HsGbNGp555hlGjx7N9u3bmT9/Pq1atQLg/fffJz4+nieeeILk5GSSk5OJiIggOTmZ2267jcaNG7NmzRrmz5/PoUOH6Nq1q+O8zz//PEuWLGH27NksWLCApUuXsnbt2kK1qU+fPsyaNYszZ84Axgdyu3btCA4OLtK1FdT+wjh16hTt27dn0aJFrF+/nrZt29KhQweSkpKK1IaLLVmyhF27drFkyRI+++wzpk2b5gim3377LU888QTx8fEkJyfz7bffAkZIWLNmDXPmzCEhIQG73U779u0dQQeMuS9jxozh008/ZcuWLVStWhWA9957j+bNm7N+/XruvfdeevToQc+ePenevTvr1q2jdu3a9OzZ0/FzlpGRQVxcHHPnzmXz5s3861//okePHvz+++9Fek/Xrl1L165deeihh9i0aROjRo1ixIgReUL4uHHjaNKkCevXr+fJJ5+kX79+bNu27Yre40vRkJCImN8bYQW/dkMb+Of/YN+qvD0rTuzG6/tWQVRLo2h8Azhz1LnaqLRiNXHy5Ml0794dgHbt2nHq1CkWL15M69atjW81fjy9e/fm8ccfB+C1115j0aJFeXpZiiIpKYkKFSpw33334efnR2RkpONuzwEBAXh5eVG+fHmn7v6JEycSGxvLG2+84SibMmUKERER7Nixg7CwMCZPnsz06dO5++67AWNuRrVq1QrVpsaNG1OrVi2+/vprevTowbRp03j33XfZvXt3ka6toPYXRqNGjWjUqJHj+Wuvvcbs2bOZM2cOTz31VJHOdaGKFSvy4Ycf4u7uTt26dbn33ntZvHgxTzzxBJUqVaJ8+fJOwys7d+5kzpw5rFy5kmbNmgHwxRdfEBERwXfffceDDz4IQHZ2NhMmTHBqM0D79u3597//DcDLL7/MxIkTufnmmx3HDRkyhPj4eA4dOkRISAjh4eE899xzjuOffvpp5s+fz//+9z+aNm1a6Pf03Xff5a677mLEiBEA3HjjjWzdupW3337bqXevffv2PPnkk462vPfeeyxdupS6desW+z2+FPWwiEjZcOrQ1a1XBNu3b2f16tU89NBDgDE00q1bN6ehkcTERKd5HUCe50V19913ExkZSc2aNenRowdffPGFo2ejIGvXrmXJkiWO+TK+vr6OD5hdu3axa9cusrKynNpWqVIl6tSpU+h29e7dm6lTp7Js2TJHb0dJOn36NC+88ALR0dEEBgbi6+vLtm3brriHpX79+ri7uzueh4aGcvjw4QLrJyYm4uHhQdOmTR1lQUFB1KlTh8TEREeZl5cXDRs2zHP8hWW5PVQNGjTIU5bbBqvVyuuvv07Dhg0JCgrC19eXBQsWFPm6ExMTad68uVNZ8+bN2blzJ1br+R7KC9tnsVgICQm55PtxpdTDIiLm9+Ilek4s5z5AfAs55HBhvYGbit+mC0yePJmcnBzCw8MdZXa7HU9PT44fP07FihWLdV43NzenYSXAaSjBz8+PdevWsXTpUhYsWMDLL7/MqFGj+OOPP/LM88hls9no0KEDY8eOzfNaaGio0/yK4vrnP//JCy+8wKhRo+jZsyceHnk/ai53bQUpzHHPP/88P//8M++88w61a9emXLlyPPDAA2RlZRXjas7z9PR0em6xWLDZbAXWv7idF5ZfOOemXLlyTs/z+365r+dXltuGcePG8d577zF+/HgaNGhAhQoVGDhwYJGv++L2FXQtRX0/rpR6WETE/LwqFPzw9DHqRDYzVgOR9z9+gwX8w416lzpvEeXk5DB9+nTGjRvHhg0bHI+NGzcSGRnJF198AUC9evX47bffnI69+PnFqlSpQnJysuO51Wpl8+bNTnU8PDxo3bo1b731Fn/++Sd79+7ll19+MS7Py8vpN2KA2NhYtmzZQo0aNahdu7bTo0KFCtSuXRtPT0+nth0/fpwdO3YU+j2pVKkSHTt2ZNmyZfTu3btQ1wawYcMGp+f5tb9KlSqcPHmS06dPF3jc8uXL6dWrF/fffz8NGjQgJCTEMeG4JEVHR5OTk+OYQwLGxNgdO3bku/z7Si1fvpxOnTrRvXt3GjVqRM2aNfME0Pze0/zavWLFCqeyVatWceONNzr1MJU0BRYRKRvc3I2ly0De0HLuebs3jXpX0dy5czl+/Dh9+vQhJibG6fHAAw8wefJkAAYMGMCUKVOYMmUKO3bsYOTIkWzZsuWS577zzjv58ccf+fHHH9m2bRtPPvmk00qOuXPn8sEHH7Bhwwb27dvH9OnTsdlsjuGbGjVq8Pvvv7N3715SU1Ox2Wz079+fY8eO8fDDD7N69Wp2797NggUL6N27N1arFV9fX/r06cPzzz/P4sWL2bx5M7169SryMttp06aRmppa4HyGO++8kzVr1jB9+nR27tzJyJEj84Sx/NrftGlTypcvz4svvshff/3Fl19+mWcyaO3atfn2228dwfGRRx65pr/5F+SGG26gU6dOPPHEE6xYsYKNGzfSvXt3wsPD6dSp01X/frVr12bhwoWsWrWKxMRE/v3vf5OSkuJUJ7/39GLPPvssixcv5tVXX2XHjh189tlnfPjhh07zY1xBgUVEyo7ojtB1OviHOpf7hxnl12AflsmTJ9O6det8N93q0qULGzZsYN26dXTr1o2XX36ZIUOGEBcXx759++jXr98lz927d28effRRevbsyW233UZUVBR33HGH4/XAwEC+/fZb7rzzTurVq8ekSZOYMWMG9evXB4yN6Nzd3YmOjqZKlSokJSURFhbGypUrsVqttG3blpiYGAYMGEBAQIAjlLz99tu0atWKjh070rp1a1q0aEFcXFyR3pdy5coRFBRU4Ott27ZlxIgRvPDCC9x8882cPHmSnj17OtXJr/2VKlXi888/Z968eY5lu6NGjXI67r333qNixYo0a9aMDh060LZtW2JjY4vU/qtl6tSpxMXFcd999xEfH4/dbmfevHl5hlOuhhEjRhAbG0vbtm25/fbbCQkJybN0PL/39GKxsbF89dVXzJw5k5iYGF5++WVGjx7tNOHWFSz2ggbZSpn09HQCAgJIS0vD39/f1c0RkSLKyMhgz549REVF4ePjc2Uns1mN1UCnDhlzViKbXfWelWth2LBhLF++PE93vEhpd6l/34X9/NakWxEpe9zczy9dLgXsdju7d+9m8eLFjmXJIuJMQ0IiIi6WlpZGdHQ0Xl5evPjii65ujogpqYdFRMTFAgMDyczMdHUzRExNPSwiIiJiegosImIqZWQdgIhc4Gr8u1ZgERFTyF3mebmt5UWk9Mn9d30ly7k1h0VETMHd3Z3AwEDHvUjKly+f73blIlJ62O12zpw5w+HDhwkMDLyinXIVWETENHLvIHstb6AmIiUvMDCwyHfdvpgCi4iYhsViITQ0lKpVqxbqRngiYn6enp5X5R5ECiwiYjru7u4uvcmaiJiPJt2KiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkVK7BMmDCBqKgofHx8iIuLY/ny5Zesv2zZMuLi4vDx8aFmzZpMmjQpT53x48dTp04dypUrR0REBIMGDSIjI6M4zRMREZEypsiBZdasWQwcOJDhw4ezfv16WrZsyT333ENSUlK+9ffs2UP79u1p2bIl69ev58UXX+SZZ57hm2++cdT54osvGDp0KCNHjiQxMZHJkycza9Yshg0bVvwrExERkTLDYrfb7UU5oGnTpsTGxjJx4kRHWb169ejcuTNjxozJU3/IkCHMmTOHxMRER1nfvn3ZuHEjCQkJADz11FMkJiayePFiR51nn32W1atXX7b3Jld6ejoBAQGkpaXh7+9flEsSERERFyns53eReliysrJYu3Ytbdq0cSpv06YNq1atyveYhISEPPXbtm3LmjVryM7OBqBFixasXbuW1atXA7B7927mzZvHvffeW2BbMjMzSU9Pd3qIiIhI2eRRlMqpqalYrVaCg4OdyoODg0lJScn3mJSUlHzr5+TkkJqaSmhoKA899BBHjhyhRYsW2O12cnJy6NevH0OHDi2wLWPGjOGVV14pSvNFRESklCrWpFuLxeL03G635ym7XP0Ly5cuXcrrr7/OhAkTWLduHd9++y1z587l1VdfLfCcw4YNIy0tzfHYv39/cS5FRERESoEi9bBUrlwZd3f3PL0phw8fztOLkiskJCTf+h4eHgQFBQEwYsQIevToweOPPw5AgwYNOH36NP/6178YPnw4bm55c5W3tzfe3t5Fab6IiIiUUkXqYfHy8iIuLo6FCxc6lS9cuJBmzZrle0x8fHye+gsWLKBJkyZ4enoCcObMmTyhxN3dHbvdThHnBIuIiEgZVOQhocGDB/Ppp58yZcoUEhMTGTRoEElJSfTt2xcwhmp69uzpqN+3b1/27dvH4MGDSUxMZMqUKUyePJnnnnvOUadDhw5MnDiRmTNnsmfPHhYuXMiIESPo2LEj7u7uV+EyRUREpDQr0pAQQLdu3Th69CijR48mOTmZmJgY5s2bR2RkJADJyclOe7JERUUxb948Bg0axEcffURYWBgffPABXbp0cdR56aWXsFgsvPTSSxw4cIAqVarQoUMHXn/99atwiSIiIlLaFXkfFrPSPiwiIiKlzzXZh0VERETEFRRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0ihVYJkyYQFRUFD4+PsTFxbF8+fJL1l+2bBlxcXH4+PhQs2ZNJk2alKfOiRMn6N+/P6Ghofj4+FCvXj3mzZtXnOaJiIhIGVPkwDJr1iwGDhzI8OHDWb9+PS1btuSee+4hKSkp3/p79uyhffv2tGzZkvXr1/Piiy/yzDPP8M033zjqZGVlcffdd7N3716+/vprtm/fzieffEJ4eHjxr0xERETKDIvdbrcX5YCmTZsSGxvLxIkTHWX16tWjc+fOjBkzJk/9IUOGMGfOHBITEx1lffv2ZePGjSQkJAAwadIk3n77bbZt24anp2exLiQ9PZ2AgADS0tLw9/cv1jlERESkZBX287tIPSxZWVmsXbuWNm3aOJW3adOGVatW5XtMQkJCnvpt27ZlzZo1ZGdnAzBnzhzi4+Pp378/wcHBxMTE8MYbb2C1WgtsS2ZmJunp6U4PERERKZuKFFhSU1OxWq0EBwc7lQcHB5OSkpLvMSkpKfnWz8nJITU1FYDdu3fz9ddfY7VamTdvHi+99BLjxo3j9ddfL7AtY8aMISAgwPGIiIgoyqWIiIhIKVKsSbcWi8Xpud1uz1N2ufoXlttsNqpWrcrHH39MXFwcDz30EMOHD3cadrrYsGHDSEtLczz2799fnEsRERGRUsCjKJUrV66Mu7t7nt6Uw4cP5+lFyRUSEpJvfQ8PD4KCggAIDQ3F09MTd3d3R5169eqRkpJCVlYWXl5eec7r7e2Nt7d3UZovIiIipVSReli8vLyIi4tj4cKFTuULFy6kWbNm+R4THx+fp/6CBQto0qSJY4Jt8+bN+euvv7DZbI46O3bsIDQ0NN+wIiIiIteXIg8JDR48mE8//ZQpU6aQmJjIoEGDSEpKom/fvoAxVNOzZ09H/b59+7Jv3z4GDx5MYmIiU6ZMYfLkyTz33HOOOv369ePo0aMMGDCAHTt28OOPP/LGG2/Qv3//q3CJIiIiUtoVaUgIoFu3bhw9epTRo0eTnJxMTEwM8+bNIzIyEoDk5GSnPVmioqKYN28egwYN4qOPPiIsLIwPPviALl26OOpERESwYMECBg0aRMOGDQkPD2fAgAEMGTLkKlyiiIiIlHZF3ofFrLQPi4iISOlzTfZhEREREXEFBRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMb1iBZYJEyYQFRWFj48PcXFxLF++/JL1ly1bRlxcHD4+PtSsWZNJkyYVWHfmzJlYLBY6d+5cnKaJiIhIGVTkwDJr1iwGDhzI8OHDWb9+PS1btuSee+4hKSkp3/p79uyhffv2tGzZkvXr1/Piiy/yzDPP8M033+Spu2/fPp577jlatmxZ9CsRERGRMstit9vtRTmgadOmxMbGMnHiREdZvXr16Ny5M2PGjMlTf8iQIcyZM4fExERHWd++fdm4cSMJCQmOMqvVym233cZjjz3G8uXLOXHiBN99912h25Wenk5AQABpaWn4+/sX5ZJERETERQr7+V2kHpasrCzWrl1LmzZtnMrbtGnDqlWr8j0mISEhT/22bduyZs0asrOzHWWjR4+mSpUq9OnTp1BtyczMJD093ekhIiIiZVORAktqaipWq5Xg4GCn8uDgYFJSUvI9JiUlJd/6OTk5pKamArBy5UomT57MJ598Uui2jBkzhoCAAMcjIiKiKJciIiIipUixJt1aLBan53a7PU/Z5ernlp88eZLu3bvzySefULly5UK3YdiwYaSlpTke+/fvL8IViIiISGniUZTKlStXxt3dPU9vyuHDh/P0ouQKCQnJt76HhwdBQUFs2bKFvXv30qFDB8frNpvNaJyHB9u3b6dWrVp5zuvt7Y23t3dRmi8iIiKlVJF6WLy8vIiLi2PhwoVO5QsXLqRZs2b5HhMfH5+n/oIFC2jSpAmenp7UrVuXTZs2sWHDBsejY8eO3HHHHWzYsEFDPSIiIlK0HhaAwYMH06NHD5o0aUJ8fDwff/wxSUlJ9O3bFzCGag4cOMD06dMBY0XQhx9+yODBg3niiSdISEhg8uTJzJgxAwAfHx9iYmKcvkdgYCBAnnIRERG5PhU5sHTr1o2jR48yevRokpOTiYmJYd68eURGRgKQnJzstCdLVFQU8+bNY9CgQXz00UeEhYXxwQcf0KVLl6t3FSIiIlKmFXkfFrPSPiwiIiKlzzXZh0VERETEFRRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQ8XN0AERGRkma12Vm95xiHT2ZQ1c+HW6Iq4e5mcXWz5BIUWERE5Loyf3Myr/ywleS0DEdZaIAPIztE0y4m1IUtk0vRkJCIiFw35m9Opt/n65zCCkBKWgb9Pl/H/M3JLmqZXI4Ci4iIXBesNjuv/LAVez6v5Za98sNWrLb8aoirKbCIiMh1YfWeY3l6Vi5kB5LTMli951jJNUoKTYFFRETKvGyrjZ8KOdxz+GTBoUZcR5NuRUSkzOs1dTUr/zpaqLqB5TyvcWukONTDIiIiZc6e1NNkZFsdz9vFhBJUwRNf74J/T7cAnu4WXv5+M7/tLly4kZKjwCIiImWC3W5n1a5UHv/sD+4ct5Tv1h9wvNa1STVWDbuLdx5siAUjnFzIgjGHpYKXB/uOneWhj39j+OxNnMzILsErkEvRkJCIiJRqmTlWftiYzOQVe0hMTneUbz900vG1t4c7YPS0TOwey6tzNhFxaiNVOcFhAtnv24gRHRvQrHZl3vxpG1/+nsQXvyfxy7bDvH5/DHfWDS7x6xJnFrvdXibWb6WnpxMQEEBaWhr+/v6ubo6IiFxjdrudD3/5i+m/7ePIyUwAfDzdeCCuGo81j6JWFd/8D9w6B/v8IVjSD54/l38YlnZjIbojAAm7jjL02z/Zd/QMAJ0bh/Ha/Q0uOaQkxVPYz2+98yJiatpCXQpisVhYm3ScIyczCfH3oWezSB65pTqB5b0KPmjrHPiqJ5aLdmOxpCfDVz2h63SI7kh8rSDmD2jFe4t28Ony3exJPY2Ph2ZRuJJ6WETEtLSFuuSy2+38ujOVaSv38MY/GhAaUA6AdUnH2X/sDO0bhOLpfplAYbPC+Bi4oGfFmQX8w2DgJnBzd5Ru3H+Ccl7u3BjsB0BGtpUTZ7IJCfC5Gpd23Svs57fiooiYkrZQFzDCwYzVSbR571cenbKaJduP8NmqfY7XY6tXpFPj8MuHFYA9yy8RVgDskH4A9q1yKm0UEegIKwAfLN7J3e8u48vfk7BpV9wSoyEhETGdy22hbsHYQv3u6BAND5VRh09m8HnCPj7/PYljp7MA8PX2oGuTCB65pXrhTmLNgZyz4H0ubPy1uHDHnTpU8Cltdv7Ye4yTmTm8OHsTczYe4M1/NKRG5QqFO7cUmwKLiJhOUbZQj68VVHINkxKRmWOlzXu/cuKMsaS4WsVy9GpWg643R+Dvc4lN3aw5kLwR9i6HvSsgKQGa9oW7Rhiv12gGCR9cvgG+Ba8IcnezMPNf8UxbtZd3ft7Ob7uP0e79X3n27jr0bhGlAH0NKbCIiKkcP53Ft+v/LlTd1XuPahJuGWCz2fltz1HiawZhsVjw9nCnU6MwNh9M5/EWUdwdHYxHQUM+2Rnw+6TzASXrlPPrB9ef//qGNsYclfRkyLf/7twclshmsP0nqN0a3PMGJHc3C31aRHF3vWCGzf6TlX8d5fV5icz98yDvdWtMzYJWJ8kV0aRbEXG5jGwrixMPM3v9AZbtOEy2tfD/LVUs78mS526/9MoQMaXTmTl8s+5vpqzYw96jZ/imXzxxkZUAyMqx4XXxqhxrDqRshFNHoE47o8xmg7drwtnjxnOfQKjR4vyjan1wu+A851YJGS78OTsXertOh3IV4bP7oHIduOdNqHVngddgt9v535q/efVH4y7PCwa1olrF8sV+T65HWtYsIqZmtdn5ffdRZq8/wPzNKZzMzHG8Vi/Ej+T0DNLOZOf7ezAY+214uFkI8vV2Citvzd9GBW8Pbq9ThehQfywW9b6YzcETZ/ksYS8zfk8iPcP4e/fz8WD/sbPERRp1vDzczgeUvSuMx74EyDoJ/uFwY1uwWIww0nwAuHsbASU4xjmgXCy6oxFK5g9xnoDrHwbt3jReT5wL5YMgdTv8936oex+0eQ0qReU5ncVioevNEdxWpwpbD6Y7hZX9x84QUUnh5WpRD4uIuETqqUyavrEY67lVFuGB5ejUOIzON4VzY7CfY5UQ5Pt7MBO7x3JXvWBS0jIcHwqZOVYav7KQs+fuIRPs780ddapye52qtLihsjb9crETZ7IY8f0W5m1Kdvy91wgqz2PNo3ggrhoVLvz7+WkIrP/CCCgX8gmAyObwj0/A+wqGXmxWYzXQqUPGnJXIZk5LmTl7ApaNhd//A3arEYiaPwMtBoHX5SfYrvorlR5TVtOrWQ2ebXMj5b30s1eQwn5+K7CIyDV34MRZvt9wgKSjZ3izS0NH+YCZ6ynv5UHnxmHcXKMSbhfNRSnqPixns6x8u/5vlmw7wsq/Uh3BBYyb2j0aX4OX7ou+BlcohWG12bn9nSXsP3aWW2tW4vFm1bkjIAX3pBVGeHhwGnga+6vw01D4feL5gJI7xBMc4xwsrrXD2+CnF2DPMuN5tZuhz0Kjd+cS3pq/jQlLdwEQUakcb/6jIc1rV77WrS2VFFhExKXSzmQzb3Mys9cfYPWeY47ylUPvJDywXKHPU9ydbjOyrazec4wl2w+zdPsR9qSeZki7uvS7vRYAx05n8f6iHdxetyrxNYPw8SzBD8HrwMmMbGb9sZ+fNqcw44lbHUM861b/SuiJNYQeW2NMks08f+8fHv0BoloZXx/dZUygLemAkh+7HbbNhZ9fhLtGQoMHCnXYku2HGf7tJg6eC9wP3RzBsPb1CCh3iZVO1yEFFhFxiT/2HuPT5btZsu0IWVYbYPwy2jSqEvffFM69DcNcMjSzJ/U0vt4eVPHzBmD2+r8ZNGsjYMyHaVarMnfUqcLtdapq3sEV2H/sDFNX7uWrNfs5m5mJOzbGdmvC/TdVg98mwvyhzgd4B0CNcz0o0Z0hINwl7S6U7Azw8D7fu7JhBvy9Gu4cAeUr5XvIyYxs3pq/nf/+Zmx2V9XPm7FdGnJH3aol1WrT06RbESkRNpudLKvN0UOx/9gZft5ibLxVN8SPzjeF07FRGGFF6FW5FqIu2tirVhVfHmlanSXbDpOclsEv2w7zy7bDwBZqV/Vl3IONaBQR6JK2ljZ2u501+44z5de/OLjtd26xbOV9t0Ru9dnOnw1epHH9DkbFyGbOAcUVQzxXwvOCrfizM2Dhy3D6MGz+Fu58CeIeA3fnj1U/H09e7RzDfQ1DGfrtJvaknubUBRPMpfDUwyIixbI95SSz1x9gzoYDPNqsBv++zRhqOZ2Zw//98hedGodRL9T8/xbtdjvbD51kybYjLNl+mLX7jhvDUC/eRVV/4wPql22HSD2Zxe11qjjK5Jwzx0hdMYUNy+dyi9s2/C1nnV+PfRQ6ntuszWYD7KUnoFzO3hXG5OBDm43nVevDPWMhqmW+1TOyrfyw8SAPxFVzrF47cOIsYQE+1/VqNg0JichVl5x2ljkbDvLdhoMkJp+fe3BzjYr8r28zF7bs6kk7k826/ce5o875Lvvun/7Oir9SAagf5s+ddY2VR40jAq+vTetsVkj5k1Onz/KHtbYxrHE6Fd6u5ahi9fLH/cIelJAGZSeg5MeaA+umwS+vnd8LJrqzsQw6MOKShx4/nUXrd5fRoFoAr9/foEhzu8oSBRYRuWrsdjt9PlvDku2Hyf0fw9Pdwh11qnL/TeHcUbdqmZ60+tGSv1iw9RB//n2CC//HrFjekzbRIbzZpUHZ/A35XEDJ3QfFtnclblkn+c1en545I1g59E5jTtCPz2GvGImlRsuyH1AKcuYYLHkD1kwGuw36LIKImy95yKKth3jyi3VkWW1U8HJn6D11+WfTyDyr5co6zWERkWLLyrHxx95jjmWYFouFcl7u2O1wS41KdL4pnPYNQq6b3WX731Gb/nfUJvVUJsu2G0NHv+44wvEz2SSnZziFlc9/20fjiEDqh5XyTeu+eQJ2zHdaxeMGpNvLk2rzpWbl8hxKzzACy73vUIqv9OooXwnufQfiesGuX5zDypHtUPnGPEuhW0cHM29AS4Z88ydr9x1nxPdb+GFjMm92aaDt/fOhHhYRAYzJs2uTjjN7/QHmbUrmxJlsFg1uRe2qxp1udx85hZeHm7YdPyfHamNd0gnc3XBsJ5+SlsGtY4w7Alf1Mzatu6NuFVrcUKVkVkZdbjO0/Orn9qAc2Q6dPjz/2hddYefPnLJUICGnDr/Z6vGbLZrQG5vQu2Vt4msFle5AVlKO7YGPmkL1ptBuLATn3QfIZrPz39/2MXb+Ns5kWfHycGNQ6xvpe1vN6+I91pCQiBTKX4eNybPfbzjI38fPT5is6ufN2AcaOs3lkEv76/Apxs7fxsq/UjmT5bxp3c01KvFEq5rX7v3cOqeA7ebHGtvNQ54hHvYlQGba+fqDtmD3Dzc+JA+s48TZHOKnHQGLOw/EVeOx5jX0m39Rbf4GZvcDayZY3OHmx+GOYcb9ii6y/9gZXpy9ieU7U+napBpvPdDIBQ0ueQosInJZy3ceocfk1Y7nvt4etIsJoXPjcOJrBV1fE0qvosycc5vWnVt5tCf1NAAfPHwTHRuFAcaH087DJ4mvWZlyXlc458NxQ7+L/zu/4IZ+0R2NFS2/T3Ku4u0P1eM5UvlmJp24hR2nfPhvn6aOl5duP0zjiMDrZvjvmji+Fxa8BIk/GM/LVYK7XobYnnl6wOx2O99tOMCddYMdG8wdO51FBW93vD3K5twgBRYRcXIyI5uftxzCzQL/iK0GGB+szd/8hcYRgXRqHE7resFX/uEpeexJPc3S7Ye5/6Zwxwf/R0v+4u2ft+Pt4UZ8rSDurFuVO4qzaZ3NCu/FwMmDBdfxD4eBm2Dr9/DDAKgeDzVaYK/Rgl/TQ/h01X6W70x1VP95YCvqhPgV51LlUnYtMTbOO7LNeB7RFB6bf8mbNdrtdnpN/YMDJ84ytksDx/BjWaJJtyJCttXGrzuOMHv9ARYlHiIj20b1SuW5/yaj29/bw50VQ+4s0yt8zCCqcgWiKjvf6dfbw42wAB8OpmWwdPsRlm4/AmyhVpUK3FPLh771s/G1njSWyp49du7Pc4+4XlDzduNEKz+4dFgBSD9gzG2pex/U60iGzcLs9QeYMmsPOw8nA+BmgTbRIfRpGcWNwRr2uSZq3QF9V8Afk40VRbXuvPSdpYHktAy2HEwn9VQmD0xK4NH4Gjzfto7zjSKvE9ffFYtcBzbuP8E36/5m7p/JHDud5SivWbkCnW8KJ8tqc3Qvmz6sFHUiqRnkZBrBwsv3/B2Fj+6CbT86AsjjZ4/TJ+w4mb5HyT51lE/LPcaHhxuy68hpdqetxHfDWwWfP+LW84ElJ6Pgehc6dQg8jN6dhZsPMuzbTYAxDNi1SQSPNa+hWxKUBHdPuLWvcT8izwve76TfYN9KuLW/0466YYHlWDS4Fa/9mMjXa/9m2qq9LNx6iDH/aECrG6u44AJcR4FFpAz68vckZq3ZD0BlX286NArl/pvCaRAeULpWHRRmIum1lJ2Rt4cjtPH5DcH2/wEJ/2fswXH2xPk62cacFe7/DzR6yPg6dScsHOF0egvgc+4x6I4Aeje6m+U7j2D/2wZ/1TQmZparyMK92fyd4YOXbyVCQsII9W5AHZvdmGMUcWuhLmXDcW8an/u6XUwIN9eoSNv6IXS9OQJ/H92Mr8RVuODOzTYrzHsOUjbBuv9C2zegzj2OZdCB5b1458FGdGwUxrBvN3HgxFl6TlnNA3HVGNkhGr/r5O9Pc1hESrEjJzP5YeNBvt9wgFc6xdD43L1v/th7jC9/T6LzTeE0rxWEh/ulu51NqbATSQsjJwvOpJ4PFGcuGmJp8CCExBh1t/8Ecwcb5Tln856r8yRo/LDx9Y6f4cuu+X9Pixu0f9tYFQJweBuseNeYcHkuiDge5StCYA2oEJTnNCczsukxeTUbL9q0LrC8J61uqEKHBsG0+PEOvM+kkN8caZsdUgiim89/WPLCXaXzZ6Gss9th0/9gwQg4lWKU1boL2r0JVW50qno6M4e3f97OZwl7iQqqwLwBLc3fS3oZmnQrUkadzsxhwdYUZq8/yMq/UrHajH/CvZrVYFTH+i5u3VVis8L4GOeeFScWY3io3ZvGstyLA8jZE8bS0RotjOobZ8Lsfxf8/S4VQizuzuGi+QCod5/xWtrfRsApVxHKBV5Qr5Kx+uYy8xOKIvVUJr/uOMKS7UdYtv0w6RnGDfS63RyBJfEH3sg2hpAuDC3nfjTolz0Qz5hOvNophooVtNrHtDJPwvJxkPARWLPAzQOa9oXbXgCfAKeqa/cdw2KxEFvdWB5ttdk5fiaLyr7ermj5FVFgESljTpzJYuScLSzYcoiz2ef3+GgcEcj9N4Vzb8PQUvmfVb72LIfP7ruyc3SeCI0fMb7esQBmPHSuNyOfHo6YLlCtiVH37Ak4vuf8a97+eXYodbUcq431+0+wZNthgip48eqPibR1W81Iz+mEWY456h20B/FKdg9+tt3CjCduJb5W3h4cMaGju+DnF42dhgH+8Qk0LKAn75xPft3Nh0v+4qV76zndXLE00CohEROz2uys3nOMwyczqOrnwy1RlfLseWK32zmUnklIgDEBz8/Hk4RdRzmbbaVGUHk63xROp8bhRFWu4IpLuLZOHSpcvUq1jS5zR/gIPN/DkRtAAGq3hpePFi54lAuEcjcVp9UlxsPdjZtrVOLmGpX4fsMBAH623cLCzCbc4raNqpzgMIGsttXFhtHLc/hkISfniusF1YJHZsHOhbDpa4h54PxrGWl5eltsNjsLtqaQdjab57/+kzkbD/LG/Q3K3CRqBRaREjZ/czKv/LCV5LTzHyChAT6M7BBNu5hQ9qae5rsNxs6zpzJz+G3YXbi7WXB3s/Bq5xiC/X1oVK2UTZ4tqnKF7AnoMB6iWl6+3lUcmjGbqn7nV5TYcOM3W96t3y+uJ6XEDXcbj1yZJ41t/mveAa1HgV8wAG5uFmY8cSufLN/De4t2sHxnKm3H/8oLbevQM75GmbmZooaERErQ/M3J9Pt8XZ5ppLmiKpdnT+oZx3MfTze+69+cuiHX0c/06VSY1QOSVl2iksVYLTRwk/mXOF9jVpudFmN/ISUtI9+fKwsQEuDDiiF3aufi0m7LbPhfL+NrLz9jbkvTvo7l6mDc82voN5tYvdcYGmwSWZG3H2xk6p7Ywn5+l91fO0RMxmqz88oPWwsMKwB7Us9gAVrdWIX3ujVi7Ut3X19hBeCvRUZY8cjtEbj4Q/bc83ZvXvdhBcDdzcLIDkavSgHvFCM7RCuslAX174fHF0NYLGSdNJbJT4w3ho7OqVnFl5n/upVXO9Wngpc7G/8+QVaOzYWNvnrUwyJyjWRkW0lOyyD5xFkOnDjL6j3H+N/avy973MTusdwTE1oCLTSxX9+Beh2MOwjn2Ycl3AgrJbEPSylyuaFGKUNsNtg4AxaNgtOHjbIb20HX/zr1thw4cZY1e4/RqXG4o+zIyUyq+Jlrcr4m3YpcQzlWG4dPZpKcdpaDJzI4eOIs9cMCaHGDsRnUtpR02o1fXqxzl5XfhgrNZoPfPoLG/zRW8AC0es74s0odqHtv6dvp1gXaxYRyd3TIZSdzSxng5gY3/dNYXr/sLeOGlt5+TmEFIDywHOEXhJXNB9L4x8RV9GkRxYC7bih1+7cosIhcxG63c/xMNgdPnMXX24Ma58Z+D5w4y4AZ6zl44iyHTmY69j/J1TM+0hFYQvyN4Yxynu6EBfoQFlgOdwss3ZHK5VxXkyMzT8LsvrBtLuxcAD2+zztB1s29cBNrBXc3i5YuX098AqDt6xD7qBFYcp1Ign0JxlLoCybn/7wlhawcGxOX7uLnzSmMfaAhN9coPTdTVGCR6441d0tzIO1MNpNX7iH5xFmS04yekoNpZ8nINno5Ho2P5JVOxg6oFbzcWbPvuOM8Hm4Wgv19CA8sR2igD42qBTpeCyjnyfoRdxNY3tOxmqewkyNviSo9/4FckWO7YcYjcCQR3L2g0cNlejWPyDVz0W64LBgBW7+DNZPhnrEQZizTf7ZNHWLCAxjx3WZ2p57mwUkJ9IyP5IV2dfEtBTdTLFYLJ0yYwNtvv01ycjL169dn/PjxtGxZ8G9Ay5YtY/DgwWzZsoWwsDBeeOEF+vbt63j9k08+Yfr06WzevBmAuLg43njjDW655ZbiNE+uc1k5NtYnHTcCSNpZDp44S/KJDA6cCyWdGocx+lwIsWPng8U78z1PZV9vPC/YxjygnCcT/hlLSIARUir7ehfY3W6xWPLsKJo7ObLf5+uw4Lzh/HU3OXL3UmO1w9nj4BsCD33hvG+KiBSP3Q6hjYyJuPt/h4/vgNiecNfLUKEybeuHcGvNIN74MZFZa/YzPWEfi7YeYkyXhtxm8pspFjmwzJo1i4EDBzJhwgSaN2/Of/7zH+655x62bt1K9erV89Tfs2cP7du354knnuDzzz9n5cqVPPnkk1SpUoUuXboAsHTpUh5++GGaNWuGj48Pb731Fm3atGHLli2Eh4fnOaeYT2E2QrtSNpud1NOZJJ/I7Qkx/kxOO8tNERV5olVNAM5mWen28W8FnufgifOTEgPKedLj1kiC/b0JDShHWGA5wgJ9CAnwcdzNOJfFYqF9gyubvNguJpSJ3WPzTI4MuV4mR9rt8Pt/jF087VYIj4NuX4B/Gb9ukZJisUDLwcZNNxeOhE1fwbrPYMt3xu0qbn6cgHKejH2gIR0bhzH02z/Zf+ws+46eBswdWIq8Sqhp06bExsYyceJER1m9evXo3LkzY8aMyVN/yJAhzJkzh8TEREdZ37592bhxIwkJCfl+D6vVSsWKFfnwww/p2bNnodqlVUKuc7VWJ6RnZF8QRs4S7OdD62hjY6STGdnEvbqILGv+E1Jb16vKp4/eDBhzUO55fzmB5T2NAHIuiIQG+pz72sfldzctiYBnSpmnYFJzOL4XGj4EHd4Hz+tozo5ISduXAD+9ACl/Gs/bvA7NnnK8fCYrh5mr99Or2fkN5o6eyqRSBa8S25zymqwSysrKYu3atQwdOtSpvE2bNqxalf8mTwkJCbRp08aprG3btkyePJns7Gw8PfN+cJw5c4bs7GwqVSp4LD8zM5PMzEzH8/T09KJcilwlBW2ElpKWQb/P1zGxeyztYkLJyLaSkpaBzW6nZhVfwFj2++//rnWstDmVmeN0jtb1qjoCi6+3Bx7uFnJsxqTU0HMTWcMCfAgNKEfdkPMTziwWC/MHtrqm132lrtvJkd6+8PBM2LUEbu1nunv0iJQ5kfHwr6Ww/r/wx2Ro8tj512w2ynt50LtFlKPoTFYOnSespF6IP691jqGqv49pfsEqUmBJTU3FarUSHBzsVB4cHExKSkq+x6SkpORbPycnh9TUVEJD8/4GPnToUMLDw2ndunWBbRkzZgyvvPJKUZovV9mlNkLLLXt6xnr8vDdx7Ew24NwT4u3hxh97j3Em6/yN/ALLexIaUI7wQB9iIys6yi0WC0ueu51KFbyc5pVIKXBwPaTuPH/ztqr1jIeIlAw3d4jrZawmyv0lwWaDafcadzRvMRC8jNWQf+w9TvKJDPYfO0vC7qN0bhzOwsRDpJhgf59iTbq9uJvIbrdfsusov/r5lQO89dZbzJgxg6VLl+LjU3BX8bBhwxg8eLDjeXp6OhEREYVqv1wZq83OntRTzF5/wGkYKD/ZVrsjrJTzdMftgr9zi8XC2w80wr+cx7n5Iz6U9yr4RzLYX0MHpc6mr+H7/mDLgYo1IEIT6UVc5sLP3J0LjB2lk1bBhi+gzatQ/x/cdmMV5j7Tghe+/pM//07jv7/ty3Oai3vQS0qRAkvlypVxd3fP05ty+PDhPL0ouUJCQvKt7+HhQVCQc5f4O++8wxtvvMGiRYto2LDhJdvi7e2Nt7e5dusry+ZsPMiGpBNsPpDG5oNpTr0il/N82zo8ckt1pyW+ue5tqMmWZZLNCotHw8rxxvMb2hqbwImIOdzY1tgZd8FwY9+Wr3sbQ0b3jKVuSAP+9+94mry+iJMZOXkOtWOsbHzlh63cHR1SYsNDRepb9/LyIi4ujoULFzqVL1y4kGbNmuV7THx8fJ76CxYsoEmTJk7zV95++21effVV5s+fT5MmWt7oClabnb8On2T2+r/5dPlup9c++uUvpqzcw+pzQzjlPN25Mdi3UOeNrV6RiiU4gUtcLCMNZjx0Pqy0GAQPzzA2uRIRc7BYjNtb9F8NdwwHj3KwbyX8pxX8+Cwbdic7woobNm5120pHt1Xc6rYVN2zYgeS0DFbvOVZiTS7ykNDgwYPp0aMHTZo0IT4+no8//pikpCTHvirDhg3jwIEDTJ8+HTBWBH344YcMHjyYJ554goSEBCZPnsyMGTMc53zrrbcYMWIEX375JTVq1HD0yPj6+uLrW7gPRSm6Pamn2bD/OH/+ncbmA2lsOZju6Dnx9nCjV7MaeJybL9KxcRhHTmbSIDyABtUCqHVu4qw2QhMnqX8ZYeXoTuPmhZ0+ggYPuLpVIlIQz3LGXZ8bPWzcTHHLbDiwjpQw43/1tm6rGek5nTDL+WBy0F6JV7J78rPtFg6fvPS0gKupyIGlW7duHD16lNGjR5OcnExMTAzz5s0jMjISgOTkZJKSkhz1o6KimDdvHoMGDeKjjz4iLCyMDz74wLEHCxgb0WVlZfHAA87/sY0cOZJRo0YV89IkV+6cky0H0+nYKMzR0zFmXiILth5yqlvO0536Yf7EhAdwJtuK/7nA0v+O2vmeWxuhiZPt84yw4h9ubAZ3bodNETG5wAh4cBo06QPeflQ9W562bquZ6Dk+z13AQzjGRM/x9MseSFW/W0usibpbcxmTG07+/DuNTQfy9pysGHIH1SqWB+DT5buZvzmFmPAAp56TogYM3SVWHOx2407LcY+Cb1VXt0ZEismak0PqazdS1X40390HbHY4bAmiyks7cPe4sm39C/v5rcBSilltdnYfOUW1iuUp52XsyvrmT9uYtGxXnrq5PSevdo6hXujVf3/Msk5fSlh2Bix/B5oPNPZYEZGyYc9y+Oy+y9d7dO4V35z0mmwcJ66TG042Hcjbc/LF401pXtu4S3B0mL/TsE6D8AAaVgugZjF6Toriut0I7XqWfhBm/hMOroOju+DBqa5ukYhcLacOXb5OUepdBQosJmS12cmx2Rz3svl5SwqDZm3IdylxOU93p0lP7eqHcG+DUPVuyLW1/w+Y1R1OpUC5isamVCJSdvjmv1VJsetdBQosLnZhz0nuap2tyekMv7ce/2xqTGQODfBxLCWuH+ZPg2rn5pyE5+058fLQLrByjW34En4YANYsqFLPWLJcKeryx4lI6RHZDPzDID0ZCloH6h9m1CshCiwusif1NM//byNbk9Pz7TnZevD8vZHqhvizcFCraz6sI3JJ1hxj2eNvE4znde+D+yeBt9+ljxOR0sfNHdqNha96QkHrQNu9adQrIQosl3AlE0mtNju7jpxi0wWrdZrXrsygu28EoGJ5T9bsOw5AeS/nOSe5PSe5vDzcuCFYHwriYmeOGlvtA9w2BG4bCm7q0RMps6I7QtfpMH+IMWctl3+YEVaiO5ZocxRYClCcpboZ2Vbe/GmbY0Ls2WznnpMLh2sCy3sx4Z+x3FDVVz0nUjr4BUO3z+FkMtTv7OrWiEhJiO4Ide+FfauMCba+wcYwUAn2rOTSsuZ8zN+cTL/P1+UZtcuNFCPuiyagnCebDqTh5+PBs22Me6TY7XZiX13I8XM3+7u456RRRKBjh1iRUmH7fLBlQ70Orm6JiJRRWtZcTFabnVd+2JrvFKPcstFztzrKwgPLOQKLxWJhcJs6VPByp2G1AKIqq+dESim7HZaPg19eM7bu/tcyqHKjq1slItcxBZaLrN5zzGkYqCB1gn1pXrsKDar5Y7fbHdvd97g18lo3UeTayjoD3/eHLd8azxs/olVAIuJyCiwXKeyNnJ68ozadGodf49aIlLAT+2HmI5DyJ7h5QPu3oUlvV7dKRESB5WJV/Xyuaj2RUmPfKpjVA86kQvnK0O2/JbrHgojIpWhN4kVuiapEaIBPnrtT5rJgrBa6JapSSTZL5Nrb/pMRVkIawL+WKKyIiKkosFzE3c3CyA7RAHlCS+7zkR2iNZlWyp67RhqP3j9DYHVXt0ZExIkCSz7axYQysXssIQHOwz4hAT5M7B5b4D4sIqXK6aPw83DIyTKeu3tAy8HgVcG17RIRyYfmsBSgXUwod0eHFHunWxFTO7QFZjwEJ5LAlgP3jHV1i0RELkmB5RLc3SzE1wpydTNErq6tc2B2X8g+DRWjdKdlESkVFFhErhc2GywbC8veNJ7XvB0emArlNYFcRMxPgeVSbFZT3D9B5IplnoLZ/4Ztc43ntz4Jd79qzFsRESkF9L9VQbbOKeAOlWNL/A6VIlfs1CHYsxzcveC+9+Cm7q5ukYhIkSiw5GfrHPiqJ1x8R6H0ZKO863SFFildgmrBg1PB2w8ibnF1a0REikzLmi9msxo9K5e6/eH8oUY9EbOy22H1J7B76fmy2ncprIhIqaXAcrF9q5yHgfKwQ/oB2LeyxJokUiQ5WfDDMzDvOfhfLzh5yNUtEhG5YhoSutipQv7nPuMRYxJutSbQ6nmwaH8WMYFTh437Ae3/DSxu0GIw+FZ1datERK6YAsvFfIMLVy/rJOz8GY7thtteOF++ZAx4+0JYLIQ2Mr4WKQkHN8DMf0L63+AdAA9MhhvudnWrRESuCgWWi0U2M1YDpSeT/zwWC/iHQpepkLIR3D3Pv2Szwqr/MzbkAuM33Mp1IDwOwm+CiFshJKYkrkKuN5u+hu+fgpyzEFQbHp4JlW9wdatERK4azWG5mJu7sXQZKPD2h+3GQuSt0PTf0KT3+ZdzMqHVs1D3PvALA7sNjiTChs/hx2dh8Wjn023+Bo7sMDb0ErkSfy02wkrtu+HxxQorIlLmqIclP9EdjaXL+e7D8mbBS5q9ykPLZ88/P5kCB9bBgbVwcB1EtTr/WvpB+Ppc2PH2h7DGxjBSeByEx4J/uObFSOHd957xM3Tz49rcUETKJIvdbs9v3KPUSU9PJyAggLS0NPz9/a/OSa/lTreHtsLcQZC80fjN+GItBkHrUcbXOVmQdUpbqMt5R3cZy5bbvgFu6igVkdKrsJ/f6mG5FDd3iGp5bc4dHA19fgZrjjFsdGFPzKGtUDX6fN2//4Bp7aFiDaMHJrcnJrQheFW4Nu0T8/prMXz9GGSkgW8V5149EZEySoHF1dw9IKSB8Yh71CjLOuM8HHR0p/Hn8b3GY/M3xnOLG1SpB21fh1p3lGSrxRXsdkj4CBaOMOZHVbsFGv/T1a0SESkRCixm5FXe+XlcL4juBAfXn+uJWWf0xJxMhsNbwOuCpdN//g9W/+eCnphYqFRLwwalXXaGMYS48UvjeePucN+74OHt2naJiJQQBZbSolxFqHWn8ciVftAILyENzpclJRhDSH//cb7MO8CYkBkeB7f200ZipU16MszqDgfWgMXd6FFr2leTskXkuqLAUpr5hxmPCzUfANXjz8+HSd4ImWmwZ5nxiO9/vu6f/zOGmMJjIewmTeo1q1MpcGgz+ATCg9M0/Cci1yUFlrKmYqTxaPig8dyaDYcTjfBydBdUqHy+7sYvYdcv559Xqnl+GCk8zpgjUdihpGu5oup6F3YTPDAVqtY1/o5ERK5DWtZ8PVv7Gez51Qgzx3Y7v+YdAEP2ng8su5caw1JVo5139wXYOqeAPWvGFrxnjRTMmgO/vGrMWwqPdXVrRESuqcJ+fiuwiOHMMWNS78Fzk3q9KkCXT8+//n5jOL4HPHwgpOH5De7OnoCfXiDvbQzOza/oOl2hpSjOHjc2FNz1i7F54FNr8k7CFhEpQ7QPixRN+UpQ+y7jcbGcLGOY6cwxYz7M36uNxyWdCzA/PQ81WoKPv4aILpTfENrRv2DGw3BsF3iWNybXKqyIiADqYZGisNmMoaPcCb27lkLqtsIf71keYv4BnT4yntvt8GVXozfHqwJ4+Rl3t/aqYCzVDqrtPMH0yHbjHF4VwNsv79BUaZHfEFr5IGP/nZyzEBABD31pbAwoIlLGqYdFrj43N6hc23g06mbcIfibPoU/PvuMEVIufL5zQcH163U4H1jsdphwq7FhWi537/MBp+Yd0PGD86/9+KyxsZ6X7/mAk/u1fzhE3Hy+7tnjRhBy97r2S4W3zoGvepJnCO3MUePPynWg14/GDrYiIuKgwCLF5xtcuHrdv4XQxpB10ggZudw8oNME4z5JmSch6/S5r08ZdatdECpyMo1lvVmnwJpllFkz4Uym8WGf+4EPRrhZOw1sOfm3J6oVPPrD+ecf3GSEFjcPI9R4+53v5QlrDPeOO193+bvGcI5XhXNh6dzD29foJalSx7kdFwYgm9XoWckz3+cCumeUiEi+FFik+CKbGauB0pPJ/0PYYrxe83Zj/kqFIOeXPbzhpkJuLe/pA0P2GF/n3gzSEW5OO99TyW6HO18699pF9bJOQXCM87mzTht/2nIg44TxcHzfcs51V/0fnD2WfxtDG8O/l51//n+xcDr1fM8OdudhoPykHzDmtlyre1iJiJRSCixSfG7uxtLlr3pirAq6MLSc61lo9+bVn2zr4QUelQruiXBzM+52XVjDUy4KNSfPhx3vi8ZTY3savTEX1886DYHVnetmpEPmuUdRnDpUtPoiItcBBRa5MtEdjaXL+e7D8mbpWNLs5g4+Acbjcu5+pfDn7bfq3FDXuUCT9Dv8MvryxxV2qE1E5DqiwCJXLroj1L1XO91ezC/YeOSqHg9rPr38EFpks5JqoYhIqaHAIleHm7vmXVyOq4bQRETKgELeKEZErorcITT/UOdy/zDtCiwicgnqYREpaRpCExEpMgUWEVfQEJqISJFoSEhERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETK9YgWXChAlERUXh4+NDXFwcy5cvv2T9ZcuWERcXh4+PDzVr1mTSpEl56nzzzTdER0fj7e1NdHQ0s2fPLk7TREREpAwqcmCZNWsWAwcOZPjw4axfv56WLVtyzz33kJSUlG/9PXv20L59e1q2bMn69et58cUXeeaZZ/jmm28cdRISEujWrRs9evRg48aN9OjRg65du/L7778X/8pERESkzLDY7XZ7UQ5o2rQpsbGxTJw40VFWr149OnfuzJgxY/LUHzJkCHPmzCExMdFR1rdvXzZu3EhCQgIA3bp1Iz09nZ9++slRp127dlSsWJEZM2YUql3p6ekEBASQlpaGv79/US5JREREXKSwn98eRTlpVlYWa9euZejQoU7lbdq0YdWqVfkek5CQQJs2bZzK2rZty+TJk8nOzsbT05OEhAQGDRqUp8748eMLbEtmZiaZmZmO52lpaYBx4SIiIlI65H5uX67/pEiBJTU1FavVSnBwsFN5cHAwKSkp+R6TkpKSb/2cnBxSU1MJDQ0tsE5B5wQYM2YMr7zySp7yiIiIwl6OiIiImMTJkycJCAgo8PUiBZZcFovF6bndbs9Tdrn6F5cX9ZzDhg1j8ODBjuc2m41jx44RFBR0yeOKKj09nYiICPbv36+hpsvQe1U0er8KT+9V4em9Kjy9V4V3Ld8ru93OyZMnCQsLu2S9IgWWypUr4+7unqfn4/Dhw3l6SHKFhITkW9/Dw4OgoKBL1inonADe3t54e3s7lQUGBhb2UorM399fP9CFpPeqaPR+FZ7eq8LTe1V4eq8K71q9V5fqWclVpFVCXl5exMXFsXDhQqfyhQsX0qxZs3yPiY+Pz1N/wYIFNGnSBE9Pz0vWKeicIiIicn0p8pDQ4MGD6dGjB02aNCE+Pp6PP/6YpKQk+vbtCxhDNQcOHGD69OmAsSLoww8/ZPDgwTzxxBMkJCQwefJkp9U/AwYMoFWrVowdO5ZOnTrx/fffs2jRIlasWHGVLlNERERKsyIHlm7dunH06FFGjx5NcnIyMTExzJs3j8jISACSk5Od9mSJiopi3rx5DBo0iI8++oiwsDA++OADunTp4qjTrFkzZs6cyUsvvcSIESOoVasWs2bNomnTplfhEq+Mt7c3I0eOzDP8JHnpvSoavV+Fp/eq8PReFZ7eq8Izw3tV5H1YREREREqa7iUkIiIipqfAIiIiIqanwCIiIiKmp8AiIiIipqfAUoAxY8Zw88034+fnR9WqVencuTPbt293dbNMaeLEiTRs2NCxoVB8fLzTjSylYGPGjMFisTBw4EBXN8V0Ro0ahcVicXqEhIS4ulmmdeDAAbp3705QUBDly5encePGrF271tXNMqUaNWrk+dmyWCz079/f1U0znZycHF566SWioqIoV64cNWvWZPTo0dhsthJvS7G25r8eLFu2jP79+3PzzTeTk5PD8OHDadOmDVu3bqVChQqubp6pVKtWjTfffJPatWsD8Nlnn9GpUyfWr19P/fr1Xdw68/rjjz/4+OOPadiwoaubYlr169dn0aJFjufu7u4ubI15HT9+nObNm3PHHXfw008/UbVqVXbt2nVNd/8uzf744w+sVqvj+ebNm7n77rt58MEHXdgqcxo7diyTJk3is88+o379+qxZs4bHHnuMgIAABgwYUKJt0bLmQjpy5AhVq1Zl2bJltGrVytXNMb1KlSrx9ttv06dPH1c3xZROnTpFbGwsEyZM4LXXXqNx48aXvDv59WjUqFF89913bNiwwdVNMb2hQ4eycuVKli9f7uqmlEoDBw5k7ty57Ny586rei64suO+++wgODmby5MmOsi5dulC+fHn++9//lmhbNCRUSGlpaYDxQSwFs1qtzJw5k9OnTxMfH+/q5phW//79uffee2ndurWrm2JqO3fuJCwsjKioKB566CF2797t6iaZ0pw5c2jSpAkPPvggVatW5aabbuKTTz5xdbNKhaysLD7//HN69+6tsJKPFi1asHjxYnbs2AHAxo0bWbFiBe3bty/xtmhIqBDsdjuDBw+mRYsWxMTEuLo5prRp0ybi4+PJyMjA19eX2bNnEx0d7epmmdLMmTNZt24df/zxh6ubYmpNmzZl+vTp3HjjjRw6dIjXXnuNZs2asWXLFseNU8Wwe/duJk6cyODBg3nxxRdZvXo1zzzzDN7e3vTs2dPVzTO17777jhMnTtCrVy9XN8WUhgwZQlpaGnXr1sXd3R2r1crrr7/Oww8/XPKNsctlPfnkk/bIyEj7/v37Xd0U08rMzLTv3LnT/scff9iHDh1qr1y5sn3Lli2ubpbpJCUl2atWrWrfsGGDo+y2226zDxgwwHWNKiVOnTplDw4Oto8bN87VTTEdT09Pe3x8vFPZ008/bb/11ltd1KLSo02bNvb77rvP1c0wrRkzZtirVatmnzFjhv3PP/+0T58+3V6pUiX7tGnTSrwtCiyX8dRTT9mrVatm3717t6ubUqrcdddd9n/961+ubobpzJ492w7Y3d3dHQ/AbrFY7O7u7vacnBxXN9HUWrdube/bt6+rm2E61atXt/fp08epbMKECfawsDAXtah02Lt3r93Nzc3+3XffubopplWtWjX7hx9+6FT26quv2uvUqVPibdGQUAHsdjtPP/00s2fPZunSpURFRbm6SaWK3W4nMzPT1c0wnbvuuotNmzY5lT322GPUrVuXIUOGaBXMJWRmZpKYmEjLli1d3RTTad68eZ5tF3bs2OG4Ka3kb+rUqVStWpV7773X1U0xrTNnzuDm5jzd1d3dXcuazaR///58+eWXfP/99/j5+ZGSkgJAQEAA5cqVc3HrzOXFF1/knnvuISIigpMnTzJz5kyWLl3K/PnzXd000/Hz88szD6pChQoEBQVpftRFnnvuOTp06ED16tU5fPgwr732Gunp6Tz66KOubprpDBo0iGbNmvHGG2/QtWtXVq9ezccff8zHH3/s6qaZls1mY+rUqTz66KN4eOijsCAdOnTg9ddfp3r16tSvX5/169fz7rvv0rt375JvTIn36ZQSQL6PqVOnurppptO7d297ZGSk3cvLy16lShX7XXfdZV+wYIGrm1VqaA5L/rp162YPDQ21e3p62sPCwuz/+Mc/NC/qEn744Qd7TEyM3dvb2163bl37xx9/7OommdrPP/9sB+zbt293dVNMLT093T5gwAB79erV7T4+PvaaNWvahw8fbs/MzCzxtmgfFhERETE97cMiIiIipqfAIiIiIqanwCIiIiKmp8AiIiIipqfAIiIiIqanwCIiIiKmp8AiIiIipqfAIiKmdfvttzNw4EBXN0NETECBRURERExPgUVERERMT4FFREqN+fPnExAQwPTp013dFBEpYQosIlIqzJw5k65duzJ9+nR69uzp6uaISAlTYBER05swYQJ9+/bl+++/p1OnTq5ujoi4gIerGyAicinffPMNhw4dYsWKFdxyyy2ubo6IuIh6WETE1Bo3bkyVKlWYOnUqdrvd1c0RERdRYBERU6tVqxZLlizh+++/5+mnn3Z1c0TERTQkJCKmd+ONN7JkyRJuv/12PDw8GD9+vKubJCIlTIFFREqFOnXq8Msvv3D77bfj7u7OuHHjXN0kESlBFrsGhUVERMTkNIdFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERM7/8BvMP8VKfpxR0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sklearn import metrics\n",
    "\n",
    "ARI = []\n",
    "AMI = []\n",
    "x = range(2, 9)\n",
    "for k in x:\n",
    "    ARI.append(metrics.adjusted_rand_score(overlap['Verhaak_subtype'], overlap[f'k = {k}']))\n",
    "    AMI.append(metrics.adjusted_mutual_info_score(overlap[f'k = {k}'], overlap['Verhaak_subtype']))\n",
    "    \n",
    "plt.plot(x, ARI, 'o--', label = 'Adjusted Rand Index')\n",
    "plt.plot(x, AMI, 'o--', label = 'Adjusted Mutual Information')\n",
    "plt.ylim(0, 0.1)\n",
    "plt.xticks(x)\n",
    "plt.xlabel('k')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ca799357-bf72-4b3a-9e91-5df84b5f5af8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.01541584781345078,\n",
       " 0.01948010903298123,\n",
       " 0.021814501721202834,\n",
       " 0.029093830262392522,\n",
       " 0.023193176824508634,\n",
       " 0.03217692513218214,\n",
       " 0.018973277517232698]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ARI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb12b446-42c1-4e21-8f76-5b96a2449d8c",
   "metadata": {},
   "source": [
    "## Plot overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "859d1a3b-94a3-4068-bdb4-198e42a6aa6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESENCHYMAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CLASSICAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLASSICAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEURAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESENCHYMAL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Verhaak_subtype  k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8  count\n",
       "0     MESENCHYMAL      2      2      2      3      3      3      3     23\n",
       "1       CLASSICAL      2      2      2      2      2      2      2     19\n",
       "2       CLASSICAL      2      2      2      3      3      3      3     14\n",
       "3          NEURAL      2      2      2      3      3      3      3     10\n",
       "4     MESENCHYMAL      1      1      4      5      5      4      6      9"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = overlap.iloc[:,6:].value_counts().reset_index()\n",
    "counts.columns = counts.columns[:-1].tolist() + [\"count\"]\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "495fc472-b0a0-4a74-85f6-1f628b38a2cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "06431653-b2a7-41a1-b098-59784043306a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: ggalluvial\n",
       "Loading required package: ggplot2\n",
       "In addition: Warning messages:\n",
       "1: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n",
       "2: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n",
       "3: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i counts -w 800 -h 600\n",
    "\n",
    "require(ggalluvial)\n",
    "\n",
    "cluster_alluvial <- ggplot(counts, aes(y = count,\n",
    "                   axis1 = Verhaak_subtype,\n",
    "                   axis2 = `k = 2`,\n",
    "                   axis3 = `k = 3`,\n",
    "                   axis4 = `k = 4`,\n",
    "                   axis5 = `k = 5`,\n",
    "                   axis6 = `k = 6`,\n",
    "                   axis7 = `k = 7`,\n",
    "                   axis8 = `k = 8`)) +\n",
    "    geom_alluvium(aes(fill = Verhaak_subtype), width = 1/12) +\n",
    "    geom_stratum(width = 1/6, fill = 'grey', color = 'black') +\n",
    "    geom_label(stat = 'stratum', aes(label = after_stat(stratum))) +\n",
    "    #scale_x_discrete(limits = c('Verhaak subtype', 'k = 2', 'k = 3', 'k = 4', 'k = 5'), expand = c(0.1, 0.05)) +\n",
    "    scale_x_discrete(limits = c('Verhaak subtype', 'k = 2', 'k = 3', 'k = 4', 'k = 5', 'k = 6', 'k = 7', 'k = 8'), expand = c(0.1, 0.05)) +\n",
    "    scale_fill_brewer(type = 'qual', palette = 'Dark2') +\n",
    "    theme_classic(base_size = 16)\n",
    "\n",
    "pdf(\"cluster_alluvial.pdf\", width=15, height=12)\n",
    "print(cluster_alluvial)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "696cc8c9-0f40-49a2-9403-aede0bf85ce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading required package: ggpubr\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i overlap -w 1200\n",
    "\n",
    "library(\"tidyr\")\n",
    "require(ggpubr)\n",
    "\n",
    "png(filename=\"cluster_ssgsea_scores.png\", width=1200, height=600)\n",
    "\n",
    "overlap_long <- overlap %>%\n",
    "    pivot_longer(cols = starts_with('VERHAAK_GLIOBLASTOMA'), names_to = 'subtype', values_to = 'score')\n",
    "\n",
    "#p2 <- ggboxplot(overlap_long, x = 'k = 3', y = 'score', facet.by = 'subtype') +\n",
    "#    stat_compare_means()\n",
    "    \n",
    "p3 <- ggboxplot(overlap_long, x = 'k = 5', y = 'score', facet.by = 'subtype') +\n",
    "    stat_compare_means()\n",
    "\n",
    "p4 <- ggboxplot(overlap_long, x = 'k = 6', y = 'score', facet.by = 'subtype') +\n",
    "    stat_compare_means()\n",
    "\n",
    "ggarrange(p3, p4, ncol=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a3aa41-ce0b-4314-af30-2a2becbed4d7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Assess enrichment of variants within clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2c6829-a140-4507-b30b-da944c6f6e38",
   "metadata": {},
   "source": [
    "## Statistically (Fisher Exact test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d33b9306-3870-444a-9ab4-e8edd47b3a17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify the clustering in which to test enrichments\n",
    "k = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8cd4f618-e0c3-4796-b421-7deb5fe970c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ABL1_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>AKT2_mut</th>\n",
       "      <th>AKT3_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF217_del</th>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <th>ZRSR2_del</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 659 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ABL1_mut  AKT1_mut  AKT2_mut  AKT3_mut  ALK_mut  APC_mut  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0       0.0      0.0      0.0   \n",
       "\n",
       "                 ARAF_mut  ARID1A_mut  ARID1B_mut  ARID2_mut  ...  ZNF217_del  \\\n",
       "sample                                                        ...               \n",
       "GLSS-19-0266-R1       0.0         0.0         0.0        0.0  ...         0.0   \n",
       "\n",
       "                 ZRSR2_amp  ZRSR2_del  k = 2  k = 3  k = 4  k = 5  k = 6  \\\n",
       "sample                                                                     \n",
       "GLSS-19-0266-R1        0.0        0.0      1      1      1      1      1   \n",
       "\n",
       "                 k = 7  k = 8  \n",
       "sample                         \n",
       "GLSS-19-0266-R1      1      1  \n",
       "\n",
       "[1 rows x 659 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = matrix.T.join(memberships)\n",
    "df.to_csv('../results/variants_with_memberships_v2.tsv', sep = '\\t')\n",
    "df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9bec0429-8896-40d4-902c-abbcafd7ea58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>cluster</th>\n",
       "      <th>p_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1071</th>\n",
       "      <td>JAK3_amp</td>\n",
       "      <td>14.859649</td>\n",
       "      <td>6.114934e-26</td>\n",
       "      <td>2</td>\n",
       "      <td>1.993468e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>BRD4_amp</td>\n",
       "      <td>14.643418</td>\n",
       "      <td>1.261330e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>2.055969e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>MEF2B_amp</td>\n",
       "      <td>14.193548</td>\n",
       "      <td>3.326003e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.614256e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>CRTC1_amp</td>\n",
       "      <td>13.584553</td>\n",
       "      <td>1.048592e-24</td>\n",
       "      <td>2</td>\n",
       "      <td>8.546023e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>AKT2_amp</td>\n",
       "      <td>13.800000</td>\n",
       "      <td>1.022993e-23</td>\n",
       "      <td>2</td>\n",
       "      <td>5.558260e-21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant  odds_ratio       p_value  cluster         p_adj\n",
       "1071   JAK3_amp   14.859649  6.114934e-26        2  1.993468e-22\n",
       "890    BRD4_amp   14.643418  1.261330e-25        2  2.055969e-22\n",
       "1101  MEF2B_amp   14.193548  3.326003e-25        2  3.614256e-22\n",
       "946   CRTC1_amp   13.584553  1.048592e-24        2  8.546023e-22\n",
       "834    AKT2_amp   13.800000  1.022993e-23        2  5.558260e-21"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_column = f'k = {k}'\n",
    "results = []\n",
    "\n",
    "for cluster in range(1, k + 1):\n",
    "    for variant in df:\n",
    "        if variant.startswith('k ='):\n",
    "            continue\n",
    "\n",
    "        # contingency table\n",
    "        cluster_with_variant = len(df[(df[cluster_column] == cluster) & (df[variant] != 0)])\n",
    "        cluster_without_variant = len(df[(df[cluster_column] == cluster) & (df[variant] == 0)])\n",
    "        non_cluster_with_variant = len(df[(df[cluster_column] != cluster) & (df[variant] != 0)])\n",
    "        non_cluster_without_variant = len(df[(df[cluster_column] != cluster) & (df[variant] == 0)])\n",
    "\n",
    "        contingency_table = [\n",
    "            [cluster_with_variant, cluster_without_variant],\n",
    "            [non_cluster_with_variant, non_cluster_without_variant]\n",
    "        ]\n",
    "\n",
    "        # Fisher's exact test\n",
    "        oddsratio, pvalue = stats.fisher_exact(contingency_table)\n",
    "\n",
    "        results.append({\n",
    "            'variant': variant,\n",
    "            'odds_ratio': oddsratio,\n",
    "            'p_value': pvalue,\n",
    "            'cluster': cluster\n",
    "        })\n",
    "\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df['p_adj'] = multipletests(results_df['p_value'], alpha = 0.1, method = 'fdr_bh')[1]\n",
    "results_df = results_df.sort_values('p_adj')\n",
    "\n",
    "results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc12ea7e-d774-4203-8157-5ef2b49f75fd",
   "metadata": {},
   "source": [
    "Get chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1219951f-2791-458d-aa69-4bb5a0ea8c1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5_2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5_2</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485.5_5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2_2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020.3_4</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52473</td>\n",
       "      <td>53312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id         gene chrom  start    end\n",
       "0  ENSG00000223972.5_2      DDX11L1  chr1  11869  14409\n",
       "1  ENSG00000227232.5_2       WASH7P  chr1  14404  29570\n",
       "2  ENSG00000243485.5_5  MIR1302-2HG  chr1  29554  31109\n",
       "3  ENSG00000237613.2_2      FAM138A  chr1  34554  36081\n",
       "4  ENSG00000268020.3_4       OR4G4P  chr1  52473  53312"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed = pd.read_csv('../data/example_bed_gencode30.v30lift37.annotation.bed',\n",
    "                  sep = '\\t')\n",
    "bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3c1bdae4-c0d5-4962-a5c6-05eb4816ecc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>cluster</th>\n",
       "      <th>p_adj</th>\n",
       "      <th>gene</th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JAK3_amp</td>\n",
       "      <td>14.859649</td>\n",
       "      <td>6.114934e-26</td>\n",
       "      <td>2</td>\n",
       "      <td>1.993468e-22</td>\n",
       "      <td>JAK3</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRD4_amp</td>\n",
       "      <td>14.643418</td>\n",
       "      <td>1.261330e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>2.055969e-22</td>\n",
       "      <td>BRD4</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEF2B_amp</td>\n",
       "      <td>14.193548</td>\n",
       "      <td>3.326003e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.614256e-22</td>\n",
       "      <td>MEF2B</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRTC1_amp</td>\n",
       "      <td>13.584553</td>\n",
       "      <td>1.048592e-24</td>\n",
       "      <td>2</td>\n",
       "      <td>8.546023e-22</td>\n",
       "      <td>CRTC1</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT2_amp</td>\n",
       "      <td>13.800000</td>\n",
       "      <td>1.022993e-23</td>\n",
       "      <td>2</td>\n",
       "      <td>5.558260e-21</td>\n",
       "      <td>AKT2</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3265</th>\n",
       "      <td>SOX9_del</td>\n",
       "      <td>0.937143</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>del</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3266</th>\n",
       "      <td>RB1_mut</td>\n",
       "      <td>0.932613</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>RB1</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3267</th>\n",
       "      <td>REL_mut</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>REL</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3268</th>\n",
       "      <td>PRF1_mut</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>PRF1</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3269</th>\n",
       "      <td>ABL1_mut</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3270 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant  odds_ratio       p_value  cluster         p_adj   gene type  \\\n",
       "0      JAK3_amp   14.859649  6.114934e-26        2  1.993468e-22   JAK3  amp   \n",
       "1      BRD4_amp   14.643418  1.261330e-25        2  2.055969e-22   BRD4  amp   \n",
       "2     MEF2B_amp   14.193548  3.326003e-25        2  3.614256e-22  MEF2B  amp   \n",
       "3     CRTC1_amp   13.584553  1.048592e-24        2  8.546023e-22  CRTC1  amp   \n",
       "4      AKT2_amp   13.800000  1.022993e-23        2  5.558260e-21   AKT2  amp   \n",
       "...         ...         ...           ...      ...           ...    ...  ...   \n",
       "3265   SOX9_del    0.937143  1.000000e+00        3  1.000000e+00   SOX9  del   \n",
       "3266    RB1_mut    0.932613  1.000000e+00        5  1.000000e+00    RB1  mut   \n",
       "3267    REL_mut    0.000000  1.000000e+00        5  1.000000e+00    REL  mut   \n",
       "3268   PRF1_mut    0.000000  1.000000e+00        5  1.000000e+00   PRF1  mut   \n",
       "3269   ABL1_mut    0.000000  1.000000e+00        1  1.000000e+00   ABL1  mut   \n",
       "\n",
       "      chrom  \n",
       "0     chr19  \n",
       "1     chr19  \n",
       "2     chr19  \n",
       "3     chr19  \n",
       "4     chr19  \n",
       "...     ...  \n",
       "3265  chr17  \n",
       "3266  chr13  \n",
       "3267   chr2  \n",
       "3268  chr10  \n",
       "3269   chr9  \n",
       "\n",
       "[3270 rows x 8 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df['gene'] = results_df['variant'].str.split('_').str[0]\n",
    "results_df['type'] = results_df['variant'].str.split('_').str[-1]\n",
    "results_df = results_df.merge(bed[['gene', 'chrom']], on = 'gene', how = 'left')\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "846d1f77-149b-4aba-b8f5-599974ebb96d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    cluster  chrom type  odds_ratio     p_adj\n",
      "31        1  chr19  del   15.142711  0.000021\n",
      "63        1   chr9  amp   12.840870  0.004407\n",
      "60        1   chr8  amp    8.037606  0.004500\n",
      "57        1   chr7  amp    0.311201  0.005256\n",
      "4         1  chr10  del    0.313054  0.011933\n",
      "52        1   chr5  del    4.912999  0.047965\n",
      "30        1  chr19  amp    0.262997  0.067320\n",
      "9         1  chr12  amp    4.007076  0.106587\n",
      "54        1   chr6  amp   10.334962  0.114856\n",
      "58        1   chr7  del    5.072003  0.196899\n",
      "     cluster  chrom type  odds_ratio         p_adj\n",
      "101        2  chr19  amp   12.829273  2.236061e-19\n",
      "107        2  chr20  amp    3.743551  1.692482e-06\n",
      "87         2  chr14  del    0.096228  4.662876e-06\n",
      "128        2   chr7  amp    3.666771  6.065128e-04\n",
      "84         2  chr13  del    0.284737  1.104095e-03\n",
      "90         2  chr15  del    0.204977  5.584740e-03\n",
      "78         2  chr11  del    0.110984  7.389302e-03\n",
      "114        2  chr22  del    0.403311  7.730819e-03\n",
      "111        2  chr21  del    0.048195  1.472848e-02\n",
      "75         2  chr10  del    2.408003  1.622160e-02\n",
      "     cluster  chrom type  odds_ratio         p_adj\n",
      "172        3  chr19  amp    0.084415  4.676460e-10\n",
      "158        3  chr14  del    6.186742  4.043066e-09\n",
      "178        3  chr20  amp    0.157827  1.667353e-08\n",
      "146        3  chr10  del    3.631652  3.376942e-04\n",
      "185        3  chr22  del    2.744805  9.381112e-04\n",
      "199        3   chr7  amp    2.977558  1.441051e-03\n",
      "142        3   chr1  amp    0.188100  2.162526e-02\n",
      "155        3  chr13  del    2.221836  2.823381e-02\n",
      "205        3   chr9  amp    0.180608  4.995241e-02\n",
      "166        3  chr17  amp    0.169105  6.412813e-02\n",
      "     cluster  chrom type  odds_ratio     p_adj\n",
      "254        4  chr21  mut   48.571429  0.006436\n",
      "257        4  chr22  mut         NaN  0.006722\n",
      "245        4  chr19  mut         NaN  0.009455\n",
      "236        4  chr16  mut         NaN  0.017170\n",
      "281        4   chrX  mut         NaN  0.027588\n",
      "248        4   chr2  mut         NaN  0.033413\n",
      "260        4   chr3  mut         NaN  0.073480\n",
      "275        4   chr8  mut         NaN  0.077648\n",
      "215        4   chr1  mut         NaN  0.100820\n",
      "272        4   chr7  mut         NaN  0.103691\n",
      "     cluster  chrom type  odds_ratio     p_adj\n",
      "284        5   chr1  amp    7.435227  0.000093\n",
      "341        5   chr7  amp    0.287380  0.000165\n",
      "335        5   chr5  amp   10.026398  0.000224\n",
      "305        5  chr16  amp   22.515268  0.000299\n",
      "311        5  chr18  amp    6.332659  0.000951\n",
      "317        5   chr2  amp   16.459054  0.001568\n",
      "288        5  chr10  del    0.338588  0.001641\n",
      "329        5   chr3  amp    4.757587  0.002777\n",
      "323        5  chr21  amp    4.967653  0.003060\n",
      "291        5  chr11  del    6.695545  0.004958\n"
     ]
    }
   ],
   "source": [
    "hallmarks = results_df.groupby(['cluster', 'chrom', 'type']).agg({'odds_ratio': 'mean', 'p_adj': 'mean'}).reset_index().sort_values(['cluster', 'p_adj'])\n",
    "\n",
    "for cluster in range(1, k + 1):\n",
    "    print(hallmarks[hallmarks['cluster'] == cluster].head(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9f7591e-811b-4658-8e38-15e6c4550319",
   "metadata": {},
   "source": [
    "## Visually (heatmap)\n",
    "Sort according to genomic coordinate so we can better visualize clustering of CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9bb68c66-282c-40bd-8f89-b72d42168105",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_old</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MTOR_mut</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MTOR_amp</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>SDHB_amp</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>ID3_amp</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ENSG00000117318.9_3</td>\n",
       "      <td>1</td>\n",
       "      <td>23884417</td>\n",
       "      <td>23885992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ARID1A_mut</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ENSG00000117713.20_5</td>\n",
       "      <td>1</td>\n",
       "      <td>27019727</td>\n",
       "      <td>27108595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>BCORL1_mut</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>ENSG00000085185.15_5</td>\n",
       "      <td>23</td>\n",
       "      <td>129115083</td>\n",
       "      <td>129192058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>BCORL1_amp</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>ENSG00000085185.15_5</td>\n",
       "      <td>23</td>\n",
       "      <td>129115083</td>\n",
       "      <td>129192058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>GPC3_amp</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>ENSG00000147257.13_3</td>\n",
       "      <td>23</td>\n",
       "      <td>132669773</td>\n",
       "      <td>133119922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>GPC3_mut</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>ENSG00000147257.13_3</td>\n",
       "      <td>23</td>\n",
       "      <td>132669773</td>\n",
       "      <td>133119922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>PHF6_amp</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>ENSG00000156531.16_3</td>\n",
       "      <td>23</td>\n",
       "      <td>133507283</td>\n",
       "      <td>133562821</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>540 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant    gene gene_old               gene_id  chrom      start  \\\n",
       "69     MTOR_mut    MTOR     MTOR  ENSG00000198793.12_4      1   11166592   \n",
       "156    MTOR_amp    MTOR     MTOR  ENSG00000198793.12_4      1   11166592   \n",
       "195    SDHB_amp    SDHB     SDHB   ENSG00000117118.9_2      1   17345217   \n",
       "188     ID3_amp     ID3      ID3   ENSG00000117318.9_3      1   23884417   \n",
       "109  ARID1A_mut  ARID1A   ARID1A  ENSG00000117713.20_5      1   27019727   \n",
       "..          ...     ...      ...                   ...    ...        ...   \n",
       "272  BCORL1_mut  BCORL1   BCORL1  ENSG00000085185.15_5     23  129115083   \n",
       "339  BCORL1_amp  BCORL1   BCORL1  ENSG00000085185.15_5     23  129115083   \n",
       "312    GPC3_amp    GPC3     GPC3  ENSG00000147257.13_3     23  132669773   \n",
       "428    GPC3_mut    GPC3     GPC3  ENSG00000147257.13_3     23  132669773   \n",
       "394    PHF6_amp    PHF6     PHF6  ENSG00000156531.16_3     23  133507283   \n",
       "\n",
       "           end  \n",
       "69    11322564  \n",
       "156   11322564  \n",
       "195   17380665  \n",
       "188   23885992  \n",
       "109   27108595  \n",
       "..         ...  \n",
       "272  129192058  \n",
       "339  129192058  \n",
       "312  133119922  \n",
       "428  133119922  \n",
       "394  133562821  \n",
       "\n",
       "[540 rows x 7 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_variants = results_df[results_df['p_adj'] < 0.1][['variant']].drop_duplicates()\n",
    "sig_variants['gene'] = sig_variants['variant'].str.split('_').str[0]\n",
    "\n",
    "# symbol mismatches\n",
    "sig_variants['gene_old'] = sig_variants['gene']\n",
    "sig_variants.loc[sig_variants['gene'] == 'FAM46C', 'gene'] = 'TENT5C'\n",
    "sig_variants.loc[sig_variants['gene'] == 'PARK2', 'gene'] = 'PRKN'\n",
    "\n",
    "sig_variants = sig_variants.merge(bed, on = 'gene', how = 'left')\n",
    "\n",
    "assert len(sig_variants[sig_variants['chrom'].isnull()]) == 0\n",
    "\n",
    "sig_variants['chrom'] = sig_variants['chrom'].str.replace('chr', '').replace('X', '23').astype(int)\n",
    "\n",
    "sig_variants.sort_values(['chrom', 'start'], inplace = True)\n",
    "\n",
    "sig_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d0f9e94e-0ab2-40cc-9d6a-4aeab8531910",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "217 184 139\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    sum(sig_variants['variant'].str.contains('_amp')),\n",
    "    sum(sig_variants['variant'].str.contains('_del')),\n",
    "    sum(sig_variants['variant'].str.contains('_mut')),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd6162ea-ce20-463e-a8bb-b5a2c8c4c292",
   "metadata": {},
   "source": [
    "Combine amps and dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6907e483-01c9-48f5-b445-98b275433b26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTOR_mut</th>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <th>ID3_cnv</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>MYCL_mut</th>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <th>MPL_cnv</th>\n",
       "      <th>MUTYH_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>GPC3_mut</th>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <th>PHF6_cnv</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 383 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 MTOR_mut  MTOR_cnv  SDHB_cnv  ID3_cnv  ARID1A_mut  \\\n",
       "sample                                                               \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0266-TP       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0267-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0267-TP       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0268-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "\n",
       "                 ARID1A_cnv  MYCL_mut  MYCL_cnv  MPL_cnv  MUTYH_cnv  ...  \\\n",
       "sample                                                               ...   \n",
       "GLSS-19-0266-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0266-TP         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-TP        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0268-R1         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "\n",
       "                 GPC3_mut  GPC3_cnv  PHF6_cnv  k = 2  k = 3  k = 4  k = 5  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0      1      1      1      1   \n",
       "GLSS-19-0266-TP       0.0       0.0       0.0      2      2      2      2   \n",
       "GLSS-19-0267-R1       0.0       0.0       0.0      1      1      1      1   \n",
       "GLSS-19-0267-TP       0.0       0.0       0.0      2      2      2      3   \n",
       "GLSS-19-0268-R1       0.0       0.0       0.0      2      3      3      4   \n",
       "\n",
       "                 k = 6  k = 7  k = 8  \n",
       "sample                                \n",
       "GLSS-19-0266-R1      1      1      1  \n",
       "GLSS-19-0266-TP      2      2      2  \n",
       "GLSS-19-0267-R1      1      1      1  \n",
       "GLSS-19-0267-TP      3      3      3  \n",
       "GLSS-19-0268-R1      4      3      3  \n",
       "\n",
       "[5 rows x 383 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix_sig = matrix.T.loc[:, sig_variants['variant']]\n",
    "new_df = {}\n",
    "\n",
    "for gene in sig_variants['gene_old']:\n",
    "    if f'{gene}_mut' in matrix_sig:\n",
    "        new_df[f'{gene}_mut'] = matrix_sig[f'{gene}_mut']\n",
    "        \n",
    "    if f'{gene}_amp' in matrix_sig and f'{gene}_del' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = matrix_sig[f'{gene}_amp'] + (-1 * matrix_sig[f'{gene}_del'])\n",
    "    elif f'{gene}_del' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = -1 * matrix_sig[f'{gene}_del']\n",
    "    elif f'{gene}_amp' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = matrix_sig[f'{gene}_amp']\n",
    "\n",
    "matrix_sig = pd.DataFrame.from_dict(new_df)\n",
    "df_sig = matrix_sig.join(memberships)\n",
    "df_sig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "2df67b8b-7005-455f-ae18-18d2a80273ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_cnv</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID3_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATRX_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STAG2_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORL1_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHF6_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chrom\n",
       "gene_cnv         \n",
       "MTOR_cnv        1\n",
       "SDHB_cnv        1\n",
       "ID3_cnv         1\n",
       "ARID1A_cnv      1\n",
       "MYCL_cnv        1\n",
       "...           ...\n",
       "ATRX_cnv       23\n",
       "STAG2_cnv      23\n",
       "BCORL1_cnv     23\n",
       "GPC3_cnv       23\n",
       "PHF6_cnv       23\n",
       "\n",
       "[237 rows x 1 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_chrom_df = sig_variants[['gene_old', 'chrom']].drop_duplicates()\n",
    "cnv_chrom_df['gene_cnv'] = cnv_chrom_df['gene_old'] + '_cnv'\n",
    "cnv_chrom_df = cnv_chrom_df[cnv_chrom_df['gene_cnv'].isin(list(df_sig))]\n",
    "cnv_chrom_df.drop(columns = 'gene_old', inplace = True)\n",
    "cnv_chrom_df.set_index('gene_cnv', inplace = True)\n",
    "cnv_chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0b1b4efb-ae12-4644-9355-41fb98a55072",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_old</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MTOR_mut</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MTOR_amp</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>SDHB_amp</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>ID3_amp</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ENSG00000117318.9_3</td>\n",
       "      <td>1</td>\n",
       "      <td>23884417</td>\n",
       "      <td>23885992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ARID1A_mut</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ENSG00000117713.20_5</td>\n",
       "      <td>1</td>\n",
       "      <td>27019727</td>\n",
       "      <td>27108595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant    gene gene_old               gene_id  chrom     start  \\\n",
       "69     MTOR_mut    MTOR     MTOR  ENSG00000198793.12_4      1  11166592   \n",
       "156    MTOR_amp    MTOR     MTOR  ENSG00000198793.12_4      1  11166592   \n",
       "195    SDHB_amp    SDHB     SDHB   ENSG00000117118.9_2      1  17345217   \n",
       "188     ID3_amp     ID3      ID3   ENSG00000117318.9_3      1  23884417   \n",
       "109  ARID1A_mut  ARID1A   ARID1A  ENSG00000117713.20_5      1  27019727   \n",
       "\n",
       "          end  \n",
       "69   11322564  \n",
       "156  11322564  \n",
       "195  17380665  \n",
       "188  23885992  \n",
       "109  27108595  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "eaf9db51-fc4b-4308-987d-02f13cd466c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('int64')"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_chrom_df[\"chrom\"].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bc78f3b3-d4ef-4415-82b3-9557792489e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: grid\n",
       "========================================\n",
       "ComplexHeatmap version 2.16.0\n",
       "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
       "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
       "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
       "\n",
       "If you use it in published research, please cite either one:\n",
       "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
       "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
       "    genomic data. Bioinformatics 2016.\n",
       "\n",
       "\n",
       "The new InteractiveComplexHeatmap package can directly export static \n",
       "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
       "\n",
       "This message can be suppressed by:\n",
       "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
       "========================================\n",
       "\n",
       "========================================\n",
       "circlize version 0.4.16\n",
       "CRAN page: https://cran.r-project.org/package=circlize\n",
       "Github page: https://github.com/jokergoo/circlize\n",
       "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
       "\n",
       "If you use it in published research, please cite:\n",
       "Gu, Z. circlize implements and enhances circular visualization\n",
       "  in R. Bioinformatics 2014.\n",
       "\n",
       "This message can be suppressed by:\n",
       "  suppressPackageStartupMessages(library(circlize))\n",
       "========================================\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i df_sig -i cnv_chrom_df -w 3000 -h 1200\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(RColorBrewer)\n",
    "\n",
    "df_sig <- df_sig[order(df_sig$`k = 6`),]\n",
    "\n",
    "muts <- as.matrix(df_sig[, grep('_mut', names(df_sig))])\n",
    "cnvs <- as.matrix(df_sig[, grep('_cnv', names(df_sig))])\n",
    "\n",
    "clusters <- df_sig[, grep('=', names(df_sig))]\n",
    "\n",
    "cluster_colors <- brewer.pal(n = 8, name = 'Dark2')\n",
    "\n",
    "chrom_colors <- rep(c(cluster_colors[6], cluster_colors[8]), length.out = 23)\n",
    "chrom_colors <- setNames(chrom_colors, 1:23)\n",
    "\n",
    "row_ha <- rowAnnotation(#Cluster_2 = clusters$`k = 2`,\n",
    "                        #Cluster_3 = clusters$`k = 3`,\n",
    "                        #Cluster_4 = clusters$`k = 4`,\n",
    "                        Cluster_5 = clusters$`k = 5`,\n",
    "                        Cluster_6 = clusters$`k = 6`,\n",
    "                        col = list(#Cluster_2 = c(`1` = cluster_colors[1], `2` = cluster_colors[2]),\n",
    "                                   #Cluster_3 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3]),\n",
    "                                   #Cluster_4 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4]),\n",
    "                                   Cluster_5 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5]),\n",
    "                                   Cluster_6 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5], `6` = cluster_colors[6])),\n",
    "                        simple_anno_size = unit(2, 'cm'))\n",
    "\n",
    "col_ha <- columnAnnotation(df = cnv_chrom_df,\n",
    "                          simple_anno_size = unit(2, 'cm'),\n",
    "                          col = list(chrom = chrom_colors))\n",
    "\n",
    "heatmap_muts <- Heatmap(muts, name = 'SNV/indels',\n",
    "                        col = colorRamp2(c(0, 2, 4), c('white', 'grey', 'black')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "heatmap_cnvs <- Heatmap(cnvs, name = 'CNVs', right_annotation = row_ha, top_annotation = col_ha,\n",
    "                        col = colorRamp2(c(-2, 0, 2), c('blue', 'white', 'red')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "full_heatmap <- heatmap_muts + heatmap_cnvs\n",
    "pdf(\"glass_heatmap.pdf\", width=50, height=20)\n",
    "draw(full_heatmap)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7771136b-2b05-43ff-8ee3-b47a0d6e6790",
   "metadata": {},
   "source": [
    "# Assess prognostic value of clusters\n",
    "\n",
    "Adjust for MGMT methylation status?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b2ccc9-3821-4c95-a07d-96ab23016954",
   "metadata": {},
   "source": [
    "## Merge cluster assignments with OS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7e736dca-6a2d-42e6-9910-5a77fefffde4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>case_vital_status</th>\n",
       "      <th>case_overall_survival_mo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GLSS-DF-0002</td>\n",
       "      <td>dead</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GLSS-DF-0003</td>\n",
       "      <td>dead</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLSS-HK-0001</td>\n",
       "      <td>dead</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GLSS-MD-0006</td>\n",
       "      <td>dead</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GLSS-MD-0002</td>\n",
       "      <td>dead</td>\n",
       "      <td>262.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   case_barcode case_vital_status  case_overall_survival_mo\n",
       "0  GLSS-DF-0002              dead                      11.0\n",
       "1  GLSS-DF-0003              dead                       9.0\n",
       "2  GLSS-HK-0001              dead                      25.0\n",
       "3  GLSS-MD-0006              dead                      93.0\n",
       "4  GLSS-MD-0002              dead                     262.0"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_cases_df = pd.read_csv('../data/glass_clinical_cases_table.tsv', sep = '\\t',\n",
    "                            usecols = ['case_barcode', 'case_overall_survival_mo', 'case_vital_status'])\n",
    "glass_cases_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6f140c05-f4f8-4cfe-a9a7-8b4a53d009d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_sample_memberships = glass_samples_df.join(memberships, how=\"inner\")#.to_csv(\"glass_sample_cluster_memberships.tsv\", sep=\"\\t\")\n",
    "glass_sample_memberships.index.name = \"sample_id\"\n",
    "glass_sample_memberships.to_csv(\"glass_sample_cluster_memberships.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "35638595-2eb5-4225-b094-d714ac6ee179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NaN    272\n",
       "U      128\n",
       "M       78\n",
       "Name: mgmt_methylation, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_samples_df['mgmt_methylation'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "31eaac00-eff7-434d-b393-ef103f478b08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-1389-R1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-1389-TP</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-4065-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-4065-TP</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-FG-5963-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8\n",
       "Name                                                            \n",
       "GLSS-19-0266-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0266-TP      2      2      2      2      2      2      2\n",
       "GLSS-19-0267-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0267-TP      2      2      2      3      3      3      3\n",
       "GLSS-19-0268-R1      2      3      3      4      4      3      3\n",
       "...                ...    ...    ...    ...    ...    ...    ...\n",
       "TCGA-19-1389-R1      2      2      2      3      3      3      3\n",
       "TCGA-19-1389-TP      1      1      4      5      6      5      4\n",
       "TCGA-19-4065-R1      1      1      1      1      1      5      4\n",
       "TCGA-19-4065-TP      1      1      4      5      1      5      4\n",
       "TCGA-FG-5963-R1      1      1      4      5      3      3      3\n",
       "\n",
       "[365 rows x 7 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "d0f16b2b-3998-4a18-9be4-ac4ebdfd9e73",
   "metadata": {},
   "outputs": [],
   "source": [
    "memberships_with_os = memberships.\\\n",
    "    join(glass_samples_df[['case_barcode', 'surgical_interval_mo', 'mgmt_methylation']], how = 'left').\\\n",
    "    reset_index(names = 'sample_barcode').\\\n",
    "    merge(glass_cases_df, on = 'case_barcode', how = 'left').\\\n",
    "    replace({'dead': True, 'alive': False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c692eb0b-c6fd-48ee-a6d5-728459a73847",
   "metadata": {},
   "outputs": [],
   "source": [
    "memberships_with_os = memberships_with_os[memberships_with_os['case_overall_survival_mo'].notnull() &\n",
    "                                          memberships_with_os['surgical_interval_mo'].notnull()]\n",
    "\n",
    "# THIS ANALYSIS APPLIES THE CORRECT NOTION OF SURVIVAL FOR THE GLASS COHORT\n",
    "memberships_with_os['os_harmonized'] = np.where(\n",
    "    memberships_with_os['sample_barcode'].str.endswith('-TP'),\n",
    "    memberships_with_os['case_overall_survival_mo'],\n",
    "    memberships_with_os['case_overall_survival_mo'] - memberships_with_os['surgical_interval_mo'],\n",
    ")\n",
    "\n",
    "memberships_with_os['setting'] = np.where(\n",
    "    memberships_with_os['sample_barcode'].str.endswith('-TP'),\n",
    "    'Newly diagnosed',\n",
    "    'Recurrent'\n",
    ")\n",
    "\n",
    "memberships_with_mgmt = memberships_with_os[memberships_with_os[\"mgmt_methylation\"].notnull()].copy()\n",
    "memberships_with_mgmt['mgmt_methylation'] = np.where(\n",
    "    memberships_with_mgmt['mgmt_methylation'] == 'M', 1.0, 0.0\n",
    ")\n",
    "\n",
    "#memberships_with_os = memberships_with_os.drop(columns=[\"mgmt_methylation\"])\n",
    "#memberships_with_os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "7cbf3ee7-73e7-4074-9486-07b52ba1d903",
   "metadata": {},
   "outputs": [],
   "source": [
    "memberships_with_os.to_csv(\"../results/glass_sample_w_cluster_assignment.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57cc77ec-e874-45f8-af63-132b1cc2c63a",
   "metadata": {},
   "source": [
    "# k = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f44dbc54-340c-49e6-a260-f8c570445e70",
   "metadata": {},
   "outputs": [],
   "source": [
    "for clust in range(1, 1+5):\n",
    "    memberships_with_os[f\"cluster_{clust}\"] = memberships_with_os[\"k = 5\"].map(lambda x: x == clust).astype(int)\n",
    "    memberships_with_mgmt[f\"cluster_{clust}\"] = memberships_with_mgmt[\"k = 5\"].map(lambda x: x == clust).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21ca1095-05ed-4e42-8f62-41811fc67150",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 5`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_5_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_5_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_5_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "#print(summary(model_5_combined))\n",
    "#print(summary(model_5_nd))\n",
    "#print(summary(model_5_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c53969bf-1e17-4fd1-811b-ebf9d52f6c29",
   "metadata": {},
   "source": [
    "## with MGMT methylation status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4206029-11e6-4094-a400-f02ee159f766",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i memberships_with_mgmt\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_mgmt %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 5`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_5_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_5_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_5_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "\n",
    "#print(summary(model_5_combined))\n",
    "#print(summary(model_5_nd))\n",
    "#print(summary(model_5_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8ceeb91-32ac-496a-a4b7-63a75ec8fafa",
   "metadata": {},
   "source": [
    "## with primary/recurrent setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45758307-c57a-49c1-88d8-74191b692b34",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 5`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_5_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_5_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_5_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + setting,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "#print(summary(model_5_combined))\n",
    "#print(summary(model_5_nd))\n",
    "#print(summary(model_5_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec110c60-9278-4178-81c4-04e400ae6c30",
   "metadata": {},
   "source": [
    "## k = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfa8c236-2abf-4586-b832-d9758c2e5879",
   "metadata": {},
   "outputs": [],
   "source": [
    "for clust in range(1, 1+6):\n",
    "    memberships_with_os[f\"cluster_{clust}\"] = memberships_with_os[\"k = 6\"].map(lambda x: x == clust).astype(int)\n",
    "    memberships_with_mgmt[f\"cluster_{clust}\"] = memberships_with_mgmt[\"k = 6\"].map(lambda x: x == clust).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd3eee7-6ef5-4fb5-9fb6-5e2a797597b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 6`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_6_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_6_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_6_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5,\n",
    "                      data = memberships_with_os)\n",
    "univariate_6 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_6,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fbde3d2-edda-4731-b21e-ad145422b44a",
   "metadata": {},
   "source": [
    "## With MGMT methylation status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793eb381-671b-4a07-a26a-ba6bc11299dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "memberships_with_os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e820e48d-2d35-4ccf-b794-4792d834b052",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i memberships_with_mgmt\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_mgmt %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 6`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_6_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_6_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_6_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_6 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_6 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e31284a0-eb4c-4161-b6b2-03b8336e7e42",
   "metadata": {},
   "source": [
    "## With primary/recurrent setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c2944d-0f97-4c38-8e33-f6d9bf055539",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 6`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_6_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_6_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_6_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_6 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_6 + setting,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "521976f5-67be-4759-ae23-8c6ecf35459f",
   "metadata": {},
   "source": [
    "# Find a performant classifier\n",
    "Borrowing some code from David's train_classifier.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc0a451-73ec-4040-acea-a61f80f3b601",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.neighbors import KNeighborsClassifier, NearestCentroid\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.linear_model import LogisticRegression, SGDClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.model_selection import GridSearchCV, cross_validate, cross_val_score, StratifiedKFold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc64daf5-9b22-4443-b57f-d6fa5e39a508",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdf38170-eb92-4f90-8b7e-9b05df8b1459",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = matrix_sig\n",
    "y = df_sig[f'k = {k}']\n",
    "\n",
    "counts = y.value_counts()\n",
    "dist = counts / counts.sum()\n",
    "\n",
    "print('Class distribution')\n",
    "print(counts)\n",
    "print(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa16a03-62c9-4f9b-ba1f-ff450c56757e",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "knn = KNeighborsClassifier()\n",
    "cent = NearestCentroid()\n",
    "rf = RandomForestClassifier(n_estimators = 1000)\n",
    "lr = LogisticRegression(solver = 'saga', max_iter = 20000)\n",
    "sgd = SGDClassifier(max_iter = 20000)\n",
    "\n",
    "n_folds = 5\n",
    "\n",
    "p_grid = [\n",
    "    {\n",
    "        'classifier': [knn],\n",
    "        'classifier__n_neighbors': [1,2,3,4,5,6,7,8],\n",
    "        'classifier__weights': ['uniform', 'distance']\n",
    "    },\n",
    "    {\n",
    "        'classifier': [cent],\n",
    "        'classifier__metric': ['euclidean', 'manhattan']\n",
    "    },\n",
    "    {\n",
    "        'classifier': [rf],\n",
    "        'classifier__criterion': ['gini', 'entropy', 'log_loss']\n",
    "    },\n",
    "    {\n",
    "        'classifier': [lr],\n",
    "        'classifier__penalty': ['l1', 'l2'],\n",
    "        'classifier__C': [1e-1, 1, 10, 20]\n",
    "    },\n",
    "    {\n",
    "        'classifier': [sgd],\n",
    "#        'classifier__loss': ['hinge', 'log_loss'],\n",
    "        'classifier__loss': ['log_loss'],\n",
    "        'classifier__penalty': ['l1', 'l2'],\n",
    "        'classifier__alpha': [1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 10]\n",
    "    }\n",
    "]\n",
    "\n",
    "pipeline = Pipeline([('classifier', rf)])\n",
    "\n",
    "gscv = GridSearchCV(pipeline, p_grid, scoring = 'accuracy', n_jobs = 8, cv = n_folds, verbose = 1)\n",
    "\n",
    "scores = cross_validate(gscv, X, y, cv = n_folds)\n",
    "\n",
    "best_model = gscv.fit(X, y).best_estimator_.named_steps['classifier']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa9e38ce-b8a3-488e-b22a-ae7eaa8e8361",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gscv.best_params_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f637e700-cf21-4172-b163-f65413805e44",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39c1ae9f-4509-485c-a172-de380eac7bf9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cross_val_score(best_model, X, y, cv = 5).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67a058bf-8021-4291-a4b1-75a4bf3cf939",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_model.coef_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d775a1a0-efa3-47fe-8894-143a6178cdb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c56d69af-03b3-4b25-8e6b-e3e2bfaa62b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "best_model.fit(X, y)\n",
    "\n",
    "coeffs = best_model.coef_\n",
    "coeffs = pd.DataFrame(coeffs).T\n",
    "coeffs.index = [c for c in df_sig.columns if not c.startswith(\"k =\")]\n",
    "coeffs.sort_values(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83344ada-a467-4835-90ed-9edd2ce45b27",
   "metadata": {},
   "source": [
    "# Does PCA tell us anything?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c21255f2-3f4d-4edb-b45e-db4565103d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "\n",
    "pca = PCA(n_components = 5)\n",
    "X_r = pca.fit_transform(X)\n",
    "# X_icb_r = pca.transform(op_matrix_sig)\n",
    "\n",
    "print(pca.explained_variance_ratio_)\n",
    "\n",
    "colors = ['red', 'blue', 'orange', 'brown', 'green', 'gray']\n",
    "# plt.scatter(X_icb_r[:,0], X_icb_r[:,1], s = 40, color = 'gray', alpha = 0.8, lw = 1, label = 'ICB')\n",
    "\n",
    "for i, color in enumerate(colors):\n",
    "    plt.scatter(X_r[y == i + 1, 0], X_r[y == i + 1, 1], s = 20, color = color, alpha = 0.5, lw = 1, label = i + 1)\n",
    "plt.legend(loc = 'best')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.xscale('symlog')\n",
    "plt.yscale('symlog')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db7e40e-c8c9-4181-99b2-77557d5f560e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3', 'PC4', 'PC5'], index=list(X)).sort_values('PC2').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5c90e1e-2ff1-446f-9605-ff1fb57d0892",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3', 'PC4', 'PC5'], index=list(X)).sort_values('PC2').tail(10)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1e2105a6-eca9-4b77-9dd0-3b73aca0f956",
   "metadata": {},
   "source": [
    "# Prepare the ICB OncoPanel data\n",
    "\n",
    "Preprocess the data identically to GLASS... to the extent possible. Some details:\n",
    "- [x] Remove likely germline variants from the ICB OncoPanel data before classification using the gnomAD frequency filter we used for the Cox regressions.\n",
    "- [ ] Add in WES mutational and CNV data for the ~one dozen ICB patients with WES and no OncoPanel\n",
    "    - This would involve preprocessing akin to that done in `aggregate_muts_for_pathways.ipynb`, just without the filters for GISTIC peaks, OncoKB, and CADD\n",
    "- [x] Determine how to handle patients in ICB with multiple OncoPanel / WES samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69cbdfe-01b8-448e-9574-4c9e72e6c51a",
   "metadata": {},
   "source": [
    "### Load ICB cohort sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4449ed75-ff09-47ac-baf6-b1e49a442fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_path = '../data/sample_cohort_sheet.tsv'\n",
    "pt_path = '../data/participant_cohort_sheet.tsv'\n",
    "\n",
    "icb_sample_df = pd.read_csv(clinical_path, sep=\"\\t\")\n",
    "icb_sample_df[\"sample_id_legacy\"] = icb_sample_df[\"sample_id_legacy\"].astype(str)\n",
    "\n",
    "icb_pt_df = pd.read_csv(pt_path, sep=\"\\t\")\n",
    "# icb_clinical = pd.read_csv(clinical_path, sep = '\\t')\n",
    "# icb_pt = pd.read_csv(pt_path, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77310d88-0355-46bd-959d-70917afcc3c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_sample_df[(icb_sample_df[\"usable_wes\"] | icb_sample_df[\"usable_oncopanel\"]) & (icb_sample_df[\"tumor_normal\"] == \"tumor\") & (icb_sample_df[\"pre_post\"] == \"Pre\")][\"participant_id\"].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c6106d6-49bf-4a52-8405-fe946a2e5ae3",
   "metadata": {},
   "source": [
    "### Load OncoPanel variants\n",
    "\n",
    "...and clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90f2079c-98b1-475e-9838-5540929b30bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "op_mutations_path = '../data/oncopanel_mutations.csv' \n",
    "op_cnvs_path = '../data/oncopanel_cnvs.csv'\n",
    "op_clinical_path = '../data/oncopanel_clinical.xlsx'\n",
    "\n",
    "op_mut_df = pd.read_csv(op_mutations_path, sep = '\\t')\n",
    "op_mut_df[\"id_uniquesample\"] = op_mut_df[\"id_uniquesample\"].astype(str)\n",
    "\n",
    "op_cnv_df = pd.read_csv(op_cnvs_path, sep = '\\t')\n",
    "op_cnv_df[\"id_uniquesample\"] = op_cnv_df[\"id_uniquesample\"].astype(str)\n",
    "\n",
    "op_sample_df = pd.read_excel(op_clinical_path)\n",
    "op_sample_df[\"id_uniquesample\"] = op_sample_df[\"id_uniquesample\"].astype(str)\n",
    "op_sample_df[\"id_drspatient\"] = op_sample_df[\"id_drspatient\"].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8a5ec2-3c01-4039-8034-1cc291b5bb79",
   "metadata": {},
   "source": [
    "Filter to keep variants belonging to known ICB samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dcdb9af-eb8a-41d3-93a7-8b46bd3c80f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Harmonize sample IDs in the OncoPanel data\n",
    "op_sample_df['sample_id_merge'] = np.where(op_sample_df['id_broad_sample'].notnull(),\n",
    "                                           op_sample_df['id_broad_sample'],\n",
    "                                           op_sample_df['id_uniquesample']).astype(str)\n",
    "\n",
    "op_sample_df.replace({'GBM.ICB-3-14.637.Post.FFbank': 'GBM.ICB-3-14.233.Pre'}, inplace = True)\n",
    "\n",
    "icb_op_samples = icb_sample_df[icb_sample_df['usable_oncopanel']]['sample_id_legacy'].astype(str)\n",
    "op_sample_df = op_sample_df[op_sample_df['sample_id_merge'].isin(icb_op_samples)]\n",
    "\n",
    "len(op_sample_df) == icb_sample_df['usable_oncopanel'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64ba3adf-efa7-44f6-9e55-a8508e5d144d",
   "metadata": {},
   "outputs": [],
   "source": [
    "kept_op_samples = op_sample_df['id_uniquesample'].tolist()\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df['id_uniquesample'].isin(kept_op_samples)]\n",
    "op_cnv_df = op_cnv_df[op_cnv_df['id_uniquesample'].isin(kept_op_samples)]\n",
    "\n",
    "len(kept_op_samples)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e3b589aa-5b41-482e-90b5-a578185b3433",
   "metadata": {},
   "source": [
    "### Filter mutations\n",
    "\n",
    "Select mutations by several criteria:\n",
    "* Filter mutations by VAF\n",
    "* Keep only variants belonging to genes with exons covered by all versions of OncoPanel\n",
    "* Keep only non-synonymous mutations\n",
    "* Filter likely germline events by gnomAD frequency\n",
    "* Keep mutations belonging to the latest pre-ICB timepoint for each participant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0db565c-62f3-4177-99a8-240cc7de9323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants above the VAF threshold (0.1)\n",
    "op_mut_df = op_mut_df[op_mut_df['ALLELE_FRACTION'] >= 100*VAF_THRESHOLD]\n",
    "\n",
    "# Keep only genes with exons covered by all versions of OncoPanel\n",
    "op_mut_df = op_mut_df[op_mut_df['HARMONIZED_HUGO_GENE_NAME'].isin(oncopanel_genes['Gene'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22010afe-b10a-4857-88df-fd67930e5cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Non-synonymous mutations\n",
    "nonsyn_classifications = ['Missense',\n",
    "                          'Frame_Shift_Del',\n",
    "                          'Nonsense',\n",
    "                          'Splice_Region',\n",
    "                          'Frame_Shift_Ins',\n",
    "                          'Splice_Site',\n",
    "                          'In_Frame_Ins',\n",
    "                          'In_Frame_Del',\n",
    "                          'Translation_Start_Site']\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df['HARMONIZED_VARIANT_CLASS'].isin(nonsyn_classifications)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6039fbe6-49f7-4eea-b8b9-9e65dd8cae76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gnomAD germline mutation filtering\n",
    "gnomad_threshold = 2e-4\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df[\"MAX_GNOMAD_FREQUENCY\"].fillna(0.0).astype(float).lt(gnomad_threshold)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1883ad81-6a8c-4643-a754-5bef0e6cfa02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Latest pre-ICB timepoint for each participant\n",
    "def keep_latest_preicb_timepoint_variants(variant_df, sample_df, sample_id_col=\"sample_id_merge\"):\n",
    "    variant_df = variant_df.join(sample_df[[\"sample_id_legacy\", \"participant_id_legacy\", \"collection_date_dfd\", \"pre_post\"]].set_index(\"sample_id_legacy\"), on=sample_id_col)\n",
    "    variant_df = variant_df[variant_df[\"pre_post\"] == \"Pre\"]\n",
    "    latest_timepoints = variant_df[[\"participant_id_legacy\", \"collection_date_dfd\"]].groupby(\"participant_id_legacy\").max()\n",
    "    latest_timepoints.columns = [\"latest_timepoint\"]\n",
    "    variant_df = variant_df.join(latest_timepoints, on=\"participant_id_legacy\")\n",
    "\n",
    "    variant_df = variant_df[variant_df[\"collection_date_dfd\"] == variant_df[\"latest_timepoint\"]]\n",
    "    return variant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f815699a-877d-410c-a59f-2dd36f497798",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_mut_df = op_mut_df.join(op_sample_df[[\"id_uniquesample\", \"sample_id_merge\"]].set_index(\"id_uniquesample\"), on=\"id_uniquesample\")\n",
    "op_mut_df = keep_latest_preicb_timepoint_variants(op_mut_df, icb_sample_df)\n",
    "\n",
    "op_mut_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3b770f8-e853-4051-b016-5e872952707e",
   "metadata": {},
   "source": [
    "### Construct pt-by-variant matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec429b49-5e0e-42d5-8980-32520dc2b1e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_mut_df['gene_mut'] = op_mut_df['HARMONIZED_HUGO_GENE_NAME'] + '_mut'\n",
    "\n",
    "op_variants_matrix = op_mut_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# scoring: sequence variants = 2\n",
    "op_variants_matrix *= 2\n",
    "\n",
    "op_variants_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad545274-6ca3-4204-8bd5-ad7d6590f6e4",
   "metadata": {},
   "source": [
    "### Filter CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b601d63-6bd2-46b9-8a81-65cf910a0ee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_cnv_df = op_cnv_df[op_cnv_df['HARMONIZED_HUGO_GENE_NAME'].isin(oncopanel_genes['Gene'])]\n",
    "\n",
    "op_cnv_df['hlvl_call'] = op_cnv_df['HARMONIZED_VARIANT_CLASS'].map(\n",
    "    {'2DEL': -2, '1DEL': -1, 'LA': 1, 'HA': 2}\n",
    ")\n",
    "\n",
    "op_cnv_df['gene_cnv'] = op_cnv_df['HARMONIZED_HUGO_GENE_NAME'] + '_cnv'\n",
    "\n",
    "op_cnv_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49f77769-172e-4d3f-85d3-159db637f5fe",
   "metadata": {},
   "source": [
    "Ensure no duplicate CNVs in any sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c25b8108-9e92-46a2-97d8-9bf8dcefdc7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_cnv_df.groupby(['id_uniquesample', 'HARMONIZED_HUGO_GENE_NAME']).size().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bd3f2f5-1af0-4c44-8a26-055b272f5ae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_cnv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55c63f1b-a354-4555-9fe1-ee9096c348b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_cnv_df = op_cnv_df.join(op_sample_df[[\"id_uniquesample\", \"sample_id_merge\"]].set_index(\"id_uniquesample\"), on=\"id_uniquesample\")\n",
    "op_cnv_df = keep_latest_preicb_timepoint_variants(op_cnv_df, icb_sample_df, sample_id_col=\"sample_id_merge\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29912f4c-a459-4201-b192-8695aea1bb45",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_cnvs_matrix = op_cnv_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_cnv', values = 'hlvl_call', aggfunc = 'sum', fill_value = 0)\n",
    "\n",
    "op_cnvs_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4b2a6aa-02e2-43ae-9770-eaae83e33f57",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_variant_median_rates(op_cnv_df, pt_col=\"participant_id_legacy\", variant_col=\"hlvl_call\")\n",
    "plt.title(\"Median CNV counts per pt in OncoPanel data\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b2800fe-6bdb-4a47-b7e1-09ef645fc540",
   "metadata": {},
   "source": [
    "# Prepare the ICB WES data\n",
    "\n",
    "Get the participants who have pre-ICB WES tumor data but NOT OncoPanel call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdb7439e-01da-47ca-b9cd-13850d589104",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_sample_df = icb_sample_df[(icb_sample_df[\"pre_post\"] ==  \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\")]\n",
    "pt_usable_data = pre_sample_df[[\"participant_id_legacy\", \"usable_wes\", \"usable_oncopanel\"]].groupby(\"participant_id_legacy\").any()\n",
    "pt_usable_data = pt_usable_data[pt_usable_data[\"usable_wes\"] & (~pt_usable_data[\"usable_oncopanel\"])]\n",
    "\n",
    "wes_pt_df = icb_pt_df[icb_pt_df[\"participant_id_legacy\"].isin(pt_usable_data.index)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596f1605-7297-426a-98c7-dd44a854acf2",
   "metadata": {},
   "source": [
    "### Load and process ICB WES sequence variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9688a22b-1975-4bd6-9403-4d9f364831c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "wes_mut_df = pd.read_pickle('../data/icb_wes_mut_ccfs_maf.pickle')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f23ca0c2-dcd7-4548-bc90-54e95f592660",
   "metadata": {},
   "source": [
    "Filter WES sequence variants by various criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13c81242-ebf2-4787-8527-655c14f08d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants for pts without OncoPanel data\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"Patient_ID\"].isin(wes_pt_df[\"participant_id_legacy\"])]\n",
    "\n",
    "# Correct some sample IDs that erroneously end with \"_pair\"\n",
    "wes_mut_df[\"Sample_ID\"] = wes_mut_df[\"Sample_ID\"].map(lambda x: x[:-5] if x.endswith(\"_pair\") else x)\n",
    "\n",
    "# Filter by CCF\n",
    "cluster_ccf_threshold = 0.1\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"clust_ccf_mean\"] > cluster_ccf_threshold]\n",
    "\n",
    "# Restrict to genes covered by OncoPanel\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"Hugo_Symbol\"].isin(oncopanel_genes[\"Gene\"])]\n",
    "\n",
    "# Remove silent and non-coding mutations. \n",
    "# This is consistent with the filtering of GLASS variants\n",
    "wes_mut_df = wes_mut_df[~wes_mut_df[\"variant_type\"].isin([\"silent\", \"non_coding\"])]\n",
    "\n",
    "# Keep only variants belonging to their latest pre-ICB timepoint.\n",
    "wes_mut_df = keep_latest_preicb_timepoint_variants(wes_mut_df, pre_sample_df, sample_id_col=\"Sample_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4200ae71-3748-4d01-87c9-150e7d6c63ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_df[\"variant_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e45eb41-967d-4d23-a17a-e4846f3b8838",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_df[\"gene_mut\"] = wes_mut_df[\"Hugo_Symbol\"]+\"_mut\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6c736e7-fe32-4515-bfbf-602afc6f86c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_matrix = wes_mut_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# In this methodology, sequence variants are given weight = 2\n",
    "wes_mut_matrix *= 2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c760f96b-ab72-482e-8fe0-60f6c0492ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3800996-3675-4244-9470-a9560a59295b",
   "metadata": {},
   "source": [
    "### Load and process ICB WES CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982f90c4-3727-404f-aa85-b9088d269910",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df = pd.concat([\n",
    "    pd.read_csv('../data/icb_wes_tnpair_gene_allelic_cn.maf', sep = '\\t'),\n",
    "    pd.read_csv('../data/icb_wes_tonly_gene_allelic_cn.maf', sep = '\\t')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cad9964-a6ef-4c82-b590-718b2a56cd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some helper functions\n",
    "def call_hlvl(cn_col, thresholds=[0.8, 1.8, 2.2, 3.2], values=[-2,-1,0,1,2]): \n",
    "    dig = np.digitize(cn_col.values, thresholds)\n",
    "    hlvl = np.array(values)[dig]\n",
    "    return pd.Series(data=hlvl, index=cn_col.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "644b4b32-63a1-4fe2-8eef-5bda3e4e1c9e",
   "metadata": {},
   "source": [
    "Filter the variants by various criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c9037c7-cf2b-403a-9658-85842481f8f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants belonging to the latest pre-ICB timepoint for each pt\n",
    "wes_cnv_df = keep_latest_preicb_timepoint_variants(wes_cnv_df, icb_sample_df, sample_id_col=\"sample\")\n",
    "\n",
    "# Restrict to ICB pts that only have WES data (and not OncoPanel)\n",
    "wes_cnv_df = wes_cnv_df[wes_cnv_df[\"participant_id_legacy\"].isin(wes_pt_df[\"participant_id_legacy\"])]\n",
    "\n",
    "# Restrict to genes covered by OncoPanel\n",
    "wes_cnv_df = wes_cnv_df[wes_cnv_df[\"gene\"].isin(oncopanel_genes[\"Gene\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e6916e7-1cc0-411d-8230-d91d507e4dc8",
   "metadata": {},
   "source": [
    "Annotate gene-level copy number gains, losses, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad22c0b-1572-4cbd-8bfc-66730b4bb605",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df[\"hlvl_call\"] = call_hlvl(wes_cnv_df[\"corrected_total_cn\"])\n",
    "\n",
    "# Use ABSOLUTE's homozygous deletions\n",
    "wes_cnv_df.loc[wes_cnv_df[\"HZ\"] == 1, \"hlvl_call\"] = -2\n",
    "\n",
    "# Finally: keep only the actual CNVs\n",
    "wes_cnv_df = wes_cnv_df.loc[wes_cnv_df[\"hlvl_call\"] != 0,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa5055d1-39cf-43ad-821e-a43c65c8580d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df[\"gene_cnv\"] = wes_cnv_df[\"gene\"] + \"_cnv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "651320b3-ae3e-4616-b214-31d0a20b99e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_variant_median_rates(wes_cnv_df, pt_col=\"participant_id_legacy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204cc9d2-fd78-45f7-800b-ac352bde7987",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_matrix = wes_cnv_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_cnv', aggfunc = 'sum', values = \"hlvl_call\", fill_value = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "812517ac-9778-4704-afd7-7f002f722f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55bc7889-e530-4e53-bfd7-b73ad29fa5c4",
   "metadata": {},
   "source": [
    "# Combine all filtered variants for ICB cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd3e94fd-0650-4588-b270-cc14aefbf8f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_matrix = op_variants_matrix.join(op_cnvs_matrix, how = 'outer').fillna(0)\n",
    "print(op_matrix.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dd30925-560e-406a-b46d-08d4cf816dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_matrix = wes_mut_matrix[~wes_mut_matrix.index.isin(op_matrix.index)]\n",
    "icb_wes_matrix = wes_mut_matrix.join(wes_cnv_matrix, how=\"outer\").fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ec53d92-cbe0-49b6-b1f6-49dce8809ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_matrix = pd.concat((op_matrix, icb_wes_matrix)).fillna(0)\n",
    "icb_matrix = icb_matrix.fillna(0)\n",
    "icb_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f947d998-d575-42b9-9751-5c365267133a",
   "metadata": {},
   "source": [
    "Keep only alteration types significantly associated with clusters in GLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19554dd1-5331-435e-bd2f-0b28f9477586",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_alterations = list(matrix_sig)\n",
    "missing = set(glass_alterations) - set(icb_matrix)\n",
    "len(missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a0e9f94-351d-4d14-812b-0ac2f5bf2a67",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for c in missing:\n",
    "    icb_matrix[c] = 0.0\n",
    "\n",
    "icb_matrix = icb_matrix[list(matrix_sig)]\n",
    "\n",
    "print(icb_matrix.shape, matrix_sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c73c84e0-95a3-4d35-8c11-939c937bf136",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_matrix.index.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c55572a-290b-4107-a513-9a97072cf82d",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_samples = icb_sample_df[(icb_sample_df[\"pre_post\"] == \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\") & (icb_sample_df[\"usable_oncopanel\"] | icb_sample_df[\"usable_wes\"])][\"participant_id_legacy\"]\n",
    "#icb_sample_df[(icb_sample_df[\"pre_post\"] == \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\") & icb_sample_df[\"usable_wes\"]][\"participant_id_legacy\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "478d9361-2c13-4130-b080-087eacbe0556",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These pts have no variants that were included in clustering\n",
    "no_variant_pts = list(set(icb_samples) - set(icb_matrix.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2a43f7d-f4d7-407c-87a7-74ac2024a445",
   "metadata": {},
   "outputs": [],
   "source": [
    "for pt in no_variant_pts:\n",
    "    icb_matrix.loc[pt, :] = 0.0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0da39d67-4654-46bf-b549-ca43a864b683",
   "metadata": {},
   "source": [
    "# Classify the ICB participants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08d29fc6-9055-4af7-b8f5-cea53f0f1e92",
   "metadata": {},
   "source": [
    "## Apply the best classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4bbd46d-4153-43c8-9d79-a9164e12b9f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#best_model.fit(X, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a509105d-b3b4-4e64-904d-1c850ca9006b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_labels = [f\"cluster_{c}\" for c in best_model.classes_]\n",
    "\n",
    "icb_pred = best_model.predict_proba(icb_matrix)\n",
    "icb_pred_df = pd.DataFrame(data=icb_pred, index = icb_matrix.index, columns=cluster_labels)\n",
    "icb_pred_df[\"cluster\"] = icb_pred_df.apply(lambda row: cluster_labels[np.argmax(row)], axis=1)\n",
    "\n",
    "icb_pred_df['cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5d09203-1758-48d3-b383-0546cd4869e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_results = icb_pred_df.copy()\n",
    "\n",
    "icb_results = icb_results.merge(icb_pt_df, left_on='participant_id_legacy', right_on=\"participant_id_legacy\")\n",
    "\n",
    "#icb_results = icb_results[icb_results['MGMT methylated Y/N'].notnull()]\n",
    "\n",
    "icb_results['setting'] = icb_results['ICB for Newly diagnosed']\n",
    "icb_results['vital_status'] = icb_results['Deceased'].map({'Deceased': 1, 'Alive': 0})\n",
    "icb_results['mgmt_methylated'] = icb_results['MGMT methylated Y/N'].replace({'Unmethylated': 0.0, 'Methylated': 1.0, 'Partially': 1.0})\n",
    "\n",
    "print(icb_results['participant_id'].nunique()) #, icb_results['sample_id_legacy'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c0a3d5-cfaa-439f-b68a-752844941f08",
   "metadata": {},
   "source": [
    "## Output predicted cluster assignments to TSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59c4c4b6-d749-4554-9514-6c0c72a7a2e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_pt_df = icb_pt_df.merge(icb_pred_df, left_on=\"participant_id_legacy\", right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3752fd1f-b28a-4c60-b1a2-1273a60fccaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_pt_df.to_csv(\"../results/icb_pt_w_cluster_assignments.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e11757a-03bf-48d7-b395-01a9ac4ff7c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_results.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5875261-a890-4431-827f-8bd2f313f1e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_results.to_csv(\"../results/icb_pt_w_cluster_assignments.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8574c80-74d9-4722-b809-5bb683c6bb3b",
   "metadata": {},
   "source": [
    "## Assess prognostic value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f077fbdf-b50a-4670-9eab-8b6dd5577536",
   "metadata": {},
   "source": [
    "## With no covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f1e062-3882-4410-b6da-b189a6eb110b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i icb_results\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- icb_results %>%\n",
    "    mutate(cluster = as.factor(cluster))\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly-diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "# model_6_combined <- coxph(Surv(osicb, vital_status) ~ setting + mgmt_methylated + cluster,\n",
    "#                           data = combined)\n",
    "# model_6_nd <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = nd)\n",
    "# model_6_rc <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(osicb, vital_status) ~ cluster_1,\n",
    "                      data = combined)\n",
    "univariate_2 <- coxph(Surv(osicb, vital_status) ~ cluster_2,\n",
    "                      data = combined)\n",
    "univariate_3 <- coxph(Surv(osicb, vital_status) ~ cluster_3,\n",
    "                      data = combined)\n",
    "univariate_4 <- coxph(Surv(osicb, vital_status) ~ cluster_4,\n",
    "                      data = combined)\n",
    "univariate_5 <- coxph(Surv(osicb, vital_status) ~ cluster_5,\n",
    "                      data = combined)\n",
    "# univariate_6 <- coxph(Surv(osicb, vital_status) ~ cluster_6,\n",
    "#                       data = combined)\n",
    "\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "#print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de3def5b-8377-44e9-b9f1-5c3251d2067f",
   "metadata": {},
   "source": [
    "## With MGMT methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff9ab233-500d-4dee-89b0-ebb082bb9761",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i icb_results\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- icb_results %>%\n",
    "    mutate(cluster = as.factor(cluster))\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly-diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "# model_6_combined <- coxph(Surv(osicb, vital_status) ~ setting + mgmt_methylated + cluster,\n",
    "#                           data = combined)\n",
    "# model_6_nd <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = nd)\n",
    "# model_6_rc <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(osicb, vital_status) ~ cluster_1 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_2 <- coxph(Surv(osicb, vital_status) ~ cluster_2 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_3 <- coxph(Surv(osicb, vital_status) ~ cluster_3 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_4 <- coxph(Surv(osicb, vital_status) ~ cluster_4 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_5 <- coxph(Surv(osicb, vital_status) ~ cluster_5 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "# univariate_6 <- coxph(Surv(osicb, vital_status) ~ cluster_6 + mgmt_methylated,\n",
    "#                       data = combined)\n",
    "\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "#print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77e71e7f-9a53-4129-ba62-503336df7775",
   "metadata": {},
   "source": [
    "## With Newly-diagnosed/recurrent as a covariate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b867e7d-f78e-41f1-921f-f0272df2106a",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i icb_results\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- icb_results %>%\n",
    "    mutate(cluster = as.factor(cluster))\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly-diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "# model_6_combined <- coxph(Surv(osicb, vital_status) ~ setting + mgmt_methylated + cluster,\n",
    "#                           data = combined)\n",
    "# model_6_nd <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = nd)\n",
    "# model_6_rc <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(osicb, vital_status) ~ cluster_1 + setting,\n",
    "                      data = combined)\n",
    "univariate_2 <- coxph(Surv(osicb, vital_status) ~ cluster_2 + setting,\n",
    "                      data = combined)\n",
    "univariate_3 <- coxph(Surv(osicb, vital_status) ~ cluster_3 + setting,\n",
    "                      data = combined)\n",
    "univariate_4 <- coxph(Surv(osicb, vital_status) ~ cluster_4 + setting,\n",
    "                      data = combined)\n",
    "univariate_5 <- coxph(Surv(osicb, vital_status) ~ cluster_5 + setting,\n",
    "                      data = combined)\n",
    "#univariate_6 <- coxph(Surv(osicb, vital_status) ~ cluster_6 + setting,\n",
    "#                      data = combined)\n",
    "\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "#print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a1e240e-3b0f-4f23-a1bd-8f1b446156fc",
   "metadata": {},
   "source": [
    "## Create a heatmap of ICB cohort, grouped by predicted cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "509b1b70-e24a-41ba-a6a5-03989816c074",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "icb_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0869cc10-268f-4e42-b3c5-88734a497bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_pred_df[\"k = 5\"] = icb_pred_df[\"cluster\"].map(lambda x: x.split(\"_\")[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88b27ac7-07c9-4304-a3e0-d25e8758af46",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_matrix = icb_matrix.join(icb_pred_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa9e8ebc-fda6-4d68-8e91-12e1da5e6450",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f30e5a0-67fa-4dd9-a293-c20b1532093c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i icb_matrix -i cnv_chrom_df -w 3000 -h 1200\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(RColorBrewer)\n",
    "\n",
    "#icb_matrix <- icb_matrix[order(icb_matrix$`k = 6`),]\n",
    "icb_matrix <- icb_matrix[order(icb_matrix$`k = 5`),]\n",
    "\n",
    "muts <- as.matrix(icb_matrix[, grep('_mut', names(icb_matrix))])\n",
    "cnvs <- as.matrix(icb_matrix[, grep('_cnv', names(icb_matrix))])\n",
    "\n",
    "clusters <- icb_matrix[, grep('=', names(icb_matrix))]\n",
    "cluster_colors <- brewer.pal(n = 8, name = 'Dark2')\n",
    "\n",
    "chrom_colors <- rep(c(cluster_colors[6], cluster_colors[8]), length.out = 23)\n",
    "chrom_colors <- setNames(chrom_colors, 1:23)\n",
    "\n",
    "row_ha <- rowAnnotation(#Cluster_2 = clusters$`k = 2`,\n",
    "                        #Cluster_3 = clusters$`k = 3`,\n",
    "                        #Cluster_4 = clusters$`k = 4`,\n",
    "                        Cluster_5 = clusters, #clusters$`k = 5`,\n",
    "                        #Cluster_6 = clusters, #clusters$`k = 5`,\n",
    "                        col = list(#Cluster_2 = c(`1` = cluster_colors[1], `2` = cluster_colors[2]),\n",
    "                                   #Cluster_3 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3]),\n",
    "                                   #Cluster_4 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4]),\n",
    "                                   Cluster_5 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5])),\n",
    "                                   #Cluster_6 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5], `6` = cluster_colors[6])),\n",
    "                        simple_anno_size = unit(2, 'cm'))\n",
    "\n",
    "col_ha <- columnAnnotation(df = cnv_chrom_df,\n",
    "                          simple_anno_size = unit(2, 'cm'),\n",
    "                          col = list(chrom = chrom_colors))\n",
    "\n",
    "heatmap_muts <- Heatmap(muts, name = 'SNV/indels',\n",
    "                        col = colorRamp2(c(0, 2, 4), c('white', 'grey', 'black')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "heatmap_cnvs <- Heatmap(cnvs, name = 'CNVs', right_annotation = row_ha, top_annotation = col_ha,\n",
    "                        col = colorRamp2(c(-2, 0, 2), c('blue', 'white', 'red')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "full_heatmap <- heatmap_muts + heatmap_cnvs\n",
    "pdf(\"icb_heatmap.pdf\", width=50, height=10)\n",
    "draw(full_heatmap)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b114d27f-1a66-4cb2-952f-c5298182fd0b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
