{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "327e58cd-d270-4af2-b3be-4199b82e5a5e",
   "metadata": {},
   "source": [
    "# DNA Clustering and Classification\n",
    "\n",
    "This notebook explores whether we can use mutational and copy number information to identify clusters of GLASS samples that overlap with Verhaak transcriptional subtype.\n",
    "\n",
    "The actual consensus clustering is done by the NMFConsensus script from GenePattern here: https://github.com/genepattern/NMFConsensus/blob/develop/nmfconsensus.R, and the clustering parameters are as listed here: gs://fc-secure-3f0f305e-f110-432b-bc9a-c8d96824cea2/jweiss/local_files/results/2024-07-22_GLASS-Verhaak-classifier/GLASS_sample_clustering.params.txt\n",
    "\n",
    "The notebook also trains a classifier to predict membership in these DNA-based clusters and then applies this classifier to the ICB OncoPanel data.\n",
    "\n",
    "This notebook generates the following figures:\n",
    "* The clustering heatmap shown in Fig. 1c\n",
    "* The sankey plot shown in Extended Data Fig. 3d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1ad4df9b-047d-408a-acbf-7fddc296f9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn import metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9527aa9a-803b-4a97-91a1-1b63e67486e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "45b03d54-e19d-47d6-ae00-a6b78cf90767",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dir = '../../results'\n",
    "if not os.path.exists(results_dir):\n",
    "    os.mkdir(results_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e690a97a-9ae1-4557-bbb3-1581b6574563",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load and preprocess GLASS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e68c7b-6a7d-424d-b669-e3b902ea8d4f",
   "metadata": {},
   "source": [
    "## Clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "82324a6c-798f-4b2a-bad9-4fcd54acf3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_surgeries_df = pd.read_csv('../../data/glass_clinical_surgeries_table.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a62849fe-9475-4602-96a8-420cbf344afb",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_adjusted_os = pd.read_csv(\"../../data/sample_clinical_data_GLASS_removenonGBM_removeIDHmut_flagICB.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ab2684b5-4846-46ef-9b80-8f3be5be8207",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/1784923559.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  (glass_surgeries_df[\"histology\"].str.match(\"Glioblastoma|Gliosarcoma\").fillna(False).astype(bool))]#[\"case_barcode\"].nunique()\n"
     ]
    }
   ],
   "source": [
    "glass_surgeries_df = (glass_surgeries_df[(glass_surgeries_df[\"sample_barcode\"].notnull()) & \\\n",
    "                                        (glass_surgeries_df[\"idh_status\"] != \"IDHmut\") & \\\n",
    "                                        (glass_surgeries_df[\"histology\"].str.match(\"Glioblastoma|Gliosarcoma\").fillna(False).astype(bool))]#[\"case_barcode\"].nunique()\n",
    "                     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3174a055-06b1-4b25-b46a-eb961175ef37",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_samples_df = (glass_surgeries_df.merge(glass_adjusted_os[[\"sample_barcode\", \"sample_overall_survival_days\"]], \n",
    "                                             on=\"sample_barcode\", how=\"left\")\n",
    "                    .set_index(\"sample_barcode\")\n",
    "                   )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bc56e85-de94-481a-8728-29c9b8872aa7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## WES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaec0811-e0f2-4d4d-9384-6d89c83f7aad",
   "metadata": {},
   "source": [
    "### Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f04862e1-7178-4640-82cd-da64c18b59b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "glass_variants_passgeno_df = pd.read_csv('../../data/glass_variants_passgeno.csv', sep = ',', low_memory = False)\n",
    "glass_variants_anno_df = pd.read_csv('../../data/glass_variants_anno.csv', sep = ',', low_memory = False)\n",
    "oncopanel_genes = pd.read_csv(os.path.join('../../data/oncopanel_exonic_v123.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6e5ceaec-7a11-4f0a-84b6-8ff79f50053c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create and save \"sample\" column\n",
    "glass_variants_passgeno_df[\"sample\"] = glass_variants_passgeno_df[\"aliquot_barcode\"].str[:15]\n",
    "wes_mut_samples = set(glass_variants_passgeno_df[\"sample\"])\n",
    "\n",
    "# Keep only the variants detected by single-sample MuTect2\n",
    "glass_variants_passgeno_df = glass_variants_passgeno_df[glass_variants_passgeno_df[\"ssm2_pass_call\"] == \"t\"]\n",
    "\n",
    "# Keep only variants above the VAF threshold (0.1)\n",
    "VAF_THRESHOLD = 0.1\n",
    "glass_variants_passgeno_df = glass_variants_passgeno_df[glass_variants_passgeno_df['af'] >= VAF_THRESHOLD]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b6cfe8d-3f95-4f18-b310-cd2a764c8977",
   "metadata": {},
   "source": [
    "Join the variant annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3d8aa671-2b32-4491-85df-482e8d00bbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def append_variant_id(df, colnames=[\"chrom\",\"start\",\"end\",\"alt\"]):\n",
    "    df[\"variant_id\"] = df[colnames[0]].astype(str)\n",
    "    for c in colnames[1:]:\n",
    "        df[\"variant_id\"] = df[\"variant_id\"] + \":\" + df[c].astype(str)\n",
    "\n",
    "def chrom_float_to_int(chrom_float):\n",
    "    if np.isnan(chrom_float):\n",
    "        return -1\n",
    "    else:\n",
    "        return int(chrom_float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "85dc6542-ec07-4c3d-a956-4085ae9cb74a",
   "metadata": {},
   "outputs": [],
   "source": [
    "append_variant_id(glass_variants_passgeno_df)\n",
    "append_variant_id(glass_variants_anno_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d0fb47b8-e924-4f11-9576-6b90699861e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_variants_anno_df.set_index(\"variant_id\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c9aba2a5-8ea1-4eca-ab21-00665149208f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3014019, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:168513274:168513274:G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168513274.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>XCL2</td>\n",
       "      <td>FIVE_PRIME_FLANK</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:168545864:168545864:G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168545864.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>XCL1</td>\n",
       "      <td>FIVE_PRIME_UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:197398786:197398786:A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>197398786.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CRB1</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:211652463:211652463:C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>211652463.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>RD3</td>\n",
       "      <td>MISSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.T168S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:229771597:229771597:T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>229771597.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>URB2</td>\n",
       "      <td>NONSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.K413*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample               variant_id  chrom        start ref alt  \\\n",
       "583  GLSS-19-0266-R1  1:168513274:168513274:G    1.0  168513274.0   A   G   \n",
       "584  GLSS-19-0266-R1  1:168545864:168545864:G    1.0  168545864.0   A   G   \n",
       "593  GLSS-19-0266-R1  1:197398786:197398786:A    1.0  197398786.0   G   A   \n",
       "598  GLSS-19-0266-R1  1:211652463:211652463:C    1.0  211652463.0   G   C   \n",
       "601  GLSS-19-0266-R1  1:229771597:229771597:T    1.0  229771597.0   A   T   \n",
       "\n",
       "    gene_symbol variant_classification variant_type protein_change  \n",
       "583        XCL2       FIVE_PRIME_FLANK          SNP            NaN  \n",
       "584        XCL1         FIVE_PRIME_UTR          SNP            NaN  \n",
       "593        CRB1                 INTRON          SNP            NaN  \n",
       "598         RD3               MISSENSE          SNP        p.T168S  \n",
       "601        URB2               NONSENSE          SNP        p.K413*  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_passgeno_cols = [\"sample\", \"variant_id\"] \n",
    "kept_anno_cols = [\"chrom\", \"start\", \"ref\", \"alt\", \"gene_symbol\", \"variant_classification\", \"variant_type\", \"protein_change\"]\n",
    "\n",
    "glass_variants_df = glass_variants_passgeno_df[kept_passgeno_cols].join(glass_variants_anno_df[kept_anno_cols], on=\"variant_id\", how=\"left\")\n",
    "print(glass_variants_df.shape)\n",
    "glass_variants_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04fe63b0-45c8-4ee8-8070-daf63fd2812a",
   "metadata": {},
   "source": [
    "Reorder/filter the columns to a more standard format.\n",
    "\n",
    "Keep only genes with exons covered by all versions (v1, v2, v3) of OncoPanel.\n",
    "\n",
    "Keep only variants on valid chromosomes (1-X).\n",
    "\n",
    "Make sure variant positions are stored as integers.\n",
    "\n",
    "Remove duplicate rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "556aab90-65e1-4d75-8eee-3c12f880ad57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29772, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>3</td>\n",
       "      <td>70005722</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MITF</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>13</td>\n",
       "      <td>28913225</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>FLT1</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>X</td>\n",
       "      <td>123220608</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>STAG2</td>\n",
       "      <td>NONSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.Q1089*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1267</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>69990330</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>MITF</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1422</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>7</td>\n",
       "      <td>140481402</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>MISSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.G469A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample chrom      start ref alt gene_symbol  \\\n",
       "687   GLSS-19-0266-R1     3   70005722   G   A        MITF   \n",
       "970   GLSS-19-0266-R1    13   28913225   T   G        FLT1   \n",
       "1160  GLSS-19-0266-R1     X  123220608   C   T       STAG2   \n",
       "1267  GLSS-19-0266-TP     3   69990330   G   T        MITF   \n",
       "1422  GLSS-19-0266-TP     7  140481402   C   G        BRAF   \n",
       "\n",
       "     variant_classification variant_type protein_change  \n",
       "687                  INTRON          SNP            NaN  \n",
       "970                  INTRON          SNP            NaN  \n",
       "1160               NONSENSE          SNP       p.Q1089*  \n",
       "1267                 INTRON          SNP            NaN  \n",
       "1422               MISSENSE          SNP        p.G469A  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_glass_cols = [\"sample\", \"chrom\", \"start\", \"ref\", \"alt\", \"gene_symbol\", \"variant_classification\", \"variant_type\", \"protein_change\"]\n",
    "glass_variants_df = glass_variants_df.loc[:, kept_glass_cols]\n",
    "\n",
    "glass_variants_df = glass_variants_df[glass_variants_df['gene_symbol'].isin(oncopanel_genes['Gene'])]\n",
    "\n",
    "glass_variants_df[\"chrom\"] = glass_variants_df[\"chrom\"].map(chrom_float_to_int).astype(int)\n",
    "glass_variants_df = glass_variants_df[glass_variants_df[\"chrom\"] > 0] # Missing chromosome labels were mapped to -1\n",
    "glass_variants_df[\"chrom\"] = np.where(glass_variants_df[\"chrom\"] == 23, \"X\", glass_variants_df[\"chrom\"])\n",
    "glass_variants_df[\"start\"] = glass_variants_df[\"start\"].astype(int)\n",
    "\n",
    "glass_variants_df = glass_variants_df.drop_duplicates()\n",
    "print(glass_variants_df.shape)\n",
    "glass_variants_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4176a6db-26ad-46a0-9334-608a48c5fac2",
   "metadata": {},
   "source": [
    "Keep only non-synonymous variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0c43605d-182c-4644-9148-8022bd5bf297",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_classification\n",
       "INTRON                     23579\n",
       "MISSENSE                    2521\n",
       "FIVE_PRIME_FLANK            1102\n",
       "SILENT                       799\n",
       "THREE_PRIME_UTR              792\n",
       "SPLICE_SITE                  253\n",
       "FRAME_SHIFT_DEL              230\n",
       "NONSENSE                     213\n",
       "FIVE_PRIME_UTR               180\n",
       "FRAME_SHIFT_INS               49\n",
       "IN_FRAME_DEL                  40\n",
       "IN_FRAME_INS                   4\n",
       "START_CODON_SNP                3\n",
       "DE_NOVO_START_IN_FRAME         3\n",
       "DE_NOVO_START_OUT_FRAME        3\n",
       "NONSTOP                        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_variants_df['variant_classification'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "14c9d1b5-8366-446c-90d4-e20d41c58c93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_classification\n",
       "MISSENSE                   2521\n",
       "SPLICE_SITE                 253\n",
       "FRAME_SHIFT_DEL             230\n",
       "NONSENSE                    213\n",
       "FRAME_SHIFT_INS              49\n",
       "IN_FRAME_DEL                 40\n",
       "IN_FRAME_INS                  4\n",
       "START_CODON_SNP               3\n",
       "DE_NOVO_START_IN_FRAME        3\n",
       "DE_NOVO_START_OUT_FRAME       3\n",
       "NONSTOP                       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonsyn_classifications = ['MISSENSE',\n",
    "                          'SPLICE_SITE',\n",
    "                          'FRAME_SHIFT_DEL',\n",
    "                          'NONSENSE',\n",
    "                          'FRAME_SHIFT_INS',\n",
    "                          'IN_FRAME_DEL',\n",
    "                          'IN_FRAME_INS',\n",
    "                          'START_CODON_SNP',\n",
    "                          'DE_NOVO_START_IN_FRAME',\n",
    "                          'DE_NOVO_START_OUT_FRAME',\n",
    "                          'NONSTOP']\n",
    "\n",
    "glass_variants_df = glass_variants_df[glass_variants_df['variant_classification'].isin(nonsyn_classifications)]\n",
    "glass_variants_df['variant_classification'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cb43ab0-c6e3-4583-8067-2e8be6e2690f",
   "metadata": {},
   "source": [
    "### CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7841966b-e6a6-430d-aaa4-8915c1d6962e",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_cnvs_df = pd.read_csv('../../data/glass_cnvs.csv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bde655e2-960d-4910-af81-3e2e55719666",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35563, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliquot_barcode</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>hlvl_call</th>\n",
       "      <th>wcr</th>\n",
       "      <th>sample</th>\n",
       "      <th>cnv</th>\n",
       "      <th>hlvl_call_abs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2228044</th>\n",
       "      <td>GLSS-AT-GP01-TP-02D-WXS-DPVLWZ</td>\n",
       "      <td>PDCD1LG2</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.807710</td>\n",
       "      <td>GLSS-AT-GP01-TP</td>\n",
       "      <td>del</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156480</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.141363</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156491</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>EP300</td>\n",
       "      <td>1</td>\n",
       "      <td>1.100724</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156625</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>MCL1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.145647</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156707</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>CREBBP</td>\n",
       "      <td>1</td>\n",
       "      <td>1.109192</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         aliquot_barcode gene_symbol  hlvl_call       wcr  \\\n",
       "2228044   GLSS-AT-GP01-TP-02D-WXS-DPVLWZ    PDCD1LG2         -1  0.807710   \n",
       "29156480  GLSS-MD-0044-R1-01D-WGS-VG0QP6        ABL1          1  1.141363   \n",
       "29156491  GLSS-MD-0044-R1-01D-WGS-VG0QP6       EP300          1  1.100724   \n",
       "29156625  GLSS-MD-0044-R1-01D-WGS-VG0QP6        MCL1          1  1.145647   \n",
       "29156707  GLSS-MD-0044-R1-01D-WGS-VG0QP6      CREBBP          1  1.109192   \n",
       "\n",
       "                   sample  cnv  hlvl_call_abs  \n",
       "2228044   GLSS-AT-GP01-TP  del              1  \n",
       "29156480  GLSS-MD-0044-R1  amp              1  \n",
       "29156491  GLSS-MD-0044-R1  amp              1  \n",
       "29156625  GLSS-MD-0044-R1  amp              1  \n",
       "29156707  GLSS-MD-0044-R1  amp              1  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create and save \"sample\" column\n",
    "glass_cnvs_df[\"sample\"] = glass_cnvs_df[\"aliquot_barcode\"].str[:15]\n",
    "wes_cnv_samples = set(glass_cnvs_df[\"sample\"])\n",
    "\n",
    "# Keep only CNVs (not copy-neutral segments)\n",
    "glass_cnvs_df = glass_cnvs_df[glass_cnvs_df['hlvl_call'] != 0]\n",
    "\n",
    "# Keep only genes with exons covered by all versions (v1, v2, v3) of OncoPanel.\n",
    "glass_cnvs_df = glass_cnvs_df[glass_cnvs_df[\"gene_symbol\"].isin(oncopanel_genes[\"Gene\"])]\n",
    "\n",
    "# Simplify CNV calls to just amp or del\n",
    "glass_cnvs_df['cnv'] = glass_cnvs_df['hlvl_call'].map({2: 'amp', 1: 'amp', -1: 'del', -2: 'del'})\n",
    "\n",
    "# Scoring\n",
    "glass_cnvs_df['hlvl_call_abs'] = np.abs(glass_cnvs_df['hlvl_call'])\n",
    "\n",
    "# Deduplicate, keeping strongest event\n",
    "glass_cnvs_df.sort_values('hlvl_call_abs', inplace = True)\n",
    "glass_cnvs_df = glass_cnvs_df.drop_duplicates(['sample', 'gene_symbol', 'cnv'], keep = 'first')\n",
    "\n",
    "print(glass_cnvs_df.shape)\n",
    "glass_cnvs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc7a0373-09de-47f1-b46e-998531885a4c",
   "metadata": {},
   "source": [
    "### Intersect sample lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "48113151-34b6-40ac-a71e-0da8f533005c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(478, 30)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_samples_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f526a132-40c4-4fa0-813e-a6066aa2e50b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "389"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_intersection = (set(glass_samples_df.index).\\\n",
    "    intersection(wes_mut_samples).\\\n",
    "    intersection(wes_cnv_samples))\n",
    "\n",
    "len(sample_intersection)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c9be5e-17b8-4a8f-9fb7-2cf4bdbd6bc6",
   "metadata": {},
   "source": [
    "### Subset dataframes to these samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d34953d1-af8f-49f7-a359-f0778de5bac7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "330 361\n"
     ]
    }
   ],
   "source": [
    "variants_subset = glass_variants_df[glass_variants_df['sample'].isin(sample_intersection)]\n",
    "\n",
    "cnvs_subset = glass_cnvs_df[glass_cnvs_df['sample'].isin(sample_intersection)]\n",
    "\n",
    "print(variants_subset['sample'].nunique(),\n",
    "      cnvs_subset['sample'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4e8315f1-e262-4f01-9609-9047b2979f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_variant_median_rates(variant_df, pt_col=\"sample\", variant_col=\"hlvl_call\"):\n",
    "    counts = variant_df[variant_col].value_counts()\n",
    "    per_pt_median_rates = variant_df.groupby([variant_col, pt_col]).size().reset_index().pivot_table(index=pt_col, columns=variant_col, fill_value=0.0).droplevel(0,axis=1).median()\n",
    "    plt.bar(per_pt_median_rates.index, per_pt_median_rates.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dc0b2c55-e873-4444-822a-1c1a5e4617b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGxCAYAAADCo9TSAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAAJTtJREFUeJzt3XlU1XX+x/HXjeWCghShbBIQWWouMy7jUgpmcaQy0zKXMmxwjpXWOM7Uz2UcUEtKHccac2kjO7llpS16SipF56ilnpzKltEJRwzJpQS3MPDz+6PDPV4vCOjlcwGfj3PuOXO/93vvfX+/fQeefu+CwxhjBAAAYMllvh4AAABcWogPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gP1Norr7wih8Mhh8OhDRs2eNxujNE111wjh8OhlJQUrz53QkKCRo4c6bq+YcOGKuewZdOmTbrnnnsUGxurwMBAhYWFqWfPnlqwYIFOnDjhWi8hIUEOh0MPPvigx2NUbMcbb7whSRo4cKCCg4N19OjRKp/33nvvVUBAgH744Qevb5Mta9euVVZWlq/HaNAKCwuVlZWlnTt3etyWlZUlh8NR68ccOXKkEhISLn44L+E4aXyID1yw0NBQvfTSSx7L8/Ly9N///lehoaF1PkOnTp20ZcsWderUqc6fqzKZmZnq3bu3vv/+e02fPl25ublavny5+vbtq6ysLP31r3/1uM9LL72kb7/99ryPm5GRoZ9//llLly6t9Pbi4mKtWrVKt99+uyIjI72yLb6wdu1aTZ061ddjNGiFhYWaOnVqpfExatQobdmypdaPOWXKFK1atcoL03kHx0njQ3zggg0ZMkRvvvmmSkpK3Ja/9NJL6tGjh6666qo6n6FZs2bq3r27mjVrVufPda6VK1dq2rRpysjI0L/+9S/9/ve/V3JystLS0jR9+nTt2bNHaWlpbvfp0aOHmjZtqkmTJp33sdPS0hQTE6OXX3650tuXLVumU6dOKSMjw2vbg5orLy9XaWlpvZ+hZcuW6t69e60fOykpSb/97W8vdDSgWsQHLtiwYcMk/fqLsEJxcbHefPNN/f73v6/0PqdPn9YTTzyh1q1by+l0qnnz5nrggQd06NAht/V++eUXPf7444qKilKTJk1044036tNPP/V4vMpedtm+fbuGDh2qhIQEBQcHKyEhQcOGDdP//vc/t/tWvHy0fv16PfTQQ4qIiNCVV16pQYMGqbCwsNrtnzZtmq644go9++yzlZ7aDg0NVWpqqtuy8PBwTZgwQW+99Za2bt1a5WP7+fkpPT1dO3bs0BdffOFxe05OjqKjoz3ipjJLly5Vjx49FBISopCQEP3mN7/xOGP18ssvq2PHjgoKClJ4eLgGDhyor7/+2m2dlJSUSl9GO/cU/d69e+VwODR79mzNmTNHiYmJCgkJUY8ePdy2eeTIkXruueckyfUynsPh0N69eyX9GnfdunVTWFiYmjRpoquvvrrK4+psDodDY8eO1aJFi3TttdfK6XSqbdu2Wr58uce6RUVFGj16tFq2bKnAwEAlJiZq6tSpKisr89iemTNn6oknnlBiYqKcTqfWr19/0TMcOnRIDz/8sNq2bauQkBC1aNFCN910kzZt2uS23vlm6Nq1qyTpgQcecO3DipcoqnrZpbpjorKXXepqmy72OEEDZYBaysnJMZLMtm3bzIgRI8zvfvc7120LFiwwTZs2NSUlJeb66683ycnJrtvKy8tNv379TNOmTc3UqVNNbm6uefHFF01sbKxp27atOXnypGvd9PR043A4zGOPPWbWrVtn5syZY2JjY02zZs1Menq6a73169cbSWb9+vWuZStXrjR/+9vfzKpVq0xeXp5Zvny5SU5ONs2bNzeHDh3y2I6rr77aPPLII+aDDz4wL774orniiitMnz59zrsPCgsLjSQzZMiQGu+3+Ph4c9ttt5mTJ0+a2NhY06tXL4/tWLlypWvZ7t27jcPhMOPGjXN7nF27dhlJZsKECdU+55QpU4wkM2jQILNy5UrXvpwyZYprnRkzZhhJZtiwYWbNmjXm1VdfNVdffbUJCwsz//nPf1zrJScnu/33rJCenm7i4+Nd1/Pz840kk5CQYPr162dWr15tVq9ebdq3b2+uuOIKc/ToUWOMMXv27DF33323kWS2bNniuvz8889m8+bNxuFwmKFDh5q1a9eajz/+2OTk5JgRI0ZUu82STFxcnGnbtq1ZtmyZeeedd0y/fv089u+BAwdMXFyciY+PN4sWLTIffvihmT59unE6nWbkyJEe2xMbG2v69Olj3njjDbNu3TqTn59/0TN888035qGHHjLLly83GzZsMO+9957JyMgwl112mdsxXdUM//73v13H8V//+lfXPiwoKDDGGJOZmWnO/TFfk2Pi3P+mdblNF3OcoOEiPlBrZ8dHxS/NL7/80hhjTNeuXV0/uM+Nj2XLlhlJ5s0333R7vG3bthlJZv78+cYYY77++msjyfzpT39yW2/JkiVGUrXxca6ysjJz/Phx07RpU/PMM894bMfDDz/stv7MmTONJHPgwIEqH3Pr1q01DoAKFfFhjDEvvPCCkWTeffddt+04+4e4Mb/+wo+IiDCnT592Lfvzn/9sJLmFQWW+++474+fnZ+69994q1/npp59McHCwufXWW92W79u3zzidTjN8+HC3WWoTH+3btzdlZWWu5Z9++qmRZJYtW+ZaNmbMGI9fjsYYM3v2bCPJ9QuoNiSZ4OBgU1RU5FpWVlZmWrduba655hrXstGjR5uQkBDzv//9r9Ln3rVrl9v2JCUluf138MYM5yorKzO//PKL6du3rxk4cKBr+flmqPj/T05OjsfjnRsfNTkmjKk6Pupimy7mOEHDxcsuuCjJyclKSkrSyy+/rC+++ELbtm2r8tT4e++9p8svv1z9+/dXWVmZ6/Kb3/xGUVFRrpdOKk5n33vvvW73v+eee+Tv71/tTMePH9f//d//6ZprrpG/v7/8/f0VEhKiEydOeLyUIEl33HGH2/UOHTpIksfLNN70wAMPqG3btpowYYLOnDlT5XoZGRk6fPiw3nnnHUlSWVmZXnvtNfXq1UutWrU673Pk5uaqvLxcY8aMqXKdLVu26NSpU26fIJKkuLg43XTTTfroo49qvlHnuO222+Tn5+e6Xpv9WvFSwj333KPXX39d33//fa2eu2/fvm5vxPXz89OQIUO0Z88e7d+/X9Kvx2OfPn0UExPjdjxWvJSVl5fn9ph33HGHAgICvDqDJC1cuFCdOnVSUFCQ/P39FRAQoI8++qjKY7U2M5yrJsfE+dTFNl3McYKGi/jARXE4HHrggQf02muvaeHChbr22mvVq1evStf94YcfdPToUQUGBiogIMDtUlRUpMOHD0uSjhw5IkmKiopyu7+/v7+uvPLKamcaPny45s2bp1GjRumDDz7Qp59+qm3btql58+Y6deqUx/rnPqbT6ZSkStetUPFm2vz8/GrnqYyfn59mzJihXbt2afHixVWud/fddyssLEw5OTmSfn3X/w8//FCjN5pWvI+mZcuWVa5Tsa+jo6M9bouJiXHdfiEuZL9W6N27t1avXq2ysjLdf//9atmypdq1a+f2/qLzOffYOXtZxTb98MMPevfddz2Oxeuvv16SXMdjhcr20cXOMGfOHD300EPq1q2b3nzzTW3dulXbtm1Tv379Kt1PtZ3hXDU5Js6nLrbpYo4TNFzV/zMSqMbIkSP1t7/9TQsXLtSTTz5Z5XoVb+h8//33K7294qO5FT+MioqKFBsb67q9rKys2l+GxcXFeu+995SZmakJEya4lpeWlurHH3+s8TZVJzo6Wu3bt9e6det08uRJNWnSpNaPMWDAAN1www3KzMzU888/X+k6wcHBGjZsmF544QUdOHBAL7/8skJDQzV48OBqH7958+aSpP379ysuLq7SdSr29YEDBzxuKywsVEREhOt6UFCQiouLPdY795e0twwYMEADBgxQaWmptm7dquzsbA0fPlwJCQnq0aPHee9bVFRU5bKKbY6IiFCHDh2qPGZjYmLcrtf2+zJqMsNrr72mlJQULViwwG29Y8eOVfqYF/KdHWeryTFxPnWxTbg0ceYDFy02NlaPPfaY+vfvr/T09CrXu/3223XkyBGVl5erS5cuHpfrrrtOklyfqFiyZInb/V9//XW3TyFUxuFwyBjj+tdThRdffFHl5eUXsHVVmzJlin766Sc9+uijMsZ43H78+HGtW7fuvI/x9NNPq6CgQM8++2yV62RkZKi8vFyzZs3S2rVrNXTo0BrFTmpqqvz8/Dx+CZytR48eCg4O1muvvea2fP/+/fr444/Vt29f17KEhAT95z//cft455EjR7R58+ZqZ6lKTf6V63Q6lZycrKefflqS9Nlnn1X7uB999JHbl6+Vl5drxYoVSkpKcv2r//bbb9eXX36ppKSkSo/Hc+Ojtmoyg8Ph8DhWP//881p9N0dtzhTU5Jg4H1vbdC7OhjQ+nPmAVzz11FPVrjN06FAtWbJEt956q/74xz/qd7/7nQICArR//36tX79eAwYM0MCBA9WmTRvdd999mjt3rgICAnTzzTfryy+/1OzZs6v9Po9mzZqpd+/emjVrliIiIpSQkKC8vDy99NJLuvzyy720tb8aPHiwpkyZounTp+ubb75RRkaGkpKSdPLkSX3yySdatGiRhgwZ4vFx27PdcMMNGjBggN5+++0q1+nSpYs6dOiguXPnyhhT4+/2SEhI0KRJkzR9+nSdOnVKw4YNU1hYmL766isdPnxYU6dO1eWXX64pU6Zo0qRJuv/++zVs2DAdOXJEU6dOVVBQkDIzM12PN2LECC1atEj33Xef/vCHP+jIkSOaOXPmRX3HSvv27SX9GmFpaWny8/NThw4d9MQTT2j//v3q27evWrZsqaNHj+qZZ55RQECAkpOTq33ciIgI3XTTTZoyZYqaNm2q+fPn65tvvnH7WOi0adOUm5urnj176tFHH9V1112nn3/+WXv37tXatWu1cOHCC355oqYz3H777Zo+fboyMzOVnJysb7/9VtOmTVNiYmK1oV0hKSlJwcHBWrJkidq0aaOQkBDFxMRUGk81OSbqwzadq6rjJDAw8IIeD/WAj9/wigbo7E+7nM+5n3YxxphffvnFzJ4923Ts2NEEBQWZkJAQ07p1azN69Gize/du13qlpaXmz3/+s2nRooUJCgoy3bt3N1u2bDHx8fHVftpl//795q677jJXXHGFCQ0NNf369TNffvmlx32r2o6afILmbHl5eebuu+820dHRJiAgwDRr1sz06NHDzJo1y5SUlLjWO/vTLmf76quvjJ+fX6WfdqnwzDPPGEmmbdu2NZrpbK+++qrp2rWra3//9re/9fhkxIsvvmg6dOhgAgMDTVhYmBkwYIDr0x5nW7x4sWnTpo0JCgoybdu2NStWrKjy0y6zZs3yuL8kk5mZ6bpeWlpqRo0aZZo3b24cDoeRZPLz8817771n0tLSTGxsrAkMDDQtWrQwt956q9m0aVO12yvJjBkzxsyfP98kJSWZgIAA07p1a7NkyRKPdQ8dOmQeffRRk5iYaAICAkx4eLjp3LmzmTx5sjl+/Hi123OxM5SWlpq//OUvJjY21gQFBZlOnTqZ1atX12qfGvPrJ8lat25tAgIC3PZxZR+1Nab6Y6KqT7vY2qaaHidouBzGVHK+GAAaKIfDoTFjxmjevHmX9Aze1hi3Cb7Dez4AAIBVxAcAALCKl10AAIBVnPkAAABWER8AAMAq4gMAAFhV775k7MyZMyosLFRoaOhFf5UwAACwwxijY8eOKSYmRpdddv5zG/UuPgoLCy/obw4AAADfKygoqPbbgetdfFT8cbGCgoKL+tpmAABgT0lJieLi4ly/x8+n3sVHxUstzZo1Iz4AAGhgavKWCd5wCgAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFjl7+sBcGlImLDG1yM0GHufus3XIwBAneLMBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGCVv68HAIDGJmHCGl+P0GDsfeo2X48AH+DMBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVtUqPrKzs9W1a1eFhoaqRYsWuvPOO/Xtt9+6rWOMUVZWlmJiYhQcHKyUlBTt2rXLq0MDAICGq1bxkZeXpzFjxmjr1q3Kzc1VWVmZUlNTdeLECdc6M2fO1Jw5czRv3jxt27ZNUVFRuuWWW3Ts2DGvDw8AABqeWv1huffff9/tek5Ojlq0aKEdO3aod+/eMsZo7ty5mjx5sgYNGiRJWrx4sSIjI7V06VKNHj3ae5MDAIAG6aLe81FcXCxJCg8PlyTl5+erqKhIqamprnWcTqeSk5O1efPmSh+jtLRUJSUlbhcAANB4XXB8GGM0fvx43XjjjWrXrp0kqaioSJIUGRnptm5kZKTrtnNlZ2crLCzMdYmLi7vQkQAAQANwwfExduxYff7551q2bJnHbQ6Hw+26McZjWYWJEyequLjYdSkoKLjQkQAAQANQq/d8VHjkkUf0zjvvaOPGjWrZsqVreVRUlKRfz4BER0e7lh88eNDjbEgFp9Mpp9N5IWMAAIAGqFZnPowxGjt2rN566y19/PHHSkxMdLs9MTFRUVFRys3NdS07ffq08vLy1LNnT+9MDAAAGrRanfkYM2aMli5dqrfffluhoaGu93GEhYUpODhYDodD48aN04wZM9SqVSu1atVKM2bMUJMmTTR8+PA62QAAANCw1Co+FixYIElKSUlxW56Tk6ORI0dKkh5//HGdOnVKDz/8sH766Sd169ZN69atU2hoqFcGBgAADVut4sMYU+06DodDWVlZysrKutCZAABAI8bfdgEAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwKpax8fGjRvVv39/xcTEyOFwaPXq1W63jxw5Ug6Hw+3SvXt3b80LAAAauFrHx4kTJ9SxY0fNmzevynX69eunAwcOuC5r1669qCEBAEDj4V/bO6SlpSktLe286zidTkVFRV3wUAAAoPGqk/d8bNiwQS1atNC1116rP/zhDzp48GCV65aWlqqkpMTtAgAAGi+vx0daWpqWLFmijz/+WH//+9+1bds23XTTTSotLa10/ezsbIWFhbkucXFx3h4JAADUI7V+2aU6Q4YMcf3vdu3aqUuXLoqPj9eaNWs0aNAgj/UnTpyo8ePHu66XlJQQIAAANGJej49zRUdHKz4+Xrt37670dqfTKafTWddjAACAeqLOv+fjyJEjKigoUHR0dF0/FQAAaABqfebj+PHj2rNnj+t6fn6+du7cqfDwcIWHhysrK0t33XWXoqOjtXfvXk2aNEkREREaOHCgVwcHAAANU63jY/v27erTp4/resX7NdLT07VgwQJ98cUXevXVV3X06FFFR0erT58+WrFihUJDQ703NQAAaLBqHR8pKSkyxlR5+wcffHBRAwEAgMaNv+0CAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsqnV8bNy4Uf3791dMTIwcDodWr17tdrsxRllZWYqJiVFwcLBSUlK0a9cub80LAAAauFrHx4kTJ9SxY0fNmzev0ttnzpypOXPmaN68edq2bZuioqJ0yy236NixYxc9LAAAaPj8a3uHtLQ0paWlVXqbMUZz587V5MmTNWjQIEnS4sWLFRkZqaVLl2r06NEXNy0AAGjwvPqej/z8fBUVFSk1NdW1zOl0Kjk5WZs3b670PqWlpSopKXG7AACAxsur8VFUVCRJioyMdFseGRnpuu1c2dnZCgsLc13i4uK8ORIAAKhn6uTTLg6Hw+26McZjWYWJEyequLjYdSkoKKiLkQAAQD1R6/d8nE9UVJSkX8+AREdHu5YfPHjQ42xIBafTKafT6c0xAABAPebVMx+JiYmKiopSbm6ua9np06eVl5ennj17evOpAABAA1XrMx/Hjx/Xnj17XNfz8/O1c+dOhYeH66qrrtK4ceM0Y8YMtWrVSq1atdKMGTPUpEkTDR8+3KuDAwCAhqnW8bF9+3b16dPHdX38+PGSpPT0dL3yyit6/PHHderUKT388MP66aef1K1bN61bt06hoaHemxoAADRYtY6PlJQUGWOqvN3hcCgrK0tZWVkXMxcAAGik+NsuAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABY5fX4yMrKksPhcLtERUV5+2kAAEAD5V8XD3r99dfrww8/dF338/Ori6cBAAANUJ3Eh7+/P2c7AABAperkPR+7d+9WTEyMEhMTNXToUH333XdVrltaWqqSkhK3CwAAaLy8Hh/dunXTq6++qg8++EAvvPCCioqK1LNnTx05cqTS9bOzsxUWFua6xMXFeXskAABQj3g9PtLS0nTXXXepffv2uvnmm7VmzRpJ0uLFiytdf+LEiSouLnZdCgoKvD0SAACoR+rkPR9na9q0qdq3b6/du3dXervT6ZTT6azrMQAAQD1R59/zUVpaqq+//lrR0dF1/VQAAKAB8Hp8/OUvf1FeXp7y8/P1ySef6O6771ZJSYnS09O9/VQAAKAB8vrLLvv379ewYcN0+PBhNW/eXN27d9fWrVsVHx/v7acCAAANkNfjY/ny5d5+SAAA0Ijwt10AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABglb+vB7AtYcIaX4/QYOx96jZfjwAAaIQ48wEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABY5e/rAQAA8IaECWt8PUKDsfep23z6/Jz5AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAqjqLj/nz5ysxMVFBQUHq3LmzNm3aVFdPBQAAGpA6iY8VK1Zo3Lhxmjx5sj777DP16tVLaWlp2rdvX108HQAAaEDqJD7mzJmjjIwMjRo1Sm3atNHcuXMVFxenBQsW1MXTAQCABsTrf9X29OnT2rFjhyZMmOC2PDU1VZs3b/ZYv7S0VKWlpa7rxcXFkqSSkhJvjyZJOlN6sk4etzHy5n8D9nvN1dWxD3s43muOnzO+URc/Zyoe0xhT7bpej4/Dhw+rvLxckZGRbssjIyNVVFTksX52dramTp3qsTwuLs7bo6GWwub6eoJLE/sdlxKOd9+oy/1+7NgxhYWFnXcdr8dHBYfD4XbdGOOxTJImTpyo8ePHu66fOXNGP/74o6688spK12+MSkpKFBcXp4KCAjVr1szX41wy2O/2sc99g/3uG5fafjfG6NixY4qJial2Xa/HR0REhPz8/DzOchw8eNDjbIgkOZ1OOZ1Ot2WXX365t8dqEJo1a3ZJHKD1DfvdPva5b7DffeNS2u/VnfGo4PU3nAYGBqpz587Kzc11W56bm6uePXt6++kAAEADUycvu4wfP14jRoxQly5d1KNHDz3//PPat2+fHnzwwbp4OgAA0IDUSXwMGTJER44c0bRp03TgwAG1a9dOa9euVXx8fF08XYPndDqVmZnp8fIT6hb73T72uW+w332D/V41h6nJZ2IAAAC8hL/tAgAArCI+AACAVcQHAACwivgAAABWER8AAMAq4qMe2bt3rzIyMpSYmKjg4GAlJSUpMzNTp0+f9vVojd6TTz6pnj17qkmTJpfsN+zaMH/+fCUmJiooKEidO3fWpk2bfD1So7dx40b1799fMTExcjgcWr16ta9HavSys7PVtWtXhYaGqkWLFrrzzjv17bff+nqseoX4qEe++eYbnTlzRosWLdKuXbv0j3/8QwsXLtSkSZN8PVqjd/r0aQ0ePFgPPfSQr0dptFasWKFx48Zp8uTJ+uyzz9SrVy+lpaVp3759vh6tUTtx4oQ6duyoefPm+XqUS0ZeXp7GjBmjrVu3Kjc3V2VlZUpNTdWJEyd8PVq9wfd81HOzZs3SggUL9N133/l6lEvCK6+8onHjxuno0aO+HqXR6datmzp16qQFCxa4lrVp00Z33nmnsrOzfTjZpcPhcGjVqlW68847fT3KJeXQoUNq0aKF8vLy1Lt3b1+PUy9w5qOeKy4uVnh4uK/HAC7K6dOntWPHDqWmprotT01N1ebNm300FWBHcXGxJPGz/CzERz323//+V//85z/5mzho8A4fPqzy8nKPv2wdGRnp8RewgcbEGKPx48frxhtvVLt27Xw9Tr1BfFiQlZUlh8Nx3sv27dvd7lNYWKh+/fpp8ODBGjVqlI8mb9guZL+jbjkcDrfrxhiPZUBjMnbsWH3++edatmyZr0epV+rkD8vB3dixYzV06NDzrpOQkOD634WFherTp4/rLwLjwtR2v6PuREREyM/Pz+Msx8GDBz3OhgCNxSOPPKJ33nlHGzduVMuWLX09Tr1CfFgQERGhiIiIGq37/fffq0+fPurcubNycnJ02WWcnLpQtdnvqFuBgYHq3LmzcnNzNXDgQNfy3NxcDRgwwIeTAd5njNEjjzyiVatWacOGDUpMTPT1SPUO8VGPFBYWKiUlRVdddZVmz56tQ4cOuW6Liory4WSN3759+/Tjjz9q3759Ki8v186dOyVJ11xzjUJCQnw7XCMxfvx4jRgxQl26dHGd1du3bx/vaapjx48f1549e1zX8/PztXPnToWHh+uqq67y4WSN15gxY7R06VK9/fbbCg0NdZ3xCwsLU3BwsI+nqycM6o2cnBwjqdIL6lZ6enql+339+vW+Hq1Ree6550x8fLwJDAw0nTp1Mnl5eb4eqdFbv359pcd2enq6r0drtKr6OZ6Tk+Pr0eoNvucDAABYxRsKAACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABW/T+gDX+/LXEhtwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_variant_median_rates(cnvs_subset, pt_col=\"sample\", variant_col=\"hlvl_call\")\n",
    "plt.title(\"Median CNV counts per participant\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00682ef6-0bd3-40de-8014-0bada5fd0d49",
   "metadata": {},
   "source": [
    "### Create a combined mutation + CNV + subtype matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "865ddd1e-cef4-4f2f-b8ba-fd84270b056c",
   "metadata": {},
   "source": [
    "#### Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "970b4eea-37c0-40ba-acea-862b73f2aeca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/389233590.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  variants_subset['gene_mut'] = variants_subset['gene_symbol'] + '_mut'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_mut</th>\n",
       "      <th>ABL1_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>AKT2_mut</th>\n",
       "      <th>AKT3_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>TSC1_mut</th>\n",
       "      <th>TSC2_mut</th>\n",
       "      <th>U2AF1_mut</th>\n",
       "      <th>WRN_mut</th>\n",
       "      <th>WT1_mut</th>\n",
       "      <th>XPA_mut</th>\n",
       "      <th>XPC_mut</th>\n",
       "      <th>XPO1_mut</th>\n",
       "      <th>ZNF217_mut</th>\n",
       "      <th>ZRSR2_mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 211 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_mut         ABL1_mut  AKT1_mut  AKT2_mut  AKT3_mut  ALK_mut  APC_mut  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1         0         0         0         0        0        0   \n",
       "GLSS-19-0266-TP         0         0         0         0        0        0   \n",
       "GLSS-19-0267-R1         0         0         0         0        0        0   \n",
       "GLSS-19-0267-TP         0         0         0         0        0        0   \n",
       "GLSS-19-0268-R1         0         0         0         0        0        0   \n",
       "\n",
       "gene_mut         ARAF_mut  ARID1A_mut  ARID1B_mut  ARID2_mut  ...  TSC1_mut  \\\n",
       "sample                                                        ...             \n",
       "GLSS-19-0266-R1         0           0           0          0  ...         0   \n",
       "GLSS-19-0266-TP         0           0           0          0  ...         0   \n",
       "GLSS-19-0267-R1         0           0           0          0  ...         0   \n",
       "GLSS-19-0267-TP         0           0           0          0  ...         0   \n",
       "GLSS-19-0268-R1         0           0           0          0  ...         0   \n",
       "\n",
       "gene_mut         TSC2_mut  U2AF1_mut  WRN_mut  WT1_mut  XPA_mut  XPC_mut  \\\n",
       "sample                                                                     \n",
       "GLSS-19-0266-R1         0          0        0        0        0        0   \n",
       "GLSS-19-0266-TP         0          0        0        0        0        0   \n",
       "GLSS-19-0267-R1         0          0        0        0        0        0   \n",
       "GLSS-19-0267-TP         0          0        0        0        0        0   \n",
       "GLSS-19-0268-R1         0          0        0        0        0        0   \n",
       "\n",
       "gene_mut         XPO1_mut  ZNF217_mut  ZRSR2_mut  \n",
       "sample                                            \n",
       "GLSS-19-0266-R1         0           0          0  \n",
       "GLSS-19-0266-TP         0           0          0  \n",
       "GLSS-19-0267-R1         0           0          0  \n",
       "GLSS-19-0267-TP         0           0          0  \n",
       "GLSS-19-0268-R1         0           0          0  \n",
       "\n",
       "[5 rows x 211 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_subset['gene_mut'] = variants_subset['gene_symbol'] + '_mut'\n",
    "variants_matrix = variants_subset.pivot_table(index = 'sample', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# Scoring: point mutations = 2\n",
    "variants_matrix = variants_matrix * 2\n",
    "\n",
    "variants_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c255550-1889-4b0a-b22b-c3a8fd51b6a3",
   "metadata": {},
   "source": [
    "#### CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "242fad5a-5ae5-4015-82a9-e7edbb9cbd12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/4001559487.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cnvs_subset['gene_cnv'] = cnvs_subset['gene_symbol'] + '_' + cnvs_subset['cnv']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_cnv</th>\n",
       "      <th>ABL1_amp</th>\n",
       "      <th>ABL1_del</th>\n",
       "      <th>AKT1_amp</th>\n",
       "      <th>AKT1_del</th>\n",
       "      <th>AKT2_amp</th>\n",
       "      <th>AKT2_del</th>\n",
       "      <th>AKT3_amp</th>\n",
       "      <th>AKT3_del</th>\n",
       "      <th>ALK_amp</th>\n",
       "      <th>ALK_del</th>\n",
       "      <th>...</th>\n",
       "      <th>XPA_amp</th>\n",
       "      <th>XPA_del</th>\n",
       "      <th>XPC_amp</th>\n",
       "      <th>XPC_del</th>\n",
       "      <th>XPO1_amp</th>\n",
       "      <th>XPO1_del</th>\n",
       "      <th>ZNF217_amp</th>\n",
       "      <th>ZNF217_del</th>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <th>ZRSR2_del</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 475 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_cnv         ABL1_amp  ABL1_del  AKT1_amp  AKT1_del  AKT2_amp  AKT2_del  \\\n",
       "sample                                                                        \n",
       "GLSS-19-0266-R1         0         0         0         0         0         0   \n",
       "GLSS-19-0266-TP         0         0         0         0         1         0   \n",
       "GLSS-19-0267-R1         0         0         0         0         0         0   \n",
       "GLSS-19-0267-TP         0         1         0         0         0         0   \n",
       "GLSS-19-0268-TP         0         0         0         0         0         0   \n",
       "\n",
       "gene_cnv         AKT3_amp  AKT3_del  ALK_amp  ALK_del  ...  XPA_amp  XPA_del  \\\n",
       "sample                                                 ...                     \n",
       "GLSS-19-0266-R1         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0266-TP         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0267-R1         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0267-TP         0         0        0        0  ...        0        1   \n",
       "GLSS-19-0268-TP         0         0        0        0  ...        0        1   \n",
       "\n",
       "gene_cnv         XPC_amp  XPC_del  XPO1_amp  XPO1_del  ZNF217_amp  ZNF217_del  \\\n",
       "sample                                                                          \n",
       "GLSS-19-0266-R1        0        0         0         0           1           0   \n",
       "GLSS-19-0266-TP        0        0         0         0           1           0   \n",
       "GLSS-19-0267-R1        0        0         0         0           0           0   \n",
       "GLSS-19-0267-TP        0        0         0         0           0           0   \n",
       "GLSS-19-0268-TP        1        0         0         0           0           1   \n",
       "\n",
       "gene_cnv         ZRSR2_amp  ZRSR2_del  \n",
       "sample                                 \n",
       "GLSS-19-0266-R1          0          0  \n",
       "GLSS-19-0266-TP          0          0  \n",
       "GLSS-19-0267-R1          0          0  \n",
       "GLSS-19-0267-TP          0          0  \n",
       "GLSS-19-0268-TP          0          0  \n",
       "\n",
       "[5 rows x 475 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnvs_subset['gene_cnv'] = cnvs_subset['gene_symbol'] + '_' + cnvs_subset['cnv']\n",
    "cnvs_matrix = cnvs_subset.pivot_table(index = 'sample', columns = 'gene_cnv', values = 'hlvl_call_abs', aggfunc = 'sum', fill_value = 0)\n",
    "\n",
    "cnvs_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e8107a5-914f-40f8-bd38-6a61296bebc8",
   "metadata": {},
   "source": [
    "#### Combined\n",
    "Keep only variant types with 2 or more samples with an alteration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "005bb477-f614-4df7-93b4-87e22d4eaed6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(652, 365)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix = variants_matrix.join(cnvs_matrix, how = 'outer').fillna(0)\n",
    "matrix = matrix.loc[:, matrix.astype(bool).astype(int).sum() >= 2].T\n",
    "\n",
    "matrix.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8dde6a-0e03-4e3b-84c5-9f413863bf77",
   "metadata": {},
   "source": [
    "#### Write to file\n",
    "Will need to convert this to .gct format manually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "514f448e-4d43-4318-abd2-974d729f05db",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.to_csv('../../data/gene_sample_matrix.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c13384b8-67c9-465d-b533-fde8c490ea39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <th>GLSS-19-0269-R1</th>\n",
       "      <th>GLSS-19-0269-TP</th>\n",
       "      <th>GLSS-19-0270-R1</th>\n",
       "      <th>GLSS-19-0270-TP</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-14-1034-TP</th>\n",
       "      <th>TCGA-14-1402-R1</th>\n",
       "      <th>TCGA-14-1402-TP</th>\n",
       "      <th>TCGA-19-0957-R1</th>\n",
       "      <th>TCGA-19-0957-TP</th>\n",
       "      <th>TCGA-19-1389-R1</th>\n",
       "      <th>TCGA-19-1389-TP</th>\n",
       "      <th>TCGA-19-4065-R1</th>\n",
       "      <th>TCGA-19-4065-TP</th>\n",
       "      <th>TCGA-FG-5963-R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT2_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT3_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALK_mut</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPO1_del</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF217_amp</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF217_del</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRSR2_del</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>652 rows Ã— 365 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample      GLSS-19-0266-R1  GLSS-19-0266-TP  GLSS-19-0267-R1  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              1.0              1.0              0.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      GLSS-19-0267-TP  GLSS-19-0268-R1  GLSS-19-0268-TP  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              0.0              0.0              0.0   \n",
       "ZNF217_del              0.0              0.0              1.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      GLSS-19-0269-R1  GLSS-19-0269-TP  GLSS-19-0270-R1  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              0.0              0.0              1.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      GLSS-19-0270-TP  ...  TCGA-14-1034-TP  TCGA-14-1402-R1  \\\n",
       "ABL1_mut                0.0  ...              0.0              2.0   \n",
       "AKT1_mut                0.0  ...              0.0              6.0   \n",
       "AKT2_mut                0.0  ...              0.0              0.0   \n",
       "AKT3_mut                0.0  ...              0.0              2.0   \n",
       "ALK_mut                 0.0  ...              0.0              6.0   \n",
       "...                     ...  ...              ...              ...   \n",
       "XPO1_del                0.0  ...              0.0              0.0   \n",
       "ZNF217_amp              1.0  ...              0.0              1.0   \n",
       "ZNF217_del              0.0  ...              0.0              0.0   \n",
       "ZRSR2_amp               0.0  ...              0.0              0.0   \n",
       "ZRSR2_del               0.0  ...              0.0              0.0   \n",
       "\n",
       "sample      TCGA-14-1402-TP  TCGA-19-0957-R1  TCGA-19-0957-TP  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              1.0              0.0              1.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               0.0              0.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      TCGA-19-1389-R1  TCGA-19-1389-TP  TCGA-19-4065-R1  \\\n",
       "ABL1_mut                0.0              0.0              0.0   \n",
       "AKT1_mut                0.0              0.0              0.0   \n",
       "AKT2_mut                0.0              0.0              0.0   \n",
       "AKT3_mut                0.0              0.0              0.0   \n",
       "ALK_mut                 0.0              0.0              0.0   \n",
       "...                     ...              ...              ...   \n",
       "XPO1_del                0.0              0.0              0.0   \n",
       "ZNF217_amp              0.0              1.0              0.0   \n",
       "ZNF217_del              0.0              0.0              0.0   \n",
       "ZRSR2_amp               2.0              1.0              0.0   \n",
       "ZRSR2_del               0.0              0.0              0.0   \n",
       "\n",
       "sample      TCGA-19-4065-TP  TCGA-FG-5963-R1  \n",
       "ABL1_mut                0.0              0.0  \n",
       "AKT1_mut                0.0              0.0  \n",
       "AKT2_mut                0.0              0.0  \n",
       "AKT3_mut                0.0              0.0  \n",
       "ALK_mut                 0.0              0.0  \n",
       "...                     ...              ...  \n",
       "XPO1_del                0.0              0.0  \n",
       "ZNF217_amp              0.0              0.0  \n",
       "ZNF217_del              0.0              0.0  \n",
       "ZRSR2_amp               2.0              0.0  \n",
       "ZRSR2_del               0.0              0.0  \n",
       "\n",
       "[652 rows x 365 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10dd3b58-6ad9-47fa-b85c-57cba2a500e8",
   "metadata": {},
   "source": [
    "## Run clustering with external script\n",
    "\n",
    "As noted above:\n",
    "> The actual consensus clustering is done by the NMFConsensus script from GenePattern here: https://github.com/genepattern/NMFConsensus/blob/develop/nmfconsensus.R, and the clustering parameters are as listed here: gs://fc-secure-3f0f305e-f110-432b-bc9a-c8d96824cea2/jweiss/local_files/results/2024-07-22_GLASS-Verhaak-classifier/GLASS_sample_clustering.params.txt\n",
    "\n",
    "The outputs appear to be located at `os.path.join(results_dir, 'GLASS_sample_clustering_v2.membership.gct')`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c03f0ec-69bd-4c1b-a243-17fa17ef92bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Assess clustering overlap with Verhaak subtype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3897da0-3ee4-412a-b82e-651facf565b3",
   "metadata": {},
   "source": [
    "## Load clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "03204775-058e-4831-bf2d-fc6ff4baf7ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>membership</th>\n",
       "      <th>membership.1</th>\n",
       "      <th>membership.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0266.R1</th>\n",
       "      <td>GLSS.19.0266.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0266.TP</th>\n",
       "      <td>GLSS.19.0266.TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0267.R1</th>\n",
       "      <td>GLSS.19.0267.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0267.TP</th>\n",
       "      <td>GLSS.19.0267.TP</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0268.R1</th>\n",
       "      <td>GLSS.19.0268.R1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.1389.R1</th>\n",
       "      <td>TCGA.19.1389.R1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.1389.TP</th>\n",
       "      <td>TCGA.19.1389.TP</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.4065.R1</th>\n",
       "      <td>TCGA.19.4065.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.4065.TP</th>\n",
       "      <td>TCGA.19.4065.TP</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.FG.5963.R1</th>\n",
       "      <td>TCGA.FG.5963.R1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Description  membership  membership.1  membership.2\n",
       "Name                                                                    \n",
       "GLSS.19.0266.R1  GLSS.19.0266.R1           1             1             1\n",
       "GLSS.19.0266.TP  GLSS.19.0266.TP           2             2             2\n",
       "GLSS.19.0267.R1  GLSS.19.0267.R1           1             1             1\n",
       "GLSS.19.0267.TP  GLSS.19.0267.TP           3             3             3\n",
       "GLSS.19.0268.R1  GLSS.19.0268.R1           4             3             3\n",
       "...                          ...         ...           ...           ...\n",
       "TCGA.19.1389.R1  TCGA.19.1389.R1           3             3             3\n",
       "TCGA.19.1389.TP  TCGA.19.1389.TP           6             5             4\n",
       "TCGA.19.4065.R1  TCGA.19.4065.R1           1             5             4\n",
       "TCGA.19.4065.TP  TCGA.19.4065.TP           1             5             4\n",
       "TCGA.FG.5963.R1  TCGA.FG.5963.R1           3             3             3\n",
       "\n",
       "[365 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_68 = pd.read_csv(\"../../data/GLASS_sample_clustering_revision.membership.gct\", sep=\"\\t\", skiprows=2, index_col=0)\n",
    "memberships_25 = pd.read_csv(\"../../data/GLASS_sample_clustering_v2.membership.gct\", sep=\"\\t\", skiprows=2, index_col=0)\n",
    "memberships_68"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0736b97c-5ea9-457f-bec4-96f576083745",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8\n",
       "Name                                                            \n",
       "GLSS-19-0266-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0266-TP      2      2      2      2      2      2      2\n",
       "GLSS-19-0267-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0267-TP      2      2      2      3      3      3      3\n",
       "GLSS-19-0268-R1      2      3      3      4      4      3      3"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_25.rename(columns = {**{'membership': 'k = 2'}, **{f'membership.{i}': f'k = {i+2}' for i in range(1,8)}}, inplace = True)\n",
    "memberships_68.rename(columns = {**{'membership': 'k = 6'}, **{f'membership.{i}': f'k = {i+6}' for i in range(1,8)}}, inplace = True)\n",
    "\n",
    "memberships = memberships_25.join(memberships_68[[c for c in memberships_68.columns if c.startswith(\"k\")]])\n",
    "\n",
    "memberships.drop(columns = 'Description', inplace = True)\n",
    "memberships.index = memberships.index.str.replace('.', '-')\n",
    "memberships.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc7c6fc-3237-410b-93b3-8c3160cc2205",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load ssGSEA\n",
    "(Verhaak subtype scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a6ed5f47-3e3a-49d3-be5f-9b4425301066",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(425, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_CLASSICAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_NEURAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_PRONEURAL</th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "      <th>surgery_number</th>\n",
       "      <th>surgical_interval_mo</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>4499.507042</td>\n",
       "      <td>3725.238591</td>\n",
       "      <td>1945.457473</td>\n",
       "      <td>2028.442349</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>3866.366259</td>\n",
       "      <td>3899.952531</td>\n",
       "      <td>3055.641397</td>\n",
       "      <td>1970.036575</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</td>\n",
       "      <td>2</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3998.177157</td>\n",
       "      <td>2373.110794</td>\n",
       "      <td>2792.203914</td>\n",
       "      <td>3845.965090</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3403.500159</td>\n",
       "      <td>1489.313078</td>\n",
       "      <td>4904.914629</td>\n",
       "      <td>3993.835349</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_NEURAL</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <td>GLSS-19-0268</td>\n",
       "      <td>4367.052085</td>\n",
       "      <td>2494.211618</td>\n",
       "      <td>813.639386</td>\n",
       "      <td>3709.522026</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 case_barcode  VERHAAK_GLIOBLASTOMA_CLASSICAL  \\\n",
       "sample_barcode                                                  \n",
       "GLSS-19-0266-TP  GLSS-19-0266                     4499.507042   \n",
       "GLSS-19-0266-R1  GLSS-19-0266                     3866.366259   \n",
       "GLSS-19-0267-TP  GLSS-19-0267                     3998.177157   \n",
       "GLSS-19-0267-R1  GLSS-19-0267                     3403.500159   \n",
       "GLSS-19-0268-TP  GLSS-19-0268                     4367.052085   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_MESENCHYMAL  \\\n",
       "sample_barcode                                      \n",
       "GLSS-19-0266-TP                       3725.238591   \n",
       "GLSS-19-0266-R1                       3899.952531   \n",
       "GLSS-19-0267-TP                       2373.110794   \n",
       "GLSS-19-0267-R1                       1489.313078   \n",
       "GLSS-19-0268-TP                       2494.211618   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_NEURAL  VERHAAK_GLIOBLASTOMA_PRONEURAL  \\\n",
       "sample_barcode                                                                 \n",
       "GLSS-19-0266-TP                  1945.457473                     2028.442349   \n",
       "GLSS-19-0266-R1                  3055.641397                     1970.036575   \n",
       "GLSS-19-0267-TP                  2792.203914                     3845.965090   \n",
       "GLSS-19-0267-R1                  4904.914629                     3993.835349   \n",
       "GLSS-19-0268-TP                   813.639386                     3709.522026   \n",
       "\n",
       "                                  Verhaak_subtype  surgery_number  \\\n",
       "sample_barcode                                                      \n",
       "GLSS-19-0266-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "GLSS-19-0266-R1  VERHAAK_GLIOBLASTOMA_MESENCHYMAL               2   \n",
       "GLSS-19-0267-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "GLSS-19-0267-R1       VERHAAK_GLIOBLASTOMA_NEURAL               2   \n",
       "GLSS-19-0268-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "\n",
       "                 surgical_interval_mo  \n",
       "sample_barcode                         \n",
       "GLSS-19-0266-TP                   0.0  \n",
       "GLSS-19-0266-R1                  18.0  \n",
       "GLSS-19-0267-TP                   0.0  \n",
       "GLSS-19-0267-R1                  10.0  \n",
       "GLSS-19-0268-TP                   0.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea = pd.read_csv('../../data/glass_verhaak_scores.tsv', sep = '\\t', index_col = 1)\n",
    "print(ssgsea.shape)\n",
    "ssgsea.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e2be0ae-3832-49fb-a5db-5c4cff83a80a",
   "metadata": {},
   "source": [
    "Combine duplicated sample barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "11bb92a4-3e47-4fd0-8eb2-5cef6fe8c58a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(419, 6)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea = ssgsea.groupby('sample_barcode').agg({'case_barcode': 'min',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_CLASSICAL' : 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_MESENCHYMAL': 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_PRONEURAL' : 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_NEURAL' : 'mean'})\n",
    "\n",
    "ssgsea['Verhaak_subtype'] = ssgsea.iloc[:, 1:].idxmax(axis=1)\n",
    "ssgsea.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "380dcc00-8286-45a5-a143-1da7b9cd0e0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_CLASSICAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_PRONEURAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_NEURAL</th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>3866.366259</td>\n",
       "      <td>3899.952531</td>\n",
       "      <td>1970.036575</td>\n",
       "      <td>3055.641397</td>\n",
       "      <td>MESENCHYMAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>4499.507042</td>\n",
       "      <td>3725.238591</td>\n",
       "      <td>2028.442349</td>\n",
       "      <td>1945.457473</td>\n",
       "      <td>CLASSICAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3403.500159</td>\n",
       "      <td>1489.313078</td>\n",
       "      <td>3993.835349</td>\n",
       "      <td>4904.914629</td>\n",
       "      <td>NEURAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3998.177157</td>\n",
       "      <td>2373.110794</td>\n",
       "      <td>3845.965090</td>\n",
       "      <td>2792.203914</td>\n",
       "      <td>CLASSICAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>GLSS-19-0268</td>\n",
       "      <td>3271.104740</td>\n",
       "      <td>3855.965091</td>\n",
       "      <td>2225.823185</td>\n",
       "      <td>3986.308979</td>\n",
       "      <td>NEURAL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 case_barcode  VERHAAK_GLIOBLASTOMA_CLASSICAL  \\\n",
       "sample_barcode                                                  \n",
       "GLSS-19-0266-R1  GLSS-19-0266                     3866.366259   \n",
       "GLSS-19-0266-TP  GLSS-19-0266                     4499.507042   \n",
       "GLSS-19-0267-R1  GLSS-19-0267                     3403.500159   \n",
       "GLSS-19-0267-TP  GLSS-19-0267                     3998.177157   \n",
       "GLSS-19-0268-R1  GLSS-19-0268                     3271.104740   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_MESENCHYMAL  \\\n",
       "sample_barcode                                      \n",
       "GLSS-19-0266-R1                       3899.952531   \n",
       "GLSS-19-0266-TP                       3725.238591   \n",
       "GLSS-19-0267-R1                       1489.313078   \n",
       "GLSS-19-0267-TP                       2373.110794   \n",
       "GLSS-19-0268-R1                       3855.965091   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_PRONEURAL  VERHAAK_GLIOBLASTOMA_NEURAL  \\\n",
       "sample_barcode                                                                 \n",
       "GLSS-19-0266-R1                     1970.036575                  3055.641397   \n",
       "GLSS-19-0266-TP                     2028.442349                  1945.457473   \n",
       "GLSS-19-0267-R1                     3993.835349                  4904.914629   \n",
       "GLSS-19-0267-TP                     3845.965090                  2792.203914   \n",
       "GLSS-19-0268-R1                     2225.823185                  3986.308979   \n",
       "\n",
       "                Verhaak_subtype  \n",
       "sample_barcode                   \n",
       "GLSS-19-0266-R1     MESENCHYMAL  \n",
       "GLSS-19-0266-TP       CLASSICAL  \n",
       "GLSS-19-0267-R1          NEURAL  \n",
       "GLSS-19-0267-TP       CLASSICAL  \n",
       "GLSS-19-0268-R1          NEURAL  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea['Verhaak_subtype'] = ssgsea['Verhaak_subtype'].apply(lambda x: x.split('_')[-1])\n",
    "ssgsea.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e84501f-0905-4451-9220-2093fb09ad68",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b24d5c90-508e-4eb9-8c35-29948f196b65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "365 419 246\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Verhaak_subtype\n",
       "CLASSICAL      80\n",
       "MESENCHYMAL    79\n",
       "NEURAL         58\n",
       "PRONEURAL      29\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap = ssgsea.join(memberships, how = 'inner').reset_index()\n",
    "# overlap = overlap[overlap['Verhaak_subtype'].isin(['PRONEURAL', 'MESENCHYMAL'])]\n",
    "print(len(memberships), len(ssgsea), len(overlap))\n",
    "overlap['Verhaak_subtype'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b577a3-1f1d-44cc-9950-1d7d892391e8",
   "metadata": {},
   "source": [
    "## Calculate clustering overlap metrics\n",
    "- Adjusted Rand Index\n",
    "- Adjusted Mutual Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6eb1a6c3-f311-4697-a481-b2a4f81249b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x38b628ec0>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAG2CAYAAABcYt1RAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAAYJJJREFUeJzt3XmczfXix/HXmR2zMZjNGIPCGEszSmNrE1KWrqK6SNS9pLK0IIm0SKXUr3ArS26FbqUkyRKyTMmaZSzZRsxgLDO22c45vz++5nDMDDODOd8Z7+fjcR7mfM7n+53P9xjOez7b12K32+2IiIiImJibqxsgIiIicjkKLCIiImJ6CiwiIiJiegosIiIiYnoKLCIiImJ6CiwiIiJiegosIiIiYnoKLCIiImJ6CiwiIiJiegosIiIiYnrFCiwTJkwgKioKHx8f4uLiWL58eYF1k5OTeeSRR6hTpw5ubm4MHDgw33rffPMN0dHReHt7Ex0dzezZs4vTNBERESmDihxYZs2axcCBAxk+fDjr16+nZcuW3HPPPSQlJeVbPzMzkypVqjB8+HAaNWqUb52EhAS6detGjx492LhxIz169KBr1678/vvvRW2eiIiIlEGWot78sGnTpsTGxjJx4kRHWb169ejcuTNjxoy55LG33347jRs3Zvz48U7l3bp1Iz09nZ9++slR1q5dOypWrMiMGTOK0jwREREpgzyKUjkrK4u1a9cydOhQp/I2bdqwatWqYjciISGBQYMGOZW1bds2T7C5UGZmJpmZmY7nNpuNY8eOERQUhMViKXZbREREpOTY7XZOnjxJWFgYbm4FD/wUKbCkpqZitVoJDg52Kg8ODiYlJaV4LQVSUlKKfM4xY8bwyiuvFPt7ioiIiHns37+fatWqFfh6kQJLrot7MOx2+xX3ahT1nMOGDWPw4MGO52lpaVSvXp39+/fj7+9/RW0RERGRkpGenk5ERAR+fn6XrFekwFK5cmXc3d3z9HwcPnw4Tw9JUYSEhBT5nN7e3nh7e+cp9/f3V2AREREpZS7X8VGkVUJeXl7ExcWxcOFCp/KFCxfSrFmzorfunPj4+DznXLBgwRWdU0RERMqOIg8JDR48mB49etCkSRPi4+P5+OOPSUpKom/fvoAxVHPgwAGmT5/uOGbDhg0AnDp1iiNHjrBhwwa8vLyIjo4GYMCAAbRq1YqxY8fSqVMnvv/+exYtWsSKFSuuwiWKiIhIaVfkwNKtWzeOHj3K6NGjSU5OJiYmhnnz5hEZGQkYG8VdvCfLTTfd5Ph67dq1fPnll0RGRrJ3714AmjVrxsyZM3nppZcYMWIEtWrVYtasWTRt2vQKLk1ERETKiiLvw2JW6enpBAQEkJaWpjksIqWc1WolOzvb1c0QkavA09MTd3f3Al8v7Od3sVYJiYhcC3a7nZSUFE6cOOHqpojIVRQYGEhISMgVrShWYBER08gNK1WrVqV8+fLaBFKklLPb7Zw5c4bDhw8DEBoaWuxzKbCIiClYrVZHWAkKCnJ1c0TkKilXrhxgbFdStWrVSw4PXUqx7tYsInK15c5ZKV++vItbIiJXW+6/6yuZm6bAIiKmomEgkbLnavy7VmARERER01NgERFxkVGjRtG4ceMCn5vJ7bffzsCBA13djAL16tWLzp07l5nvI3kpsIhImWO12UnYdZTvNxwgYddRrLaS2W5q1apVuLu7065du2Id/9xzz7F48eKr1p6SDBnTpk3DYrE4HsHBwXTo0IEtW7aUyPe/nKVLl2KxWLRkvhTTKiERKVPmb07mlR+2kpyW4SgLDfBhZIdo2sUUf0llYUyZMoWnn36aTz/9lKSkJKpXr16k4319ffH19b1Grbv2/P392b59O3a7nQMHDvDCCy9w7733smPHDry8vFzdPCnl1MMiImXG/M3J9Pt8nVNYAUhJy6Df5+uYvzn5mn3v06dP89VXX9GvXz/uu+8+pk2blqfOm2++SXBwMH5+fvTp04eMDOd2XjwklF8PSefOnenVq5fj+YQJE7jhhhvw8fEhODiYBx54ADCGLpYtW8b777/v6PXIvR3K1q1bad++Pb6+vgQHB9OjRw9SU1OdrqVnz574+voSGhrKuHHjCvUeWCwWQkJCCA0NpUmTJgwaNIh9+/axfft2R513332XBg0aUKFCBSIiInjyySc5deqU4/Vp06YRGBjIzz//TL169fD19aVdu3YkJ5//u7NarQwePJjAwECCgoJ44YUXKOqm7Vfr+9jtdt566y1q1qxJuXLlaNSoEV9//bXjtdatW9OuXTvHcSdOnKB69eoMHz68SO0VBRYRKQXOZOUU+MjItgLGMNArP2wlv4+t3LJRP2x1Gh7K73zFNWvWLOrUqUOdOnXo3r07U6dOdfpw++qrrxg5ciSvv/46a9asITQ0lAkTJhT7+wGsWbOGZ555htGjR7N9+3bmz59Pq1atAHj//feJj4/niSeeIDk5meTkZCIiIkhOTua2226jcePGrFmzhvnz53Po0CG6du3qOO/zzz/PkiVLmD17NgsWLGDp0qWsXbu2SG07ceIEX375JWBszZ7Lzc2NDz74gM2bN/PZZ5/xyy+/8MILLzgde+bMGd555x3++9//8uuvv5KUlMRzzz3neH3cuHFMmTKFyZMns2LFCo4dO8bs2bOL/P5dje/z0ksvMXXqVCZOnMiWLVsYNGgQ3bt3Z9myZVgsFj777DNWr17NBx98AEDfvn0JDg5m1KhRRW7v9U5DQiJietEv/1zga3fUqcLUx25h9Z5jeXpWLmTH6GlZvecY8bWMjelajF3CsdNZTvX2vnlvsdo4efJkunfvDkC7du04deoUixcvpnXr1gCMHz+e3r178/jjjwPw2muvsWjRojy9LEWRlJREhQoVuO+++/Dz8yMyMtJxs9mAgAC8vLwoX748ISEhjmMmTpxIbGwsb7zxhqNsypQpREREsGPHDsLCwpg8eTLTp0/n7rvvBuCzzz6jWrVql21PWloavr6+jt1NATp27EjdunUddS7sMYqKiuLVV1+lX79+TuEtOzubSZMmUatWLQCeeuopRo8e7Xh9/PjxDBs2jC5dugAwadIkfv654J+Rglzp9zl9+jTvvvsuv/zyC/Hx8QDUrFmTFStW8J///IfbbruN8PBw/vOf/9CjRw8OHTrEDz/8wPr1651CnBSOAouIlAmHTxbug7+w9Ypi+/btrF69mm+//RYADw8PunXrxpQpUxyBJTExkb59+zodFx8fz5IlS4r9fe+++24iIyOpWbMm7dq1o127dtx///2X3Hxv7dq1LFmyJN+5Mrt27eLs2bNkZWU5PoABKlWqRJ06dS7bHj8/P9atW0dOTg7Lli3j7bffZtKkSU51lixZwhtvvMHWrVtJT08nJyeHjIwMTp8+TYUKFQBjk7HcEAHGdu65W7unpaWRnJzs1D4PDw+aNGlS5GGhK/0+W7duJSMjwxHscmVlZTmCI8CDDz7I7NmzGTNmDBMnTuTGG28sUjvFoMAiIqa3dXTbAl9zO7chVVU/n0Kd68J6K4bccWUNO2fy5Mnk5OQQHh7uKLPb7Xh6enL8+HEqVqxYrPO6ubnl+RC+cKfQ3ICwdOlSFixYwMsvv8yoUaP4448/CAwMzPecNpuNDh06MHbs2DyvhYaGsnPnzmK1Nbe9tWvXBqBu3bqkpKTQrVs3fv31VwD27dtH+/bt6du3L6+++iqVKlVixYoV9OnTx+m6Lu59sFgsRQ4jhXGl38dmswHw448/Ov3dA3h7ezu+PnPmDGvXrsXd3f2K3t/rneawiIjplffyKPDh42ncl+SWqEqEBvhQ0H6aFozVQrdEVbrkeYsqJyeH6dOnM27cODZs2OB4bNy4kcjISL744gsA6tWrx2+//eZ07MXPL1alSpU8k0A3b97sVMfDw4PWrVvz1ltv8eeff7J3715++eUXALy8vLBarU71Y2Nj2bJlCzVq1KB27dpOjwoVKlC7dm08PT2d2nb8+HF27NhR5Pdm0KBBbNy40THvY82aNeTk5DBu3DhuvfVWbrzxRg4ePFikcwYEBBAaGurUvpycnCLPsbka3yc6Ohpvb2+SkpLyvJcRERGOes8++yxubm789NNPfPDBB46/Hyka9bCISJng7mZhZIdo+n2+Dgs4Tb7NDTEjO0Tj7nZ1t/6fO3cux48fp0+fPgQEBDi99sADDzB58mSeeuopBgwYwKOPPkqTJk1o0aIFX3zxBVu2bKFmzZoFnvvOO+9k8ODB/Pjjj9SqVYv33nvPaR+RuXPnsnv3blq1akXFihWZN28eNpvNMXxTo0YNfv/9d/bu3Yuvry+VKlWif//+fPLJJzz88MM8//zzVK5cmb/++ouZM2fyySef4OvrS58+fXj++ecJCgoiODiY4cOH4+ZW9N9v/f39efzxxxk5ciSdO3emVq1a5OTk8H//93906NCBlStX5hkyKowBAwbw5ptvcsMNN1CvXj3efffda7K/yuW+j5+fH8899xyDBg3CZrPRokUL0tPTWbVqFb6+vjz66KP8+OOPTJkyhYSEBGJjYxk6dCiPPvoof/75Z7F73q5X6mERkTKjXUwoE7vHEhLgPDwUEuDDxO6x12QflsmTJ9O6des8YQWgS5cubNiwgXXr1tGtWzdefvllhgwZQlxcHPv27aNfv36XPHfv3r159NFH6dmzJ7fddhtRUVHcccf5YazAwEC+/fZb7rzzTurVq8ekSZOYMWMG9evXB4yN6Nzd3YmOjqZKlSokJSURFhbGypUrsVqttG3blpiYGAYMGEBAQIAjlLz99tu0atWKjh070rp1a1q0aEFcXFyx3p8BAwaQmJjI//73Pxo3bsy7777L2LFjiYmJ4YsvvmDMmDFFPuezzz5Lz5496dWrF/Hx8fj5+XH//fcXq31X+n1effVVXn75ZcaMGUO9evVo27YtP/zwA1FRURw5coQ+ffowatQoYmNjARg5ciRhYWF55jPJ5Vns12Jg0AXS09MJCAggLS0Nf39/VzdHRIooIyODPXv2EBUVhY9P4eajFMRqs7N6zzEOn8ygqp8xDHS1e1auhWHDhrF8+XJWrFjh6qaIXFWX+vdd2M9vDQmJSJnj7mZxLF0uDex2O7t372bx4sVOq0tE5DwNCYmIuFhaWhrR0dF4eXnx4osvuro5IqakHhYRERcLDAwkMzPT1c0QMTX1sIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIuMioUaNo3Lhxgc/N5Pbbb2fgwIGubsZVU6NGDcaPH39Vz3nmzBm6dOmCv78/Fovlmtzf6FpYunRpqWivAouIlD02K+xZDpu+Nv60WS9/zFWwatUq3N3dadeuXbGOf+6551i8ePFVa09Jhoxp06ZhsVioV69ente++uorLBYLNWrUKNI5XRmSihMeP/vsM5YvX86qVatITk7O9/5Srpbfe9qsWTPTtvdC2jhORMqWrXNg/hBIP3i+zD8M2o2F6I7X9FtPmTKFp59+mk8//ZSkpCSqV69epON9fX3x9fW9Rq279ipUqMDhw4dJSEggPj7eUT5lypQivxel0a5du6hXrx4xMTHFPofVasVisRTr7tjF5eXlRUhISIl9v+JSD4uIlB1b58BXPZ3DCkB6slG+dc41+9anT5/mq6++ol+/ftx3331MmzYtT50333yT4OBg/Pz86NOnDxkZGU6vX/xbfX6/DXfu3JlevXo5nk+YMIEbbrgBHx8fgoODeeCBBwDo1asXy5Yt4/3338disWCxWNi7dy8AW7dupX379vj6+hIcHEyPHj1ITU11upaePXvi6+tLaGgo48aNK9R74OHhwSOPPMKUKVMcZX///TdLly7lkUcecarbq1cvOnfu7FQ2cOBAbr/99ku2f9q0aQQGBjod991332GxnL+55a5du+jUqRPBwcH4+vpy8803s2jRokJdQ0Fy2/vOO+8QGhpKUFAQ/fv3Jzs7GzD+rsaNG8evv/6KxWJxXMfx48fp2bMnFStWpHz58txzzz3s3LnTcd7c65k7dy7R0dF4e3uzb98+atSowWuvveb4e4iMjOT777/nyJEjdOrUCV9fXxo0aMCaNWsc5zp69CgPP/ww1apVo3z58jRo0IAZM2Y4XUN+72l+Q0LffPMN9evXx9vbmxo1auT5GahRowZvvPEGvXv3xs/Pj+rVq/Pxxx9f0Xt8OQosImJ+WacLfmSf+9C3WY2eFfK7Af25svlDnIeH8jtfMc2aNYs6depQp04dunfvztSpU7Hbz7flq6++YuTIkbz++uusWbOG0NBQJkyYUOzvB7BmzRqeeeYZRo8ezfbt25k/fz6tWrUC4P333yc+Pp4nnniC5ORkkpOTiYiIIDk5mdtuu43GjRuzZs0a5s+fz6FDh+jatavjvM8//zxLlixh9uzZLFiwgKVLl7J27dpCtalPnz7MmjWLM2fOAMYHcrt27QgODi7StRXU/sI4deoU7du3Z9GiRaxfv562bdvSoUMHkpKSitSGiy1ZsoRdu3axZMkSPvvsM6ZNm+YIpt9++y1PPPEE8fHxJCcn8+233wJGSFizZg1z5swhISEBu91O+/btHUEHjLkvY8aM4dNPP2XLli1UrVoVgPfee4/mzZuzfv167r33Xnr06EHPnj3p3r0769ato3bt2vTs2dPxc5aRkUFcXBxz585l8+bN/Otf/6JHjx78/vvvRXpP165dS9euXXnooYfYtGkTo0aNYsSIEXlC+Lhx42jSpAnr16/nySefpF+/fmzbtu2K3uNL0ZCQiJjfG2EFv3ZDG/jn/2Dfqrw9K07sxuv7VkFUS6NofAM4c9S52qi0YjVx8uTJdO/eHYB27dpx6tQpFi9eTOvWrY1vNX48vXv35vHHHwfgtddeY9GiRXl6WYoiKSmJChUqcN999+Hn50dkZKTjbs8BAQF4eXlRvnx5p+7+iRMnEhsbyxtvvOEomzJlChEREezYsYOwsDAmT57M9OnTufvuuwFjbka1atUK1abGjRtTq1Ytvv76a3r06MG0adN499132b17d5GuraD2F0ajRo1o1KiR4/lrr73G7NmzmTNnDk899VSRznWhihUr8uGHH+Lu7k7dunW59957Wbx4MU888QSVKlWifPnyTsMrO3fuZM6cOaxcuZJmzZoB8MUXXxAREcF3333Hgw8+CEB2djYTJkxwajNA+/bt+fe//w3Ayy+/zMSJE7n55psdxw0ZMoT4+HgOHTpESEgI4eHhPPfcc47jn376aebPn8///vc/mjZtWuj39N133+Wuu+5ixIgRANx4441s3bqVt99+26l3r3379jz55JOOtrz33nssXbqUunXrFvs9vhT1sIhI2XDq0NWtVwTbt29n9erVPPTQQ4AxNNKtWzenoZHExESneR1AnudFdffddxMZGUnNmjXp0aMHX3zxhaNnoyBr165lyZIljvkyvr6+jg+YXbt2sWvXLrKyspzaVqlSJerUqVPodvXu3ZupU6eybNkyR29HSTp9+jQvvPAC0dHRBAYG4uvry7Zt2664h6V+/fq4u7s7noeGhnL48OEC6ycmJuLh4UHTpk0dZUFBQdSpU4fExERHmZeXFw0bNsxz/IVluT1UDRo0yFOW2war1crrr79Ow4YNCQoKwtfXlwULFhT5uhMTE2nevLlTWfPmzdm5cydW6/keygvbZ7FYCAkJueT7caXUwyIi5vfiJXpOLOc+QHwLOeRwYb2Bm4rfpgtMnjyZnJwcwsPDHWV2ux1PT0+OHz9OxYoVi3VeNzc3p2ElwGkowc/Pj3Xr1rF06VIWLFjAyy+/zKhRo/jjjz/yzPPIZbPZ6NChA2PHjs3zWmhoqNP8iuL65z//yQsvvMCoUaPo2bMnHh55P2oud20FKcxxzz//PD///DPvvPMOtWvXply5cjzwwANkZWUV42rO8/T0dHpusViw2WwF1r+4nReWXzjnply5ck7P8/t+ua/nV5bbhnHjxvHee+8xfvx4GjRoQIUKFRg4cGCRr/vi9hV0LUV9P66UelhExPy8KhT88PQx6kQ2M1YDkfc/foMF/MONepc6bxHl5OQwffp0xo0bx4YNGxyPjRs3EhkZyRdffAFAvXr1+O2335yOvfj5xapUqUJycrLjudVqZfPmzU51PDw8aN26NW+99RZ//vkne/fu5ZdffjEuz8vL6TdigNjYWLZs2UKNGjWoXbu206NChQrUrl0bT09Pp7YdP36cHTt2FPo9qVSpEh07dmTZsmX07t27UNcGsGHDBqfn+bW/SpUqnDx5ktOnTxd43PLly+nVqxf3338/DRo0ICQkxDHhuCRFR0eTk5PjmEMCxsTYHTt25Lv8+0otX76cTp060b17dxo1akTNmjXzBND83tP82r1ixQqnslWrVnHjjTc69TCVNAUWESkb3NyNpctA3tBy7nm7N416V9HcuXM5fvw4ffr0ISYmxunxwAMPMHnyZAAGDBjAlClTmDJlCjt27GDkyJFs2bLlkue+8847+fHHH/nxxx/Ztm0bTz75pNNKjrlz5/LBBx+wYcMG9u3bx/Tp07HZbI7hmxo1avD777+zd+9eUlNTsdls9O/fn2PHjvHwww+zevVqdu/ezYIFC+jduzdWqxVfX1/69OnD888/z+LFi9m8eTO9evUq8jLbadOmkZqaWuB8hjvvvJM1a9Ywffp0du7cyciRI/OEsfza37RpU8qXL8+LL77IX3/9xZdffplnMmjt2rX59ttvHcHxkUceuaa/+RfkhhtuoFOnTjzxxBOsWLGCjRs30r17d8LDw+nUqdNV/361a9dm4cKFrFq1isTERP7973+TkpLiVCe/9/Rizz77LIsXL+bVV19lx44dfPbZZ3z44YdO82NcQYFFRMqO6I7QdTr4hzqX+4cZ5ddgH5bJkyfTunXrfDfd6tKlCxs2bGDdunV069aNl19+mSFDhhAXF8e+ffvo16/fJc/du3dvHn30UXr27Mltt91GVFQUd9xxh+P1wMBAvv32W+68807q1avHpEmTmDFjBvXr1weMjejc3d2Jjo6mSpUqJCUlERYWxsqVK7FarbRt25aYmBgGDBhAQECAI5S8/fbbtGrVio4dO9K6dWtatGhBXFxckd6XcuXKERQUVODrbdu2ZcSIEbzwwgvcfPPNnDx5kp49ezrVya/9lSpV4vPPP2fevHmOZbujRo1yOu69996jYsWKNGvWjA4dOtC2bVtiY2OL1P6rZerUqcTFxXHfffcRHx+P3W5n3rx5eYZTroYRI0YQGxtL27Ztuf322wkJCcmzdDy/9/RisbGxfPXVV8ycOZOYmBhefvllRo8e7TTh1hUs9oIG2UqZ9PR0AgICSEtLw9/f39XNEZEiysjIYM+ePURFReHj43NlJ7NZjdVApw4Zc1Yim131npVrYdiwYSxfvjxPd7xIaXepf9+F/fzWpFsRKXvc3M8vXS4F7HY7u3fvZvHixY5lySLiTENCIiIulpaWRnR0NF5eXrz44ouubo6IKamHRUTExQIDA8nMzHR1M0RMTT0sIiIiYnoKLCJiKmVkHYCIXOBq/LtWYBERU8hd5nm5reVFpPTJ/Xd9Jcu5NYdFREzB3d2dwMBAx71Iypcvn+925SJSetjtds6cOcPhw4cJDAy8op1yFVhExDRy7yB7LW+gJiIlLzAwsMh33b6YAouImIbFYiE0NJSqVasW6kZ4ImJ+np6eV+UeRAosImI67u7uLr3JmoiYjybdioiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpKbCIiIiI6SmwiIiIiOkpsIiIiIjpFSuwTJgwgaioKHx8fIiLi2P58uWXrL9s2TLi4uLw8fGhZs2aTJo0KU+d8ePHU6dOHcqVK0dERASDBg0iIyOjOM0TERGRMqbIgWXWrFkMHDiQ4cOHs379elq2bMk999xDUlJSvvX37NlD+/btadmyJevXr+fFF1/kmWee4ZtvvnHU+eKLLxg6dCgjR44kMTGRyZMnM2vWLIYNG1b8KxMREZEyw2K32+1FOaBp06bExsYyceJER1m9evXo3LkzY8aMyVN/yJAhzJkzh8TEREdZ37592bhxIwkJCQA89dRTJCYmsnjxYkedZ599ltWrV1+29yZXeno6AQEBpKWl4e/vX5RLEhERERcp7Od3kXpYsrKyWLt2LW3atHEqb9OmDatWrcr3mISEhDz127Zty5o1a8jOzgagRYsWrF27ltWrVwOwe/du5s2bx7333ltgWzIzM0lPT3d6iIiISNnkUZTKqampWK1WgoODncqDg4NJSUnJ95iUlJR86+fk5JCamkpoaCgPPfQQR44coUWLFtjtdnJycujXrx9Dhw4tsC1jxozhlVdeKUrzRUREpJQq1qRbi8Xi9Nxut+cpu1z9C8uXLl3K66+/zoQJE1i3bh3ffvstc+fO5dVXXy3wnMOGDSMtLc3x2L9/f3EuRUREREqBIvWwVK5cGXd39zy9KYcPH87Ti5IrJCQk3/oeHh4EBQUBMGLECHr06MHjjz8OQIMGDTh9+jT/+te/GD58OG5ueXOVt7c33t7eRWm+iIiIlFJF6mHx8vIiLi6OhQsXOpUvXLiQZs2a5XtMfHx8nvoLFiygSZMmeHp6AnDmzJk8ocTd3R273U4R5wSLiIhIGVTkIaHBgwfz6aefMmXKFBITExk0aBBJSUn07dsXMIZqevbs6ajft29f9u3bx+DBg0lMTGTKlClMnjyZ5557zlGnQ4cOTJw4kZkzZ7Jnzx4WLlzIiBEj6NixI+7u7lfhMkVERKQ0K9KQEEC3bt04evQoo0ePJjk5mZiYGObNm0dkZCQAycnJTnuyREVFMW/ePAYNGsRHH31EWFgYH3zwAV26dHHUeemll7BYLLz00kscOHCAKlWq0KFDB15//fWrcIkiIiJS2hV5Hxaz0j4sIiIipc812YdFRERExBUUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9IoVWCZMmEBUVBQ+Pj7ExcWxfPnyS9ZftmwZcXFx+Pj4ULNmTSZNmpSnzokTJ+jfvz+hoaH4+PhQr1495s2bV5zmiYiISBlT5MAya9YsBg4cyPDhw1m/fj0tW7bknnvuISkpKd/6e/bsoX379rRs2ZL169fz4osv8swzz/DNN9846mRlZXH33Xezd+9evv76a7Zv384nn3xCeHh48a9MREREygyL3W63F+WApk2bEhsby8SJEx1l9erVo3PnzowZMyZP/SFDhjBnzhwSExMdZX379mXjxo0kJCQAMGnSJN5++222bduGp6dnsS4kPT2dgIAA0tLS8Pf3L9Y5REREpGQV9vO7SD0sWVlZrF27ljZt2jiVt2nThlWrVuV7TEJCQp76bdu2Zc2aNWRnZwMwZ84c4uPj6d+/P8HBwcTExPDGG29gtVoLbEtmZibp6elODxERESmbihRYUlNTsVqtBAcHO5UHBweTkpKS7zEpKSn51s/JySE1NRWA3bt38/XXX2O1Wpk3bx4vvfQS48aN4/XXXy+wLWPGjCEgIMDxiIiIKMqliIiISClSrEm3FovF6bndbs9Tdrn6F5bbbDaqVq3Kxx9/TFxcHA899BDDhw93Gna62LBhw0hLS3M89u/fX5xLERERkVLAoyiVK1eujLu7e57elMOHD+fpRckVEhKSb30PDw+CgoIACA0NxdPTE3d3d0edevXqkZKSQlZWFl5eXnnO6+3tjbe3d1GaLyIiIqVUkXpYvLy8iIuLY+HChU7lCxcupFmzZvkeEx8fn6f+ggULaNKkiWOCbfPmzfnrr7+w2WyOOjt27CA0NDTfsCIiIiLXlyIPCQ0ePJhPP/2UKVOmkJiYyKBBg0hKSqJv376AMVTTs2dPR/2+ffuyb98+Bg8eTGJiIlOmTGHy5Mk899xzjjr9+vXj6NGjDBgwgB07dvDjjz/yxhtv0L9//6twiSIiIlLaFWlICKBbt24cPXqU0aNHk5ycTExMDPPmzSMyMhKA5ORkpz1ZoqKimDdvHoMGDeKjjz4iLCyMDz74gC5dujjqREREsGDBAgYNGkTDhg0JDw9nwIABDBky5CpcooiIiJR2Rd6Hxay0D4uIiEjpc032YRERERFxBQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETG9YgWWCRMmEBUVhY+PD3FxcSxfvvyS9ZctW0ZcXBw+Pj7UrFmTSZMmFVh35syZWCwWOnfuXJymiYiISBlU5MAya9YsBg4cyPDhw1m/fj0tW7bknnvuISkpKd/6e/bsoX379rRs2ZL169fz4osv8swzz/DNN9/kqbtv3z6ee+45WrZsWfQrERERkTLLYrfb7UU5oGnTpsTGxjJx4kRHWb169ejcuTNjxozJU3/IkCHMmTOHxMRER1nfvn3ZuHEjCQkJjjKr1cptt93GY489xvLlyzlx4gTfffddoduVnp5OQEAAaWlp+Pv7F+WSRERExEUK+/ldpB6WrKws1q5dS5s2bZzK27Rpw6pVq/I9JiEhIU/9tm3bsmbNGrKzsx1lo0ePpkqVKvTp06dQbcnMzCQ9Pd3pISIiImVTkQJLamoqVquV4OBgp/Lg4GBSUlLyPSYlJSXf+jk5OaSmpgKwcuVKJk+ezCeffFLotowZM4aAgADHIyIioiiXIiIiIqVIsSbdWiwWp+d2uz1P2eXq55afPHmS7t2788knn1C5cuVCt2HYsGGkpaU5Hvv37y/CFYiIiEhp4lGUypUrV8bd3T1Pb8rhw4fz9KLkCgkJybe+h4cHQUFBbNmyhb1799KhQwfH6zabzWichwfbt2+nVq1aec7r7e2Nt7d3UZovIiIipVSReli8vLyIi4tj4cKFTuULFy6kWbNm+R4THx+fp/6CBQto0qQJnp6e1K1bl02bNrFhwwbHo2PHjtxxxx1s2LBBQz0iIiJStB4WgMGDB9OjRw+aNGlCfHw8H3/8MUlJSfTt2xcwhmoOHDjA9OnTAWNF0IcffsjgwYN54oknSEhIYPLkycyYMQMAHx8fYmJinL5HYGAgQJ5yERERuT4VObB069aNo0ePMnr0aJKTk4mJiWHevHlERkYCkJyc7LQnS1RUFPPmzWPQoEF89NFHhIWF8cEHH9ClS5erdxUiIiJSphV5Hxaz0j4sIiIipc812YdFRERExBUUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0FFhERETE9BRYRERExPQUWERERMT0PFzdABERkZJmtdlZvecYh09mUNXPh1uiKuHuZnF1s+QSFFhEROS6Mn9zMq/8sJXktAxHWWiADyM7RNMuJtSFLZNL0ZCQiIhcN+ZvTqbf5+ucwgpASloG/T5fx/zNyS5qmVyOAouIiFwXrDY7r/ywFXs+r+WWvfLDVqy2/GqIqymwiIjIdWH1nmN5elYuZAeS0zJYvedYyTVKCk2BRUREyrxsq42fCjncc/hkwaFGXEeTbkVEpMzrNXU1K/86Wqi6geU8r3FrpDjUwyIiImXOntTTZGRbHc/bxYQSVMETX++Cf0+3AJ7uFl7+fjO/7S5cuJGSo8AiIiJlgt1uZ9WuVB7/7A/uHLeU79YfcLzWtUk1Vg27i3cebIgFI5xcyIIxh6WClwf7jp3loY9/Y/jsTZzMyC7BK5BL0ZCQiIiUapk5Vn7YmMzkFXtITE53lG8/dNLxtbeHO2D0tEzsHsurczYRcWojVTnBYQLZ79uIER0b0Kx2Zd78aRtf/p7EF78n8cu2w7x+fwx31g0u8esSZxa73V4m1m+lp6cTEBBAWloa/v7+rm6OiIhcY3a7nQ9/+Yvpv+3jyMlMAHw83XggrhqPNY+iVhXf/A/cOgf7/CFY0g+eP5d/GJZ2YyG6IwAJu44y9Ns/2Xf0DACdG4fx2v0NLjmkJMVT2M9vvfMiYmraQl0KYrFYWJt0nCMnMwnx96Fns0geuaU6geW9Cj5o6xz4qieWi3ZjsaQnw1c9oet0iO5IfK0g5g9oxXuLdvDp8t3sST2Nj4dmUbiSelhExLS0hbrkstvt/LozlWkr9/DGPxoQGlAOgHVJx9l/7AztG4Ti6X6ZQGGzwvgYuKBnxZkF/MNg4CZwc3eUbtx/gnJe7twY7AdARraVE2eyCQnwuRqXdt0r7Oe34qKImJK2UBcwwsGM1Um0ee9XHp2ymiXbj/DZqn2O12OrV6RT4/DLhxWAPcsvEVYA7JB+APatciptFBHoCCsAHyzeyd3vLuPL35OwaVfcEqMhIRExncttoW7B2EL97ugQDQ+VUYdPZvB5wj4+/z2JY6ezAPD19qBrkwgeuaV64U5izYGcs+B9Lmz8tbhwx506VPApbXb+2HuMk5k5vDh7E3M2HuDNfzSkRuUKhTu3FJsCi4iYTlG2UI+vFVRyDZMSkZljpc17v3LijLGkuFrFcvRqVoOuN0fg73OJTd2sOZC8EfYuh70rICkBmvaFu0YYr9doBgkfXL4BvgWvCHJ3szDzX/FMW7WXd37ezm+7j9Hu/V959u469G4RpQB9DSmwiIipHD+dxbfr/y5U3dV7j2oSbhlgs9n5bc9R4msGYbFY8PZwp1OjMDYfTOfxFlHcHR2MR0FDPtkZ8Puk8wEl65Tz6wfXn//6hjbGHJX0ZMi3/+7cHJbIZrD9J6jdGtzzBiR3Nwt9WkRxd71ghs3+k5V/HeX1eYnM/fMg73VrTM2CVifJFdGkWxFxuYxsK4sTDzN7/QGW7ThMtrXw/y1VLO/Jkuduv/TKEDGl05k5fLPub6as2MPeo2f4pl88cZGVAMjKseF18aocaw6kbIRTR6BOO6PMZoO3a8LZ48Zzn0Co0eL8o2p9cLvgPOdWCRku/Dk7F3q7TodyFeGz+6ByHbjnTah1Z4HXYLfb+d+av3n1R+MuzwsGtaJaxfLFfk+uR1rWLCKmZrXZ+X33UWavP8D8zSmczMxxvFYvxI/k9AzSzmTn+3swGPtteLhZCPL1dgorb83fRgVvD26vU4XoUH8sFvW+mM3BE2f5LGEvM35PIj3D+Hv38/Fg/7GzxEUadbw83M4HlL0rjMe+BMg6Cf7hcGNbsFiMMNJ8ALh7GwElOMY5oFwsuqMRSuYPcZ6A6x8G7d40Xk+cC+WDIHU7/Pd+qHsftHkNKkXlOZ3FYqHrzRHcVqcKWw+mO4WV/cfOEFFJ4eVqUQ+LiLhE6qlMmr6xGOu5VRbhgeXo1DiMzjeFc2Own2OVEOT7ezATu8dyV71gUtIyHB8KmTlWGr+ykLPn7iET7O/NHXWqcnudqrS4obI2/XKxE2eyGPH9FuZtSnb8vdcIKs9jzaN4IK4aFS78+/lpCKz/wggoF/IJgMjm8I9PwPsKhl5sVmM10KlDxpyVyGZOS5k5ewKWjYXf/wN2qxGImj8DLQaB1+Un2K76K5UeU1bTq1kNnm1zI+W99LNXkMJ+fiuwiMg1d+DEWb7fcICko2d4s0tDR/mAmesp7+VB58Zh3FyjEm4XzUUp6j4sZ7OsfLv+b5ZsO8LKv1IdwQWMm9o9Gl+Dl+6LvgZXKIVhtdm5/Z0l7D92lltrVuLxZtW5IyAF96QVRnh4cBp4Gvur8NNQ+H3i+YCSO8QTHOMcLK61w9vgpxdgzzLjebWboc9Co3fnEt6av40JS3cBEFGpHG/+oyHNa1e+1q0tlRRYRMSl0s5kM29zMrPXH2D1nmOO8pVD7yQ8sFyhz1PcnW4zsq2s3nOMJdsPs3T7EfaknmZIu7r0u70WAMdOZ/H+oh3cXrcq8TWD8PEswQ/B68DJjGxm/bGfnzanMOOJWx1DPOtW/0roiTWEHltjTJLNPH/vHx79AaJaGV8f3WVMoC3pgJIfux22zYWfX4S7RkKDBwp12JLthxn+7SYOngvcD90cwbD29Qgod4mVTtchBRYRcYk/9h7j0+W7WbLtCFlWG2D8Mto0qhL33xTOvQ3DXDI0syf1NL7eHlTx8wZg9vq/GTRrI2DMh2lWqzJ31KnC7XWqat7BFdh/7AxTV+7lqzX7OZuZiTs2xnZrwv03VYPfJsL8oc4HeAdAjXM9KNGdISDcJe0ulOwM8PA+37uyYQb8vRruHAHlK+V7yMmMbN6av53//mZsdlfVz5uxXRpyR92qJdVq09OkWxEpETabnSyrzdFDsf/YGX7eYmy8VTfEj843hdOxURhhRehVuRaiLtrYq1YVXx5pWp0l2w6TnJbBL9sO88u2w8AWalf1ZdyDjWgUEeiStpY2drudNfuOM+XXvzi47XdusWzlfbdEbvXZzp8NXqRx/Q5GxchmzgHFFUM8V8Lzgq34szNg4ctw+jBs/hbufAniHgN3549VPx9PXu0cw30NQxn67Sb2pJ7m1AUTzKXw1MMiIsWyPeUks9cfYM6GAzzarAb/vs0YajmdmcP//fIXnRqHUS/U/P8W7XY72w+dZMm2IyzZfpi1+44bw1Av3kVVf+MD6pdth0g9mcXtdao4yuScM8dIXTGFDcvncovbNvwtZ51fj30UOp7brM1mA+ylJ6Bczt4VxuTgQ5uN51Xrwz1jIaplvtUzsq38sPEgD8RVc6xeO3DiLGEBPtf1ajYNCYnIVZecdpY5Gw7y3YaDJCafn3twc42K/K9vMxe27OpJO5PNuv3HuaPO+S777p/+zoq/UgGoH+bPnXWNlUeNIwKvr03rbFZI+ZNTp8/yh7W2MaxxOhXeruWoYvXyx/3CHpSQBmUnoOTHmgPrpsEvr53fCya6s7EMOjDikoceP51F63eX0aBaAK/f36BIc7vKEgUWEblq7HY7fT5bw5Lth8n9H8PT3cIddapy/03h3FG3apmetPrRkr9YsPUQf/59ggv/x6xY3pM20SG82aVB2fwN+VxAyd0HxbZ3JW5ZJ/nNXp+eOSNYOfROY07Qj89hrxiJpUbLsh9QCnLmGCx5A9ZMBrsN+iyCiJsveciirYd48ot1ZFltVPByZ+g9dfln08g8q+XKOs1hEZFiy8qx8cfeY45lmBaLhXJe7tjtcEuNSnS+KZz2DUKum91l+99Rm/531Cb1VCbLthtDR7/uOMLxM9kkp2c4hZXPf9tH44hA6oeV8k3rvnkCdsx3WsXjBqTby5Nq86Vm5fIcSs8wAsu971CKr/TqKF8J7n0H4nrBrl+cw8qR7VD5xjxLoVtHBzNvQEuGfPMna/cdZ8T3W/hhYzJvdmmg7f3zoR4WEQGMybNrk44ze/0B5m1K5sSZbBYNbkXtqsadbncfOYWXh5u2HT8nx2pjXdIJ3N1wbCefkpbBrWOMOwJX9TM2rbujbhVa3FClZFZGXW4ztPzq5/agHNkOnT48/9oXXWHnz5yyVCAhpw6/2erxmy2a0Bub0LtlbeJrBZXuQFZSju2Bj5pC9abQbiwE590HyGaz89/f9jF2/jbOZFnx8nBjUOsb6XtbzeviPdaQkIgUyl+Hjcmz3284yN/Hz0+YrOrnzdgHGjrN5ZBL++vwKcbO38bKv1I5k+W8ad3NNSrxRKua1+793DqngO3mxxrbzUOeIR72JUBm2vn6g7Zg9w83PiQPrOPE2Rzipx0BizsPxFXjseY19Jt/UW3+Bmb3A2smWNzh5sfhjmHG/Yousv/YGV6cvYnlO1Pp2qQabz3QyAUNLnkKLCJyWct3HqHH5NWO577eHrSLCaFz43DiawVdXxNKr6LMnHOb1p1bebQn9TQAHzx8Ex0bhQHGh9POwyeJr1mZcl5XOOfDcUO/i/87v+CGftEdjRUtv09yruLtD9XjOVL5ZiaduIUdp3z4b5+mjpeXbj9M44jA62b475o4vhcWvASJPxjPy1WCu16G2J55esDsdjvfbTjAnXWDHRvMHTudRQVvd7w9yubcIAUWEXFyMiObn7ccws0C/4itBhgfrM3f/IXGEYF0ahxO63rBV/7hKXnsST3N0u2Huf+mcMcH/0dL/uLtn7fj7eFGfK0g7qxblTuKs2mdzQrvxcDJgwXX8Q+HgZtg6/fwwwCoHg81WmCv0YJf00P4dNV+lu9MdVT/eWAr6oT4FedS5VJ2LTE2zjuyzXge0RQem3/JmzXa7XZ6Tf2DAyfOMrZLA8fwY1miSbciQrbVxq87jjB7/QEWJR4iI9tG9Urluf8mo9vf28OdFUPuLNMrfMwgqnIFoio73+nX28ONsAAfDqZlsHT7EZZuPwJsoVaVCtxTy4e+9bPxtZ40lsqePXbuz3OPuF5Q83bjRCs/uHRYAUg/YMxtqXsf1OtIhs3C7PUHmDJrDzsPJwPgZoE20SH0aRnFjcEa9rkmat0BfVfAH5ONFUW17rz0naWB5LQMthxMJ/VUJg9MSuDR+Bo837aO840irxPX3xWLXAc27j/BN+v+Zu6fyRw7neUor1m5Ap1vCifLanN0L5s+rBR1IqkZ5GQawcLL9/wdhY/ugm0/OgLI42eP0yfsOJm+R8k+dZRPyz3Gh4cbsuvIaXanrcR3w1sFnz/i1vOBJSej4HoXOnUIPIzenYWbDzLs202AMQzYtUkEjzWvoVsSlAR3T7i1r3E/Is8L3u+k32DfSri1v9OOumGB5Vg0uBWv/ZjI12v/ZtqqvSzceogx/2hAqxuruOACXEeBRaQM+vL3JGat2Q9AZV9vOjQK5f6bwmkQHlC6Vh0UZiLptZSdkbeHI7Tx+Q3B9v8BCf9n7MFx9sT5OtnGnBXu/w80esj4OnUnLBzhdHoL4HPuMeiOAHo3upvlO49g/9sGf9U0JmaWq8jCvdn8neGDl28lQkLCCPVuQB2b3ZhjFHFroS5lw3FvGp/7ul1MCDfXqEjb+iF0vTkCfx/djK/EVbjgzs02K8x7DlI2wbr/Qts3oM49jmXQgeW9eOfBRnRsFMawbzdx4MRZek5ZzQNx1RjZIRq/6+TvT3NYREqxIycz+WHjQb7fcIBXOsXQ+Ny9b/7Ye4wvf0+i803hNK8VhIf7pbudTamwE0kLIycLzqSeDxRnLhpiafAghMQYdbf/BHMHG+U5Z/Oeq/MkaPyw8fWOn+HLrvl/T4sbtH/bWBUCcHgbrHjXmHB5Log4HuUrQmANqBCU5zQnM7LpMXk1Gy/atC6wvCetbqhChwbBtPjxDrzPpJDfHGmbHVIIopvPf1jywl2l82ehrLPbYdP/YMEIOJVilNW6C9q9CVVudKp6OjOHt3/ezmcJe4kKqsC8AS3N30t6GZp0K1JGnc7MYcHWFGavP8jKv1Kx2ox/wr2a1WBUx/oubt1VYrPC+BjnnhUnFmN4qN2bxrLciwPI2RPG0tEaLYzqG2fC7H8X/P0uFUIs7s7hovkAqHef8Vra30bAKVcRygVeUK+SsfrmMvMTiiL1VCa/7jjCku1HWLb9MOkZxg30ut0cgSXxB97INoaQLgwt53406Jc9EM+YTrzaKYaKFbTax7QyT8LycZDwEVizwM0DmvaF214AnwCnqmv3HcNisRBb3VgebbXZOX4mi8q+3q5o+RVRYBEpY06cyWLknC0s2HKIs9nn9/hoHBHI/TeFc2/D0FL5n1W+9iyHz+67snN0ngiNHzG+3rEAZjx0rjcjnx6OmC5QrYlR9+wJOL7n/Gve/nl2KHW1HKuN9ftPsGTbYYIqePHqj4m0dVvNSM/phFmOOeodtAfxSnYPfrbdwownbiW+Vt4eHDGho7vg5xeNnYYB/vEJNCygJ++cT37dzYdL/uKle+s53VyxNNAqIRETs9rsrN5zjMMnM6jq58MtUZXy7Hlit9s5lJ5JSIAxAc/Px5OEXUc5m22lRlB5Ot8UTqfG4URVruCKS7i2Th0qXL1KtY0uc0f4CDzfw5EbQABqt4aXjxYueJQLhHI3FafVJcbD3Y2ba1Ti5hqV+H7DAQB+tt3Cwswm3OK2jaqc4DCBrLbVxYbRy3P4ZCEn54rrBdWCR2bBzoWw6WuIeeD8axlpeXpbbDY7C7amkHY2m+e//pM5Gw/yxv0NytwkagUWkRI2f3Myr/ywleS08x8goQE+jOwQTbuYUPamnua7DcbOs6cyc/ht2F24u1lwd7PwaucYgv19aFStlE2eLapyhewJ6DAeolpevt5VHJoxm6p+51eU2HDjN1verd8vrielxA13G49cmSeNbf5r3gGtR4FfMABubhZmPHErnyzfw3uLdrB8Zyptx//KC23r0DO+Rpm5maKGhERK0PzNyfT7fF2eaaS5oiqXZ0/qGcdzH083vuvfnLoh19HP9OlUmNUDklZdopLFWC00cJP5lzhfY1abnRZjfyElLSPfnysLEBLgw4ohd2rn4tJuy2z4Xy/jay8/Y25L076O5epg3PNr6DebWL3XGBpsElmRtx9sZOqe2MJ+fpfdXztETMZqs/PKD1sLDCsAe1LPYAFa3ViF97o1Yu1Ld19fYQXgr0VGWPHI7RG4+EP23PN2b173YQXA3c3CyA5Gr0oB7xQjO0QrrJQF9e+HxxdDWCxknTSWyU+MN4aOzqlZxZeZ/7qVVzvVp4KXOxv/PkFWjs2Fjb561MMico1kZFtJTssg+cRZDpw4y+o9x/jf2r8ve9zE7rHcExNaAi00sV/fgXodjDsI59mHJdwIKyWxD0spcrmhRilDbDbYOAMWjYLTh42yG9tB1/869bYcOHGWNXuP0alxuKPsyMlMqviZa3K+Jt2KXEM5VhuHT2aSnHaWgycyOHjiLPXDAmhxg7EZ1LaUdNqNX16sc5eV34YKzWaD3z6Cxv80VvAAtHrO+LNKHah7b+nb6dYF2sWEcnd0yGUnc0sZ4OYGN/3TWF6/7C3jhpbefk5hBSA8sBzhF4SVzQfS+MfEVfRpEcWAu24odfu3KLCIXMRut3P8TDYHT5zF19uDGufGfg+cOMuAGes5eOIsh05mOvY/ydUzPtIRWEL8jeGMcp7uhAX6EBZYDncLLN2RyuVcV5MjM0/C7L6wbS7sXAA9vs87QdbNvXATawV3N4uWLl9PfAKg7esQ+6gRWHKdSIJ9CcZS6Asm5/+8JYWsHBsTl+7i580pjH2gITfXKD03U1RgkeuONXdLcyDtTDaTV+4h+cRZktOMnpKDaWfJyDZ6OR6Nj+SVTsYOqBW83Fmz77jjPB5uFoL9fQgPLEdooA+NqgU6Xgso58n6EXcTWN7TsZqnsJMjb4kqPf+BXJFju2HGI3AkEdy9oNHDZXo1j8g1c9FuuCwYAVu/gzWT4Z6xEGYs03+2TR1iwgMY8d1mdqee5sFJCfSMj+SFdnXxLQU3UyxWCydMmMDbb79NcnIy9evXZ/z48bRsWfBvQMuWLWPw4MFs2bKFsLAwXnjhBfr27et4/ZNPPmH69Ols3rwZgLi4ON544w1uueWW4jRPrnNZOTbWJx03AkjaWQ6eOEvyiQwOnAslnRqHMfpcCLFj54PFO/M9T2Vfbzwv2MY8oJwnE/4ZS0iAEVIq+3oX2N1usVjy7CiaOzmy3+frsOC84fx1Nzly91JjtcPZ4+AbAg994bxviogUj90OoY2Mibj7f4eP74DYnnDXy1ChMm3rh3BrzSDe+DGRWWv2Mz1hH4u2HmJMl4bcZvKbKRY5sMyaNYuBAwcyYcIEmjdvzn/+8x/uuecetm7dSvXq1fPU37NnD+3bt+eJJ57g888/Z+XKlTz55JNUqVKFLl26ALB06VIefvhhmjVrho+PD2+99RZt2rRhy5YthIeH5zmnmE9hNkK7UjabndTTmSSfyO0JMf5MTjvLTREVeaJVTQDOZlnp9vFvBZ7n4InzkxIDynnS49ZIgv29CQ0oR1hgOcICfQgJ8HHczTiXxWKhfYMrm7zYLiaUid1j80yODLleJkfa7fD7f4xdPO1WCI+Dbl+Afxm/bpGSYrFAy8HGTTcXjoRNX8G6z2DLd8btKm5+nIBynox9oCEdG4cx9Ns/2X/sLPuOngbMHViKvEqoadOmxMbGMnHiREdZvXr16Ny5M2PGjMlTf8iQIcyZM4fExERHWd++fdm4cSMJCQn5fg+r1UrFihX58MMP6dmzZ6HapVVCrnO1ViekZ2RfEEbOEuznQ+toY2OkkxnZxL26iCxr/hNSW9eryqeP3gwYc1DueX85geU9jQByLoiEBvqc+9rH5Xc3LYmAZ0qZp2BSczi+Fxo+BB3eB8/raM6OSEnblwA/vQApfxrP27wOzZ5yvHwmK4eZq/fTq9n5DeaOnsqkUgWvEtuc8pqsEsrKymLt2rUMHTrUqbxNmzasWpX/Jk8JCQm0adPGqaxt27ZMnjyZ7OxsPD3zfnCcOXOG7OxsKlUqeCw/MzOTzMxMx/P09PSiXIpcJQVthJaSlkG/z9cxsXss7WJCyci2kpKWgc1up2YVX8BY9vvv/651rLQ5lZnjdI7W9ao6Aouvtwce7hZybMak1NBzE1nDAnwIDShH3ZDzE84sFgvzB7a6ptd9pa7byZHevvDwTNi1BG7tZ7p79IiUOZHx8K+lsP6/8MdkaPLY+ddsNsp7edC7RZSj6ExWDp0nrKReiD+vdY6hqr+PaX7BKlJgSU1NxWq1Ehwc7FQeHBxMSkpKvsekpKTkWz8nJ4fU1FRCQ/P+Bj506FDCw8Np3bp1gW0ZM2YMr7zySlGaL1fZpTZCyy17esZ6/Lw3cexMNuDcE+Lt4cYfe49xJuv8jfwCy3sSGlCO8EAfYiMrOsotFgtLnrudShW8nOaVSClwcD2k7jx/87aq9YyHiJQMN3eI62WsJsr9JcFmg2n3Gnc0bzEQvIzVkH/sPU7yiQz2HztLwu6jdG4czsLEQ6SYYH+fYk26vbibyG63X7LrKL/6+ZUDvPXWW8yYMYOlS5fi41NwV/GwYcMYPHiw43l6ejoRERGFar9cGavNzp7UU8xef8BpGCg/2Va7I6yU83TH7YK/c4vFwtsPNMK/nMe5+SM+lPcq+Ecy2F9DB6XOpq/h+/5gy4GKNSBCE+lFXObCz9ydC4wdpZNWwYYvoM2rUP8f3HZjFeY+04IXvv6TP/9O47+/7ctzmot70EtKkQJL5cqVcXd3z9Obcvjw4Ty9KLlCQkLyre/h4UFQkHOX+DvvvMMbb7zBokWLaNiw4SXb4u3tjbe3uXbrK8vmbDzIhqQTbD6QxuaDaU69IpfzfNs6PHJLdaclvrnubajJlmWSzQqLR8PK8cbzG9oam8CJiDnc2NbYGXfBcGPflq97G0NG94ylbkgD/vfveJq8voiTGTl5DrVjrGx85Yet3B0dUmLDQ0XqW/fy8iIuLo6FCxc6lS9cuJBmzZrle0x8fHye+gsWLKBJkyZO81fefvttXn31VebPn0+TJlre6ApWm52/Dp9k9vq/+XT5bqfXPvrlL6as3MPqc0M45TzduTHYt1Dnja1ekYolOIFLXCwjDWY8dD6stBgED88wNrkSEXOwWIzbW/RfDXcMB49ysG8l/KcV/PgsG3YnO8KKGzZuddtKR7dV3Oq2FTds2IHktAxW7zlWYk0u8pDQ4MGD6dGjB02aNCE+Pp6PP/6YpKQkx74qw4YN48CBA0yfPh0wVgR9+OGHDB48mCeeeIKEhAQmT57MjBkzHOd86623GDFiBF9++SU1atRw9Mj4+vri61u4D0Upuj2pp9mw/zh//p3G5gNpbDmY7ug58fZwo1ezGnicmy/SsXEYR05m0iA8gAbVAqh1buKsNkITJ6l/GWHl6E7j5oWdPoIGD7i6VSJSEM9yxl2fGz1s3Exxy2w4sI6UMON/9bZuqxnpOZ0wy/lgctBeiVeye/Kz7RYOn7z0tICrqciBpVu3bhw9epTRo0eTnJxMTEwM8+bNIzIyEoDk5GSSkpIc9aOiopg3bx6DBg3io48+IiwsjA8++MCxBwsYG9FlZWXxwAPO/7GNHDmSUaNGFfPSJFfunJMtB9Pp2CjM0dMxZl4iC7YecqpbztOd+mH+xIQHcCbbiv+5wNL/jtr5nlsboYmT7fOMsOIfbmwGd26HTRExucAIeHAaNOkD3n5UPVuetm6rmeg5Ps9dwEM4xkTP8fTLHkhVv1tLrIm6W3MZkxtO/vw7jU0H8vacrBhyB9Uqlgfg0+W7mb85hZjwAKeek6IGDN0lVhzsduNOy3GPgm9VV7dGRIrJmpND6ms3UtV+NN/dB2x2OGwJospLO3D3uLJt/Qv7+a3AUopZbXZ2HzlFtYrlKedl7Mr65k/bmLRsV566uT0nr3aOoV7o1X9/zLJOX0pYdgYsfweaDzT2WBGRsmHPcvjsvsvXe3TuFd+c9JpsHCeukxtONh3I23PyxeNNaV7buEtwdJi/07BOg/AAGlYLoGYxek6K4rrdCO16ln4QZv4TDq6Do7vgwamubpGIXC2nDl2+TlHqXQUKLCZktdnJsdkc97L5eUsKg2ZtyHcpcTlPd6dJT+3qh3Bvg1D1bsi1tf8PmNUdTqVAuYrGplQiUnb45r9VSbHrXQUKLC52Yc9J7mqdrcnpDL+3Hv9sakxkDg3wcSwlrh/mT4Nq5+achOftOfHy0C6wco1t+BJ+GADWLKhSz1iyXCnq8seJSOkR2Qz8wyA9GQpaB+ofZtQrIQosLrIn9TTP/28jW5PT8+052Xrw/L2R6ob4s3BQq2s+rCNySdYcY9njbxOM53Xvg/sngbffpY8TkdLHzR3ajYWvekJB60DbvWnUKyEKLJdwJRNJrTY7u46cYtMFq3Wa167MoLtvBKBieU/W7DsOQHkv5zknuT0nubw83LghWB8K4mJnjhpb7QPcNgRuGwpu6tETKbOiO0LX6TB/iDFnLZd/mBFWojuWaHMUWApQnKW6GdlW3vxpm2NC7Nls556TC4drAst7MeGfsdxQ1Vc9J1I6+AVDt8/hZDLU7+zq1ohISYjuCHXvhX2rjAm2vsHGMFAJ9qzk0rLmfMzfnEy/z9flGbXLjRQj7osmoJwnmw6k4efjwbNtjHuk2O12Yl9dyPFzN/u7uOekUUSgY4dYkVJh+3ywZUO9Dq5uiYiUUVrWXExWm51Xftia7xSj3LLRc7c6ysIDyzkCi8ViYXCbOlTwcqdhtQCiKqvnREopux2Wj4NfXjO27v7XMqhyo6tbJSLXMQWWi6zec8xpGKggdYJ9aV67Cg2q+WO32x3b3fe4NfJaN1Hk2so6A9/3hy3fGs8bP6JVQCLicgosFynsjZyevKM2nRqHX+PWiJSwE/th5iOQ8ie4eUD7t6FJb1e3SkREgeViVf18rmo9kVJj3yqY1QPOpEL5ytDtvyW6x4KIyKVoTeJFbomqRGiAT567U+ayYKwWuiWqUkk2S+Ta2/6TEVZCGsC/liisiIipKLBcxN3NwsgO0QB5Qkvu85EdojWZVsqeu0Yaj94/Q2B1V7dGRMSJAks+2sWEMrF7LCEBzsM+IQE+TOweW+A+LCKlyumj8PNwyMkynrt7QMvB4FXBte0SEcmH5rAUoF1MKHdHhxR7p1sRUzu0BWY8BCeSwJYD94x1dYtERC5JgeUS3N0sxNcKcnUzRK6urXNgdl/IPg0Vo3SnZREpFRRYRK4XNhssGwvL3jSe17wdHpgK5TWBXETMT4HlUmxWU9w/QeSKZZ6C2f+GbXON57c+CXe/asxbEREpBfS/VUG2zingDpVjS/wOlSJX7NQh2LMc3L3gvvfgpu6ubpGISJEosORn6xz4qidcfEeh9GSjvOt0hRYpXYJqwYNTwdsPIm5xdWtERIpMy5ovZrMaPSuXuv3h/KFGPRGzstth9Sewe+n5stp3KayISKmlwHKxfauch4HysEP6Adi3ssSaJFIkOVnwwzMw7zn4Xy84ecjVLRIRuWIaErrYqUL+5z7jEWMSbrUm0Op5sGh/FjGBU4eN+wHt/w0sbtBiMPhWdXWrRESumALLxXyDC1cv6yTs/BmO7YbbXjhfvmQMePtCWCyENjK+FikJBzfAzH9C+t/gHQAPTIYb7nZ1q0RErgoFlotFNjNWA6Unk/88Fgv4h0KXqZCyEdw9z79ks8Kq/zM25ALjN9zKdSA8DsJvgohbISSmJK5Crjebvobvn4KcsxBUGx6eCZVvcHWrRESuGs1huZibu7F0GSjw9oftxkLkrdD039Ck9/mXczKh1bNQ9z7wCwO7DY4kwobP4cdnYfFo59Nt/gaO7DA29BK5En8tNsJK7bvh8cUKKyJS5qiHJT/RHY2ly/nuw/JmwUuavcpDy2fPPz+ZAgfWwYG1cHAdRLU6/1r6Qfj6XNjx9oewxsYwUngchMeCf7jmxUjh3fee8TN08+Pa3FBEyiSL3W7Pb9yj1ElPTycgIIC0tDT8/f2vzkmv5U63h7bC3EGQvNH4zfhiLQZB61HG1zlZkHVKW6jLeUd3GcuW274BbuooFZHSq7Cf3+phuRQ3d4hqeW3OHRwNfX4Ga44xbHRhT8yhrVA1+nzdv/+Aae2hYg2jBya3Jya0IXhVuDbtE/P6azF8/RhkpIFvFedePRGRMkqBxdXcPSCkgfGIe9QoyzrjPBx0dKfx5/G9xmPzN8ZzixtUqQdtX4dad5Rkq8UV7HZI+AgWjjDmR1W7BRr/09WtEhEpEQosZuRV3vl5XC+I7gQH15/riVln9MScTIbDW8DrgqXTf/4PVv/ngp6YWKhUS8MGpV12hjGEuPFL43nj7nDfu+Dh7dp2iYiUEAWW0qJcRah1p/HIlX7QCC8hDc6XJSUYQ0h//3G+zDvAmJAZHge39tNGYqVNejLM6g4H1oDF3ehRa9pXk7JF5LqiwFKa+YcZjws1HwDV48/Ph0neCJlpsGeZ8Yjvf77un/8zhpjCYyHsJk3qNatTKXBoM/gEwoPTNPwnItclBZaypmKk8Wj4oPHcmg2HE43wcnQXVKh8vu7GL2HXL+efV6p5fhgpPM6YI1HYoaRruaLqehd2EzwwFarWNf6ORESuQ1rWfD1b+xns+dUIM8d2O7/mHQBD9p4PLLuXGsNSVaOdd/cF2DqngD1rxha8Z40UzJoDv7xqzFsKj3V1a0RErqnCfn4rsIjhzDFjUu/Bc5N6vSpAl0/Pv/5+Yzi+Bzx8IKTh+Q3uzp6An14g720Mzs2v6DpdoaUozh43NhTc9YuxeeBTa/JOwhYRKUO0D4sUTflKUPsu43GxnCxjmOnMMWM+zN+rjcclnQswPz0PNVqCj7+GiC6U3xDa0b9gxsNwbBd4ljcm1yqsiIgA6mGRorDZjKGj3Am9u5ZC6rbCH+9ZHmL+AZ0+Mp7b7fBlV6M3x6sCePkZd7f2qmAs1Q6q7TzB9Mh24xxeFcDbL+/QVGmR3xBa+SBj/52csxAQAQ99aWwMKCJSxqmHRa4+NzeoXNt4NOpm3CH4mz6FPz77jBFSLny+c0HB9et1OB9Y7HaYcKuxYVoud+/zAafmHdDxg/Ov/fissbGel+/5gJP7tX84RNx8vu7Z40YQcve69kuFt86Br3qSZwjtzFHjz8p1oNePxg62IiLioMAixecbXLh63b+F0MaQddIIGbncPKDTBOM+SZknIev0ua9PGXWrXRAqcjKNZb1Zp8CaZZRZM+FMpvFhn/uBD0a4WTsNbDn5tyeqFTz6w/nnH9xkhBY3DyPUePud7+UJawz3jjtfd/m7xnCOV4VzYencw9vX6CWpUse5HRcGIJvV6FnJM9/nArpnlIhIvhRYpPgimxmrgdKTyf9D2GK8XvN2Y/5KhSDnlz284aZCbi3v6QND9hhf594M0hFuTjvfU8luhztfOvfaRfWyTkFwjPO5s04bf9pyIOOE8XB833LOdVf9H5w9ln8bQxvDv5edf/5/sXA69XzPDnbnYaD8pB8w5rZcq3tYiYiUUgosUnxu7sbS5a96YqwKujC0nOtZaPfm1Z9s6+EFHpUK7olwczPudl1Yw1MuCjUnz4cd74vGU2N7Gr0xF9fPOg2B1Z3rZqRD5rlHUZw6VLT6IiLXAQUWuTLRHY2ly/nuw/Jm6VjS7OYOPgHG43LufqXw5+236txQ17lAk/Q7/DL68scVdqhNROQ6osAiVy66I9S9VzvdXswv2Hjkqh4Paz69/BBaZLOSaqGISKmhwCJXh5u75l1cjquG0EREyoBC3ihGRK6K3CE0/1Dncv8w7QosInIJ6mERKWkaQhMRKTIFFhFX0BCaiEiRaEhIRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRUREREyvWIFlwoQJREVF4ePjQ1xcHMuXL79k/WXLlhEXF4ePjw81a9Zk0qRJeep88803REdH4+3tTXR0NLNnzy5O00RERKQMKnJgmTVrFgMHDmT48OGsX7+eli1bcs8995CUlJRv/T179tC+fXtatmzJ+vXrefHFF3nmmWf45ptvHHUSEhLo1q0bPXr0YOPGjfTo0YOuXbvy+++/F//KREREpMyw2O12e1EOaNq0KbGxsUycONFRVq9ePTp37syYMWPy1B8yZAhz5swhMTHRUda3b182btxIQkICAN26dSM9PZ2ffvrJUaddu3ZUrFiRGTNmFKpd6enpBAQEkJaWhr+/f1EuSURERFyksJ/fHkU5aVZWFmvXrmXo0KFO5W3atGHVqlX5HpOQkECbNm2cytq2bcvkyZPJzs7G09OThIQEBg0alKfO+PHjC2xLZmYmmZmZjudpaWmAceEiIiJSOuR+bl+u/6RIgSU1NRWr1UpwcLBTeXBwMCkpKfkek5KSkm/9nJwcUlNTCQ0NLbBOQecEGDNmDK+88kqe8oiIiMJejoiIiJjEyZMnCQgIKPD1IgWWXBaLxem53W7PU3a5+heXF/Wcw4YNY/DgwY7nNpuNY8eOERQUdMnjiio9PZ2IiAj279+voabL0HtVNHq/Ck/vVeHpvSo8vVeFdy3fK7vdzsmTJwkLC7tkvSIFlsqVK+Pu7p6n5+Pw4cN5ekhyhYSE5Fvfw8ODoKCgS9Yp6JwA3t7eeHt7O5UFBgYW9lKKzN/fXz/QhaT3qmj0fhWe3qvC03tVeHqvCu9avVeX6lnJVaRVQl5eXsTFxbFw4UKn8oULF9KsWbN8j4mPj89Tf8GCBTRp0gRPT89L1inonCIiInJ9KfKQ0ODBg+nRowdNmjQhPj6ejz/+mKSkJPr27QsYQzUHDhxg+vTpgLEi6MMPP2Tw4ME88cQTJCQkMHnyZKfVPwMGDKBVq1aMHTuWTp068f3337No0SJWrFhxlS5TRERESrMiB5Zu3bpx9OhRRo8eTXJyMjExMcybN4/IyEgAkpOTnfZkiYqKYt68eQwaNIiPPvqIsLAwPvjgA7p06eKo06xZM2bOnMlLL73EiBEjqFWrFrNmzaJp06ZX4RKvjLe3NyNHjswz/CR56b0qGr1fhaf3qvD0XhWe3qvCM8N7VeR9WERERERKmu4lJCIiIqanwCIiIiKmp8AiIiIipqfAIiIiIqanwFKAMWPGcPPNN+Pn50fVqlXp3Lkz27dvd3WzTGnixIk0bNjQsaFQfHy8040spWBjxozBYrEwcOBAVzfFdEaNGoXFYnF6hISEuLpZpnXgwAG6d+9OUFAQ5cuXp3Hjxqxdu9bVzTKlGjVq5PnZslgs9O/f39VNM52cnBxeeukloqKiKFeuHDVr1mT06NHYbLYSb0uxtua/Hixbtoz+/ftz8803k5OTw/Dhw2nTpg1bt26lQoUKrm6eqVSrVo0333yT2rVrA/DZZ5/RqVMn1q9fT/369V3cOvP6448/+Pjjj2nYsKGrm2Ja9evXZ9GiRY7n7u7uLmyNeR0/fpzmzZtzxx138NNPP1G1alV27dp1TXf/Ls3++OMPrFar4/nmzZu5++67efDBB13YKnMaO3YskyZN4rPPPqN+/fqsWbOGxx57jICAAAYMGFCibdGy5kI6cuQIVatWZdmyZbRq1crVzTG9SpUq8fbbb9OnTx9XN8WUTp06RWxsLBMmTOC1116jcePGl7w7+fVo1KhRfPfdd2zYsMHVTTG9oUOHsnLlSpYvX+7qppRKAwcOZO7cuezcufOq3ouuLLjvvvsIDg5m8uTJjrIuXbpQvnx5/vvf/5ZoWzQkVEhpaWmA8UEsBbNarcycOZPTp08THx/v6uaYVv/+/bn33ntp3bq1q5tiajt37iQsLIyoqCgeeughdu/e7eommdKcOXNo0qQJDz74IFWrVuWmm27ik08+cXWzSoWsrCw+//xzevfurbCSjxYtWrB48WJ27NgBwMaNG1mxYgXt27cv8bZoSKgQ7HY7gwcPpkWLFsTExLi6Oaa0adMm4uPjycjIwNfXl9mzZxMdHe3qZpnSzJkzWbduHX/88Yerm2JqTZs2Zfr06dx4440cOnSI1157jWbNmrFlyxbHjVPFsHv3biZOnMjgwYN58cUXWb16Nc888wze3t707NnT1c0zte+++44TJ07Qq1cvVzfFlIYMGUJaWhp169bF3d0dq9XK66+/zsMPP1zyjbHLZT355JP2yMhI+/79+13dFNPKzMy079y50/7HH3/Yhw4daq9cubJ9y5Ytrm6W6SQlJdmrVq1q37Bhg6Pstttusw8YMMB1jSolTp06ZQ8ODraPGzfO1U0xHU9PT3t8fLxT2dNPP22/9dZbXdSi0qNNmzb2++67z9XNMK0ZM2bYq1WrZp8xY4b9zz//tE+fPt1eqVIl+7Rp00q8LQosl/HUU0/Zq1WrZt+9e7erm1Kq3HXXXfZ//etfrm6G6cyePdsO2N3d3R0PwG6xWOzu7u72nJwcVzfR1Fq3bm3v27evq5thOtWrV7f36dPHqWzChAn2sLAwF7WodNi7d6/dzc3N/t1337m6KaZVrVo1+4cffuhU9uqrr9rr1KlT4m3RkFAB7HY7Tz/9NLNnz2bp0qVERUW5ukmlit1uJzMz09XNMJ277rqLTZs2OZU99thj1K1blyFDhmgVzCVkZmaSmJhIy5YtXd0U02nevHmebRd27NjhuCmt5G/q1KlUrVqVe++919VNMa0zZ87g5uY83dXd3V3Lms2kf//+fPnll3z//ff4+fmRkpICQEBAAOXKlXNx68zlxRdf5J577iEiIoKTJ08yc+ZMli5dyvz5813dNNPx8/PLMw+qQoUKBAUFaX7URZ577jk6dOhA9erVOXz4MK+99hrp6ek8+uijrm6a6QwaNIhmzZrxxhtv0LVrV1avXs3HH3/Mxx9/7OqmmZbNZmPq1Kk8+uijeHjoo7AgHTp04PXXX6d69erUr1+f9evX8+6779K7d++Sb0yJ9+mUEkC+j6lTp7q6aabTu3dve2RkpN3Ly8tepUoV+1133WVfsGCBq5tVamgOS/66detmDw0NtXt6etrDwsLs//jHPzQv6hJ++OEHe0xMjN3b29tet25d+8cff+zqJpnazz//bAfs27dvd3VTTC09Pd0+YMAAe/Xq1e0+Pj72mjVr2ocPH27PzMws8bZoHxYRERExPe3DIiIiIqanwCIiIiKmp8AiIiIipqfAIiIiIqanwCIiIiKmp8AiIiIipqfAIiIiIqanwCIipnX77bczcOBAVzdDRExAgUVERERMT4FFRERETE+BRURKjfnz5xMQEMD06dNd3RQRKWEKLCJSKsycOZOuXbsyffp0evbs6ermiEgJU2AREdObMGECffv25fvvv6dTp06ubo6IuICHqxsgInIp33zzDYcOHWLFihXccsstrm6OiLiIelhExNQaN25MlSpVmDp1Kna73dXNEREXUWAREVOrVasWS5Ys4fvvv+fpp592dXNExEU0JCQipnfjjTeyZMkSbr/9djw8PBg/fryrmyQiJUyBRURKhTp16vDLL79w++234+7uzrhx41zdJBEpQRa7BoVFRETE5DSHRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETE+BRURERExPgUVERERMT4FFRERETO//AbzD/FSn6cUdAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sklearn import metrics\n",
    "\n",
    "ARI = []\n",
    "AMI = []\n",
    "x = range(2, 9)\n",
    "for k in x:\n",
    "    ARI.append(metrics.adjusted_rand_score(overlap['Verhaak_subtype'], overlap[f'k = {k}']))\n",
    "    AMI.append(metrics.adjusted_mutual_info_score(overlap[f'k = {k}'], overlap['Verhaak_subtype']))\n",
    "    \n",
    "plt.plot(x, ARI, 'o--', label = 'Adjusted Rand Index')\n",
    "plt.plot(x, AMI, 'o--', label = 'Adjusted Mutual Information')\n",
    "plt.ylim(0, 0.1)\n",
    "plt.xticks(x)\n",
    "plt.xlabel('k')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ca799357-bf72-4b3a-9e91-5df84b5f5af8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.01541584781345078,\n",
       " 0.01948010903298123,\n",
       " 0.021814501721202834,\n",
       " 0.029093830262392522,\n",
       " 0.023193176824508634,\n",
       " 0.03217692513218214,\n",
       " 0.018973277517232698]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ARI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb12b446-42c1-4e21-8f76-5b96a2449d8c",
   "metadata": {},
   "source": [
    "## Plot overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "859d1a3b-94a3-4068-bdb4-198e42a6aa6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESENCHYMAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CLASSICAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLASSICAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEURAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESENCHYMAL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Verhaak_subtype  k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8  count\n",
       "0     MESENCHYMAL      2      2      2      3      3      3      3     23\n",
       "1       CLASSICAL      2      2      2      2      2      2      2     19\n",
       "2       CLASSICAL      2      2      2      3      3      3      3     14\n",
       "3          NEURAL      2      2      2      3      3      3      3     10\n",
       "4     MESENCHYMAL      1      1      4      5      5      4      6      9"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = overlap.iloc[:,6:].value_counts().reset_index()\n",
    "counts.columns = counts.columns[:-1].tolist() + [\"count\"]\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "495fc472-b0a0-4a74-85f6-1f628b38a2cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "06431653-b2a7-41a1-b098-59784043306a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: ggalluvial\n",
       "Loading required package: ggplot2\n",
       "In addition: Warning messages:\n",
       "1: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n",
       "2: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n",
       "3: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i counts -w 800 -h 600\n",
    "\n",
    "require(ggalluvial)\n",
    "\n",
    "cluster_alluvial <- ggplot(counts, aes(y = count,\n",
    "                   axis1 = Verhaak_subtype,\n",
    "                   axis2 = `k = 2`,\n",
    "                   axis3 = `k = 3`,\n",
    "                   axis4 = `k = 4`,\n",
    "                   axis5 = `k = 5`,\n",
    "                   axis6 = `k = 6`,\n",
    "                   axis7 = `k = 7`,\n",
    "                   axis8 = `k = 8`)) +\n",
    "    geom_alluvium(aes(fill = Verhaak_subtype), width = 1/12) +\n",
    "    geom_stratum(width = 1/6, fill = 'grey', color = 'black') +\n",
    "    geom_label(stat = 'stratum', aes(label = after_stat(stratum))) +\n",
    "    #scale_x_discrete(limits = c('Verhaak subtype', 'k = 2', 'k = 3', 'k = 4', 'k = 5'), expand = c(0.1, 0.05)) +\n",
    "    scale_x_discrete(limits = c('Verhaak subtype', 'k = 2', 'k = 3', 'k = 4', 'k = 5', 'k = 6', 'k = 7', 'k = 8'), expand = c(0.1, 0.05)) +\n",
    "    scale_fill_brewer(type = 'qual', palette = 'Dark2') +\n",
    "    theme_classic(base_size = 16)\n",
    "\n",
    "pdf(\"../../results/cluster_alluvial.pdf\", width=15, height=12)\n",
    "print(cluster_alluvial)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "696cc8c9-0f40-49a2-9403-aede0bf85ce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading required package: ggpubr\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i overlap -w 1200\n",
    "\n",
    "library(\"tidyr\")\n",
    "require(ggpubr)\n",
    "\n",
    "png(filename=\"../../results/cluster_ssgsea_scores.png\", width=1200, height=600)\n",
    "\n",
    "overlap_long <- overlap %>%\n",
    "    pivot_longer(cols = starts_with('VERHAAK_GLIOBLASTOMA'), names_to = 'subtype', values_to = 'score')\n",
    "\n",
    "#p2 <- ggboxplot(overlap_long, x = 'k = 3', y = 'score', facet.by = 'subtype') +\n",
    "#    stat_compare_means()\n",
    "    \n",
    "p3 <- ggboxplot(overlap_long, x = 'k = 5', y = 'score', facet.by = 'subtype') +\n",
    "    stat_compare_means()\n",
    "\n",
    "p4 <- ggboxplot(overlap_long, x = 'k = 6', y = 'score', facet.by = 'subtype') +\n",
    "    stat_compare_means()\n",
    "\n",
    "ggarrange(p3, p4, ncol=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a3aa41-ce0b-4314-af30-2a2becbed4d7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Assess enrichment of variants within clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2c6829-a140-4507-b30b-da944c6f6e38",
   "metadata": {},
   "source": [
    "## Statistically (Fisher Exact test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d33b9306-3870-444a-9ab4-e8edd47b3a17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify the clustering in which to test enrichments\n",
    "k = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8cd4f618-e0c3-4796-b421-7deb5fe970c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ABL1_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>AKT2_mut</th>\n",
       "      <th>AKT3_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF217_del</th>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <th>ZRSR2_del</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 659 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ABL1_mut  AKT1_mut  AKT2_mut  AKT3_mut  ALK_mut  APC_mut  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0       0.0      0.0      0.0   \n",
       "\n",
       "                 ARAF_mut  ARID1A_mut  ARID1B_mut  ARID2_mut  ...  ZNF217_del  \\\n",
       "sample                                                        ...               \n",
       "GLSS-19-0266-R1       0.0         0.0         0.0        0.0  ...         0.0   \n",
       "\n",
       "                 ZRSR2_amp  ZRSR2_del  k = 2  k = 3  k = 4  k = 5  k = 6  \\\n",
       "sample                                                                     \n",
       "GLSS-19-0266-R1        0.0        0.0      1      1      1      1      1   \n",
       "\n",
       "                 k = 7  k = 8  \n",
       "sample                         \n",
       "GLSS-19-0266-R1      1      1  \n",
       "\n",
       "[1 rows x 659 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = matrix.T.join(memberships)\n",
    "df.to_csv('../../results/variants_with_memberships_v2.tsv', sep = '\\t')\n",
    "df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9bec0429-8896-40d4-902c-abbcafd7ea58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>cluster</th>\n",
       "      <th>p_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1071</th>\n",
       "      <td>JAK3_amp</td>\n",
       "      <td>14.859649</td>\n",
       "      <td>6.114934e-26</td>\n",
       "      <td>2</td>\n",
       "      <td>1.993468e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>BRD4_amp</td>\n",
       "      <td>14.643418</td>\n",
       "      <td>1.261330e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>2.055969e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>MEF2B_amp</td>\n",
       "      <td>14.193548</td>\n",
       "      <td>3.326003e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.614256e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>CRTC1_amp</td>\n",
       "      <td>13.584553</td>\n",
       "      <td>1.048592e-24</td>\n",
       "      <td>2</td>\n",
       "      <td>8.546023e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>AKT2_amp</td>\n",
       "      <td>13.800000</td>\n",
       "      <td>1.022993e-23</td>\n",
       "      <td>2</td>\n",
       "      <td>5.558260e-21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant  odds_ratio       p_value  cluster         p_adj\n",
       "1071   JAK3_amp   14.859649  6.114934e-26        2  1.993468e-22\n",
       "890    BRD4_amp   14.643418  1.261330e-25        2  2.055969e-22\n",
       "1101  MEF2B_amp   14.193548  3.326003e-25        2  3.614256e-22\n",
       "946   CRTC1_amp   13.584553  1.048592e-24        2  8.546023e-22\n",
       "834    AKT2_amp   13.800000  1.022993e-23        2  5.558260e-21"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_column = f'k = {k}'\n",
    "results = []\n",
    "\n",
    "for cluster in range(1, k + 1):\n",
    "    for variant in df:\n",
    "        if variant.startswith('k ='):\n",
    "            continue\n",
    "\n",
    "        # contingency table\n",
    "        cluster_with_variant = len(df[(df[cluster_column] == cluster) & (df[variant] != 0)])\n",
    "        cluster_without_variant = len(df[(df[cluster_column] == cluster) & (df[variant] == 0)])\n",
    "        non_cluster_with_variant = len(df[(df[cluster_column] != cluster) & (df[variant] != 0)])\n",
    "        non_cluster_without_variant = len(df[(df[cluster_column] != cluster) & (df[variant] == 0)])\n",
    "\n",
    "        contingency_table = [\n",
    "            [cluster_with_variant, cluster_without_variant],\n",
    "            [non_cluster_with_variant, non_cluster_without_variant]\n",
    "        ]\n",
    "\n",
    "        # Fisher's exact test\n",
    "        oddsratio, pvalue = stats.fisher_exact(contingency_table)\n",
    "\n",
    "        results.append({\n",
    "            'variant': variant,\n",
    "            'odds_ratio': oddsratio,\n",
    "            'p_value': pvalue,\n",
    "            'cluster': cluster\n",
    "        })\n",
    "\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df['p_adj'] = multipletests(results_df['p_value'], alpha = 0.1, method = 'fdr_bh')[1]\n",
    "results_df = results_df.sort_values('p_adj')\n",
    "\n",
    "results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc12ea7e-d774-4203-8157-5ef2b49f75fd",
   "metadata": {},
   "source": [
    "Get chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1219951f-2791-458d-aa69-4bb5a0ea8c1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5_2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5_2</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485.5_5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2_2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020.3_4</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52473</td>\n",
       "      <td>53312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id         gene chrom  start    end\n",
       "0  ENSG00000223972.5_2      DDX11L1  chr1  11869  14409\n",
       "1  ENSG00000227232.5_2       WASH7P  chr1  14404  29570\n",
       "2  ENSG00000243485.5_5  MIR1302-2HG  chr1  29554  31109\n",
       "3  ENSG00000237613.2_2      FAM138A  chr1  34554  36081\n",
       "4  ENSG00000268020.3_4       OR4G4P  chr1  52473  53312"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed = pd.read_csv('../../data/example_bed_gencode30.v30lift37.annotation.bed',\n",
    "                  sep = '\\t')\n",
    "bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3c1bdae4-c0d5-4962-a5c6-05eb4816ecc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>cluster</th>\n",
       "      <th>p_adj</th>\n",
       "      <th>gene</th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JAK3_amp</td>\n",
       "      <td>14.859649</td>\n",
       "      <td>6.114934e-26</td>\n",
       "      <td>2</td>\n",
       "      <td>1.993468e-22</td>\n",
       "      <td>JAK3</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRD4_amp</td>\n",
       "      <td>14.643418</td>\n",
       "      <td>1.261330e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>2.055969e-22</td>\n",
       "      <td>BRD4</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEF2B_amp</td>\n",
       "      <td>14.193548</td>\n",
       "      <td>3.326003e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.614256e-22</td>\n",
       "      <td>MEF2B</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRTC1_amp</td>\n",
       "      <td>13.584553</td>\n",
       "      <td>1.048592e-24</td>\n",
       "      <td>2</td>\n",
       "      <td>8.546023e-22</td>\n",
       "      <td>CRTC1</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT2_amp</td>\n",
       "      <td>13.800000</td>\n",
       "      <td>1.022993e-23</td>\n",
       "      <td>2</td>\n",
       "      <td>5.558260e-21</td>\n",
       "      <td>AKT2</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3265</th>\n",
       "      <td>SOX9_del</td>\n",
       "      <td>0.937143</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>del</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3266</th>\n",
       "      <td>RB1_mut</td>\n",
       "      <td>0.932613</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>RB1</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3267</th>\n",
       "      <td>REL_mut</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>REL</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3268</th>\n",
       "      <td>PRF1_mut</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>PRF1</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3269</th>\n",
       "      <td>ABL1_mut</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>mut</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3270 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant  odds_ratio       p_value  cluster         p_adj   gene type  \\\n",
       "0      JAK3_amp   14.859649  6.114934e-26        2  1.993468e-22   JAK3  amp   \n",
       "1      BRD4_amp   14.643418  1.261330e-25        2  2.055969e-22   BRD4  amp   \n",
       "2     MEF2B_amp   14.193548  3.326003e-25        2  3.614256e-22  MEF2B  amp   \n",
       "3     CRTC1_amp   13.584553  1.048592e-24        2  8.546023e-22  CRTC1  amp   \n",
       "4      AKT2_amp   13.800000  1.022993e-23        2  5.558260e-21   AKT2  amp   \n",
       "...         ...         ...           ...      ...           ...    ...  ...   \n",
       "3265   SOX9_del    0.937143  1.000000e+00        3  1.000000e+00   SOX9  del   \n",
       "3266    RB1_mut    0.932613  1.000000e+00        5  1.000000e+00    RB1  mut   \n",
       "3267    REL_mut    0.000000  1.000000e+00        5  1.000000e+00    REL  mut   \n",
       "3268   PRF1_mut    0.000000  1.000000e+00        5  1.000000e+00   PRF1  mut   \n",
       "3269   ABL1_mut    0.000000  1.000000e+00        1  1.000000e+00   ABL1  mut   \n",
       "\n",
       "      chrom  \n",
       "0     chr19  \n",
       "1     chr19  \n",
       "2     chr19  \n",
       "3     chr19  \n",
       "4     chr19  \n",
       "...     ...  \n",
       "3265  chr17  \n",
       "3266  chr13  \n",
       "3267   chr2  \n",
       "3268  chr10  \n",
       "3269   chr9  \n",
       "\n",
       "[3270 rows x 8 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df['gene'] = results_df['variant'].str.split('_').str[0]\n",
    "results_df['type'] = results_df['variant'].str.split('_').str[-1]\n",
    "results_df = results_df.merge(bed[['gene', 'chrom']], on = 'gene', how = 'left')\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "846d1f77-149b-4aba-b8f5-599974ebb96d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    cluster  chrom type  odds_ratio     p_adj\n",
      "31        1  chr19  del   15.142711  0.000021\n",
      "63        1   chr9  amp   12.840870  0.004407\n",
      "60        1   chr8  amp    8.037606  0.004500\n",
      "57        1   chr7  amp    0.311201  0.005256\n",
      "4         1  chr10  del    0.313054  0.011933\n",
      "52        1   chr5  del    4.912999  0.047965\n",
      "30        1  chr19  amp    0.262997  0.067320\n",
      "9         1  chr12  amp    4.007076  0.106587\n",
      "54        1   chr6  amp   10.334962  0.114856\n",
      "58        1   chr7  del    5.072003  0.196899\n",
      "     cluster  chrom type  odds_ratio         p_adj\n",
      "101        2  chr19  amp   12.829273  2.236061e-19\n",
      "107        2  chr20  amp    3.743551  1.692482e-06\n",
      "87         2  chr14  del    0.096228  4.662876e-06\n",
      "128        2   chr7  amp    3.666771  6.065128e-04\n",
      "84         2  chr13  del    0.284737  1.104095e-03\n",
      "90         2  chr15  del    0.204977  5.584740e-03\n",
      "78         2  chr11  del    0.110984  7.389302e-03\n",
      "114        2  chr22  del    0.403311  7.730819e-03\n",
      "111        2  chr21  del    0.048195  1.472848e-02\n",
      "75         2  chr10  del    2.408003  1.622160e-02\n",
      "     cluster  chrom type  odds_ratio         p_adj\n",
      "172        3  chr19  amp    0.084415  4.676460e-10\n",
      "158        3  chr14  del    6.186742  4.043066e-09\n",
      "178        3  chr20  amp    0.157827  1.667353e-08\n",
      "146        3  chr10  del    3.631652  3.376942e-04\n",
      "185        3  chr22  del    2.744805  9.381112e-04\n",
      "199        3   chr7  amp    2.977558  1.441051e-03\n",
      "142        3   chr1  amp    0.188100  2.162526e-02\n",
      "155        3  chr13  del    2.221836  2.823381e-02\n",
      "205        3   chr9  amp    0.180608  4.995241e-02\n",
      "166        3  chr17  amp    0.169105  6.412813e-02\n",
      "     cluster  chrom type  odds_ratio     p_adj\n",
      "254        4  chr21  mut   48.571429  0.006436\n",
      "257        4  chr22  mut         inf  0.006722\n",
      "245        4  chr19  mut         inf  0.009455\n",
      "236        4  chr16  mut         inf  0.017170\n",
      "281        4   chrX  mut         inf  0.027588\n",
      "248        4   chr2  mut         inf  0.033413\n",
      "260        4   chr3  mut         inf  0.073480\n",
      "275        4   chr8  mut         inf  0.077648\n",
      "215        4   chr1  mut         inf  0.100820\n",
      "272        4   chr7  mut         inf  0.103691\n",
      "     cluster  chrom type  odds_ratio     p_adj\n",
      "284        5   chr1  amp    7.435227  0.000093\n",
      "341        5   chr7  amp    0.287380  0.000165\n",
      "335        5   chr5  amp   10.026398  0.000224\n",
      "305        5  chr16  amp   22.515268  0.000299\n",
      "311        5  chr18  amp    6.332659  0.000951\n",
      "317        5   chr2  amp   16.459054  0.001568\n",
      "288        5  chr10  del    0.338588  0.001641\n",
      "329        5   chr3  amp    4.757587  0.002777\n",
      "323        5  chr21  amp    4.967653  0.003060\n",
      "291        5  chr11  del    6.695545  0.004958\n"
     ]
    }
   ],
   "source": [
    "hallmarks = results_df.groupby(['cluster', 'chrom', 'type']).agg({'odds_ratio': 'mean', 'p_adj': 'mean'}).reset_index().sort_values(['cluster', 'p_adj'])\n",
    "\n",
    "for cluster in range(1, k + 1):\n",
    "    print(hallmarks[hallmarks['cluster'] == cluster].head(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9f7591e-811b-4658-8e38-15e6c4550319",
   "metadata": {},
   "source": [
    "## Visually (heatmap)\n",
    "Sort according to genomic coordinate so we can better visualize clustering of CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9bb68c66-282c-40bd-8f89-b72d42168105",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_old</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MTOR_mut</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MTOR_amp</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>SDHB_amp</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>ID3_amp</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ENSG00000117318.9_3</td>\n",
       "      <td>1</td>\n",
       "      <td>23884417</td>\n",
       "      <td>23885992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ARID1A_mut</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ENSG00000117713.20_5</td>\n",
       "      <td>1</td>\n",
       "      <td>27019727</td>\n",
       "      <td>27108595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>BCORL1_mut</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>ENSG00000085185.15_5</td>\n",
       "      <td>23</td>\n",
       "      <td>129115083</td>\n",
       "      <td>129192058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>BCORL1_amp</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>ENSG00000085185.15_5</td>\n",
       "      <td>23</td>\n",
       "      <td>129115083</td>\n",
       "      <td>129192058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>GPC3_amp</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>ENSG00000147257.13_3</td>\n",
       "      <td>23</td>\n",
       "      <td>132669773</td>\n",
       "      <td>133119922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>GPC3_mut</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>ENSG00000147257.13_3</td>\n",
       "      <td>23</td>\n",
       "      <td>132669773</td>\n",
       "      <td>133119922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>PHF6_amp</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>ENSG00000156531.16_3</td>\n",
       "      <td>23</td>\n",
       "      <td>133507283</td>\n",
       "      <td>133562821</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>540 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant    gene gene_old               gene_id  chrom      start  \\\n",
       "69     MTOR_mut    MTOR     MTOR  ENSG00000198793.12_4      1   11166592   \n",
       "156    MTOR_amp    MTOR     MTOR  ENSG00000198793.12_4      1   11166592   \n",
       "195    SDHB_amp    SDHB     SDHB   ENSG00000117118.9_2      1   17345217   \n",
       "188     ID3_amp     ID3      ID3   ENSG00000117318.9_3      1   23884417   \n",
       "109  ARID1A_mut  ARID1A   ARID1A  ENSG00000117713.20_5      1   27019727   \n",
       "..          ...     ...      ...                   ...    ...        ...   \n",
       "272  BCORL1_mut  BCORL1   BCORL1  ENSG00000085185.15_5     23  129115083   \n",
       "339  BCORL1_amp  BCORL1   BCORL1  ENSG00000085185.15_5     23  129115083   \n",
       "312    GPC3_amp    GPC3     GPC3  ENSG00000147257.13_3     23  132669773   \n",
       "428    GPC3_mut    GPC3     GPC3  ENSG00000147257.13_3     23  132669773   \n",
       "394    PHF6_amp    PHF6     PHF6  ENSG00000156531.16_3     23  133507283   \n",
       "\n",
       "           end  \n",
       "69    11322564  \n",
       "156   11322564  \n",
       "195   17380665  \n",
       "188   23885992  \n",
       "109   27108595  \n",
       "..         ...  \n",
       "272  129192058  \n",
       "339  129192058  \n",
       "312  133119922  \n",
       "428  133119922  \n",
       "394  133562821  \n",
       "\n",
       "[540 rows x 7 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_variants = results_df[results_df['p_adj'] < 0.1][['variant']].drop_duplicates()\n",
    "sig_variants['gene'] = sig_variants['variant'].str.split('_').str[0]\n",
    "\n",
    "# symbol mismatches\n",
    "sig_variants['gene_old'] = sig_variants['gene']\n",
    "sig_variants.loc[sig_variants['gene'] == 'FAM46C', 'gene'] = 'TENT5C'\n",
    "sig_variants.loc[sig_variants['gene'] == 'PARK2', 'gene'] = 'PRKN'\n",
    "\n",
    "sig_variants = sig_variants.merge(bed, on = 'gene', how = 'left')\n",
    "\n",
    "assert len(sig_variants[sig_variants['chrom'].isnull()]) == 0\n",
    "\n",
    "sig_variants['chrom'] = sig_variants['chrom'].str.replace('chr', '').replace('X', '23').astype(int)\n",
    "\n",
    "sig_variants.sort_values(['chrom', 'start'], inplace = True)\n",
    "\n",
    "sig_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d0f9e94e-0ab2-40cc-9d6a-4aeab8531910",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "217 184 139\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    sum(sig_variants['variant'].str.contains('_amp')),\n",
    "    sum(sig_variants['variant'].str.contains('_del')),\n",
    "    sum(sig_variants['variant'].str.contains('_mut')),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd6162ea-ce20-463e-a8bb-b5a2c8c4c292",
   "metadata": {},
   "source": [
    "Combine amps and dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6907e483-01c9-48f5-b445-98b275433b26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTOR_mut</th>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <th>ID3_cnv</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>MYCL_mut</th>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <th>MPL_cnv</th>\n",
       "      <th>MUTYH_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>GPC3_mut</th>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <th>PHF6_cnv</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 383 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 MTOR_mut  MTOR_cnv  SDHB_cnv  ID3_cnv  ARID1A_mut  \\\n",
       "sample                                                               \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0266-TP       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0267-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0267-TP       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0268-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "\n",
       "                 ARID1A_cnv  MYCL_mut  MYCL_cnv  MPL_cnv  MUTYH_cnv  ...  \\\n",
       "sample                                                               ...   \n",
       "GLSS-19-0266-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0266-TP         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-TP        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0268-R1         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "\n",
       "                 GPC3_mut  GPC3_cnv  PHF6_cnv  k = 2  k = 3  k = 4  k = 5  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0      1      1      1      1   \n",
       "GLSS-19-0266-TP       0.0       0.0       0.0      2      2      2      2   \n",
       "GLSS-19-0267-R1       0.0       0.0       0.0      1      1      1      1   \n",
       "GLSS-19-0267-TP       0.0       0.0       0.0      2      2      2      3   \n",
       "GLSS-19-0268-R1       0.0       0.0       0.0      2      3      3      4   \n",
       "\n",
       "                 k = 6  k = 7  k = 8  \n",
       "sample                                \n",
       "GLSS-19-0266-R1      1      1      1  \n",
       "GLSS-19-0266-TP      2      2      2  \n",
       "GLSS-19-0267-R1      1      1      1  \n",
       "GLSS-19-0267-TP      3      3      3  \n",
       "GLSS-19-0268-R1      4      3      3  \n",
       "\n",
       "[5 rows x 383 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix_sig = matrix.T.loc[:, sig_variants['variant']]\n",
    "new_df = {}\n",
    "\n",
    "for gene in sig_variants['gene_old']:\n",
    "    if f'{gene}_mut' in matrix_sig:\n",
    "        new_df[f'{gene}_mut'] = matrix_sig[f'{gene}_mut']\n",
    "        \n",
    "    if f'{gene}_amp' in matrix_sig and f'{gene}_del' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = matrix_sig[f'{gene}_amp'] + (-1 * matrix_sig[f'{gene}_del'])\n",
    "    elif f'{gene}_del' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = -1 * matrix_sig[f'{gene}_del']\n",
    "    elif f'{gene}_amp' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = matrix_sig[f'{gene}_amp']\n",
    "\n",
    "matrix_sig = pd.DataFrame.from_dict(new_df)\n",
    "df_sig = matrix_sig.join(memberships)\n",
    "df_sig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "2df67b8b-7005-455f-ae18-18d2a80273ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_cnv</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID3_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATRX_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STAG2_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORL1_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHF6_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chrom\n",
       "gene_cnv         \n",
       "MTOR_cnv        1\n",
       "SDHB_cnv        1\n",
       "ID3_cnv         1\n",
       "ARID1A_cnv      1\n",
       "MYCL_cnv        1\n",
       "...           ...\n",
       "ATRX_cnv       23\n",
       "STAG2_cnv      23\n",
       "BCORL1_cnv     23\n",
       "GPC3_cnv       23\n",
       "PHF6_cnv       23\n",
       "\n",
       "[237 rows x 1 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_chrom_df = sig_variants[['gene_old', 'chrom']].drop_duplicates()\n",
    "cnv_chrom_df['gene_cnv'] = cnv_chrom_df['gene_old'] + '_cnv'\n",
    "cnv_chrom_df = cnv_chrom_df[cnv_chrom_df['gene_cnv'].isin(list(df_sig))]\n",
    "cnv_chrom_df.drop(columns = 'gene_old', inplace = True)\n",
    "cnv_chrom_df.set_index('gene_cnv', inplace = True)\n",
    "cnv_chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0b1b4efb-ae12-4644-9355-41fb98a55072",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_old</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MTOR_mut</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MTOR_amp</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>SDHB_amp</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>ID3_amp</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ID3</td>\n",
       "      <td>ENSG00000117318.9_3</td>\n",
       "      <td>1</td>\n",
       "      <td>23884417</td>\n",
       "      <td>23885992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ARID1A_mut</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ENSG00000117713.20_5</td>\n",
       "      <td>1</td>\n",
       "      <td>27019727</td>\n",
       "      <td>27108595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant    gene gene_old               gene_id  chrom     start  \\\n",
       "69     MTOR_mut    MTOR     MTOR  ENSG00000198793.12_4      1  11166592   \n",
       "156    MTOR_amp    MTOR     MTOR  ENSG00000198793.12_4      1  11166592   \n",
       "195    SDHB_amp    SDHB     SDHB   ENSG00000117118.9_2      1  17345217   \n",
       "188     ID3_amp     ID3      ID3   ENSG00000117318.9_3      1  23884417   \n",
       "109  ARID1A_mut  ARID1A   ARID1A  ENSG00000117713.20_5      1  27019727   \n",
       "\n",
       "          end  \n",
       "69   11322564  \n",
       "156  11322564  \n",
       "195  17380665  \n",
       "188  23885992  \n",
       "109  27108595  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "eaf9db51-fc4b-4308-987d-02f13cd466c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('int64')"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_chrom_df[\"chrom\"].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bc78f3b3-d4ef-4415-82b3-9557792489e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: grid\n",
       "========================================\n",
       "ComplexHeatmap version 2.25.2\n",
       "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
       "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
       "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
       "\n",
       "If you use it in published research, please cite either one:\n",
       "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
       "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
       "    genomic data. Bioinformatics 2016.\n",
       "\n",
       "\n",
       "The new InteractiveComplexHeatmap package can directly export static \n",
       "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
       "\n",
       "This message can be suppressed by:\n",
       "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
       "========================================\n",
       "\n",
       "========================================\n",
       "circlize version 0.4.16\n",
       "CRAN page: https://cran.r-project.org/package=circlize\n",
       "Github page: https://github.com/jokergoo/circlize\n",
       "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
       "\n",
       "If you use it in published research, please cite:\n",
       "Gu, Z. circlize implements and enhances circular visualization\n",
       "  in R. Bioinformatics 2014.\n",
       "\n",
       "This message can be suppressed by:\n",
       "  suppressPackageStartupMessages(library(circlize))\n",
       "========================================\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i df_sig -i cnv_chrom_df -w 3000 -h 1200\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(RColorBrewer)\n",
    "\n",
    "df_sig <- df_sig[order(df_sig$`k = 6`),]\n",
    "\n",
    "muts <- as.matrix(df_sig[, grep('_mut', names(df_sig))])\n",
    "cnvs <- as.matrix(df_sig[, grep('_cnv', names(df_sig))])\n",
    "\n",
    "clusters <- df_sig[, grep('=', names(df_sig))]\n",
    "\n",
    "cluster_colors <- brewer.pal(n = 8, name = 'Dark2')\n",
    "\n",
    "chrom_colors <- rep(c(cluster_colors[6], cluster_colors[8]), length.out = 23)\n",
    "chrom_colors <- setNames(chrom_colors, 1:23)\n",
    "\n",
    "row_ha <- rowAnnotation(#Cluster_2 = clusters$`k = 2`,\n",
    "                        #Cluster_3 = clusters$`k = 3`,\n",
    "                        #Cluster_4 = clusters$`k = 4`,\n",
    "                        Cluster_5 = clusters$`k = 5`,\n",
    "                        Cluster_6 = clusters$`k = 6`,\n",
    "                        col = list(#Cluster_2 = c(`1` = cluster_colors[1], `2` = cluster_colors[2]),\n",
    "                                   #Cluster_3 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3]),\n",
    "                                   #Cluster_4 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4]),\n",
    "                                   Cluster_5 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5]),\n",
    "                                   Cluster_6 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5], `6` = cluster_colors[6])),\n",
    "                        simple_anno_size = unit(2, 'cm'))\n",
    "\n",
    "col_ha <- columnAnnotation(df = cnv_chrom_df,\n",
    "                          simple_anno_size = unit(2, 'cm'),\n",
    "                          col = list(chrom = chrom_colors))\n",
    "\n",
    "heatmap_muts <- Heatmap(muts, name = 'SNV/indels',\n",
    "                        col = colorRamp2(c(0, 2, 4), c('white', 'grey', 'black')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "heatmap_cnvs <- Heatmap(cnvs, name = 'CNVs', right_annotation = row_ha, top_annotation = col_ha,\n",
    "                        col = colorRamp2(c(-2, 0, 2), c('blue', 'white', 'red')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "full_heatmap <- heatmap_muts + heatmap_cnvs\n",
    "pdf(\"../../results/glass_heatmap.pdf\", width=50, height=20)\n",
    "draw(full_heatmap)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7771136b-2b05-43ff-8ee3-b47a0d6e6790",
   "metadata": {},
   "source": [
    "# Assess prognostic value of clusters\n",
    "\n",
    "Adjust for MGMT methylation status"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b2ccc9-3821-4c95-a07d-96ab23016954",
   "metadata": {},
   "source": [
    "## Merge cluster assignments with OS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7e736dca-6a2d-42e6-9910-5a77fefffde4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>case_vital_status</th>\n",
       "      <th>case_overall_survival_mo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GLSS-DF-0002</td>\n",
       "      <td>dead</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GLSS-DF-0003</td>\n",
       "      <td>dead</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLSS-HK-0001</td>\n",
       "      <td>dead</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GLSS-MD-0006</td>\n",
       "      <td>dead</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GLSS-MD-0002</td>\n",
       "      <td>dead</td>\n",
       "      <td>262.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   case_barcode case_vital_status  case_overall_survival_mo\n",
       "0  GLSS-DF-0002              dead                      11.0\n",
       "1  GLSS-DF-0003              dead                       9.0\n",
       "2  GLSS-HK-0001              dead                      25.0\n",
       "3  GLSS-MD-0006              dead                      93.0\n",
       "4  GLSS-MD-0002              dead                     262.0"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_cases_df = pd.read_csv('../../data/glass_clinical_cases_table.tsv', sep = '\\t',\n",
    "                            usecols = ['case_barcode', 'case_overall_survival_mo', 'case_vital_status'])\n",
    "glass_cases_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6f140c05-f4f8-4cfe-a9a7-8b4a53d009d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_sample_memberships = glass_samples_df.join(memberships, how=\"inner\")#.to_csv(\"glass_sample_cluster_memberships.tsv\", sep=\"\\t\")\n",
    "glass_sample_memberships.index.name = \"sample_id\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "35638595-2eb5-4225-b094-d714ac6ee179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mgmt_methylation\n",
       "NaN    272\n",
       "U      128\n",
       "M       78\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_samples_df['mgmt_methylation'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "31eaac00-eff7-434d-b393-ef103f478b08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-1389-R1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-1389-TP</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-4065-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-4065-TP</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-FG-5963-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8\n",
       "Name                                                            \n",
       "GLSS-19-0266-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0266-TP      2      2      2      2      2      2      2\n",
       "GLSS-19-0267-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0267-TP      2      2      2      3      3      3      3\n",
       "GLSS-19-0268-R1      2      3      3      4      4      3      3\n",
       "...                ...    ...    ...    ...    ...    ...    ...\n",
       "TCGA-19-1389-R1      2      2      2      3      3      3      3\n",
       "TCGA-19-1389-TP      1      1      4      5      6      5      4\n",
       "TCGA-19-4065-R1      1      1      1      1      1      5      4\n",
       "TCGA-19-4065-TP      1      1      4      5      1      5      4\n",
       "TCGA-FG-5963-R1      1      1      4      5      3      3      3\n",
       "\n",
       "[365 rows x 7 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "d0f16b2b-3998-4a18-9be4-ac4ebdfd9e73",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/724823768.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  replace({'dead': True, 'alive': False})\n"
     ]
    }
   ],
   "source": [
    "memberships_with_os = memberships.\\\n",
    "    join(glass_samples_df[['case_barcode', 'surgical_interval_mo', 'mgmt_methylation']], how = 'left').\\\n",
    "    reset_index(names = 'sample_barcode').\\\n",
    "    merge(glass_cases_df, on = 'case_barcode', how = 'left').\\\n",
    "    replace({'dead': True, 'alive': False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c692eb0b-c6fd-48ee-a6d5-728459a73847",
   "metadata": {},
   "outputs": [],
   "source": [
    "memberships_with_os = memberships_with_os[memberships_with_os['case_overall_survival_mo'].notnull() &\n",
    "                                          memberships_with_os['surgical_interval_mo'].notnull()]\n",
    "\n",
    "# THIS ANALYSIS APPLIES THE CORRECT NOTION OF SURVIVAL FOR THE GLASS COHORT\n",
    "memberships_with_os['os_harmonized'] = np.where(\n",
    "    memberships_with_os['sample_barcode'].str.endswith('-TP'),\n",
    "    memberships_with_os['case_overall_survival_mo'],\n",
    "    memberships_with_os['case_overall_survival_mo'] - memberships_with_os['surgical_interval_mo'],\n",
    ")\n",
    "\n",
    "memberships_with_os['setting'] = np.where(\n",
    "    memberships_with_os['sample_barcode'].str.endswith('-TP'),\n",
    "    'Newly diagnosed',\n",
    "    'Recurrent'\n",
    ")\n",
    "\n",
    "memberships_with_mgmt = memberships_with_os[memberships_with_os[\"mgmt_methylation\"].notnull()].copy()\n",
    "memberships_with_mgmt['mgmt_methylation'] = np.where(\n",
    "    memberships_with_mgmt['mgmt_methylation'] == 'M', 1.0, 0.0\n",
    ")\n",
    "\n",
    "#memberships_with_os = memberships_with_os.drop(columns=[\"mgmt_methylation\"])\n",
    "#memberships_with_os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "7cbf3ee7-73e7-4074-9486-07b52ba1d903",
   "metadata": {},
   "outputs": [],
   "source": [
    "memberships_with_os.to_csv(\"../../results/glass_sample_cluster_memberships.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57cc77ec-e874-45f8-af63-132b1cc2c63a",
   "metadata": {},
   "source": [
    "# k = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f44dbc54-340c-49e6-a260-f8c570445e70",
   "metadata": {},
   "outputs": [],
   "source": [
    "for clust in range(1, 1+5):\n",
    "    memberships_with_os[f\"cluster_{clust}\"] = memberships_with_os[\"k = 5\"].map(lambda x: x == clust).astype(int)\n",
    "    memberships_with_mgmt[f\"cluster_{clust}\"] = memberships_with_mgmt[\"k = 5\"].map(lambda x: x == clust).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "21ca1095-05ed-4e42-8f62-41811fc67150",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€\n",
      "âœ” dplyr     1.1.4     âœ” readr     2.1.5\n",
      "âœ” forcats   1.0.0     âœ” stringr   1.5.2\n",
      "âœ” lubridate 1.9.4     âœ” tibble    3.3.0\n",
      "âœ” purrr     1.1.0     \n",
      "â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€\n",
      "âœ– dplyr::filter() masks stats::filter()\n",
      "âœ– dplyr::lag()    masks stats::lag()\n",
      "â„¹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gbm_icb_manuscript/lib/python3.13/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"mgmt_methylation\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 360  20\n",
      "[1] 161  20\n",
      "[1] 199  20\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_1 0.1884    1.2073   0.1645 1.145    0.252\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1     1.207     0.8283    0.8746     1.667\n",
      "\n",
      "Concordance= 0.504  (se = 0.01 )\n",
      "Likelihood ratio test= 1.25  on 1 df,   p=0.3\n",
      "Wald test            = 1.31  on 1 df,   p=0.3\n",
      "Score (logrank) test = 1.32  on 1 df,   p=0.3\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)   \n",
      "cluster_2 -0.3829    0.6819   0.1259 -3.041  0.00236 **\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2    0.6819      1.467    0.5327    0.8727\n",
      "\n",
      "Concordance= 0.531  (se = 0.016 )\n",
      "Likelihood ratio test= 9.7  on 1 df,   p=0.002\n",
      "Wald test            = 9.25  on 1 df,   p=0.002\n",
      "Score (logrank) test = 9.35  on 1 df,   p=0.002\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)   \n",
      "cluster_3 0.3698    1.4474   0.1233 2.998  0.00271 **\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3     1.447     0.6909     1.137     1.843\n",
      "\n",
      "Concordance= 0.525  (se = 0.014 )\n",
      "Likelihood ratio test= 8.59  on 1 df,   p=0.003\n",
      "Wald test            = 8.99  on 1 df,   p=0.003\n",
      "Score (logrank) test = 9.09  on 1 df,   p=0.003\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "            coef exp(coef) se(coef)    z Pr(>|z|)  \n",
      "cluster_4 0.4352    1.5452   0.2220 1.96     0.05 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4     1.545     0.6472         1     2.388\n",
      "\n",
      "Concordance= 0.518  (se = 0.009 )\n",
      "Likelihood ratio test= 3.4  on 1 df,   p=0.07\n",
      "Wald test            = 3.84  on 1 df,   p=0.05\n",
      "Score (logrank) test = 3.9  on 1 df,   p=0.05\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_5 -0.1942    0.8235   0.1390 -1.398    0.162\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5    0.8235      1.214    0.6272     1.081\n",
      "\n",
      "Concordance= 0.517  (se = 0.013 )\n",
      "Likelihood ratio test= 2.03  on 1 df,   p=0.2\n",
      "Wald test            = 1.95  on 1 df,   p=0.2\n",
      "Score (logrank) test = 1.96  on 1 df,   p=0.2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 5`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_5_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_5_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_5_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "#print(summary(model_5_combined))\n",
    "#print(summary(model_5_nd))\n",
    "#print(summary(model_5_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c53969bf-1e17-4fd1-811b-ebf9d52f6c29",
   "metadata": {},
   "source": [
    "## with MGMT methylation status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "c4206029-11e6-4094-a400-f02ee159f766",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 161  20\n",
      "[1] 92 20\n",
      "[1] 69 20\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                     coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_1         0.09148   1.09579  0.21277  0.430   0.6672  \n",
      "mgmt_methylation -0.41324   0.66150  0.16772 -2.464   0.0137 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1           1.0958     0.9126    0.7221     1.663\n",
      "mgmt_methylation    0.6615     1.5117    0.4762     0.919\n",
      "\n",
      "Concordance= 0.547  (se = 0.027 )\n",
      "Likelihood ratio test= 6.52  on 2 df,   p=0.04\n",
      "Wald test            = 6.37  on 2 df,   p=0.04\n",
      "Score (logrank) test = 6.45  on 2 df,   p=0.04\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_2        -0.4491    0.6382   0.1860 -2.415   0.0157 *\n",
      "mgmt_methylation -0.4115    0.6626   0.1681 -2.448   0.0144 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2           0.6382      1.567    0.4433    0.9189\n",
      "mgmt_methylation    0.6626      1.509    0.4766    0.9213\n",
      "\n",
      "Concordance= 0.57  (se = 0.025 )\n",
      "Likelihood ratio test= 12.54  on 2 df,   p=0.002\n",
      "Wald test            = 11.91  on 2 df,   p=0.003\n",
      "Score (logrank) test = 12.09  on 2 df,   p=0.002\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_3         0.2716    1.3121   0.1748  1.554   0.1201  \n",
      "mgmt_methylation -0.4117    0.6626   0.1679 -2.452   0.0142 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3           1.3121     0.7621    0.9315    1.8482\n",
      "mgmt_methylation    0.6626     1.5093    0.4768    0.9207\n",
      "\n",
      "Concordance= 0.553  (se = 0.026 )\n",
      "Likelihood ratio test= 8.68  on 2 df,   p=0.01\n",
      "Wald test            = 8.57  on 2 df,   p=0.01\n",
      "Score (logrank) test = 8.67  on 2 df,   p=0.01\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)   \n",
      "cluster_4         0.7398    2.0955   0.3313  2.233  0.02556 * \n",
      "mgmt_methylation -0.4446    0.6411   0.1683 -2.642  0.00824 **\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4           2.0955     0.4772    1.0946    4.0113\n",
      "mgmt_methylation    0.6411     1.5599    0.4609    0.8916\n",
      "\n",
      "Concordance= 0.57  (se = 0.024 )\n",
      "Likelihood ratio test= 10.42  on 2 df,   p=0.005\n",
      "Wald test            = 10.88  on 2 df,   p=0.004\n",
      "Score (logrank) test = 11.1  on 2 df,   p=0.004\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                     coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_5        -0.05171   0.94960  0.20660 -0.250    0.802  \n",
      "mgmt_methylation -0.41600   0.65968  0.16756 -2.483    0.013 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5           0.9496      1.053    0.6334    1.4236\n",
      "mgmt_methylation    0.6597      1.516    0.4750    0.9161\n",
      "\n",
      "Concordance= 0.543  (se = 0.026 )\n",
      "Likelihood ratio test= 6.4  on 2 df,   p=0.04\n",
      "Wald test            = 6.24  on 2 df,   p=0.04\n",
      "Score (logrank) test = 6.32  on 2 df,   p=0.04\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i memberships_with_mgmt\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_mgmt %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 5`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_5_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_5_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_5_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "\n",
    "#print(summary(model_5_combined))\n",
    "#print(summary(model_5_nd))\n",
    "#print(summary(model_5_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8ceeb91-32ac-496a-a4b7-63a75ec8fafa",
   "metadata": {},
   "source": [
    "## with primary/recurrent setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "45758307-c57a-49c1-88d8-74191b692b34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 360  20\n",
      "[1] 161  20\n",
      "[1] 199  20\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                   coef exp(coef) se(coef)     z Pr(>|z|)    \n",
      "cluster_1        0.1355    1.1452   0.1650 0.822    0.411    \n",
      "settingRecurrent 0.9739    2.6482   0.1162 8.379   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1            1.145     0.8732    0.8288     1.582\n",
      "settingRecurrent     2.648     0.3776    2.1087     3.326\n",
      "\n",
      "Concordance= 0.664  (se = 0.012 )\n",
      "Likelihood ratio test= 71.08  on 2 df,   p=4e-16\n",
      "Wald test            = 71.37  on 2 df,   p=3e-16\n",
      "Score (logrank) test = 76.13  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_2        -0.3927    0.6752   0.1271 -3.089  0.00201 ** \n",
      "settingRecurrent  0.9822    2.6703   0.1165  8.430  < 2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2           0.6752     1.4810    0.5263    0.8663\n",
      "settingRecurrent    2.6703     0.3745    2.1251    3.3553\n",
      "\n",
      "Concordance= 0.678  (se = 0.013 )\n",
      "Likelihood ratio test= 80.44  on 2 df,   p=<2e-16\n",
      "Wald test            = 79.66  on 2 df,   p=<2e-16\n",
      "Score (logrank) test = 84.6  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                   coef exp(coef) se(coef)     z Pr(>|z|)    \n",
      "cluster_3        0.5681    1.7648   0.1265 4.490 7.12e-06 ***\n",
      "settingRecurrent 1.0678    2.9090   0.1187 8.994  < 2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3            1.765     0.5666     1.377     2.261\n",
      "settingRecurrent     2.909     0.3438     2.305     3.671\n",
      "\n",
      "Concordance= 0.69  (se = 0.013 )\n",
      "Likelihood ratio test= 89.34  on 2 df,   p=<2e-16\n",
      "Wald test            = 88.63  on 2 df,   p=<2e-16\n",
      "Score (logrank) test = 93.8  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                   coef exp(coef) se(coef)     z Pr(>|z|)    \n",
      "cluster_4        0.1280    1.1365   0.2246 0.570    0.569    \n",
      "settingRecurrent 0.9670    2.6300   0.1174 8.235   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4            1.137     0.8799    0.7318     1.765\n",
      "settingRecurrent     2.630     0.3802    2.0893     3.311\n",
      "\n",
      "Concordance= 0.673  (se = 0.013 )\n",
      "Likelihood ratio test= 70.74  on 2 df,   p=4e-16\n",
      "Wald test            = 71.22  on 2 df,   p=3e-16\n",
      "Score (logrank) test = 75.99  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_5        -0.2379    0.7883   0.1390 -1.711   0.0871 .  \n",
      "settingRecurrent  0.9844    2.6763   0.1161  8.477   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5           0.7883     1.2685    0.6003     1.035\n",
      "settingRecurrent    2.6763     0.3737    2.1315     3.360\n",
      "\n",
      "Concordance= 0.681  (se = 0.013 )\n",
      "Likelihood ratio test= 73.49  on 2 df,   p=<2e-16\n",
      "Wald test            = 73.82  on 2 df,   p=<2e-16\n",
      "Score (logrank) test = 78.56  on 2 df,   p=<2e-16\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gbm_icb_manuscript/lib/python3.13/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"mgmt_methylation\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    }
   ],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 5`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_5_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_5_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_5_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + setting,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "#print(summary(model_5_combined))\n",
    "#print(summary(model_5_nd))\n",
    "#print(summary(model_5_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec110c60-9278-4178-81c4-04e400ae6c30",
   "metadata": {},
   "source": [
    "## k = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "dfa8c236-2abf-4586-b832-d9758c2e5879",
   "metadata": {},
   "outputs": [],
   "source": [
    "for clust in range(1, 1+6):\n",
    "    memberships_with_os[f\"cluster_{clust}\"] = memberships_with_os[\"k = 6\"].map(lambda x: x == clust).astype(int)\n",
    "    memberships_with_mgmt[f\"cluster_{clust}\"] = memberships_with_mgmt[\"k = 6\"].map(lambda x: x == clust).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "1cd3eee7-6ef5-4fb5-9fb6-5e2a797597b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 360  21\n",
      "[1] 161  21\n",
      "[1] 199  21\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_1 0.1145    1.1213   0.1624 0.705    0.481\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1     1.121     0.8919    0.8157     1.541\n",
      "\n",
      "Concordance= 0.502  (se = 0.01 )\n",
      "Likelihood ratio test= 0.48  on 1 df,   p=0.5\n",
      "Wald test            = 0.5  on 1 df,   p=0.5\n",
      "Score (logrank) test = 0.5  on 1 df,   p=0.5\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)   \n",
      "cluster_2 -0.3615    0.6966   0.1332 -2.713  0.00666"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gbm_icb_manuscript/lib/python3.13/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"mgmt_methylation\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " **\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2    0.6966      1.436    0.5365    0.9045\n",
      "\n",
      "Concordance= 0.526  (se = 0.015 )\n",
      "Likelihood ratio test= 7.8  on 1 df,   p=0.005\n",
      "Wald test            = 7.36  on 1 df,   p=0.007\n",
      "Score (logrank) test = 7.43  on 1 df,   p=0.006\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)   \n",
      "cluster_3 0.3452    1.4123   0.1204 2.868  0.00413 **\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3     1.412     0.7081     1.116     1.788\n",
      "\n",
      "Concordance= 0.527  (se = 0.015 )\n",
      "Likelihood ratio test= 7.92  on 1 df,   p=0.005\n",
      "Wald test            = 8.23  on 1 df,   p=0.004\n",
      "Score (logrank) test = 8.31  on 1 df,   p=0.004\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "            coef exp(coef) se(coef)    z Pr(>|z|)  \n",
      "cluster_4 0.4352    1.5452   0.2220 1.96     0.05 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4     1.545     0.6472         1     2.388\n",
      "\n",
      "Concordance= 0.518  (se = 0.009 )\n",
      "Likelihood ratio test= 3.4  on 1 df,   p=0.07\n",
      "Wald test            = 3.84  on 1 df,   p=0.05\n",
      "Score (logrank) test = 3.9  on 1 df,   p=0.05\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_5 -0.2128    0.8083   0.1922 -1.107    0.268\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5    0.8083      1.237    0.5547     1.178\n",
      "\n",
      "Concordance= 0.51  (se = 0.01 )\n",
      "Likelihood ratio test= 1.3  on 1 df,   p=0.3\n",
      "Wald test            = 1.23  on 1 df,   p=0.3\n",
      "Score (logrank) test = 1.23  on 1 df,   p=0.3\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_6, \n",
      "    data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_6 -0.1420    0.8676   0.1641 -0.865    0.387\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_6    0.8676      1.153    0.6289     1.197\n",
      "\n",
      "Concordance= 0.511  (se = 0.011 )\n",
      "Likelihood ratio test= 0.78  on 1 df,   p=0.4\n",
      "Wald test            = 0.75  on 1 df,   p=0.4\n",
      "Score (logrank) test = 0.75  on 1 df,   p=0.4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 6`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_6_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_6_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_6_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5,\n",
    "                      data = memberships_with_os)\n",
    "univariate_6 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_6,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fbde3d2-edda-4731-b21e-ad145422b44a",
   "metadata": {},
   "source": [
    "## With MGMT methylation status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "793eb381-671b-4a07-a26a-ba6bc11299dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>surgical_interval_mo</th>\n",
       "      <th>...</th>\n",
       "      <th>case_vital_status</th>\n",
       "      <th>case_overall_survival_mo</th>\n",
       "      <th>os_harmonized</th>\n",
       "      <th>setting</th>\n",
       "      <th>cluster_1</th>\n",
       "      <th>cluster_2</th>\n",
       "      <th>cluster_3</th>\n",
       "      <th>cluster_4</th>\n",
       "      <th>cluster_5</th>\n",
       "      <th>cluster_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>18.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>35.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLSS-19-0267-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>14.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GLSS-19-0267-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GLSS-19-0268-R1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-19-0268</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>TCGA-19-1389-R1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-19-1389</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>TCGA-19-1389-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-1389</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>TCGA-19-4065-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-4065</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>TCGA-19-4065-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-4065</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>TCGA-FG-5963-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-FG-5963</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>27.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>360 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_barcode  k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8  \\\n",
       "0    GLSS-19-0266-R1      1      1      1      1      1      1      1   \n",
       "1    GLSS-19-0266-TP      2      2      2      2      2      2      2   \n",
       "2    GLSS-19-0267-R1      1      1      1      1      1      1      1   \n",
       "3    GLSS-19-0267-TP      2      2      2      3      3      3      3   \n",
       "4    GLSS-19-0268-R1      2      3      3      4      4      3      3   \n",
       "..               ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "360  TCGA-19-1389-R1      2      2      2      3      3      3      3   \n",
       "361  TCGA-19-1389-TP      1      1      4      5      6      5      4   \n",
       "362  TCGA-19-4065-R1      1      1      1      1      1      5      4   \n",
       "363  TCGA-19-4065-TP      1      1      4      5      1      5      4   \n",
       "364  TCGA-FG-5963-R1      1      1      4      5      3      3      3   \n",
       "\n",
       "     case_barcode  surgical_interval_mo  ... case_vital_status  \\\n",
       "0    GLSS-19-0266                  18.0  ...              True   \n",
       "1    GLSS-19-0266                   0.0  ...              True   \n",
       "2    GLSS-19-0267                  10.0  ...              True   \n",
       "3    GLSS-19-0267                   0.0  ...              True   \n",
       "4    GLSS-19-0268                  10.0  ...              True   \n",
       "..            ...                   ...  ...               ...   \n",
       "360  TCGA-19-1389                   3.0  ...              True   \n",
       "361  TCGA-19-1389                   0.0  ...              True   \n",
       "362  TCGA-19-4065                   2.0  ...              True   \n",
       "363  TCGA-19-4065                   0.0  ...              True   \n",
       "364  TCGA-FG-5963                  14.0  ...              True   \n",
       "\n",
       "     case_overall_survival_mo  os_harmonized          setting cluster_1  \\\n",
       "0                        35.0           17.0        Recurrent         1   \n",
       "1                        35.0           35.0  Newly diagnosed         0   \n",
       "2                        14.0            4.0        Recurrent         1   \n",
       "3                        14.0           14.0  Newly diagnosed         0   \n",
       "4                        12.0            2.0        Recurrent         0   \n",
       "..                        ...            ...              ...       ...   \n",
       "360                       5.0            2.0        Recurrent         0   \n",
       "361                       5.0            5.0  Newly diagnosed         0   \n",
       "362                      15.0           13.0        Recurrent         1   \n",
       "363                      15.0           15.0  Newly diagnosed         1   \n",
       "364                      27.0           13.0        Recurrent         0   \n",
       "\n",
       "     cluster_2  cluster_3  cluster_4  cluster_5  cluster_6  \n",
       "0            0          0          0          0          0  \n",
       "1            1          0          0          0          0  \n",
       "2            0          0          0          0          0  \n",
       "3            0          1          0          0          0  \n",
       "4            0          0          1          0          0  \n",
       "..         ...        ...        ...        ...        ...  \n",
       "360          0          1          0          0          0  \n",
       "361          0          0          0          0          1  \n",
       "362          0          0          0          0          0  \n",
       "363          0          0          0          0          0  \n",
       "364          0          1          0          0          0  \n",
       "\n",
       "[360 rows x 21 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_with_os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e820e48d-2d35-4ccf-b794-4792d834b052",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 161  21\n",
      "[1] 92 21\n",
      "[1] 69 21\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_1         0.2530    1.2878   0.2196  1.152   0.2495  \n",
      "mgmt_methylation -0.4103    0.6635   0.1678 -2.445   0.0145 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1           1.2878     0.7765    0.8373    1.9807\n",
      "mgmt_methylation    0.6635     1.5072    0.4775    0.9218\n",
      "\n",
      "Concordance= 0.551  (se = 0.026 )\n",
      "Likelihood ratio test= 7.59  on 2 df,   p=0.02\n",
      "Wald test            = 7.51  on 2 df,   p=0.02\n",
      "Score (logrank) test = 7.61  on 2 df,   p=0.02\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)   \n",
      "cluster_2        -0.5624    0.5698   0.2066 -2.722  0.00649 **\n",
      "mgmt_methylation -0.4117    0.6625   0.1686 -2.442  0.01461 * \n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2           0.5698      1.755    0.3801    0.8543\n",
      "mgmt_methylation    0.6625      1.509    0.4761    0.9220\n",
      "\n",
      "Concordance= 0.573  (se = 0.025 )\n",
      "Likelihood ratio test= 14.51  on 2 df,   p=7e-04\n",
      "Wald test            = 13.4  on 2 df,   p=0.001\n",
      "Score (logrank) test = 13.66  on 2 df,   p=0.001\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_3         0.2638    1.3018   0.1738  1.517    0.129  \n",
      "mgmt_methylation -0.4015    0.6693   0.1683 -2.386    0.017 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3           1.3018     0.7682    0.9260    1.8303\n",
      "mgmt_methylation    0.6693     1.4940    0.4813    0.9308\n",
      "\n",
      "Concordance= 0.551  (se = 0.026 )\n",
      "Likelihood ratio test= 8.58  on 2 df,   p=0.01\n",
      "Wald test            = 8.46  on 2 df,   p=0.01\n",
      "Score (logrank) test = 8.56  on 2 df,   p=0.01\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)   \n",
      "cluster_4         0.7398    2.0955   0.3313  2.233  0.02556 * \n",
      "mgmt_methylation -0.4446    0.6411   0.1683 -2.642  0.00824 **\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4           2.0955     0.4772    1.0946    4.0113\n",
      "mgmt_methylation    0.6411     1.5599    0.4609    0.8916\n",
      "\n",
      "Concordance= 0.57  (se = 0.024 )\n",
      "Likelihood ratio test= 10.42  on 2 df,   p=0.005\n",
      "Wald test            = 10.88  on 2 df,   p=0.004\n",
      "Score (logrank) test = 11.1  on 2 df,   p=0.004\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_5         0.3674    1.4440   0.2728  1.347   0.1780  \n",
      "mgmt_methylation -0.4186    0.6580   0.1677 -2.496   0.0126 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5            1.444     0.6925    0.8460    2.4646\n",
      "mgmt_methylation     0.658     1.5198    0.4736    0.9141\n",
      "\n",
      "Concordance= 0.552  (se = 0.025 )\n",
      "Likelihood ratio test= 7.98  on 2 df,   p=0.02\n",
      "Wald test            = 7.97  on 2 df,   p=0.02\n",
      "Score (logrank) test = 8.07  on 2 df,   p=0.02\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_6 + \n",
      "    mgmt_methylation, data = memberships_with_mgmt)\n",
      "\n",
      "  n= 161, number of events= 159 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_6        -0.3252    0.7224   0.2235 -1.455   0.1458  \n",
      "mgmt_methylation -0.4076    0.6653   0.1677 -2.430   0.0151 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_6           0.7224      1.384    0.4661    1.1196\n",
      "mgmt_methylation    0.6653      1.503    0.4789    0.9242\n",
      "\n",
      "Concordance= 0.554  (se = 0.025 )\n",
      "Likelihood ratio test= 8.61  on 2 df,   p=0.01\n",
      "Wald test            = 8.26  on 2 df,   p=0.02\n",
      "Score (logrank) test = 8.36  on 2 df,   p=0.02\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i memberships_with_mgmt\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_mgmt %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 6`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_6_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_6_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_6_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "univariate_6 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_6 + mgmt_methylation,\n",
    "                      data = memberships_with_mgmt)\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e31284a0-eb4c-4161-b6b2-03b8336e7e42",
   "metadata": {},
   "source": [
    "## With primary/recurrent setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d8c2944d-0f97-4c38-8e33-f6d9bf055539",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 360  21\n",
      "[1] 161  21\n",
      "[1] 199  21\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                   coef exp(coef) se(coef)    z Pr(>|z|)    \n",
      "cluster_1        0.1072    1.1132   0.1625 0.66    0.509    \n",
      "settingRecurrent 0.9767    2.6557   0.1161 8.41   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1            1.113     0.8983    0.8096     1.531\n",
      "settingRecurrent     2.656     0.3765    2.1151     3.335\n",
      "\n",
      "Concordance= 0.665  (se = 0.012 )\n",
      "Likelihood ratio test= 70.86  on 2 df,   p=4e-16\n",
      "Wald test            = 71.15  on 2 df,   p=4e-16\n",
      "Score (logrank) test = 75.9  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_2        -0.2975    0.7426   0.1342 -2.218   0.0266 *  \n",
      "settingRecurrent  0.9605    2.6130   0.1164  8.252   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2           0.7426     1.3465    0.5709     0.966\n",
      "settingRecurrent    2.6130     0.3827    2.0800     3.283\n",
      "\n",
      "Concordance= 0.675  (se = 0.013 )\n",
      "Likelihood ratio test= 75.59  on 2 df,   p=<2e-16\n",
      "Wald test            = 75.34  on 2 df,   p=<2e-16\n",
      "Score (logrank) test = 80.18  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                   coef exp(coef) se(coef)     z Pr(>|z|)    \n",
      "cluster_3        0.4397    1.5523   0.1208 3.641 0.000272 ***\n",
      "settingRecurrent 1.0142    2.7571   0.1167 8.692  < 2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3            1.552     0.6442     1.225     1.967\n",
      "settingRecurrent     2.757     0.3627     2.194     3.466\n",
      "\n",
      "Concordance= 0.689  (se = 0.013 )\n",
      "Likelihood ratio test= 83.04  on 2 df,   p=<2e-16\n",
      "Wald test            = 83.51  on 2 df,   p=<2e-16\n",
      "Score (logrank) test = 88.35  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                   coef exp(coef) se(coef)     z Pr(>|z|)    \n",
      "cluster_4        0.1280    1.1365   0.2246 0.570    0.569    \n",
      "settingRecurrent 0.9670    2.6300   0.1174 8.235   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4            1.137     0.8799    0.7318     1.765\n",
      "settingRecurrent     2.630     0.3802    2.0893     3.311\n",
      "\n",
      "Concordance= 0.673  (se = 0.013 )\n",
      "Likelihood ratio test= 70.74  on 2 df,   p=4e-16\n",
      "Wald test            = 71.22  on 2 df,   p=3e-16\n",
      "Score (logrank) test = 75.99  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_5        -0.4095    0.6640   0.1936 -2.115   0.0344 *  \n",
      "settingRecurrent  1.0108    2.7477   0.1171  8.633   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5            0.664     1.5061    0.4543    0.9704\n",
      "settingRecurrent     2.748     0.3639    2.1843    3.4564\n",
      "\n",
      "Concordance= 0.678  (se = 0.012 )\n",
      "Likelihood ratio test= 75.41  on 2 df,   p=<2e-16\n",
      "Wald test            = 75.76  on 2 df,   p=<2e-16\n",
      "Score (logrank) test = 80.49  on 2 df,   p=<2e-16\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_6 + \n",
      "    setting, data = memberships_with_os)\n",
      "\n",
      "  n= 360, number of events= 324 \n",
      "\n",
      "                     coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_6        -0.09606   0.90841  0.16424 -0.585    0.559    \n",
      "settingRecurrent  0.97382   2.64803  0.11612  8.386   <2e-16 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_6           0.9084     1.1008    0.6584     1.253\n",
      "settingRecurrent    2.6480     0.3776    2.1090     3.325\n",
      "\n",
      "Concordance= 0.675  (se = 0.013 )\n",
      "Likelihood ratio test= 70.78  on 2 df,   p=4e-16\n",
      "Wald test            = 71.18  on 2 df,   p=3e-16\n",
      "Score (logrank) test = 75.91  on 2 df,   p=<2e-16\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gbm_icb_manuscript/lib/python3.13/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"mgmt_methylation\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    }
   ],
   "source": [
    "%%R -i memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 6`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "#model_6_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "#                          data = combined)\n",
    "#model_6_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = nd)\n",
    "#model_6_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "#                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5 + setting,\n",
    "                      data = memberships_with_os)\n",
    "univariate_6 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_6 + setting,\n",
    "                      data = memberships_with_os)\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "521976f5-67be-4759-ae23-8c6ecf35459f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Find a performant classifier\n",
    "Borrowing some code from David's train_classifier.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bfc0a451-73ec-4040-acea-a61f80f3b601",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sklearn.neighbors import KNeighborsClassifier, NearestCentroid\n",
    "# from sklearn.ensemble import RandomForestClassifier\n",
    "# from sklearn.linear_model import LogisticRegression, SGDClassifier\n",
    "# from sklearn.svm import SVC\n",
    "# from sklearn.pipeline import Pipeline\n",
    "# from sklearn.model_selection import GridSearchCV, cross_validate, cross_val_score, StratifiedKFold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "bc64daf5-9b22-4443-b57f-d6fa5e39a508",
   "metadata": {},
   "outputs": [],
   "source": [
    "# k = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "fdf38170-eb92-4f90-8b7e-9b05df8b1459",
   "metadata": {},
   "outputs": [],
   "source": [
    "# X = matrix_sig\n",
    "# y = df_sig[f'k = {k}']\n",
    "\n",
    "# counts = y.value_counts()\n",
    "# dist = counts / counts.sum()\n",
    "\n",
    "# print('Class distribution')\n",
    "# print(counts)\n",
    "# print(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "afa16a03-62c9-4f9b-ba1f-ff450c56757e",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# knn = KNeighborsClassifier()\n",
    "# cent = NearestCentroid()\n",
    "# rf = RandomForestClassifier(n_estimators = 1000)\n",
    "# lr = LogisticRegression(solver = 'saga', max_iter = 20000)\n",
    "# sgd = SGDClassifier(max_iter = 20000)\n",
    "\n",
    "# n_folds = 5\n",
    "\n",
    "# p_grid = [\n",
    "#     {\n",
    "#         'classifier': [knn],\n",
    "#         'classifier__n_neighbors': [1,2,3,4,5,6,7,8],\n",
    "#         'classifier__weights': ['uniform', 'distance']\n",
    "#     },\n",
    "#     {\n",
    "#         'classifier': [cent],\n",
    "#         'classifier__metric': ['euclidean', 'manhattan']\n",
    "#     },\n",
    "#     {\n",
    "#         'classifier': [rf],\n",
    "#         'classifier__criterion': ['gini', 'entropy', 'log_loss']\n",
    "#     },\n",
    "#     {\n",
    "#         'classifier': [lr],\n",
    "#         'classifier__penalty': ['l1', 'l2'],\n",
    "#         'classifier__C': [1e-1, 1, 10, 20]\n",
    "#     },\n",
    "#     {\n",
    "#         'classifier': [sgd],\n",
    "# #        'classifier__loss': ['hinge', 'log_loss'],\n",
    "#         'classifier__loss': ['log_loss'],\n",
    "#         'classifier__penalty': ['l1', 'l2'],\n",
    "#         'classifier__alpha': [1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 10]\n",
    "#     }\n",
    "# ]\n",
    "\n",
    "# pipeline = Pipeline([('classifier', rf)])\n",
    "\n",
    "# gscv = GridSearchCV(pipeline, p_grid, scoring = 'accuracy', n_jobs = 8, cv = n_folds, verbose = 1)\n",
    "\n",
    "# scores = cross_validate(gscv, X, y, cv = n_folds)\n",
    "\n",
    "# best_model = gscv.fit(X, y).best_estimator_.named_steps['classifier']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "aa9e38ce-b8a3-488e-b22a-ae7eaa8e8361",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# gscv.best_params_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f637e700-cf21-4172-b163-f65413805e44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# best_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "39c1ae9f-4509-485c-a172-de380eac7bf9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# cross_val_score(best_model, X, y, cv = 5).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "67a058bf-8021-4291-a4b1-75a4bf3cf939",
   "metadata": {},
   "outputs": [],
   "source": [
    "# best_model.coef_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d775a1a0-efa3-47fe-8894-143a6178cdb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "c56d69af-03b3-4b25-8e6b-e3e2bfaa62b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# best_model.fit(X, y)\n",
    "\n",
    "# coeffs = best_model.coef_\n",
    "# coeffs = pd.DataFrame(coeffs).T\n",
    "# coeffs.index = [c for c in df_sig.columns if not c.startswith(\"k =\")]\n",
    "# coeffs.sort_values(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83344ada-a467-4835-90ed-9edd2ce45b27",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Does PCA tell us anything?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "c21255f2-3f4d-4edb-b45e-db4565103d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sklearn.decomposition import PCA\n",
    "\n",
    "# pca = PCA(n_components = 5)\n",
    "# X_r = pca.fit_transform(X)\n",
    "# # X_icb_r = pca.transform(op_matrix_sig)\n",
    "\n",
    "# print(pca.explained_variance_ratio_)\n",
    "\n",
    "# colors = ['red', 'blue', 'orange', 'brown', 'green', 'gray']\n",
    "# # plt.scatter(X_icb_r[:,0], X_icb_r[:,1], s = 40, color = 'gray', alpha = 0.8, lw = 1, label = 'ICB')\n",
    "\n",
    "# for i, color in enumerate(colors):\n",
    "#     plt.scatter(X_r[y == i + 1, 0], X_r[y == i + 1, 1], s = 20, color = color, alpha = 0.5, lw = 1, label = i + 1)\n",
    "# plt.legend(loc = 'best')\n",
    "# plt.xlabel('PC1')\n",
    "# plt.ylabel('PC2')\n",
    "# plt.xscale('symlog')\n",
    "# plt.yscale('symlog')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "6db7e40e-c8c9-4181-99b2-77557d5f560e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3', 'PC4', 'PC5'], index=list(X)).sort_values('PC2').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "b5c90e1e-2ff1-446f-9605-ff1fb57d0892",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3', 'PC4', 'PC5'], index=list(X)).sort_values('PC2').tail(10)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1e2105a6-eca9-4b77-9dd0-3b73aca0f956",
   "metadata": {},
   "source": [
    "# Prepare the ICB OncoPanel data\n",
    "\n",
    "Preprocess the data identically to GLASS... to the extent possible. Some details:\n",
    "- [x] Remove likely germline variants from the ICB OncoPanel data before classification using the gnomAD frequency filter we used for the Cox regressions.\n",
    "- [ ] Add in WES mutational and CNV data for the ~one dozen ICB patients with WES and no OncoPanel\n",
    "    - This would involve preprocessing akin to that done in `aggregate_muts_for_pathways.ipynb`, just without the filters for GISTIC peaks, OncoKB, and CADD\n",
    "- [x] Determine how to handle patients in ICB with multiple OncoPanel / WES samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69cbdfe-01b8-448e-9574-4c9e72e6c51a",
   "metadata": {},
   "source": [
    "### Load ICB cohort sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "4449ed75-ff09-47ac-baf6-b1e49a442fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_path = '../../data/sample_cohort_sheet.tsv'\n",
    "pt_path = '../../data/participant_cohort_sheet.tsv'\n",
    "\n",
    "icb_sample_df = pd.read_csv(clinical_path, sep=\"\\t\")\n",
    "icb_sample_df[\"sample_id_legacy\"] = icb_sample_df[\"sample_id_legacy\"].astype(str)\n",
    "\n",
    "icb_pt_df = pd.read_csv(pt_path, sep=\"\\t\")\n",
    "# icb_clinical = pd.read_csv(clinical_path, sep = '\\t')\n",
    "# icb_pt = pd.read_csv(pt_path, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "77310d88-0355-46bd-959d-70917afcc3c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "167"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_sample_df[(icb_sample_df[\"usable_wes\"] | icb_sample_df[\"usable_oncopanel\"]) & (icb_sample_df[\"tumor_normal\"] == \"tumor\") & (icb_sample_df[\"pre_post\"] == \"Pre\")][\"participant_id\"].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c6106d6-49bf-4a52-8405-fe946a2e5ae3",
   "metadata": {},
   "source": [
    "### Load OncoPanel variants\n",
    "\n",
    "...and clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "90f2079c-98b1-475e-9838-5540929b30bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "op_mutations_path = '../../data/oncopanel_mutations.csv' \n",
    "op_cnvs_path = '../../data/oncopanel_cnvs.csv'\n",
    "op_clinical_path = '../../data/oncopanel_clinical.xlsx'\n",
    "\n",
    "op_mut_df = pd.read_csv(op_mutations_path, sep = '\\t')\n",
    "op_mut_df[\"id_uniquesample\"] = op_mut_df[\"id_uniquesample\"].astype(str)\n",
    "\n",
    "op_cnv_df = pd.read_csv(op_cnvs_path, sep = '\\t')\n",
    "op_cnv_df[\"id_uniquesample\"] = op_cnv_df[\"id_uniquesample\"].astype(str)\n",
    "\n",
    "op_sample_df = pd.read_excel(op_clinical_path)\n",
    "op_sample_df[\"id_uniquesample\"] = op_sample_df[\"id_uniquesample\"].astype(str)\n",
    "op_sample_df[\"id_drspatient\"] = op_sample_df[\"id_drspatient\"].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8a5ec2-3c01-4039-8034-1cc291b5bb79",
   "metadata": {},
   "source": [
    "Filter to keep variants belonging to known ICB samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "6dcdb9af-eb8a-41d3-93a7-8b46bd3c80f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.True_"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Harmonize sample IDs in the OncoPanel data\n",
    "op_sample_df['sample_id_merge'] = np.where(op_sample_df['id_broad_sample'].notnull(),\n",
    "                                           op_sample_df['id_broad_sample'],\n",
    "                                           op_sample_df['id_uniquesample']).astype(str)\n",
    "\n",
    "op_sample_df.replace({'GBM.ICB-3-14.637.Post.FFbank': 'GBM.ICB-3-14.233.Pre'}, inplace = True)\n",
    "\n",
    "icb_op_samples = icb_sample_df[icb_sample_df['usable_oncopanel']]['sample_id_legacy'].astype(str)\n",
    "op_sample_df = op_sample_df[op_sample_df['sample_id_merge'].isin(icb_op_samples)]\n",
    "\n",
    "len(op_sample_df) == icb_sample_df['usable_oncopanel'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "64ba3adf-efa7-44f6-9e55-a8508e5d144d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "159"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_op_samples = op_sample_df['id_uniquesample'].tolist()\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df['id_uniquesample'].isin(kept_op_samples)]\n",
    "op_cnv_df = op_cnv_df[op_cnv_df['id_uniquesample'].isin(kept_op_samples)]\n",
    "\n",
    "len(kept_op_samples)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e3b589aa-5b41-482e-90b5-a578185b3433",
   "metadata": {},
   "source": [
    "### Filter mutations\n",
    "\n",
    "Select mutations by several criteria:\n",
    "* Filter mutations by VAF\n",
    "* Keep only variants belonging to genes with exons covered by all versions of OncoPanel\n",
    "* Keep only non-synonymous mutations\n",
    "* Filter likely germline events by gnomAD frequency\n",
    "* Keep mutations belonging to the latest pre-ICB timepoint for each participant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "e0db565c-62f3-4177-99a8-240cc7de9323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants above the VAF threshold (0.1)\n",
    "op_mut_df = op_mut_df[op_mut_df['ALLELE_FRACTION'] >= 100*VAF_THRESHOLD]\n",
    "\n",
    "# Keep only genes with exons covered by all versions of OncoPanel\n",
    "op_mut_df = op_mut_df[op_mut_df['HARMONIZED_HUGO_GENE_NAME'].isin(oncopanel_genes['Gene'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "22010afe-b10a-4857-88df-fd67930e5cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Non-synonymous mutations\n",
    "nonsyn_classifications = ['Missense',\n",
    "                          'Frame_Shift_Del',\n",
    "                          'Nonsense',\n",
    "                          'Splice_Region',\n",
    "                          'Frame_Shift_Ins',\n",
    "                          'Splice_Site',\n",
    "                          'In_Frame_Ins',\n",
    "                          'In_Frame_Del',\n",
    "                          'Translation_Start_Site']\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df['HARMONIZED_VARIANT_CLASS'].isin(nonsyn_classifications)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "6039fbe6-49f7-4eea-b8b9-9e65dd8cae76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gnomAD germline mutation filtering\n",
    "gnomad_threshold = 2e-4\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df[\"MAX_GNOMAD_FREQUENCY\"].fillna(0.0).astype(float).lt(gnomad_threshold)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "1883ad81-6a8c-4643-a754-5bef0e6cfa02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Latest pre-ICB timepoint for each participant\n",
    "def keep_latest_preicb_timepoint_variants(variant_df, sample_df, sample_id_col=\"sample_id_merge\"):\n",
    "    variant_df = variant_df.join(sample_df[[\"sample_id_legacy\", \"participant_id_legacy\", \"collection_date_dfd\", \"pre_post\"]].set_index(\"sample_id_legacy\"), on=sample_id_col)\n",
    "    variant_df = variant_df[variant_df[\"pre_post\"] == \"Pre\"]\n",
    "    latest_timepoints = variant_df[[\"participant_id_legacy\", \"collection_date_dfd\"]].groupby(\"participant_id_legacy\").max()\n",
    "    latest_timepoints.columns = [\"latest_timepoint\"]\n",
    "    variant_df = variant_df.join(latest_timepoints, on=\"participant_id_legacy\")\n",
    "\n",
    "    variant_df = variant_df[variant_df[\"collection_date_dfd\"] == variant_df[\"latest_timepoint\"]]\n",
    "    return variant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "f815699a-877d-410c-a59f-2dd36f497798",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_drspatient</th>\n",
       "      <th>id_uniquesample</th>\n",
       "      <th>id_icb</th>\n",
       "      <th>id_mj</th>\n",
       "      <th>VARIANT_CALL_ID</th>\n",
       "      <th>CANONICAL_GENE</th>\n",
       "      <th>BEST_EFF_GENE</th>\n",
       "      <th>CANONICAL_VARIANT_CLASS</th>\n",
       "      <th>BEST_EFF_VARIANT_CLASS</th>\n",
       "      <th>HARMONIZED_HUGO_GENE_NAME</th>\n",
       "      <th>...</th>\n",
       "      <th>CANONICAL_REF_SEQ_TSCP_ID</th>\n",
       "      <th>BEST_EFFECT_ENSEMBL_TSCP_ID</th>\n",
       "      <th>BEST_EFFECT_REF_SEQ_TSCP_ID</th>\n",
       "      <th>HARMONIZED_TRANSCRIPT_ID</th>\n",
       "      <th>MAX_GNOMAD_FREQUENCY</th>\n",
       "      <th>sample_id_merge</th>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th>collection_date_dfd</th>\n",
       "      <th>pre_post</th>\n",
       "      <th>latest_timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>101954597</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>221.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>210153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001126112</td>\n",
       "      <td>ENST00000269305</td>\n",
       "      <td>8.80300000000e-06</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>GBM.ICB-221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>101954597</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>221.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>210001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RB1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_000321</td>\n",
       "      <td>ENST00000267163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>GBM.ICB-221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>656799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAF</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001654</td>\n",
       "      <td>ENST00000377045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>GBM.ICB-105</td>\n",
       "      <td>19.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1019542469</td>\n",
       "      <td>1019543884287</td>\n",
       "      <td>1.0</td>\n",
       "      <td>720.0</td>\n",
       "      <td>798723</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MPL</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_005373</td>\n",
       "      <td>ENST00000372470</td>\n",
       "      <td>.000028908</td>\n",
       "      <td>GBM.ICB-1-10.533.A2-Pre</td>\n",
       "      <td>GBM.ICB-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1019542469</td>\n",
       "      <td>1019543884287</td>\n",
       "      <td>1.0</td>\n",
       "      <td>720.0</td>\n",
       "      <td>798828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SETBP1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_015559</td>\n",
       "      <td>ENST00000282030</td>\n",
       "      <td>.000086861</td>\n",
       "      <td>GBM.ICB-1-10.533.A2-Pre</td>\n",
       "      <td>GBM.ICB-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4193</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264657.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264657</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925965</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Missense</td>\n",
       "      <td>GNAS</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371100.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371100</td>\n",
       "      <td>.000086133</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4195</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925895</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense</td>\n",
       "      <td>CBLB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264122.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264122</td>\n",
       "      <td>.000115821</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4196</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Frameshift</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371953.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4255</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>1480352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>ATM</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>GBM.ICB-108</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>378 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id_drspatient id_uniquesample  id_icb   id_mj  VARIANT_CALL_ID  \\\n",
       "2         101954597   1019543883876   221.0   192.0           210153   \n",
       "3         101954597   1019543883876   221.0   192.0           210001   \n",
       "14        101954688   1019543884068   105.0   221.0           656799   \n",
       "24       1019542469   1019543884287     1.0   720.0           798723   \n",
       "25       1019542469   1019543884287     1.0   720.0           798828   \n",
       "...             ...             ...     ...     ...              ...   \n",
       "4193     1019543398   1019543916503   244.0  1027.0         10925958   \n",
       "4194     1019543398   1019543916503   244.0  1027.0         10925965   \n",
       "4195     1019543398   1019543916503   244.0  1027.0         10925895   \n",
       "4196     1019543398   1019543916503   244.0  1027.0         10925856   \n",
       "4255       92361207    923612315805   108.0   255.0          1480352   \n",
       "\n",
       "     CANONICAL_GENE BEST_EFF_GENE CANONICAL_VARIANT_CLASS  \\\n",
       "2               NaN           NaN                     NaN   \n",
       "3               NaN           NaN                     NaN   \n",
       "14              NaN           NaN                     NaN   \n",
       "24              NaN           NaN                     NaN   \n",
       "25              NaN           NaN                     NaN   \n",
       "...             ...           ...                     ...   \n",
       "4193            NaN           NaN                     NaN   \n",
       "4194            NaN           NaN      intergenic_variant   \n",
       "4195            NaN           NaN                     NaN   \n",
       "4196            NaN           NaN                     NaN   \n",
       "4255            NaN           ATM                     NaN   \n",
       "\n",
       "     BEST_EFF_VARIANT_CLASS HARMONIZED_HUGO_GENE_NAME  ...  \\\n",
       "2                       NaN                      TP53  ...   \n",
       "3                       NaN                       RB1  ...   \n",
       "14                      NaN                      ARAF  ...   \n",
       "24                      NaN                       MPL  ...   \n",
       "25                      NaN                    SETBP1  ...   \n",
       "...                     ...                       ...  ...   \n",
       "4193                    NaN                     STAT3  ...   \n",
       "4194               Missense                      GNAS  ...   \n",
       "4195               Missense                      CBLB  ...   \n",
       "4196             Frameshift                      PTEN  ...   \n",
       "4255            Splice_Site                       ATM  ...   \n",
       "\n",
       "     CANONICAL_REF_SEQ_TSCP_ID BEST_EFFECT_ENSEMBL_TSCP_ID  \\\n",
       "2                          NaN                         NaN   \n",
       "3                          NaN                         NaN   \n",
       "14                         NaN                         NaN   \n",
       "24                         NaN                         NaN   \n",
       "25                         NaN                         NaN   \n",
       "...                        ...                         ...   \n",
       "4193                       NaN           ENST00000264657.5   \n",
       "4194                       NaN           ENST00000371100.4   \n",
       "4195                       NaN           ENST00000264122.4   \n",
       "4196                       NaN           ENST00000371953.3   \n",
       "4255                       NaN                         NaN   \n",
       "\n",
       "     BEST_EFFECT_REF_SEQ_TSCP_ID HARMONIZED_TRANSCRIPT_ID  \\\n",
       "2                   NM_001126112          ENST00000269305   \n",
       "3                      NM_000321          ENST00000267163   \n",
       "14                     NM_001654          ENST00000377045   \n",
       "24                     NM_005373          ENST00000372470   \n",
       "25                     NM_015559          ENST00000282030   \n",
       "...                          ...                      ...   \n",
       "4193                         NaN          ENST00000264657   \n",
       "4194                         NaN          ENST00000371100   \n",
       "4195                         NaN          ENST00000264122   \n",
       "4196                         NaN          ENST00000371953   \n",
       "4255                         NaN                      NaN   \n",
       "\n",
       "     MAX_GNOMAD_FREQUENCY          sample_id_merge participant_id_legacy  \\\n",
       "2       8.80300000000e-06            1019543883876           GBM.ICB-221   \n",
       "3                     NaN            1019543883876           GBM.ICB-221   \n",
       "14                    NaN            1019543884068           GBM.ICB-105   \n",
       "24             .000028908  GBM.ICB-1-10.533.A2-Pre             GBM.ICB-1   \n",
       "25             .000086861  GBM.ICB-1-10.533.A2-Pre             GBM.ICB-1   \n",
       "...                   ...                      ...                   ...   \n",
       "4193                  NaN            1019543916503           GBM.ICB-244   \n",
       "4194           .000086133            1019543916503           GBM.ICB-244   \n",
       "4195           .000115821            1019543916503           GBM.ICB-244   \n",
       "4196                  NaN            1019543916503           GBM.ICB-244   \n",
       "4255                  NaN             923612315805           GBM.ICB-108   \n",
       "\n",
       "      collection_date_dfd  pre_post latest_timepoint  \n",
       "2                     0.0       Pre              0.0  \n",
       "3                     0.0       Pre              0.0  \n",
       "14                   19.0       Pre             19.0  \n",
       "24                    0.0       Pre              0.0  \n",
       "25                    0.0       Pre              0.0  \n",
       "...                   ...       ...              ...  \n",
       "4193                  0.0       Pre              0.0  \n",
       "4194                  0.0       Pre              0.0  \n",
       "4195                  0.0       Pre              0.0  \n",
       "4196                  0.0       Pre              0.0  \n",
       "4255                  0.0       Pre              0.0  \n",
       "\n",
       "[378 rows x 44 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_mut_df = op_mut_df.join(op_sample_df[[\"id_uniquesample\", \"sample_id_merge\"]].set_index(\"id_uniquesample\"), on=\"id_uniquesample\")\n",
    "op_mut_df = keep_latest_preicb_timepoint_variants(op_mut_df, icb_sample_df)\n",
    "\n",
    "op_mut_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3b770f8-e853-4051-b016-5e872952707e",
   "metadata": {},
   "source": [
    "### Construct pt-by-variant matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "ec429b49-5e0e-42d5-8980-32520dc2b1e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>ASXL1_mut</th>\n",
       "      <th>ATM_mut</th>\n",
       "      <th>ATRX_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>TCF3_mut</th>\n",
       "      <th>TCF7L2_mut</th>\n",
       "      <th>TERT_mut</th>\n",
       "      <th>TET2_mut</th>\n",
       "      <th>TNFAIP3_mut</th>\n",
       "      <th>TP53_mut</th>\n",
       "      <th>TSC1_mut</th>\n",
       "      <th>TSC2_mut</th>\n",
       "      <th>WRN_mut</th>\n",
       "      <th>WT1_mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-102</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-103</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-104</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-87</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-88</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-89</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-90</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-94</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows Ã— 127 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_mut               AKT1_mut  ALK_mut  APC_mut  ARAF_mut  ARID1A_mut  \\\n",
       "participant_id_legacy                                                     \n",
       "GBM.ICB-1                     0        0        0         0           0   \n",
       "GBM.ICB-10                    0        0        0         0           0   \n",
       "GBM.ICB-102                   0        0        0         0           0   \n",
       "GBM.ICB-103                   0        0        2         0           0   \n",
       "GBM.ICB-104                   0        0        0         0           0   \n",
       "...                         ...      ...      ...       ...         ...   \n",
       "GBM.ICB-87                    0        0        0         0           0   \n",
       "GBM.ICB-88                    0        0        0         0           0   \n",
       "GBM.ICB-89                    0        0        0         0           0   \n",
       "GBM.ICB-90                    0        0        0         0           0   \n",
       "GBM.ICB-94                    0        0        0         0           0   \n",
       "\n",
       "gene_mut               ARID1B_mut  ARID2_mut  ASXL1_mut  ATM_mut  ATRX_mut  \\\n",
       "participant_id_legacy                                                        \n",
       "GBM.ICB-1                       0          0          0        0         0   \n",
       "GBM.ICB-10                      0          0          0        0         0   \n",
       "GBM.ICB-102                     0          0          0        0         0   \n",
       "GBM.ICB-103                     0          0          0        0         0   \n",
       "GBM.ICB-104                     0          0          0        0         0   \n",
       "...                           ...        ...        ...      ...       ...   \n",
       "GBM.ICB-87                      0          0          0        0         0   \n",
       "GBM.ICB-88                      0          0          0        0         0   \n",
       "GBM.ICB-89                      0          0          0        0         0   \n",
       "GBM.ICB-90                      0          0          0        0         0   \n",
       "GBM.ICB-94                      0          0          0        0         2   \n",
       "\n",
       "gene_mut               ...  TCF3_mut  TCF7L2_mut  TERT_mut  TET2_mut  \\\n",
       "participant_id_legacy  ...                                             \n",
       "GBM.ICB-1              ...         0           0         0         0   \n",
       "GBM.ICB-10             ...         0           0         0         0   \n",
       "GBM.ICB-102            ...         0           0         0         0   \n",
       "GBM.ICB-103            ...         0           0         0         0   \n",
       "GBM.ICB-104            ...         0           0         0         0   \n",
       "...                    ...       ...         ...       ...       ...   \n",
       "GBM.ICB-87             ...         0           0         0         0   \n",
       "GBM.ICB-88             ...         0           0         0         0   \n",
       "GBM.ICB-89             ...         0           0         0         0   \n",
       "GBM.ICB-90             ...         0           0         0         0   \n",
       "GBM.ICB-94             ...         0           0         0         0   \n",
       "\n",
       "gene_mut               TNFAIP3_mut  TP53_mut  TSC1_mut  TSC2_mut  WRN_mut  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-1                        0         0         0         0        0   \n",
       "GBM.ICB-10                       0         0         0         0        0   \n",
       "GBM.ICB-102                      0         0         0         0        0   \n",
       "GBM.ICB-103                      0         2         0         0        0   \n",
       "GBM.ICB-104                      0         0         0         0        0   \n",
       "...                            ...       ...       ...       ...      ...   \n",
       "GBM.ICB-87                       0         0         0         0        0   \n",
       "GBM.ICB-88                       0         0         0         0        0   \n",
       "GBM.ICB-89                       0         0         0         0        0   \n",
       "GBM.ICB-90                       0         0         0         0        0   \n",
       "GBM.ICB-94                       0         2         0         0        0   \n",
       "\n",
       "gene_mut               WT1_mut  \n",
       "participant_id_legacy           \n",
       "GBM.ICB-1                    0  \n",
       "GBM.ICB-10                   0  \n",
       "GBM.ICB-102                  0  \n",
       "GBM.ICB-103                  0  \n",
       "GBM.ICB-104                  0  \n",
       "...                        ...  \n",
       "GBM.ICB-87                   0  \n",
       "GBM.ICB-88                   0  \n",
       "GBM.ICB-89                   0  \n",
       "GBM.ICB-90                   0  \n",
       "GBM.ICB-94                   0  \n",
       "\n",
       "[137 rows x 127 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_mut_df['gene_mut'] = op_mut_df['HARMONIZED_HUGO_GENE_NAME'] + '_mut'\n",
    "\n",
    "op_variants_matrix = op_mut_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# scoring: sequence variants = 2\n",
    "op_variants_matrix *= 2\n",
    "\n",
    "op_variants_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad545274-6ca3-4204-8bd5-ad7d6590f6e4",
   "metadata": {},
   "source": [
    "### Filter CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "8b601d63-6bd2-46b9-8a81-65cf910a0ee5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_drspatient</th>\n",
       "      <th>id_uniquesample</th>\n",
       "      <th>id_icb</th>\n",
       "      <th>id_mj</th>\n",
       "      <th>VARIANT_CALL_ID</th>\n",
       "      <th>CNV_COPY_COUNT</th>\n",
       "      <th>CNV_COPY_COUNT_IN_REPORT</th>\n",
       "      <th>TUMOR_PURITY</th>\n",
       "      <th>PANEL_VERSION</th>\n",
       "      <th>CANONICAL_GENE</th>\n",
       "      <th>HARMONIZED_HUGO_GENE_NAME</th>\n",
       "      <th>HARMONIZED_VARIANT_CLASS</th>\n",
       "      <th>CNV_GROUP_ID</th>\n",
       "      <th>CNV_GROUP_CYTOBAND</th>\n",
       "      <th>CNV_GROUP_TIER</th>\n",
       "      <th>CNV_GROUP_EVENT</th>\n",
       "      <th>hlvl_call</th>\n",
       "      <th>gene_cnv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129364723</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>EGFR_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>CDKN2A_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129332074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRF1</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>PRF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IKZF1</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>IKZF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129333379</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RET</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>RET_cnv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     id_drspatient id_uniquesample  id_icb  id_mj  VARIANT_CALL_ID  \\\n",
       "124      101954688   1019543884068   105.0  221.0        129364723   \n",
       "125      101954688   1019543884068   105.0  221.0        129354263   \n",
       "126      101954688   1019543884068   105.0  221.0        129332074   \n",
       "128      101954688   1019543884068   105.0  221.0        129354974   \n",
       "129      101954688   1019543884068   105.0  221.0        129333379   \n",
       "\n",
       "     CNV_COPY_COUNT  CNV_COPY_COUNT_IN_REPORT  TUMOR_PURITY  PANEL_VERSION  \\\n",
       "124             NaN                       NaN          70.0            1.0   \n",
       "125             NaN                       NaN          70.0            1.0   \n",
       "126             NaN                       NaN          70.0            1.0   \n",
       "128             NaN                       NaN          70.0            1.0   \n",
       "129             NaN                       NaN          70.0            1.0   \n",
       "\n",
       "    CANONICAL_GENE HARMONIZED_HUGO_GENE_NAME HARMONIZED_VARIANT_CLASS  \\\n",
       "124            NaN                      EGFR                       LA   \n",
       "125            NaN                    CDKN2A                     1DEL   \n",
       "126            NaN                      PRF1                     1DEL   \n",
       "128            NaN                     IKZF1                       LA   \n",
       "129            NaN                       RET                     1DEL   \n",
       "\n",
       "     CNV_GROUP_ID CNV_GROUP_CYTOBAND CNV_GROUP_TIER CNV_GROUP_EVENT  \\\n",
       "124           NaN                NaN            NaN             NaN   \n",
       "125           NaN                NaN            NaN             NaN   \n",
       "126           NaN                NaN            NaN             NaN   \n",
       "128           NaN                NaN            NaN             NaN   \n",
       "129           NaN                NaN            NaN             NaN   \n",
       "\n",
       "     hlvl_call    gene_cnv  \n",
       "124          1    EGFR_cnv  \n",
       "125         -1  CDKN2A_cnv  \n",
       "126         -1    PRF1_cnv  \n",
       "128          1   IKZF1_cnv  \n",
       "129         -1     RET_cnv  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnv_df = op_cnv_df[op_cnv_df['HARMONIZED_HUGO_GENE_NAME'].isin(oncopanel_genes['Gene'])]\n",
    "\n",
    "op_cnv_df['hlvl_call'] = op_cnv_df['HARMONIZED_VARIANT_CLASS'].map(\n",
    "    {'2DEL': -2, '1DEL': -1, 'LA': 1, 'HA': 2}\n",
    ")\n",
    "\n",
    "op_cnv_df['gene_cnv'] = op_cnv_df['HARMONIZED_HUGO_GENE_NAME'] + '_cnv'\n",
    "\n",
    "op_cnv_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49f77769-172e-4d3f-85d3-159db637f5fe",
   "metadata": {},
   "source": [
    "Ensure no duplicate CNVs in any sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "c25b8108-9e92-46a2-97d8-9bf8dcefdc7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnv_df.groupby(['id_uniquesample', 'HARMONIZED_HUGO_GENE_NAME']).size().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "9bd3f2f5-1af0-4c44-8a26-055b272f5ae1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_drspatient</th>\n",
       "      <th>id_uniquesample</th>\n",
       "      <th>id_icb</th>\n",
       "      <th>id_mj</th>\n",
       "      <th>VARIANT_CALL_ID</th>\n",
       "      <th>CNV_COPY_COUNT</th>\n",
       "      <th>CNV_COPY_COUNT_IN_REPORT</th>\n",
       "      <th>TUMOR_PURITY</th>\n",
       "      <th>PANEL_VERSION</th>\n",
       "      <th>CANONICAL_GENE</th>\n",
       "      <th>HARMONIZED_HUGO_GENE_NAME</th>\n",
       "      <th>HARMONIZED_VARIANT_CLASS</th>\n",
       "      <th>CNV_GROUP_ID</th>\n",
       "      <th>CNV_GROUP_CYTOBAND</th>\n",
       "      <th>CNV_GROUP_TIER</th>\n",
       "      <th>CNV_GROUP_EVENT</th>\n",
       "      <th>hlvl_call</th>\n",
       "      <th>gene_cnv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129364723</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>EGFR_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>CDKN2A_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129332074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRF1</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>PRF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IKZF1</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>IKZF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129333379</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RET</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>RET_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21926</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>77678303</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARD11</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>CARD11_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21928</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>77404498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MET</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>MET_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21930</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>77336483</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ETV1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21931</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>76593171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BUB1B</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>BUB1B_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21933</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>76385523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>SMARCB1_cnv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3286 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id_drspatient id_uniquesample  id_icb  id_mj  VARIANT_CALL_ID  \\\n",
       "124        101954688   1019543884068   105.0  221.0        129364723   \n",
       "125        101954688   1019543884068   105.0  221.0        129354263   \n",
       "126        101954688   1019543884068   105.0  221.0        129332074   \n",
       "128        101954688   1019543884068   105.0  221.0        129354974   \n",
       "129        101954688   1019543884068   105.0  221.0        129333379   \n",
       "...              ...             ...     ...    ...              ...   \n",
       "21926       92361207    923612315805   108.0  255.0         77678303   \n",
       "21928       92361207    923612315805   108.0  255.0         77404498   \n",
       "21930       92361207    923612315805   108.0  255.0         77336483   \n",
       "21931       92361207    923612315805   108.0  255.0         76593171   \n",
       "21933       92361207    923612315805   108.0  255.0         76385523   \n",
       "\n",
       "       CNV_COPY_COUNT  CNV_COPY_COUNT_IN_REPORT  TUMOR_PURITY  PANEL_VERSION  \\\n",
       "124               NaN                       NaN          70.0            1.0   \n",
       "125               NaN                       NaN          70.0            1.0   \n",
       "126               NaN                       NaN          70.0            1.0   \n",
       "128               NaN                       NaN          70.0            1.0   \n",
       "129               NaN                       NaN          70.0            1.0   \n",
       "...               ...                       ...           ...            ...   \n",
       "21926             NaN                       NaN          80.0            1.0   \n",
       "21928             NaN                       NaN          80.0            1.0   \n",
       "21930             NaN                       NaN          80.0            1.0   \n",
       "21931             NaN                       NaN          80.0            1.0   \n",
       "21933             NaN                       NaN          80.0            1.0   \n",
       "\n",
       "      CANONICAL_GENE HARMONIZED_HUGO_GENE_NAME HARMONIZED_VARIANT_CLASS  \\\n",
       "124              NaN                      EGFR                       LA   \n",
       "125              NaN                    CDKN2A                     1DEL   \n",
       "126              NaN                      PRF1                     1DEL   \n",
       "128              NaN                     IKZF1                       LA   \n",
       "129              NaN                       RET                     1DEL   \n",
       "...              ...                       ...                      ...   \n",
       "21926            NaN                    CARD11                       LA   \n",
       "21928            NaN                       MET                       LA   \n",
       "21930            NaN                      ETV1                       LA   \n",
       "21931            NaN                     BUB1B                     1DEL   \n",
       "21933            NaN                   SMARCB1                     1DEL   \n",
       "\n",
       "       CNV_GROUP_ID CNV_GROUP_CYTOBAND CNV_GROUP_TIER CNV_GROUP_EVENT  \\\n",
       "124             NaN                NaN            NaN             NaN   \n",
       "125             NaN                NaN            NaN             NaN   \n",
       "126             NaN                NaN            NaN             NaN   \n",
       "128             NaN                NaN            NaN             NaN   \n",
       "129             NaN                NaN            NaN             NaN   \n",
       "...             ...                ...            ...             ...   \n",
       "21926           NaN                NaN            NaN             NaN   \n",
       "21928           NaN                NaN            NaN             NaN   \n",
       "21930           NaN                NaN            NaN             NaN   \n",
       "21931           NaN                NaN            NaN             NaN   \n",
       "21933           NaN                NaN            NaN             NaN   \n",
       "\n",
       "       hlvl_call     gene_cnv  \n",
       "124            1     EGFR_cnv  \n",
       "125           -1   CDKN2A_cnv  \n",
       "126           -1     PRF1_cnv  \n",
       "128            1    IKZF1_cnv  \n",
       "129           -1      RET_cnv  \n",
       "...          ...          ...  \n",
       "21926          1   CARD11_cnv  \n",
       "21928          1      MET_cnv  \n",
       "21930          1     ETV1_cnv  \n",
       "21931         -1    BUB1B_cnv  \n",
       "21933         -1  SMARCB1_cnv  \n",
       "\n",
       "[3286 rows x 18 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "55c63f1b-a354-4555-9fe1-ee9096c348b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_cnv_df = op_cnv_df.join(op_sample_df[[\"id_uniquesample\", \"sample_id_merge\"]].set_index(\"id_uniquesample\"), on=\"id_uniquesample\")\n",
    "op_cnv_df = keep_latest_preicb_timepoint_variants(op_cnv_df, icb_sample_df, sample_id_col=\"sample_id_merge\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "29912f4c-a459-4201-b192-8695aea1bb45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_cnv</th>\n",
       "      <th>ABL1_cnv</th>\n",
       "      <th>AKT1_cnv</th>\n",
       "      <th>AKT2_cnv</th>\n",
       "      <th>AKT3_cnv</th>\n",
       "      <th>ALK_cnv</th>\n",
       "      <th>APC_cnv</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>ARID1B_cnv</th>\n",
       "      <th>ARID2_cnv</th>\n",
       "      <th>AR_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>TP53_cnv</th>\n",
       "      <th>TSC1_cnv</th>\n",
       "      <th>TSC2_cnv</th>\n",
       "      <th>U2AF1_cnv</th>\n",
       "      <th>VHL_cnv</th>\n",
       "      <th>WRN_cnv</th>\n",
       "      <th>WT1_cnv</th>\n",
       "      <th>XPA_cnv</th>\n",
       "      <th>XPC_cnv</th>\n",
       "      <th>ZNF217_cnv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-102</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-103</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-104</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 229 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_cnv               ABL1_cnv  AKT1_cnv  AKT2_cnv  AKT3_cnv  ALK_cnv  \\\n",
       "participant_id_legacy                                                    \n",
       "GBM.ICB-1                     0         0         0         0        0   \n",
       "GBM.ICB-10                    0         0         0         0        0   \n",
       "GBM.ICB-102                   0         0         0         0        0   \n",
       "GBM.ICB-103                   0         0        -1         0        0   \n",
       "GBM.ICB-104                   0         0         0         0        0   \n",
       "\n",
       "gene_cnv               APC_cnv  ARID1A_cnv  ARID1B_cnv  ARID2_cnv  AR_cnv  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-1                    0           0           0          0       0   \n",
       "GBM.ICB-10                   0           0          -1          0       0   \n",
       "GBM.ICB-102                  0           0          -1          0       0   \n",
       "GBM.ICB-103                  0           0           0         -1       0   \n",
       "GBM.ICB-104                  0           0           0          0       0   \n",
       "\n",
       "gene_cnv               ...  TP53_cnv  TSC1_cnv  TSC2_cnv  U2AF1_cnv  VHL_cnv  \\\n",
       "participant_id_legacy  ...                                                     \n",
       "GBM.ICB-1              ...         0         0         0          1        0   \n",
       "GBM.ICB-10             ...         0         0         0          0        0   \n",
       "GBM.ICB-102            ...         0         0         0          0        0   \n",
       "GBM.ICB-103            ...         0         0         0          0        0   \n",
       "GBM.ICB-104            ...         0         0         0          0        0   \n",
       "\n",
       "gene_cnv               WRN_cnv  WT1_cnv  XPA_cnv  XPC_cnv  ZNF217_cnv  \n",
       "participant_id_legacy                                                  \n",
       "GBM.ICB-1                    0        0        0        0           0  \n",
       "GBM.ICB-10                   0        0        0        0           0  \n",
       "GBM.ICB-102                  0        0        0        0           0  \n",
       "GBM.ICB-103                  0        0        0        0           0  \n",
       "GBM.ICB-104                  0        0        0        0           0  \n",
       "\n",
       "[5 rows x 229 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnvs_matrix = op_cnv_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_cnv', values = 'hlvl_call', aggfunc = 'sum', fill_value = 0)\n",
    "\n",
    "op_cnvs_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "d4b2a6aa-02e2-43ae-9770-eaae83e33f57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGxCAYAAADCo9TSAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAAK6tJREFUeJzt3XtUVWXCx/HfEeEAgnjFKwqpk/e7maYpWYx4Ga03UytFo3m9pTn2VlpjiJrkZTk2Y14zL6/XrNFMnZJMyXnVQitLGytNU/NuCqmJgc/7R4uzPB5AqMNzQL+ftc5anX02ez97s9l+2+dscBhjjAAAACwp4esBAACA2wvxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxUYQsWrRIDodDDodDW7du9XjdGKPatWvL4XCoY8eOXl13ZGSkBgwY4Hq+devWXMdhy7Zt2/TII4+oWrVqCggIUFhYmNq2bavZs2fr0qVLrvkiIyPlcDg0ePBgj2Vkb8dbb70lSXrwwQcVFBSkCxcu5Lrexx57TP7+/jp16pTXt8mWjRs3aty4cb4ehs989dVXGjdunA4fPpyv+bN/9vI7f35dunRJr7zyipo1a6aQkBCVKlVKTZs21aRJk9yOYV86fPiw67zjcDhUokQJlS9fXl26dNGOHTt8PTxJUseOHX/XOe/G81t+Xb58WePGjfPpefBWRXwUQaGhoVqwYIHH9JSUFB08eFChoaGFPobmzZtrx44dat68eaGvKycJCQm699579cMPP2jChAlKTk7WypUr1alTJ40bN05//etfPb5mwYIF+vrrr/Ncbnx8vK5cuaLly5fn+HpaWprWrFmjbt26qVKlSl7ZFl/YuHGjEhMTfT0Mn/nqq6+UmJiY75jo2rWrduzYoSpVqnhtDKdOndLdd9+t8ePH649//KPWrFmjtWvXKjY2VhMnTtTdd99dpAJ3+PDh2rFjh7Zt26akpCTt2bNH0dHR+uyzz3w9NJ+5fPmyEhMTiY9CUNLXA4Cn3r17a9myZXrttddUunRp1/QFCxaoTZs2Sk9PL/QxlC5dWnfffXehrycnq1ev1vjx4xUfH6/58+fL4XC4XouNjdVzzz3n8X9kbdq00VdffaUXXnhBb7/9dq7Ljo2NVdWqVfXGG29o6NChHq+vWLFCP//8s+Lj4723Qci3rKwsZWZmyul0Wl1vxYoVVbFiRa8us3///tq/f7+2bNmidu3auaY/8MAD6tq1q6KjoxUXF6f33nvPq+v9rWrUqOH6mb/nnntUu3ZtderUSbNmzdL8+fN9PDrcarjyUQT17dtX0q//EGZLS0vT22+/rSeeeCLHr7l69aomTpyounXryul0qmLFiho4cKDOnDnjNt8vv/yi5557TpUrV1ZwcLDatWunTz75xGN5Ob3tsmvXLvXp00eRkZEKCgpSZGSk+vbtq++//97ta7MvYW/ZskVDhgxRhQoVVL58eT300EM6fvz4Tbd//PjxKlu2rP7+97+7hUe20NBQxcTEuE0rV66cRo8erX/+85/auXNnrsv28/NTXFycdu/erS+//NLj9YULF6pKlSqKjY296TiXL1+uNm3aKCQkRCEhIWratKnHFas33nhDTZo0UWBgoMqVK6cHH3xQ//nPf9zmye2S8oABAxQZGel6nn15fNq0aZo+fbqioqIUEhKiNm3auG3zgAED9Nprr0mS2+X07KsAq1evVuvWrRUWFqbg4GDdcccduR5X13M4HHrqqac0d+5c/eEPf5DT6VT9+vW1cuVKj3lPnjypQYMGqXr16goICFBUVJQSExOVmZnpsT1TpkzRxIkTFRUVJafTqS1btvyuMSxatEi9evWSJEVHR7u2f9GiRbkuN6e3XTp27KiGDRsqNTVV7du3d+2rV155RdeuXctzX+3atUubNm1SfHy8W3hka9eunZ544gm9//772r17t8f2/e///q/q1aun4OBgNWnSROvXr/dYxv79+9W3b19VqlRJTqdTNWrUUP/+/ZWRkeGaZ+/everRo4fKli2rwMBANW3aVIsXL85z7NmyQyT753vVqlWKiYlRlSpVFBQUpHr16mn06NEebx8NGDBAISEhOnDggLp06aKQkBBFRETomWeecRublP/zVn7l9/x25swZDR06VPXr11dISIjCw8N13333adu2ba55Dh8+7ArSxMRE13GU/fbNgQMHNHDgQNWpU0fBwcGqVq2aunfvnuN5BTkwKDIWLlxoJJnU1FTTr18/c9ddd7lemz17tilVqpRJT083DRo0MB06dHC9lpWVZTp37mxKlSplEhMTTXJysnn99ddNtWrVTP369c3ly5dd88bFxRmHw2GeffZZs2nTJjN9+nRTrVo1U7p0aRMXF+eab8uWLUaS2bJli2va6tWrzUsvvWTWrFljUlJSzMqVK02HDh1MxYoVzZkzZzy244477jDDhw8377//vnn99ddN2bJlTXR0dJ774Pjx40aS6d27d773W82aNU3Xrl3N5cuXTbVq1Uz79u09tmP16tWuad9++61xOBxm5MiRbsvZt2+fkWRGjx5903WOHTvWSDIPPfSQWb16tWtfjh071jXPpEmTjCTTt29fs2HDBrNkyRJzxx13mLCwMPPNN9+45uvQoYPb9zNbXFycqVmzpuv5oUOHjCQTGRlpOnfubNauXWvWrl1rGjVqZMqWLWsuXLhgjDHmwIED5uGHHzaSzI4dO1yPK1eumO3btxuHw2H69OljNm7caD788EOzcOFC069fv5tusyQTERFh6tevb1asWGHWrVtnOnfu7LF/T5w4YSIiIkzNmjXN3LlzzQcffGAmTJhgnE6nGTBggMf2VKtWzURHR5u33nrLbNq0yRw6dOh3jeH06dOuff/aa6+5tv/06dO5Ljf7mL1+3R06dDDly5c3derUMXPmzDHJyclm6NChRpJZvHhxnvsqe/3/+te/cp1n48aNRpJJSkpy277IyEhz1113mTfffNNs3LjRdOzY0ZQsWdIcPHjQNd/nn39uQkJCTGRkpJkzZ47ZvHmzWbp0qXnkkUdMenq6McaY/fv3m9DQUFOrVi2zZMkSs2HDBtO3b18jyUyePNm1rOzvw9SpU93Gt2fPHiPJPProo8YYYyZMmGD+9re/mQ0bNpitW7eaOXPmmKioKI+f6bi4OBMQEGDq1atnpk2bZj744APz0ksvGYfDYRITE13zFeS8ldvPyI3ye37bv3+/GTJkiFm5cqXZunWrWb9+vYmPjzclSpRwnfOuXLli3nvvPSPJxMfHu46jAwcOGGOMSUlJMc8884x56623TEpKilmzZo3p2bOnCQoKMvv377/pWG93xEcRcn18ZP+juXfvXmOMMa1atXKduG+MjxUrVhhJ5u2333ZbXmpqqpFkZs2aZYwx5j//+Y+RZP7yl7+4zbds2TIj6abxcaPMzExz8eJFU6pUKfPqq696bMfQoUPd5p8yZYqRZE6cOJHrMnfu3JnvAMiWHR/GGDN//nwjybz77rtu23H9P47G/Hoyq1Chgrl69apr2jPPPGMkuYVBTr777jvj5+dnHnvssVznOX/+vAkKCjJdunRxm37kyBHjdDpdJ/TssRQkPho1amQyMzNd0z/55BMjyaxYscI1bdiwYSan/7eYNm2akeQKlYKQZIKCgszJkydd0zIzM03dunVN7dq1XdMGDRpkQkJCzPfff5/juvft2+e2PbVq1XL7PnhjDKtXr77p8Xu93OJDkvn444/d5q1fv7754x//mOfyBg8ebCTl+Y9Q9s/jkCFD3LavUqVKroAwxpiTJ0+aEiVKuEXKfffdZ8qUKZNnUPXp08c4nU5z5MgRt+mxsbEmODjYdQxkfx8mT55sfvnlF3PlyhWze/du06pVKyPJbNiwwWPZ165dM7/88otJSUkxksyePXtcr8XFxRlJ5s0333T7mi5dupg777zT9Ty/5y1j8hcfBTm/3SgzM9P88ssvplOnTubBBx90TT9z5oyRZBISEvJcd/Yyrl69aurUqeMxBnjibZciqkOHDqpVq5beeOMNffnll0pNTc310vj69etVpkwZde/eXZmZma5H06ZNVblyZddbJ9mXsx977DG3r3/kkUdUsuTNP/5z8eJFPf/886pdu7ZKliypkiVLKiQkRJcuXfJ4K0GS/vSnP7k9b9y4sSR5vE3jTQMHDlT9+vU1evToPC+Nx8fH6+zZs1q3bp0kKTMzU0uXLlX79u1Vp06dPNeRnJysrKwsDRs2LNd5duzYoZ9//tnjE/YRERG67777tHnz5vxv1A26du0qPz8/1/OC7NdWrVpJ+vV7/uabb+qHH34o0Lo7derk9kFcPz8/9e7dWwcOHNCxY8ck/Xo8RkdHq2rVqm7HY/ZbWSkpKW7L/NOf/iR/f3+vjsFbKleurLvuusttWuPGjb1yDBtjJMnjrcXo6Gi3D5VXqlRJ4eHhrnVevnxZKSkpeuSRR/L8nMqHH36oTp06KSIiwm36gAEDdPnyZY/PTT3//PPy9/dXYGCgWrRooSNHjmju3Lnq0qWLJOm7777To48+qsqVK8vPz0/+/v7q0KGDJHn8/DscDnXv3t1t2o37Lb/nrfwq6Pltzpw5at68uQIDA1WyZEn5+/tr8+bNOZ7LcpKZmalJkyapfv36CggIUMmSJRUQEKBvv/0238u4nREfRZTD4dDAgQO1dOlSzZkzR3/4wx/Uvn37HOc9deqULly4oICAAPn7+7s9Tp48qbNnz0qSzp07J+nXE+r1SpYsqfLly990TI8++qhmzpypJ598Uu+//74++eQTpaamqmLFivr555895r9xmdkfIsxp3mw1atSQJB06dOim48mJn5+fJk2apH379uX53vbDDz+ssLAwLVy4UNKvd4ecOnUqXx80zX4/unr16rnOk72vc7p7omrVqq7Xf4vfsl+z3XvvvVq7dq0yMzPVv39/Va9eXQ0bNnT7fFFebjx2rp+WvU2nTp3Su+++63EsNmjQQJJcx2O2gt5hkp8xeEtOPxdOp/Om+zo/x3H250tujIObrfP8+fPKysrK8/iTft0XuR1/2a9f7+mnn1Zqaqp2796tgwcP6sSJE/rv//5vSb/+j0f79u318ccfa+LEidq6datSU1P1z3/+U5LnsRccHKzAwECPbbhy5YrreX7PW/lVkPPb9OnTNWTIELVu3Vpvv/22du7cqdTUVHXu3DlfP0eSNGrUKI0dO1Y9e/bUu+++q48//lipqalq0qRJvpdxO+NulyJswIABeumllzRnzhy9/PLLuc6X/YHO3D41n/1/Udk/gCdPnlS1atVcr2dmZt70pJ2Wlqb169crISFBo0ePdk3PyMjQjz/+mO9tupkqVaqoUaNG2rRpky5fvqzg4OACL6NHjx665557lJCQoHnz5uU4T1BQkPr27av58+frxIkTeuONNxQaGur6oGJesv9v89ixYx7/cGTL3tcnTpzweO348eOqUKGC63lgYKDS0tI85ivoyTe/evTooR49eigjI0M7d+5UUlKSHn30UUVGRqpNmzZ5fu3JkydznZa9zRUqVFDjxo1zPWaz//HLltOHin/vGHztgQce0AsvvKC1a9eqc+fOOc6zdu1a17wFUa5cOfn5+d30Kk/58uVzPf4kuR2D0q8x3bJlyxyX9eGHH+r48ePaunWr62qHpDx/X87N5Pe8lV8FOb8tXbpUHTt21OzZs92m//TTT/le39KlS9W/f39NmjTJbfrZs2dVpkyZAo39dsSVjyKsWrVqevbZZ9W9e3fFxcXlOl+3bt107tw5ZWVlqWXLlh6PO++8U5Jcd1QsW7bM7evffPNNt7sQcuJwOGSM8bgF8vXXX1dWVtZv2LrcjR07VufPn9eIESNcl6avd/HiRW3atCnPZUyePFlHjx7V3//+91zniY+PV1ZWlqZOnaqNGzeqT58++YqdmJgY+fn5eZy4rtemTRsFBQVp6dKlbtOPHTvmuhyeLTIyUt98843bnQDnzp3T9u3bbzqW3OTnaojT6VSHDh00efJkScrX73PYvHmz2++myMrK0qpVq1SrVi3X/4l369ZNe/fuVa1atXI8Hm+Mj4LKzxgKcjWoMLRs2VIxMTFasGCB/u///s/j9X//+99644031LlzZ7Vo0aJAyw4KClKHDh20evXqPAO1U6dOrmi43pIlSxQcHFygW+mzA/HGn/+5c+cWYOTu8nveyq+CnN8cDofHtnzxxRceb0XldRzltIwNGzYU+K3M2xVXPoq4V1555abz9OnTR8uWLVOXLl309NNP66677pK/v7+OHTumLVu2qEePHnrwwQdVr149Pf7445oxY4b8/f11//33a+/evZo2bZrb7xPJSenSpXXvvfdq6tSpqlChgiIjI5WSkqIFCxZ4vfJ79eqlsWPHasKECdq/f7/i4+NVq1YtXb58WR9//LHmzp2r3r17e9xue7177rlHPXr00DvvvJPrPC1btlTjxo01Y8YMGWPy/bs9IiMj9cILL2jChAn6+eef1bdvX4WFhemrr77S2bNnlZiYqDJlymjs2LF64YUX1L9/f/Xt21fnzp1TYmKiAgMDlZCQ4Fpev379NHfuXD3++OP685//rHPnzmnKlCk3/Z7kpVGjRpJ+jbDY2Fj5+fmpcePGmjhxoo4dO6ZOnTqpevXqunDhgl599VW39+/zUqFCBd13330aO3asSpUqpVmzZmn//v1ut7qOHz9eycnJatu2rUaMGKE777xTV65c0eHDh7Vx40bNmTPnpm8Z/N4xNGzYUJI0b948hYaGKjAwUFFRUVavjCxZskT333+/YmJiNGLECFdwfvjhh3r11VdVt27dPG//zcv06dPVrl07tW7dWqNHj1bt2rV16tQprVu3TnPnzlVoaKgSEhJcn7956aWXVK5cOS1btkwbNmzQlClTFBYWlu/1tW3bVmXLltXgwYOVkJAgf39/LVu2THv27PlN45fyf97Kr4Kc37p166YJEyYoISFBHTp00Ndff63x48crKirKLVRCQ0NVs2ZNvfPOO+rUqZPKlSvnOv9169ZNixYtUt26ddW4cWPt3r1bU6dO/V3H9m3Ft593xfWuv9slLzfe7WKMMb/88ouZNm2aadKkiQkMDDQhISGmbt26ZtCgQebbb791zZeRkWGeeeYZEx4ebgIDA83dd99tduzYYWrWrHnTu12OHTtm/uu//suULVvWhIaGms6dO5u9e/d6fG1u25GfO2iul5KSYh5++GFTpUoV4+/vb0qXLm3atGljpk6d6nY3wPV3u1zvq6++Mn5+fjne7ZLt1VdfNZJM/fr18zWm6y1ZssS0atXKtb+bNWtmFi5c6DbP66+/bho3bmwCAgJMWFiY6dGjh+tuj+stXrzY1KtXzwQGBpr69eubVatW5Xq3y423RBpjPD6Rn5GRYZ588klTsWJF43A4XHdyrF+/3sTGxppq1aqZgIAAEx4ebrp06WK2bdt20+2VZIYNG2ZmzZplatWqZfz9/U3dunXNsmXLPOY9c+aMGTFihImKijL+/v6mXLlypkWLFubFF180Fy9evOn2eGMMM2bMMFFRUa5j4MbvzfVyu9ulQYMGHvPe+H3Jy8WLF82kSZNM06ZNTXBwsAkODjaNGzc2EydOdO2HnLbvRjf+jBnz6/Hdq1cvU758eRMQEGBq1KhhBgwYYK5cueKa58svvzTdu3c3YWFhJiAgwDRp0sRjP+T3+7B9+3bTpk0bExwcbCpWrGiefPJJ8+mnn3rs27i4OFOqVCmPr09ISPC4Ayu/56383mqb3/NbRkaG+Z//+R9TrVo1ExgYaJo3b27Wrl2b4/f2gw8+MM2aNTNOp9Ptrpnz58+b+Ph4Ex4eboKDg027du3Mtm3b8j3W253DmByuawPADRwOh4YNG6aZM2fe1mMA8PvxmQ8AAGAV8QEAAKzibRcAAGAVVz4AAIBVxAcAALCK+AAAAFYVuV8ydu3aNR0/flyhoaEF/rXLAADAN4wx+umnn1S1alWVKJH3tY0iFx/Hjx/P9e9lAACAou3o0aM3/U2vRS4+sv+Y0NGjR3/Xr5cGAAD2pKenKyIiIl9/FLDIxUf2Wy2lS5cmPgAAKGby85EJPnAKAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVBY6Pjz76SN27d1fVqlXlcDi0du1at9eNMRo3bpyqVq2qoKAgdezYUfv27fPWeAEAQDFX4Pi4dOmSmjRpopkzZ+b4+pQpUzR9+nTNnDlTqampqly5sh544AH99NNPv3uwAACg+CvwH5aLjY1VbGxsjq8ZYzRjxgy9+OKLeuihhyRJixcvVqVKlbR8+XINGjTo940WAAAUe179zMehQ4d08uRJxcTEuKY5nU516NBB27dvz/FrMjIylJ6e7vYAAAC3rgJf+cjLyZMnJUmVKlVym16pUiV9//33OX5NUlKSEhMTvTkMFEGRozf4egjFxuFXuvp6CABQqArlbheHw+H23BjjMS3bmDFjlJaW5nocPXq0MIYEAACKCK9e+ahcubKkX6+AVKlSxTX99OnTHldDsjmdTjmdTm8OAwAAFGFevfIRFRWlypUrKzk52TXt6tWrSklJUdu2bb25KgAAUEwV+MrHxYsXdeDAAdfzQ4cO6fPPP1e5cuVUo0YNjRw5UpMmTVKdOnVUp04dTZo0ScHBwXr00Ue9OnAAAFA8FTg+du3apejoaNfzUaNGSZLi4uK0aNEiPffcc/r55581dOhQnT9/Xq1bt9amTZsUGhrqvVEDAIBiy2GMMb4exPXS09MVFhamtLQ0lS5d2tfDgZdwt0v+cbcLgOKoIP9+87ddAACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAqr8dHZmam/vrXvyoqKkpBQUG64447NH78eF27ds3bqwIAAMVQSW8vcPLkyZozZ44WL16sBg0aaNeuXRo4cKDCwsL09NNPe3t1AACgmPF6fOzYsUM9evRQ165dJUmRkZFasWKFdu3a5e1VAQCAYsjrb7u0a9dOmzdv1jfffCNJ2rNnj/7973+rS5cuOc6fkZGh9PR0twcAALh1ef3Kx/PPP6+0tDTVrVtXfn5+ysrK0ssvv6y+ffvmOH9SUpISExO9PQwA8JnI0Rt8PYRi4/ArXX09BPiA1698rFq1SkuXLtXy5cv16aefavHixZo2bZoWL16c4/xjxoxRWlqa63H06FFvDwkAABQhXr/y8eyzz2r06NHq06ePJKlRo0b6/vvvlZSUpLi4OI/5nU6nnE6nt4cBAACKKK9f+bh8+bJKlHBfrJ+fH7faAgAASYVw5aN79+56+eWXVaNGDTVo0ECfffaZpk+frieeeMLbqwIAAMWQ1+PjH//4h8aOHauhQ4fq9OnTqlq1qgYNGqSXXnrJ26sCAADFkNfjIzQ0VDNmzNCMGTO8vWgAAHAL4G+7AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsKpT4+OGHH/T444+rfPnyCg4OVtOmTbV79+7CWBUAAChmSnp7gefPn9c999yj6Oho/etf/1J4eLgOHjyoMmXKeHtVAACgGPJ6fEyePFkRERFauHCha1pkZKS3VwMAAIopr7/tsm7dOrVs2VK9evVSeHi4mjVrpvnz5+c6f0ZGhtLT090eAADg1uX1+Pjuu+80e/Zs1alTR++//74GDx6sESNGaMmSJTnOn5SUpLCwMNcjIiLC20MCAABFiNfj49q1a2revLkmTZqkZs2aadCgQfrzn/+s2bNn5zj/mDFjlJaW5nocPXrU20MCAABFiNfjo0qVKqpfv77btHr16unIkSM5zu90OlW6dGm3BwAAuHV5PT7uueceff31127TvvnmG9WsWdPbqwIAAMWQ1+PjL3/5i3bu3KlJkybpwIEDWr58uebNm6dhw4Z5e1UAAKAY8np8tGrVSmvWrNGKFSvUsGFDTZgwQTNmzNBjjz3m7VUBAIBiyOu/50OSunXrpm7duhXGogEAQDHH33YBAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKsKPT6SkpLkcDg0cuTIwl4VAAAoBgo1PlJTUzVv3jw1bty4MFcDAACKkUKLj4sXL+qxxx7T/PnzVbZs2Vzny8jIUHp6utsDAADcugotPoYNG6auXbvq/vvvz3O+pKQkhYWFuR4RERGFNSQAAFAEFEp8rFy5Up9++qmSkpJuOu+YMWOUlpbmehw9erQwhgQAAIqIkt5e4NGjR/X0009r06ZNCgwMvOn8TqdTTqfT28MAAABFlNfjY/fu3Tp9+rRatGjhmpaVlaWPPvpIM2fOVEZGhvz8/Ly9WgAAUEx4PT46deqkL7/80m3awIEDVbduXT3//POEBwAAtzmvx0doaKgaNmzoNq1UqVIqX768x3QAAHD74TecAgAAq7x+5SMnW7dutbEaAABQDHDlAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgldfjIykpSa1atVJoaKjCw8PVs2dPff31195eDQAAKKa8Hh8pKSkaNmyYdu7cqeTkZGVmZiomJkaXLl3y9qoAAEAxVNLbC3zvvffcni9cuFDh4eHavXu37r33Xm+vDgAAFDNej48bpaWlSZLKlSuX4+sZGRnKyMhwPU9PTy/sIQEAAB8q1A+cGmM0atQotWvXTg0bNsxxnqSkJIWFhbkeERERhTkkAADgY4UaH0899ZS++OILrVixItd5xowZo7S0NNfj6NGjhTkkAADgY4X2tsvw4cO1bt06ffTRR6pevXqu8zmdTjmdzsIaBgAAKGK8Hh/GGA0fPlxr1qzR1q1bFRUV5e1VAACAYszr8TFs2DAtX75c77zzjkJDQ3Xy5ElJUlhYmIKCgry9OgAAUMx4/TMfs2fPVlpamjp27KgqVaq4HqtWrfL2qgAAQDFUKG+7AAAA5Ia/7QIAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsIr4AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFXEBwAAsKrQ4mPWrFmKiopSYGCgWrRooW3bthXWqgAAQDFSKPGxatUqjRw5Ui+++KI+++wztW/fXrGxsTpy5EhhrA4AABQjhRIf06dPV3x8vJ588knVq1dPM2bMUEREhGbPnl0YqwMAAMVISW8v8OrVq9q9e7dGjx7tNj0mJkbbt2/3mD8jI0MZGRmu52lpaZKk9PR0bw8NPnQt47Kvh1BscOwXfxzv+cfxfuvI/l4aY246r9fj4+zZs8rKylKlSpXcpleqVEknT570mD8pKUmJiYke0yMiIrw9NKBYCJvh6xEA9nC833p++uknhYWF5TmP1+Mjm8PhcHtujPGYJkljxozRqFGjXM+vXbumH3/8UeXLl89x/ltRenq6IiIidPToUZUuXdrXw7ltsN/tY5/7BvvdN263/W6M0U8//aSqVavedF6vx0eFChXk5+fncZXj9OnTHldDJMnpdMrpdLpNK1OmjLeHVSyULl36tjhAixr2u33sc99gv/vG7bTfb3bFI5vXP3AaEBCgFi1aKDk52W16cnKy2rZt6+3VAQCAYqZQ3nYZNWqU+vXrp5YtW6pNmzaaN2+ejhw5osGDBxfG6gAAQDFSKPHRu3dvnTt3TuPHj9eJEyfUsGFDbdy4UTVr1iyM1RV7TqdTCQkJHm8/oXCx3+1jn/sG+9032O+5c5j83BMDAADgJfxtFwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER9FyOHDhxUfH6+oqCgFBQWpVq1aSkhI0NWrV309tFveyy+/rLZt2yo4OPi2/Q27NsyaNUtRUVEKDAxUixYttG3bNl8P6Zb30UcfqXv37qpataocDofWrl3r6yHd8pKSktSqVSuFhoYqPDxcPXv21Ndff+3rYRUpxEcRsn//fl27dk1z587Vvn379Le//U1z5szRCy+84Ouh3fKuXr2qXr16aciQIb4eyi1r1apVGjlypF588UV99tlnat++vWJjY3XkyBFfD+2WdunSJTVp0kQzZ8709VBuGykpKRo2bJh27typ5ORkZWZmKiYmRpcuXfL10IoMfs9HETd16lTNnj1b3333na+HcltYtGiRRo4cqQsXLvh6KLec1q1bq3nz5po9e7ZrWr169dSzZ08lJSX5cGS3D4fDoTVr1qhnz56+Hspt5cyZMwoPD1dKSoruvfdeXw+nSODKRxGXlpamcuXK+XoYwO9y9epV7d69WzExMW7TY2JitH37dh+NCrAjLS1NkjiXX4f4KMIOHjyof/zjH/xNHBR7Z8+eVVZWlsdftq5UqZLHX8AGbiXGGI0aNUrt2rVTw4YNfT2cIoP4sGDcuHFyOBx5Pnbt2uX2NcePH1fnzp3Vq1cvPfnkkz4aefH2W/Y7CpfD4XB7bozxmAbcSp566il98cUXWrFiha+HUqQUyh+Wg7unnnpKffr0yXOeyMhI138fP35c0dHRrr8IjN+moPsdhadChQry8/PzuMpx+vRpj6shwK1i+PDhWrdunT766CNVr17d18MpUogPCypUqKAKFSrka94ffvhB0dHRatGihRYuXKgSJbg49VsVZL+jcAUEBKhFixZKTk7Wgw8+6JqenJysHj16+HBkgPcZYzR8+HCtWbNGW7duVVRUlK+HVOQQH0XI8ePH1bFjR9WoUUPTpk3TmTNnXK9VrlzZhyO79R05ckQ//vijjhw5oqysLH3++eeSpNq1ayskJMS3g7tFjBo1Sv369VPLli1dV/WOHDnCZ5oK2cWLF3XgwAHX80OHDunzzz9XuXLlVKNGDR+O7NY1bNgwLV++XO+8845CQ0NdV/zCwsIUFBTk49EVEQZFxsKFC42kHB8oXHFxcTnu9y1btvh6aLeU1157zdSsWdMEBASY5s2bm5SUFF8P6Za3ZcuWHI/tuLg4Xw/tlpXbeXzhwoW+HlqRwe/5AAAAVvGBAgAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVf8PiFsCobkay2YAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_variant_median_rates(op_cnv_df, pt_col=\"participant_id_legacy\", variant_col=\"hlvl_call\")\n",
    "plt.title(\"Median CNV counts per pt in OncoPanel data\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b2800fe-6bdb-4a47-b7e1-09ef645fc540",
   "metadata": {},
   "source": [
    "# Prepare the ICB WES data\n",
    "\n",
    "Get the participants who have pre-ICB WES tumor data but NOT OncoPanel call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "cdb7439e-01da-47ca-b9cd-13850d589104",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_sample_df = icb_sample_df[(icb_sample_df[\"pre_post\"] ==  \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\")]\n",
    "pt_usable_data = pre_sample_df[[\"participant_id_legacy\", \"usable_wes\", \"usable_oncopanel\"]].groupby(\"participant_id_legacy\").any()\n",
    "pt_usable_data = pt_usable_data[pt_usable_data[\"usable_wes\"] & (~pt_usable_data[\"usable_oncopanel\"])]\n",
    "\n",
    "wes_pt_df = icb_pt_df[icb_pt_df[\"participant_id_legacy\"].isin(pt_usable_data.index)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596f1605-7297-426a-98c7-dd44a854acf2",
   "metadata": {},
   "source": [
    "### Load and process ICB WES sequence variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "9688a22b-1975-4bd6-9403-4d9f364831c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "wes_mut_df = pd.read_pickle('../../data/icb_wes_mut_ccfs_maf.pickle')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f23ca0c2-dcd7-4548-bc90-54e95f592660",
   "metadata": {},
   "source": [
    "Filter WES sequence variants by various criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "13c81242-ebf2-4787-8527-655c14f08d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants for pts without OncoPanel data\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"Patient_ID\"].isin(wes_pt_df[\"participant_id_legacy\"])]\n",
    "\n",
    "# Correct some sample IDs that erroneously end with \"_pair\"\n",
    "wes_mut_df[\"Sample_ID\"] = wes_mut_df[\"Sample_ID\"].map(lambda x: x[:-5] if x.endswith(\"_pair\") else x)\n",
    "\n",
    "# Filter by CCF\n",
    "cluster_ccf_threshold = 0.1\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"clust_ccf_mean\"] > cluster_ccf_threshold]\n",
    "\n",
    "# Restrict to genes covered by OncoPanel\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"Hugo_Symbol\"].isin(oncopanel_genes[\"Gene\"])]\n",
    "\n",
    "# Remove silent and non-coding mutations. \n",
    "# This is consistent with the filtering of GLASS variants\n",
    "wes_mut_df = wes_mut_df[~wes_mut_df[\"variant_type\"].isin([\"silent\", \"non_coding\"])]\n",
    "\n",
    "# Keep only variants belonging to their latest pre-ICB timepoint.\n",
    "wes_mut_df = keep_latest_preicb_timepoint_variants(wes_mut_df, pre_sample_df, sample_id_col=\"Sample_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "4200ae71-3748-4d01-87c9-150e7d6c63ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_type\n",
       "neoantigen    26\n",
       "nonsense       4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_mut_df[\"variant_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "2e45eb41-967d-4d23-a17a-e4846f3b8838",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_df[\"gene_mut\"] = wes_mut_df[\"Hugo_Symbol\"]+\"_mut\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "a6c736e7-fe32-4515-bfbf-602afc6f86c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_matrix = wes_mut_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# In this methodology, sequence variants are given weight = 2\n",
    "wes_mut_matrix *= 2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "c760f96b-ab72-482e-8fe0-60f6c0492ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_mut</th>\n",
       "      <th>BRD4_mut</th>\n",
       "      <th>EGFR_mut</th>\n",
       "      <th>ERCC2_mut</th>\n",
       "      <th>FLT3_mut</th>\n",
       "      <th>GNAS_mut</th>\n",
       "      <th>KDR_mut</th>\n",
       "      <th>MITF_mut</th>\n",
       "      <th>MSH6_mut</th>\n",
       "      <th>NOTCH1_mut</th>\n",
       "      <th>PDGFRA_mut</th>\n",
       "      <th>PIK3CA_mut</th>\n",
       "      <th>PTEN_mut</th>\n",
       "      <th>RB1_mut</th>\n",
       "      <th>TET2_mut</th>\n",
       "      <th>TP53_mut</th>\n",
       "      <th>TSC1_mut</th>\n",
       "      <th>WT1_mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-139</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-144</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-212</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-253</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-28</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-30</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-32</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-37</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-43</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-69</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-84</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_mut               BRD4_mut  EGFR_mut  ERCC2_mut  FLT3_mut  GNAS_mut  \\\n",
       "participant_id_legacy                                                      \n",
       "GBM.ICB-139                 0.0       4.0        0.0       0.0       0.0   \n",
       "GBM.ICB-144                 0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-212                 0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-253                 2.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-28                  0.0       2.0        0.0       0.0       0.0   \n",
       "GBM.ICB-30                  0.0       2.0        0.0       0.0       0.0   \n",
       "GBM.ICB-32                  0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-37                  0.0       0.0        0.0       0.0       2.0   \n",
       "GBM.ICB-42                  0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-43                  0.0       0.0        2.0       0.0       0.0   \n",
       "GBM.ICB-69                  0.0       0.0        0.0       2.0       0.0   \n",
       "GBM.ICB-84                  0.0       0.0        0.0       0.0       0.0   \n",
       "\n",
       "gene_mut               KDR_mut  MITF_mut  MSH6_mut  NOTCH1_mut  PDGFRA_mut  \\\n",
       "participant_id_legacy                                                        \n",
       "GBM.ICB-139                0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-144                0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-212                0.0       0.0       0.0         2.0         0.0   \n",
       "GBM.ICB-253                0.0       2.0       0.0         0.0         0.0   \n",
       "GBM.ICB-28                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-30                 2.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-32                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-37                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-42                 0.0       0.0       0.0         0.0         2.0   \n",
       "GBM.ICB-43                 0.0       0.0       2.0         0.0         0.0   \n",
       "GBM.ICB-69                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-84                 0.0       0.0       0.0         0.0         0.0   \n",
       "\n",
       "gene_mut               PIK3CA_mut  PTEN_mut  RB1_mut  TET2_mut  TP53_mut  \\\n",
       "participant_id_legacy                                                      \n",
       "GBM.ICB-139                   0.0       6.0      0.0       0.0       0.0   \n",
       "GBM.ICB-144                   0.0       2.0      0.0       0.0       2.0   \n",
       "GBM.ICB-212                   2.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-253                   0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-28                    0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-30                    0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-32                    0.0       4.0      0.0       0.0       0.0   \n",
       "GBM.ICB-37                    0.0       2.0      0.0       0.0       2.0   \n",
       "GBM.ICB-42                    0.0       2.0      0.0       0.0       0.0   \n",
       "GBM.ICB-43                    0.0       0.0      0.0       2.0       0.0   \n",
       "GBM.ICB-69                    0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-84                    0.0       2.0      2.0       0.0       2.0   \n",
       "\n",
       "gene_mut               TSC1_mut  WT1_mut  \n",
       "participant_id_legacy                     \n",
       "GBM.ICB-139                 0.0      0.0  \n",
       "GBM.ICB-144                 0.0      0.0  \n",
       "GBM.ICB-212                 0.0      0.0  \n",
       "GBM.ICB-253                 0.0      0.0  \n",
       "GBM.ICB-28                  0.0      0.0  \n",
       "GBM.ICB-30                  0.0      0.0  \n",
       "GBM.ICB-32                  0.0      0.0  \n",
       "GBM.ICB-37                  0.0      2.0  \n",
       "GBM.ICB-42                  0.0      0.0  \n",
       "GBM.ICB-43                  2.0      0.0  \n",
       "GBM.ICB-69                  0.0      0.0  \n",
       "GBM.ICB-84                  0.0      0.0  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_mut_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3800996-3675-4244-9470-a9560a59295b",
   "metadata": {},
   "source": [
    "### Load and process ICB WES CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "982f90c4-3727-404f-aa85-b9088d269910",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df = pd.concat([\n",
    "    pd.read_csv('../../data/icb_wes_tnpair_gene_allelic_cn.maf', sep = '\\t'),\n",
    "    pd.read_csv('../../data/icb_wes_tonly_gene_allelic_cn.maf', sep = '\\t')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "1cad9964-a6ef-4c82-b590-718b2a56cd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some helper functions\n",
    "def call_hlvl(cn_col, thresholds=[0.8, 1.8, 2.2, 3.2], values=[-2,-1,0,1,2]): \n",
    "    dig = np.digitize(cn_col.values, thresholds)\n",
    "    hlvl = np.array(values)[dig]\n",
    "    return pd.Series(data=hlvl, index=cn_col.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "644b4b32-63a1-4fe2-8eef-5bda3e4e1c9e",
   "metadata": {},
   "source": [
    "Filter the variants by various criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "0c9037c7-cf2b-403a-9658-85842481f8f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants belonging to the latest pre-ICB timepoint for each pt\n",
    "wes_cnv_df = keep_latest_preicb_timepoint_variants(wes_cnv_df, icb_sample_df, sample_id_col=\"sample\")\n",
    "\n",
    "# Restrict to ICB pts that only have WES data (and not OncoPanel)\n",
    "wes_cnv_df = wes_cnv_df[wes_cnv_df[\"participant_id_legacy\"].isin(wes_pt_df[\"participant_id_legacy\"])]\n",
    "\n",
    "# Restrict to genes covered by OncoPanel\n",
    "wes_cnv_df = wes_cnv_df[wes_cnv_df[\"gene\"].isin(oncopanel_genes[\"Gene\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e6916e7-1cc0-411d-8230-d91d507e4dc8",
   "metadata": {},
   "source": [
    "Annotate gene-level copy number gains, losses, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "8ad22c0b-1572-4cbd-8bfc-66730b4bb605",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df[\"hlvl_call\"] = call_hlvl(wes_cnv_df[\"corrected_total_cn\"])\n",
    "\n",
    "# Use ABSOLUTE's homozygous deletions\n",
    "wes_cnv_df.loc[wes_cnv_df[\"HZ\"] == 1, \"hlvl_call\"] = -2\n",
    "\n",
    "# Finally: keep only the actual CNVs\n",
    "wes_cnv_df = wes_cnv_df.loc[wes_cnv_df[\"hlvl_call\"] != 0,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "aa5055d1-39cf-43ad-821e-a43c65c8580d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df[\"gene_cnv\"] = wes_cnv_df[\"gene\"] + \"_cnv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "651320b3-ae3e-4616-b214-31d0a20b99e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhYAAAGdCAYAAABO2DpVAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAAEuVJREFUeJzt3W9oXYXdwPFfbOltp0m2tms1NK1B2ebsKpiKa1Fp96csSLGDyRyjhKEvJmlZCWMz+kI7JhE29od1ZjpGZQxt2YuqsK2QgU0EyWg7i9JtorOSbFFrJ8tt87BbTO/z4sGwPP2jN/3d3Nz084EDnpNzPD8Ol/TLuSf3NpTL5XIAACS4rNYDAABzh7AAANIICwAgjbAAANIICwAgjbAAANIICwAgjbAAANLMn+kTnjlzJkZHR6OxsTEaGhpm+vQAwDSUy+U4efJktLS0xGWXnf++xIyHxejoaLS2ts70aQGABCMjI7FixYrz/nzGw6KxsTEi/m+wpqammT49ADANxWIxWltbJ/8dP58ZD4v33/5oamoSFgBQZz7oMQYPbwIAaYQFAJBGWAAAaYQFAJBGWAAAaYQFAJBGWAAAaYQFAJBGWAAAaYQFAJCmorB46KGHoqGhYcpy5ZVXVms2AKDOVPxdIddff3388Y9/nFyfN29e6kAAQP2qOCzmz5/vLgUAcE4VP2Px6quvRktLS7S1tcVdd90Vr7/++gX3L5VKUSwWpywAwNxU0R2Lm2++OX7961/HJz7xiXj77bfj+9//fqxfvz6OHj0aS5YsOecxvb29sXPnzpRhAWrt6vt+V+sR6sYbj9xe6xGogYZyuVye7sHj4+NxzTXXxHe+853o7u4+5z6lUilKpdLkerFYjNbW1hgbG4umpqbpnhqgJoTFhycs5pZisRjNzc0f+O93xc9Y/LfLL788PvOZz8Srr7563n0KhUIUCoWLOQ0AUCcu6nMsSqVS/PWvf42rrroqax4AoI5VFBbf/va3Y2BgII4dOxZ/+tOf4itf+UoUi8Xo7Oys1nwAQB2p6K2Qf/zjH/G1r30tTpw4ER//+Mfjs5/9bAwNDcWqVauqNR8AUEcqCos9e/ZUaw4AYA7wXSEAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkuaiw6O3tjYaGhtixY0fSOABAPZt2WBw8eDAef/zxWLNmTeY8AEAdm1ZYnDp1Kr7+9a/HL3/5y/jYxz6WPRMAUKemFRZdXV1x++23xxe+8IUP3LdUKkWxWJyyAABz0/xKD9izZ0/8+c9/joMHD36o/Xt7e2Pnzp0VDwYA1J+K7liMjIzEt771rfjNb34TCxcu/FDH9PT0xNjY2OQyMjIyrUEBgNmvojsWhw8fjuPHj0d7e/vktomJiRgcHIxdu3ZFqVSKefPmTTmmUChEoVDImRYAmNUqCovPf/7z8fLLL0/Z9o1vfCM+9alPxXe/+92zogIAuLRUFBaNjY2xevXqKdsuv/zyWLJkyVnbAYBLj0/eBADSVPxXIf/fgQMHEsYAAOYCdywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDQVhUVfX1+sWbMmmpqaoqmpKdatWxd/+MMfqjUbAFBnKgqLFStWxCOPPBKHDh2KQ4cOxec+97m444474ujRo9WaDwCoI/Mr2Xnz5s1T1h9++OHo6+uLoaGhuP7661MHAwDqT0Vh8d8mJibit7/9bYyPj8e6devOu1+pVIpSqTS5XiwWp3tKAGCWqzgsXn755Vi3bl385z//iSuuuCL27dsXn/70p8+7f29vb+zcufOihmR2u/q+39V6hLrxxiO313oEgKqq+K9CPvnJT8aRI0diaGgo7r333ujs7Iy//OUv592/p6cnxsbGJpeRkZGLGhgAmL0qvmOxYMGCuPbaayMiYu3atXHw4MH46U9/Go899tg59y8UClEoFC5uSgCgLlz051iUy+Upz1AAAJeuiu5Y3H///dHR0RGtra1x8uTJ2LNnTxw4cCD2799frfkAgDpSUVi8/fbbsXXr1njzzTejubk51qxZE/v3748vfvGL1ZoPAKgjFYXFr371q2rNAQDMAb4rBABIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIU1FY9Pb2xk033RSNjY2xbNmy2LJlS7zyyivVmg0AqDMVhcXAwEB0dXXF0NBQ9Pf3x3vvvRebNm2K8fHxas0HANSR+ZXsvH///inru3fvjmXLlsXhw4fjtttuSx0MAKg/FYXF/zc2NhYREYsXLz7vPqVSKUql0uR6sVi8mFMCALPYtB/eLJfL0d3dHbfcckusXr36vPv19vZGc3Pz5NLa2jrdUwIAs9y0w2Lbtm3x0ksvxVNPPXXB/Xp6emJsbGxyGRkZme4pAYBZblpvhWzfvj2effbZGBwcjBUrVlxw30KhEIVCYVrDAQD1paKwKJfLsX379ti3b18cOHAg2traqjUXAFCHKgqLrq6uePLJJ+OZZ56JxsbGeOuttyIiorm5ORYtWlSVAQGA+lHRMxZ9fX0xNjYWGzZsiKuuumpy2bt3b7XmAwDqSMVvhQAAnI/vCgEA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACBNxWExODgYmzdvjpaWlmhoaIinn366CmMBAPWo4rAYHx+PG264IXbt2lWNeQCAOja/0gM6Ojqio6OjGrMAAHWu4rCoVKlUilKpNLleLBarfUoAoEaq/vBmb29vNDc3Ty6tra3VPiUAUCNVD4uenp4YGxubXEZGRqp9SgCgRqr+VkihUIhCoVDt0wAAs4DPsQAA0lR8x+LUqVPx2muvTa4fO3Ysjhw5EosXL46VK1emDgcA1JeKw+LQoUOxcePGyfXu7u6IiOjs7IwnnngibTAAoP5UHBYbNmyIcrlcjVkAgDrnGQsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSzK/1AADwQa6+73e1HqFuvPHI7TU9vzsWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApJlWWDz66KPR1tYWCxcujPb29nj++eez5wIA6lDFYbF3797YsWNHPPDAA/Hiiy/GrbfeGh0dHTE8PFyN+QCAOlJxWPzoRz+Ku+++O+6555647rrr4ic/+Um0trZGX19fNeYDAOrI/Ep2Pn36dBw+fDjuu+++Kds3bdoUL7zwwjmPKZVKUSqVJtfHxsYiIqJYLFY6K7PUmdL/1HqEuuF1X/+83j+8zNe76/7hVev3zPv/33K5fMH9KgqLEydOxMTERCxfvnzK9uXLl8dbb711zmN6e3tj586dZ21vbW2t5NQwJzT/pNYTwMzxeq+Nal/3kydPRnNz83l/XlFYvK+hoWHKerlcPmvb+3p6eqK7u3ty/cyZM/Huu+/GkiVLznvMXFIsFqO1tTVGRkaiqamp1uNcMlz32nDda8N1n3mX4jUvl8tx8uTJaGlpueB+FYXF0qVLY968eWfdnTh+/PhZdzHeVygUolAoTNn20Y9+tJLTzglNTU2XzItvNnHda8N1rw3XfeZdatf8Qncq3lfRw5sLFiyI9vb26O/vn7K9v78/1q9fX9l0AMCcU/FbId3d3bF169ZYu3ZtrFu3Lh5//PEYHh6Ob37zm9WYDwCoIxWHxVe/+tX417/+Fd/73vfizTffjNWrV8fvf//7WLVqVTXmq3uFQiEefPDBs94Oorpc99pw3WvDdZ95rvn5NZQ/6O9GAAA+JN8VAgCkERYAQBphAQCkERYAQBphMUPeeOONuPvuu6OtrS0WLVoU11xzTTz44INx+vTpWo825z388MOxfv36+MhHPnJJfjjbTHn00Uejra0tFi5cGO3t7fH888/XeqQ5b3BwMDZv3hwtLS3R0NAQTz/9dK1HmvN6e3vjpptuisbGxli2bFls2bIlXnnllVqPNasIixnyt7/9Lc6cOROPPfZYHD16NH784x/HL37xi7j//vtrPdqcd/r06bjzzjvj3nvvrfUoc9bevXtjx44d8cADD8SLL74Yt956a3R0dMTw8HCtR5vTxsfH44Ybbohdu3bVepRLxsDAQHR1dcXQ0FD09/fHe++9F5s2bYrx8fFajzZr+HPTGvrBD34QfX198frrr9d6lEvCE088ETt27Ih///vftR5lzrn55pvjxhtvjL6+vslt1113XWzZsiV6e3trONmlo6GhIfbt2xdbtmyp9SiXlHfeeSeWLVsWAwMDcdttt9V6nFnBHYsaGhsbi8WLF9d6DLgop0+fjsOHD8emTZumbN+0aVO88MILNZoKZsbY2FhEhN/l/0VY1Mjf//73+NnPfuaj0Kl7J06ciImJibO+iHD58uVnfWEhzCXlcjm6u7vjlltuidWrV9d6nFlDWFykhx56KBoaGi64HDp0aMoxo6Oj8aUvfSnuvPPOuOeee2o0eX2bznWnuhoaGqasl8vls7bBXLJt27Z46aWX4qmnnqr1KLNKxd8VwlTbtm2Lu+6664L7XH311ZP/PTo6Ghs3bpz8Ajemp9LrTvUsXbo05s2bd9bdiePHj591FwPmiu3bt8ezzz4bg4ODsWLFilqPM6sIi4u0dOnSWLp06Yfa95///Gds3Lgx2tvbY/fu3XHZZW4YTVcl153qWrBgQbS3t0d/f398+ctfntze398fd9xxRw0ng3zlcjm2b98e+/btiwMHDkRbW1utR5p1hMUMGR0djQ0bNsTKlSvjhz/8YbzzzjuTP7vyyitrONncNzw8HO+++24MDw/HxMREHDlyJCIirr322rjiiitqO9wc0d3dHVu3bo21a9dO3o0bHh72DFGVnTp1Kl577bXJ9WPHjsWRI0di8eLFsXLlyhpONnd1dXXFk08+Gc8880w0NjZO3qlrbm6ORYsW1Xi6WaLMjNi9e3c5Is65UF2dnZ3nvO7PPfdcrUebU37+85+XV61aVV6wYEH5xhtvLA8MDNR6pDnvueeeO+dru7Ozs9ajzVnn+z2+e/fuWo82a/gcCwAgjTf5AYA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASPO/49O26AH5kqwAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_variant_median_rates(wes_cnv_df, pt_col=\"participant_id_legacy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "204cc9d2-fd78-45f7-800b-ac352bde7987",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_matrix = wes_cnv_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_cnv', aggfunc = 'sum', values = \"hlvl_call\", fill_value = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "812517ac-9778-4704-afd7-7f002f722f15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_cnv</th>\n",
       "      <th>AKT1_cnv</th>\n",
       "      <th>AKT3_cnv</th>\n",
       "      <th>ARID2_cnv</th>\n",
       "      <th>BAP1_cnv</th>\n",
       "      <th>BUB1B_cnv</th>\n",
       "      <th>CARD11_cnv</th>\n",
       "      <th>CCND1_cnv</th>\n",
       "      <th>CCND2_cnv</th>\n",
       "      <th>CCNE1_cnv</th>\n",
       "      <th>CDK4_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>PTEN_cnv</th>\n",
       "      <th>RAD21_cnv</th>\n",
       "      <th>RARA_cnv</th>\n",
       "      <th>RB1_cnv</th>\n",
       "      <th>RET_cnv</th>\n",
       "      <th>SETD2_cnv</th>\n",
       "      <th>SMARCB1_cnv</th>\n",
       "      <th>SOX2_cnv</th>\n",
       "      <th>TNFAIP3_cnv</th>\n",
       "      <th>TP53_cnv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-139</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-144</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-212</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-253</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-28</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-30</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-32</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-37</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-42</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-43</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-69</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-84</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_cnv               AKT1_cnv  AKT3_cnv  ARID2_cnv  BAP1_cnv  BUB1B_cnv  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-139                  -1         0          0         0         -1   \n",
       "GBM.ICB-144                   0         1          0         1          0   \n",
       "GBM.ICB-212                  -1         0          0         1          0   \n",
       "GBM.ICB-253                   2         1          2         1          2   \n",
       "GBM.ICB-28                    0         0          0         0          0   \n",
       "GBM.ICB-30                    0         1          0         0         -1   \n",
       "GBM.ICB-32                    0         0          0         0          0   \n",
       "GBM.ICB-37                    2         2          1         1          0   \n",
       "GBM.ICB-42                    0         0          0         0         -1   \n",
       "GBM.ICB-43                   -1         0          0         0         -1   \n",
       "GBM.ICB-69                   -1         0          0         0          0   \n",
       "GBM.ICB-84                    0         0          0         0          0   \n",
       "\n",
       "gene_cnv               CARD11_cnv  CCND1_cnv  CCND2_cnv  CCNE1_cnv  CDK4_cnv  \\\n",
       "participant_id_legacy                                                          \n",
       "GBM.ICB-139                     3          0          0          3         0   \n",
       "GBM.ICB-144                     1          0          2          1         2   \n",
       "GBM.ICB-212                     1          0          0          0         0   \n",
       "GBM.ICB-253                     4          1          2          2         2   \n",
       "GBM.ICB-28                      1          0          0          0         0   \n",
       "GBM.ICB-30                      1          0          0          1         0   \n",
       "GBM.ICB-32                      0         -2          0          0         0   \n",
       "GBM.ICB-37                      2          1          1          1         1   \n",
       "GBM.ICB-42                      1          0          0          0         0   \n",
       "GBM.ICB-43                      0          0          0         -1         0   \n",
       "GBM.ICB-69                      1          0          0          0         0   \n",
       "GBM.ICB-84                      1          0          0          1         0   \n",
       "\n",
       "gene_cnv               ...  PTEN_cnv  RAD21_cnv  RARA_cnv  RB1_cnv  RET_cnv  \\\n",
       "participant_id_legacy  ...                                                    \n",
       "GBM.ICB-139            ...        -3          0         0        0       -3   \n",
       "GBM.ICB-144            ...         0          1         1        0        1   \n",
       "GBM.ICB-212            ...         0          0         0        0        0   \n",
       "GBM.ICB-253            ...         0          2         0        2        0   \n",
       "GBM.ICB-28             ...        -1          0         0        0        0   \n",
       "GBM.ICB-30             ...        -1          1         0        0       -1   \n",
       "GBM.ICB-32             ...         0          0         0        0        0   \n",
       "GBM.ICB-37             ...         0          0         1        0        0   \n",
       "GBM.ICB-42             ...        -1          0         0       -1       -1   \n",
       "GBM.ICB-43             ...        -1          0         0       -1       -1   \n",
       "GBM.ICB-69             ...        -1          0         0        0       -1   \n",
       "GBM.ICB-84             ...        -1          0         0       -1       -1   \n",
       "\n",
       "gene_cnv               SETD2_cnv  SMARCB1_cnv  SOX2_cnv  TNFAIP3_cnv  TP53_cnv  \n",
       "participant_id_legacy                                                           \n",
       "GBM.ICB-139                    0            0         0            0         0  \n",
       "GBM.ICB-144                    1            0         1            1         0  \n",
       "GBM.ICB-212                    1            0         0            0         0  \n",
       "GBM.ICB-253                    1            2         2            1         2  \n",
       "GBM.ICB-28                     0            2         0            0         0  \n",
       "GBM.ICB-30                     0            0         0            0         0  \n",
       "GBM.ICB-32                     0            0         0            0         0  \n",
       "GBM.ICB-37                     1            0         1            2         0  \n",
       "GBM.ICB-42                     0            0         0            0         0  \n",
       "GBM.ICB-43                     0           -1         0           -1         0  \n",
       "GBM.ICB-69                     0            0         0            0         0  \n",
       "GBM.ICB-84                     0            0         0            0         0  \n",
       "\n",
       "[12 rows x 45 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_cnv_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55bc7889-e530-4e53-bfd7-b73ad29fa5c4",
   "metadata": {},
   "source": [
    "# Combine all filtered variants for ICB cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "dd3e94fd-0650-4588-b270-cc14aefbf8f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(154, 356)\n"
     ]
    }
   ],
   "source": [
    "op_matrix = op_variants_matrix.join(op_cnvs_matrix, how = 'outer').fillna(0)\n",
    "print(op_matrix.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "9dd30925-560e-406a-b46d-08d4cf816dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_matrix = wes_mut_matrix[~wes_mut_matrix.index.isin(op_matrix.index)]\n",
    "icb_wes_matrix = wes_mut_matrix.join(wes_cnv_matrix, how=\"outer\").fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "5ec53d92-cbe0-49b6-b1f6-49dce8809ed8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>ASXL1_mut</th>\n",
       "      <th>ATM_mut</th>\n",
       "      <th>ATRX_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>U2AF1_cnv</th>\n",
       "      <th>VHL_cnv</th>\n",
       "      <th>WRN_cnv</th>\n",
       "      <th>WT1_cnv</th>\n",
       "      <th>XPA_cnv</th>\n",
       "      <th>XPC_cnv</th>\n",
       "      <th>ZNF217_cnv</th>\n",
       "      <th>MITF_mut</th>\n",
       "      <th>MSH6_mut</th>\n",
       "      <th>H3F3A_cnv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-102</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-103</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-104</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-37</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-43</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-69</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-84</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>166 rows Ã— 359 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AKT1_mut  ALK_mut  APC_mut  ARAF_mut  ARID1A_mut  \\\n",
       "participant_id_legacy                                                     \n",
       "GBM.ICB-1                   0.0      0.0      0.0       0.0         0.0   \n",
       "GBM.ICB-10                  0.0      0.0      0.0       0.0         0.0   \n",
       "GBM.ICB-102                 0.0      0.0      0.0       0.0         0.0   \n",
       "GBM.ICB-103                 0.0      0.0      2.0       0.0         0.0   \n",
       "GBM.ICB-104                 0.0      0.0      0.0       0.0         0.0   \n",
       "...                         ...      ...      ...       ...         ...   \n",
       "GBM.ICB-37                  0.0      0.0      0.0       0.0         0.0   \n",
       "GBM.ICB-42                  0.0      0.0      0.0       0.0         0.0   \n",
       "GBM.ICB-43                  0.0      0.0      0.0       0.0         0.0   \n",
       "GBM.ICB-69                  0.0      0.0      0.0       0.0         0.0   \n",
       "GBM.ICB-84                  0.0      0.0      0.0       0.0         0.0   \n",
       "\n",
       "                       ARID1B_mut  ARID2_mut  ASXL1_mut  ATM_mut  ATRX_mut  \\\n",
       "participant_id_legacy                                                        \n",
       "GBM.ICB-1                     0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-10                    0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-102                   0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-103                   0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-104                   0.0        0.0        0.0      0.0       0.0   \n",
       "...                           ...        ...        ...      ...       ...   \n",
       "GBM.ICB-37                    0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-42                    0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-43                    0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-69                    0.0        0.0        0.0      0.0       0.0   \n",
       "GBM.ICB-84                    0.0        0.0        0.0      0.0       0.0   \n",
       "\n",
       "                       ...  U2AF1_cnv  VHL_cnv  WRN_cnv  WT1_cnv  XPA_cnv  \\\n",
       "participant_id_legacy  ...                                                  \n",
       "GBM.ICB-1              ...        1.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-10             ...        0.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-102            ...        0.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-103            ...        0.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-104            ...        0.0      0.0      0.0      0.0      0.0   \n",
       "...                    ...        ...      ...      ...      ...      ...   \n",
       "GBM.ICB-37             ...        0.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-42             ...        0.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-43             ...        0.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-69             ...        0.0      0.0      0.0      0.0      0.0   \n",
       "GBM.ICB-84             ...        0.0      0.0      0.0      0.0      0.0   \n",
       "\n",
       "                       XPC_cnv  ZNF217_cnv  MITF_mut  MSH6_mut  H3F3A_cnv  \n",
       "participant_id_legacy                                                      \n",
       "GBM.ICB-1                  0.0         0.0       0.0       0.0        0.0  \n",
       "GBM.ICB-10                 0.0         0.0       0.0       0.0        0.0  \n",
       "GBM.ICB-102                0.0         0.0       0.0       0.0        0.0  \n",
       "GBM.ICB-103                0.0         0.0       0.0       0.0        0.0  \n",
       "GBM.ICB-104                0.0         0.0       0.0       0.0        0.0  \n",
       "...                        ...         ...       ...       ...        ...  \n",
       "GBM.ICB-37                 0.0         0.0       0.0       0.0        2.0  \n",
       "GBM.ICB-42                 0.0         0.0       0.0       0.0        0.0  \n",
       "GBM.ICB-43                 0.0         0.0       0.0       2.0        0.0  \n",
       "GBM.ICB-69                 0.0         0.0       0.0       0.0        0.0  \n",
       "GBM.ICB-84                 0.0         0.0       0.0       0.0        0.0  \n",
       "\n",
       "[166 rows x 359 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_matrix = pd.concat((op_matrix, icb_wes_matrix)).fillna(0)\n",
    "icb_matrix = icb_matrix.fillna(0)\n",
    "icb_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f947d998-d575-42b9-9751-5c365267133a",
   "metadata": {},
   "source": [
    "Keep only alteration types significantly associated with clusters in GLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "19554dd1-5331-435e-bd2f-0b28f9477586",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "53"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_alterations = list(matrix_sig)\n",
    "missing = set(glass_alterations) - set(icb_matrix)\n",
    "len(missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "3a0e9f94-351d-4d14-812b-0ac2f5bf2a67",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(166, 376) (365, 376)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n"
     ]
    }
   ],
   "source": [
    "for c in missing:\n",
    "    icb_matrix[c] = 0.0\n",
    "\n",
    "icb_matrix = icb_matrix[list(matrix_sig)]\n",
    "\n",
    "print(icb_matrix.shape, matrix_sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "c73c84e0-95a3-4d35-8c11-939c937bf136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "166"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_matrix.index.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "6c55572a-290b-4107-a513-9a97072cf82d",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_samples = icb_sample_df[(icb_sample_df[\"pre_post\"] == \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\") & (icb_sample_df[\"usable_oncopanel\"] | icb_sample_df[\"usable_wes\"])][\"participant_id_legacy\"]\n",
    "#icb_sample_df[(icb_sample_df[\"pre_post\"] == \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\") & icb_sample_df[\"usable_wes\"]][\"participant_id_legacy\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "478d9361-2c13-4130-b080-087eacbe0556",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These pts have no variants that were included in clustering\n",
    "no_variant_pts = list(set(icb_samples) - set(icb_matrix.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "e2a43f7d-f4d7-407c-87a7-74ac2024a445",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jf/x908w3hn72zftc125z_ydf640000gn/T/ipykernel_41322/2180289332.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  icb_matrix.loc[pt, :] = 0.0\n"
     ]
    }
   ],
   "source": [
    "for pt in no_variant_pts:\n",
    "    icb_matrix.loc[pt, :] = 0.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "b7a81e43-85c0-4904-9783-655d235ab8a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "167"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_matrix.index.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0da39d67-4654-46bf-b549-ca43a864b683",
   "metadata": {},
   "source": [
    "# Classify the ICB participants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08d29fc6-9055-4af7-b8f5-cea53f0f1e92",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Apply the best classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "d4bbd46d-4153-43c8-9d79-a9164e12b9f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#best_model.fit(X, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "a509105d-b3b4-4e64-904d-1c850ca9006b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster_labels = [f\"cluster_{c}\" for c in best_model.classes_]\n",
    "\n",
    "# icb_pred = best_model.predict_proba(icb_matrix)\n",
    "# icb_pred_df = pd.DataFrame(data=icb_pred, index = icb_matrix.index, columns=cluster_labels)\n",
    "# icb_pred_df[\"cluster\"] = icb_pred_df.apply(lambda row: cluster_labels[np.argmax(row)], axis=1)\n",
    "\n",
    "# icb_pred_df['cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "a5d09203-1758-48d3-b383-0546cd4869e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# icb_results = icb_pred_df.copy()\n",
    "\n",
    "# icb_results = icb_results.merge(icb_pt_df, left_on='participant_id_legacy', right_on=\"participant_id_legacy\")\n",
    "\n",
    "# #icb_results = icb_results[icb_results['MGMT methylated Y/N'].notnull()]\n",
    "\n",
    "# icb_results['setting'] = icb_results['ICB for Newly diagnosed']\n",
    "# icb_results['vital_status'] = icb_results['Deceased'].map({'Deceased': 1, 'Alive': 0})\n",
    "# icb_results['mgmt_methylated'] = icb_results['MGMT methylated Y/N'].replace({'Unmethylated': 0.0, 'Methylated': 1.0, 'Partially': 1.0})\n",
    "\n",
    "# print(icb_results['participant_id'].nunique()) #, icb_results['sample_id_legacy'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c0a3d5-cfaa-439f-b68a-752844941f08",
   "metadata": {},
   "source": [
    "## Output predicted cluster assignments to TSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "1e11757a-03bf-48d7-b395-01a9ac4ff7c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# icb_results.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "c5875261-a890-4431-827f-8bd2f313f1e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# icb_results.to_csv(\"../results/icb_pt_w_cluster_assignments.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2650456-8462-40af-9dbe-035cc5d5b353",
   "metadata": {},
   "source": [
    "(To skip model training, resume here)\n",
    "## Read in predicted cluster assignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "d47ae844-bcb0-41d7-b463-225ae0084251",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_results = pd.read_csv('../../results/icb_pt_w_cluster_assignments.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8574c80-74d9-4722-b809-5bb683c6bb3b",
   "metadata": {},
   "source": [
    "# Assess prognostic value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f077fbdf-b50a-4670-9eab-8b6dd5577536",
   "metadata": {},
   "source": [
    "## With no covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "a4f1e062-3882-4410-b6da-b189a6eb110b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 167  57\n",
      "[1] 38 57\n",
      "[1] 129  57\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_1, data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_1 -1.1238    0.3250   0.8338 -1.348    0.178\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1     0.325      3.077   0.06342     1.666\n",
      "\n",
      "Concordance= 0.534  (se = 0.025 )\n",
      "Likelihood ratio test= 1.9  on 1 df,   p=0.2\n",
      "Wald test            = 1.82  on 1 df,   p=0.2\n",
      "Score (logrank) test = 1.82  on 1 df,   p=0.2\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_2, data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "            coef exp(coef) se(coef)    z Pr(>|z|)\n",
      "cluster_2 0.1685    1.1835   0.3305 0.51     0.61\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2     1.183      0.845    0.6192     2.262\n",
      "\n",
      "Concordance= 0.518  (se = 0.025 )\n",
      "Likelihood ratio test= 0.26  on 1 df,   p=0.6\n",
      "Wald test            = 0.26  on 1 df,   p=0.6\n",
      "Score (logrank) test = 0.26  on 1 df,   p=0.6\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_3, data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_3 0.4551    1.5763   0.3568 1.275    0.202\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3     1.576     0.6344    0.7832     3.172\n",
      "\n",
      "Concordance= 0.514  (se = 0.026 )\n",
      "Likelihood ratio test= 1.59  on 1 df,   p=0.2\n",
      "Wald test            = 1.63  on 1 df,   p=0.2\n",
      "Score (logrank) test = 1.63  on 1 df,   p=0.2\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_4, data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_4 -0.9798    0.3754   1.3508 -0.725    0.468\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4    0.3754      2.664   0.02659       5.3\n",
      "\n",
      "Concordance= 0.51  (se = 0.028 )\n",
      "Likelihood ratio test= 0.54  on 1 df,   p=0.5\n",
      "Wald test            = 0.53  on 1 df,   p=0.5\n",
      "Score (logrank) test = 0.53  on 1 df,   p=0.5\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_5, data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_5 -1.3572    0.2574   0.7621 -1.781   0.0749 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5    0.2574      3.885   0.05779     1.146\n",
      "\n",
      "Concordance= 0.517  (se = 0.028 )\n",
      "Likelihood ratio test= 3.46  on 1 df,   p=0.06\n",
      "Wald test            = 3.17  on 1 df,   p=0.07\n",
      "Score (logrank) test = 3.17  on 1 df,   p=0.07\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gbm_icb_manuscript/lib/python3.13/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"AE\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    }
   ],
   "source": [
    "%%R -i icb_results\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- icb_results %>%\n",
    "    mutate(cluster = as.factor(cluster))\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly-diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "# model_6_combined <- coxph(Surv(osicb, vital_status) ~ setting + mgmt_methylated + cluster,\n",
    "#                           data = combined)\n",
    "# model_6_nd <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = nd)\n",
    "# model_6_rc <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(osicb, vital_status) ~ cluster_1,\n",
    "                      data = combined)\n",
    "univariate_2 <- coxph(Surv(osicb, vital_status) ~ cluster_2,\n",
    "                      data = combined)\n",
    "univariate_3 <- coxph(Surv(osicb, vital_status) ~ cluster_3,\n",
    "                      data = combined)\n",
    "univariate_4 <- coxph(Surv(osicb, vital_status) ~ cluster_4,\n",
    "                      data = combined)\n",
    "univariate_5 <- coxph(Surv(osicb, vital_status) ~ cluster_5,\n",
    "                      data = combined)\n",
    "# univariate_6 <- coxph(Surv(osicb, vital_status) ~ cluster_6,\n",
    "#                       data = combined)\n",
    "\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "#print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de3def5b-8377-44e9-b9f1-5c3251d2067f",
   "metadata": {},
   "source": [
    "## With MGMT methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "ff9ab233-500d-4dee-89b0-ebb082bb9761",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 167  57\n",
      "[1] 38 57\n",
      "[1] 129  57\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_1 + mgmt_methylated, \n",
      "    data = combined)\n",
      "\n",
      "  n= 157, number of events= 145 \n",
      "   (10 observations deleted due to missingness)\n",
      "\n",
      "                   coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_1       -1.0098    0.3643   0.8633 -1.170   0.2421  \n",
      "mgmt_methylated -0.3151    0.7297   0.1719 -1.833   0.0667 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1          0.3643      2.745   0.06709     1.978\n",
      "mgmt_methylated    0.7297      1.370   0.52104     1.022\n",
      "\n",
      "Concordance= "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gbm_icb_manuscript/lib/python3.13/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"AE\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.546  (se = 0.027 )\n",
      "Likelihood ratio test= 5.25  on 2 df,   p=0.07\n",
      "Wald test            = 5.13  on 2 df,   p=0.08\n",
      "Score (logrank) test = 5.16  on 2 df,   p=0.08\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_2 + mgmt_methylated, \n",
      "    data = combined)\n",
      "\n",
      "  n= 157, number of events= 145 \n",
      "   (10 observations deleted due to missingness)\n",
      "\n",
      "                   coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_2        0.1658    1.1804   0.3384  0.490   0.6241  \n",
      "mgmt_methylated -0.3372    0.7137   0.1715 -1.966   0.0492 *\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2          1.1804     0.8472     0.608    2.2915\n",
      "mgmt_methylated    0.7137     1.4011     0.510    0.9989\n",
      "\n",
      "Concordance= 0.544  (se = 0.027 )\n",
      "Likelihood ratio test= 4.07  on 2 df,   p=0.1\n",
      "Wald test            = 4.01  on 2 df,   p=0.1\n",
      "Score (logrank) test = 4.04  on 2 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_3 + mgmt_methylated, \n",
      "    data = combined)\n",
      "\n",
      "  n= 157, number of events= 145 \n",
      "   (10 observations deleted due to missingness)\n",
      "\n",
      "                   coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_3        0.4163    1.5163   0.3690  1.128   0.2592  \n",
      "mgmt_methylated -0.3004    0.7405   0.1740 -1.727   0.0842 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3          1.5163     0.6595    0.7358     3.125\n",
      "mgmt_methylated    0.7405     1.3504    0.5266     1.041\n",
      "\n",
      "Concordance= 0.539  (se = 0.027 )\n",
      "Likelihood ratio test= 5.08  on 2 df,   p=0.08\n",
      "Wald test            = 5.1  on 2 df,   p=0.08\n",
      "Score (logrank) test = 5.14  on 2 df,   p=0.08\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_4 + mgmt_methylated, \n",
      "    data = combined)\n",
      "\n",
      "  n= 157, number of events= 145 \n",
      "   (10 observations deleted due to missingness)\n",
      "\n",
      "                   coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_4       -0.2741    0.7602   1.3708 -0.200   0.8415  \n",
      "mgmt_methylated -0.3248    0.7227   0.1752 -1.854   0.0638 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4          0.7602      1.315   0.05178    11.162\n",
      "mgmt_methylated    0.7227      1.384   0.51261     1.019\n",
      "\n",
      "Concordance= 0.532  (se = 0.028 )\n",
      "Likelihood ratio test= 3.87  on 2 df,   p=0.1\n",
      "Wald test            = 3.81  on 2 df,   p=0.1\n",
      "Score (logrank) test = 3.84  on 2 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_5 + mgmt_methylated, \n",
      "    data = combined)\n",
      "\n",
      "  n= 157, number of events= 145 \n",
      "   (10 observations deleted due to missingness)\n",
      "\n",
      "                   coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_5       -1.4052    0.2453   0.7630 -1.842   0.0655 .\n",
      "mgmt_methylated -0.3291    0.7196   0.1715 -1.919   0.0550 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5          0.2453      4.076   0.05499     1.094\n",
      "mgmt_methylated    0.7196      1.390   0.51412     1.007\n",
      "\n",
      "Concordance= 0.546  (se = 0.028 )\n",
      "Likelihood ratio test= 7.58  on 2 df,   p=0.02\n",
      "Wald test            = 7.31  on 2 df,   p=0.03\n",
      "Score (logrank) test = 7.38  on 2 df,   p=0.02\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i icb_results\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- icb_results %>%\n",
    "    mutate(cluster = as.factor(cluster))\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly-diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "# model_6_combined <- coxph(Surv(osicb, vital_status) ~ setting + mgmt_methylated + cluster,\n",
    "#                           data = combined)\n",
    "# model_6_nd <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = nd)\n",
    "# model_6_rc <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(osicb, vital_status) ~ cluster_1 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_2 <- coxph(Surv(osicb, vital_status) ~ cluster_2 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_3 <- coxph(Surv(osicb, vital_status) ~ cluster_3 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_4 <- coxph(Surv(osicb, vital_status) ~ cluster_4 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "univariate_5 <- coxph(Surv(osicb, vital_status) ~ cluster_5 + mgmt_methylated,\n",
    "                      data = combined)\n",
    "# univariate_6 <- coxph(Surv(osicb, vital_status) ~ cluster_6 + mgmt_methylated,\n",
    "#                       data = combined)\n",
    "\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "#print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77e71e7f-9a53-4129-ba62-503336df7775",
   "metadata": {},
   "source": [
    "## With Newly-diagnosed/recurrent as a covariate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "9b867e7d-f78e-41f1-921f-f0272df2106a",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 167  57"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gbm_icb_manuscript/lib/python3.13/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"AE\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[1] 38 57\n",
      "[1] 129  57\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_1 + setting, \n",
      "    data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_1        -0.8118    0.4440   0.8297 -0.978 0.327858    \n",
      "settingRecurrent  0.6655    1.9454   0.1997  3.332 0.000861 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1            0.444      2.252   0.08733     2.258\n",
      "settingRecurrent     1.945      0.514   1.31532     2.877\n",
      "\n",
      "Concordance= 0.595  (se = 0.026 )\n",
      "Likelihood ratio test= 14.28  on 2 df,   p=8e-04\n",
      "Wald test            = 12.82  on 2 df,   p=0.002\n",
      "Score (logrank) test = 13.26  on 2 df,   p=0.001\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_2 + setting, \n",
      "    data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "                   coef exp(coef) se(coef)     z Pr(>|z|)    \n",
      "cluster_2        0.1286    1.1372   0.3306 0.389   0.6973    \n",
      "settingRecurrent 0.6826    1.9790   0.1989 3.432   0.0006 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2            1.137     0.8794    0.5949     2.174\n",
      "settingRecurrent     1.979     0.5053    1.3401     2.923\n",
      "\n",
      "Concordance= 0.59  (se = 0.026 )\n",
      "Likelihood ratio test= 13.44  on 2 df,   p=0.001\n",
      "Wald test            = 12.01  on 2 df,   p=0.002\n",
      "Score (logrank) test = 12.45  on 2 df,   p=0.002\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_3 + setting, \n",
      "    data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "                   coef exp(coef) se(coef)     z Pr(>|z|)    \n",
      "cluster_3        0.4124    1.5104   0.3523 1.170  0.24183    \n",
      "settingRecurrent 0.6789    1.9718   0.1989 3.414  0.00064 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3            1.510     0.6621    0.7572     3.013\n",
      "settingRecurrent     1.972     0.5072    1.3353     2.912\n",
      "\n",
      "Concordance= 0.587  (se = 0.026 )\n",
      "Likelihood ratio test= 14.63  on 2 df,   p=7e-04\n",
      "Wald test            = 13.27  on 2 df,   p=0.001\n",
      "Score (logrank) test = 13.72  on 2 df,   p=0.001\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_4 + setting, \n",
      "    data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_4        -0.5746    0.5629   1.3643 -0.421 0.673641    \n",
      "settingRecurrent  0.6780    1.9699   0.1994  3.401 0.000673 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4           0.5629     1.7764   0.03883     8.161\n",
      "settingRecurrent    1.9699     0.5076   1.33272     2.912\n",
      "\n",
      "Concordance= 0.583  (se = 0.027 )\n",
      "Likelihood ratio test= 13.47  on 2 df,   p=0.001\n",
      "Wald test            = 12.04  on 2 df,   p=0.002\n",
      "Score (logrank) test = 12.48  on 2 df,   p=0.002\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ cluster_5 + setting, \n",
      "    data = combined)\n",
      "\n",
      "  n= 167, number of events= 154 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "cluster_5        -1.4450    0.2357   0.7684 -1.880 0.060045 .  \n",
      "settingRecurrent  0.6966    2.0068   0.1992  3.496 0.000472 ***\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5           0.2357     4.2419   0.05228     1.063\n",
      "settingRecurrent    2.0068     0.4983   1.35806     2.966\n",
      "\n",
      "Concordance= 0.586  (se = 0.027 )\n",
      "Likelihood ratio test= 17.16  on 2 df,   p=2e-04\n",
      "Wald test            = 15.37  on 2 df,   p=5e-04\n",
      "Score (logrank) test = 15.77  on 2 df,   p=4e-04\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i icb_results\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- icb_results %>%\n",
    "    mutate(cluster = as.factor(cluster))\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly-diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "# model_6_combined <- coxph(Surv(osicb, vital_status) ~ setting + mgmt_methylated + cluster,\n",
    "#                           data = combined)\n",
    "# model_6_nd <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = nd)\n",
    "# model_6_rc <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "#                           data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(osicb, vital_status) ~ cluster_1 + setting,\n",
    "                      data = combined)\n",
    "univariate_2 <- coxph(Surv(osicb, vital_status) ~ cluster_2 + setting,\n",
    "                      data = combined)\n",
    "univariate_3 <- coxph(Surv(osicb, vital_status) ~ cluster_3 + setting,\n",
    "                      data = combined)\n",
    "univariate_4 <- coxph(Surv(osicb, vital_status) ~ cluster_4 + setting,\n",
    "                      data = combined)\n",
    "univariate_5 <- coxph(Surv(osicb, vital_status) ~ cluster_5 + setting,\n",
    "                      data = combined)\n",
    "#univariate_6 <- coxph(Surv(osicb, vital_status) ~ cluster_6 + setting,\n",
    "#                      data = combined)\n",
    "\n",
    "\n",
    "# print(summary(model_6_combined))\n",
    "# print(summary(model_6_nd))\n",
    "# print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "#print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a1e240e-3b0f-4f23-a1bd-8f1b446156fc",
   "metadata": {},
   "source": [
    "# Create a heatmap of ICB cohort, grouped by predicted cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "509b1b70-e24a-41ba-a6a5-03989816c074",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(167, 376)"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "0869cc10-268f-4e42-b3c5-88734a497bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_results[\"k = 5\"] = icb_results[\"cluster\"].map(lambda x: x.split(\"_\")[-1])\n",
    "icb_results.set_index('participant_id_legacy', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "88b27ac7-07c9-4304-a3e0-d25e8758af46",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_matrix = icb_matrix.join(icb_results[\"k = 5\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "1f30e5a0-67fa-4dd9-a293-c20b1532093c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    }
   ],
   "source": [
    "%%R -i icb_matrix -i cnv_chrom_df -w 3000 -h 1200\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(RColorBrewer)\n",
    "\n",
    "#icb_matrix <- icb_matrix[order(icb_matrix$`k = 6`),]\n",
    "icb_matrix <- icb_matrix[order(icb_matrix$`k = 5`),]\n",
    "\n",
    "muts <- as.matrix(icb_matrix[, grep('_mut', names(icb_matrix))])\n",
    "cnvs <- as.matrix(icb_matrix[, grep('_cnv', names(icb_matrix))])\n",
    "\n",
    "clusters <- icb_matrix[, grep('=', names(icb_matrix))]\n",
    "cluster_colors <- brewer.pal(n = 8, name = 'Dark2')\n",
    "\n",
    "chrom_colors <- rep(c(cluster_colors[6], cluster_colors[8]), length.out = 23)\n",
    "chrom_colors <- setNames(chrom_colors, 1:23)\n",
    "\n",
    "row_ha <- rowAnnotation(#Cluster_2 = clusters$`k = 2`,\n",
    "                        #Cluster_3 = clusters$`k = 3`,\n",
    "                        #Cluster_4 = clusters$`k = 4`,\n",
    "                        Cluster_5 = clusters, #clusters$`k = 5`,\n",
    "                        #Cluster_6 = clusters, #clusters$`k = 5`,\n",
    "                        col = list(#Cluster_2 = c(`1` = cluster_colors[1], `2` = cluster_colors[2]),\n",
    "                                   #Cluster_3 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3]),\n",
    "                                   #Cluster_4 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4]),\n",
    "                                   Cluster_5 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5])),\n",
    "                                   #Cluster_6 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5], `6` = cluster_colors[6])),\n",
    "                        simple_anno_size = unit(2, 'cm'))\n",
    "\n",
    "col_ha <- columnAnnotation(df = cnv_chrom_df,\n",
    "                          simple_anno_size = unit(2, 'cm'),\n",
    "                          col = list(chrom = chrom_colors))\n",
    "\n",
    "heatmap_muts <- Heatmap(muts, name = 'SNV/indels',\n",
    "                        col = colorRamp2(c(0, 2, 4), c('white', 'grey', 'black')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "heatmap_cnvs <- Heatmap(cnvs, name = 'CNVs', right_annotation = row_ha, top_annotation = col_ha,\n",
    "                        col = colorRamp2(c(-2, 0, 2), c('blue', 'white', 'red')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "full_heatmap <- heatmap_muts + heatmap_cnvs\n",
    "pdf(\"../../results/icb_heatmap.pdf\", width=50, height=10)\n",
    "draw(full_heatmap)\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
