{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "327e58cd-d270-4af2-b3be-4199b82e5a5e",
   "metadata": {},
   "source": [
    "# DNA Clustering and Classification\n",
    "\n",
    "This notebook explores whether we can use mutational and copy number information to identify clusters of GLASS samples that overlap with Verhaak transcriptional subtype.\n",
    "\n",
    "The actual consensus clustering is done by the NMFConsensus script from GenePattern here: https://github.com/genepattern/NMFConsensus/blob/develop/nmfconsensus.R, and the clustering parameters are as listed here: gs://fc-secure-3f0f305e-f110-432b-bc9a-c8d96824cea2/jweiss/local_files/results/2024-07-22_GLASS-Verhaak-classifier/GLASS_sample_clustering.params.txt\n",
    "\n",
    "The notebook also trains a classifier to predict membership in these DNA-based clusters and then applies this classifier to the ICB OncoPanel data.\n",
    "\n",
    "This notebook generates the following figures:\n",
    "* The clustering heatmap shown in Fig. 1c\n",
    "* The sankey plot shown in Extended Data Fig. 3d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1ad4df9b-047d-408a-acbf-7fddc296f9e5",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'sklearn'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 7\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mscipy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mstats\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mstats\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mstatsmodels\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mstats\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmultitest\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m multipletests\n\u001b[0;32m----> 7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msklearn\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m metrics\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'sklearn'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn import metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9527aa9a-803b-4a97-91a1-1b63e67486e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b03d54-e19d-47d6-ae00-a6b78cf90767",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dir = '../results'\n",
    "if not os.path.exists(results_dir):\n",
    "    os.mkdir(results_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e690a97a-9ae1-4557-bbb3-1581b6574563",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load and preprocess GLASS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e68c7b-6a7d-424d-b669-e3b902ea8d4f",
   "metadata": {},
   "source": [
    "## Clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "82324a6c-798f-4b2a-bad9-4fcd54acf3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_surgeries_df = pd.read_csv('../data/glass_clinical_surgeries_table.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ab2684b5-4846-46ef-9b80-8f3be5be8207",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/2887213321.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  glass_surgeries_df[(glass_surgeries_df[\"sample_barcode\"].notnull()) & (glass_surgeries_df[\"idh_status\"] != \"IDHmut\") & (glass_surgeries_df[\"histology\"].str.match(\"Glioblastoma|Gliosarcoma\").fillna(False).astype(bool))]#[\"case_barcode\"].nunique()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>surgery_number</th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>surgical_interval_mo</th>\n",
       "      <th>histology</th>\n",
       "      <th>grade</th>\n",
       "      <th>idh_status</th>\n",
       "      <th>codel_status</th>\n",
       "      <th>who_classification</th>\n",
       "      <th>mgmt_methylation</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment_radiotherapy</th>\n",
       "      <th>treatment_radiation_dose_gy</th>\n",
       "      <th>treatment_radiation_fractions</th>\n",
       "      <th>treatment_radiation_other</th>\n",
       "      <th>treatment_chemotherapy_other</th>\n",
       "      <th>treatment_chemotherapy_other_cycles</th>\n",
       "      <th>comments</th>\n",
       "      <th>idh_codel_subtype</th>\n",
       "      <th>treatment_alkylating_agent</th>\n",
       "      <th>mgmt_methylation_method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GLSS-LU-00C4</td>\n",
       "      <td>1</td>\n",
       "      <td>GLSS-LU-00C4-TP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>noncodel</td>\n",
       "      <td>Glioblastoma, IDH-wildtype</td>\n",
       "      <td>M</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>1</td>\n",
       "      <td>array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLSS-LU-00C7</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-LU-00C7-R1</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>noncodel</td>\n",
       "      <td>Glioblastoma, IDH-wildtype</td>\n",
       "      <td>U</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GLSS-LU-00C7</td>\n",
       "      <td>1</td>\n",
       "      <td>GLSS-LU-00C7-TP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>noncodel</td>\n",
       "      <td>Glioblastoma, IDH-wildtype</td>\n",
       "      <td>U</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>1</td>\n",
       "      <td>array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GLSS-LU-0B13</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-LU-0B13-R1</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>noncodel</td>\n",
       "      <td>Glioblastoma, IDH-wildtype</td>\n",
       "      <td>M</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GLSS-LU-0B13</td>\n",
       "      <td>1</td>\n",
       "      <td>GLSS-LU-0B13-TP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>noncodel</td>\n",
       "      <td>Glioblastoma, IDH-wildtype</td>\n",
       "      <td>M</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>1</td>\n",
       "      <td>array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>GLSS-HF-887A</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-HF-887A-R1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>32.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CPT-11; Bevacizumab/Avastin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tumor was highly necrotic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>977</th>\n",
       "      <td>GLSS-HF-891E</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-HF-891E-R1</td>\n",
       "      <td>39.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Procarbazine Hydrochloride; Vincristine Sulfat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978</th>\n",
       "      <td>GLSS-HF-891E</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-HF-891E-R2</td>\n",
       "      <td>41.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carmustine; Tamoxifen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carmustine as Gliadel wafer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>GLSS-HF-D48E</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-HF-D48E-R1</td>\n",
       "      <td>36.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>noncodel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U</td>\n",
       "      <td>...</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bevacizumab/Avastin; CPT-11; Carboplatin; Etop...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>PCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985</th>\n",
       "      <td>GLSS-HF-E362</td>\n",
       "      <td>1</td>\n",
       "      <td>GLSS-HF-E362-TP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>478 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     case_barcode  surgery_number   sample_barcode  surgical_interval_mo  \\\n",
       "1    GLSS-LU-00C4               1  GLSS-LU-00C4-TP                   0.0   \n",
       "2    GLSS-LU-00C7               2  GLSS-LU-00C7-R1                  17.0   \n",
       "3    GLSS-LU-00C7               1  GLSS-LU-00C7-TP                   0.0   \n",
       "4    GLSS-LU-0B13               2  GLSS-LU-0B13-R1                  14.0   \n",
       "5    GLSS-LU-0B13               1  GLSS-LU-0B13-TP                   0.0   \n",
       "..            ...             ...              ...                   ...   \n",
       "975  GLSS-HF-887A               3  GLSS-HF-887A-R1                   8.0   \n",
       "977  GLSS-HF-891E               2  GLSS-HF-891E-R1                  39.0   \n",
       "978  GLSS-HF-891E               3  GLSS-HF-891E-R2                  41.0   \n",
       "984  GLSS-HF-D48E               2  GLSS-HF-D48E-R1                  36.0   \n",
       "985  GLSS-HF-E362               1  GLSS-HF-E362-TP                   0.0   \n",
       "\n",
       "        histology grade idh_status codel_status          who_classification  \\\n",
       "1    Glioblastoma    IV      IDHwt     noncodel  Glioblastoma, IDH-wildtype   \n",
       "2    Glioblastoma    IV      IDHwt     noncodel  Glioblastoma, IDH-wildtype   \n",
       "3    Glioblastoma    IV      IDHwt     noncodel  Glioblastoma, IDH-wildtype   \n",
       "4    Glioblastoma    IV      IDHwt     noncodel  Glioblastoma, IDH-wildtype   \n",
       "5    Glioblastoma    IV      IDHwt     noncodel  Glioblastoma, IDH-wildtype   \n",
       "..            ...   ...        ...          ...                         ...   \n",
       "975  Glioblastoma    IV        NaN          NaN                         NaN   \n",
       "977  Glioblastoma    IV        NaN          NaN                         NaN   \n",
       "978  Glioblastoma    IV        NaN          NaN                         NaN   \n",
       "984  Glioblastoma    IV        NaN     noncodel                         NaN   \n",
       "985  Glioblastoma    IV        NaN          NaN                         NaN   \n",
       "\n",
       "    mgmt_methylation  ... treatment_radiotherapy treatment_radiation_dose_gy  \\\n",
       "1                  M  ...                      1                         NaN   \n",
       "2                  U  ...                      0                         NaN   \n",
       "3                  U  ...                      1                         NaN   \n",
       "4                  M  ...                      0                         NaN   \n",
       "5                  M  ...                      1                         NaN   \n",
       "..               ...  ...                    ...                         ...   \n",
       "975              NaN  ...                   TRUE                        32.0   \n",
       "977              NaN  ...                  FALSE                         NaN   \n",
       "978              NaN  ...                  FALSE                         NaN   \n",
       "984                U  ...                  FALSE                         NaN   \n",
       "985              NaN  ...                   TRUE                         NaN   \n",
       "\n",
       "    treatment_radiation_fractions treatment_radiation_other  \\\n",
       "1                             NaN                       NaN   \n",
       "2                             NaN                       NaN   \n",
       "3                             NaN                       NaN   \n",
       "4                             NaN                       NaN   \n",
       "5                             NaN                       NaN   \n",
       "..                            ...                       ...   \n",
       "975                           NaN                       NaN   \n",
       "977                           NaN                       NaN   \n",
       "978                           NaN                       NaN   \n",
       "984                           NaN                       NaN   \n",
       "985                           NaN                       NaN   \n",
       "\n",
       "                          treatment_chemotherapy_other  \\\n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "..                                                 ...   \n",
       "975                        CPT-11; Bevacizumab/Avastin   \n",
       "977  Procarbazine Hydrochloride; Vincristine Sulfat...   \n",
       "978                              Carmustine; Tamoxifen   \n",
       "984  Bevacizumab/Avastin; CPT-11; Carboplatin; Etop...   \n",
       "985                                                NaN   \n",
       "\n",
       "    treatment_chemotherapy_other_cycles                     comments  \\\n",
       "1                                   NaN                          NaN   \n",
       "2                                   NaN                          NaN   \n",
       "3                                   NaN                          NaN   \n",
       "4                                   NaN                          NaN   \n",
       "5                                   NaN                          NaN   \n",
       "..                                  ...                          ...   \n",
       "975                                 NaN    Tumor was highly necrotic   \n",
       "977                                 NaN                          NaN   \n",
       "978                                 NaN  Carmustine as Gliadel wafer   \n",
       "984                                 NaN                          NaN   \n",
       "985                                 NaN                          NaN   \n",
       "\n",
       "    idh_codel_subtype treatment_alkylating_agent mgmt_methylation_method  \n",
       "1               IDHwt                          1                   array  \n",
       "2               IDHwt                        NaN                   array  \n",
       "3               IDHwt                          1                   array  \n",
       "4               IDHwt                        NaN                   array  \n",
       "5               IDHwt                          1                   array  \n",
       "..                ...                        ...                     ...  \n",
       "975               NaN                      FALSE                     NaN  \n",
       "977               NaN                       TRUE                     NaN  \n",
       "978               NaN                       TRUE                     NaN  \n",
       "984               NaN                       TRUE                     PCR  \n",
       "985               NaN                        NaN                     NaN  \n",
       "\n",
       "[478 rows x 30 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_surgeries_df[(glass_surgeries_df[\"sample_barcode\"].notnull()) & (glass_surgeries_df[\"idh_status\"] != \"IDHmut\") & (glass_surgeries_df[\"histology\"].str.match(\"Glioblastoma|Gliosarcoma\").fillna(False).astype(bool))]#[\"case_barcode\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "68180e00-ef4c-45c0-be30-fbc491838c0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(828, 29)\n"
     ]
    }
   ],
   "source": [
    "glass_surgeries_df = glass_surgeries_df[~glass_surgeries_df['sample_barcode'].isnull()]\n",
    "glass_samples_df = glass_surgeries_df.set_index('sample_barcode')\n",
    "print(glass_samples_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89dd99c3-8e6e-423e-9f60-25074de69c76",
   "metadata": {},
   "source": [
    "### Discard samples with non-GBM histology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b2f3cfc2-9bd2-4e3a-8bdd-441a888ee895",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(557, 29)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3198457000.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  glass_samples_df = glass_samples_df.loc[glass_samples_df['histology'].str.match(\"Glioblastoma|Gliosarcoma\").fillna(False).astype(bool), :]\n"
     ]
    }
   ],
   "source": [
    "glass_samples_df = glass_samples_df.loc[glass_samples_df['histology'].str.match(\"Glioblastoma|Gliosarcoma\").fillna(False).astype(bool), :]\n",
    "print(glass_samples_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d697789c-9370-4493-bcf2-a1b48910d1c0",
   "metadata": {},
   "source": [
    "### Discard samples with IDH-mut status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "15efa798-a929-4629-a757-637003b14ad5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(478, 29)\n"
     ]
    }
   ],
   "source": [
    "glass_samples_df = glass_samples_df.loc[glass_samples_df['idh_status'] != 'IDHmut']\n",
    "print(glass_samples_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bc56e85-de94-481a-8728-29c9b8872aa7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## WES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaec0811-e0f2-4d4d-9384-6d89c83f7aad",
   "metadata": {},
   "source": [
    "### Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f04862e1-7178-4640-82cd-da64c18b59b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_variants_passgeno_df = pd.read_csv('../../../data/glass/variants_passgeno.csv', sep = ',', low_memory = False)\n",
    "glass_variants_anno_df = pd.read_csv('../../../data/glass/variants_anno.csv', sep = ',', low_memory = False)\n",
    "oncopanel_genes = pd.read_csv(os.path.join('../data/oncopanel_exonic_v123.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6e5ceaec-7a11-4f0a-84b6-8ff79f50053c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create and save \"sample\" column\n",
    "glass_variants_passgeno_df[\"sample\"] = glass_variants_passgeno_df[\"aliquot_barcode\"].str[:15]\n",
    "wes_mut_samples = set(glass_variants_passgeno_df[\"sample\"])\n",
    "\n",
    "# Keep only the variants detected by single-sample MuTect2\n",
    "glass_variants_passgeno_df = glass_variants_passgeno_df[glass_variants_passgeno_df[\"ssm2_pass_call\"] == \"t\"]\n",
    "\n",
    "# Keep only variants above the VAF threshold (0.1)\n",
    "VAF_THRESHOLD = 0.1\n",
    "glass_variants_passgeno_df = glass_variants_passgeno_df[glass_variants_passgeno_df['af'] >= VAF_THRESHOLD]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b6cfe8d-3f95-4f18-b310-cd2a764c8977",
   "metadata": {},
   "source": [
    "Join the variant annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3d8aa671-2b32-4491-85df-482e8d00bbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def append_variant_id(df, colnames=[\"chrom\",\"start\",\"end\",\"alt\"]):\n",
    "    df[\"variant_id\"] = df[colnames[0]].astype(str)\n",
    "    for c in colnames[1:]:\n",
    "        df[\"variant_id\"] = df[\"variant_id\"] + \":\" + df[c].astype(str)\n",
    "\n",
    "def chrom_float_to_int(chrom_float):\n",
    "    if np.isnan(chrom_float):\n",
    "        return -1\n",
    "    else:\n",
    "        return int(chrom_float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "85dc6542-ec07-4c3d-a956-4085ae9cb74a",
   "metadata": {},
   "outputs": [],
   "source": [
    "append_variant_id(glass_variants_passgeno_df)\n",
    "append_variant_id(glass_variants_anno_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d0fb47b8-e924-4f11-9576-6b90699861e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_variants_anno_df.set_index(\"variant_id\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c9aba2a5-8ea1-4eca-ab21-00665149208f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3014019, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:168513274:168513274:G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168513274.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>XCL2</td>\n",
       "      <td>FIVE_PRIME_FLANK</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:168545864:168545864:G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168545864.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>XCL1</td>\n",
       "      <td>FIVE_PRIME_UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:197398786:197398786:A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>197398786.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CRB1</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:211652463:211652463:C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>211652463.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>RD3</td>\n",
       "      <td>MISSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.T168S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>1:229771597:229771597:T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>229771597.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>URB2</td>\n",
       "      <td>NONSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.K413*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample               variant_id  chrom        start ref alt  \\\n",
       "583  GLSS-19-0266-R1  1:168513274:168513274:G    1.0  168513274.0   A   G   \n",
       "584  GLSS-19-0266-R1  1:168545864:168545864:G    1.0  168545864.0   A   G   \n",
       "593  GLSS-19-0266-R1  1:197398786:197398786:A    1.0  197398786.0   G   A   \n",
       "598  GLSS-19-0266-R1  1:211652463:211652463:C    1.0  211652463.0   G   C   \n",
       "601  GLSS-19-0266-R1  1:229771597:229771597:T    1.0  229771597.0   A   T   \n",
       "\n",
       "    gene_symbol variant_classification variant_type protein_change  \n",
       "583        XCL2       FIVE_PRIME_FLANK          SNP            NaN  \n",
       "584        XCL1         FIVE_PRIME_UTR          SNP            NaN  \n",
       "593        CRB1                 INTRON          SNP            NaN  \n",
       "598         RD3               MISSENSE          SNP        p.T168S  \n",
       "601        URB2               NONSENSE          SNP        p.K413*  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_passgeno_cols = [\"sample\", \"variant_id\"] \n",
    "kept_anno_cols = [\"chrom\", \"start\", \"ref\", \"alt\", \"gene_symbol\", \"variant_classification\", \"variant_type\", \"protein_change\"]\n",
    "\n",
    "glass_variants_df = glass_variants_passgeno_df[kept_passgeno_cols].join(glass_variants_anno_df[kept_anno_cols], on=\"variant_id\", how=\"left\")\n",
    "print(glass_variants_df.shape)\n",
    "glass_variants_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04fe63b0-45c8-4ee8-8070-daf63fd2812a",
   "metadata": {},
   "source": [
    "Reorder/filter the columns to a more standard format.\n",
    "\n",
    "Keep only genes with exons covered by all versions (v1, v2, v3) of OncoPanel.\n",
    "\n",
    "Keep only variants on valid chromosomes (1-X).\n",
    "\n",
    "Make sure variant positions are stored as integers.\n",
    "\n",
    "Remove duplicate rows (not sure why these exist)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "556aab90-65e1-4d75-8eee-3c12f880ad57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29772, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>3</td>\n",
       "      <td>70005722</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MITF</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>13</td>\n",
       "      <td>28913225</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>FLT1</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>GLSS-19-0266-R1</td>\n",
       "      <td>X</td>\n",
       "      <td>123220608</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>STAG2</td>\n",
       "      <td>NONSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.Q1089*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1267</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>69990330</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>MITF</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1422</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>7</td>\n",
       "      <td>140481402</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>MISSENSE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>p.G469A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample chrom      start ref alt gene_symbol  \\\n",
       "687   GLSS-19-0266-R1     3   70005722   G   A        MITF   \n",
       "970   GLSS-19-0266-R1    13   28913225   T   G        FLT1   \n",
       "1160  GLSS-19-0266-R1     X  123220608   C   T       STAG2   \n",
       "1267  GLSS-19-0266-TP     3   69990330   G   T        MITF   \n",
       "1422  GLSS-19-0266-TP     7  140481402   C   G        BRAF   \n",
       "\n",
       "     variant_classification variant_type protein_change  \n",
       "687                  INTRON          SNP            NaN  \n",
       "970                  INTRON          SNP            NaN  \n",
       "1160               NONSENSE          SNP       p.Q1089*  \n",
       "1267                 INTRON          SNP            NaN  \n",
       "1422               MISSENSE          SNP        p.G469A  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_glass_cols = [\"sample\", \"chrom\", \"start\", \"ref\", \"alt\", \"gene_symbol\", \"variant_classification\", \"variant_type\", \"protein_change\"]\n",
    "glass_variants_df = glass_variants_df.loc[:, kept_glass_cols]\n",
    "\n",
    "glass_variants_df = glass_variants_df[glass_variants_df['gene_symbol'].isin(oncopanel_genes['Gene'])]\n",
    "\n",
    "glass_variants_df[\"chrom\"] = glass_variants_df[\"chrom\"].map(chrom_float_to_int).astype(int)\n",
    "glass_variants_df = glass_variants_df[glass_variants_df[\"chrom\"] > 0] # Missing chromosome labels were mapped to -1\n",
    "glass_variants_df[\"chrom\"] = np.where(glass_variants_df[\"chrom\"] == 23, \"X\", glass_variants_df[\"chrom\"])\n",
    "glass_variants_df[\"start\"] = glass_variants_df[\"start\"].astype(int)\n",
    "\n",
    "glass_variants_df = glass_variants_df.drop_duplicates()\n",
    "print(glass_variants_df.shape)\n",
    "glass_variants_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4176a6db-26ad-46a0-9334-608a48c5fac2",
   "metadata": {},
   "source": [
    "Keep only non-synonymous variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0c43605d-182c-4644-9148-8022bd5bf297",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_classification\n",
       "INTRON                     23579\n",
       "MISSENSE                    2521\n",
       "FIVE_PRIME_FLANK            1102\n",
       "SILENT                       799\n",
       "THREE_PRIME_UTR              792\n",
       "SPLICE_SITE                  253\n",
       "FRAME_SHIFT_DEL              230\n",
       "NONSENSE                     213\n",
       "FIVE_PRIME_UTR               180\n",
       "FRAME_SHIFT_INS               49\n",
       "IN_FRAME_DEL                  40\n",
       "IN_FRAME_INS                   4\n",
       "START_CODON_SNP                3\n",
       "DE_NOVO_START_IN_FRAME         3\n",
       "DE_NOVO_START_OUT_FRAME        3\n",
       "NONSTOP                        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_variants_df['variant_classification'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "14c9d1b5-8366-446c-90d4-e20d41c58c93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_classification\n",
       "MISSENSE                   2521\n",
       "SPLICE_SITE                 253\n",
       "FRAME_SHIFT_DEL             230\n",
       "NONSENSE                    213\n",
       "FRAME_SHIFT_INS              49\n",
       "IN_FRAME_DEL                 40\n",
       "IN_FRAME_INS                  4\n",
       "START_CODON_SNP               3\n",
       "DE_NOVO_START_IN_FRAME        3\n",
       "DE_NOVO_START_OUT_FRAME       3\n",
       "NONSTOP                       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonsyn_classifications = ['MISSENSE',\n",
    "                          'SPLICE_SITE',\n",
    "                          'FRAME_SHIFT_DEL',\n",
    "                          'NONSENSE',\n",
    "                          'FRAME_SHIFT_INS',\n",
    "                          'IN_FRAME_DEL',\n",
    "                          'IN_FRAME_INS',\n",
    "                          'START_CODON_SNP',\n",
    "                          'DE_NOVO_START_IN_FRAME',\n",
    "                          'DE_NOVO_START_OUT_FRAME',\n",
    "                          'NONSTOP']\n",
    "\n",
    "glass_variants_df = glass_variants_df[glass_variants_df['variant_classification'].isin(nonsyn_classifications)]\n",
    "glass_variants_df['variant_classification'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cb43ab0-c6e3-4583-8067-2e8be6e2690f",
   "metadata": {},
   "source": [
    "### CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7841966b-e6a6-430d-aaa4-8915c1d6962e",
   "metadata": {},
   "outputs": [],
   "source": [
    "glass_cnvs_df = pd.read_csv('../../../data/glass/cnvs.csv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bde655e2-960d-4910-af81-3e2e55719666",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35563, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliquot_barcode</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>hlvl_call</th>\n",
       "      <th>wcr</th>\n",
       "      <th>sample</th>\n",
       "      <th>cnv</th>\n",
       "      <th>hlvl_call_abs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2228044</th>\n",
       "      <td>GLSS-AT-GP01-TP-02D-WXS-DPVLWZ</td>\n",
       "      <td>PDCD1LG2</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.807710</td>\n",
       "      <td>GLSS-AT-GP01-TP</td>\n",
       "      <td>del</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156480</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.141363</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156491</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>EP300</td>\n",
       "      <td>1</td>\n",
       "      <td>1.100724</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156625</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>MCL1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.145647</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29156707</th>\n",
       "      <td>GLSS-MD-0044-R1-01D-WGS-VG0QP6</td>\n",
       "      <td>CREBBP</td>\n",
       "      <td>1</td>\n",
       "      <td>1.109192</td>\n",
       "      <td>GLSS-MD-0044-R1</td>\n",
       "      <td>amp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         aliquot_barcode gene_symbol  hlvl_call       wcr  \\\n",
       "2228044   GLSS-AT-GP01-TP-02D-WXS-DPVLWZ    PDCD1LG2         -1  0.807710   \n",
       "29156480  GLSS-MD-0044-R1-01D-WGS-VG0QP6        ABL1          1  1.141363   \n",
       "29156491  GLSS-MD-0044-R1-01D-WGS-VG0QP6       EP300          1  1.100724   \n",
       "29156625  GLSS-MD-0044-R1-01D-WGS-VG0QP6        MCL1          1  1.145647   \n",
       "29156707  GLSS-MD-0044-R1-01D-WGS-VG0QP6      CREBBP          1  1.109192   \n",
       "\n",
       "                   sample  cnv  hlvl_call_abs  \n",
       "2228044   GLSS-AT-GP01-TP  del              1  \n",
       "29156480  GLSS-MD-0044-R1  amp              1  \n",
       "29156491  GLSS-MD-0044-R1  amp              1  \n",
       "29156625  GLSS-MD-0044-R1  amp              1  \n",
       "29156707  GLSS-MD-0044-R1  amp              1  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create and save \"sample\" column\n",
    "glass_cnvs_df[\"sample\"] = glass_cnvs_df[\"aliquot_barcode\"].str[:15]\n",
    "wes_cnv_samples = set(glass_cnvs_df[\"sample\"])\n",
    "\n",
    "# Keep only CNVs (not copy-neutral segments)\n",
    "glass_cnvs_df = glass_cnvs_df[glass_cnvs_df['hlvl_call'] != 0]\n",
    "\n",
    "# Keep only genes with exons covered by all versions (v1, v2, v3) of OncoPanel.\n",
    "glass_cnvs_df = glass_cnvs_df[glass_cnvs_df[\"gene_symbol\"].isin(oncopanel_genes[\"Gene\"])]\n",
    "\n",
    "# Simplify CNV calls to just amp or del\n",
    "glass_cnvs_df['cnv'] = glass_cnvs_df['hlvl_call'].map({2: 'amp', 1: 'amp', -1: 'del', -2: 'del'})\n",
    "\n",
    "# Scoring\n",
    "glass_cnvs_df['hlvl_call_abs'] = np.abs(glass_cnvs_df['hlvl_call'])\n",
    "\n",
    "# Deduplicate, keeping strongest event\n",
    "glass_cnvs_df.sort_values('hlvl_call_abs', inplace = True)\n",
    "glass_cnvs_df = glass_cnvs_df.drop_duplicates(['sample', 'gene_symbol', 'cnv'], keep = 'first')\n",
    "\n",
    "print(glass_cnvs_df.shape)\n",
    "glass_cnvs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc7a0373-09de-47f1-b46e-998531885a4c",
   "metadata": {},
   "source": [
    "### Intersect sample lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "48113151-34b6-40ac-a71e-0da8f533005c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(478, 29)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_samples_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f526a132-40c4-4fa0-813e-a6066aa2e50b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "389"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_intersection = (set(glass_samples_df.index).\\\n",
    "    intersection(wes_mut_samples).\\\n",
    "    intersection(wes_cnv_samples))\n",
    "\n",
    "len(sample_intersection)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c9be5e-17b8-4a8f-9fb7-2cf4bdbd6bc6",
   "metadata": {},
   "source": [
    "### Subset dataframes to these samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d34953d1-af8f-49f7-a359-f0778de5bac7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "330 361\n"
     ]
    }
   ],
   "source": [
    "variants_subset = glass_variants_df[glass_variants_df['sample'].isin(sample_intersection)]\n",
    "\n",
    "cnvs_subset = glass_cnvs_df[glass_cnvs_df['sample'].isin(sample_intersection)]\n",
    "\n",
    "print(variants_subset['sample'].nunique(),\n",
    "      cnvs_subset['sample'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4e8315f1-e262-4f01-9609-9047b2979f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_variant_median_rates(variant_df, pt_col=\"sample\", variant_col=\"hlvl_call\"):\n",
    "    counts = variant_df[variant_col].value_counts()\n",
    "    per_pt_median_rates = variant_df.groupby([variant_col, pt_col]).size().reset_index().pivot_table(index=pt_col, columns=variant_col, fill_value=0.0).droplevel(0,axis=1).median()\n",
    "    plt.bar(per_pt_median_rates.index, per_pt_median_rates.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dc0b2c55-e873-4444-822a-1c1a5e4617b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGzCAYAAACPa3XZAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAI3pJREFUeJzt3Qu8VFW9B/DFQx4+OITIK0HwlZlJhYrkIygu6PWSmplWt8C82jW1EMsrVpqvi2UppQj5+RRk5SMrsdTwGil0P0ImXistvWIQmImvCwgmGsz9/HefOZ05nHPgwDlrOOd8v5/P5jAze2avWbNn5rfXXmtNp1KpVEoAAJl0zrUhAIAgfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV80KYMHTo0TZo0qfbyAw88kDp16lT8Bf7hS1/6UvHeaK7Ro0cXC7Qm4YNmmzNnTvGhFst///d/b3Z7zNg/ePDg4vZ/+Zd/Se3dHXfckY499tjUt2/f1K1btzRo0KD0oQ99KP3iF7/YLCTFsmTJks0eIwLVrrvuWvz/+eefT127dk3/+q//2ug2X3nlldSzZ8/0gQ98ILVl//mf/5nmzp1b7WK0Wa+++moRMtpr+G7vz68jEz7YZj169Eg333zzZtcvWLAgPfPMM6l79+6tXoajjz46/fWvfy3+5hYh67TTTisCwKpVq9KUKVPSrFmz0tlnn53++Mc/pve9733pwQcf3Ox+8WHalH79+qV/+qd/SnfeeWfx4duQH//4x+m1115rMqC0BcLH9on949JLL23wy/kLX/hC8d5orv/6r/8qlh39+dG2CR9ss3/+539Ot99+e/rb3/5WcX0EkhEjRqQBAwa0ehk6d+5chKD4m9vXvva1ohVo8uTJRWvGRRddlD7xiU+kz3/+8+nhhx9ON910U9GCUdc73vGOdNddd6VHHnmkycf+6Ec/mtatW5d+8pOfNHh71HFNTU067rjjWvQ5sXUaC4W5bNq0qQifTYl9L94bzRWtd7FAq4pftYXmmD17dvwScun2228vderUqXTPPffU3rZhw4bSm970ptLXvva10l577VU67rjjKu67cePG0rXXXls68MADS927dy/169evdOaZZ5ZefvnlivU2bdpUuvzyy0tvfvObSz179iyNHj269NhjjxWPOXHixNr17r///qIs8bds4cKFpQ9+8IOlwYMHl7p161bac889S5MnTy69+uqrFduIx9lll11KzzzzTOn4448v/t+3b9/S+eefX/rb3/7WZB3EY/Xp06d0wAEHbHHduuX81re+VdTPhAkTGixL2bp164rL9dcLq1atKnXp0qV0+umnb3G7f/3rX0uXXHJJab/99ivqe8CAAaUTTzyxtHTp0optTZkypainqK/999+/dPXVVxevQdmyZcuK8sdrX19cH9soi//HdU899VTxvGpqakq9evUqTZo0qbR+/fqK+9Vfyq/t2rVrS5/5zGeK1zvKtMcee5TGjh1bWrJkSZPPt7ztP/zhD6WTTz65tNtuuxWv06c//emiLur77ne/W3rXu95V6tGjR/G6nHLKKaUVK1ZUrPOe97yn9La3va308MMPl4466qhif4yyNab8Wj799NOlcePGlXbeeefSwIEDS5deemlFnYao51GjRhVljDJEWeJ91VAdn3322aXvfe97xXuna9euxfuooTosvxblumjoOR966KHF8+jdu3fxnO69996K5xtL/X331ltvLU2dOrXUv3//4jnFvlm/rlryvVfe5xp7frRtWj7Yrs6fo0aNSrfcckvtdT/72c/SmjVr0qmnntrgfT75yU+mz33uc+mII45IX//614vTFt///vfT+PHj0xtvvFG73sUXX5y++MUvpuHDh6err7467b333mncuHFp/fr1WyxXtMbEkelZZ52VrrvuuuKx4+/HP/7xzdbduHFjcfvuu++evvrVr6b3vOc9RYvGjTfe2OQ2oq/Lyy+/nD7ykY+kLl26pK3Vq1evdN5556Wf/vSnTbZ+7LLLLun4449P9957b7Gdum677bai3NE60pRYJ/rcRLN1tETF8/rMZz5TvD6PPfZYsU58r73//e9P1157bTrmmGPSNddck97ylrcUr1GcRtoe0e8l+qZMmzat+H+0EkVZyr773e8Wp+aOOuqo4v+xxP4R/v3f/z3NnDkznXTSSemGG25In/3sZ4s+Ln/4wx+2etvRMhDbjha6b3zjG+nMM8+sWOfKK68s9on99tuveN7RgjV//vziFN7q1asr1n3ppZeKfj3RcjV9+vQ0ZsyYLdZ91Gf//v3TV77ylaL+L7nkkmKpK94D73znO9Nll11WnIKK1oqTTz453X333Zs9ZvQhin3nlFNOKe536KGHFnUUTjzxxNo6bKofUNT/xz72sbTTTjsV24zL0T+rbv+kxkR9Rbn+4z/+I336059O9913Xxo7dmzFqZ2WfO/tsccezX5+tCHVTj+03ZaPX//616Xrr7++OLosH9nE0eaYMWOK/9dv+fjlL39Z3O/73/9+xePNmzev4vrnn3++OGqK+9Y9Urzooosqjo4ba/mof5QVpk2bVrTS/OlPf6q9Lh4n7nvZZZdVrPvOd76zNGLEiCbr4Otf/3px3zvuuGMrauwf5Yyj2tWrVxdH2e9///sbbfkId999d3Gfb37zmxXXH3744UWLULQiNeXb3/52cf9rrrlms9vK9Tp37txinSuuuKLi9jh6jfoqt5BsS8vHJz7xiYr1osVl9913r7gunnPd17MsWkviSL+5ytuuW7fhU5/6VHH9b37zm+Ly8uXLi9ajK6+8smK93/3ud0WrQt3roxUg7jtr1qytKkN5vzr33HMr6jv259ivX3jhhUb31ddff7100EEHld773vdWXB+P17lz59Ljjz9ecX08VmOtAfVbPqIlKh4jXof6+07d91ljLR+xz0WLVNkPfvCD4vp4LzT2fLb3vdfU86Nt0/LBdokjzDjyiX4McZQbf6M1oCFxVBT9FKIz5Ysvvli7xFFhjPS4//77i/V+/vOfp9dffz2de+65FUMF48h0a8QRclm0lMQ23v3udxdH+f/zP/+z2fpxlF1XHIlHh9GmrF27tvi72267peaKOojnEv05GipPWbT0xNFf3U69y5YtS4sXL04f/vCHt9jP5Uc/+lExAifqsb5yvd5zzz1Fy00cydZ1/vnnF/UVLVnbqqF6jRaEct01pXfv3ulXv/pVevbZZ7dp29Hpt65yHcTzLXfYjX4Tsf/W3Rejn1K0hJT3xbJooYlWuuY455xzKuo7Lsd+Hft3Q/vq//3f/xWtUlFPDbWKRcvAgQcemLZVdOyN5xytivX3na0ZkhutF3X39w9+8INp4MCBtXWa671H+yB8sF3iyzGaXuMLMj7Qoyk1PpQa8tRTTxUfrjGaI+5Xd4nOlTHENPzpT38q/saXQP1tvelNb9pimVasWFEMXe3Tp08RauJ+8cEdYvt1RYe8uL2u2EZ8EWzp9EmIwLUt4vRHfME2NfIlmuCjif2Xv/xl+vOf/1xcVw4iWzrlEp5++uniFEr9Tq91RV3H0OD6Ieqtb31r7e3basiQIRWXy6/dluo2xKmKODUUpwQOO+ywop6a86VUf9/ZZ599ii/c5cuX1+6L8YUY69XfF+PUTnlfLHvzm9/crE6Ysa04VVjX/vvvX/wtlyFEWD/88MOL/TD21/Kphvr7aRg2bFjaHrE/RLm2NcDUr9MILPvuu2/F88nx3qN9aPxTCbZStHScccYZ6bnnnivOi8eXakPiqCuCR/TxaEj9D6JtEeEnWlain0Scmz7ggAOK/hPx5R0filGGuprTX6OueNzwu9/9Lp1wwgnb3PoRX6pNtX7EUNrrr7++6FcT/R7ib3x5RN+DnBo7Mo76bkxjdfv3swhNixaJOAqOOVRi2Gf0+/nyl79cBNzYx7a3/LEfxHXRstNQOctzrjR0RN9SIlRGf5voYxL9WqIVIfpizJ49u8Eh7K1RhpaU671H+yB8sN2iM1h0FIzTAdEZsjFx9BlNztHZtKkP0r322qv26LTu0eMLL7ywxaOiCAP/+7//m77zne9UdHKLznEt6cgjjyyO0iIMxBDbbfkgjfARnRej019jgW3kyJFFvcWXUXywP/7440XHv60R94tTF9GRN77UGqvreE2iBadu68cTTzxRe3vdVov6HTG3p2VkS8398WX8qU99qliiJeJd73pX8dy3JnzEvlO3pWDp0qXFl190ki7XTYSgWKfcItGSYlvRUlP3sWO/DOUyxGmxOPqPTsV158SJ8LG1mjODaTznKNfvf//7bQqvUad1Rf1FvR588MGt9t7blhlaaRucdmG7xVFiNBXHUfyECROaPJqNo6PLL798s9tirpDyF1ucxokvy+glX/coOb6ot6QcAureL/4fowNa0s4771wc3UUTffxt6Gj+e9/7XnrooYe22PoRk4k9+uijja4Xp1iidSRGSsSHcWN9auqLkSJxzj1aTuorlzdGgsRrUn+dGP0S2yp/0cdppug/snDhwor14oh9e8SRcf1AE+Wp30QfLWZxemjDhg1b9bgzZsyouBz7Uig/nxgxEftKBL/6r11cjr4p26tuncZjxuXYr2PyuRDbjzqu23oUpzCaM+la7Iehfh02JFro4rRLjHKp3wqxNa1RMW9N3dOMP/zhD9Nf/vKX2jptjfdec54fbYuWD1rExIkTt7hOnPuNFpIY/hhfttGhMj6M44gqOqPGh1T0F4nTL3GKIdaLoaLxBRlfvtFEHl+ATYmm3jjCi/tHc298acYRZmucR47hqNESEcMDo4NilD06LMbpp/gCieDR0Ayn9ft+xBf9b37zm+KLuLFTL/GFESElWo3KR85bEkef8YURQ2ajLHEaIzoBRktHtCbEUN4IizFsNCZGiy++GNocpzliWxGMoi7L/u3f/i1dddVVxd9DDjmkCCLlo/ltFZ2Nozwx1DXCRbRERD+VPffcs6jPKE+E21jn17/+dVHXWyM65sYpjRjuumjRoiIIRmiLxwvxvK644oo0derU4nnHF3O0/MT94lRPDMuNfWhbRYvGvHnzivdFtF7FvhvDVKOVrHx6MSaIi+cdZYyyRetOhKboR/Hb3/52q7YTLYhxGi5aHKOVJfpaHHTQQcVSXzxuvM4R/mNfiAAWLS5Rr1H38X5rSjx2tPhFx9uY0TcOBuIx45Rra733mvP8aGOqPdyGtj3UtikNTTIWbrzxxmI4XUxyFMN03/72t5cuuOCC0rPPPlu7TgwFjEmZYnKm5k4y9vvf/76YkGrXXXctJi4644wziiGW9YeKNjS8tanJmRrzwx/+sJhMKiaKimGaUeaYrOqBBx7YrJwNTSBV3l5DZSmLSaFinRtuuKHUHDH08fOf/3xp2LBhpZ122qmYZCyG0cYEWGWvvPJK6bzzzisNGjSoWCcmJKs/yVj5sWJisxgGG6/bhz70oWJYdGNDbesOKa2738Sw3bInnniidPTRRxevcXkYdUxU97nPfa40fPjwYjtRL/H/rXnu5W3HPhDPM+4fw5rPOeecBicZ+9GPflQ68sgji23EEpPGxRDfJ598crNJxrZWQ5OMxcRcUbb6Q1xj0rnyBHCx7aijhva/8iRjDXnwwQeL91MM492aScZiCHYMaY1tRt3E87vvvvu2ONT2lltuKSYZi4kB4/WK93bd4bOt9d5r7PnRtnWKf6odgABaQpz6i1Mp0T9oS61krSU6V8YpiRjB1R7E76pE61i0TjY2kg2aS58PACAr4QMAyEr4AACy0ucDAMhKywcAkJXwAQB07EnGYua9+CXLmPDH1LoA0DZEL46YBTcmrdvSr27vcOEjgkf8kiUA0PasXLmymKW4TYWP8o9bReHLP1sOAOzY1q5dWzQe1P2RyjYTPsqnWiJ4CB8A0LZsTZcJHU4BgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgq655N0dHNfTCu6tdhDZj+VXHVbsIAK1KywcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZNU17+YA2r+hF95d7SK0GcuvOq7aRaAKtHwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAMCOGz6mTZuWDj300LTbbrulfv36pRNOOCE9+eSTFeu89tpr6eyzz06777572nXXXdNJJ52UVq1a1dLlBgA6QvhYsGBBESwWL16c7rvvvvTGG2+kcePGpfXr19euc95556Wf/vSn6fbbby/Wf/bZZ9MHPvCB1ig7ANDef1hu3rx5FZfnzJlTtIAsWbIkHX300WnNmjXpW9/6Vrr55pvTe9/73mKd2bNnp7e+9a1FYDn88MNbtvQAQMfq8xFhI/Tp06f4GyEkWkPGjh1bu84BBxyQhgwZkhYtWtTgY2zYsCGtXbu2YgEA2q9tDh+bNm1KkydPTkcccUQ66KCDiuuee+651K1bt9S7d++Kdfv371/c1lg/kpqamtpl8ODB21okAKA9h4/o+/HYY4+lW2+9dbsKMHXq1KIFpbysXLlyux4PAGhHfT7KzjnnnHTXXXelhQsXpj333LP2+gEDBqTXX389rV69uqL1I0a7xG0N6d69e7EAAB1Ds1o+SqVSETzuuOOO9Itf/CINGzas4vYRI0aknXbaKc2fP7/2uhiKu2LFijRq1KiWKzUA0DFaPuJUS4xkufPOO4u5Psr9OKKvRs+ePYu/p59+epoyZUrRCbVXr17p3HPPLYKHkS4AQLPDx8yZM4u/o0ePrrg+htNOmjSp+P+1116bOnfuXEwuFiNZxo8fn2644Qa1DQA0P3zEaZct6dGjR5oxY0axAADU57ddAICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPAGDHDh8LFy5MEyZMSIMGDUqdOnVKc+fOrbh90qRJxfV1l2OOOaYlywwAdKTwsX79+jR8+PA0Y8aMRteJsPGXv/yldrnlllu2t5wAQDvRtbl3OPbYY4ulKd27d08DBgzYnnIBAO1Uq/T5eOCBB1K/fv3SW97ylnTWWWell156qdF1N2zYkNauXVuxAADtV4uHjzjlctNNN6X58+enL3/5y2nBggVFS8nGjRsbXH/atGmppqamdhk8eHBLFwkAaMunXbbk1FNPrf3/29/+9nTwwQenffbZp2gNed/73rfZ+lOnTk1TpkypvRwtHwIIALRfrT7Udu+99059+/ZNS5cubbR/SK9evSoWAKD9avXw8cwzzxR9PgYOHNjamwIA2uNpl3Xr1lW0Yixbtiw9+uijqU+fPsVy6aWXppNOOqkY7fL000+nCy64IO27775p/PjxLV12AKAjhI+HH344jRkzpvZyub/GxIkT08yZM9Nvf/vb9J3vfCetXr26mIhs3Lhx6fLLLy9OrwAANDt8jB49OpVKpUZvv/fee7e3TABAO+a3XQCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgB07fCxcuDBNmDAhDRo0KHXq1CnNnTu34vZSqZQuvvjiNHDgwNSzZ880duzY9NRTT7VkmQGAjhQ+1q9fn4YPH55mzJjR4O1f+cpX0je+8Y00a9as9Ktf/Srtsssuafz48em1115rifICAG1c1+be4dhjjy2WhkSrx/Tp09MXvvCFdPzxxxfX3XTTTal///5FC8mpp566/SUGANq0Fu3zsWzZsvTcc88Vp1rKampq0siRI9OiRYsavM+GDRvS2rVrKxYAoP1q0fARwSNES0ddcbl8W33Tpk0rAkp5GTx4cEsWCQDYwVR9tMvUqVPTmjVrapeVK1dWu0gAQFsJHwMGDCj+rlq1quL6uFy+rb7u3bunXr16VSwAQPvVouFj2LBhRciYP39+7XXRhyNGvYwaNaolNwUAdJTRLuvWrUtLly6t6GT66KOPpj59+qQhQ4akyZMnpyuuuCLtt99+RRj54he/WMwJcsIJJ7R02QGAjhA+Hn744TRmzJjay1OmTCn+Tpw4Mc2ZMyddcMEFxVwgZ555Zlq9enU68sgj07x581KPHj1atuQAQMcIH6NHjy7m82hMzHp62WWXFQsAwA432gUA6FiEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AoG2Hjy996UupU6dOFcsBBxzQ0psBANqorq3xoG9729vSz3/+839spGurbAYAaINaJRVE2BgwYEBrPDQA0Ma1Sp+Pp556Kg0aNCjtvffe6aMf/WhasWJFo+tu2LAhrV27tmIBANqvFg8fI0eOTHPmzEnz5s1LM2fOTMuWLUtHHXVUeuWVVxpcf9q0aammpqZ2GTx4cEsXCQBoz+Hj2GOPTSeffHI6+OCD0/jx49M999yTVq9enX7wgx80uP7UqVPTmjVrapeVK1e2dJEAgB1Iq/cE7d27d9p///3T0qVLG7y9e/fuxQIAdAytPs/HunXr0tNPP50GDhzY2psCADpi+PjsZz+bFixYkJYvX54efPDBdOKJJ6YuXbqkD3/4wy29KQCgDWrx0y7PPPNMETReeumltMcee6QjjzwyLV68uPg/AECLh49bb721pR8SAGhH/LYLAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZdU0dzNAL7652EdqM5VcdV+0iANAOafkAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArLrm3RwAtI6hF95d7SK0GcuvOq6q29fyAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAA7SN8zJgxIw0dOjT16NEjjRw5Mj300EOttSkAoKOHj9tuuy1NmTIlXXLJJemRRx5Jw4cPT+PHj0/PP/98a2wOAOjo4eOaa65JZ5xxRjrttNPSgQcemGbNmpV23nnn9O1vf7s1NgcAdORftX399dfTkiVL0tSpU2uv69y5cxo7dmxatGjRZutv2LChWMrWrFlT/F27dm1qDZs2vNoqj9seteRroN63Xmvt++Rjf996Pmfaz+dM+TFLpVL+8PHiiy+mjRs3pv79+1dcH5efeOKJzdafNm1auvTSSze7fvDgwS1dNJqpZnq1S9AxqXc6Evt7+6v3V155JdXU1OQNH80VLSTRP6Rs06ZN6eWXX06777576tSpU2rvIilG0Fq5cmXq1atXtYvTYaj36lDv1aHeq6Oj1XupVCqCx6BBg7a4bouHj759+6YuXbqkVatWVVwflwcMGLDZ+t27dy+Wunr37p06mtgxO8LOuaNR79Wh3qtDvVdHR6r3mi20eLRah9Nu3bqlESNGpPnz51e0ZsTlUaNGtfTmAIA2plVOu8RplIkTJ6ZDDjkkHXbYYWn69Olp/fr1xegXAKBja5Xwccopp6QXXnghXXzxxem5555L73jHO9K8efM264TK3087xXwo9U890brUe3Wo9+pQ79Wh3hvXqbQ1Y2IAAFqI33YBALISPgCArIQPACAr4QMAyEr4AACyEj52IMuXL0+nn356GjZsWOrZs2faZ599imFa8WN9tK4rr7wyvfvd7y5+fbkjzrCby4wZM9LQoUNTjx490siRI9NDDz1U7SK1awsXLkwTJkwopruOn6uYO3dutYvUIcRvlh166KFpt912S/369UsnnHBCevLJJ6tdrB2K8LEDiR/ei9lgv/nNb6bHH388XXvttWnWrFnpoosuqnbR2r0IeCeffHI666yzql2Uduu2224rJiCMQP3II4+k4cOHp/Hjx6fnn3++2kVrt2Jyx6jnCH3ks2DBgnT22WenxYsXp/vuuy+98cYbady4ccXrwd+Z52MHd/XVV6eZM2emP/7xj9UuSocwZ86cNHny5LR69epqF6XdiZaOOBq8/vrri8sRtONHt84999x04YUXVrt47V60fNxxxx3FUTh5xaSb0QISoeToo4+udnF2CFo+dnBr1qxJffr0qXYxYLtblpYsWZLGjh1be13nzp2Ly4sWLapq2SDH53jwWf4PwscObOnSpem6665Ln/zkJ6tdFNguL774Ytq4ceNmP7EQl+MnGKC9iha+aE094ogj0kEHHVTt4uwwhI8Mokk5mjybWqK/R11//vOf0zHHHFP0QzjjjDOqVvaOVu8ALSn6fjz22GPp1ltvrXZR2v8Py1Hp/PPPT5MmTWpynb333rv2/88++2waM2ZMMfrixhtvzFDC9qm59U7r6du3b+rSpUtatWpVxfVxecCAAVUrF7Smc845J911113FqKM999yz2sXZoQgfGeyxxx7FsjWixSOCx4gRI9Ls2bOL8+K0fr3Turp161bs0/Pnz6/t8BjN0XE5PqChPYlxHNGROjr4PvDAA8X0CVQSPnYgETxGjx6d9tprr/TVr3616CFd5uiwda1YsSK9/PLLxd/om/Doo48W1++7775p1113rXbx2oUYZjtx4sR0yCGHpMMOOyxNnz69GHp42mmnVbto7da6deuKvmNly5YtK/bt6Pg4ZMiQqpatvZ9qufnmm9Odd95ZzPVR7tdUU1NTzOHE3xMaO4jZs2fHsOcGF1rXxIkTG6z3+++/v9pFa1euu+660pAhQ0rdunUrHXbYYaXFixdXu0jtWuy/De3Xsb/Tehr7HI/PeP7OPB8AQFY6FAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQcvp/Qi0Z+RK8+dwAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_variant_median_rates(cnvs_subset, pt_col=\"sample\", variant_col=\"hlvl_call\")\n",
    "plt.title(\"Median CNV counts per participant\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00682ef6-0bd3-40de-8014-0bada5fd0d49",
   "metadata": {},
   "source": [
    "### Create a combined mutation + CNV + subtype matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "865ddd1e-cef4-4f2f-b8ba-fd84270b056c",
   "metadata": {},
   "source": [
    "#### Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "970b4eea-37c0-40ba-acea-862b73f2aeca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/389233590.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  variants_subset['gene_mut'] = variants_subset['gene_symbol'] + '_mut'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_mut</th>\n",
       "      <th>ABL1_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>AKT2_mut</th>\n",
       "      <th>AKT3_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>TSC1_mut</th>\n",
       "      <th>TSC2_mut</th>\n",
       "      <th>U2AF1_mut</th>\n",
       "      <th>WRN_mut</th>\n",
       "      <th>WT1_mut</th>\n",
       "      <th>XPA_mut</th>\n",
       "      <th>XPC_mut</th>\n",
       "      <th>XPO1_mut</th>\n",
       "      <th>ZNF217_mut</th>\n",
       "      <th>ZRSR2_mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 211 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_mut         ABL1_mut  AKT1_mut  AKT2_mut  AKT3_mut  ALK_mut  APC_mut  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1         0         0         0         0        0        0   \n",
       "GLSS-19-0266-TP         0         0         0         0        0        0   \n",
       "GLSS-19-0267-R1         0         0         0         0        0        0   \n",
       "GLSS-19-0267-TP         0         0         0         0        0        0   \n",
       "GLSS-19-0268-R1         0         0         0         0        0        0   \n",
       "\n",
       "gene_mut         ARAF_mut  ARID1A_mut  ARID1B_mut  ARID2_mut  ...  TSC1_mut  \\\n",
       "sample                                                        ...             \n",
       "GLSS-19-0266-R1         0           0           0          0  ...         0   \n",
       "GLSS-19-0266-TP         0           0           0          0  ...         0   \n",
       "GLSS-19-0267-R1         0           0           0          0  ...         0   \n",
       "GLSS-19-0267-TP         0           0           0          0  ...         0   \n",
       "GLSS-19-0268-R1         0           0           0          0  ...         0   \n",
       "\n",
       "gene_mut         TSC2_mut  U2AF1_mut  WRN_mut  WT1_mut  XPA_mut  XPC_mut  \\\n",
       "sample                                                                     \n",
       "GLSS-19-0266-R1         0          0        0        0        0        0   \n",
       "GLSS-19-0266-TP         0          0        0        0        0        0   \n",
       "GLSS-19-0267-R1         0          0        0        0        0        0   \n",
       "GLSS-19-0267-TP         0          0        0        0        0        0   \n",
       "GLSS-19-0268-R1         0          0        0        0        0        0   \n",
       "\n",
       "gene_mut         XPO1_mut  ZNF217_mut  ZRSR2_mut  \n",
       "sample                                            \n",
       "GLSS-19-0266-R1         0           0          0  \n",
       "GLSS-19-0266-TP         0           0          0  \n",
       "GLSS-19-0267-R1         0           0          0  \n",
       "GLSS-19-0267-TP         0           0          0  \n",
       "GLSS-19-0268-R1         0           0          0  \n",
       "\n",
       "[5 rows x 211 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_subset['gene_mut'] = variants_subset['gene_symbol'] + '_mut'\n",
    "variants_matrix = variants_subset.pivot_table(index = 'sample', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# Scoring: point mutations = 2\n",
    "variants_matrix = variants_matrix * 2\n",
    "\n",
    "variants_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c255550-1889-4b0a-b22b-c3a8fd51b6a3",
   "metadata": {},
   "source": [
    "#### CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "242fad5a-5ae5-4015-82a9-e7edbb9cbd12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/4001559487.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cnvs_subset['gene_cnv'] = cnvs_subset['gene_symbol'] + '_' + cnvs_subset['cnv']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_cnv</th>\n",
       "      <th>ABL1_amp</th>\n",
       "      <th>ABL1_del</th>\n",
       "      <th>AKT1_amp</th>\n",
       "      <th>AKT1_del</th>\n",
       "      <th>AKT2_amp</th>\n",
       "      <th>AKT2_del</th>\n",
       "      <th>AKT3_amp</th>\n",
       "      <th>AKT3_del</th>\n",
       "      <th>ALK_amp</th>\n",
       "      <th>ALK_del</th>\n",
       "      <th>...</th>\n",
       "      <th>XPA_amp</th>\n",
       "      <th>XPA_del</th>\n",
       "      <th>XPC_amp</th>\n",
       "      <th>XPC_del</th>\n",
       "      <th>XPO1_amp</th>\n",
       "      <th>XPO1_del</th>\n",
       "      <th>ZNF217_amp</th>\n",
       "      <th>ZNF217_del</th>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <th>ZRSR2_del</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 475 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_cnv         ABL1_amp  ABL1_del  AKT1_amp  AKT1_del  AKT2_amp  AKT2_del  \\\n",
       "sample                                                                        \n",
       "GLSS-19-0266-R1         0         0         0         0         0         0   \n",
       "GLSS-19-0266-TP         0         0         0         0         1         0   \n",
       "GLSS-19-0267-R1         0         0         0         0         0         0   \n",
       "GLSS-19-0267-TP         0         1         0         0         0         0   \n",
       "GLSS-19-0268-TP         0         0         0         0         0         0   \n",
       "\n",
       "gene_cnv         AKT3_amp  AKT3_del  ALK_amp  ALK_del  ...  XPA_amp  XPA_del  \\\n",
       "sample                                                 ...                     \n",
       "GLSS-19-0266-R1         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0266-TP         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0267-R1         0         0        0        0  ...        0        0   \n",
       "GLSS-19-0267-TP         0         0        0        0  ...        0        1   \n",
       "GLSS-19-0268-TP         0         0        0        0  ...        0        1   \n",
       "\n",
       "gene_cnv         XPC_amp  XPC_del  XPO1_amp  XPO1_del  ZNF217_amp  ZNF217_del  \\\n",
       "sample                                                                          \n",
       "GLSS-19-0266-R1        0        0         0         0           1           0   \n",
       "GLSS-19-0266-TP        0        0         0         0           1           0   \n",
       "GLSS-19-0267-R1        0        0         0         0           0           0   \n",
       "GLSS-19-0267-TP        0        0         0         0           0           0   \n",
       "GLSS-19-0268-TP        1        0         0         0           0           1   \n",
       "\n",
       "gene_cnv         ZRSR2_amp  ZRSR2_del  \n",
       "sample                                 \n",
       "GLSS-19-0266-R1          0          0  \n",
       "GLSS-19-0266-TP          0          0  \n",
       "GLSS-19-0267-R1          0          0  \n",
       "GLSS-19-0267-TP          0          0  \n",
       "GLSS-19-0268-TP          0          0  \n",
       "\n",
       "[5 rows x 475 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnvs_subset['gene_cnv'] = cnvs_subset['gene_symbol'] + '_' + cnvs_subset['cnv']\n",
    "cnvs_matrix = cnvs_subset.pivot_table(index = 'sample', columns = 'gene_cnv', values = 'hlvl_call_abs', aggfunc = 'sum', fill_value = 0)\n",
    "\n",
    "cnvs_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e8107a5-914f-40f8-bd38-6a61296bebc8",
   "metadata": {},
   "source": [
    "#### Combined\n",
    "Keep only variant types with 2 or more samples with an alteration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "005bb477-f614-4df7-93b4-87e22d4eaed6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(652, 365)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix = variants_matrix.join(cnvs_matrix, how = 'outer').fillna(0)\n",
    "matrix = matrix.loc[:, matrix.astype(bool).astype(int).sum() >= 2].T\n",
    "\n",
    "matrix.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8dde6a-0e03-4e3b-84c5-9f413863bf77",
   "metadata": {},
   "source": [
    "#### Write to file\n",
    "Will need to convert this to .gct format manually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "514f448e-4d43-4318-abd2-974d729f05db",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.to_csv('../data/gene_sample_matrix.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "516ee05d-aaee-4c61-ba27-cd1a2bb9aa02",
   "metadata": {},
   "source": [
    "Clustering on GLASS DNA, apply those to our data, assess overlap with Verhaak subtype, use those in our regressions\n",
    "- Richter's and DLBCL papers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10dd3b58-6ad9-47fa-b85c-57cba2a500e8",
   "metadata": {},
   "source": [
    "## Run clustering with external script\n",
    "\n",
    "As noted above:\n",
    "> The actual consensus clustering is done by the NMFConsensus script from GenePattern here: https://github.com/genepattern/NMFConsensus/blob/develop/nmfconsensus.R, and the clustering parameters are as listed here: gs://fc-secure-3f0f305e-f110-432b-bc9a-c8d96824cea2/jweiss/local_files/results/2024-07-22_GLASS-Verhaak-classifier/GLASS_sample_clustering.params.txt\n",
    "\n",
    "The outputs appear to be located at `os.path.join(results_dir, 'GLASS_sample_clustering_v2.membership.gct')`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c03f0ec-69bd-4c1b-a243-17fa17ef92bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Assess clustering overlap with Verhaak subtype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3897da0-3ee4-412a-b82e-651facf565b3",
   "metadata": {},
   "source": [
    "## Load clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "03204775-058e-4831-bf2d-fc6ff4baf7ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>membership</th>\n",
       "      <th>membership.1</th>\n",
       "      <th>membership.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0266.R1</th>\n",
       "      <td>GLSS.19.0266.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0266.TP</th>\n",
       "      <td>GLSS.19.0266.TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0267.R1</th>\n",
       "      <td>GLSS.19.0267.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0267.TP</th>\n",
       "      <td>GLSS.19.0267.TP</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS.19.0268.R1</th>\n",
       "      <td>GLSS.19.0268.R1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.1389.R1</th>\n",
       "      <td>TCGA.19.1389.R1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.1389.TP</th>\n",
       "      <td>TCGA.19.1389.TP</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.4065.R1</th>\n",
       "      <td>TCGA.19.4065.R1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.19.4065.TP</th>\n",
       "      <td>TCGA.19.4065.TP</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.FG.5963.R1</th>\n",
       "      <td>TCGA.FG.5963.R1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Description  membership  membership.1  membership.2\n",
       "Name                                                                    \n",
       "GLSS.19.0266.R1  GLSS.19.0266.R1           1             1             1\n",
       "GLSS.19.0266.TP  GLSS.19.0266.TP           2             2             2\n",
       "GLSS.19.0267.R1  GLSS.19.0267.R1           1             1             1\n",
       "GLSS.19.0267.TP  GLSS.19.0267.TP           3             3             3\n",
       "GLSS.19.0268.R1  GLSS.19.0268.R1           4             3             3\n",
       "...                          ...         ...           ...           ...\n",
       "TCGA.19.1389.R1  TCGA.19.1389.R1           3             3             3\n",
       "TCGA.19.1389.TP  TCGA.19.1389.TP           6             5             4\n",
       "TCGA.19.4065.R1  TCGA.19.4065.R1           1             5             4\n",
       "TCGA.19.4065.TP  TCGA.19.4065.TP           1             5             4\n",
       "TCGA.FG.5963.R1  TCGA.FG.5963.R1           3             3             3\n",
       "\n",
       "[365 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_68 = pd.read_csv(\"../data/GLASS_sample_clustering_revision.membership.gct\", sep=\"\\t\", skiprows=2, index_col=0)\n",
    "memberships_25 = pd.read_csv(\"../data/GLASS_sample_clustering_v2.membership.gct\", sep=\"\\t\", skiprows=2, index_col=0)\n",
    "memberships_68"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0736b97c-5ea9-457f-bec4-96f576083745",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8\n",
       "Name                                                            \n",
       "GLSS-19-0266-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0266-TP      2      2      2      2      2      2      2\n",
       "GLSS-19-0267-R1      1      1      1      1      1      1      1\n",
       "GLSS-19-0267-TP      2      2      2      3      3      3      3\n",
       "GLSS-19-0268-R1      2      3      3      4      4      3      3"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_25.rename(columns = {**{'membership': 'k = 2'}, **{f'membership.{i}': f'k = {i+2}' for i in range(1,8)}}, inplace = True)\n",
    "memberships_68.rename(columns = {**{'membership': 'k = 6'}, **{f'membership.{i}': f'k = {i+6}' for i in range(1,8)}}, inplace = True)\n",
    "\n",
    "memberships = memberships_25.join(memberships_68[[c for c in memberships_68.columns if c.startswith(\"k\")]])\n",
    "\n",
    "memberships.drop(columns = 'Description', inplace = True)\n",
    "memberships.index = memberships.index.str.replace('.', '-')\n",
    "memberships.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc7c6fc-3237-410b-93b3-8c3160cc2205",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load ssGSEA\n",
    "(Verhaak subtype scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a6ed5f47-3e3a-49d3-be5f-9b4425301066",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(425, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_CLASSICAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_NEURAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_PRONEURAL</th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "      <th>surgery_number</th>\n",
       "      <th>surgical_interval_mo</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>4499.507042</td>\n",
       "      <td>3725.238591</td>\n",
       "      <td>1945.457473</td>\n",
       "      <td>2028.442349</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>3866.366259</td>\n",
       "      <td>3899.952531</td>\n",
       "      <td>3055.641397</td>\n",
       "      <td>1970.036575</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</td>\n",
       "      <td>2</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3998.177157</td>\n",
       "      <td>2373.110794</td>\n",
       "      <td>2792.203914</td>\n",
       "      <td>3845.965090</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3403.500159</td>\n",
       "      <td>1489.313078</td>\n",
       "      <td>4904.914629</td>\n",
       "      <td>3993.835349</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_NEURAL</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-TP</th>\n",
       "      <td>GLSS-19-0268</td>\n",
       "      <td>4367.052085</td>\n",
       "      <td>2494.211618</td>\n",
       "      <td>813.639386</td>\n",
       "      <td>3709.522026</td>\n",
       "      <td>VERHAAK_GLIOBLASTOMA_CLASSICAL</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 case_barcode  VERHAAK_GLIOBLASTOMA_CLASSICAL  \\\n",
       "sample_barcode                                                  \n",
       "GLSS-19-0266-TP  GLSS-19-0266                     4499.507042   \n",
       "GLSS-19-0266-R1  GLSS-19-0266                     3866.366259   \n",
       "GLSS-19-0267-TP  GLSS-19-0267                     3998.177157   \n",
       "GLSS-19-0267-R1  GLSS-19-0267                     3403.500159   \n",
       "GLSS-19-0268-TP  GLSS-19-0268                     4367.052085   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_MESENCHYMAL  \\\n",
       "sample_barcode                                      \n",
       "GLSS-19-0266-TP                       3725.238591   \n",
       "GLSS-19-0266-R1                       3899.952531   \n",
       "GLSS-19-0267-TP                       2373.110794   \n",
       "GLSS-19-0267-R1                       1489.313078   \n",
       "GLSS-19-0268-TP                       2494.211618   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_NEURAL  VERHAAK_GLIOBLASTOMA_PRONEURAL  \\\n",
       "sample_barcode                                                                 \n",
       "GLSS-19-0266-TP                  1945.457473                     2028.442349   \n",
       "GLSS-19-0266-R1                  3055.641397                     1970.036575   \n",
       "GLSS-19-0267-TP                  2792.203914                     3845.965090   \n",
       "GLSS-19-0267-R1                  4904.914629                     3993.835349   \n",
       "GLSS-19-0268-TP                   813.639386                     3709.522026   \n",
       "\n",
       "                                  Verhaak_subtype  surgery_number  \\\n",
       "sample_barcode                                                      \n",
       "GLSS-19-0266-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "GLSS-19-0266-R1  VERHAAK_GLIOBLASTOMA_MESENCHYMAL               2   \n",
       "GLSS-19-0267-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "GLSS-19-0267-R1       VERHAAK_GLIOBLASTOMA_NEURAL               2   \n",
       "GLSS-19-0268-TP    VERHAAK_GLIOBLASTOMA_CLASSICAL               1   \n",
       "\n",
       "                 surgical_interval_mo  \n",
       "sample_barcode                         \n",
       "GLSS-19-0266-TP                   0.0  \n",
       "GLSS-19-0266-R1                  18.0  \n",
       "GLSS-19-0267-TP                   0.0  \n",
       "GLSS-19-0267-R1                  10.0  \n",
       "GLSS-19-0268-TP                   0.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea = pd.read_csv(os.path.join(\"..\", \"data\", 'glass_verhaak_scores.tsv'), sep = '\\t', index_col = 1)\n",
    "print(ssgsea.shape)\n",
    "ssgsea.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e2be0ae-3832-49fb-a5db-5c4cff83a80a",
   "metadata": {},
   "source": [
    "Combine duplicated sample barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "11bb92a4-3e47-4fd0-8eb2-5cef6fe8c58a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(419, 6)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea = ssgsea.groupby('sample_barcode').agg({'case_barcode': 'min',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_CLASSICAL' : 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_MESENCHYMAL': 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_PRONEURAL' : 'mean',\n",
    "                                               'VERHAAK_GLIOBLASTOMA_NEURAL' : 'mean'})\n",
    "\n",
    "ssgsea['Verhaak_subtype'] = ssgsea.iloc[:, 1:].idxmax(axis=1)\n",
    "ssgsea.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "380dcc00-8286-45a5-a143-1da7b9cd0e0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_CLASSICAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_MESENCHYMAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_PRONEURAL</th>\n",
       "      <th>VERHAAK_GLIOBLASTOMA_NEURAL</th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>3866.366259</td>\n",
       "      <td>3899.952531</td>\n",
       "      <td>1970.036575</td>\n",
       "      <td>3055.641397</td>\n",
       "      <td>MESENCHYMAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>4499.507042</td>\n",
       "      <td>3725.238591</td>\n",
       "      <td>2028.442349</td>\n",
       "      <td>1945.457473</td>\n",
       "      <td>CLASSICAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3403.500159</td>\n",
       "      <td>1489.313078</td>\n",
       "      <td>3993.835349</td>\n",
       "      <td>4904.914629</td>\n",
       "      <td>NEURAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>3998.177157</td>\n",
       "      <td>2373.110794</td>\n",
       "      <td>3845.965090</td>\n",
       "      <td>2792.203914</td>\n",
       "      <td>CLASSICAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>GLSS-19-0268</td>\n",
       "      <td>3271.104740</td>\n",
       "      <td>3855.965091</td>\n",
       "      <td>2225.823185</td>\n",
       "      <td>3986.308979</td>\n",
       "      <td>NEURAL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 case_barcode  VERHAAK_GLIOBLASTOMA_CLASSICAL  \\\n",
       "sample_barcode                                                  \n",
       "GLSS-19-0266-R1  GLSS-19-0266                     3866.366259   \n",
       "GLSS-19-0266-TP  GLSS-19-0266                     4499.507042   \n",
       "GLSS-19-0267-R1  GLSS-19-0267                     3403.500159   \n",
       "GLSS-19-0267-TP  GLSS-19-0267                     3998.177157   \n",
       "GLSS-19-0268-R1  GLSS-19-0268                     3271.104740   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_MESENCHYMAL  \\\n",
       "sample_barcode                                      \n",
       "GLSS-19-0266-R1                       3899.952531   \n",
       "GLSS-19-0266-TP                       3725.238591   \n",
       "GLSS-19-0267-R1                       1489.313078   \n",
       "GLSS-19-0267-TP                       2373.110794   \n",
       "GLSS-19-0268-R1                       3855.965091   \n",
       "\n",
       "                 VERHAAK_GLIOBLASTOMA_PRONEURAL  VERHAAK_GLIOBLASTOMA_NEURAL  \\\n",
       "sample_barcode                                                                 \n",
       "GLSS-19-0266-R1                     1970.036575                  3055.641397   \n",
       "GLSS-19-0266-TP                     2028.442349                  1945.457473   \n",
       "GLSS-19-0267-R1                     3993.835349                  4904.914629   \n",
       "GLSS-19-0267-TP                     3845.965090                  2792.203914   \n",
       "GLSS-19-0268-R1                     2225.823185                  3986.308979   \n",
       "\n",
       "                Verhaak_subtype  \n",
       "sample_barcode                   \n",
       "GLSS-19-0266-R1     MESENCHYMAL  \n",
       "GLSS-19-0266-TP       CLASSICAL  \n",
       "GLSS-19-0267-R1          NEURAL  \n",
       "GLSS-19-0267-TP       CLASSICAL  \n",
       "GLSS-19-0268-R1          NEURAL  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssgsea['Verhaak_subtype'] = ssgsea['Verhaak_subtype'].apply(lambda x: x.split('_')[-1])\n",
    "ssgsea.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e84501f-0905-4451-9220-2093fb09ad68",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b24d5c90-508e-4eb9-8c35-29948f196b65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "365 419 246\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Verhaak_subtype\n",
       "CLASSICAL      80\n",
       "MESENCHYMAL    79\n",
       "NEURAL         58\n",
       "PRONEURAL      29\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap = ssgsea.join(memberships, how = 'inner').reset_index()\n",
    "# overlap = overlap[overlap['Verhaak_subtype'].isin(['PRONEURAL', 'MESENCHYMAL'])]\n",
    "print(len(memberships), len(ssgsea), len(overlap))\n",
    "overlap['Verhaak_subtype'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b577a3-1f1d-44cc-9950-1d7d892391e8",
   "metadata": {},
   "source": [
    "## Calculate clustering overlap metrics\n",
    "- Adjusted Rand Index\n",
    "- Adjusted Mutual Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6eb1a6c3-f311-4697-a481-b2a4f81249b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x398053510>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAG2CAYAAABcYt1RAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAVldJREFUeJzt3QlYVFX/B/AvO27gBiK44AqSuCtuuWUuaWabZppmVq9vZZplLrnk26JWmpWWf9uszDQrzcw0tyz3BTfcc1/BFQRln//zO5cZZmBABoG5M3w/z3MfmTt3hjsDcr9zzu+c42IwGAwgIiIi0jFXe58AERER0Z0wsBAREZHuMbAQERGR7jGwEBERke4xsBAREZHuMbAQERGR7jGwEBERke4xsBAREZHuMbAQERGR7jGwEBERkXMGltmzZyM4OBje3t6IiIjA9u3bczz2wIEDePTRR9XxLi4umDlz5l0/JxERERUvNgeWRYsWYeTIkZg0aRIiIyPRsGFDdO3aFTExMVaPv3XrFmrWrImpU6ciICCgQJ6TiIiIihcXWxc/lNaP5s2bY9asWep2eno6qlatimHDhmHMmDG5PlZaUEaMGKG2gnpOIiIicn7uthycnJyMXbt2YezYsaZ9rq6u6Ny5M7Zs2ZKvE8jPcyYlJanNSALOtWvXUKFCBdXtRERERPonbSY3b95EYGCguvYXWGC5cuUK0tLSUKlSJYv9cvvw4cP5Otn8POeUKVMwefLkfH0/IiIi0pezZ8+iSpUqBRdY9EJaY6TmxSg2NhbVqlVTL9jHx8eu50ZERER5ExcXp0pAypQpc8djbQosFStWhJubG6Kjoy32y+2cCmoL4zm9vLzUlpWEFQYWIiIix5KXcg6bRgl5enqiadOmWLt2rUX9iNxu1apVvk6yMJ6TiIiInIvNXULSFTNo0CA0a9YMLVq0UPOqJCQkYPDgwer+gQMHIigoSNWZGItqDx48aPr6/Pnz2LNnD0qXLo3atWvn6TmJiIioeLM5sPTt2xeXL1/GxIkTcenSJTRq1AgrV640Fc2eOXPGotL3woULaNy4sen2Bx98oLb27dvjr7/+ytNzEhERUfFm8zwsei3a8fX1VcW3rGEhcnwycjAlJcXep0FEBcDDw0PVqt7t9dshRwkRkXOSz0/Synrjxg17nwoRFaCyZcuqgTR3M1caAwsR6YYxrPj7+6NkyZKcCJLICT6E3Lp1y7TUTuXKlfP9XAwsRKSbbiBjWJFZq4nIOZQoUUL9K6FF/n/n1D1UKKs1ExEVNGPNirSsEJFzKZnx//puatMYWIhIV9gNROR8XArg/zUDCxEREekeAwsRkR28+eabas6pnG7rSYcOHTBixAjoVVGdn97fB2fHwEJETiUt3YAtx6/i1z3n1b9yuyhs2bJFFRP26NEjX49/7bXXLJYocaSL67x581STv2wycaiMBJEJQWUiUT2Q85NhteTYOEqIiJzGyqiLmPzbQVyMTTTtq+zrjUkPhqFb/fwPp8yLL7/8EsOGDVP/ygzfgYGBNj1eliuRzVHJpF9HjhxRw1hPnjyJF154AY8//ji2bdtm71MjJ8EWFiJymrDy3/mRFmFFXIpNVPvl/sISHx+PRYsW4b///a9qYZFP9FlNnTpVLTdSpkwZDBkyBImJlueZtUvIWgtJ79698fTTT5tuf/rpp6hTpw68vb3Vcz/22GNqvxyzYcMGfPTRR6aWj1OnTqn7oqKi0L17dxWO5DFPPfUUrly5YnpOWcdN1oST+6WlZPr06Xl6D+R7yMRg8pjWrVur17h9+3Y1k6nR6NGjUbduXTVipGbNmpgwYYLFqBHje/Ddd98hODhYzYD6xBNP4ObNm3d9fuYK6vskJSWpljFZP69UqVKIiIgwLTkjP9977rkHzz//vOn448ePq5//V199ZfM5EwMLEencreTUHLfElDR1jHT7SMuKtc4f4743fzto0T2U03Pmx48//ojQ0FCEhIRgwIAB6oJkvuqJ3C8XyXfffRc7d+5UF0AJG3dDnufll1/G//73P9WyIeuvtWvXTt0nQUVWu3/uuedw8eJFtVWtWlXNc9OpUye1vps8Xh4THR2NPn36mJ531KhRKuz8+uuv+PPPP9UFODIy0qZzk/k2lixZorrIzOfckIu1hDlZEFfO8fPPP8eHH35o8Vi5qC9duhTLly9Xm5yLhL2CPL+C+j4vvfSS6gpcuHAh9u3bp1qUunXrhmPHjqkQ+f333+Obb75RzyHzDMnvxv33349nnnnG5vMldgkRkc6FTVyV430dQ/zw9eAW2H7yWraWFXOGjJYWOa5VLW1SurbT1uNaQnK2Y09Ntb0GRbqB5GIk5IIl66LIxU5aSYSsQC8tDrKJt99+G2vWrMnWymILqQ+RT/U9e/ZUQaB69eqmhWalxcDT01O1ZEirh9GsWbPUMRKcjCRcSZg5evSo6saS1zJ//nzcd9996n654FapUuWO5yOvWVojjDObCglUco5G48ePN30tLRvSOiEX+9dff920Pz09XYUaeU1CWoCktuedd95RLVn5Pb+s7vb7yPv/9ddfq3+N3X/yeiQEyn55j6UVR37Wzz77rGrBOX36tApHlD8MLETk8GJuJhbocbaQ1g3p+pAWBeHu7q4KTuWCZwwshw4dwtChQy0eJy0g69evz/f3lU/qElKka0VCkmwPP/xwrhPv7d27V31Pa7Uy0uJw+/ZtJCcnq64No/Lly6uWozuRC7+0QEgXzx9//KFaF+Tib066zT7++GP1vSQUpKamZlvwToKMMUQIaY0yTusuj8vv+WV1t99n//79qtVEuriydhOZz9T86quvqpYcCYvyvnAW5/xjYCEiXTv4v6453ueaMRmVfxnvPD2X+XEbR3csgLPTWlfkwmteZCutDF5eXuoiJa0d+SGjbcy7lYR5vYcxIEhXhXRZTJw4UXU77dixI8cRMRISHnzwQUybNi3bfXLB/vfff/N1rsbzrV27tvq6Xr166qIvNT1SJyKk66R///6YPHkyunbtqt4XaV3JWhsiK/tmrY2R1pCCdrffR95L6e7atWtXtqnmzQOhhCBpvZJjpKtIgiXlD2tYiEjXSnq657h5e2gXihY1yqvRQDnNpSn75X457k7PawsJKt9++6266O7Zs8e0SUuGBJgffvjBdAHPOlpm69atuT63n5+fqj0xkk/zUjBrTlpzOnfujPfee0/VUEhh7bp169R90iUkjzHXpEkTHDhwQLUuSLgw36TrplatWupCbn6u169fVxdcW40ZM0a1qBjrPjZv3qxahN544w00a9ZMFQtLF4ktCvL87vb7SNeavL8SSLK+l+bdcFKvEh4errqUpOhYWtsofxhYiMjhubm6qKHLImtoMd6W++W4giT1CHIhk9qU+vXrW2yPPvqoan0Rw4cPV7UiUtsgF71Jkyap4JAbKY79/fff1Xb48GHVWiFFs+bfW7pXJCDJhV+Ck7QQGLstJJTIBVdCjIwCkvtefPFFXLt2Df369VMtMdIKsmrVKgwePFhdfKVlQF6LFJxK8JGAJCOOpPXEVlIXI11U0vIjJKBIvYe0qsj3lXM3dqPlVUGe391+H+kKkhYjGUn0yy+/qKHc0jU4ZcoU9TMTs2fPVi1LElbkWBnlJf9KdxPZjoGFiJyCzLPy2YAmCPC17B6S27K/MOZhkUAiLRzWun0ksMhIHGn5kJoWGcIrxaVNmzZVAUMCSG7kk/mgQYPUBbF9+/aqVqVjx8xuLOn2kQulBBtpwZkzZ45q0ZGhtMYCUOmGCAsLU601xuLQTZs2qXDSpUsX9clfhk7Lcxkvxu+//z7uvfde1XUkr61t27bqnPPjlVdeURdvuZD36tVL3ZaRNVKMKi0u8p7YqiDP726/jwRQ+flInYoERQkkEgSrVaumQqYEHhkNJuFNyNcSHvPzuglwMWTtJHVAMs5f/mBIlXrWAi4icgwyYkY+pdaoUUMNCc0vGboso4GkwFZqVqQbqKBbVgrD2LFj8c8//2Djxo32PhWiIvv/bcv1m0W3RORUJJwYhy47AvnMeOLECTWk1jgsmYiyY5cQEZEdySdL6baRItlx48bZ+3SIdIstLEREdiT1IzJ3BxHlji0sREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsRER2ICsryxT1Od3Wkw4dOqgp/J2FrLM0c+bMAn3OW7duqeUYZLZWWfnZfN0nPfvrr78c5nwZWIjIuaSnASf/Afb/pP0rt4uALHIna/f06NEjX4+XtX9ktltHDBnz5s1TFz1Z0yirxYsXq/skJDhKSMpPeJQFDmVpBVkjSVbZtra+lL11sPKetm7dWrfnmxUnjiMi53FwGbByNBB3IXOfTyDQbRoQ1qtQv7UshDhs2DD174ULF9RCg7auECyboypVqhRiYmJUcGvVqpVpv7wfshigs5MVqCWwyUrd+SWLUkq4K+jVp3MjMywHBATAEbCFhYicJ6z8ONAyrIi4i9p+ub+QxMfHY9GiRWoFZmlhkRaHrKZOnYpKlSqhTJkyGDJkiFoMLrdP9dY+DctqwE8//bTptqz+W6dOHbWYnDz3Y489pvbLMRs2bMBHH32kLoCynTp1St0XFRWF7t27q3Akj3nqqafUCsJGCQkJagViub9y5cqYPn16nt4Dd3d3PPnkk/jqq69M+86dO6e6HGS/OTk/eS3m5LXKa87t/OV9lZmBzS1dulTdbx4cHnroIfXa5DU0b94ca9aswd0wnu8HH3yg3pMKFSrgxRdfREpKirpfzlvep7///ludi/F1XL9+Xb2X5cqVQ8mSJdX7fuzYMdPzGl/PsmXL1PIMXl5ealVtaY16++23TT+H6tWrq2MuX76sXpvsa9CggVoN3Ojq1avo168fgoKC1PeSlbhl9W7z12DtPbXWJfTzzz+rVb/lfORcsv4OyL53331XrSguv88SSOfOnYvCxsBCRPqWnJDzlpJx0ZduH2lZgbXF5zP2yf3m3UM5PWc+/PjjjwgNDUVISAgGDBigLtqyqKH5/RJI5I+8XGTkoidh427I87z88sv43//+hyNHjmDlypVo166duk8uStLK8dxzz6nmftmqVq2qLkqdOnVSiyzK4+Ux0dHR6NOnj+l5R40apS5sv/76K/788091QYuMjMzTOckFTF6r1HMYL8jdunVT4cEWOZ1/XsPjAw88oLrXdu/erb7/gw8+qILA3Vi/fr0KQ/KvdP/IazMG019++UWdq5yznKvcNoYEeZ8lbEjLk/xOyLkZg46Q92ratGn44osvcODAAfj7+6v9H374Idq0aaNeg4Tgp556SgUY+f2Sn0etWrXUbePvmQTgpk2b4vfff1eh9Pnnn1eP2b59u03v6a5du9TvwxNPPIH9+/er39sJEyZkC+ESYpo1a6bO74UXXlBhXX4PCxO7hIhI397NpWulTheg/2Lg9ObsLSsWDNr9clyNe7VdM8OBW1ezH/pmrM2nKN0eciERcoGUBQ3lom/8pC0FntKqIpuQT8/yqT9rK4st5AIs3TA9e/ZUn3LlU7hxtWepR5Cmfvmkbd7cP2vWLHWMBCcjCVdy4Tp69KjqxpLXMn/+fNx3333qfrk4V6lSJU/nJM9ds2ZN/PTTT+piKRe5GTNmqNWobZHT+edFw4YN1Wb01ltvYcmSJSo0vPTSS8gvaSWR90/qlCScSoiQUCQBoHz58upczbtXpCVFvuemTZtUnYj4/vvv1XstrUKPP/642ifhRcKr+TkLCTb/+c9/1NcTJ07EZ599plqLjI8bPXq0CiASOOV7SsuK1EEZSffkqlWrVIBs0aJFnt9T+XnJz15Ciqhbty4OHjyI999/36J1T85PgorxXCRgSZiT0F5Y2MJCRI4vPrpgj7OBfKqUT7HSHG/sGunbt6+68BsdOnQIERERFo8zr/PIj/vvv1+FFAkIEg7kYmhs2cjJ3r171UXFWC8jm1x8hbQeyJacnGxxrnIxtuUiJK0sX3/9tQps0r0kF7aiJC0scuGWehLpbpHXKO//3bawSBeJhBUjaSWTmp2cyPeU3wXz91K6kuS9lPuMJERI905W5vuMLVTh4eHZ9hnPQepfJJzJMfIzk9ctgcXW1y3nJi075uS2BDD5HtbOT7qUJATl9n4UBLawEJG+jcul5cQl4wJSOo9dDubHjdiPgiDBJDU11aLIVprppf9fPpHnd/SFFF6adysJ864EaVWRrgHpspGuG/kULs33O3bsyFbnYX4xl+4R6YLISi7A//77L+5W//798frrr6tzkSAlF21bX1tO8vI4CSurV69W9Sa1a9dGiRIlVG2PBLG74eHhYXFbLtLp6em4W3J+5jU41r6f8X4PK/uM5yAtINLtI615Elqk9U3qgu72dRf1+5EbtrAQkb55lsp58/DWjqneWhsNhOx/+DUugE+QdtydntcGElS+/fZb1Z+/Z88e0yYtGRJgjEWP8ml/27ZtFo/dunVrrs/t5+en6gyM5NOt1CaYkzDQuXNnvPfee9i3b58qoly3bp328jw9LT4RiyZNmqg6CSmalIu5+SYXOKmLkAuR+blK4ah0F+WVfLrv1auXamGR1pa8vDYh75s5a+cvj7t586ZqucnpcdIFI10XDz/8sLpwyyd/Y8FxUZKfufx+mL+XUhgrLXJSYFvQNm3apApypWtSupek5S3rz83ae2rtvOW5sj63dA2ZtzDZAwMLETk+Vzdt6LKSNbRk3O42VTuuAC1fvlxd0KU2RYazmm8yiZixW2j48OGqVkS6SuQiMmnSJBUcciPFsVJAKdvhw4dVUaP5SA753h9//LG6YJ8+fVoFJ/mEa+y+kVAiF0u5WMsoILlPRrZcu3ZNdV9JS4x0AUm3weDBg9WFTLoR5LVI4a0EHwlIcvG3dZit1K7I9zR2N1l7bVKMKucsXQ3yfmQNY9bOX7pXpAZj3Lhx6twXLFiQrRhURk1J0asxOMoIpcL+5G+NnIcECKlx2bhxozoXCRNSayL7C+P7rV69Ws0DI906Uv8i9S13ek+zevXVV1VtjnQvye+q1DBJS6F5fYy9MLAQkXOQeVb6fAv4VLbcLy0vsr8Q5mGRQCItHNa6fSSwyEVZWj6kpkWKGKWrREZySMCQAJIbaZ0YNGiQGgnSvn179Ym5Y8eOpvul20cuzHLxl0/Fc+bMUS06Umsh5AIjn4jl07y0TEgtg7T6yKdlCSddunRRLRDSbSDPZQwl0rVw7733qq4jeW1t27ZV52xrN4fUa+Ska9eupvdDCkml1URepzlr5y+tN1IQvGLFCtOwXel6ylo0KgWyUugqr0G+l7Qs2YMEVHnvpDBaapakO0vOPWt3SkEYP368ep3yeqXYW1qWsg4dt/aeZiXPIYW6CxcuVMFbuhplJJp5wa29uBiydgg6oLi4OPUHQyrzZVpkInI8MmLm5MmTqFGjhppXJN9k6LKMBpICW6lZkW6gAm5ZKQxjx45VM6XKp3Gi4vL/O86G6zeLbonIuUg4MQ5ddgDymVGG/UozvHFYMhFlxy4hIiI7kk+W0kQvBZFSm0FE1rGFhYjIjqR+JCkpyd6nQaR7bGEhIiIi3WNgISJdcYJxAERUCP+vGViISBeMQz3vNL08ETke4//ruxnSzRoWItIFmR9C6jmM65HIBGHWpiwnIsdqWZGwIv+v5f/33cyWy8BCRLphXEW2sBdRI6KiJWHF1pW3s2JgISLdkBYVWYTP398/T4vhEZH+STdQQaxDxMBCRLojf9zsvdAaEekLi26JiIhI9xhYiIiISPcYWIiIiEj3GFiIiIhI9xhYiIiISPcYWIiIiEj3GFiIiIhI9xhYiIiISPcYWIiIiEj3GFiIiIhI9xhYiIiISPcYWIiIiEj3GFiIiIhI9xhYiIiIyDkDy+zZsxEcHAxvb29ERERg+/btuR6/ePFihIaGquPDw8OxYsUKi/vj4+Px0ksvoUqVKihRogTCwsIwZ86c/JwaEREROSGbA8uiRYswcuRITJo0CZGRkWjYsCG6du2KmJgYq8dv3rwZ/fr1w5AhQ7B792707t1bbVFRUaZj5PlWrlyJ+fPn49ChQxgxYoQKMMuWLbu7V0dEREROwcVgMBhseYC0qDRv3hyzZs1St9PT01G1alUMGzYMY8aMyXZ83759kZCQgOXLl5v2tWzZEo0aNTK1otSvX18dN2HCBNMxTZs2Rffu3fH222/f8Zzi4uLg6+uL2NhY+Pj42PJyiIiIyE5suX7b1MKSnJyMXbt2oXPnzplP4Oqqbm/ZssXqY2S/+fFCWmTMj2/durVqTTl//jwkP61fvx5Hjx5Fly5drD5nUlKSepHmGxERETkvmwLLlStXkJaWhkqVKlnsl9uXLl2y+hjZf6fjP/nkE1W3IjUsnp6e6Natm6qTadeundXnnDJlikpkxk1aeIiIiMh56WKUkASWrVu3qlYWacGZPn06XnzxRaxZs8bq8WPHjlXNR8bt7NmzRX7OREREVHTcbTm4YsWKcHNzQ3R0tMV+uR0QEGD1MbI/t+Nv376NcePGYcmSJejRo4fa16BBA+zZswcffPBBtu4k4eXlpTYiIiIqHmxqYZHuGimGXbt2rWmfFN3K7VatWll9jOw3P16sXr3adHxKSorapBbGnAQjeW4iIiIim1pYjEOQBw0ahGbNmqFFixaYOXOmGgU0ePBgdf/AgQMRFBSk6kzE8OHD0b59e9XNIy0oCxcuxM6dOzF37lx1v1QFy/2jRo1Sc7BUr14dGzZswLfffosZM2YU9OslIiKi4hBYZPjx5cuXMXHiRFU4K8OTZQ4VY2HtmTNnLFpLZATQggULMH78eNX1U6dOHSxdulQNZTaSECN1Kf3798e1a9dUaHnnnXcwdOjQgnqdREREVJzmYdEjzsNCRETkeAptHhYiIiIie2BgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLdY2AhIiIi3WNgISIiIt1jYCEiIiLnDCyzZ89GcHAwvL29ERERge3bt+d6/OLFixEaGqqODw8Px4oVK7Idc+jQIfTq1Qu+vr4oVaoUmjdvjjNnzuTn9IiIiKi4B5ZFixZh5MiRmDRpEiIjI9GwYUN07doVMTExVo/fvHkz+vXrhyFDhmD37t3o3bu32qKiokzHHD9+HG3btlWh5q+//sK+ffswYcIEFXCIiIiIXAwGg8GWB0iLirR+zJo1S91OT09H1apVMWzYMIwZMybb8X379kVCQgKWL19u2teyZUs0atQIc+bMUbefeOIJeHh44LvvvsvXi4iLi1MtM7GxsfDx8cnXcxAREVHRsuX6bVMLS3JyMnbt2oXOnTtnPoGrq7q9ZcsWq4+R/ebHC2mRMR4vgef3339H3bp11X5/f38VipYuXZrjeSQlJakXab4RERGR87IpsFy5cgVpaWmoVKmSxX65fenSJauPkf25HS9dSfHx8Zg6dSq6deuGP//8Ew8//DAeeeQRbNiwwepzTpkyRSUy4yYtPEREROS87D5KSFpYxEMPPYRXXnlFdRVJ11LPnj1NXUZZjR07VjUfGbezZ88W8VkTERFRUXK35eCKFSvCzc0N0dHRFvvldkBAgNXHyP7cjpfndHd3R1hYmMUx9erVw8aNG60+p5eXl9qIiIioeLCphcXT0xNNmzbF2rVrLVpI5HarVq2sPkb2mx8vVq9ebTpenlOKeI8cOWJxzNGjR1G9enVbTo+IiIiclE0tLEKGNA8aNAjNmjVDixYtMHPmTDUKaPDgwer+gQMHIigoSNWZiOHDh6N9+/aYPn06evTogYULF2Lnzp2YO3eu6TlHjRqlRhO1a9cOHTt2xMqVK/Hbb7+pIc5ERERENgcWCRaXL1/GxIkTVeGs1JxIwDAW1spkbzJyyKh169ZYsGABxo8fj3HjxqFOnTpqBFD9+vVNx0iRrdSrSMh5+eWXERISgp9//lnNzUJERERk8zwsesR5WIiIiBxPoc3DQkRERGQPDCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsRERE5JyBZfbs2QgODoa3tzciIiKwffv2XI9fvHgxQkND1fHh4eFYsWJFjscOHToULi4umDlzZn5OjYiIiJyQzYFl0aJFGDlyJCZNmoTIyEg0bNgQXbt2RUxMjNXjN2/ejH79+mHIkCHYvXs3evfurbaoqKhsxy5ZsgRbt25FYGBg/l4NEREROSWbA8uMGTPw3HPPYfDgwQgLC8OcOXNQsmRJfPXVV1aP/+ijj9CtWzeMGjUK9erVw1tvvYUmTZpg1qxZFsedP38ew4YNw/fffw8PD4/8vyIiIiIq3oElOTkZu3btQufOnTOfwNVV3d6yZYvVx8h+8+OFtMiYH5+eno6nnnpKhZp77rnnjueRlJSEuLg4i42IiIicl02B5cqVK0hLS0OlSpUs9svtS5cuWX2M7L/T8dOmTYO7uztefvnlPJ3HlClT4Ovra9qqVq1qy8sgIiIiB2P3UULSYiPdRvPmzVPFtnkxduxYxMbGmrazZ88W+nkSERGRgwSWihUrws3NDdHR0Rb75XZAQIDVx8j+3I7/559/VMFutWrVVCuLbKdPn8arr76qRiJZ4+XlBR8fH4uNiIiInJdNgcXT0xNNmzbF2rVrLepP5HarVq2sPkb2mx8vVq9ebTpealf27duHPXv2mDYZJST1LKtWrcrfqyIiIiKn4m7rA2RI86BBg9CsWTO0aNFCzZeSkJCgRg2JgQMHIigoSNWZiOHDh6N9+/aYPn06evTogYULF2Lnzp2YO3euur9ChQpqMyejhKQFJiQkpGBeJRERERWvwNK3b19cvnwZEydOVIWzjRo1wsqVK02FtWfOnFEjh4xat26NBQsWYPz48Rg3bhzq1KmDpUuXon79+gX7SoiIiMhpuRgMBgMcnAxrltFCUoDLehYiIiLnu37bfZQQERER0Z0wsBAREZHuMbAQERGR7jGwEBERke4xsBAREZHuMbAQERGR7jGwEBERke4xsBAREZHuMbAQERGR7jGwEBERke4xsBAREZHuMbAQERGR7jGwEBERke4xsBAREZHuMbAQERGR7jGwEBERke4xsBAREZHuudv7BIiIiIpaWroB209eQ8zNRPiX8UaLGuXh5upi79OiXDCwEBFRsbIy6iIm/3YQF2MTTfsq+3pj0oNh6Fa/sl3PjXLGLiEiIipWYeW/8yMtwoq4FJuo9sv9pE8MLEREVGy6gaRlxWDlPuM+uV+OI/1hYCEiomJBalaytqyYk5gi98txpD8MLERE5PRib6fgh+1n8nSsFOKS/rDoloiInFp8UiraTluHm4mpeTrev4xXoZ8T2Y4tLERE5FQMBgMOX4oz3S7t5Y52df1Q1780fEt4ILfBy3Lf7PX/4uy1W0VyrpR3Lgb5yTq4uLg4+Pr6IjY2Fj4+PvY+HSIisoPk1HT8tvcCvtp0EgcuxGHdq+1R06+0uu9mYooKLqsOXFKjgYQhS1CR2+6uLkhNN6CkpxtGdQ3BwFbBnJ9FJ9dvdgkREZFDu5aQjAXbTuObLadx+WaS2uft4Yr952NNgaWMt4f6V+ZZ+WxAE7y1bD+qxu+FP24gBmVxtnRDTOgVjpAAH4z+eZ8qvJURQ8v3XcS0R8NR27+MXV8jsYWFiIgcOKi8v+oIfok8h6TUdFP9yaDWwXiyRTWUK+Vp/YEHl8GwcjRc4i6Ydhl8AuHSbRoQ1gvp6QYs2H4GU/84rOpfPN1cMeL+OnihQ+2iemnFRhxbWIjIGXD6dMpNCQ83/BF1UYWV+kE+GNK2BnqEB8LTPZfyzIPLgB8HwiXLbCwucRfVfvT5Fq5hvTCgZXV0CvXHG0v2Y/2Ry7gWn1z4L4hyxRYWItIlTp9O5hJT0vDrnvNYf/gyPu3fBK4ZwVX2BfhoYdbF5Q5hNj0NmFkfMGtZseQC+AQCI/YDrm5qj1wiV+y/hI6hfijp6W6aFbdsSQ94e2jHUP6xhYWInGL69KyfpozTp0sNAkNL8SA1Kd9tPY3vt57G1QStlePvY5fRIcRfff1Qo6C8P9m/a3MJK8IAxJ0HTm8Gatyr9kgI6tEg83dNuoteWhCpzmXqI+GIqFkhvy+NbMTAQkQONX26fIaW++8PC2D3kBM7dDEOX248iWV7LiA5TatPCSpbAk+3DkbjauXy9iRJ8YCbJ+CeUcuy4/O8PS4+Ose7zt+4jbPXbyE6Lgl9527FgJbVMLpbqKmolwoPAwsROez06a1q8dOtM9p37gZ6zdpkut2kWlkMaVsTXe+pBHc319wDytltwKmN2nYhEuj/E1Cro3Z/1ZbAsT/vfAKlK+V4V9XyJbF6ZHtMWXEIP2w/i/lbz2DdoRi883A4OoZqrT5UOBhYiEhXDl6IzdNxP+48i0o+XqZhq+S4biWn4vClm2iS0XISHuSLewJ9UKNiKVVIm2uLyvXTwK55mQElPctsthd2ZwaWNsOBnV8AUmBrtQ0vo4alciPg2BqgTmer39LH2wNTHmmABxsEYswv+3Hm2i0MnrcDDzcOwlu966v5XqjgseiWiOzuanySmu9iye7z2HP2hk2PbV2rAhY817LQzo0Kz8XY2/hm82m1xo9ciraMvQ+lMi72MglcttE+yQlaC4p3WSCoibbvUhQwp03mMb7VtPqT4LZA9TZAuepWRwnB6tRxUKOEEHMQ+GsKUKcL0HUKUDHn4cy3k9MwY/UR1X1Vr7IPlr7YBh65tQKRBRbdEpHuyR/6Pw9ewtLd5/H3sSuqdsV42fBwd1UXrJz4eLurYaw7Tl1HTb9Spv2paekYsWgPWtasoJrnpeaB9Ee6fOQC//u+i2pWWVG9QknVUiEXfaHCijGgGLt4zu/SWlAa9AUemas9mX8Y0HQwUKW5FlKyBpSswnppoWTlaMsCXGlZ6TZVuz86CnD10LqPjq8HWv4XaDcK8M5+QS3h6YY3eoShR4NAeLm7msJKUmoabtxKQSUf74J744o5trAQkd1GAg3NmCJdNKjii96NgtCzYWVEnr6e4/TpwjhKSKZbv52SpuZoETtPXcNjc7aYjg+pVAYdQv3QKcQfTaqX4ydfOztwIRZvLjuggqaRDEd+tm0N3FevUmYRdVoK8M2DwLkd2bt4fKoA9R8Gurx9dycjQ5xlNJAU2ErNSvXWpqHMypV/gVVjM2te5JjOk7Ww5Hrn36MZq4/i600n8cYD9dC3edU7D7kupuJsuH4zsBBRoZI/MTJF+tLdF1ClXAk807aG6RPoo59tVmHiocZBqJWlFiU/87BIF8MvkTJXRwwiz1xHxod3pYy3O959OBwPNgwsrJdKd3D6agI6fPAX3Fxc1M/h2YhKuCftkNZ6khgH9Pgg8+DP2gLR+7WAYuzika1sdRlrXHQnfXQVsHIMcO2EdrvtSKDzpFwfIq2FT8zdYgpmrWpWwNRHw1G9QmZrIGkYWIjI7s5cvaUm9Vqy5zxOXE4wNfv/9VqHPH/avJuZbm/cSlZdTX8djsFfRy+radyXvNDaVMC55fhVbDlxVc1m2iDI1zQRGRUMWe34m82nEJeYgvcea6jtTL6Ff9b/jsbpUSh9YWtGF0+Kdp8MPx5zBvDI6MY7txMoVbHoA4o1qUnA1s+AzR8Dz60DygXn6XdXWlg++PMIElPS1dpGr3UJweA2NTgc3wwDCxHZzU+7zqkiyl2nM5v9pW///rBKahRFxxD/Ig8HcvGQuokGVcqaLhajf9qHRTvPqq8rlPJE+7p+6BDqj/Z1/OBbknNq5IdcTqRlS+pTVkZdgqchCckuXvj79Y6oUq4ksLA/cHi55YPMW1DueQTwLAndSrmdGajEite1mpkWzwNuHjm2Ko35eb8Kx6Jh1bKY0adhthbF4iqORbdEVJRTpksgMbaabD95VYUVyQWta1VE78ZBav4Me06sJSEl69DYDiF+uJmUgn+OXlGzlv6y+7za5LybVi+Hb55pYZqKnXKXkpaOP6Iu4bt/DsHzwg60dD2EwR4H0dj1BHb0XI1A34yLvNSJXNhj3y6eu2EeVi7uBbb/n/b1rm+AblOA2vdle4h0Ay14LgILd5zFu78fwpFLcfDIQw0MZccWFiLKV4vF1hNX1TBk+ST9/bMR6pOjkGHJUvwqNQqOMEJCLrYSsKTuZf2RGByNjkctv1JY+2oH0zHfbTkFfx9vtK1d0TTsljJcPoL9K7/A7WMb0MjlX3i6pFne/8gXQIPHta/TUrXCVkcJKLlJTwf2zAfWTAZuXdH2hfQAur4DlNfqtGBlaQmp55LWRvOZc4vzaLY4dgkRUUGTPxUHL8apYcjL9l5QU5MbDetUG692CYEzOKemXU9E0+rlTcXBjf+3GreS0+Dp5qrqaKR1RmpfZGKzYjX6I/kWcG47zhn8cdmjstZqdfBXs3lNgLQygXCr0S6zBUXqPZz5Pbp9A9jwntbaIiOa3LyA1i8B974KeOZeZLvtxFU8+cU2PNMmGCPvD1FDpIubOAYWIipI8ilw8NfbVeuDkW8JD7UonNSlNK1WzmmLVqV4d+aaY1h3OEbNE2JOioifaVMDg1rfuQjTkQOKjOIxyHZuJ1zTUzAj9TGs8Xsav7/cFi63rgGrxsEQ3AYuwfc6f0DJScxhbTTRifVAKT9g2C7A2zfXh7y/6jBmrz+uvq5WvqQaSSTdqMVJHGtYiOhuxN5KwfEr8aap0gN8vHH9VoqazKtzPX81X0r7ED94uTv/J8KyJT3xZq971HDqk1cSVHD568hlbDt5Faev3kJCcqpFuJEZex1+0rqb0cDiQdpInYxRPC4Z2wVDeSQb3NWyCDeTUuFTqgLwyP+Z5sgptvxDgaeWAEdWaC0txrAibQJXjgF+dbM9ZFTXUFUv9caSKBWGn/x8G/q1qIqxD9RT0/+TJbawEJGpeFbqOJbukXlMLqNsSQ81VbpxVI3UpsissvxDqolPSsXmf6+omVllQTwhw7iHL9yjvq5bqbQKLjIqqmlRTVp3p8nQcmlBgVcZba0dY63JezWApDhEowI2ptXD1vQw7Ha9BxGNm2Bw25qo7c9RLnmy/yfg52eBpk8DnSYAEvCykAkQp608rBZSFBIGpz7aQP3uOLs4trAQUV6kpxuw7eQ1VZeyIuoibiZmthaUL+WpajkCM1oKGmUU1ZJGFrjrck+Axb5Snu5oHlxOFfFK95ls/7fhhJq0rl0dP7zWNUTVvRQKWSPH6nTz07Tp5rMGFNnMWlBQroYKLPIZ1sXNHXj8a2y57ot+P0er4umBrYIxrkU1lCvlWTjn76zOy4zNBmDX18CBX4CO44FmzwDyHmeQEXRv9w43LaYoLXnXE5Ltetp6xBYWomLMvA/dOJPsQ42C0LtxIEID+H8pv6xNWie2jbvPNHJKQo2UejQ0mxsm30wL+mX9c262oJ+Els/aaOvkmCsTCENwWxwr0RDvXGyGhtXKY+T9dU2BdtWBS2ra/GwLEVLendoE/DFam7nXuP6RrFtUs73Vlk5pqevTLHM6fxldVMnHyykLvFl0S0TZXLhxW43uiahR3jQnyd6zNzDgy23oEV5ZBRW5z1mLZ+05BFyGssrEddJKYfTUl9vwz7ErqiVLTVoX4qf+lZoZm7uBPrwHuHkxhwNctJaWEfuBFaOAI3+YRvAkVWmNJac88dXmU6aC6oqlvbB1bCe4c92lgiU/p13zgHVvAbczJlVs+wrQ+c071pN1/nCDmo357Yfro7JxThsnwcBCRErs7RS1Jo/MlyJdP/K/vU+zKqap0uW/f1KqTBvu/MWzeiLv+8gf92LNwWhVuGokWVEKnR+pZcCTddK0C5v5JiNy5N+eM4HSftqDfnoGiPr5zt900HKgSjPA3Rsx8UmYv+U05m87Y2r9KeXphj7Nq2Jw6xqoVkHHs806OvkZrn8X2PkVMPBXbRK9XKw9FK0WAk1OS1fdkGMfCEW/5tWc5oMFa1iIijFpxl9zKFoVz645FIPk1HTTfS2Cy6NlzcyiP2li1nVYsbWIVA8kFabc0oJFmcqZ53vyH61+5PZ1uNy+jg8NN5Be/SoS464iLeEa/uP9HjZf9sbO09fxzK1fgM0/5fw9OozNDCzyHuWFvIcZM7V+tOYYvt+mFXjKaKanWwejb4uqLKguCiXLa4s8tnkZKFstc/+OL4GSFYCwhyyGhUt3nAwff/3nfdh95oYaUbRszwVVlFto9VA6xcBC5GTkb907Kw6pIbeijn9pNT3+Q40CtfVcHEVeikgLO3gkx2dv4QjtAbh7acfsng8cXmF2TEYLSFpGweQrBwDfKtrXR1cCW2ZZfAvpdDH+RBYMqYtzXrXUkOla0QeBsyFAiXJI9PDB8mOJiENplC7nh6DKgQhMLI1gKY6VH3aj/sDBpXd8OWdTyqBqxteD2wTj0MU4DGlbUy2bwO4fOzAPK7HngT/Ha0FX5rLpPg2odI/p7jqVyuCnoa3VYpLvrzqiWku7zfxb1Ro9366mU9a2WMMuISIHdiz6purukYvckhdbm+ZFkVVipWZFgkpYZR/H+4OW1yLSvEpOABKuZO9iUQHjBtBpfOY6MWvfAiK/1e43jqAxNyIKKJtx6ZeLzOZPrH9PVw/gP38DlcK02weWAsdWAyXKqiCiNvm0bfy6Yl3LtWoySHHuqz/uwamMAGokE411DPFDn6aBqPl9K3jduqS6lLJKNwCXUAHvhf6Imf2a5f09o6Ijo7c2zQQ2fQSkJgIubkDzZ4GOY7XfjSyrYI/9ZT82/nsFjzQJwow+jeDIWMNC5MRkqLE0CUtQkanyjeYMaIpu9S2H2Tok6eKYWd+yZcVaEWnf74FrxzOCxw3LFg7ZBvwCeGXMFfLri1prSE5yCyFunkCJ8pkBo/en2myu4sw2IOZARugwCx+yybTsBRgUZaircb2jbSeuqZoG8WGfhti24hu8m/Keum0eWiSsiBdSRiD43icwuluo44XX4uT6ae3379Ay7bb8Tt03AWgyyKIrVC7bP0eex32h/qZh5jIMuqSXm8NN5sgaFiInFHU+FlP/OIxNx6+o3grh7uqCDiH+anp8GWXiFKRmJcewIgxA3Hlg/dvAv2tyPkzCizGwyB9+WePFvEXDfHM3W6SxxfNAg76ZAURaPXK6yFeL0LYiIPUKNdrWwDNtayBBJq07flXNuiuFmAvjG+G66whM8vgWgbhmeoy0rExOeQqr0lvgh7r+DCt6V6460Pc74MRfwB9jgMuHtOHQNTtaLKgoP8fHmlaxCDCvLd6L09duYdqjDdREhc6IgYXIDsNct5+8hpibifAv460W07M2D4cUy8YlpqhhpkKKY6UZWDSrXk5198hwZKebyEuKQ/NCChSlv9+8i8WilUNbvFCRoaNd3rK9tkCnZMVoWfFXNpmzQ0goWZ3UDC1cD8MfNxCDstieHop0VSkD9ftGDqJmB2DoP1ohrtRRma/+nBibbY2iS3GJ2HsuFlfik/DYnM2qiHpU1xCU9HSuS7xzvRoinZMhxpN/O4iLsYkWk7XJOjXd6ldWn5Qiz1xX3T2/77uI1rUrYvaTTdRxMhX6uw+H4946FU1TwTslGQ2UF42fuuOQUBO9jyy6CxJ6jSScyBT6dzqOHICbB9ByqOW+87uAb3pp87e0egnw0H6mMjfLmpHt8NbyQ/g58hy+3nQKqw9GY+ojDdC2jvMspsgaFqIiDCsyn4K1MlLZ171+AA5ciLNYEbhq+RJY/2qH4jWKQ6aL//J+wJA5HDvHidCcOIjY0mLXdto6NRuqtT/m8vsV4OuNjaM73f2MumRfv78G7Phc+1rqqLq+C4Q8YNFlueHoZYz7Zb9aYV3IvEsTeoap6f8d/fpdjP4KEtn3oiItK9YuKMZ9f0RdUmFFJvCS6v/vhrQofmFFyJoruYUVIdOaM6woEkKkhU5kjSPG23I/w4oTeOB94JEvtPl9rp8CFj4JzH8EuHzEdIjMlrzqlXYY1Kq6ur31xDW4uzrH3xC2sBAVEqk/uXgjUQ0v3vjvZXy58dQdHzOsU2280KE2SngW44txym1g/TtApfrA2slZ5mEJ0sJKUczD4mTdjeREkuKBf6Zr8/rInD+u7kD70UD71y0O23HqmgqtzYLLmyaVvH4rGRUy6uL0gMOaiQqZLFAmTfAXYm/jwo1EXLxxW63FY5zSXCZ4mrTsgM3P+9ETjdTzFCtSRCjTlLceDmT9JOiIM906QEE3OYlrJ4BVbwBHVgAPfgQ0fTrXw+dtOomZa49hYs8wNbJQD6PGOKyZ6C7/6F++maTCSI0KpUyjcGRNj5lrjuFi7G1cic++9Hst/9KmwCIrq4qyJT1UQVwJd1dEnr1xx+9d7Aojrx4HfngCuHIUSE0GOoy2vF/CSV4La0mFk1a1MpdeICdXvibQ7wfg1EagWqvM/TIs2qMUULW5aZe0Tfy+/yJu3EpR61j9uucC3n0kXC3N4CgYWKhYkf+0MpmW8VOnTE8u/3ElhEjXjbSWyMRsqRkzbn3Wvwm6h2vN6Slp6WrVXSNvD1cEli2BQN8Squm9kk9m2JC5UQ7+r6tpWGFeCyPlE3GxIXOoyMJ90sJSJhCoc7+9z4jIMQW3tZw199eXgNizQMMngc6TgDIBqjVlwXMtMffvE/ho7TFVnNtlxgaM7h6KARHVHWIxxXx1Cc2ePRvvv/8+Ll26hIYNG+KTTz5BixYtcjx+8eLFmDBhAk6dOoU6depg2rRpeOCBB9R9KSkpGD9+PFasWIETJ06opqHOnTtj6tSpCAwMzNP5sEuIzMmMj1EXYlX9iFTKa2FE676RfdP7NMQDGSFE+v2Hzo/M9hwSaAJ8vDGmeygebKj9Hkqry56zNxBY1luFFGk9saVJ1ThKSJj/pzM+w2cDmhSPWgP5kyN976snasW1VVoAfecDZfI4nJmIciazPK8aD+zJmNnZs7RW2xLxX8Bday3+NyYeY37epxbaFM2Dy6kV3O2xmGKh1rAsWrQIAwcOxJw5cxAREYGZM2eqQHLkyBH4+/tnO37z5s1o164dpkyZgp49e2LBggUqsERGRqJ+/frqJB977DE899xzKvxcv34dw4cPR1paGnbu3FngL5gct+9cWjik9UPVjMTe1sJIxtdPt65hmm9gZdQlDJ2/K8fnGd+jHp69t6ZpunOpN1EhpKy0lJRQTaR+ZbwKpe+/2BdGpiQCvw0H9i3UbjceAPSYkbmYIBEVjHO7gD9GaXO3iAq1ga5TgLpdTAW43209jWkrDyMpNR3LXmqDewItJ6Rz+MAiIaV58+aYNUtbdTQ9PR1Vq1bFsGHDMGbMmGzH9+3bFwkJCVi+fLlpX8uWLdGoUSMVeqzZsWOHarE5ffo0qlW786yTDCz2VRAXYfk1vJqQbOqWkRASUaMCwgK1n6esofLMNztMU9LnFkIOXIjFiIV7tO6ajNaQyqrrRgsl0vUis8baS7EujLy4D/iiM5CeCnSbok2Dr4PCPyKnlJ4O7P0BWPMmkBCj7Xt+AxCYuWDiueu31N+jR5pkTvV/NT6pyEYSFVrRbXJyMnbt2oWxY8ea9rm6uqounC1btlh9jOwfOXKkxb6uXbti6dKcl0OXE5em9rJly1q9PykpSW3mL5j0NRma1GrIfmM3x83EFBVEypXyMBWW7j8Xi3dXHNK6amIT1VT0WUOIMbBUKO2pwoqnm6sKHJlBRAshLTKG7Qn5lLB6ZHvoVbEujKzcQFs8sFRFbfpxIio8rq5A4/5AvQeBv98DYs9bhBUJNFXKlVSbkdT1PfzpJgxpWwPDOtVRH+708iHLpsBy5coV1VVTqZJlX7PcPnz4sNXHSJ2LteNlvzWJiYkYPXo0+vXrl2Paku6lyZMn23LqZKfJ0F5asBve7nsRn5yWrSXEAAO2nLhqeox80PYr7WVqGTGffj40wAc73uiMCqU8HaI4jMzsXQj41wMqN9Ruhz9m7zMiKl68fYAub2v1Y0ZxF4FvegLtRgHhfUxTCvyx/yISU9Ixe/1xNZnlI42D8P22M7roxtbVKCEpwO3Tp4/qHvjss89yPE5aeMxbbaSFRbqlqOjmIDl86SaW7Tlv8UtsjYy2MYYV3xIeptE3oqZfaczs20j98ktIkVE2nu7WZ2SU/VJXQg4kLRVYM0krsPWpoi3mJqslE5F9uJh92JP/l1f/BZb8R1tksfs0IKgJRnYJQVigLyb8GoUTlxPwwZ9Hsz1N1hZ0XQaWihUrws3NDdHRlqupyu2AgACrj5H9eTneGFakbmXdunW59mV5eXmpjYpuxlYZ+ht1LlYN6z0afdMieNzJuAdC0T+iulph1lxpL3e14jA56UgFGbJ8fJ12u9GTgLf1Ll4isoP7JmpdsxveB85tBz7vBDR5Cug0Ed3qB6hu9jbT1uF2ivaB05z89ZfoIy3s94cFFFn3kE0LDHh6eqJp06ZYu3ataZ8U3crtVq3MJq0xI/vNjxerV6+2ON4YVo4dO4Y1a9agQoVi2r+vg5YTGbYrlePSLGiUlmbAhKVRWLTzLA5ejFNhpXwpTzSokreK8vCgstnCCjkxWddE/vhJWPEoCTw+D+j0RvZZbInIfty9tFWfh+0CGvTVYkjkt8AnTYHtn+NI9E1TWHFFOlq6HkQv183qX7ktoUVa2KW2pchO2dYHSFfMoEGD0KxZMzWSR4Y1yyigwYMHq/tlyHNQUJCqMxEyRLl9+/aYPn06evTogYULF6rhynPnzjWFFRnWLMOcZSSR1MgY61vKly+vQhIVPBnStu+81mIiLSfy9TGzlpO2tSuaJkyTmV6lH1MKXMODfBFepawacSOHcjI0snBkJfDzs0DyTcC3GvDE91qhLRHpk09l4JG5QLMh2jDoi3uBG6cR46l193d13Y5JHt8i0CUzmFwwlMfklIFYld5CFeLqNrDIMOXLly9j4sSJKljI8OSVK1eaCmvPnDmjRg4ZtW7dWs29IpPDjRs3Tk0cJyOEZA4Wcf78eSxbtkx9Lc9lbv369ejQgSMJCqrmJPZ2ilrJ09iVOeir7WqfOWk5qR/ki1Y1LVu5ZvS1/NkINxdtFVjpy3TJYTI0rhJbjEhBn6wJJGGlelugzzdakzMR6V+1COC59dow6Hq94H8+RYWVOR4zs30gDcA1fOYxE/9NGQH/Mi2L7BS5+KGThhNpOdl/7gb2n48ztZxUKVcCG0d3Mh370oJIxCWmokGQrwop4VV8VcuJrQtiFfvJ0CjT7RvAtjnAva8Cbh72Phsiyqe01FRcebsu/A1XrU6VJC3sMS4V4Df+KNzc89/lz8UPi1E4kZla61XO/CE/MXerqkPJSoYD1/YvreY6MY7EmfVkkwI5DwklUnilh3H6VMRizwH7FgFtR2rNdiXKAh2yTyBJRI7F7ewWVMLVzObyLOTPe4Dcf3ZLkS1QysDiQOFEJvSJyqg7MbaciKjJXU0zt8pEa2ev3dJaTO6y5cQWxXoytOLqzFZg0QAg4bI2Aqj5EHufEREVlPjogj2uADCw6DScyIyuxgnSpvxxCF/+c9LqUGJpOZHp7GVOEzGxZxje6V2/UMMJEXZ9A/z+KpCeAlQK50rLRM6mdKWCPa4AMLDorOVk37lYHIuJx5+vtEOtjBAioUTCivxrbDmRVhP5V2pFzMOJPdfIoWIgLQVYORbY8bl2O6y3NtW+Z9Gv8kpEhah6a8AnUJsRN6dxoHK/HFdEGFjsRFYUnrnmqAonMsV9VgcvxJkCy6NNqqBng8Bs4YSoSCVcBRYPAk79o93uOB5o9xoXLyRyRq5uQLdpwI8DM8Z9WhkH2m2qdlwRYWDJxd0s+GRsOdFG62itJ2MfqGcaVixkNI+QlhNji4mxBUXCiVFRrZpJlKvo/cDpTYBnaW3ehtAe9j4jIipMYb2APt8CK0cDcRcy90vLioQVub8IcVhzAQ7V/TcmHv+34bgKJ9ZaTkbeXxcv31dHfX09IRnbT12z2q1DpFuR3wFBTYFKYfY+EyIqKulpwOnNWoGt1KxIN1ABtazYcv1mYMkhrMhkaFnfGGOkGNU1BGW83VW9Sbu6fniwYaDaf+TSTXSd+bfp+IqlM2tO5N/G1cqqlhoih5CeDmyaCYQ9BFSoZe+zISInxHlY7oK0ikjLirUUZ9z33qojmccbDKbAUsuvFF7uVBv3mHXrsOWEHFJSPLD0v8ChZcDehdpKy7L2CBGRnTCwZCE1K+bdQDlpWNUX7er4Wcw94u7mqpbmJnJo108BC/sD0VGAqwfQehjDChHZHQNLFnldyOmZNjXwUKOgQj8foiJ18m/gx0HA7WtAKX+g73xtjREiIjtjYMkirzUmrEUhpyKlbDu+AP4YDRjSgMqNgCcWAL4M5USkD5nLKpMiQ5dV7UkO98t+uV+OI3Ia6anAvh+1sBL+OPDMSoYVItIVBpYsZJ4VGbossoYW4225nwv7kVORlZWl+6f7e8AjnwMeJex9RkREFhhYrJB5Vj4b0AQBZpO3Cbkt+3Oah4XIoVzcC2z6OPN2mUpAxH84cy0R6RJrWHIgoeT+sIB8z3RLpGtRPwNLXwRSbwPlgot8xkoiIlsxsORCwon5sGUip5gMbv3bwD/Ttdu1OwM12tn7rIiI7oiBhai4SIwDfnkeOPqHdrv1y0DnN4t08TIiovxiYLHT+glERerqceCHfsCVI4CbF9DrE6BhX3ufFRFRnjGw5OTgshxWqJzG/n5yzAJbCStlKgNPfK8tYEhE5EAYWHIKKz8ONFs9KEPcRW2/LLfN0EKOpP4jQGIsENIdKBNg77MhIrIZhzVb6waSlpXclj9cOUY7jkivUhKBVW9oIduo2WCGFSJyWAwsWUnNink3UDYGIO68dhyRHt28BMzrAWyZBSx+Wpt2n4jIwbFLKCspsM2LnwZrw0FDHgDCHyvssyLKm/O7tJWWb14EvMsCHcZwIjgicgpsYclKRgPlRcJlbfKtUxstm+GXvQzs/ForckxLKbTTJMpm7yLgq+5aWPELBZ5bB9TqaO+zIiIqEGxhyUqGLstoINX3b60p3UWrA5BhoRf3AEHNMu+KjgIiv9E24e4NBDQAgpoAgU2A4DaAb5UieylUTEg91ZpJwOZPtNt1uwOPzAW8fex9ZkREBYYtLFnJPCsydDm35Q9lgbg69wPtRll+gi1ZHmg7EqjRHvDyBVITgXPbgW1zgCXPA1G/ZB4bHwMcWm5ZFEmUHym3gWNrtK/vfQ14YgHDChE5HbawWCNDlmXostV5WKbmPKS5fE2g86TMKdCvndBqCi5Eav9WbZF57IkNwC/Pal/L3BjSAiMtMao1pjFQolxhvkJyJl6lgX4LtG7Iex6299kQERUKF4PB8YcQxMXFwdfXF7GxsfDx8XGMmW73/wRs/BCIOQgY0rPf3/9noE5n7WuZP8PNE/AoUTDfmxzf0T+B66eAiOftfSZEREVy/WYLS24knNS4t3CeW0YWyZacAFzcZ9kSIxeiSmGZx26dA2yYpu1TLTFNtZYYv3qAG3+ExYp8vtg0E1gzWRv9ExAOVG9l77MiIip0vNrZm2cp7YJjftG5dU2rhzG6+i9gSAMu7dc2U1FvCaByQ637qkweRzeRY9eqLBsG7F+s3W4yiFPsE1GxwcCiR+ZhRciID6mNOR+Z2RJzYQ+QFAdc2geUrJB57B+jgStHzWpimnJ2U2cQex5Y+KQ2Ms3VHeg+DWieUQNFRFQMMLA4Amn6l+HQshkLfqWoV1pebpy27BY6vl5b5O74usx9ZQK18FKlOdBmOCcSczRntgGLBgAJMUCJ8lqLWmF1VRIR6RQDi6NydQX86mpb1tYYUz3MbuDyIeDmBeDwBS3gtB2Reexf0wBvXy3MSC0Ei3r1SVrRJKxUqq8NWS5X3d5nRERU5BhYnE1gI23DEO12Urx2wZMQYx5I0lK1UUqpt7Xb0s3gH5bZjVQ1AvAL0ceIquJOun7cPID6j2lDmImIiiEOay6uZHTSltlakJFNlhowJ2sk9ftB+1p+RQ7+ClRuAJSrkb1L6eCyHOasmZbznDWUMym6Xj0R6PIW5+MhIqcWx2HNlKfRSe1fzwwksecyh1VLcW9w28xjJYgsHqR9LRdQY0Gv/HvrirZ+UtZlDGQG3x8HavUWDC15F30QWNhPG9qeeAPoO9/eZ0REpAsMLKS1mJStqm1hD2W///Y1rZtIhlTfvg4cX6ttuZIA4wL88TpQswPgVYbFvnfqPjv8O/DL80ByPFC2OtBhrL3PlIhIN9glRHmXmqwt8Ggs6D35NxB7Jo8PdgE8SwPdpgBNntJ2SQBa97a2X2ozPI1bKe12tVaAfz3t2ORbWiuQ6bhSjlkjk1P3mbxWWf1b1GgHPP5N9uHtREROhl1CVDjcPTPXO2qesbzAzxnFvXdkAJJvWoYMuWgfXZnzQ2SRSWNgkflHvu5ueb9HyczwIqOfmj6t7b9xBtjwnlkQKpXxdRntaykurlArs/hYzsuzTOHPGixhRbrJsnWfXcgMKy3+A3R9RyuyJSIiEwYWyj/pzsiL/j8BAQ20rg7zSe4kOPT6RBvJJPcl3dSKgdXX8UD5jFAh0lK0IdiyX2b9FSm3tC0hYxZY8wCw+7ucz6fjG5n1OzJnzWetta/dvTPDjTHsNHoyMwhJMezmTyxbg8y/VnPlBGnHSsOleReYdANJy0rWsGLOu6zWAuWILUdERIWMgYXyT2ovpDtDCmytXohdtPtrdcq4CGcJOFIz00RaHPKgZntgzBktCKQmasFGBRwJOwmAb9XMY+V7dhqfcUzG/dKKYvxa6kOM5LaRPK9st65m7pNzN7p5Cdg4I+dzbPWS1joipPvq48YZ3VtltH3m3UDWSJGt1LZwUjgiomwYWCj/JITI0GXVzeGSJbRktC50m1qwLQbSaiHzychWqqL1Y8pWA9qNytvzVW0BjL+cEXziswecimYT83n7ABFDzY7LONb4tXmLk+xPT9FCiGx5JYW4RESUDYtuqZAKSYO0sFJchzRLbUz8pcxAc2YL8Ocbd37coOVsYSGiYiOORbdUpCSUhPbgTLfmpIBXalqMZPbhrbPv3H0m7xsREWXDwEIFQ8IJWwb01X1GROREXO19AkTFqiVKZv71qWy5X1pWOCMwEVGu2MJCVJTYfUZElC8MLERFjd1nREQ2Y5cQERER6R4DCxEREekeAwsRERHpHgMLERER6R4DCxEREekeAwsRERHpHgMLERER6R4DCxEREekeAwsRERHpHgMLERER6R4DCxEREekeAwsRERHpHgMLERER6R4DCxERETlnYJk9ezaCg4Ph7e2NiIgIbN++PdfjFy9ejNDQUHV8eHg4VqxYYXG/wWDAxIkTUblyZZQoUQKdO3fGsWPH8nNqRERE5IRsDiyLFi3CyJEjMWnSJERGRqJhw4bo2rUrYmJirB6/efNm9OvXD0OGDMHu3bvRu3dvtUVFRZmOee+99/Dxxx9jzpw52LZtG0qVKqWeMzEx8e5eHRERETkFF4M0b9hAWlSaN2+OWbNmqdvp6emoWrUqhg0bhjFjxmQ7vm/fvkhISMDy5ctN+1q2bIlGjRqpgCLfPjAwEK+++ipee+01dX9sbCwqVaqEefPm4YknnrjjOcXFxcHX11c9zsfHx5aXQ0RERHZiy/Xb3ZYnTk5Oxq5duzB27FjTPldXV9WFs2XLFquPkf3SImNOWk+WLl2qvj558iQuXbqknsNITl6CkTzWWmBJSkpSm5G8UOMLJyIiIsdgvG7npe3EpsBy5coVpKWlqdYPc3L78OHDVh8jYcTa8bLfeL9xX07HZDVlyhRMnjw5235p6SEiIiLHcvPmTdVYUWCBRS+khce81Ua6pa5du4YKFSrAxcWlwNOfBKGzZ8+yu+kO+F7lHd+rvON7ZRu+X3nH98r+75W0rEhYkdKQO7EpsFSsWBFubm6Ijo622C+3AwICrD5G9ud2vPFf2SejhMyPkToXa7y8vNRmrmzZsihM8gPiL3Te8L3KO75Xecf3yjZ8v/KO75V936s7tazka5SQp6cnmjZtirVr11q0bsjtVq1aWX2M7Dc/Xqxevdp0fI0aNVRoMT9GkpyMFsrpOYmIiKh4sblLSLpiBg0ahGbNmqFFixaYOXOmGgU0ePBgdf/AgQMRFBSk6kzE8OHD0b59e0yfPh09evTAwoULsXPnTsydO1fdL104I0aMwNtvv406deqoADNhwgTVPCTDn4mIiIhsDiwyTPny5ctqojcpipVum5UrV5qKZs+cOaNGDhm1bt0aCxYswPjx4zFu3DgVSmSEUP369U3HvP766yr0PP/887hx4wbatm2rnlMmmrM36XqSOWeydkFRdnyv8o7vVd7xvbIN36+843vlWO+VzfOwEBERERU1riVEREREusfAQkRERLrHwEJERES6x8BCREREusfAkgMZli2LPJYpUwb+/v5qiPWRI0fsfVq69Nlnn6FBgwamCYVk/pw//vjD3qflEKZOnWoa2k+W3nzzTfXemG+hoaH2Pi3dOn/+PAYMGKBm/C5RogTCw8PVFBKUXXBwcLbfLdlefPFFe5+arqSlpalpRmS6EfmdqlWrFt566608rftTGBxyav6isGHDBvXLK6ElNTVVDcnu0qULDh48iFKlStn79HSlSpUq6sIrQ9blF/mbb77BQw89hN27d+Oee+6x9+np1o4dO/B///d/KuyRdfL7s2bNGtNtd3f+ybLm+vXraNOmDTp27Kg+LPj5+eHYsWMoV66cvU9Nt//35GJsFBUVhfvvvx+PP/64Xc9Lb6ZNm6Y+kMrfdPm/KAFY5lyTmWlffvnlIj8fDmvOI5l7RlpaJMi0a9fO3qeje+XLl8f777+PIUOG2PtUdCk+Ph5NmjTBp59+qiZNlPmMZBJGsmxhkTmb9uzZY+9T0b0xY8Zg06ZN+Oeff+x9Kg5JWjiXL1+uQl5Br0fnyHr27KnmWPvyyy9N+x599FHV2jJ//vwiPx92CeVRbGys6UJMOZNPLTKbsUwEyKUVciatdzLzc+fOne19KromFxCZ9bpmzZro37+/mpiSslu2bJmafVxaCOSDVePGjfH555/b+7QcQnJysrr4PvPMMwwrWcjEr7JsztGjR9XtvXv3YuPGjejevTvsge2reSDrJUkClyZX8xl6KdP+/ftVQElMTETp0qWxZMkShIWF2fu0dEkCXWRkpGqWppxFRERg3rx5CAkJwcWLFzF58mTce++9qvleasso04kTJ1TTvSydIt3X8rslTfay/psspUI5k1Y8mWH96aeftvep6LLlLi4uTtWOycLH8oH0nXfeUR8e7EK6hCh3Q4cONVSvXt1w9uxZe5+KbiUlJRmOHTtm2Llzp2HMmDGGihUrGg4cOGDv09KdM2fOGPz9/Q179+417Wvfvr1h+PDhdj0vR3D9+nWDj4+P4YsvvrD3qeiOh4eHoVWrVhb7hg0bZmjZsqXdzslRdOnSxdCzZ097n4Yu/fDDD4YqVaqof/ft22f49ttvDeXLlzfMmzfPLufDwHIHL774ovqBnThxwt6n4lDuu+8+w/PPP2/v09CdJUuWSM2Ywc3NzbTJbRcXF/V1amqqvU9R15o1a6YCMVmqVq2aYciQIRb7Pv30U0NgYKDdzskRnDp1yuDq6mpYunSpvU9Fl6pUqWKYNWuWxb633nrLEBISYpfzYZdQDiTMDRs2THVt/PXXX2pYF9nWjZaUlGTv09Cd++67T3WfmZOqe2lyHT16tGp2pZwLlY8fP46nnnrK3qeiO9JdnXXaBak7qF69ut3OyRF8/fXXquZH6skou1u3blksZizkb5T8fbcHBpZciiJllelff/1V9ZfLytRChnNJhTRlGjt2rCrCqlatGm7evKneNwl5q1atsvep6Y78LmWtg5Jh8jJ3BuujLL322mt48MEH1UX3woULaqVY+WPZr18/e5+a7rzyyiuqQPLdd99Fnz59sH37dsydO1dtZJ1cdCWwSI0Ph8tbJ///pGZF/rbLsGaZqmLGjBmqQNku7NKu4wDkrbG2ff311/Y+Nd155plnVI2Pp6enwc/PT3UH/fnnn/Y+LYfBGhbr+vbta6hcubL6vQoKClK3//33X3uflm799ttvhvr16xu8vLwMoaGhhrlz59r7lHRt1apV6m/6kSNH7H0quhUXF6f+NkmXo7e3t6FmzZqGN954Q9Us2gPnYSEiIiLd4zwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLERERKR7DCxERESkewwsREREpHsMLESkWx06dFArpRMRMbAQERGR7jGwEBERke4xsBCRw/j999/VAqTff/+9vU+FiIoYl6gkIocgq4APHTpU/duzZ097nw4RFTG2sBCR7s2ePRsvvPACfvvtN4YVomKKLSxEpGs//fQTYmJisGnTJjRv3tzep0NEdsIWFiLStcaNG8PPzw9fffUVDAaDvU+HiOyEgYWIdK1WrVpYv349fv31VwwbNszep0NEdsIuISLSvbp166rQIhPJubu7Y+bMmfY+JSIqYgwsROQQQkJCsG7dOhVa3NzcMH36dHufEhEVIRcDO4WJiIhI51jDQkRERLrHwEJERES6x8BCREREusfAQkRERLrHwEJERES6x8BCREREusfAQkRERLrHwEJERES6x8BCREREusfAQkRERLrHwEJERES6x8BCRERE0Lv/B10zBFQ+qmX3AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sklearn import metrics\n",
    "\n",
    "ARI = []\n",
    "AMI = []\n",
    "x = range(2, 9)\n",
    "for k in x:\n",
    "    ARI.append(metrics.adjusted_rand_score(overlap['Verhaak_subtype'], overlap[f'k = {k}']))\n",
    "    AMI.append(metrics.adjusted_mutual_info_score(overlap[f'k = {k}'], overlap['Verhaak_subtype']))\n",
    "    \n",
    "plt.plot(x, ARI, 'o--', label = 'Adjusted Rand Index')\n",
    "plt.plot(x, AMI, 'o--', label = 'Adjusted Mutual Information')\n",
    "plt.ylim(0, 0.1)\n",
    "plt.xticks(x)\n",
    "plt.xlabel('k')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb12b446-42c1-4e21-8f76-5b96a2449d8c",
   "metadata": {},
   "source": [
    "## Plot overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "859d1a3b-94a3-4068-bdb4-198e42a6aa6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Verhaak_subtype</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESENCHYMAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CLASSICAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLASSICAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEURAL</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESENCHYMAL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Verhaak_subtype  k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8  count\n",
       "0     MESENCHYMAL      2      2      2      3      3      3      3     23\n",
       "1       CLASSICAL      2      2      2      2      2      2      2     19\n",
       "2       CLASSICAL      2      2      2      3      3      3      3     14\n",
       "3          NEURAL      2      2      2      3      3      3      3     10\n",
       "4     MESENCHYMAL      1      1      4      5      5      4      6      9"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = overlap.iloc[:,6:].value_counts().reset_index()\n",
    "counts.columns = counts.columns[:-1].tolist() + [\"count\"]\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "495fc472-b0a0-4a74-85f6-1f628b38a2cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error importing in API mode: ImportError(\"dlopen(/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/_rinterface_cffi_api.abi3.so, 0x0002): Library not loaded: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.dylib\\n  Referenced from: <3D361A20-57F1-3C9B-9D2B-6A0F70CF5F35> /Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/_rinterface_cffi_api.abi3.so\\n  Reason: tried: '/Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.dylib' (no such file), '/Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.dylib' (no such file)\")\n",
      "Trying to import in ABI mode.\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/rpy2/rinterface/__init__.py:1211: UserWarning: Environment variable \"R_SESSION_TMPDIR\" redefined by R and overriding existing variable. Current: \"/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T//Rtmp4yeeZC\", R: \"/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T//RtmpWHapsQ\"\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "%reload_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "06431653-b2a7-41a1-b098-59784043306a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: ggalluvial\n",
       "Loading required package: ggplot2\n",
       "Want to understand how all the pieces fit together? Read R for Data\n",
       "Science: https://r4ds.hadley.nz/\n",
       "In addition: Warning messages:\n",
       "1: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n",
       "2: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n",
       "3: In to_lodes_form(data = data, axes = axis_ind, discern = params$discern) :\n",
       "  Some strata appear at multiple axes.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i counts -w 800 -h 600\n",
    "\n",
    "require(ggalluvial)\n",
    "\n",
    "cluster_alluvial <- ggplot(counts, aes(y = count,\n",
    "                   axis1 = Verhaak_subtype,\n",
    "                   axis2 = `k = 2`,\n",
    "                   axis3 = `k = 3`,\n",
    "                   axis4 = `k = 4`,\n",
    "                   axis5 = `k = 5`,\n",
    "                   axis6 = `k = 6`,\n",
    "                   axis7 = `k = 7`,\n",
    "                   axis8 = `k = 8`)) +\n",
    "    geom_alluvium(aes(fill = Verhaak_subtype), width = 1/12) +\n",
    "    geom_stratum(width = 1/6, fill = 'grey', color = 'black') +\n",
    "    geom_label(stat = 'stratum', aes(label = after_stat(stratum))) +\n",
    "    #scale_x_discrete(limits = c('Verhaak subtype', 'k = 2', 'k = 3', 'k = 4', 'k = 5'), expand = c(0.1, 0.05)) +\n",
    "    scale_x_discrete(limits = c('Verhaak subtype', 'k = 2', 'k = 3', 'k = 4', 'k = 5', 'k = 6', 'k = 7', 'k = 8'), expand = c(0.1, 0.05)) +\n",
    "    scale_fill_brewer(type = 'qual', palette = 'Dark2') +\n",
    "    theme_classic(base_size = 16)\n",
    "\n",
    "pdf(\"cluster_alluvial.pdf\", width=15, height=12)\n",
    "print(cluster_alluvial)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "696cc8c9-0f40-49a2-9403-aede0bf85ce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading required package: ggpubr\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i overlap -w 1200\n",
    "\n",
    "library(\"tidyr\")\n",
    "require(ggpubr)\n",
    "\n",
    "png(filename=\"cluster_ssgsea_scores.png\", width=1200, height=600)\n",
    "\n",
    "overlap_long <- overlap %>%\n",
    "    pivot_longer(cols = starts_with('VERHAAK_GLIOBLASTOMA'), names_to = 'subtype', values_to = 'score')\n",
    "\n",
    "#p2 <- ggboxplot(overlap_long, x = 'k = 3', y = 'score', facet.by = 'subtype') +\n",
    "#    stat_compare_means()\n",
    "    \n",
    "p3 <- ggboxplot(overlap_long, x = 'k = 5', y = 'score', facet.by = 'subtype') +\n",
    "    stat_compare_means()\n",
    "\n",
    "p4 <- ggboxplot(overlap_long, x = 'k = 6', y = 'score', facet.by = 'subtype') +\n",
    "    stat_compare_means()\n",
    "\n",
    "ggarrange(p3, p4, ncol=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a3aa41-ce0b-4314-af30-2a2becbed4d7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Assess enrichment of variants within clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2c6829-a140-4507-b30b-da944c6f6e38",
   "metadata": {},
   "source": [
    "## Statistically (Fisher Exact test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d33b9306-3870-444a-9ab4-e8edd47b3a17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify the clustering in which to test enrichments\n",
    "k = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8cd4f618-e0c3-4796-b421-7deb5fe970c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ABL1_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>AKT2_mut</th>\n",
       "      <th>AKT3_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF217_del</th>\n",
       "      <th>ZRSR2_amp</th>\n",
       "      <th>ZRSR2_del</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 659 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ABL1_mut  AKT1_mut  AKT2_mut  AKT3_mut  ALK_mut  APC_mut  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0       0.0      0.0      0.0   \n",
       "\n",
       "                 ARAF_mut  ARID1A_mut  ARID1B_mut  ARID2_mut  ...  ZNF217_del  \\\n",
       "sample                                                        ...               \n",
       "GLSS-19-0266-R1       0.0         0.0         0.0        0.0  ...         0.0   \n",
       "\n",
       "                 ZRSR2_amp  ZRSR2_del  k = 2  k = 3  k = 4  k = 5  k = 6  \\\n",
       "sample                                                                     \n",
       "GLSS-19-0266-R1        0.0        0.0      1      1      1      1      1   \n",
       "\n",
       "                 k = 7  k = 8  \n",
       "sample                         \n",
       "GLSS-19-0266-R1      1      1  \n",
       "\n",
       "[1 rows x 659 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = matrix.T.join(memberships)\n",
    "df.to_csv('../results/variants_with_memberships_v2.tsv'), sep = '\\t')\n",
    "df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9bec0429-8896-40d4-902c-abbcafd7ea58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>cluster</th>\n",
       "      <th>p_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1071</th>\n",
       "      <td>JAK3_amp</td>\n",
       "      <td>20.558824</td>\n",
       "      <td>1.768781e-27</td>\n",
       "      <td>2</td>\n",
       "      <td>6.919471e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>BRD4_amp</td>\n",
       "      <td>19.644444</td>\n",
       "      <td>6.445656e-27</td>\n",
       "      <td>2</td>\n",
       "      <td>1.260770e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>MEF2B_amp</td>\n",
       "      <td>17.406495</td>\n",
       "      <td>2.649448e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.454881e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>CRTC1_amp</td>\n",
       "      <td>17.302222</td>\n",
       "      <td>3.689232e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.608069e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1242</th>\n",
       "      <td>SMARCA4_amp</td>\n",
       "      <td>16.911765</td>\n",
       "      <td>9.990847e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>7.816838e-22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant  odds_ratio       p_value  cluster         p_adj\n",
       "1071     JAK3_amp   20.558824  1.768781e-27        2  6.919471e-24\n",
       "890      BRD4_amp   19.644444  6.445656e-27        2  1.260770e-23\n",
       "1101    MEF2B_amp   17.406495  2.649448e-25        2  3.454881e-22\n",
       "946     CRTC1_amp   17.302222  3.689232e-25        2  3.608069e-22\n",
       "1242  SMARCA4_amp   16.911765  9.990847e-25        2  7.816838e-22"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_column = f'k = {k}'\n",
    "results = []\n",
    "\n",
    "for cluster in range(1, k + 1):\n",
    "    for variant in df:\n",
    "        if variant.startswith('k ='):\n",
    "            continue\n",
    "\n",
    "        # contingency table\n",
    "        cluster_with_variant = len(df[(df[cluster_column] == cluster) & (df[variant] != 0)])\n",
    "        cluster_without_variant = len(df[(df[cluster_column] == cluster) & (df[variant] == 0)])\n",
    "        non_cluster_with_variant = len(df[(df[cluster_column] != cluster) & (df[variant] != 0)])\n",
    "        non_cluster_without_variant = len(df[(df[cluster_column] != cluster) & (df[variant] == 0)])\n",
    "\n",
    "        contingency_table = [\n",
    "            [cluster_with_variant, cluster_without_variant],\n",
    "            [non_cluster_with_variant, non_cluster_without_variant]\n",
    "        ]\n",
    "\n",
    "        # Fisher's exact test\n",
    "        oddsratio, pvalue = stats.fisher_exact(contingency_table)\n",
    "\n",
    "        results.append({\n",
    "            'variant': variant,\n",
    "            'odds_ratio': oddsratio,\n",
    "            'p_value': pvalue,\n",
    "            'cluster': cluster\n",
    "        })\n",
    "\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df['p_adj'] = multipletests(results_df['p_value'], alpha = 0.1, method = 'fdr_bh')[1]\n",
    "results_df = results_df.sort_values('p_adj')\n",
    "\n",
    "results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc12ea7e-d774-4203-8157-5ef2b49f75fd",
   "metadata": {},
   "source": [
    "Get chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1219951f-2791-458d-aa69-4bb5a0ea8c1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5_2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5_2</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485.5_5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2_2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020.3_4</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52473</td>\n",
       "      <td>53312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id         gene chrom  start    end\n",
       "0  ENSG00000223972.5_2      DDX11L1  chr1  11869  14409\n",
       "1  ENSG00000227232.5_2       WASH7P  chr1  14404  29570\n",
       "2  ENSG00000243485.5_5  MIR1302-2HG  chr1  29554  31109\n",
       "3  ENSG00000237613.2_2      FAM138A  chr1  34554  36081\n",
       "4  ENSG00000268020.3_4       OR4G4P  chr1  52473  53312"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed = pd.read_csv('gs://fc-secure-3f0f305e-f110-432b-bc9a-c8d96824cea2/cmesser/local_files/data/2023-08-15_gencode-bed/example_bed_gencode30.v30lift37.annotation.bed',\n",
    "                  sep = '\\t')\n",
    "bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3c1bdae4-c0d5-4962-a5c6-05eb4816ecc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>cluster</th>\n",
       "      <th>p_adj</th>\n",
       "      <th>gene</th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JAK3_amp</td>\n",
       "      <td>20.558824</td>\n",
       "      <td>1.768781e-27</td>\n",
       "      <td>2</td>\n",
       "      <td>6.919471e-24</td>\n",
       "      <td>JAK3</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRD4_amp</td>\n",
       "      <td>19.644444</td>\n",
       "      <td>6.445656e-27</td>\n",
       "      <td>2</td>\n",
       "      <td>1.260770e-23</td>\n",
       "      <td>BRD4</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEF2B_amp</td>\n",
       "      <td>17.406495</td>\n",
       "      <td>2.649448e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.454881e-22</td>\n",
       "      <td>MEF2B</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRTC1_amp</td>\n",
       "      <td>17.302222</td>\n",
       "      <td>3.689232e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>3.608069e-22</td>\n",
       "      <td>CRTC1</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SMARCA4_amp</td>\n",
       "      <td>16.911765</td>\n",
       "      <td>9.990847e-25</td>\n",
       "      <td>2</td>\n",
       "      <td>7.816838e-22</td>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3919</th>\n",
       "      <td>JAK3_amp</td>\n",
       "      <td>0.888344</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>JAK3</td>\n",
       "      <td>amp</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3920</th>\n",
       "      <td>JAK2_del</td>\n",
       "      <td>0.888344</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>del</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3921</th>\n",
       "      <td>IGF1R_del</td>\n",
       "      <td>0.684091</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>IGF1R</td>\n",
       "      <td>del</td>\n",
       "      <td>chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3922</th>\n",
       "      <td>MECOM_del</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>MECOM</td>\n",
       "      <td>del</td>\n",
       "      <td>chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3923</th>\n",
       "      <td>ZRSR2_del</td>\n",
       "      <td>0.802083</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>del</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3924 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant  odds_ratio       p_value  cluster         p_adj     gene  \\\n",
       "0        JAK3_amp   20.558824  1.768781e-27        2  6.919471e-24     JAK3   \n",
       "1        BRD4_amp   19.644444  6.445656e-27        2  1.260770e-23     BRD4   \n",
       "2       MEF2B_amp   17.406495  2.649448e-25        2  3.454881e-22    MEF2B   \n",
       "3       CRTC1_amp   17.302222  3.689232e-25        2  3.608069e-22    CRTC1   \n",
       "4     SMARCA4_amp   16.911765  9.990847e-25        2  7.816838e-22  SMARCA4   \n",
       "...           ...         ...           ...      ...           ...      ...   \n",
       "3919     JAK3_amp    0.888344  1.000000e+00        4  1.000000e+00     JAK3   \n",
       "3920     JAK2_del    0.888344  1.000000e+00        4  1.000000e+00     JAK2   \n",
       "3921    IGF1R_del    0.684091  1.000000e+00        4  1.000000e+00    IGF1R   \n",
       "3922    MECOM_del    0.000000  1.000000e+00        4  1.000000e+00    MECOM   \n",
       "3923    ZRSR2_del    0.802083  1.000000e+00        6  1.000000e+00    ZRSR2   \n",
       "\n",
       "     type  chrom  \n",
       "0     amp  chr19  \n",
       "1     amp  chr19  \n",
       "2     amp  chr19  \n",
       "3     amp  chr19  \n",
       "4     amp  chr19  \n",
       "...   ...    ...  \n",
       "3919  amp  chr19  \n",
       "3920  del   chr9  \n",
       "3921  del  chr15  \n",
       "3922  del   chr3  \n",
       "3923  del   chrX  \n",
       "\n",
       "[3924 rows x 8 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df['gene'] = results_df['variant'].str.split('_').str[0]\n",
    "results_df['type'] = results_df['variant'].str.split('_').str[-1]\n",
    "results_df = results_df.merge(bed[['gene', 'chrom']], on = 'gene', how = 'left')\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "846d1f77-149b-4aba-b8f5-599974ebb96d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    cluster  chrom type  odds_ratio     p_adj\n",
      "31        1  chr19  del   12.176314  0.000125\n",
      "60        1   chr8  amp    9.526699  0.001363\n",
      "63        1   chr9  amp   16.660055  0.005155\n",
      "52        1   chr5  del    8.058133  0.006654\n",
      "37        1  chr20  del   11.083325  0.006927\n",
      "42        1  chr22  amp    7.183503  0.055846\n",
      "54        1   chr6  amp   15.624538  0.130625\n",
      "19        1  chr15  del    2.457044  0.140570\n",
      "9         1  chr12  amp    4.636399  0.162452\n",
      "58        1   chr7  del    6.214976  0.167237\n",
      "     cluster  chrom type  odds_ratio         p_adj\n",
      "101        2  chr19  amp   16.136399  4.180030e-19\n",
      "107        2  chr20  amp    4.136348  1.801327e-06\n",
      "87         2  chr14  del    0.115763  1.578115e-04\n",
      "84         2  chr13  del    0.311947  6.983918e-03\n",
      "90         2  chr15  del    0.133368  9.535650e-03\n",
      "128        2   chr7  amp    4.096283  9.700154e-03\n",
      "114        2  chr22  del    0.376857  1.254387e-02\n",
      "111        2  chr21  del    0.000000  3.086038e-02\n",
      "78         2  chr11  del    0.117898  3.467000e-02\n",
      "102        2  chr19  del    0.142667  5.056781e-02\n",
      "     cluster  chrom type  odds_ratio         p_adj\n",
      "172        3  chr19  amp    0.116939  6.677804e-10\n",
      "178        3  chr20  amp    0.191591  9.281713e-08\n",
      "158        3  chr14  del    4.993871  2.575379e-07\n",
      "146        3  chr10  del    3.453008  3.973870e-04\n",
      "185        3  chr22  del    2.592405  1.735585e-03\n",
      "199        3   chr7  amp    2.958923  1.903624e-03\n",
      "142        3   chr1  amp    0.172806  1.227181e-02\n",
      "205        3   chr9  amp    0.144074  1.786725e-02\n",
      "166        3  chr17  amp    0.116958  2.286494e-02\n",
      "193        3   chr5  amp    0.120777  4.184635e-02\n",
      "     cluster  chrom type  odds_ratio     p_adj\n",
      "254        4  chr21  mut   48.571429  0.007676\n",
      "257        4  chr22  mut         inf  0.007898\n",
      "245        4  chr19  mut         inf  0.011003\n",
      "236        4  chr16  mut         inf  0.019762\n",
      "281        4   chrX  mut         inf  0.030841\n",
      "248        4   chr2  mut         inf  0.038343\n",
      "260        4   chr3  mut         inf  0.082431\n",
      "275        4   chr8  mut         inf  0.086891\n",
      "215        4   chr1  mut         inf  0.111717\n",
      "272        4   chr7  mut         inf  0.116290\n",
      "     cluster  chrom type  odds_ratio     p_adj\n",
      "341        5   chr7  amp    0.146857  0.000017\n",
      "320        5  chr20  amp    5.585162  0.000166\n",
      "288        5  chr10  del    0.202841  0.000875\n",
      "306        5  chr16  del    5.477925  0.084608\n",
      "315        5  chr19  del    3.510996  0.120128\n",
      "314        5  chr19  amp    0.422621  0.216245\n",
      "309        5  chr17  del    5.098711  0.216720\n",
      "318        5   chr2  del   21.989801  0.231789\n",
      "299        5  chr14  amp    4.112079  0.237664\n",
      "296        5  chr13  amp         inf  0.283761\n",
      "     cluster  chrom type  odds_ratio     p_adj\n",
      "355        6   chr1  amp   11.467411  0.000009\n",
      "394        6  chr21  amp   10.349747  0.000027\n",
      "376        6  chr16  amp   20.203169  0.000147\n",
      "406        6   chr5  amp   11.610196  0.000295\n",
      "388        6   chr2  amp   26.875542  0.000747\n",
      "379        6  chr17  amp    7.608420  0.002963\n",
      "362        6  chr11  del    5.401250  0.004172\n",
      "400        6   chr3  amp    4.574519  0.012013\n",
      "382        6  chr18  amp    4.494738  0.020094\n",
      "403        6   chr4  amp    7.830426  0.021323\n"
     ]
    }
   ],
   "source": [
    "hallmarks = results_df.groupby(['cluster', 'chrom', 'type']).agg({'odds_ratio': 'mean', 'p_adj': 'mean'}).reset_index().sort_values(['cluster', 'p_adj'])\n",
    "\n",
    "for cluster in range(1, k + 1):\n",
    "    print(hallmarks[hallmarks['cluster'] == cluster].head(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9f7591e-811b-4658-8e38-15e6c4550319",
   "metadata": {},
   "source": [
    "## Visually (heatmap)\n",
    "Sort according to genomic coordinate so we can better visualize clustering of CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "9bb68c66-282c-40bd-8f89-b72d42168105",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_old</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>MTOR_mut</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>MTOR_amp</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>MTOR_del</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SDHB_amp</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>SDHB_del</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>BCORL1_del</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>ENSG00000085185.15_5</td>\n",
       "      <td>23</td>\n",
       "      <td>129115083</td>\n",
       "      <td>129192058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>GPC3_mut</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>ENSG00000147257.13_3</td>\n",
       "      <td>23</td>\n",
       "      <td>132669773</td>\n",
       "      <td>133119922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>GPC3_amp</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>GPC3</td>\n",
       "      <td>ENSG00000147257.13_3</td>\n",
       "      <td>23</td>\n",
       "      <td>132669773</td>\n",
       "      <td>133119922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>PHF6_del</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>ENSG00000156531.16_3</td>\n",
       "      <td>23</td>\n",
       "      <td>133507283</td>\n",
       "      <td>133562821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>PHF6_amp</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>ENSG00000156531.16_3</td>\n",
       "      <td>23</td>\n",
       "      <td>133507283</td>\n",
       "      <td>133562821</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>550 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant    gene gene_old               gene_id  chrom      start  \\\n",
       "67     MTOR_mut    MTOR     MTOR  ENSG00000198793.12_4      1   11166592   \n",
       "164    MTOR_amp    MTOR     MTOR  ENSG00000198793.12_4      1   11166592   \n",
       "248    MTOR_del    MTOR     MTOR  ENSG00000198793.12_4      1   11166592   \n",
       "149    SDHB_amp    SDHB     SDHB   ENSG00000117118.9_2      1   17345217   \n",
       "303    SDHB_del    SDHB     SDHB   ENSG00000117118.9_2      1   17345217   \n",
       "..          ...     ...      ...                   ...    ...        ...   \n",
       "535  BCORL1_del  BCORL1   BCORL1  ENSG00000085185.15_5     23  129115083   \n",
       "438    GPC3_mut    GPC3     GPC3  ENSG00000147257.13_3     23  132669773   \n",
       "469    GPC3_amp    GPC3     GPC3  ENSG00000147257.13_3     23  132669773   \n",
       "351    PHF6_del    PHF6     PHF6  ENSG00000156531.16_3     23  133507283   \n",
       "461    PHF6_amp    PHF6     PHF6  ENSG00000156531.16_3     23  133507283   \n",
       "\n",
       "           end  \n",
       "67    11322564  \n",
       "164   11322564  \n",
       "248   11322564  \n",
       "149   17380665  \n",
       "303   17380665  \n",
       "..         ...  \n",
       "535  129192058  \n",
       "438  133119922  \n",
       "469  133119922  \n",
       "351  133562821  \n",
       "461  133562821  \n",
       "\n",
       "[550 rows x 7 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_variants = results_df[results_df['p_adj'] < 0.1][['variant']].drop_duplicates()\n",
    "sig_variants['gene'] = sig_variants['variant'].str.split('_').str[0]\n",
    "\n",
    "# symbol mismatches\n",
    "sig_variants['gene_old'] = sig_variants['gene']\n",
    "sig_variants.loc[sig_variants['gene'] == 'FAM46C', 'gene'] = 'TENT5C'\n",
    "sig_variants.loc[sig_variants['gene'] == 'PARK2', 'gene'] = 'PRKN'\n",
    "\n",
    "sig_variants = sig_variants.merge(bed, on = 'gene', how = 'left')\n",
    "\n",
    "assert len(sig_variants[sig_variants['chrom'].isnull()]) == 0\n",
    "\n",
    "sig_variants['chrom'] = sig_variants['chrom'].str.replace('chr', '').replace('X', '23').astype(int)\n",
    "\n",
    "sig_variants.sort_values(['chrom', 'start'], inplace = True)\n",
    "\n",
    "sig_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "d0f9e94e-0ab2-40cc-9d6a-4aeab8531910",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "220 193 137\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    sum(sig_variants['variant'].str.contains('_amp')),\n",
    "    sum(sig_variants['variant'].str.contains('_del')),\n",
    "    sum(sig_variants['variant'].str.contains('_mut')),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd6162ea-ce20-463e-a8bb-b5a2c8c4c292",
   "metadata": {},
   "source": [
    "Combine amps and dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6907e483-01c9-48f5-b445-98b275433b26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTOR_mut</th>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <th>ID3_cnv</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>MYCL_mut</th>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <th>MPL_cnv</th>\n",
       "      <th>MUTYH_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>GPC3_mut</th>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <th>PHF6_cnv</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 380 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 MTOR_mut  MTOR_cnv  SDHB_cnv  ID3_cnv  ARID1A_mut  \\\n",
       "sample                                                               \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0266-TP       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0267-R1       0.0       0.0      -1.0     -1.0         0.0   \n",
       "GLSS-19-0267-TP       0.0      -1.0      -1.0     -1.0         0.0   \n",
       "GLSS-19-0268-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "\n",
       "                 ARID1A_cnv  MYCL_mut  MYCL_cnv  MPL_cnv  MUTYH_cnv  ...  \\\n",
       "sample                                                               ...   \n",
       "GLSS-19-0266-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0266-TP         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-TP        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0268-R1         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "\n",
       "                 GPC3_mut  GPC3_cnv  PHF6_cnv  k = 2  k = 3  k = 4  k = 5  \\\n",
       "sample                                                                      \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0      1      1      1      1   \n",
       "GLSS-19-0266-TP       0.0       0.0       0.0      2      2      2      2   \n",
       "GLSS-19-0267-R1       0.0       0.0       0.0      1      1      1      1   \n",
       "GLSS-19-0267-TP       0.0       0.0       0.0      2      2      2      3   \n",
       "GLSS-19-0268-R1       0.0       0.0       0.0      2      3      3      4   \n",
       "\n",
       "                 k = 6  k = 7  k = 8  \n",
       "sample                                \n",
       "GLSS-19-0266-R1      1      1      1  \n",
       "GLSS-19-0266-TP      2      2      2  \n",
       "GLSS-19-0267-R1      1      1      1  \n",
       "GLSS-19-0267-TP      3      3      3  \n",
       "GLSS-19-0268-R1      4      3      3  \n",
       "\n",
       "[5 rows x 380 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix_sig = matrix.T.loc[:, sig_variants['variant']]\n",
    "new_df = {}\n",
    "\n",
    "for gene in sig_variants['gene_old']:\n",
    "    if f'{gene}_mut' in matrix_sig:\n",
    "        new_df[f'{gene}_mut'] = matrix_sig[f'{gene}_mut']\n",
    "        \n",
    "    if f'{gene}_amp' in matrix_sig and f'{gene}_del' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = matrix_sig[f'{gene}_amp'] + (-1 * matrix_sig[f'{gene}_del'])\n",
    "    elif f'{gene}_del' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = -1 * matrix_sig[f'{gene}_del']\n",
    "    elif f'{gene}_amp' in matrix_sig:\n",
    "        new_df[f'{gene}_cnv'] = matrix_sig[f'{gene}_amp']\n",
    "\n",
    "matrix_sig = pd.DataFrame.from_dict(new_df)\n",
    "df_sig = matrix_sig.join(memberships)\n",
    "df_sig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2df67b8b-7005-455f-ae18-18d2a80273ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_cnv</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID3_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATRX_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STAG2_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORL1_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHF6_cnv</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chrom\n",
       "gene_cnv         \n",
       "MTOR_cnv        1\n",
       "SDHB_cnv        1\n",
       "ID3_cnv         1\n",
       "ARID1A_cnv      1\n",
       "MYCL_cnv        1\n",
       "...           ...\n",
       "ATRX_cnv       23\n",
       "STAG2_cnv      23\n",
       "BCORL1_cnv     23\n",
       "GPC3_cnv       23\n",
       "PHF6_cnv       23\n",
       "\n",
       "[236 rows x 1 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_chrom_df = sig_variants[['gene_old', 'chrom']].drop_duplicates()\n",
    "cnv_chrom_df['gene_cnv'] = cnv_chrom_df['gene_old'] + '_cnv'\n",
    "cnv_chrom_df = cnv_chrom_df[cnv_chrom_df['gene_cnv'].isin(list(df_sig))]\n",
    "cnv_chrom_df.drop(columns = 'gene_old', inplace = True)\n",
    "cnv_chrom_df.set_index('gene_cnv', inplace = True)\n",
    "cnv_chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0b1b4efb-ae12-4644-9355-41fb98a55072",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_old</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>MTOR_mut</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>MTOR_amp</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>MTOR_del</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>ENSG00000198793.12_4</td>\n",
       "      <td>1</td>\n",
       "      <td>11166592</td>\n",
       "      <td>11322564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SDHB_amp</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>SDHB_del</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>ENSG00000117118.9_2</td>\n",
       "      <td>1</td>\n",
       "      <td>17345217</td>\n",
       "      <td>17380665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant  gene gene_old               gene_id  chrom     start       end\n",
       "67   MTOR_mut  MTOR     MTOR  ENSG00000198793.12_4      1  11166592  11322564\n",
       "164  MTOR_amp  MTOR     MTOR  ENSG00000198793.12_4      1  11166592  11322564\n",
       "248  MTOR_del  MTOR     MTOR  ENSG00000198793.12_4      1  11166592  11322564\n",
       "149  SDHB_amp  SDHB     SDHB   ENSG00000117118.9_2      1  17345217  17380665\n",
       "303  SDHB_del  SDHB     SDHB   ENSG00000117118.9_2      1  17345217  17380665"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "eaf9db51-fc4b-4308-987d-02f13cd466c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('int64')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_chrom_df[\"chrom\"].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "bc78f3b3-d4ef-4415-82b3-9557792489e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: grid\n",
       "========================================\n",
       "ComplexHeatmap version 2.16.0\n",
       "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
       "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
       "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
       "\n",
       "If you use it in published research, please cite either one:\n",
       "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
       "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
       "    genomic data. Bioinformatics 2016.\n",
       "\n",
       "\n",
       "The new InteractiveComplexHeatmap package can directly export static \n",
       "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
       "\n",
       "This message can be suppressed by:\n",
       "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
       "========================================\n",
       "\n",
       "========================================\n",
       "circlize version 0.4.16\n",
       "CRAN page: https://cran.r-project.org/package=circlize\n",
       "Github page: https://github.com/jokergoo/circlize\n",
       "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
       "\n",
       "If you use it in published research, please cite:\n",
       "Gu, Z. circlize implements and enhances circular visualization\n",
       "  in R. Bioinformatics 2014.\n",
       "\n",
       "This message can be suppressed by:\n",
       "  suppressPackageStartupMessages(library(circlize))\n",
       "========================================\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i df_sig -i cnv_chrom_df -w 3000 -h 1200\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(RColorBrewer)\n",
    "\n",
    "df_sig <- df_sig[order(df_sig$`k = 6`),]\n",
    "\n",
    "muts <- as.matrix(df_sig[, grep('_mut', names(df_sig))])\n",
    "cnvs <- as.matrix(df_sig[, grep('_cnv', names(df_sig))])\n",
    "\n",
    "clusters <- df_sig[, grep('=', names(df_sig))]\n",
    "\n",
    "cluster_colors <- brewer.pal(n = 8, name = 'Dark2')\n",
    "\n",
    "chrom_colors <- rep(c(cluster_colors[6], cluster_colors[8]), length.out = 23)\n",
    "chrom_colors <- setNames(chrom_colors, 1:23)\n",
    "\n",
    "row_ha <- rowAnnotation(#Cluster_2 = clusters$`k = 2`,\n",
    "                        #Cluster_3 = clusters$`k = 3`,\n",
    "                        #Cluster_4 = clusters$`k = 4`,\n",
    "                        Cluster_5 = clusters$`k = 5`,\n",
    "                        Cluster_6 = clusters$`k = 6`,\n",
    "                        col = list(#Cluster_2 = c(`1` = cluster_colors[1], `2` = cluster_colors[2]),\n",
    "                                   #Cluster_3 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3]),\n",
    "                                   #Cluster_4 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4]),\n",
    "                                   Cluster_5 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5]),\n",
    "                                   Cluster_6 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5], `6` = cluster_colors[6])),\n",
    "                        simple_anno_size = unit(2, 'cm'))\n",
    "\n",
    "col_ha <- columnAnnotation(df = cnv_chrom_df,\n",
    "                          simple_anno_size = unit(2, 'cm'),\n",
    "                          col = list(chrom = chrom_colors))\n",
    "\n",
    "heatmap_muts <- Heatmap(muts, name = 'SNV/indels',\n",
    "                        col = colorRamp2(c(0, 2, 4), c('white', 'grey', 'black')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "heatmap_cnvs <- Heatmap(cnvs, name = 'CNVs', right_annotation = row_ha, top_annotation = col_ha,\n",
    "                        col = colorRamp2(c(-2, 0, 2), c('blue', 'white', 'red')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "full_heatmap <- heatmap_muts + heatmap_cnvs\n",
    "pdf(\"glass_heatmap.pdf\", width=50, height=20)\n",
    "draw(full_heatmap)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7771136b-2b05-43ff-8ee3-b47a0d6e6790",
   "metadata": {},
   "source": [
    "# Assess prognostic value of clusters\n",
    "\n",
    "Adjust for MGMT methylation status!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b2ccc9-3821-4c95-a07d-96ab23016954",
   "metadata": {},
   "source": [
    "## Merge cluster assignments with OS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "7e736dca-6a2d-42e6-9910-5a77fefffde4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>case_vital_status</th>\n",
       "      <th>case_overall_survival_mo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GLSS-DF-0002</td>\n",
       "      <td>dead</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GLSS-DF-0003</td>\n",
       "      <td>dead</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLSS-HK-0001</td>\n",
       "      <td>dead</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GLSS-MD-0006</td>\n",
       "      <td>dead</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GLSS-MD-0002</td>\n",
       "      <td>dead</td>\n",
       "      <td>262.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   case_barcode case_vital_status  case_overall_survival_mo\n",
       "0  GLSS-DF-0002              dead                      11.0\n",
       "1  GLSS-DF-0003              dead                       9.0\n",
       "2  GLSS-HK-0001              dead                      25.0\n",
       "3  GLSS-MD-0006              dead                      93.0\n",
       "4  GLSS-MD-0002              dead                     262.0"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_cases_df = pd.read_csv('../../../data/2022-08-10_glass/clinical_cases_table.tsv', sep = '\\t',\n",
    "                            usecols = ['case_barcode', 'case_overall_survival_mo', 'case_vital_status'])\n",
    "glass_cases_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "35638595-2eb5-4225-b094-d714ac6ee179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mgmt_methylation\n",
       "NaN    272\n",
       "U      128\n",
       "M       78\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_samples_df['mgmt_methylation'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c0db4cc5-a0bf-4f2c-9ae4-3c7b041516df",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/2411381081.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  replace({'dead': True, 'alive': False})\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>surgical_interval_mo</th>\n",
       "      <th>mgmt_methylation</th>\n",
       "      <th>case_vital_status</th>\n",
       "      <th>case_overall_survival_mo</th>\n",
       "      <th>os_harmonized</th>\n",
       "      <th>setting</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GLSS-19-0267-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GLSS-19-0271-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-19-0271</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GLSS-AT-00P1-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-AT-00P1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>37.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>GLSS-AT-00P4-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-AT-00P4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>TCGA-19-1389-R1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-19-1389</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Recurrent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>TCGA-19-1389-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-1389</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>TCGA-19-4065-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-4065</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Recurrent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>TCGA-19-4065-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-4065</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>TCGA-FG-5963-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-FG-5963</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>27.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Recurrent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_barcode  k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8  \\\n",
       "1    GLSS-19-0266-TP      2      2      2      2      2      2      2   \n",
       "3    GLSS-19-0267-TP      2      2      2      3      3      3      3   \n",
       "11   GLSS-19-0271-TP      2      2      2      2      2      2      2   \n",
       "27   GLSS-AT-00P1-TP      1      2      4      2      2      2      2   \n",
       "33   GLSS-AT-00P4-TP      2      2      2      3      3      3      3   \n",
       "..               ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "360  TCGA-19-1389-R1      2      2      2      3      3      3      3   \n",
       "361  TCGA-19-1389-TP      1      1      4      5      6      5      4   \n",
       "362  TCGA-19-4065-R1      1      1      1      1      1      5      4   \n",
       "363  TCGA-19-4065-TP      1      1      4      5      1      5      4   \n",
       "364  TCGA-FG-5963-R1      1      1      4      5      3      3      3   \n",
       "\n",
       "     case_barcode  surgical_interval_mo  mgmt_methylation  case_vital_status  \\\n",
       "1    GLSS-19-0266                   0.0               1.0               True   \n",
       "3    GLSS-19-0267                   0.0               0.0               True   \n",
       "11   GLSS-19-0271                   0.0               0.0               True   \n",
       "27   GLSS-AT-00P1                   0.0               0.0               True   \n",
       "33   GLSS-AT-00P4                   0.0               1.0               True   \n",
       "..            ...                   ...               ...                ...   \n",
       "360  TCGA-19-1389                   3.0               0.0               True   \n",
       "361  TCGA-19-1389                   0.0               0.0               True   \n",
       "362  TCGA-19-4065                   2.0               0.0               True   \n",
       "363  TCGA-19-4065                   0.0               0.0               True   \n",
       "364  TCGA-FG-5963                  14.0               0.0               True   \n",
       "\n",
       "     case_overall_survival_mo  os_harmonized          setting  \n",
       "1                        35.0           35.0  Newly diagnosed  \n",
       "3                        14.0           14.0  Newly diagnosed  \n",
       "11                       14.0           14.0  Newly diagnosed  \n",
       "27                       37.0           37.0  Newly diagnosed  \n",
       "33                       14.0           14.0  Newly diagnosed  \n",
       "..                        ...            ...              ...  \n",
       "360                       5.0            2.0        Recurrent  \n",
       "361                       5.0            5.0  Newly diagnosed  \n",
       "362                      15.0           13.0        Recurrent  \n",
       "363                      15.0           15.0  Newly diagnosed  \n",
       "364                      27.0           13.0        Recurrent  \n",
       "\n",
       "[161 rows x 15 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_with_os = memberships.\\\n",
    "    join(glass_samples_df[['case_barcode', 'surgical_interval_mo', 'mgmt_methylation']], how = 'left').\\\n",
    "    reset_index(names = 'sample_barcode').\\\n",
    "    merge(glass_cases_df, on = 'case_barcode', how = 'left').\\\n",
    "    replace({'dead': True, 'alive': False})\n",
    "\n",
    "memberships_with_os = memberships_with_os[memberships_with_os['case_overall_survival_mo'].notnull() &\n",
    "                                          memberships_with_os['surgical_interval_mo'].notnull() &\n",
    "                                          memberships_with_os['mgmt_methylation'].notnull()]\n",
    "\n",
    "memberships_with_os['os_harmonized'] = np.where(\n",
    "    memberships_with_os['sample_barcode'].str.endswith('-TP'),\n",
    "    memberships_with_os['case_overall_survival_mo'],\n",
    "    memberships_with_os['case_overall_survival_mo'] - memberships_with_os['surgical_interval_mo'],\n",
    ")\n",
    "\n",
    "memberships_with_os['setting'] = np.where(\n",
    "    memberships_with_os['sample_barcode'].str.endswith('-TP'),\n",
    "    'Newly diagnosed',\n",
    "    'Recurrent'\n",
    ")\n",
    "\n",
    "memberships_with_os['mgmt_methylation'] = np.where(\n",
    "    memberships_with_os['mgmt_methylation'] == 'M', 1.0, 0.0\n",
    ")\n",
    "\n",
    "memberships_with_os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f44dbc54-340c-49e6-a260-f8c570445e70",
   "metadata": {},
   "outputs": [],
   "source": [
    "for clust in range(1, 1+8):\n",
    "    memberships_with_os[f\"cluster_{clust}\"] = memberships_with_os[\"k = 6\"].map(lambda x: x == clust).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "bc64f1e8-5d0c-4cf7-a11c-c5a84c918b75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>k = 2</th>\n",
       "      <th>k = 3</th>\n",
       "      <th>k = 4</th>\n",
       "      <th>k = 5</th>\n",
       "      <th>k = 6</th>\n",
       "      <th>k = 7</th>\n",
       "      <th>k = 8</th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>surgical_interval_mo</th>\n",
       "      <th>...</th>\n",
       "      <th>os_harmonized</th>\n",
       "      <th>setting</th>\n",
       "      <th>cluster_1</th>\n",
       "      <th>cluster_2</th>\n",
       "      <th>cluster_3</th>\n",
       "      <th>cluster_4</th>\n",
       "      <th>cluster_5</th>\n",
       "      <th>cluster_6</th>\n",
       "      <th>cluster_7</th>\n",
       "      <th>cluster_8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GLSS-19-0266-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-19-0266</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GLSS-19-0267-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-19-0267</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GLSS-19-0271-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-19-0271</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GLSS-AT-00P1-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GLSS-AT-00P1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>GLSS-AT-00P4-TP</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>GLSS-AT-00P4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>TCGA-19-1389-R1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-19-1389</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>TCGA-19-1389-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-1389</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>TCGA-19-4065-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-4065</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>TCGA-19-4065-TP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-19-4065</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Newly diagnosed</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>TCGA-FG-5963-R1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-FG-5963</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Recurrent</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_barcode  k = 2  k = 3  k = 4  k = 5  k = 6  k = 7  k = 8  \\\n",
       "1    GLSS-19-0266-TP      2      2      2      2      2      2      2   \n",
       "3    GLSS-19-0267-TP      2      2      2      3      3      3      3   \n",
       "11   GLSS-19-0271-TP      2      2      2      2      2      2      2   \n",
       "27   GLSS-AT-00P1-TP      1      2      4      2      2      2      2   \n",
       "33   GLSS-AT-00P4-TP      2      2      2      3      3      3      3   \n",
       "..               ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "360  TCGA-19-1389-R1      2      2      2      3      3      3      3   \n",
       "361  TCGA-19-1389-TP      1      1      4      5      6      5      4   \n",
       "362  TCGA-19-4065-R1      1      1      1      1      1      5      4   \n",
       "363  TCGA-19-4065-TP      1      1      4      5      1      5      4   \n",
       "364  TCGA-FG-5963-R1      1      1      4      5      3      3      3   \n",
       "\n",
       "     case_barcode  surgical_interval_mo  ...  os_harmonized          setting  \\\n",
       "1    GLSS-19-0266                   0.0  ...           35.0  Newly diagnosed   \n",
       "3    GLSS-19-0267                   0.0  ...           14.0  Newly diagnosed   \n",
       "11   GLSS-19-0271                   0.0  ...           14.0  Newly diagnosed   \n",
       "27   GLSS-AT-00P1                   0.0  ...           37.0  Newly diagnosed   \n",
       "33   GLSS-AT-00P4                   0.0  ...           14.0  Newly diagnosed   \n",
       "..            ...                   ...  ...            ...              ...   \n",
       "360  TCGA-19-1389                   3.0  ...            2.0        Recurrent   \n",
       "361  TCGA-19-1389                   0.0  ...            5.0  Newly diagnosed   \n",
       "362  TCGA-19-4065                   2.0  ...           13.0        Recurrent   \n",
       "363  TCGA-19-4065                   0.0  ...           15.0  Newly diagnosed   \n",
       "364  TCGA-FG-5963                  14.0  ...           13.0        Recurrent   \n",
       "\n",
       "     cluster_1  cluster_2 cluster_3  cluster_4  cluster_5  cluster_6  \\\n",
       "1            0          1         0          0          0          0   \n",
       "3            0          0         1          0          0          0   \n",
       "11           0          1         0          0          0          0   \n",
       "27           0          1         0          0          0          0   \n",
       "33           0          0         1          0          0          0   \n",
       "..         ...        ...       ...        ...        ...        ...   \n",
       "360          0          0         1          0          0          0   \n",
       "361          0          0         0          0          0          1   \n",
       "362          1          0         0          0          0          0   \n",
       "363          1          0         0          0          0          0   \n",
       "364          0          0         1          0          0          0   \n",
       "\n",
       "     cluster_7  cluster_8  \n",
       "1            0          0  \n",
       "3            0          0  \n",
       "11           0          0  \n",
       "27           0          0  \n",
       "33           0          0  \n",
       "..         ...        ...  \n",
       "360          0          0  \n",
       "361          0          0  \n",
       "362          0          0  \n",
       "363          0          0  \n",
       "364          0          0  \n",
       "\n",
       "[161 rows x 23 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "memberships_with_os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "da85d697-ec02-4d08-819e-819817417948",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get participant-level cluster memberships;\n",
    "# Just choose the modal value.\n",
    "pt_memberships_with_os = memberships_with_os.groupby(\"case_barcode\").agg(lambda x: x.value_counts().index[0]).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57cc77ec-e874-45f8-af63-132b1cc2c63a",
   "metadata": {},
   "source": [
    "## k = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "21ca1095-05ed-4e42-8f62-41811fc67150",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€\n",
      "âœ” dplyr     1.1.4     âœ” readr     2.1.5\n",
      "âœ” forcats   1.0.0     âœ” stringr   1.5.1\n",
      "âœ” lubridate 1.9.3     âœ” tibble    3.2.1\n",
      "âœ” purrr     1.0.2     \n",
      "â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€\n",
      "âœ– dplyr::filter() masks stats::filter()\n",
      "âœ– dplyr::lag()    masks stats::lag()\n",
      "â„¹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors\n",
      "[1] 105  23\n",
      "[1] 41 23\n",
      "[1] 64 23\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ setting + \n",
      "    mgmt_methylation + cluster, data = combined)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "settingRecurrent  0.9563    2.6020   0.2237  4.275 1.91e-05 ***\n",
      "mgmt_methylation -0.1061    0.8994   0.2218 -0.478   0.6325    \n",
      "cluster2         -0.6115    0.5425   0.3137 -1.950   0.0512 .  \n",
      "cluster3          0.1880    1.2068   0.3010  0.624   0.5324    \n",
      "cluster4         -0.2082    0.8120   0.5013 -0.415   0.6778    \n",
      "cluster5         -0.1727    0.8414   0.3383 -0.511   0.6096    \n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "settingRecurrent    2.6020     0.3843    1.6785     4.034\n",
      "mgmt_methylation    0.8994     1.1119    0.5823     1.389\n",
      "cluster2            0.5425     1.8432    0.2934     1.003\n",
      "cluster3            1.2068     0.8286    0.6689     2.177\n",
      "cluster4            0.8120     1.2315    0.3040     2.169\n",
      "cluster5            0.8414     1.1886    0.4335     1.633\n",
      "\n",
      "Concordance= 0.681  (se = 0.024 )\n",
      "Likelihood ratio test= 25.76  on 6 df,   p=2e-04\n",
      "Wald test            = 24.03  on 6 df,   p=5e-04\n",
      "Score (logrank) test = 25.11  on 6 df,   p=3e-04\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + \n",
      "    cluster, data = nd)\n",
      "\n",
      "  n= 41, number of events= 40 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "mgmt_methylation -0.3959    0.6731   0.3865 -1.024    0.306\n",
      "cluster2         -0.1832    0.8326   0.5280 -0.347    0.729\n",
      "cluster3          0.3791    1.4610   0.4861  0.780    0.435\n",
      "cluster4              NA        NA   0.0000     NA       NA\n",
      "cluster5         -0.7536    0.4707   0.7322 -1.029    0.303\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "mgmt_methylation    0.6731     1.4857    0.3155     1.436\n",
      "cluster2            0.8326     1.2011    0.2958     2.343\n",
      "cluster3            1.4610     0.6845    0.5635     3.788\n",
      "cluster4                NA         NA        NA        NA\n",
      "cluster5            0.4707     2.1246    0.1121     1.977\n",
      "\n",
      "Concordance= 0.628  (se = 0.053 )\n",
      "Likelihood ratio test= 7.97  on 4 df,   p=0.09\n",
      "Wald test            = 7.09  on 4 df,   p=0.1\n",
      "Score (logrank) test = 7.64  on 4 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + \n",
      "    cluster, data = rc)\n",
      "\n",
      "  n= 64, number of events= 63 \n",
      "\n",
      "                     coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "mgmt_methylation  0.19045   1.20979  0.30230  0.630    0.529\n",
      "cluster2         -0.60834   0.54425  0.40182 -1.514    0.130\n",
      "cluster3          0.05748   1.05917  0.38943  0.148    0.883\n",
      "cluster4         -0.31837   0.72734  0.52343 -0.608    0.543\n",
      "cluster5          0.19985   1.22122  0.42066  0.475    0.635\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "mgmt_methylation    1.2098     0.8266    0.6689     2.188\n",
      "cluster2            0.5443     1.8374    0.2476     1.196\n",
      "cluster3            1.0592     0.9441    0.4937     2.272\n",
      "cluster4            0.7273     1.3749    0.2607     2.029\n",
      "cluster5            1.2212     0.8189    0.5355     2.785\n",
      "\n",
      "Concordance= 0.55  (se = 0.047 )\n",
      "Likelihood ratio test= 5.69  on 5 df,   p=0.3\n",
      "Wald test            = 5.29  on 5 df,   p=0.4\n",
      "Score (logrank) test = 5.44  on 5 df,   p=0.4\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_1 0.3828    1.4664   0.2577 1.486    0.137\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1     1.466     0.6819     0.885      2.43\n",
      "\n",
      "Concordance= 0.526  (se = 0.022 )\n",
      "Likelihood ratio test= 2.04  on 1 df,   p=0.2\n",
      "Wald test            = 2.21  on 1 df,   p=0.1\n",
      "Score (logrank) test = 2.23  on 1 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_2 -0.4792    0.6193   0.2530 -1.894   0.0582 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2    0.6193      1.615    0.3772     1.017\n",
      "\n",
      "Concordance= 0.544  (se = 0.025 )\n",
      "Likelihood ratio test= 3.92  on 1 df,   p=0.05\n",
      "Wald test            = 3.59  on 1 df,   p=0.06\n",
      "Score (logrank) test = 3.65  on 1 df,   p=0.06\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_3 0.3187    1.3754   0.2189 1.456    0.145\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3     1.375     0.7271    0.8956     2.112\n",
      "\n",
      "Concordance= 0.514  (se = 0.026 )\n",
      "Likelihood ratio test= 2.05  on 1 df,   p=0.2\n",
      "Wald test            = 2.12  on 1 df,   p=0.1\n",
      "Score (logrank) test = 2.14  on 1 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_4 0.3754    1.4556   0.4225 0.889    0.374\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4     1.456      0.687     0.636     3.331\n",
      "\n",
      "Concordance= 0.514  (se = 0.016 )\n",
      "Likelihood ratio test= 0.71  on 1 df,   p=0.4\n",
      "Wald test            = 0.79  on 1 df,   p=0.4\n",
      "Score (logrank) test = 0.8  on 1 df,   p=0.4\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "              coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_5 -0.07423   0.92846  0.33507 -0.222    0.825\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5    0.9285      1.077    0.4814     1.791\n",
      "\n",
      "Concordance= 0.5  (se = 0.021 )\n",
      "Likelihood ratio test= 0.05  on 1 df,   p=0.8\n",
      "Wald test            = 0.05  on 1 df,   p=0.8\n",
      "Score (logrank) test = 0.05  on 1 df,   p=0.8\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i pt_memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- pt_memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 5`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "model_5_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "                          data = combined)\n",
    "model_5_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "                          data = nd)\n",
    "model_5_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5,\n",
    "                      data = pt_memberships_with_os)\n",
    "\n",
    "print(summary(model_5_combined))\n",
    "print(summary(model_5_nd))\n",
    "print(summary(model_5_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec110c60-9278-4178-81c4-04e400ae6c30",
   "metadata": {},
   "source": [
    "## k = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "1cd3eee7-6ef5-4fb5-9fb6-5e2a797597b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 105  23\n",
      "[1] 41 23\n",
      "[1] 64 23\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ setting + \n",
      "    mgmt_methylation + cluster, data = combined)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "                     coef exp(coef) se(coef)      z Pr(>|z|)    \n",
      "settingRecurrent  0.92710   2.52716  0.22275  4.162 3.15e-05 ***\n",
      "mgmt_methylation -0.03383   0.96673  0.23593 -0.143   0.8860    \n",
      "cluster2         -0.84931   0.42771  0.33879 -2.507   0.0122 *  \n",
      "cluster3          0.01776   1.01792  0.29073  0.061   0.9513    \n",
      "cluster4         -0.39478   0.67383  0.51785 -0.762   0.4459    \n",
      "cluster5         -0.45068   0.63720  0.39982 -1.127   0.2597    \n",
      "cluster6         -0.51068   0.60009  0.37985 -1.344   0.1788    \n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "settingRecurrent    2.5272     0.3957    1.6331    3.9106\n",
      "mgmt_methylation    0.9667     1.0344    0.6088    1.5351\n",
      "cluster2            0.4277     2.3380    0.2202    0.8309\n",
      "cluster3            1.0179     0.9824    0.5758    1.7996\n",
      "cluster4            0.6738     1.4841    0.2442    1.8593\n",
      "cluster5            0.6372     1.5694    0.2910    1.3951\n",
      "cluster6            0.6001     1.6664    0.2850    1.2634\n",
      "\n",
      "Concordance= 0.679  (se = 0.023 )\n",
      "Likelihood ratio test= 27.19  on 7 df,   p=3e-04\n",
      "Wald test            = 25.26  on 7 df,   p=7e-04\n",
      "Score (logrank) test = 26.53  on 7 df,   p=4e-04\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + \n",
      "    cluster, data = nd)\n",
      "\n",
      "  n= 41, number of events= 40 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "mgmt_methylation -0.3494    0.7051   0.3795 -0.921    0.357\n",
      "cluster2         -0.2853    0.7518   0.5269 -0.542    0.588\n",
      "cluster3          0.1512    1.1632   0.4845  0.312    0.755\n",
      "cluster4              NA        NA   0.0000     NA       NA\n",
      "cluster5         -0.7933    0.4524   0.7367 -1.077    0.282\n",
      "cluster6         -1.0186    0.3611   0.7215 -1.412    0.158\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "mgmt_methylation    0.7051     1.4182   0.33516     1.483\n",
      "cluster2            0.7518     1.3302   0.26768     2.111\n",
      "cluster3            1.1632     0.8597   0.45007     3.006\n",
      "cluster4                NA         NA        NA        NA\n",
      "cluster5            0.4524     2.2107   0.10675     1.917\n",
      "cluster6            0.3611     2.7692   0.08781     1.485\n",
      "\n",
      "Concordance= 0.643  (se = 0.051 )\n",
      "Likelihood ratio test= 8.63  on 5 df,   p=0.1\n",
      "Wald test            = 7.47  on 5 df,   p=0.2\n",
      "Score (logrank) test = 8.13  on 5 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + \n",
      "    cluster, data = rc)\n",
      "\n",
      "  n= 64, number of events= 63 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "mgmt_methylation  0.1586    1.1719   0.2980  0.532   0.5946  \n",
      "cluster2         -1.0158    0.3621   0.4665 -2.178   0.0294 *\n",
      "cluster3         -0.1638    0.8489   0.3700 -0.443   0.6579  \n",
      "cluster4         -0.4863    0.6149   0.5572 -0.873   0.3828  \n",
      "cluster5         -0.2347    0.7908   0.4835 -0.486   0.6273  \n",
      "cluster6         -0.2331    0.7921   0.4515 -0.516   0.6057  \n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "mgmt_methylation    1.1719     0.8533    0.6535    2.1013\n",
      "cluster2            0.3621     2.7616    0.1451    0.9034\n",
      "cluster3            0.8489     1.1780    0.4110    1.7531\n",
      "cluster4            0.6149     1.6262    0.2063    1.8328\n",
      "cluster5            0.7908     1.2646    0.3066    2.0398\n",
      "cluster6            0.7921     1.2625    0.3269    1.9191\n",
      "\n",
      "Concordance= 0.552  (se = 0.049 )\n",
      "Likelihood ratio test= 6.41  on 6 df,   p=0.4\n",
      "Wald test            = 5.51  on 6 df,   p=0.5\n",
      "Score (logrank) test = 5.77  on 6 df,   p=0.4\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_1, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_1 0.3828    1.4664   0.2577 1.486    0.137\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_1     1.466     0.6819     0.885      2.43\n",
      "\n",
      "Concordance= 0.526  (se = 0.022 )\n",
      "Likelihood ratio test= 2.04  on 1 df,   p=0.2\n",
      "Wald test            = 2.21  on 1 df,   p=0.1\n",
      "Score (logrank) test = 2.23  on 1 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_2, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)  \n",
      "cluster_2 -0.4792    0.6193   0.2530 -1.894   0.0582 .\n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_2    0.6193      1.615    0.3772     1.017\n",
      "\n",
      "Concordance= 0.544  (se = 0.025 )\n",
      "Likelihood ratio test= 3.92  on 1 df,   p=0.05\n",
      "Wald test            = 3.59  on 1 df,   p=0.06\n",
      "Score (logrank) test = 3.65  on 1 df,   p=0.06\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_3, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_3 0.3187    1.3754   0.2189 1.456    0.145\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_3     1.375     0.7271    0.8956     2.112\n",
      "\n",
      "Concordance= 0.514  (se = 0.026 )\n",
      "Likelihood ratio test= 2.05  on 1 df,   p=0.2\n",
      "Wald test            = 2.12  on 1 df,   p=0.1\n",
      "Score (logrank) test = 2.14  on 1 df,   p=0.1\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_4, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "            coef exp(coef) se(coef)     z Pr(>|z|)\n",
      "cluster_4 0.3754    1.4556   0.4225 0.889    0.374\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_4     1.456      0.687     0.636     3.331\n",
      "\n",
      "Concordance= 0.514  (se = 0.016 )\n",
      "Likelihood ratio test= 0.71  on 1 df,   p=0.4\n",
      "Wald test            = 0.79  on 1 df,   p=0.4\n",
      "Score (logrank) test = 0.8  on 1 df,   p=0.4\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_5, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "              coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_5 -0.07423   0.92846  0.33507 -0.222    0.825\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_5    0.9285      1.077    0.4814     1.791\n",
      "\n",
      "Concordance= 0.5  (se = 0.021 )\n",
      "Likelihood ratio test= 0.05  on 1 df,   p=0.8\n",
      "Wald test            = 0.05  on 1 df,   p=0.8\n",
      "Score (logrank) test = 0.05  on 1 df,   p=0.8\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(os_harmonized, case_vital_status) ~ cluster_6, \n",
      "    data = pt_memberships_with_os)\n",
      "\n",
      "  n= 105, number of events= 103 \n",
      "\n",
      "             coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "cluster_6 -0.3114    0.7324   0.2929 -1.063    0.288\n",
      "\n",
      "          exp(coef) exp(-coef) lower .95 upper .95\n",
      "cluster_6    0.7324      1.365    0.4125       1.3\n",
      "\n",
      "Concordance= 0.51  (se = 0.023 )\n",
      "Likelihood ratio test= 1.21  on 1 df,   p=0.3\n",
      "Wald test            = 1.13  on 1 df,   p=0.3\n",
      "Score (logrank) test = 1.14  on 1 df,   p=0.3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i pt_memberships_with_os\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- pt_memberships_with_os %>%\n",
    "    mutate(across(starts_with('k = '), as.factor)) %>%\n",
    "    rename(cluster = `k = 6`)\n",
    "\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "model_6_combined <- coxph(Surv(os_harmonized, case_vital_status) ~ setting + mgmt_methylation + cluster,\n",
    "                          data = combined)\n",
    "model_6_nd <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "                          data = nd)\n",
    "model_6_rc <- coxph(Surv(os_harmonized, case_vital_status) ~ mgmt_methylation + cluster,\n",
    "                          data = rc)\n",
    "\n",
    "univariate_1 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_1,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_2 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_2,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_3 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_3,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_4 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_4,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_5 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_5,\n",
    "                      data = pt_memberships_with_os)\n",
    "univariate_6 <- coxph(Surv(os_harmonized, case_vital_status) ~ cluster_6,\n",
    "                      data = pt_memberships_with_os)\n",
    "\n",
    "print(summary(model_6_combined))\n",
    "print(summary(model_6_nd))\n",
    "print(summary(model_6_rc))\n",
    "\n",
    "print(summary(univariate_1))\n",
    "print(summary(univariate_2))\n",
    "print(summary(univariate_3))\n",
    "print(summary(univariate_4))\n",
    "print(summary(univariate_5))\n",
    "print(summary(univariate_6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "521976f5-67be-4759-ae23-8c6ecf35459f",
   "metadata": {},
   "source": [
    "# Find a performant classifier\n",
    "Borrowing some code from David's train_classifier.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "bfc0a451-73ec-4040-acea-a61f80f3b601",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.neighbors import KNeighborsClassifier, NearestCentroid\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.linear_model import LogisticRegression, SGDClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.model_selection import GridSearchCV, cross_validate, cross_val_score, StratifiedKFold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fdf38170-eb92-4f90-8b7e-9b05df8b1459",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class distribution\n",
      "k = 6\n",
      "3    120\n",
      "2     86\n",
      "6     49\n",
      "1     47\n",
      "5     39\n",
      "4     24\n",
      "Name: count, dtype: int64\n",
      "k = 6\n",
      "3    0.328767\n",
      "2    0.235616\n",
      "6    0.134247\n",
      "1    0.128767\n",
      "5    0.106849\n",
      "4    0.065753\n",
      "Name: count, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "X = matrix_sig\n",
    "y = df_sig['k = 6']\n",
    "\n",
    "counts = y.value_counts()\n",
    "dist = counts / counts.sum()\n",
    "\n",
    "print('Class distribution')\n",
    "print(counts)\n",
    "print(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "afa16a03-62c9-4f9b-ba1f-ff450c56757e",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 43 candidates, totalling 215 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 43 candidates, totalling 215 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 43 candidates, totalling 215 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 43 candidates, totalling 215 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 43 candidates, totalling 215 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 43 candidates, totalling 215 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n",
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/sklearn/neighbors/_nearest_centroid.py:244: UserWarning: self.within_class_std_dev_ has at least 1 zero standard deviation.Inputs within the same classes for at least 1 feature are identical.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "knn = KNeighborsClassifier()\n",
    "cent = NearestCentroid()\n",
    "rf = RandomForestClassifier(n_estimators = 1000)\n",
    "lr = LogisticRegression(solver = 'saga', max_iter = 20000)\n",
    "sgd = SGDClassifier(max_iter = 20000)\n",
    "\n",
    "n_folds = 5\n",
    "\n",
    "p_grid = [\n",
    "    {\n",
    "        'classifier': [knn],\n",
    "        'classifier__n_neighbors': [1,2,3,4,5,6,7,8],\n",
    "        'classifier__weights': ['uniform', 'distance']\n",
    "    },\n",
    "    {\n",
    "        'classifier': [cent],\n",
    "        'classifier__metric': ['euclidean', 'manhattan']\n",
    "    },\n",
    "    {\n",
    "        'classifier': [rf],\n",
    "        'classifier__criterion': ['gini', 'entropy', 'log_loss']\n",
    "    },\n",
    "    {\n",
    "        'classifier': [lr],\n",
    "        'classifier__penalty': ['l1', 'l2'],\n",
    "        'classifier__C': [1e-1, 1, 10, 20]\n",
    "    },\n",
    "    {\n",
    "        'classifier': [sgd],\n",
    "#        'classifier__loss': ['hinge', 'log_loss'],\n",
    "        'classifier__loss': ['log_loss'],\n",
    "        'classifier__penalty': ['l1', 'l2'],\n",
    "        'classifier__alpha': [1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 10]\n",
    "    }\n",
    "]\n",
    "\n",
    "pipeline = Pipeline([('classifier', rf)])\n",
    "\n",
    "gscv = GridSearchCV(pipeline, p_grid, scoring = 'accuracy', n_jobs = 8, cv = n_folds, verbose = 1)\n",
    "\n",
    "scores = cross_validate(gscv, X, y, cv = n_folds)\n",
    "\n",
    "best_model = gscv.fit(X, y).best_estimator_.named_steps['classifier']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "aa9e38ce-b8a3-488e-b22a-ae7eaa8e8361",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'classifier': SGDClassifier(max_iter=20000),\n",
       " 'classifier__alpha': 1e-05,\n",
       " 'classifier__loss': 'log_loss',\n",
       " 'classifier__penalty': 'l1'}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gscv.best_params_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f637e700-cf21-4172-b163-f65413805e44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"â–¸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  display: none;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  overflow: visible;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"â–¾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".estimator-table summary {\n",
       "    padding: .5rem;\n",
       "    font-family: monospace;\n",
       "    cursor: pointer;\n",
       "}\n",
       "\n",
       ".estimator-table details[open] {\n",
       "    padding-left: 0.1rem;\n",
       "    padding-right: 0.1rem;\n",
       "    padding-bottom: 0.3rem;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table {\n",
       "    margin-left: auto !important;\n",
       "    margin-right: auto !important;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:nth-child(odd) {\n",
       "    background-color: #fff;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:nth-child(even) {\n",
       "    background-color: #f6f6f6;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:hover {\n",
       "    background-color: #e0e0e0;\n",
       "}\n",
       "\n",
       ".estimator-table table td {\n",
       "    border: 1px solid rgba(106, 105, 104, 0.232);\n",
       "}\n",
       "\n",
       ".user-set td {\n",
       "    color:rgb(255, 94, 0);\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".user-set td.value pre {\n",
       "    color:rgb(255, 94, 0) !important;\n",
       "    background-color: transparent !important;\n",
       "}\n",
       "\n",
       ".default td {\n",
       "    color: black;\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".user-set td i,\n",
       ".default td i {\n",
       "    color: black;\n",
       "}\n",
       "\n",
       ".copy-paste-icon {\n",
       "    background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA0NDggNTEyIj48IS0tIUZvbnQgQXdlc29tZSBGcmVlIDYuNy4yIGJ5IEBmb250YXdlc29tZSAtIGh0dHBzOi8vZm9udGF3ZXNvbWUuY29tIExpY2Vuc2UgLSBodHRwczovL2ZvbnRhd2Vzb21lLmNvbS9saWNlbnNlL2ZyZWUgQ29weXJpZ2h0IDIwMjUgRm9udGljb25zLCBJbmMuLS0+PHBhdGggZD0iTTIwOCAwTDMzMi4xIDBjMTIuNyAwIDI0LjkgNS4xIDMzLjkgMTQuMWw2Ny45IDY3LjljOSA5IDE0LjEgMjEuMiAxNC4xIDMzLjlMNDQ4IDMzNmMwIDI2LjUtMjEuNSA0OC00OCA0OGwtMTkyIDBjLTI2LjUgMC00OC0yMS41LTQ4LTQ4bDAtMjg4YzAtMjYuNSAyMS41LTQ4IDQ4LTQ4ek00OCAxMjhsODAgMCAwIDY0LTY0IDAgMCAyNTYgMTkyIDAgMC0zMiA2NCAwIDAgNDhjMCAyNi41LTIxLjUgNDgtNDggNDhMNDggNTEyYy0yNi41IDAtNDgtMjEuNS00OC00OEwwIDE3NmMwLTI2LjUgMjEuNS00OCA0OC00OHoiLz48L3N2Zz4=);\n",
       "    background-repeat: no-repeat;\n",
       "    background-size: 14px 14px;\n",
       "    background-position: 0;\n",
       "    display: inline-block;\n",
       "    width: 14px;\n",
       "    height: 14px;\n",
       "    cursor: pointer;\n",
       "}\n",
       "</style><body><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>SGDClassifier(alpha=1e-05, loss=&#x27;log_loss&#x27;, max_iter=20000, penalty=&#x27;l1&#x27;)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>SGDClassifier</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.7/modules/generated/sklearn.linear_model.SGDClassifier.html\">?<span>Documentation for SGDClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\" data-param-prefix=\"\">\n",
       "        <div class=\"estimator-table\">\n",
       "            <details>\n",
       "                <summary>Parameters</summary>\n",
       "                <table class=\"parameters-table\">\n",
       "                  <tbody>\n",
       "                    \n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('loss',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">loss&nbsp;</td>\n",
       "            <td class=\"value\">&#x27;log_loss&#x27;</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('penalty',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">penalty&nbsp;</td>\n",
       "            <td class=\"value\">&#x27;l1&#x27;</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('alpha',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">alpha&nbsp;</td>\n",
       "            <td class=\"value\">1e-05</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('l1_ratio',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">l1_ratio&nbsp;</td>\n",
       "            <td class=\"value\">0.15</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('fit_intercept',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">fit_intercept&nbsp;</td>\n",
       "            <td class=\"value\">True</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('max_iter',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">max_iter&nbsp;</td>\n",
       "            <td class=\"value\">20000</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('tol',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">tol&nbsp;</td>\n",
       "            <td class=\"value\">0.001</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('shuffle',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">shuffle&nbsp;</td>\n",
       "            <td class=\"value\">True</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('verbose',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">verbose&nbsp;</td>\n",
       "            <td class=\"value\">0</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('epsilon',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">epsilon&nbsp;</td>\n",
       "            <td class=\"value\">0.1</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('n_jobs',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">n_jobs&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('random_state',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">random_state&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('learning_rate',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">learning_rate&nbsp;</td>\n",
       "            <td class=\"value\">&#x27;optimal&#x27;</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('eta0',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">eta0&nbsp;</td>\n",
       "            <td class=\"value\">0.0</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('power_t',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">power_t&nbsp;</td>\n",
       "            <td class=\"value\">0.5</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('early_stopping',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">early_stopping&nbsp;</td>\n",
       "            <td class=\"value\">False</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('validation_fraction',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">validation_fraction&nbsp;</td>\n",
       "            <td class=\"value\">0.1</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('n_iter_no_change',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">n_iter_no_change&nbsp;</td>\n",
       "            <td class=\"value\">5</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('class_weight',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">class_weight&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('warm_start',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">warm_start&nbsp;</td>\n",
       "            <td class=\"value\">False</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('average',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">average&nbsp;</td>\n",
       "            <td class=\"value\">False</td>\n",
       "        </tr>\n",
       "    \n",
       "                  </tbody>\n",
       "                </table>\n",
       "            </details>\n",
       "        </div>\n",
       "    </div></div></div></div></div><script>function copyToClipboard(text, element) {\n",
       "    // Get the parameter prefix from the closest toggleable content\n",
       "    const toggleableContent = element.closest('.sk-toggleable__content');\n",
       "    const paramPrefix = toggleableContent ? toggleableContent.dataset.paramPrefix : '';\n",
       "    const fullParamName = paramPrefix ? `${paramPrefix}${text}` : text;\n",
       "\n",
       "    const originalStyle = element.style;\n",
       "    const computedStyle = window.getComputedStyle(element);\n",
       "    const originalWidth = computedStyle.width;\n",
       "    const originalHTML = element.innerHTML.replace('Copied!', '');\n",
       "\n",
       "    navigator.clipboard.writeText(fullParamName)\n",
       "        .then(() => {\n",
       "            element.style.width = originalWidth;\n",
       "            element.style.color = 'green';\n",
       "            element.innerHTML = \"Copied!\";\n",
       "\n",
       "            setTimeout(() => {\n",
       "                element.innerHTML = originalHTML;\n",
       "                element.style = originalStyle;\n",
       "            }, 2000);\n",
       "        })\n",
       "        .catch(err => {\n",
       "            console.error('Failed to copy:', err);\n",
       "            element.style.color = 'red';\n",
       "            element.innerHTML = \"Failed!\";\n",
       "            setTimeout(() => {\n",
       "                element.innerHTML = originalHTML;\n",
       "                element.style = originalStyle;\n",
       "            }, 2000);\n",
       "        });\n",
       "    return false;\n",
       "}\n",
       "\n",
       "document.querySelectorAll('.fa-regular.fa-copy').forEach(function(element) {\n",
       "    const toggleableContent = element.closest('.sk-toggleable__content');\n",
       "    const paramPrefix = toggleableContent ? toggleableContent.dataset.paramPrefix : '';\n",
       "    const paramName = element.parentElement.nextElementSibling.textContent.trim();\n",
       "    const fullParamName = paramPrefix ? `${paramPrefix}${paramName}` : paramName;\n",
       "\n",
       "    element.setAttribute('title', fullParamName);\n",
       "});\n",
       "</script></body>"
      ],
      "text/plain": [
       "SGDClassifier(alpha=1e-05, loss='log_loss', max_iter=20000, penalty='l1')"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "39c1ae9f-4509-485c-a172-de380eac7bf9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.7479452054794521)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cross_val_score(best_model, X, y, cv = 5).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "67a058bf-8021-4291-a4b1-75a4bf3cf939",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[   2.66465631,  125.01816781,   30.44080466, ...,    0.        ,\n",
       "           1.56365377,   51.04194943],\n",
       "       [  50.68805105,   45.89977385,   78.25022871, ...,    0.        ,\n",
       "         -68.63804931,  -67.47923114],\n",
       "       [-252.86212567, -127.96714129,   -9.27857909, ...,    0.        ,\n",
       "         -34.4760229 ,   78.70852831],\n",
       "       [ 118.89026094,   28.68021432,   37.69301906, ...,   34.18236582,\n",
       "          -4.38676664,   11.82497289],\n",
       "       [ -34.02555558,   14.73969189,  -42.84779585, ...,    0.        ,\n",
       "          67.21639477,   -0.77244131],\n",
       "       [ -85.59074891,  112.55975052,   48.72187545, ...,    0.        ,\n",
       "           1.98041582,   31.11846206]], shape=(6, 373))"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_model.coef_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d775a1a0-efa3-47fe-8894-143a6178cdb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTOR_mut</th>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <th>ID3_cnv</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>MYCL_mut</th>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <th>MPL_cnv</th>\n",
       "      <th>MUTYH_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>AR_mut</th>\n",
       "      <th>AR_cnv</th>\n",
       "      <th>ATRX_mut</th>\n",
       "      <th>ATRX_cnv</th>\n",
       "      <th>STAG2_cnv</th>\n",
       "      <th>BCORL1_mut</th>\n",
       "      <th>BCORL1_cnv</th>\n",
       "      <th>GPC3_mut</th>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <th>PHF6_cnv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0266-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0267-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-19-0268-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-1389-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-1389-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-4065-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-19-4065-TP</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-FG-5963-R1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365 rows Ã— 373 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 MTOR_mut  MTOR_cnv  SDHB_cnv  ID3_cnv  ARID1A_mut  \\\n",
       "sample                                                               \n",
       "GLSS-19-0266-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0266-TP       0.0       0.0       0.0      0.0         0.0   \n",
       "GLSS-19-0267-R1       0.0       0.0      -1.0     -1.0         0.0   \n",
       "GLSS-19-0267-TP       0.0      -1.0      -1.0     -1.0         0.0   \n",
       "GLSS-19-0268-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "...                   ...       ...       ...      ...         ...   \n",
       "TCGA-19-1389-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "TCGA-19-1389-TP       0.0       1.0       1.0      1.0         0.0   \n",
       "TCGA-19-4065-R1       0.0       0.0      -1.0     -1.0         0.0   \n",
       "TCGA-19-4065-TP       0.0       1.0       1.0      1.0         0.0   \n",
       "TCGA-FG-5963-R1       0.0       0.0       0.0      0.0         0.0   \n",
       "\n",
       "                 ARID1A_cnv  MYCL_mut  MYCL_cnv  MPL_cnv  MUTYH_cnv  ...  \\\n",
       "sample                                                               ...   \n",
       "GLSS-19-0266-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0266-TP         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0267-TP        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "GLSS-19-0268-R1         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "...                     ...       ...       ...      ...        ...  ...   \n",
       "TCGA-19-1389-R1         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "TCGA-19-1389-TP         1.0       0.0       0.0      0.0       -1.0  ...   \n",
       "TCGA-19-4065-R1        -1.0       0.0       0.0      0.0        0.0  ...   \n",
       "TCGA-19-4065-TP         1.0       0.0       1.0      1.0        1.0  ...   \n",
       "TCGA-FG-5963-R1         0.0       0.0       0.0      0.0        0.0  ...   \n",
       "\n",
       "                 AR_mut  AR_cnv  ATRX_mut  ATRX_cnv  STAG2_cnv  BCORL1_mut  \\\n",
       "sample                                                                       \n",
       "GLSS-19-0266-R1     0.0     0.0       0.0       2.0        0.0         0.0   \n",
       "GLSS-19-0266-TP     0.0     0.0       0.0       0.0        0.0         0.0   \n",
       "GLSS-19-0267-R1     0.0     0.0       0.0       2.0        0.0         0.0   \n",
       "GLSS-19-0267-TP     0.0     0.0       0.0       0.0        0.0         0.0   \n",
       "GLSS-19-0268-R1     0.0     0.0       0.0       0.0        0.0         0.0   \n",
       "...                 ...     ...       ...       ...        ...         ...   \n",
       "TCGA-19-1389-R1     0.0     1.0       0.0      -1.0        2.0         0.0   \n",
       "TCGA-19-1389-TP     0.0    -1.0       0.0       0.0        1.0         2.0   \n",
       "TCGA-19-4065-R1     0.0     1.0       0.0       0.0        1.0         0.0   \n",
       "TCGA-19-4065-TP     0.0     2.0       0.0       0.0        2.0         0.0   \n",
       "TCGA-FG-5963-R1     0.0     0.0       2.0       0.0        0.0         0.0   \n",
       "\n",
       "                 BCORL1_cnv  GPC3_mut  GPC3_cnv  PHF6_cnv  \n",
       "sample                                                     \n",
       "GLSS-19-0266-R1         0.0       0.0       0.0       0.0  \n",
       "GLSS-19-0266-TP         0.0       0.0       0.0       0.0  \n",
       "GLSS-19-0267-R1         0.0       0.0       0.0       0.0  \n",
       "GLSS-19-0267-TP         0.0       0.0       0.0       0.0  \n",
       "GLSS-19-0268-R1         0.0       0.0       0.0       0.0  \n",
       "...                     ...       ...       ...       ...  \n",
       "TCGA-19-1389-R1        -1.0       0.0       0.0       0.0  \n",
       "TCGA-19-1389-TP         1.0       0.0       1.0       1.0  \n",
       "TCGA-19-4065-R1         0.0       0.0       0.0       0.0  \n",
       "TCGA-19-4065-TP        -1.0       0.0       0.0       0.0  \n",
       "TCGA-FG-5963-R1         0.0       0.0       0.0       0.0  \n",
       "\n",
       "[365 rows x 373 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c56d69af-03b3-4b25-8e6b-e3e2bfaa62b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AXL_mut</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-203.397854</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>66.505159</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-87.524084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STAT6_cnv</th>\n",
       "      <td>37.061223</td>\n",
       "      <td>-198.797572</td>\n",
       "      <td>36.491025</td>\n",
       "      <td>34.366344</td>\n",
       "      <td>-33.664452</td>\n",
       "      <td>28.281377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORL1_cnv</th>\n",
       "      <td>-45.754198</td>\n",
       "      <td>-193.789628</td>\n",
       "      <td>1.580640</td>\n",
       "      <td>-80.178330</td>\n",
       "      <td>-9.704101</td>\n",
       "      <td>41.343732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL6_cnv</th>\n",
       "      <td>10.911213</td>\n",
       "      <td>-193.195327</td>\n",
       "      <td>-74.271486</td>\n",
       "      <td>-33.733633</td>\n",
       "      <td>-27.688955</td>\n",
       "      <td>43.611151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NFE2L2_cnv</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-176.105316</td>\n",
       "      <td>-85.139526</td>\n",
       "      <td>-16.750548</td>\n",
       "      <td>33.273700</td>\n",
       "      <td>48.421688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SH2B3_cnv</th>\n",
       "      <td>93.051086</td>\n",
       "      <td>159.752569</td>\n",
       "      <td>-34.743604</td>\n",
       "      <td>32.881120</td>\n",
       "      <td>-83.868684</td>\n",
       "      <td>-74.663942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPN11_cnv</th>\n",
       "      <td>64.317704</td>\n",
       "      <td>164.228864</td>\n",
       "      <td>64.623063</td>\n",
       "      <td>49.036652</td>\n",
       "      <td>-66.369075</td>\n",
       "      <td>-41.998549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53_cnv</th>\n",
       "      <td>-23.485257</td>\n",
       "      <td>165.087856</td>\n",
       "      <td>-125.997539</td>\n",
       "      <td>52.638229</td>\n",
       "      <td>44.124946</td>\n",
       "      <td>-16.600590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B2M_cnv</th>\n",
       "      <td>11.653804</td>\n",
       "      <td>190.906469</td>\n",
       "      <td>-84.013094</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-19.975612</td>\n",
       "      <td>39.904044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ESR1_mut</th>\n",
       "      <td>29.251744</td>\n",
       "      <td>209.330985</td>\n",
       "      <td>-114.225105</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>85.494841</td>\n",
       "      <td>-94.277061</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>373 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0           1           2          3          4          5\n",
       "AXL_mut      0.000000 -203.397854    0.000000  66.505159   0.000000 -87.524084\n",
       "STAT6_cnv   37.061223 -198.797572   36.491025  34.366344 -33.664452  28.281377\n",
       "BCORL1_cnv -45.754198 -193.789628    1.580640 -80.178330  -9.704101  41.343732\n",
       "BCL6_cnv    10.911213 -193.195327  -74.271486 -33.733633 -27.688955  43.611151\n",
       "NFE2L2_cnv   0.000000 -176.105316  -85.139526 -16.750548  33.273700  48.421688\n",
       "...               ...         ...         ...        ...        ...        ...\n",
       "SH2B3_cnv   93.051086  159.752569  -34.743604  32.881120 -83.868684 -74.663942\n",
       "PTPN11_cnv  64.317704  164.228864   64.623063  49.036652 -66.369075 -41.998549\n",
       "TP53_cnv   -23.485257  165.087856 -125.997539  52.638229  44.124946 -16.600590\n",
       "B2M_cnv     11.653804  190.906469  -84.013094   0.000000 -19.975612  39.904044\n",
       "ESR1_mut    29.251744  209.330985 -114.225105   0.000000  85.494841 -94.277061\n",
       "\n",
       "[373 rows x 6 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_model.fit(X, y)\n",
    "\n",
    "coeffs = best_model.coef_\n",
    "coeffs = pd.DataFrame(coeffs).T\n",
    "coeffs.index = [c for c in df_sig.columns if not c.startswith(\"k =\")]\n",
    "coeffs.sort_values(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83344ada-a467-4835-90ed-9edd2ce45b27",
   "metadata": {},
   "source": [
    "# Does PCA tell us anything?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c21255f2-3f4d-4edb-b45e-db4565103d0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.10331687 0.08294037 0.05099463 0.04129514 0.0346887 ]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkYAAAG1CAYAAAAcMztGAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAjBhJREFUeJzt/Ql4nGd59o2fs6/SSDOyZMmWvMWWHSexkzgJQZA4cchWBSVQBGkoJpT07feSUhq6kB4FSg/awL9v+fIVUtItIX2BgmlJBIEYiLMYRFbHzuLEjnfJ1mJpRhppNJtm+R/X/czII1kjzUgz2ub85Zg8c9+PNPOMJGtOXdd5XZcumUwmQQghhBBCoJ/vCyCEEEIIWShQGBFCCCGEpKAwIoQQQghJQWFECCGEEJKCwogQQgghJAWFESGEEEJICgojQgghhJAUFEaEEEIIISmM6TtkehKJBLq6ulBWVgadTjffl0MIIYSQHJBe1sPDw6irq4NeP3VMiMIoD0QU1dfXz/dlEEIIIWQGdHZ2YuXKlVN+DIVRHkikKP2FLS8vn+/LIYQQQkgODA0NqcBG+n18KiiM8iCdPhNRRGFECCGELC5yscHQfE0IIYQQkoLCiBBCCCEkBVNphBBCSAkQj8cxOjqKpYjJZILBYCjIY1EYEUIIIUu8VL2npweDg4NYylRUVGD58uWzbqdDYUQIIYQsYdKiqLq6Gna7fcn14UsmkwgGgzh79qxa19bWzurxKIwIIYSQJZw+S4sij8eDpYrNZlNHEUfyWmeTVqP5mhBCCFmipD1FEila6thTr3G2PioKI0IIIWSJs9TSZ8V8jRRGhBBCCCEp6DEihBBSdLxewOcD3G5gCVtdyBKAwogQQkjRCIWAXbuA9nYgEACcTqCpCWhtFcPsfF8dIefDVBohhJCiIaKorQ2QIqGGBu0oa9knZCr27t2L2267DXV1dco/9MQTT2AuoDAihBBStPSZRIpqarSb1XruvuzLebLI8HqBI0fm5Js3MjKCLVu24KGHHsJcwlQaIYSQoiCeIkmfSaQoE5cL6OjQztNvtEgIzX1O9JZbblG3uYYRI0IIIUUJDIjRWt4//f7x+7KWfTlPFgm7SicnyogRIYSQggcGZEQDEMJ73mPDk0/axyJFIop6e4GWFkaLFm1OVJC8qCD7zc1L6ptJYUQIIeS8wIC8/0lgQISMrIWdO6f/fOk6fPDgQXR2diIajaKy0ozrr6/HG29sRkeHSQktEUUitMgiwVdaOVEKI0IIIQULDIgoOnz4MBwOB1wuF8LhMJYtO4xPfEIecyv7GC1G3Bk50fQPxBLOidJjRAghS7u4J+/AgAQCMpG17Mv56dJnEikSUeR0OmE0GtVR1gMDnVixIkhRtBjxeLR8quRA5RYOn7sv+0vsm8qIESGEzCELueHhxMCA2ITketPXOV1gIBQKqfSZRIoysVqt8Pv96vx0w0xFXMnHybT0Uhh8umhoTeU+5QdX0mdzkBMNBAI4evTo2PrEiRM4cOAA3G43Giam9QoIhREhhCwiD89cBAZ+/GPg8GEtmiXiSIaVf+ADMr186s8XMWM2m1X6TCJFaWQt+3I+V2+SfHx9fT02b94Mk8lUyJdJZoLNpv2ASj51jma7vPrqq7juuuvG1vfdd5867ty5E9/5zneK9rxMpRFCyBJoeFio1JwEAKqr5a9zTRQ5HMDKlZpQ+trXpn58ifCImJHGfPLXfiwWU0dZy/5UEaC0N0k6HEvESY6yln2ygPB4gPXr5yR9tn37diSTyfNuxRRFAiNGhBCyiIt7ck3NiaA5dky7v25d9ucRMaTTAddeq33+8eOaKJJrO3AAePpp4O67gbvumjz1JxEeQSI/kj6TyE9jY+PYfi7eJCF9lP0NGzYwrUbmDAojQgiZo2nyhS7ukef4j/8Ann9eE1tyGxoK4qmnQojHbfjUp+xKOD38MPCDHwBnzwJ6/bmU2ac+BdTVjTdVDwycE2+HDgFvvqmJLxFB0Shw6hTw7W8Dkt2aLPUnaa+tW7cqMZOrV6gQ3iRCCgWFESGEFJCpIjhpQZL2FE3X8HAycZX5HHv2aM9jNsttFFVVB1Fd3Yny8igOHTLj2Wfr8f3vb8bjj5tUIZGImbS4kejPo49qabPycu1aKyqAtWuBri6tsbGk0+Tj05EhEXNVVVpUSZ57qvJ9ETK5ipnZeJMIKTQURoQQUkBEsMhNxIaIDhEWmebqXIp7pkuPPfKI9piyb7Fot3D4IIaHD6OiwgGj0YVAIIxnnjmMd98VY/NW9bFDQ1qltUSNEgkZ0gmcPKk95ooVmggTLTM8DOzfrx2TSe1jYzFNFMnjyGuSyFKh+vqlvUniKUpHikQUiTdJ0nCMFpG5hMKIEEIKxOnTWhRGIkBGoxbJqa8Hli0b3yBxuuKebJVrUh0WiQDf/CYQj2siRsRNWVkQtbXi6XHA4XAq8ePzOdHTI4/fCb1+A0ZG7LAbvVhR7YNvxA1fwKOiQuInkscUwSMCSJ7rwgs1b1G6Ik0+RkSRvA4RbSKWKisL29dvJt4kQooBhREhhBSI730PkLYrIiCkmksEigRBRFxI2iwzwiLHyaItU3Wf/u53taOIIvlcieLIY/v9Iej1UQwPu1RUSB5DzsXjVthsfjhsA7h+3Y/wvsZ2OK0BBMJO/OZwE/7nlVZE4zYldORaJcolwkgiSKtWATt2aKX7EjmSPYleyWOXlQFXXTV1tCjffkQz8SYRUgwojAghpACIYJDKchENEomRiFHaLiPRlyuu0CIs2XxD01WuSfRJIlJbtmhCRYSPPL5Ecvx+G0ZGzNDrw0gmnepjRZwNDYURCJjxgU2/wG2XPInewRqc6mtAhd2P27e1qev8z1/vVKk1EV+Dg5pA6u/XHvcP/1CLeIngk7SfXJd8joi+ffuAxx47v/pttv2IcvUmsREkKRYURoQQMos34LTQEc+NiBUxL4sQEkRsSHRHBEVjI/Dkk9OX1WerXJOKMkF6CsljSrWYmKQlQhSJ2HHkSD0uv/RN1Fb50W8og6NM2tSNoOP4Ctyy8afo9degd0gLQfUOWQEd8N717fjpa80IxT3KiySpMxFJkoLbtk1L1d1zD/ChDwH/9E9a9dvq1cDy5dkbU042Ky3tHZKI0GxhI0hSbCiMCCFkEjE0PDyMU6dOobe3F8PDUSQSZixbVo+Ghs2orjZBpwvixz8O4dVXbRgasqsIkQiVNWu06q/OTk08iIiRvkEiMnLpeJ2tck0EkIgiES/ymLInYkx8QQ5rCE11+7Gp8gDiCT3KKsyI6qoRNlyD2LATLmcAx7oblOiRiI+kzoZCLqxa1oH6ah9O+Tzq8eU1iEgTM7domb/8S+Df/k27ju5uYNOmqYfLTtaPSK93KsF45Ehh+hEVW3gRQmFECCGTRCM0QRRAJFKBs2drMTg4Cr3+MAYG4qiqMsBq7UQ8Lr13zPB46uHzbcbwsAlvvAFceaXm0ZEojwiOW28FXn8996n1k1WuyZ5Ug0lPor4+TRCJkJHjx5p2oXnLT+GP1sDrc8FdHsDyshdx1FqFl03NiMGJZRV+9AesSmjJ47gsftjLnWj9uBv/+pj2WCKY5CZma7nt3at10xZyaUyZ2Y9IvE+SWhSBODqqeZ3EC3XXXfYZz4RjI8jS4oEHHsCPf/xjHDp0SEVs3/ve9+LrX/+6MuUXE44EIYSQCdGIeDyuRFIwaILfH0A0OoDBQSeCQQfc7jcRjb6NU6d0OH3ahZERHQyGw6ipOYjLLtM8RlIpJuJF7ougueaa/KbWp8dSPfAA8LnPAR/5iBatec97tMcUD1G6WqxhuRfbN7XjjK8GA6EaxGJWJBJVcLmr0Ohuh9UGvHy8CetW9GLTml4gHka5uReranpRuaEJF2/TokXyeBLpkmiRRKQkMiVeIynnz0zvZTKxMWVmPyIRRRLIkShVeXlYXFL4xS9squJupqSFl5TzZyJr2ZfzZOnw/PPP4zOf+QxefPFF/OpXv1L/Jm+88UbVxqGYMGJECJkX5tI8O53heWI0Qq/XIxZLYmREQhsx6HR+xGIeBAIGLFs2hO7uOoRCTpWW6u7WIhZVVZ3weLSy+HvvPVfOLs8nzz/RNyTCQ6IpIkbGyt4jXiDiAyxuhBIePP64VokmH6c9hxbRkXEdv/mNJlyWV/rgtAVwqr8BJov2cfLl3HixC4ZwB7653YdnX2rF0R5glb0dF63tUJEiz8YWRGpaEe7QhJAgkSQxX8tziFCSKJJcb66NKdP9iN588zDOnJHXbIXDIaJoBEZjI6qq7JNGyHKFjSDnH28O/5YKxe7du8etZUZadXU19u3bh2vkr40iQWFECJlT5tI8m+scMe1jz6WBtGGVBiQSMUSjOlXtZbWGoNNFVZQmEHCMNT4UfD4rKiv9CARCcDrt580iyxQW4rcRL5IYtOWa5GOfejKE1qt2wexvB2IBwOjEq4eb8Mi/teJ0t01VgYlIEZ+PCCp5HRJBevlloMvrhn/EiTKLH/6IVYki8TnZjX7A4UTjxW40brPB692JwbPNiI34sPtZN/b9uydV0q9FoEToyGuSb0FmSu23vwXe//7sjSlvuy0Ir/ecwJXvowimWKwTTqeEmMwAJPWxeVYz4QQ2gpw/Qnn8WyoW0t9KcBeygdYkUBgRQuaUQphnc/2rNd0ocfnyIFatCqmy9rY27c1z4pyvidGIsrIymExd0OtjSiTU1Z1CJGLE8LATJlNcRVdEVEhUP5kMo6vLjIEB26SjPYS0sJAGkDLMVVJil1yizSrz7t+FbkMbVm2oAewNCA37Yelrw7Zq4N1jO1UUR6JTkuYSgSS+n+3btc9/4w0Pfn24CXdsa4PHCDRc4MLmdX4g3AusbAEs2sWIZnjyZQ8efdSjei3J80sFnXwNRQzJ46c7XMt9ETFyTlJi8vWe2JiyvHwU3d0H0d5+vsDdtm0rHn98A0ZHQ6iqkndN+6xmwmXCRpDzw64sTUeFyWbmFZpEIoHPfe5zaGpqwkUXXVTU56IwIoTMGbM1z+bzV6vWKHEUq1YdhMsluagoqqvNsFjq0d6+Gc3NpnECZmI0wmiMw2hMql/Io6MWJBI6FUXyet1wOEaUiEgmrbBYwrDZRnD8eCMuuMA+brRHJnJ9IipkxlhtrdYfSF6qTe/FZY52nOiuwbL1NbAbgGDCipM9wNYV7fhVeTNGYh4lxKTPkIgjuW5Zi8C44QZg1dpWXPAeYJWtHVZjB2BwAstb4HW2wndE+zhpFSBfO4nmSI8jET8StRJRJNcijy1veLKWSJGIPhFOIpQma0x54MDUAvfqq+1KhMrjTDcTLh/YCHLu8U7RdHQ2qdF8EK/RW2+9hd9IDrnIUBgRQuaM2U5Rz/WvVvlF/tpr8uZ+EE6nvFk7xBkjbhS1lnSUz7f1vF/m6ajDiRMnMDAwAJfLhmPH6jAw4EIwKOIoAr0+juPH16CurlelzyIRM5LJRqxevVkJCnnsbKkFERgiNKRiLf3GYjP4UGYN4JS3QQk/e+pzfcMueGwdqHT40NftUVEj8f3I48vn//Vfa4+hRc3kk3YCkWblUQol3NjV5hkTkBJlkhSWzG6T++nUnCD+JQnUiUgSESPCK3OUibymiRGeXARu644Q0DuK9rcq0OG3TjoTbjbkM6SWzA5flqajs02N5sq9996LJ598Env37sVK6VlRZCiMCCFzxmzMs9P91bpjRxA6XUhVPr34oh2Dg0FYLFJ278Dy5U5VHQU4tVJ1Vydstg1jKZ6J0Yhly5YpkVYpDmrY8PbbmpnaaNRheNiPl19eh97ei1BfH8KyZTZs3KhFRqZ7k5iseWMo7sZw2AmXzQ+b7Vy1VXWFHyMjTnT0uFWna4nwpJH0nQi/81IYkjazeLDrsfEC8tQpbVSJRINE9IjIkuuQa5BrEbEl89HkvBi808Nvs0V4phS4Ph9C3/8+PC+/jOagF5ssZcCF12Hdx1vhWUlz9GLEnaXpaCFSo1MhXr8//uM/xuOPP47nnnsOa8Q8NwdQGBFC5ozZmGez/9U6ioGBg3juuU50d0fR2WmGw1EPj6cWfn8UXV1aTx2ZHi/DUkdGrFizRkSIlHZP/nwiiMRjJGX7aeuKRFbiUupebsYFF9iwbJkda9ZI1OLc9eXyJiFma+kPpF070Ov3wGppUiM67DoZhOaCXe/HquW9+NnrLRgMaapEUnc1FV6srvNh86VutLd7Jk1hTCYgJfojniIZKSLvLSdOaPsi5gyGICKREO680warVasak1YDU0V4phS4nZ3QvdSOx1YMor3Gi8BoEM6u59D0xG/Q+r/+CTYTxdFiw5NjVWIx0mff//730dbWpv499khLdvX8rqJWIFIYEULmlJmaZ7P91RqJHERFxWHodA709MgvTPH8iPCKoqrKjFgsrCbNi3CR36Xr1oWxdm326JRm7LajsrIeZ85oAu6ii6xYuTIMv18TcFu2iH9Gm1kmkajp3iQyvVHi5ZGIzzvvaJGZigrAc2krarfJn+DtQLADNosTQXcLXu5p1T6mLIQPNO7CpSvbUbcsAIvTidc6mjDQ35pKo00tIEW8iV9Imk/K13DlBi9O9ZxFbaUXG9cO4IILNP/VmjX1uPnmzRgaMk1pbM8qcAcG0NjRgZ+uGESbqxs1cKDBUAN/0oe2zqeBlx7Bzvd9ZsrvM1mYtGapSixUanQyvv3tb6vjdqk0yODRRx/FJz/5yaI9L4URIWROmal5Vt6kZYCqCBIRGjKva2goCKOxE9XVDlitWprMZktHMHqh19egpuaEElIXXywl9WHE4yNYs+b86NT5xu7NqurLbO5EJOKHxWLG5ZdrAu7yy/N7k8j0RolAkdcinyfzyP7gD873CEkfo22XeHDLMHD0BLBj/S7ctLkNEV0NEvYGhCN+XLWyDTURefSdOQlIqX4bCoVwpmoXeiztWL1Jh/W2CtQvr8ZFqy9EfDQ+zjwtAlGq37IJpEkFbnU1an1n8M+rvUoU1UD7XlhNbiDQi/aTe9F82cfgsRfZqUsKjs02vipxLvoYSSptPqAwIoTMC/mYZ9OiRSa6S5dmERUSabnkkhDWrIniwgul95Dmn5F0mdEoikB6nqxS0+VlfIfDoYmb+vrJo1OPPKKJFxEQmrHbhGee2Yrm5g1oaRkv4MSLk+ubxFTeKPH9TOYREkQqfeYzgEXnRW1vO2Ctgclcg0BQSwfWNAJloXZNTKU+Z6q0h1zHxg/tQperDauTy9Ew5EFCF8Xp0aNwDBiwdbnWKuHEiU689toG5dOaqvJvUoEbCuHIE4+o9JlEisaIhOEyONChj8EX8lEYLWI8qarEpQyFESFkwZMZcZGouqStpLPyxRfbsHq1WXl/xOsiXhptDIVm5vZ6xZdQo8TNTTdNHp0S0SWi6Jvf1Dw3IiAk3ZXWTiIQWlrsCMGLM94zcNvcY2/subxJzLai5+Mf8aHjZwEcPNmA6LCWFpNRUes2ytCzjlSEyTNt2mNHsxf7KtvRYKpBjb4aCIhCsyERS6DT34kNng0qJfbOO37s2yfGc3tO/WrGCVy7He4rroHzwEsqfaYiRZEwEBiB/4LlcJZ51NePkIUMhREhZEEzWcQlXe7+5pt2XHrpOS9QY6NViaSzZ0dw9mwjkkkRNSIUZHCpfUrRJaJIBIqU06cySti4ETjeGcJ/vLoLR6PtCEQDcJqdaKpvQuvm1pyMxLOp6BHR1vZzN+r6nbAZ/IglrOprIKLNFPOrDtmSdssl7eGDD+3PBVBtbQAkQyFVbjHAYrTAH/YjNBpCMJhAX58ZbrdNeZtm0q/Gc+en0BQ8oHmKAr0qUiSiqLe2DC31TYwWkQUPhREhZEEzXcTF49msBIZ4XYJBPzZsMOOii0QUbVYVYOvXTy+6pGJN7osoShdZSRWapOvOVu3C82fb0FBZgwZXgxIRbYe1EMrOrTuLWtGjiTYPPnhxE7Ytb4N3BDh9yoVysx8bV43vbJ3tucceP+hWok6u3+q0am2d+rURLWa9GcloUpnL/f5GFS2acb8am01Vn4nRWjxFkj6TSFFLSkwSstChMCKELGimi7hUV0sHa83rMjBwro9RLvOcMkVXOg0nSPNDEUonz3qBi9uVKKpxauEqJSokgtLZjuYNzTlFQGZS0ZMZKTutb0V5EGiwt8Nm7MDZXidWbGlBWUPuQkOuUyJdaVHnqnQhEokgPhBHvbMeFoMFjY31eOmlzbPuVyORNKk+E6O1eIoy04+ELHQojAghC5p8Jrv/6Ed2Nfoi13lOmaIrs1+RiBJpqHjtzT68UxOAS9JPGbisLnT4O3I2Es+koidTtMWSNrwxvBNHRpqhj/lw5JQbtS0elBnzGx6XjtiIqOsIdMDpduI9m96Dm1bdhMqySvU1lDluhepXI18bCiKy2KAwIoQseHKJuMxkntNE0SWeIulwLcZuefyP3e3G/Xsy0k8pZC1pqXyNxPlU9EzaJTvhQW+fVK6lojd5jjxXkZytO1WkK1skZz761RCykKAwIoQseHKJuMy0+msyIXDXXWltMSH9ZHUpUdQ70ouWxhYlKnIM1hQnUvbYzEaeTxXJmY9+NYQsJCiMCCGLhvMiLhmqxO32zKj6azohMC795O9QkSIRRbetbcVjj+UcrJkRU0ZvijzyvBT61RAyGRRGhJAFx7RRmElSSJ6mJjRdeSfafm6ekT8mmxDIln4SUTRVsKYQkaQpRdvpeR55TsgShcKIELJgyNkyk9nxMUOVtN5iAFo+XhR/TGb6aapgzfPPa9f++uuFiyRNKtrcbnjLjfAFTsFtrYcnPRC32CPPCZkjZFaa3E6ePKnW0rH+S1/6Em655ZaiPi+FESFkwZBF74y3zEyhSmyv7MXOB25Bc7OnqP6YqfxMzz2nTbGXHkp52H7yQpox7up8Eu3rOxDoPApnqAxNxrVo9dXB1ust7shzQuaIlStX4mtf+xrWr1+v5qY99thjaGlpwf79+6cdOj0b9EV7ZEIIyYOJeke0Tvq+7Mv5capEVEgmspZ9n09pAmnsWCxtkFkxlklPjzZORBpGTvkaZsmug7uUIdywbj0aVm2BIQG0Rd7ALvMRlpCR4hLxAkNHtGORue2223DrrbcqYSR9yv7u7/5Ojf558cUXi/q8jBgRQhYEOVeVzWbGRpErxrq6tG7Z6UBW1tcwC7xBrzKC1zhSTScvXQFr42ZgoBPtZhOaf6cZnkK5vwlJEwsBHbuAvnYgFtDG0SxrAqTJqLH4P2/xeBw/+tGPMDIygquvvrqoz8WIESFk3ggGg/B6veqYLQpznt5JqxJxVcstHD53X/bnKIUkQRkJzsiMNRE9cpT1hRfm8BpmgRjAZWabtA4Yw26Hq2YVAvoYjnX7cORI4aJThChEFJ1uA3QGwN6gHWUt+0XkzTffVFEii8WCP/qjP8Ljjz+OC+UfWRFhxIgQMufIfK6DBw+q+WbRaBRmsxn19fW4+urN+OlPTdNXlS2ALoSTVozBi8dGRtH2fCWkC+NsO0dPhlTFjZt5lsIb8KPrpBPf+pUbsaHitA8gJUrEq0WKrJIfToVDDamfPdlf0TzlzL7Z0NjYiAMHDsDv9+O///u/sXPnTjz//PNFFUcURoSQOUdE0eHDh+FwOOByuRAOh9X6oosAvX7r9HpnAXUhVBVj9nPldK2DEWDkWrS/04SOmnVwuoxTajaJloVCIdhsNjWSI++ZZ6mmk68d6QXeaYHD4oGrSKZvUqJEfFr6TCJFmZhcQLBDO18kYSR/NF1wwQXq/uWXX45XXnkF/9//9//hX/7lX1AsKIwIIXOKCAGJFIkokhC5kD729HTiIx/ZgOZme256Z6F0Icwop7OtrcVOzyto7ngOvstuh/vTH5r0ErNFzaTaxmTSombZmNh00phwoux0C9ZbWqfu9Sh/+as3MXfR3sjIEsTi1jxFo/5zkSJB1rIv5+eIRCKhhh8XEwojQsicItEREQISKcrEarWqcLmc93jshdE7MxUC+XRnzNI+wINeeI7tBnCtKLico2bC1q1bp3zKiU0nB7rdePBxDzxZjOsD/SF4/PNnnCWLHItH+3kRT1E6UiSiKNwLrGwpmsi+//77Vc+ihoYGDA8P4/vf/z6ee+45/OIXv0AxoTAihMwpkjKS6IgIgXSkSJC17Mv5TGbUQXqmFTR5DmWd6ZC2qaJmsi+lybmm1dS8NkxdqFcT2QX0tmn+EEmHyJta+k1uLfNsJAcaUrlg+Tcl6TP5NyWiKL1fBM6ePYtPfOIT6O7uVn88XHLJJUoUfeADH0AxoTAihMwp8oYvKaN0dEQiRYODYfj9I8pomRYEM9Eo51XQ5CsEsnSY9MYD8LXcOOk0+pm0D8glapaLMMpl4GxrixdlofkxzpIlhNGm/duRn5c5Ssf+x3/8B+YDCiNCyJyT7lp74kQn3nnHj74+M/z+Rrz00mYcO6aJn5y6YBeygmaSlFjIasAu62G0H/omAtZn4CzTjM/i8ZF01rSqJEspWr5Rs1zIVqh3+60+4N35Mc6SJYjFs+R/XiiMCCFzjpiLxUfz2msbsG9fCG63DZWV9jHxk541NqPB8TOtoJkkJbYLB9HmPI2a4SQajB74dYaxajDx+My0fcBkUTMRRdK8LjNqlg9ZC/UiC8c4S8higMKIEJI7hRgZn/FQL75oV4Kounq8+Nm7F4jFtLEeeXeQnmkFzYSUmBdBtKMTNVEjagw2wFkBa0qwSDWYGJ/H0mozaB+QjpqJp0jSZxIpElG0+YJabeTCDFMV5xXqzZNxlpDFCoURIWR6ZmX4yd+zfPYsYDTOcOrHTIXAhJSYryKOQHIIDcM6YEO96i6trs/qUiXyUg12nt8oj/YB6aiZGK1VHyOTDvb+nwIH/7nwlWPzYJwlZLFCYUQImZ4ZG36yM5VnWbTFli3Anj3aXt4dpGcqBDJSYu5OL5x1BvjX1SKxdjNCXk0DDif8qvO0GLELgaTNVOrs+GMzM4wvUOMsIYsVCiNCyIz69ORm+MnOdJ5l0SginGY09WOmQiAjJebx+XDl6V/i26/uQfDXPugiLiQtftire/H/bG85P1o0GyJehDvbEY7WwGypgd1QpMqxEjDOEjJbKIwIIVOPqujuhj3PPj25MpVnuSBTP2YqBFIpseRvVgLvOKFb1g64OqCLOtXYjWRVK3AFCpalfGqXDyt8AfQMN0BvBOrrxYMk6TZWjhEy11AYEUKmHlWRSKB+2TJs9vthmsrwk48xO9WR2mZxY+dOz5TiZ76mfsjLeeUFG6607ESZrRkh+GCzuTFs8eCVF4APfbAw1yVZyj1PuXHPFU7UVPrRH7AiVayGrY2sHCNkrqEwIoRMP6pipURO3sFWyXNNzHmJP+axx3IzZmfpSO1paIXHs7BGU2Saw2XAhz011kM/g0BZNs2YzlI6Kj3oQxM2mttgcAFenQuDPX5EanthWduiffAsKtUIIblDYUQImX5Uxfr16NTrseHIEdgn5rzyMWbPtCP1PDCDhtZ5F/Nliq+3A1pescHWjpqyDviGnBiw34rliShw4H7OOCNkjtDP1RMRQhY+6VEV0nAwE6vTiejq1Qj9+Z8DX/kK8MADmugJBscbs+Xz0vdlX0Ii2TpSi7k4fV/25fwCIm0Ol8CY3MLhc/dlP5doUVozGgya+JGjrGV/oviKJW14Y3gn9vQ/gMePfgVPdj4Ah9MEdD0F6AyakJSjCEkRmAtARHu9XnUkZC742te+Bp1Oh8997nNFfR5GjAghuY+qqK0d6+eT9wDVmXaknkfybGg9o2K+iZV5vX4Pens9C3bG2XkeNLNZdfGWhpXSm4mQYvDKK6/gX/7lX9Qg2WLDiBEh5LxRFTKaIhAIIBaLqaOsZf+8URWZIY9MZF1hBCwD5yJBmR2pM1nAoynSlXESIMsMlOXS0zKtGSfMiVVr2ZfzgogsEVvxuCa+5Dg240yEpAjHTGQt+yIk59GDJn+5iwdNjrKWfbL0iQwOYujUKXWcK+R30F133YV/+7d/Q2VlZdGfjxEjQkhuoypS++OYLOQx7AUsrwHbyoDOB4HuDF9MqiN1KAyEYi7YjH7YsPBHU8ykMi5Xj9JimnGW1YOW+nmRLt4zmfNGFj6xcBgdu3ejb/9+xIJBGO12LLv0UjTcfDOME1LvheYzn/kMfud3fgc33HADvvrVr6LYUBgRQqYeVWGzTf1mNzHftK4LaBQz9XrA5hlnsA5Vt+LV3wDJs+0wJDsQ1zmhq27BtstasdSsxNM1sJy2LcECnHGW9qBJpCgT8aSJiJbzFEZLk47du3H6mWdg9Xhgr63F6PCwWgtrb7+9aM/7gx/8AK+99ppKpc0VFEaEkKlHVUxHZsjj7DHg7LcAm2NSX0zbr5uxq20n1q5oRk2lD70Dbhz/rQctyRlPFlnQzMajtBBnnE3rQZvh3DyysIkMDqpIkYgia1WV2jNYLOoo+yu2b4eloqLgzytRyD/5kz/Br371q/MKQooJhREhJU4+fRmnRD7Z5AN8sUl9MZGBDrz1mg81NR44PR6MQI5ATWxWk0WK81oKxKy7dy+wGWdpD5p4igR5sxJRJB40SbcyWrQ0ifj9Kn0mkaJMTGVlCHZ3q/PFEEb79u3D2bNncdlll43txeNx7N27F9/61rcQiURgkFLPAkNhREiJMl2PnRlhye6LCcecODvohru24JNFivNaCsisu3cvoBlneXnQyJLA4nIpT5Gkz9KRIkHWsi/ni8GOHTvw5ptvjtu7++67sXHjRvzlX/5lUUSRQGFESImST1/GnJnCF6NfrvliZtMwcU5fCymMB40seiwVFcponfYUSaRIRFHY68XK668vSrRIKCsrw0UXXTRuT4z/Ho/nvP1CwnJ9QkqQiT12purLmDfifxEfTDKu+WLkuLIFZRe2zrph4py/FpIVEUPyBkVRVBo03HyzEkHJREKlz+Qoa9lfajBiRMgiRMqmZ/PXej59GfNmCl/MrM3Ic/1aCCEKKcmX6jMxWitPkctVtEjRVDz33HNFfw4KI0IWEYXqOlyIOWAz8cXM2ow8icF6Tl4LIUQhYmg+BNFcQmFEyCIi3XVY8uzSS0YqgtIVQuL7KFaPnTkzI0uX7CzVV1MZrOfztRBClhYURoQsEgrddbgYaa0ZEwtpg1GlX0+WKfJTGawX1GshhCxqKIwIWSQUuutwIdJaBUNEkVSySVNIGTKb0S1b/Eq5DGRdMK+FELKoYVUaIYuEzK7Dmcy267AIiPXr51FISPqsL2OKvPQ/St+X/Yg354Gs8/5aCCGLHgojQhYJ6a7D0mU4p8n3i4XI9FPkMw3WmdBgTQgpNBRGhCwipPpMugwnk0mVPpPjou86nNktO5OMKfJps3iheyARQshE6DEiZBGxKLoOT1FZNik5TpGnwZoQMheUnDB68skn8fnPfx6JRELNWvn0pz8935dESN6IGFpwgiiHyrLZTJHPZhYXY/bp0zRcE0IKQ0kJI/Fk3HfffXj22WdVZc/ll1+OO+64Q7W1J4TMko5d8J7YBZ++HG57NTy66Lko0IQu2Od17s5jiny6B5L0NXrssYU7OJYQsjgpKWH08ssvKy/GihUr1PqWW27BL3/5S9x5553zfWmELJ5U2NARIHAScK4GyterrVDgNHYdeBTtA70IJI1wGs1o8tSj1V0J27FHgZ49QDKGUZ0TB0cuQWdwBaKxxPmdu/OYIs/BsYQsbf7mb/4GX/nKV8btiafy0KFDRX3eRWW+3rt3L2677TbU1dVBp9PhiSeeOO9jHnroIaxevVr1drnqqquUGErT1dU1JooEuX/mzJk5u35C5htJOx05kmWwqqTCjj8GHLgfePPL2lHWsi+EfcCL9wDP3wa8/GntKOuwD7ve/D7aeo7BYLCiwe6CQadDW/dh7Op4DQgcAxKjqj/RwR4HDr97DLqR4ypqK/+OpXO3dPTO93VwcCwhS5/Nmzeju7t77Pab3/ym6M+5qCJGUpa8ZcsWfOpTn8KHPvSh887/8Ic/VKmyhx9+WImiBx98EDfddJP6xVtdXT0v10zIQmCqcRpjaadpmiziwF8CXT8DTBWAtRYYHVJrbzSMdp8ONRYnakwGwGCE1eBUEaJ270k0r6iGx16PYMyKziErHLYQnIkuAJvOde4+eQQbao2wV9ROnmor5ODYfM3hhJAxvEEvfCEf3DY3PPbi//sxGo1Yvnx50Z9n3HNiESGpL7ll4xvf+Abuuece3H333WotAulnP/sZHnnkEXzhC19QkabMCJHcv/LKK7M+XiQSUbc0Q0NDBXsthMwl06adJjZZFKTRoiD75ZuAvl+nRFH1uPO+3nYE4pvR4FoLjBzXzumtcCGCjlgUPksdPEY7QmEgGgNcVhOQCALxEKA3wRo+Ar+3B6E3noTJYZk61ZZiRoNjZ2MOJ6TECY2GsOvgLrR3tiMQDcBpdqKpvgmtm1thMxXv38+RI0fUe7dkga6++mo88MADaJj4F1Epp9KmQkYl7Nu3DzfccMPYnl6vV+sXXnhBrUUEvfXWW0oQSWO8p556SkWUsiHfAAn3p2/yC5qQxUZOaafpmiwOvq0JGVP5hPPlcOuicCIOv6UOKG8EkkkVbfLH4nBaXHA7tX83NhNgNgLh6CigNwMGG+A/iPDAcbVvK6/NOdU2o75G6YiYzqBFxOQoa9knhEzJroO70Ha4DQadAQ2uBnWUtewXC8n8fOc738Hu3bvx7W9/GydOnMD73/9+DA8Po5gsGWHU39+PeDyOmvQgpRSy7unpGQvJ/eM//iOuu+461QtGyvanqki7//77VRO99E0GdRKy2JhsnIZN78XaZUfGxm1M22Sx4kJNyEj6bNz5IXgsTjSt2o7eoBe9plqEq96HXudF6LWvR9MFzfBgRPUkshvDqC/3YyQYRiBRgVgkgIDvDEZiDtRX21SUqXPIBYfNrFJtRkRVqk2G5sq/PUmvZSJpQOljFI9r6TM5Zu1rlMPYEULI5Ej6TCJFNY4a1DhrYDVa1VHWsi/ni4FkiD7ykY/gkksuUUGMn//85xgcHMQuCYEXkUWVSisEH/zgB9UtFywWi7oRspjJTDs5bSFc6NyFVbZ2xKMBXFXuxPJQE2BonbrJYtWVwLL3A2d+CiSi2nmJII0OAnW/g9bL/xiwab8kO0b64DSXoWXtzWhtvA3o/qkmPoYOYbO1F3CtQGcI8HcMwJz0o7GuCptXAkOhSVJtRnvWIbmT9jVypvxD+gn+oXRETCJFmcjrkL5JynNEvxEhkyGeIkmfSaQoE5fVhQ5/hzo/F36jiooK1dz26NGjRX2eJSOMqqqqYDAY0Cux9AxkPdfGLUIWEum0k3iKrqrehQ32NnhHauAdasCF6/0oG5R00jRNFsWfU3mp5jMKdQHhHsBYBtTeDGz9uvIY7Ny6E80bms83ZqrPDwDH/xOmcCe2VgxiQ80oQjoPbP522J11gOFa2AxBmHUGhMNROK2pVFsOQ3JVXyNXyj90NIt/KDMilvZOTRg7QgiZHPn3LJ4if9gPq/Pcvx9Zy76cnwvEAnPs2DH8/u//flGfZ8kII/nFKQ0b9+zZg9tvv13tSXdrWd97773zfXmkhJmuwmoukPSSBV7U9bWje6AGYdRg7Xpg3WYrEEuJIWmumK3JopTt9z4N1FwHJOKaOEpGgeXbAeu5X4oihs77y1EEy5mfa2k4SV3pDLBH3oVd/EgVK8bK+e2xAOr11Tg8uBLQeWBN6BBODcmV3iVTfu2mq6jLcewIIeR85N+0GK3FU5SOFIko6h3pRUtjS9GiRX/2Z3+mWvSsWrVKtdv58pe/rAIgxe49uKiEkajFzBCaGLEOHDgAt9utXOpSqr9z505s27ZNGa2lXF9+qaar1AiZS0ZHR5VpWPwxUhyQrcJqLpBgy8fu8CHyagAjaIDVAdjTARjdhHTSxCaLE/05sjZXAsnRc4Iqm7BIf66YtnVGwOAA9KlfO8FOwPMeYOQUMHJSRYg2V/UDZg86R0Lwd78Fs3vj9ENyp6uoS19fDmNHCCGTI9VngkqX+ztUpEhEUXq/GJw+fVqJIK/Xi2XLluF973sfXnzxRXW/mCwqYfTqq68q43QaEUKCiCFxrn/0ox9FX18fvvSlLynDtRisxc0+0ZBNyFwgokgqqsQ8LBVWkhKStSA/m3OOxQ2LwwmLbkKN+3TppLQ/x1wFnP6JJiqkYaMIHXMFMHx8CmGU+lxLNWAwA4kwoHdqwkWeN9StSvbhuQKwLIPJYMPWVXZsGDqLUPQ4bJfdCXvFyqlfV67+oTzGjswI9kciSxjbVOnyIvGDH/wA88GiEkbbt29HUkqBp0DSZkydkYWQPpNIkYiidBPDsWaGnZ3KQDjnabWZppPS/pyeXwLBM5qoEX9RbEQTNu9+E6j6z8mFgdwXARU5C1hqgJET2n4yDiRiwKhPE1eO1eO8P3ZHOey6DkAfyuF15ekfymPsSE6wPxIpITyTpcuXGItKGBGyWBBPkaTPJFKUSbYKq4KTLXoxk3SSfH7ZOuDkf6XSYTbV1Rp6A2CsBPpfBA7+/4DA0TFhECx/D0Ku7bD5n4N9pEM7J88Fg3ZdEjkq2wDUfwTw7ZudKXq+/UPT+ZsIIYsKCiNCioAYrcVTJOmzdKQolwqrokcv8kknZYqr8s1aKkyEjYgaaY5oqdIiR4HjwJk2oOISjFoacLAjgs7XDyCaPAyzLoR6zyXYXG6HaegAEBMBZAOcazXRsu5TgMk5e1EzX/6hXP1NhJBFA4URIUVAokFitE57iiRSJKIopwqruYheTJVOmkxcOS9QpmjodJp40Zk0b5CqTotrz2WtwcFTwOF+Kxz6YbhGTyBsqsdhbxkwWoOtdofqS6SEg+tCrfRfRFchRE2x/UPZYH8kQpYcFEaEFIl0JZV4iiR9JpGiaSusFkL0YjJx5X0RsNUBw+9qUSOpMgsPANFB7TEdqxCMAJ1ewGEBnAYRWNKPSI4xdPqADfYq2G2GVJqsTIscTdcmIF8K7R+a9vnYH4mQpQaFESFFQkrypfpMjNaF6GM0bT8kERUj0prfAySDgM0+LnrhHTwGn3GaapKpxJWIGXs94HsFCHdrwqb2Ji2KNOpHKGbVOlfL08p8DvEjJROw6kfhj+kRSpTBHpeu1OaUKVk/dZuAxcB8+5uKQGRwEBG/HxaXC5aKivm+HELmHAojQoqMiJjZCKKc+iGFQsDjvwT6DgLROIByQIYeb96MUMSLXb1daO/7FgLx2KRTscdEV6xbNVqcNDUkb/gXf0X7+P4jCBlqYFu2CfaeHylhYNPrYDZ4tM7VhoiKIiEeRHhUD7MhDlvyLBCLaINmDXZNPOQbVVmIJfFLpD9SLBxGx+7d6Nu/H7FgEEa7HcsuvRQNN98MY2Z7B0KWOBRGhOSITKEfm8nlWWD9kGSoYtse4OI6YPlpYCQEHHlbmvZjl3kIbSNATbUDDU6tY226g+3vbf69c6IrNKyZpZMObK7pgKl8pSZgMlJDo/pyHDzajc7OOKLRkzCbu1C/4kJsXp6AfeAF1DvP4nB/BeC6AFb3BoR972IkeBaN5UOwS/8k22rAuU4TRflEVRZySfx8+ZsKjIii0888A6vHA3ttLUaHh9VaWJuaJkBIKUBhRMg0SDBGdEd7uzalXorMZPaYjNnIu7gsT3WVUz8kuUC5OGlkqt8EBA8C9k7AOARvz2m0L6tGzbIL1TRsIT3rSDrYrouuw5ljHXAku+EaPY1w0I/DoTjgP4mty1/TRIx4i6JeJWIOHjqCw4cOw1HmgstVoYm0I8eBxkuxdesHsXn9WeC4F53dA/CPRGEu24DGDVdh8yonMNAODLyupeHyjaoshpL4xZgKzEifSaRIRJG1qkrtGVIDtGV/xfbtTKuRkoHCiJBpUMGYNk13NDRoU+plLch092Kqq5z6IYnQkseUi0uagOGtwMgGIDYI35lTCCwzo8E2/g1bZh2d9p7GyVMnUZHshXP0KBAPw6kfAsw2dIZWYkNI0mpvaI0c13wcwVAYna//BI5kFE5Vtl8Pp2vzeJHm3oStbmDDRD+UpMAsNwLLb9QuIJ+oCkvii454iiR9JpGiTExlZQh2d2ueIwojUiLo5/sCCFnISIAnHYyRm1gt0vdlX86r/x05klpMo64MBk3AyFHWsp9jP6RMxvVDkuiTCC1RbGkSdqDPALepBk5HpUqfZSLrMn0Z9LEYrKNdgN4CxCOqDN9q0iGKcoQMtYDnSsDRoM6FOp9FNK6H1eoEdHpg6DDgP6hEmog3EUJpRAx5JCVj1mkDaA/cD7z5ZeCdfwD6f3suRZdPSbz4nDKRtezLeTIrxGgtniJJn2Uia9mX84TMB2fOnMHHP/5x9ftEft9dfPHFajxYMWHEiJApyAzGZCLvEx0nYvD9+0/gObZ76ijQRHUlpM2sst/cnDWtllM/JLnJ86bDWHJxIpJ6e+FpaUHTekw6Fbt5TTPKOo0Ij8ThtOg0kZFMIByzwowIbIlewLRRS1ud3Qub0wOz2YpwHHBaU786RjoR1q+E2WyZvGllIVJgLIkvOhINEqN12lMkkSIRRWGvFyuvv57RIjIvDAwMoKmpSc1Ifeqpp9Tw2CNHjqCysrKoz0thRMgUZAZjMgtzZO3sPQb33ieABsfUOTZRVyKORPwEg5qQGVNXHdr5KfxGOfVDEjGWFlrymHLRLS1qvzX1r3yyqdiH9ftw+OybwKgX1lgc4aQNI3EnGsuOwg4vMHxE8xglY7A7y1HvAQ53a49nNVgQDgcwAj8aL7r8/Mq7QqXAlmBJ/EJEqs/SniJJn0mkSERRep+QnNqGFJCvf/3r6g/DRx99dGxvzZo1KDYURoRMgeiVSYMxHRG0oB0eEUVTRYEkvfTLX0ppmdbbp/xcGb2mrpya+pptPySJ1ogYk+edYO6WOM7OTc1ort0EXwJwu9aN9THavOUqoP8ldB7phD9eCRPCWO84ik3OdwGDAwh2AiubgeFjSoxsXqm9Rmnk6A/GYDaY0bgxS9PKfLpCT1eGv0RK4hcyUpIv1WditGYfIzKjtiEF5ic/+QluuukmfOQjH8Hzzz+PFStW4H//7/+Ne+65B8WEwoiQaZg0GHPtAFpffR5wjTernhcFEg/Rnj1AXR1w+rQmlN5+WxNFEg6WqE6Otf859UOSx8p8vIwyd08sAM+EMncluq66EeuiT6CvPw5rogvlRj+CIStGYh64Ks0wqDlpTqB3D+TX39aVLmyoHEIo4INt1Y2wb3rfzFNgU5Xhx4PjxdISKIlfDIgYoiAiM2obUmCOHz+Ob3/727jvvvvwV3/1V3jllVfw2c9+VomynTlXvuQPhREh0zBpMEYkwlFLlhxbKgqU6S3atEmLGnV2AkNDQHc38Lu/e051FYtcPD72OvQMrUd0+F1Y7Q6EEmbE4kbEI2EYdBG4jj+qRXlkVpoYtEc7YDc5Yd9wy9QRm1xSYGLMnnh9nf8DnP2N1lF7olhaxCXxhCxWgrm0DSlCWi2RSGDbtm34+7//e7W+9NJL8dZbb+Hhhx+mMCJkITA+GJMtx9Z7LgoklWpp57aEmuWvqg0bgMFBTTTdeOMMGiHlQY4eH2/Ag65uAy6pOo4oKhBKuFFh7oBV70PfyBqYjOthN6QEzfIdWsl9rhGbqVJg2a5v6BAw8AZQc93C7VlESAkRyqVtSBGEUW1tLS688MJxe5s2bcL//M//oJhQGBEyU3PhFIbnrM5teTwpiRbhNI23KGey+XNy9PgMnvUiEY/DG90Iu8kPq94Pgy6KUNyNaEyq4BKwV6WEizRoXPWx3KM2U3WFDp4+//okfRYdAPQmTUSJUGLPIkLmFVtG25B0pOi8tiFFQCrS0um6NO+++y5WrVqFYkJhRMhMzYVTGJ6ndm5nRJVmw3RjMnIsc3c7fOi3hHF04Eo4nQk4Dd2os76KQNgFq8kPO7qBWO3kpuk8CIYiCI0MwuZKCc7Jrk9ekzSUFOO3DKlNM8vnJoTMHHsubUOKwJ/+6Z/ive99r0qltba24uWXX8a//uu/qlsxoTAiZLbmwomG50zS0aOX9gDe04ApZbguhLdoOv9QjmXulTVuVC5zYuCEH4PDNYjbauHWW+EydcJui8AaeBWI2LXPd6zOr29QLITR4z/AwbcOoNOrRzRhhtlZjfpN12DzxVthmnh9sWEgEQPsIpoyftmyZxEh88rmXNqGFJgrrrgCjz/+OO6//3787d/+rSrVf/DBB3HXXXehmFAYEVJMc6EEmt4PYLUk6lO18/Wp/dmQa4+gXMrcLR6s3tYEC9pw9DQw6DPDvHIILms/9FYPYBZBNQSEurXZaflEbDp24eD+X+OwrxoOmxEufQTh4EkcPpBQ6bKtF01yfctvACJnNQHHnkWELAhMubQNKQLNzc3qNpdQGBFSTHNhOqpjqwHK6wpnJJaUkpo/5gFiwXPRlYkppxwnv8dXtuL0y1HUOr4Ht/sUyi39iOts0OtMWhRHHkdeQzJ+7nmnvUYvgmdeQGdgGRx2K7TZtUY4DWJOOIvOk0c0wTnx+mRcSDpFWOieRdP1SyKEzL5tyCKHwoiQYpkLizX8VHw4nY8DA/sBJAGzG7DXAzLQNVvKaZoy913/Y0P4zSi2rwH8ifVA1IRgwAFPMgCX9GpyX6aNVszH5xPxIRQOIZpYBldmhExvgdUwBH945JzgnHh9he5ZNJ0fa56R6fZsqkjIwoDCiJBimQvz6fyczxv8q58Fep7WKrhi4VRvoSHtZq7IO+XkPRuC8cQj+ODGb8JsjGM0YYdFH0Uk4cRAwAV7cEBZoxDL0+djccNmtcGsjyI8atMiRUIignDcArPVMbXgzLdn0VTRoELMbCsCsXAYHbt3qzEcMt1exnDIzDIZwyGdqAkhcw+FESHFMhcWY/jpsUeAnl9p6SbbciDUo5W3S4or1AXUfzjvlNPo0V24qKINJkMcI3EPDLqYauxYYe2Gb2Q5EtGINhokEZlcdGUTJBYP7DUXo77ztzg8uAJImmEVj1EkjBFDAxpXry9MSH66aFCxIncFQESRDG61ejyw19aqwa3pQa4ynoMQMvdQGBFSLHNhoYefyhv82b1ajx9Jn+mNgGOVZo6Oh4Gy9UDtjfmlhiJeVIy242h8BUaiXhiNMYwmnAigFmb4YDONwKjXjNJY9j6g8rIMb9MUgkSQc9592Fx5Aoh60RlcAb++AmZ7DRo3bitcNct00aBiRO4KlD6TSJGIImtVldozWCzqKPsys4xpNULmHgojQoppLizk8FN5A0/GtGhRIgzoU14oeUzV+yfVGyjPx7QaAyivbkB3dz1WVR6GUQeEI2YlhnSOOhgu+IAM0NKaO/a3nxNAiSjQ9dQ5QRLuAU5+75xISokV0/JrsLX8DDb0vYYQqmBz1cCe7AA6O2bv8ckpGlSEyF0BEE+RpM8kUpSJqaxMTbdXniMKI1LA8RpLnUSBXiOFESHFJMeqsKyMS1OlPldSZ1I6L/1+dHogPqLdr7km/8hHSjRsWufH29iMHh9U/yKH2auadds2XQuY7EDPHk14WJapijIlgOJRwLVJi175D2rpNvE5HXoQsFQDrgvPiZXYMOw6P+x6iWxdqImqQlXnTRcNKl9f2MhdgRCjtXiKJH2WjhQJspZ9OU/IbBFLgF6vR1dXF5YtW6bWOplDuIRIJpOqqrivr0+9VnmNs4HCiJC5IF8jcbY0lecKzUskyBt/YlSrTHNvAxruHC+mxp574qiQCZ6gZU0wnW7DlkYgGNuI0aEy2COvwWStAEbeAroOAtbl2nPJc4uokRYBUsYv4kdE0dBhwOjQBs2Gu4HAMc0Ibl+hfayIJhEkEvFKJs4Jptl6fHKNBhUyclcgJBokRuu0p0giRSKKwl4vVl5/PaNFpCCIUJDGiN3d3UocLWXsdjsaGhrUa54NFEaELESy+WbqbgWstcDgQU146C2aqDCXA6f/B9Cbgd7ngaG3geig9jHlFwI11wJ1twFdPz1fbMm+0NeupbgMXYDZqoke+ctSjN1D72jRKdsKTeAIkjrzvqTdF1GkUnoBwCTXFQICJ7QWAvGQFl0ST5RcXzp1VgiPT64+rtlG7oqEVJ+lPUWSPpNIkYii9D4hhUAiKCIYYrEY4vE4liIGgwFGo7Eg0TAKI0IWGtl8M2KwPvOkJoZE6BjLtDd88RaJEDjxPcDk0CI08uavM2rHkZPA6UHg7G+UCTqrSVlEw/Ax4N1vaUJHPk4eSz1+n2bAlueWx5WjiBCJIOmtqY8NaF6n8kYtujT4hhYpkrSaRIqiAaB8kxYNS2rpNfVYkhrMtWnkZOQTDco3cldkpCRfqs/EaM0+RqSYiGCQApNJ5z6ScVAYETKfTFbqPtE3IyJD0lUSgYn0AzoDUHExULlCEyuCzqyJkIotwGgPYCo/F8ERAWStBvp+DVRdPXXJetrgnY4KSUdt6zJg5DiQ0GnRHynbVwJoo+Y3EoEjgkueU0SRRInU/oh2fdE+wFKlXZ88jgg1JDShJ40pDz84u4aLCzQalA8ihiiICFkYUBgRMh9MVeo+0TeT9vBAB5grNY+PpLaSSaBsnTaFXmaLCaYy7Xxa2Eg0RxozyseKqBF/j4ixdKRpYjprMs+OiLDBt7WGkvIYsi/POSIep6gmeuQoKTvxFEV9mlBad/c5sdL1JHDiu1oESiLd0SHt8XSrC9dwcYFFgwghixMKI0Lmg+l676R9MxJVkUiRUhMihC7QPD99LwD9v9VMzuL9Ee+OrU7zBMl9+TwROKqs35wSRlGg/wVplqPtSbRGqswyTcqTeXZUZKdWq3wT35Gk7Qbf1PZFiCnBFAZGTsnUtfGprHT0x/8O4LlS+/jIAOB9WYuESbSpkGZsQgiZJRRGhMw1ufTeSftjzvwcGBUTdSXgWK2lqcS7I0ZmETqS9pI0mqylwaN8rKS0gme06Iyct68ERo5pgkh9vEMTI9KXSARR42fHC5HJPDvr/x9NmJ39NeB/SxNC8jk2EUyj2lgSEWyNn9OiWOOq4DJSgyraFNJSaXKd0mpAolCSspvnhouEECJQGBEy1+TUe8ejRY6q3gu8/tda5EWEkYgdMTyrztc1WhTGUinNbzRxUrMD8O3TxIZUpYnwkVJ7EUKuS4DRgdR4j6iWDpNUm2MNMHREu4a0PyebZ8e+Guh9DnDWaKX5ghixRRRJJEk9xgRRMzE9J69FIlbS80il9GwLouEiIYQIFEaEzDX5dGKW5oQrbtVSWyI8RICIoJDUmnOD5ulR6IHRDm0kyOqPje9jJFVfYnC21QDOVUDZBk1IBY5qqbh9n9UiQJml/crrNIlnR8aPiKFaqslyfr2TpOckWiRNKmXem6QC5bXNc8NFQggRKIwImWvynaGWmdqSNJxUpUkKS9Jqk4mqiYJGma0zIzZ2YPhdTRSJIBIkcpRZ2p/NBC1pMvEmSWRLrkM+T3xM8tgSAZPzkzExPSdRKmfKL7VAGi4SQohAYUTIfJBP752J5eg9v9RGdEj1Vy7jLSYKMfEkSZWbmKklpSVpuczSfhlMm80ELes1dwFHHtbSegimIkmV2v7Ej89sRzBZem6ydgWEEDKPUBgRMh+kxY54iAInAedqLW02FelIkJipRcjkM95inBCTnkRRwLkWiPSdS+elS/slVSYiKZsJevXHNUGlxNmAJoqW7xj//FO1IxD/1MTXRAghCwRdUqavlRAPPfQQ/uEf/gE9PT3YsmULvvnNb+LKK6/M6XOHhobgcrng9/tRXl5e9GslS5iphEOuDQ5nEm2Rz0l3txZvj5Tdi18pHTGSyrOKSzRxtPWBqR93quc//ti5dgQTo1qzGRpLyBREBgfZQZzM+v27pCJGP/zhD3Hffffh4YcfxlVXXYUHH3wQN910Ew4fPozq6ur5vjxSSkzXxygXETKTaEv6c4Z2aM83WWm/XEsuJuhsz59LOwJGiUgBiYXD6Ni9W82ciwWDauacDOiVmXMydoWQfJjdCNpFxje+8Q3cc889uPvuu3HhhRcqgSTTeB955JH5vjRSSkwUDiIa0vfTBus0sRC87zyEI+3/G959XwAO3K9FYyTiNBskMiXiR0zQqgouAeilr1AlUDEhLTbTdgTp7ttpZJ1O0U36eV6tbUDm6yckB0QUnX7mGej0ethra9VR1rJPSL6UTMQoGo1i3759uP/++8f29Ho9brjhBrzwwguTfk4kElG3zFAcIXPSx8jiQWg0hF3PfRbtJ36FAExwmuxocg2gNdgLlWybTUoq09A91AX8/OfAa68BgzHA8jrQtAtobQVsec4ty7cdQaHSiqSk02cSKbJ6PLBWVak9g8WijrIvA3qZViP5UDIRo/7+fsTjcdTUpEL7KWQtfqPJeOCBB1ROMn2rr6+fo6sli458oh2ZwiGTCcJh1+uPoO3o0zAY7Whw1sCgN6GtrxO7+no14/NMIisTr1NSWj9/DWh7CYi6gdr1gMEAtLUBu3bl//jpxxRhk+5NJC0B0vdlf2IaLZ1WlPJ/EYtylLXsEzIN4imS9JmprGzcvqxlX84Tkg8lEzGaCRJdEk9SZsSI4ojMOtqRQx8jb9CL9lN7UWMyosbmVjPQrMkAEB9A+9k+NBt64Dn6H0DjH+cWVcl2nc4dQHu7/IWg3YS0J0P2m5sBj6d47QjoRyKzRIzW4ikaHR4eixQJspZ9OU9IPpSMMKqqqoLBYEBvb2psQQpZL1++fNLPsVgs6kZIwUzUOQoHX8iHQDyGBpMj1UBRvDn9cBmM6IjH4UsY4Ol7XutmnUtKLdt1GnuBQABomJDWkzeTjg7A55uZMJrYeylb5VyOaUVCsiFpMjFai6coHSkSURT2erHy+uuZRiN5UzKpNLPZjMsvvxx79uwZ20skEmp99dVXz+u1kRIwUWcTDlISf/FXtKOsU9Eft80Np80Dv6kSiPqVKJJBsf54HE5dEu7KRk1MZD5PtnTeeVGZhDbSQyrSRt8CKozAxHSDrJ1OwD3LuWUiaqQ/UzZxk2NakZCpkOozEUHJRALB7m51lLXsE5IvJRMxEiQttnPnTmzbtk31LpJy/ZGREVWlRkjOeL1aJMUyMPtoR5aSd4/dg6b6JrS90wMY3HDF++CPxdEbT6Klei08VZeeG6cR7ALOPJk9nZeOylhrgYED54bIipdHxNxVzUDba+ciRSKKBjqAW64BnIX8whVgPAohkyAl+Wtvv10ZrdnHiMyWkhJGH/3oR9HX14cvfelLynC9detW7N69+zxDNiGTEgpphmTx3kj6SSIt7+1CqM6BoH6FKuCy2woX7WjdrKXV2k/uQUfwLJwGPVqqG9HacKnWgDHsS3XA3quZsbOl89JRmYHXgFAPYHRoAkQEU3QQuFzUTwvw0h7AdxKoGwKaTED9PuDAseJXiOUzHoUUFPGySdpWIpQixhc7IoYoiMhsKbnO17OBna9LnMce06q1REi7XAh5g3g17IVj41l4dRcgpnNhzUo/LljZC+OqGXR4ztJJWr15vfvvcHv3wlPWMD6qUrMDGHxdi/6kzcuCnJOIUrp79bsPAW//H8Bg02ajiW8pNqINo3WsBjyXA90vAQNvAAmvlv6qvBSIB+euYzXnps0ZqhXEwV1o72xHIBqA0+xUEUoR4zYTWySQpQc7XxNSjPTZhOqtXX1X4qm3a3Fz5CfYdHUvdLoOdJxyoi/Zgve/v7VglW3yl7znos8CHcvPj6pUXgb0t0+fzlt2DWD7v9IiODUPzQyUNwLOddrnh04DtjrADCBeBoS6NQ9S5da5qxDj3LQ5Q0RR2+E21Dhq0OBqgD/sV2th51aObCGlDYURIbkgnqKM6i1v0Ib2zlWodIzi5Dvb0LviEtiqgI5eN4Y6PLjwd/Io5sqlsi1blZdEWXJppmiv02agJUa1fYkcGe3AyCktnVa+ETCWaVEm+RzpPSRepLINrBBbYqhWEJ3tShTVODWRb3VqPzuy37yheUmk1QiZKSVTlUbIrJDqLKnSSlVv+UI2BKJmuDAoJY8ImVbAN7oeRodH6SfRUUWpbJtY5ZVrM8X0x4364Y0EcGRkEN6hU9qcNHMFYF2uiS+JJMlj6K2aQTsemt4zNdNRHrMYARIMBuH1etUxb0p89IhqBRENwGUd399H1rIv5wkpZRgxIiQXJPzT1KR5jEQnmfVwxgbhl2KvLfWA3T6zKvdC9PHJ0bwcqr0Nu479Bu0dv0ZgNASnyYamFVei1bkcNhE/KmJVDwwdFumhiSS5tmxDZWc6ymMWI0BGR0dx8OBBdHZ2qjE/0oZDmq5u3rwZJpNp6q8TR4+cawVhdqr0WTpSJMha9uU8IaUMhREhuSKzw4T2dnj6DqGpZiX2jF6GRIMZxlEvun0eSP/QlpY80mj5zhWbRTPFXYd/ijafFzVV70GD0Qx/LIo2/xBg8GBnItX4VDxH8tyBk4C5HMGYBaHyW2Grug2a9CtAc8uZfh6gRNHhw4fhcDiUkTIcDqu1IFWmUzKL511KjLWCSHmKJFIkoqh3pBctjS1Mo5GSh8KIkFyRevydO7UxGf1duHN4L64+9M8Y6AsgEHbilL4J5ve04sOttvnp45P+2PT0+gmVbef5SuR/hl60ByNoXrkDnqHXgXA3ULYeo3UfxsGB1ejsCyF6Vg9zZ/v4yEw6BSgGbWkWKU0j01VxUxm1ZzECRNJmEikSUeSUsJzouNRR9jds2AB7KnJXyOddioy1guhsR4e/Q0WKRBSl9wkpZSiMCMkXCQf5n4Q5uAcbNtQguLYB0RE/3qdrg2WtCKidc9/HZ5o0UdpXIhVImUi0QN4YfdU3wrP2Y2MRp4PvdOLw6VRkxmk9PzIjTSUH39D8SGLYlrSbpOEk4iTiKlsKcBapw1AopNJnEinKxGq1qhJcOZ9dGHH0SCZSki/VZ2K0Xkp9jAgpBBRGhOTLhOiDvBXbxasRnmH0Ide5YrNIE+XkK0mVy+cUmTm7V2sWabCfq2ITb5I8b9n67CnAWaQObTab8hSJSEtfjyBr2ZfzRU1ZLkFUKwgKIkLGwao0QvIlHX2QaEMmspb9dCqr0HPFZlHZlvaViI+kN9CLcCysjrKW/cw3x3RkRiIxmcha9kP+bq2ppFPCY8lzVWzQad6kyi1TzEbLsYpuEiQaJOk8GeMTCAQQi8XUUdaynzVaNMvnJYSUFhRGhOTLQht8minUpFO1CCU5ThBq4h8RH0k8GVfpMzlO5ivJjMxkMhaZ0Ye0x5XO2NIkUsRRummkrUZrJjkVkt6TVGF61pscc0wdisepsbER0rBf0mdylLXsT8ssnpcQUjowlUZIvqMqFtrgU7kuidh4XwJGh7T+QyJSxBjtWDMm1HL1laQjM2lPkUSKRBRJZEZEiL2iVhOAIr6kM7Y0gYynxJLMcJNmkkVKHYrxWzxOks6TyJaIuCkjRQV6XkJI6UBhRMh0hmYZ0BodAMyVwPIdWoRhIQ0+lTd3mZUmHh9ThSaIRCDJWA/nBee9+efiK0lHYMRTJJEZiRSNRWakKm2iMEzGsvc7muq6ZyhMRAzlLIgK+LyEkKUPhREh2Tj5PeDIt4FYENDpAJm37D+ojdW44NMLJ/ogES1JC8lYDxEn8REtOmJbru3L+TyvbdrIzEIShoQQUkAojAiZDBETJ757roxbDM1i2E3v199xLvIw39EHucZEGPBcCSQTWlpLZqHp9LMuRc8amWFaihCyRKEwImQyho9pQ1RljphEQwQ5KuPuae38QhECmWZwqUST4bCCeJ6KbQZfCMKQEEIKCKvSCFnssBSdEELmTxh1d3fju9/9Ln7+85+rniaZSNXK3/7t3xbu6giZL8rWaZ2cJQoj1VaJ2LmBqvaV2vmFxFyVopf4ZHpCyNJHl5RGIDnyyiuv4MYbb0QikVBTrlesWIEnnnhirIKlt7cXdXV1iMfjWIoMDQ2pcQRSpVNeXj7fl0OKzdF/A448DIwOa94d8eyYyoD1fwRccM/ctwko1ufkAifTE0JK5P07r4jRX/3VX+GOO+7AwMCAEkEf+MAHcO2112L//v2zvWZCzsPrBY4c0Y6TrYvOyg8DlZelujsHtaOsZT/Ha56xCDn+GHDgfuDNL2tHWct+sbpn5zpyRNoCyMgROcpa9gkpcSKDgxg6dUodSYmZr/ft24eHHnoIer0eZWVl+Od//mc0NDRgx44d+MUvfqHuEzJbQiFg1y6gvR0IBKTBIGAwABKIlGbMMiaracswWq/pha2uUhvqWgy6fgpEvcCy92mNC6VMX9ayLxVZU1yzusYmoLVVOkmjoHPP5jxyxMn0hExKLBxGx+7d6Nu/H7FgEEa7HcsuvRQNN98M44SROmQJV6VNHBPwhS98AUajUaXYHnnkkUJeGylRRGC0tQE1NYBo7ZdfBg4dAhobgasuG4X/tWNo+1UQ+L8vYecl+7MrkNmIg8nEgCCG5knEwMRr9vu1tbBz5xyKkGKkvDiZnpBJEVF0+plnYPV4YK+txejwsFoLa2+/fb4vj8yFMLrooovw29/+Fpdccsm4/T/7sz9TvqM777xzptdBiEJSUBJ1EYEht2BQExkVFZIjBhLvHEZNz0HVvLA9fBmaR4/BM1GBFEIc5CEGJl6zkP5jUfabm/MIas1WhMwm2pRNSHIyPSHnIWkziRSJKLJWVak9g8WijrK/Yvt2WOQXF1l05OUx+sQnPoHf/OY3k577i7/4C3zlK19hOo3MCp9PS0W5XOdSVFL8KF65aDCG0KmzgMMBl1uPQNwGn7NBUyOiQNLGno5diBxvw+CQAUHdDP0wFjfCMScG+/0IZlp7JhEDE685jaxlX87PyYDaidEmETHp+7KfrZJsOk8T2wEQch4Rv1+lz0xlZeP2ZS37cp6UgDD69Kc/rUr1s/GXf/mXOHHiRCGui5Qobrfmz0n/TpHsmNmsRYvM+hhsiREVjvGHrXCao3DbQuMUSGjQi3d/244X9tdgT3sN9jxnxYHDNRg1TiMOMhAx9tgPPPj+0014941evLy3F2+8HsZoYHIxMPGa08ha9uV8zsxGhKSjTRJdykTWsi/nZ2qs5mR6QsZhcbmUp0jSZ5nIWvblPCkBYST+op/85CcYnvCDkC6Fk3ORSKSQ10dKDEk5iWWot1e76fWa7pFij/JKPfQWI3p9JvSOONBU3wGPPTROgTzV5kPP6QDCcZf6PBlxJkPi3zk2jTjIIO0XOuBrxYnRFhj0cQye6cDxY5OLgYnXLDa89H3Zz9sbPlMRMpNoU65RpvQIkK0PABd/RTvKmqX6pESRNJkYrcNeL8L9/YhHIuooa9lnGq1EPEb/8i//osTPBz/4wfPOSV+Af/qnf0JHRwfuvffeQl4jKTHERy1IdqyjA1i9Gli3TqrSzOh4uxHOwaNoWfsyWtcdBXq9mgJpaYEXHvz6JeBD65zw2PwYiVuVXhJ8PX6EG5ywTuOHyfQLuZfZ8E54J05GmzEy4MNwhxt/fasHHuP01yzP29Jybj8vZjqHLB1typx6L6JIok3Zpt7n62niCBBS4oi3SNJkEhGS6rO0pyjY3a0iRSuvv35sn5SAMPre976HL37xi1nPf+5zn1OdrymMyGyQ9Jn4qMW0LP4cSUVJ1EVEi6+rDu69L8Pz+m+B7sA4BeI7DfQOetA12oSLnZo4CCdcqHL6oY/2YtDUguXTvKmn/UKZVrlQwoOoxaP0l5yfLAKU7ZpnxUxESL5T72msJmTWpflitE6LJUaKSkwYHTlyBFu2bMl6XqrV5GMIKQQiLDLFhba2Ahd/DPB+4DwFkvb6tHe2wmQCGmztcBk74B9x4shgC9ZdMH34JtMvlNmGZKJfyBv0whfywW1zw2P3ZL3mQhEMBhEKhWCz2Safdj/TaNNMokyElCDTleZTEJWoMIrFYujr68taeSbn5GMIKTppBZKe3WVxw+PxKE9PW5sNv0rsRK27GbGgDye73dhxiwee6tweVnsMbS0+JRFFqWwd7OUhPHZgF9pP7kEgPACntRJNq3egdXMrbKbC+21k9M7BgwfR2dmpZhOazWbU19erMTwmUX+FiDblG2UipMRgaX5pkZcwkl/GTz/9NC6//PJJz//yl78cm5tGSFHJ0quo9cPyZm5TXp9DJzxwOj3YcUt+Xp+p/EK73vwe2g58GzXJIBoMOvhHkmjzHVRdsXde9umCv0wRRYcPH4ZDWhS4XKoAQtbC1q1bC/Mkk0SZgnEbQv4gbLbk1BEqQkqoNF8iRRNL88VbpNJoFEalKYw+9alP4b777lPip1nMFBn89Kc/xd/93d/hG9/4RqGvkZCcGxlKzGbnzp2z8vpk9TgFvWh/57uoiftQI+kzvRXWRBhI7TdvvGNcWq0Q6TOJFIkocqZc5Omj7G/YsGH2omVCU8dRffnMIlSElEhpfjpSJLA0f2mSlzD6wz/8Q+zdu1dVpW3cuBGNMqMBMq7hEN599120traqjyGkqOQwNkPSarP1+kz0C/n8xxAIdKLBVqFFqAS9Ey5LHB2B0+p8IYWReIpEnEikKBOr9HHy+9X5GQujLBG3g4MX4vCR48WNUBGySEvz054iiRSJKJLSfKlCY7SohPsYCdLg8Yc//KH6a1XEkPzSFIH0X//1X+pGSNGZaSPDWeLWA0494E+M35e17Mv5QiJGa4nYTJxPKGvZl/MzZpKmjsETT6Lznb1jESqZgShHWUsESSJYhJQqUn0mIiiZSKj0mRxZmr80yStiFI/H8X/+z/9RvYzkL1lJp/3N3/zN7H5BE5IvMygxz7mqawo8FevQ5KlHW2+HEhIugwH+6Ah6w8NoWb5GnS8kcp2SxkpHbCRSJKJoZGRE/TEy42hRlohbaMSAaOAsXJ71hY9QEbLIMVqtqvqMpflLn7yE0d///d8rIXTDDTeoNxhp6CiVaI888kjxrpCQWZSYz7iqK8vztl50FxD9Z7QPdqEjGoRTn0CL3YzWmjWAofCiIV3MINcv4kSuX0TRrIocsjR1tNmdMOsjCAeH4bSWFzZCRcgSQcQQBdHSRpdMJpO5fvD69evxZ3/2Z/hf/+t/qbVUqP3O7/yO+ktSL7Mbljgy9kR8F/IGJZ2+yTySxSOjSswzxlQcOHBgrKprYsRlRp4Zed5XPwvv6V/Al9DBbS2Dx75MTAdA/Yenn2A/Q4KDpxHy98LmqoG9YuXsHkwiRjIoVtJo6YiREO7FgdNmHA5vg6O8sjBfL0IIWWTv33lFjGTcx6233jq2lsiRTqdDV1cXVq6c5S9rQvIhh0aG51V1xSTCI+m3CDpPHplZVVc8qAaweVZcD4+xTLsOiRRJtCpl/C5oU8SUALT3tcMuArBncgGYS8VZLhG3zRfdBgQvLmyEihBClnKDR/krMhNJR0i6gpB5YYpGhmNVXWV2wPcaMPAGMDoEa0IPf6IaoUP9sF/yyfwGoWamoTL9TdlmixWpLYEiW3Qql2halqaOpoZWbDXalGicrSeLEEKWvDCSrNsnP/lJWDL6OEio/Y/+6I/UX+VpfvzjHxf2KgmZTVVX/9twjuwD4iElZsIJC8zJQdhOPwZU2PNLf83lbLEc2hJMKsJyEVPTRNxEDFEQEUJKkbyMQdI4r7q6WuXp0rePf/zjqKurG7e30HnyySdVekA8U//+7/8+35dDioSq6qqtxMhgDwIRPWI6BwIxJ0biTtS7grDrhoCePZoAyZV0GkpSZ3KLh8/dl/1CRotm0pZgopgSIZW+L/sTX6tcb/l6zkQjhJCZRIweffRRLHYkHSjdu5999lkl4mS8yR133KEaApKlx+a1HuBoHzr79fCP2mE2JNDo7sNmt4gOANGB/NNfczVbbCbRqSwVZ0VL9RFCSCkLo6XAyy+/rIykK1asUOtbbrlFzXi788475/vSSBEwOaqxtSGBDZbXEUIFbBYT7KaYMmIrzJX5p7/ynWA/B20J5iXVRwghS5BFVWMv40huu+02lbqTargnnnhi0o976KGHsHr1amUUv+qqq5QYSiMVdGlRJMj9M2fOzMn1k3lAxMPyHbBbjfDoOmCHVxMJUZ8mcJbvmLmomYs0lEShRAQl41rER45TRafmMtVHCCFLkEUVMZJ+Klu2bFHDbD/0oQ9N+jEyrkRSZQ8//LASRQ8++CBuuukm1ctG/FGkBBERkRgFTnwXCJ7W9iTVtOauwqe/Cs1MolNzleojhJAlyKISRpL2kttUfOMb38A999yDu+++W61FIP3sZz9T3bm/8IUvqGhTZoRI7l955ZWTPlYkElG3zAZRZBEi4uKCTwP1dwDDx7S9snWLK3oyRVuCeUv1kQVNZHCQoysIWerCaDqkZ82+fftw//33j+1JR25pRPnCCy+otYigt956SwkiMV8/9dRT+OIXvzjp4z3wwAP4yle+MmfXT2ZGznPQ8hEXS4FSe71EEQuH0bF7N/r270csGITRbleT4WXYqcz7IoWHInRpsaSEUX9/vxp0W1OTMeYAUOtDhw6p+zIx/B//8R9x3XXXIZFI4C/+4i+yVqSJwJK0XGbESOZskYVBQeegEbJEEFF0+plnYPV4YK+txejwsFoLMgSVFA6K0KXJvJuvJb0lRuqpbmlRUyg++MEP4t1338XRo0fxh3/4h1k/ThpZykyVzBtZOIgoEu+Y/IxI9E+OspZ9Qko1ciFv0iKKrFVVMFgs6ihr2ZfzpPAiVKfXKxEqR1nLPlm8zHvE6POf/7zqpj0Va9euzemxquQXgcGA3t7ecfuyXr58+ayukywszpuDBowdZX9Gc9AIWeRIOkciF/ImnYmprAzB7m4t3cNUT1FEqCBCVJD9Fdu382u9SJl3YbRs2TJ1KwSSSpGGjXv27MHtqZCxpMtkfe+99xbkOcjCYGwO2oRO69KiQYafynkKI1JqiMdF0jmSPku/SQuyln05TwoDRejSZd6FUT4EAgGV/kpz4sQJHDhwAG63Gw0NWqdf8QTJ6JJt27Ypo7WU60uZf7pKjSxQvF7A5wPcbiCHLuRjc9DC4bFIkSBr2ZfzhJQa8kYsHpe0p0jepEUUhb1erLz++pJ+oy60QZoidOmyqITRq6++qkzTadLGaBFC3/nOd9T9j370o+jr68OXvvQl9PT0YOvWrdi9e/d5hmyyQAiFgF27gPZ2Ub6SDwOamoDWVlE/U89Bq69XnqJ0pEhEkYhgmYPHaBEpVcT4m07nSORC3qRFFKX3S41iGaQpQpcuumQymZzvi1gsSFWapG4kVUMjdoF47DGgrU1KBwH5C8vvF1MY0NIiinfKT2VVGiHZYQm5xvEnnhir0psoXmZbpceqtKX5/k1hlAcURkVIn0nPKYNBE0ZpRBjF49JIKqe0Ws59jAghJScOD/y//6+qFksbpIVwfz+SiQS2/umfFkQ0UoQurffveS/XJyWMeIokfTYxFy9r2ZfzOSBiSHpRURQRQiYzSEukKBNZy76cLwQihspXraIoWiJQGJH5Q4zW4ima+MtJ1rIv5yNeYOiIdiSEkBkapDOhQZosGfM1WWJImkyM1uIxEsZ5jG4B/E8CR9uBWEAbhCrT4WUQqswCI4SQaaBBmswERozI/CLVZ2K0Fk9RR4d2lPVVOuB0G6AzAPYGBEeN8L77FIJH/mu+r5gQsogQI7SIIPEUSZWeHEu5So9MD83XeUDz9Rz1MZK2RAfuV6Jo1FSDg6eBTq8MCQ7DbEigfssHsXnLVaw8I4TkDA3Spc1QHu/fTKWRhZNWS1egiadI0mf2BiWKDncDDgvgshsRDgdw+NBhlVqTHlWEEJILIoYoiEguMJVGFh4WtxI+wZEhFSkSUeS0AkZdBE6rAY4yl+pdJGX6hBBCSCGhMCLzw1TVZhaPMlqHAj6VPrMaYkBsRIsiOephdVSoho7SuyjvdN2RI9qREEIImQSm0sjcEgsBHbuAvmmqzRpaYYvEYT5zSKXPJFKE8kbAtRnhYJ7z0GY4doQQQkjpwYgRmVtEFGVUm6mjrGU/E6MN9k2fUkbrEevFCJRdjVjZRQgEI2oemoz+mLSh42SRKBFF0hJAOmzLsGE5ylr2CSGEkAwYMSJzh4gViRRZa7SbYEjNE5L9Fc1aGi0DqT6TqJJ4iqSaQCJFMiRW5qHlFIly7tAiRTJyJD12JD3DSPabmzXTd2ZVXA5jSAghhCxNKIzI3BHxjVWbjcPkAoId2vkJwkhK8qX6bMOGDVPPQ0tHokRwyeOP+rW1sVdLn0mkKBNpJil9k7q6gCefZJqNEEKIgsKIzHm1mRIt6UiRIGvZl/NZEDGUdRbaVJGo0FtAhVHrqJ057To9dmTvXmDPHi2aJOJJ9tOduHfuxEKCfVgIIaT4UBiRuSNVbaYiOelIkYiicC+wsuW8aFFBIlHy+FdtAdpeOn/syI4dwOuvT59mm2di4TA6du9G3/79avClzHiSMQfSudeYKfYIIYTMGpqvydwi1WcigpJxLX0mR1nLfiEiUZmkI1Etd00+duSaa7T02cRBkrKWffEcLQBEFMmsJ51eD3ttrTrKWvYJIYQUFkaMyNwiJflrd2pGa+Upcs88UpRrJKpipZYWkwhQpsFaDNeSTsuWZpOPKyLeoBe+kA9umxseuydr+kwiRVaPB9aqKrVnsFjUUfZXbN/OtBohhBQQCiOSFwUr3hIxk4sgyvUJ0xEn8RpJJEoiRRMjUZljR9JrMVqnPUWZaTaJKBUpjRYaDWHXS4+g/eReBPQxOMs8aKpvQuvmVthM4w3f4imS9JlEijKRKeEyEFN5jiiMCCGkYFAYkYXZIzHfJ5xpJEoeT5DnkTSbPI+IovR+oQmFsOs/Pou2zqdREzGiweCAv86HtuEedXrn1vGGbzFai6dodHh4LFIkyFr25TwhhJDCQWFEciLdI3HOirdm+oS5RqLSiMiaLM1WJLz/9Qja3/0Vamx21DjdQCQM61FNFLWXtaN5Q/O4tJpEg8RoLZ6idKRIRFHY68XK669ntIgQQgoMzdckp2xWZo9EseOk78t+wUePzfkTptJq69cXtwrN64Xvlb0I2E1w2cQwbgQcTsDpgKtrAIFhzXM0Eak+ExGUTCRU+kyOspZ9QgghhYURIzItEkiZqkeinC+onpjzJ5wjfD64h2NwVtnhRxhWOLV9ixX+QC+cCaMyYk9ESvLX3n67MlqzjxEhhBQXRozItFPnJbuULt7KpGjFW3P+hHOEW6s+awp40IsR9CKAMGLoHfWh1xJD0+prslanCSKGyletoigihJAiwohRKZOjwXnOi7fmqVqs6KReV+tPeoEGoN3pQ8doL5yhUbRs+ABar/rUfF8hISUHO8qTiVAYlTJ5GJznungr/cDBF15AqLtbm5FW1CecI1pbIZJzZ3s7mvu88JUZ4b7iGnju/BQwoVSfEFI82FGeZEOXTCaTWc+ScQwNDcHlcqkp7+Xl5Vh0ZPYEEu6/HzAYzo3DECQqI52hH3hg0sjMXA2hHx0dxcGDB9F55AiiIyMwOxyoX78emzdvVoNlFz1z9YUkhEzK8SeeUNWe0jx1YrWnePpI6b5/M2JUqimzdeuAwUFg7dq8DM4TeyQWCxFFhw8fhsPhgKuyEuFwWK2FrVu3YtEzV19IQsh5sKM8mQqar0spZSbRIUmZyVGmyp89uyANzsFgEJ2dnUoUOZ1OGI1GdZS17Mt5Mrs3haFTp9SRkFIk3VFeIkWZyFr25TwpXRgxWupM7AkkpPPn77yjRYcWmME5FAohGo2qsGcmVqtVhUHlvN1un7frW6zQU0GIBjvKk6lgxGipk+4JNNkE+epq4Nprz586P88GZzFam81mlT7LRNayL+dJ/ogoEk+FTq9Xs9fkKGvZJ6SUSHeUF09RuL8f8UhEHWUt+0yjlTaMGC11MnsCTZwgL//4/+APtHW+RuAimoclGlRfXz/mKZJIkYiikZERNDY2Mlo0A+ipIGQ86c7x8vMvHeUlUsSO8kSgMFrq5NoTKHWcVu/M0TRZqT4TxFMk6TOJFIkoSu8j4s1vWGyJk/ZUSKRooqdC3hRUHxcKI1JCsKM8yQaFUSmQQxOinPXOVL2PPpbnZPspkJJ8qT7bsGGD8hSpPkYSKYqFgOPfB/ragVgAMDqBZU1AQytgZIotG/RUEDI5IoYoiEgmFEalQA4T5HPq9ZjNyK2PAUcfBV7cAxhjBRUrIobGpc46dgGn2wBrDWBvAEb92lpYO74pJTnfUyGeImFi3xa+MRBCiAbN16VElgnyOQ+zz2bkru8CKo4B0VFNrOgMmlgREVNIJH0mkSIRRXIzWM/dl305T7Ii3gkRQclEQqXP5EhPBSGEjIcRI5L7MPvJjNz6IGA6DiScQHm9JlbkJohYWdFcOA+QpOkkfSbiKxOTCwh2pNJ49Btlg54KQgiZHkaMSghpjOj1es9rkDhxmL1N74XbdASxEe/4Xo9pI7cYt+Um5fQDnWJUAZavBWz28WJFRIyIlUIh3iVJ00n6LBNZy76cJ9MiYqh81SqKIkIImQRGjEqAsbljnZ2qcaJUeEk5vJo7NjQEj8+Hpi3L8dSzRlxVvQubKtuhiwewscoGm3sLPPEPAVg5uZG7wgSsvABYW1d8sSLRIPEupT1FIr7kecK9wMoWRosIIYTMGg6RLYEhsgcOHBibOzbWE2hoCI0+H7a+8YbKo4WslXi1rgKOqoPwjtQgNpjAGssRXFBxHEb/WuCCu+Hd0QpfyKZ5t+HFy88F8HaXG+/f9GOsM6cM0RPFSqEN0VKVJt4lVqURQgjJEQ6RJVnnjgnqePgwOk+dwgajEfaGBthGevF+688QKl+PoGUQtuFDsDtMKgoTK+/Hf//rGTzzaCcCdRug0wH79nnQ3e1BLAaU2Vvx2RbgT36vHWHdEfiSRrhrd8AjYqXQiPgRsSXeJfYxIoQQUmAojJY4k84dCwZh7eqC325HyGqFcgaVOQGHCbaubtgSZwGXSTMe6WLoiOnw6+hmGHxdWLPGhJ/ujqO7341Rk1s1zw6FbPjKD1rxo3AAV9yxF4F4DM6R17Fl+BFcs+oa1JXVwWPPUbzk2rhRzlEQEUIIKTAURkuczLlj6YiRdHMMx+Mw2+0YSz7FbYDBDsQHgbAFSE2dDiVGcXpwhRJPy8NtSP52GNsNLoSX2XE8uhXHdLfA7LIjUrcLb2EP1vhrsGaFHa91v4ZfHfsV/u8b/xeX1FyCpvomtG5uhc1km7y9NlNkhBBCFgAURkucSeeOJZMYsVhgHTiDJ/SDWGH24ArnChjOVCJiHYTFClgiI4DFgFg8ioMdl8Hj/S2ceBe9iTXoG61FmcGPxsqnEDecxmHTrUB9O5KBGiSGanDMfgA9gR4lgsKxMEbjo2g73CYucOx82zR5e+1uNm4khBAy/1AYLTUmicZI9VkgGsDJUyehH9FDP+LHCziJSKUDdn0DDsCAn0W8uOD0MmyMABbTEJY5BtBwgQ2IbsSh1y9EQ/y/Ebe4YLC6EB0BDlxwAkN1byNp/i2G8Swi9h4YzjShzB3E2/5OOMwOWI1W+MN+OM1Odb9973fR/KIDnqqGc+21peV233HgihOALdWsUShSLyTxXI0bMbKY4Hw4QggpOhRGS4Usw85Cd9yGXcd/ivbedgSjgyjrOo7+ASOqrVvgggtmnRl6GAALcGZdJQKRddiGURw8sh8n9iew1ajDZY530B02YtRUBbtpFL3r9+HMqrdhD9lQOaxDl8mAhKcHZWtfQ0XVRYiejMJldalokdlgVpEjfSiCjrOn4au5Ap6qGhU9Qnc3cPw40PemKCGg5kIE17sRSphgMwH2AjZunLJlgcmEBQ3TjIQQMmdQGC0Vsgw72zX8G7Qt86LGUYOVfYdxKtiBpGk77HDCorNABz2SkI4NSeh0egybzfhV3I83L/EjGAuhfk0ltt0eQ/XXgvAGg+hOOBFadQjlURvsETPiMAOhC+Ba7kVZ3RH0D1TCoDPAF/JBOkGsd62FYSSK/oHTcEYBt7Nau96DB1VlnOqgHY9jNAocPB5GZ98Qgk4P9DpglSuKS1c6YSpALyQRRemWBWJEF89VOr0ow2oXNJwPRwghcwaF0VIgy3DXI/pB/LxzD1xVV6ImaQSCZ2AcdcBscMKoM48TRYLcT+iSiBrs0MMGa2IUfkMMj4fa0fSBOJqe7sbpSjP2u/pxUVCPCPSIlK/DZbWdKA/40D4QQmLgAKwVOnjNEbhGLYj0d+D12CkEdVE0D5TBZQ4BFRagsxNwOLRrTZbhYGIz3jGaEAnEEUrEMZpI4ozXip7ROtxyRTlMs0iJZW1ZALmMTmzYsGHhptUmzocTijVyhRBCCIXRUhx2FsIoduEgfu4+hBeTvajoOYABuxubE1GUmWwYio9Cr0tPgznX31MnK50c9eiOD+OI6TiS3adhNJhwsqYMuGEVLtsfgi0cxbDBiPp1q+CqsEB/fD96y3W4xFyPP/ZtwP/438Cj1V4c153Fcb2kxKx4T2w1NvZVomP4ENauWKGl/iwWNVYkeOGF6IyvRgQBBACYkxFYzEaEkg4cP2vA/v37ceWVV844JTZpy4KUEV2afcn5hSuMOB+OEELmEs5KWwpMGHYmoqgNh2GLJlGZtGBUn8ThYC8OJhOoNoXhjZ9ELBlLfbJOCaFxN50RXTovhnQRRBOjiMaj8I8Oo/1CA85+/kO4dccfIP6ezYheugHRnjNKFPXakmgyrMZry6L4ibsfo5EwalGO5aiEaTSB46GTeMXeib5AAJGTJ7U5a5EI0NiI0MaNCCb1CI3aYR61wFy2EvryNbC5aqE3GHDq1Kmx+W7plJhOp1NCR46ylv1cWhZkImvZl/MLFs6HI4SQOYXCaCmQMdzVe/YU2uMnUBPSYfWQHqtdq5HU66EzWXDSYMapeAB6/RvojB1CHDHolRiS/7RUmqBP6nCpfhuMRhPiyThiiZg6esxOvDT0Cpp33IkPXXYX4uEQOjCAuNmMFjRiB9ZgD05gRB9H+agBFXDAHdGjMgSEDXG8ujwEr9uGiPRI2rYNWLMGqK2FLR6HPixl/XEYHeWAtRzQmxCLxZRwSSQSKqozMSVmNBrVUdayP3E4ruANenEmdAaVNZUYGRhA4MwZxIaGEAgEMDIyoqJNCzZalDkfTkasyC0ePndf9hktIoSQgsJU2lIhNdzV9+LPEQgOogGVQONqbN7UCAwcxsmBkxiwOhCMluHWwFn8xvQLnA4F0WB9jzJda8IogaAuhLAujHrUw5P0wKvzqoiR02BAWeBtBEZDCL35VexcfSuar/lr+Nq/CnfIDk/1KhyBFwMIyQA+WJJGJBNx6MIRWHQmhPUx+A2jGHGZYalsABIJ4Nprgddfh72jA6uqq3GmqgohpxO2REKJIkl/lZWVKeEiUZ18UmKh0RB2HdyF9s529fVwdvbjkt5lMMeWI2K2wlxdjcZrrlEpuAVPerSKeIokfSaRIplDV4yRK4QQUuJQGC0VJB20cyfcO94L53N/Db/VBmvVamVa3rp8KyqsFUos/PXqj+OpR/8cL1edRr21FzW6EBLJGExJHQZ0I4joRmGAEU44YY1boTPqYNQBFbo4oskknJZKuM12VRXlWQl43nOrVg2X7IXbZUZlEtDp4kiU2aEbjQGJOMJmIJFMwhHRY62rBpaqKqCjA7jxRuBjH1MeqUvLy9Hz0ks4fvw44oGAihSJKEp7iNKC57wu3llSYiKKpKmkVOM1dI3Af+oEnnGcRLP5YrT462F78UXY5TquuAILHs6HI4SQOYOptCWGZ+V6NF10K3rjQ+gN9KpeQnIcigzh1vW34rVIB54vD8KcMGJUF1bptJgugSFdAKO6mPIY2WCVXfjhRzKZgDGZgMfiwFBCj6aq1fCUr9IqpCSC0bIDaGlRJfeejj7sCNfBXu7BcIUV0TILAoY4/LoIrHEDPqBfhy0r1mteKBE26SaU69fDVFODW265Be9973tRW1uLyspKuN1ubNq0aSyqk+7iLSkwSYVJVGmylJikzyRSJKKoRl8G6+ke1Fg9qLG68aLpNFBl10SRVPJJRd9iQcRQ+XqKIkIIKSKMGC1BZCaZIOKgw98BZ8KIlor3YEfFZfjqiYewxnMRrN1v4217H7yGLlRjFWKIw4AILAkLrLDgBI7Ar/chHk+iymzBhrJqXLtsLVpXbh5fFaUPqUgVmptV5Ke13IbR07vx3Te/i9NDp5GM27B2IIG7YpvwybKrYOzvV14oJabSc9JSSFWZVJ9ddNFFWUvx0yJJPEWSPpNIUWNj47iUmPRQkk7fDS7p5zQCRKOAS9pZAh3ww4cQPJKOk6iVVPRNuA5CSGkQGRxExO+HxeWCRSZiE0JhtDSRTtM7t+5Ec8MO+H78PbhfPYjy4dfxTPUR2DwJOMsvhbtqLcpH3sHbkV8jYAqjxtAAZ9KJeDKGk4l38Zv4szDpk7AYjdjoqMTlTpcSRTaDafKqKBEXHo8aSvvpmk/jjk134NjAMSAcwbrn34DnxdeB/m4tUiSiKOWJmgwRQ9kM0SKepCGj9B7KJp7cNrcaQyLjSKy2Msm/AZEw/EbACTPccpWZUStCSEkRC4fRsXs3+vbvRywYhNFux7JLL0XDzTfDKE1nSUlDYbREkXSSEkW/2Ktmkx1Y7cbZxCiMgQ6Edd1wLGvAKsd7ER804enwbhgTZVgdrcJIbAgnzL2AwQD3qAWmsA5llnLs6TsGp9GEnWuu1ESRVEWJAXiKtE6ltRLuSjc8n3o/0HL+DLfZMJV48tg9aKpv0gbXOgDXyuXwH38bvQYdWgyb4PENZ41aEUKWPiKKTj/zDKweD+y1tRgdHlZrYe3tt8/35ZF5hsJoKeH1InS2C7u8e9He/RK8R38N4zo93mc1YCWWSUchrIjYcDg0AMTKYdSbkIw74BIfzvAIOswBmHV6JAw6JBCHTm/GyhErKofKoKtbhvaBLjRXHlHCI1tV1LhqsGhARW5EpEh6zzaHImRcOnGFE87EKrR0AK39DsARnzZqRQhZuukziRSJKLKK11CayUuzWSn83L8fK7ZvZ1qtxKEwWmIDZHfp3sD/VHRj2GHEAAYwYtHhHbyBW1CD7ajFZkMtEB5FZyyGgeiQqkDbhBWo9/ZgvyOMM2VRxPRJ6BJAbciMCyLlSJTp4CrbjN74MvjW/hE8yy/LGikaVw3malDpLBW5AVR6b87TiRualedI0mueEAoatSKELD7EUyTpM4kUZWIqK0Owu1vzHFEYlTQURktogKx3uQvtnhCG4zH0BL1wjOpQY3AgYIqjEwN4HSa8J+LA1mQNNqx+L3oGz+KtwXbYY3qUOSuwwTeA5SNGvFAbhz4JbPLboC9zImm1IqCLwGnzwL0suygaVw3m1OZ6WZ1avl72RaSoaNMcIs839pySeaMgIqSkEaO1eIokfZaOFAmyln05T0oblusvlQGy5eXwmWLwIoQBcwwOow3OuAHGcBTO0VH0J7vQF4+gPzKCWHU1EsNhOHROrK1woyd6Fl6nDkZDEpHEKBwhwB7VwW8GghYjQh4bvKODKiU2lbBJV4O5rON/scha9uU8IYTMJxINEqN12OtFuL8f8UhEHWUt+4wWEUaMFjtdXcAbb6jZY25DBMbL+zDiiaLG6ALMMka+DOHIEILRwxiEFWHbxfB7vTD39KAxkcAtmxtRZroJ7Qd3Y9gUhC1pwodOL0fCYsah8gCGrWFU16/AjauvGfPtZGNcNVgqUiTIWvblPCGEzDdSfZb2FEn6TCJFK6+/fmyflDYURoudvXuBnh4p04LH7cE13hG8XN4FX3QAbpMT4apyjIwaUGv2wOCPYPuBk7BV1cPmdMIuJeu792Dnjh1ojgbh043AXbsGToMTkWgUgeFejHhH4d72BdU4cjrGVYOlIkUiinpHetHS2DLnaTRCCJkMKcmX6jMxWrOPEZkIhdFiT6O9/jqwdi3Q3Q0Eg/jUQRMOQI9fNYQxmkjCMehD7fK1KDNY0XQojJWVdUCqEgPpfh1798ITi8Gz/qKxPYvFgnKTSWuCKKblmTaXNDuVKJou2kQIIXONiKFCCyI2jVz8UBgtZqTCKhAALr1UeYxw4ABsPh/+yefEIxfHsPcSF2J9MXgsQNNF16O1/1WgfoKxUIyGZ88CRqPW9DCzudkMmiBOWg3GSBEhZInDppFLBwqjxdKscTKRIYJFhEswCGzYAJw4odY2oxGfGTTgY/FL4TvbAXeHEZ73bwccRycXP1KptWULsGfPObEk+7NogjiuGowQQpY4bBq5dKAwWsBM2SzRZNMES1OTNt1+cBAYHZUcmCaUxHP0k6fhkf1YDOi8VwaNaX6kycSPNDsUkSUVbpI+y2F0ByGEEDaNXGpQGC1gcmqWmBYuEu2JRIBkEnA4tGq1cFiN9lA3EUAinLZtA+Lx88WPzTZuGCybIBJCSG6waeTSgsJogZJzs8SUoPHueC983/03uPe+DM/RLm2ivJinRSjJXzDplJuIpy9+UXuSycRPahgsIYSQ3GDTyKUFhdECJd0sUSJFmUgJvFR7yXkRRuPSbWsG4YyF0BTyorV/FDaDGZC/UpYt0wSSCKOBAe2B1k9ffk8IIST3ppFpT5FEikQUSdNI6Y/EaNHigsJogRqthVyaJY5Lt3nWwm+wo23gDGA0YGffck0YCVK9JlRW5lVlRgghZHrYNHLpQGG0gI3WyWQSXYGurM0SJ023Va0G6taj3fcmmjt98Ph12hOI0VpE0Y4dTJURQkiBexKxaeTSgcJogbDrpUe0yI+rDg0VmtH6TOCMEj3xZHzSZolZ020XXoaOeBC+niQ8R1LzyRoagLvuYpUZIYQUsSdRMZpGkrmFwmi+CYXg/a9H0H7om6iJxlFj8AL1g7BKaT2gRNGfv/fP1f2JfYyyziaLB+HctAXu3/tzoCvlKVq3jpEiQgjJAfYkKm0ojOabXbvg+2UbAqvjaHB6gEgMOHxYnXJdtFFFioT1nvX5zyaT+WYr5/j1EELIIoY9iYh+vi8ApT7rrL0dbvcKOK3l8BtjgMOJoNMC75kj6O0/Oe1UekmriQhKp9vkyNlkhBAyu55EUlmWiaxlX86TpQ0jRgtg1pmnoQFbUIMf4x28hbMYsocRjIUw2jmIDzTeCrvJnvUhOJuMEEIKB3sSEUaM5hO3GyGnFY9FXsY+dOMkBnEAPehJDsMGI9ZUrlFpMalWmw4RQ5JuoygihJDZ9ySSHkTh/n7EIxF1lLXsM4229KEwmk88Huy6xIA2HEIyNAJn3IC6qAXOUR1Wljfg2gtuwIqyFaokX0rzCSGEFB+pPpMeRMlEQvUkkiN7EpUOTKXNI6oP0Yo4akIbYDx8BAnnEJZF9QjbTOiNDSIYGjqv0/W5T/ZyphkhBKXeP6gYsCdRaUNhNI+oPkSJMBqMDiSMFphtDoSdJliTeviHBxB65w3ENqwfb8AOhVQlm5i2VTdrmYHW1HRuECwhhJRg/6BiwJ5EpQlTafOI6kOUMMLfdRz2MjfqLTUYMSbgM8VhMJow3NOB3oEOVZI/Fi0SUdTWBhgMWtNGOcpa9gkhZAn0D9Lp9ap/kBxlLfuEzBUURvOI6kNUfhF6kwH0WuJYh0rUogxBjMKqt8A8mkBL9bXnSu9T5f2oqdFu8hdU+r7sy3lCCFkC/YOkIkyOspZ9OU/IXMBU2jzTuvX3gJ//HO2Gs+g2xnAB3PgwNuEanxN1MSs82/4AMNnGlferSFEmUj7a0aGdp9+IELKI+wdJpGhi/yAxQCuvD9NaZA6gMJpnbDUrsfOyu9H85C74asrhdlbD448Cvb1Ay4SBr2K0Fk+RNBjLzLfLWvblPCGELELYP4gsFJhKWwi0tsLT3Ir10TJ4OvqAeBxoaTl/4KuIJDFai2iSWzh87r7sM1pECFmksH+Qlk4cOnWKacN5RpdMJpPzfRGLhaGhIbhcLvj9fpSXlxf+CXIpwWdVGiFkibLQqtLmilJ93Qv1/ZvCaCEJo3xgHyNCyBJlvvsYzTXHn3hCVd+J0Vw8VZI+lEiZNJWUfkpkbt+/6TFarIgYoiAihCxBSql/0MRqPCHtsZJ9aTJZKl+LhQI9RoQQQsg8V+NJpCgTWcu+nCdzC4URIYQQsgCq8TJhNd78QWFECCGEzBOsxlt40GNECCGEzCNSfZb2FEkzS4kUifE6vU/mFgojQgghZB6RknypPhOjdSlV4y1UKIwIIYSQBUApVeMtZOgxIoQQQghJQWFECCGEEJKCwogQQgghJAWFESGEEEJICgojQgghhJAUFEaEEEIIISkojAghhBBCUlAYEUIIIYSkoDAihBBCCCllYXTHHXegsrISv/u7vzvfl0IIIYSQBURJCqM/+ZM/wX/+53/O92UQQgghZIFRksJo+/btKCsrm+/LIIQQQsgCY1EJo7179+K2225DXV0ddDodnnjiiUk/7qGHHsLq1athtVpx1VVX4eWXX57zayWEEELI4mNRCaORkRFs2bJFCZ9s/PCHP8R9992HL3/5y3jttdfUx9900004e/Zs3s8XiUQwNDQ07kYIIYSQpcuiEka33HILvvrVryrzdDa+8Y1v4J577sHdd9+NCy+8EA8//DDsdjseeeSRvJ/vgQcegMvlGrvV19fP8hUQQgghZCGzqITRdESjUezbtw833HDD2J5er1frF154Ie/Hu//+++H3+8dunZ2dBb5iQgghhCwkjFhC9Pf3Ix6Po6amZty+rA8dOjS2FqH0+uuvq9TcypUr8aMf/QhXX331eY9nsVjUjRBCCCGlwbxHjL7whS8oI/VUt0xRUwiefvpp9PX1IRgM4vTp05OKIkIIIYSUHvMeMfr85z+PT37yk1N+zNq1a3N6rKqqKhgMBvT29o7bl/Xy5ctndZ2EEEIIWfrMuzBatmyZuhUCs9mMyy+/HHv27MHtt9+u9hKJhFrfe++9BXkOQgghhCxd5l0Y5UMgEMDRo0fH1idOnMCBAwfgdrvR0NCg9qRUf+fOndi2bRuuvPJKPPjgg8pLJFVqhBBCCCFLRhi9+uqruO6668bWIoIEEULf+c531P2PfvSjyj/0pS99CT09Pdi6dSt27959niGbEEIIIWQiumQymTxvl0yKNHiUfkZSul9eXj7fl0MIIYSQAr9/z3tVGiGEEELIQoHCiBBCCCEkBYURIYQQQkgKCiNCCCGEkBQURoQQQgghKSiMCCGEEEJSUBgRQgghhKSgMCKEEEIISUFhRAghhBCSgsKIEEIIISQFhREhhBBCSAoKI0IIIYSQFBRGhBBCCCEpKIwIIYQQQlJQGBFCCCGEpKAwIoQQQghJQWFECCGEEJKCwogQQgghJAWFESGEEEJICmP6DiGEEELIfBIZHETE74fF5YKlomJeroHCiBBCCCHzSiwcRsfu3ejbvx+xYBBGux3LLr0UDTffDKPVOqfXwlQaIYQQQuaVjt27cfqZZ6DT62GvrVVHWcv+XENhRAghhJB5TZ/17d8Pq8cDa1UVDBaLOspa9uX8XEJhRAghhJB5I+L3q/SZqaxs3L6sZV/OzyUURoQQQgiZNywul/IUjQ4Pj9uXtezL+bmEwogQQggh84alokIZrcNeL8L9/YhHIuooa9mf6+o0VqURQgghZF5puPlmdRRPUbC7W0WKVl5//dj+XEJhRAghhJB5xWi1Yu3tt2PF9u3sY0QIIYQQIogYmi9BlIYeI0IIIYSQFBRGhBBCCCEpKIwIIYQQQlJQGBFCCCGEpKAwIoQQQghJQWFECCGEEJKCwogQQgghJAWFESGEEEJICgojQgghhJAUFEaEEEIIISk4EiQPksmkOg4NDc33pRBCCCEkR9Lv2+n38amgMMqD4eFhdayvr5/vSyGEEELIDN7HXS7XlB+jS+Yin4gikUigq6sLZWVl0Ol08305ZI7/2hBB3NnZifLycpQq/Dpo8OtACg1/poqLSB0RRXV1ddDrp3YRMWKUB/LFXLly5XxfBplH5BcWf2nx65CGXwdSaPgzVTymixSlofmaEEIIISQFhREhhBBCSAoKI0JywGKx4Mtf/rI6ljL8Omjw60AKDX+mFg40XxNCCCGEpGDEiBBCCCEkBYURIYQQQkgKCiNCCCGEkBQURoQQQgghKSiMCCE58dBDD2H16tWwWq246qqr8PLLL8/3JRFCSMGhMCKkSDz55JNobGzE+vXr8e///u9YzPzwhz/Efffdp8qJX3vtNWzZsgU33XQTzp49O9+XtuBYSt93snC44447UFlZid/93d+d70tZ8rBcn5AiEIvFcOGFF+LZZ59Vbegvv/xy/Pa3v4XH48FiRCJEV1xxBb71rW+NzQ2UuU5//Md/jC984QvzfXkLhqX2fScLh+eee07N+nrsscfw3//93/N9OUsaRowIKQKSZtq8eTNWrFgBp9OJW265Bb/85S+xGIlGo9i3bx9uuOGGcXMDZf3CCy/M67UtNJbS950sLLZv364GmJPiQ2FEyAT27t2L2267TU1h1ul0eOKJJ/L23HR1dak3xzRy/8yZM1iM9Pf3Ix6Po6amZty+rHt6elBq3/tS+b6ThfNzReYWCiNCJjAyMqI8NPKLKhv03JTm957fdzIT+HO1uKAwImQCkv746le/qsyO2fjGN76Be+65B3fffbfylDz88MOw2+145JFH1Hn5yzAzUiD3ZW8xUlVVBYPBgN7e3nH7sl6+fDlK6XtfSt93snB+rsjcQmFESBE8N1deeSXeeust9cYYCATw1FNPqb8AFyNms1mZiPfs2TO2J+ZrWV999dUoFUrt+07mBnr4Fh7G+b4AQpaS5+bQoUPqvtFoxD/+4z/iuuuuUyLiL/7iLxZ1ZZKE+Xfu3Ilt27apN/8HH3xQpQfkL9xSoRS/72Rh/FwJIpRef/119e9u5cqV+NGPflRSf5jMJRRGpCSQkvKvf/3rU37MO++8g40bNxbsOT/4wQ+q21Lgox/9KPr6+vClL31JGa63bt2K3bt3n/fLnCyt7ztZODz99NPzfQklA4URKQk+//nP45Of/OSUH7N27dqcHquUPDeZ3HvvvepWqpTq950UF/5cLTzoMSIlwbJly1Q0aKqbeGlygZ6b0oTfd1IM+HO18GDEiJAJiGn26NGjY+sTJ07gwIEDcLvdaGhoUHv03JTm957fdzIT+HO1yJCRIISQczz77LMyJue8286dO8d93De/+c1kQ0ND0mw2J6+88srkiy++OG/XTObue8/vO8kX/lwtLjgrjRBCCCEkBT1GhBBCCCEpKIwIIYQQQlJQGBFCCCGEpKAwIoQQQghJQWFECCGEEJKCwogQQgghJAWFESGEEEJICgojQgghhJAUFEaEEEIIISkojAghS5ZPfvKT0Ol06ibDOi+44AL87d/+LWKxmDovjf//9V//FVdddRWcTicqKirUvCqZVRUMBtXHHDx4EB/+8IexevVq9ThyjhCydKEwIoQsaW6++WZ0d3fjyJEj+PznP4+/+Zu/wT/8wz+oc7//+7+Pz33uc2hpacGzzz6rBnt+8YtfRFtbG375y1+qjxGBtHbtWnzta1/D8uXL5/nVEEKKDWelEUKWdMRocHAQTzzxxNjejTfeiOHhYfzpn/4pPvrRj6pzIowykV+LQ0NDcLlc4/YlaiRCSm6EkKUJI0aEkJLCZrMhGo3ie9/7HhobG88TRYKkzCaKIkJIaUBhRAgpCSQK9PTTT+MXv/gFrr/+epVaE2FECCGZUBgRQpY0Tz75pDJWW61W3HLLLSp9Jj4juggIIZNhnHSXEEKWCNdddx2+/e1vq6q0uro6GI3ar70NGzbg0KFD8315hJAFBiNGhJAljcPhUGX6DQ0NY6JI+L3f+z28++67qgJtIhJN8vv9c3ylhJCFAIURIaQkaW1tVWm1O++8E3//93+PV199FadOnVKptxtuuEGV7wti1JYyfrnJ/TNnzqj7R48ene+XQAgpAizXJ4SUVLl+JolEQjV4fOSRR1QjR4korV+/Hp/4xCdwzz33qAq2kydPYs2aNed97rXXXovnnntuDl4FIWQuoTAihBBCCEnBVBohhBBCSAoKI0IIIYSQFBRGhBBCCCEpKIwIIYQQQlJQGBFCCCGEpKAwIoQQQghJQWFECCGEEJKCwogQQgghJAWFESGEEEJICgojQgghhJAUFEaEEEIIISkojAghhBBCoPH/B6ZSJGEsQ49GAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "\n",
    "pca = PCA(n_components = 5)\n",
    "X_r = pca.fit_transform(X)\n",
    "# X_icb_r = pca.transform(op_matrix_sig)\n",
    "\n",
    "print(pca.explained_variance_ratio_)\n",
    "\n",
    "colors = ['red', 'blue', 'orange', 'brown', 'green', 'gray']\n",
    "# plt.scatter(X_icb_r[:,0], X_icb_r[:,1], s = 40, color = 'gray', alpha = 0.8, lw = 1, label = 'ICB')\n",
    "\n",
    "for i, color in enumerate(colors):\n",
    "    plt.scatter(X_r[y == i + 1, 0], X_r[y == i + 1, 1], s = 20, color = color, alpha = 0.5, lw = 1, label = i + 1)\n",
    "plt.legend(loc = 'best')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.xscale('symlog')\n",
    "plt.yscale('symlog')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6db7e40e-c8c9-4181-99b2-77557d5f560e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TP53_mut</th>\n",
       "      <td>0.031544</td>\n",
       "      <td>-0.146018</td>\n",
       "      <td>0.035855</td>\n",
       "      <td>0.337303</td>\n",
       "      <td>0.182266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSC1_mut</th>\n",
       "      <td>0.199032</td>\n",
       "      <td>-0.136728</td>\n",
       "      <td>0.015428</td>\n",
       "      <td>-0.013127</td>\n",
       "      <td>-0.027541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AXL_mut</th>\n",
       "      <td>0.164256</td>\n",
       "      <td>-0.129870</td>\n",
       "      <td>0.023093</td>\n",
       "      <td>0.009996</td>\n",
       "      <td>-0.000764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUX1_mut</th>\n",
       "      <td>0.173836</td>\n",
       "      <td>-0.123373</td>\n",
       "      <td>0.010327</td>\n",
       "      <td>-0.002339</td>\n",
       "      <td>-0.029673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORL1_mut</th>\n",
       "      <td>0.158930</td>\n",
       "      <td>-0.116540</td>\n",
       "      <td>0.008321</td>\n",
       "      <td>-0.003080</td>\n",
       "      <td>-0.007190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA1_mut</th>\n",
       "      <td>0.160003</td>\n",
       "      <td>-0.114618</td>\n",
       "      <td>0.005291</td>\n",
       "      <td>-0.011872</td>\n",
       "      <td>-0.020714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRF1_cnv</th>\n",
       "      <td>-0.076746</td>\n",
       "      <td>-0.110522</td>\n",
       "      <td>0.100790</td>\n",
       "      <td>0.019081</td>\n",
       "      <td>0.003661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSH6_mut</th>\n",
       "      <td>0.177555</td>\n",
       "      <td>-0.107571</td>\n",
       "      <td>0.008402</td>\n",
       "      <td>-0.029613</td>\n",
       "      <td>-0.001194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NTRK2_mut</th>\n",
       "      <td>0.126712</td>\n",
       "      <td>-0.107226</td>\n",
       "      <td>0.003884</td>\n",
       "      <td>-0.002059</td>\n",
       "      <td>-0.011986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCOR_mut</th>\n",
       "      <td>0.148634</td>\n",
       "      <td>-0.106830</td>\n",
       "      <td>0.016421</td>\n",
       "      <td>-0.020640</td>\n",
       "      <td>-0.025879</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PC1       PC2       PC3       PC4       PC5\n",
       "TP53_mut    0.031544 -0.146018  0.035855  0.337303  0.182266\n",
       "TSC1_mut    0.199032 -0.136728  0.015428 -0.013127 -0.027541\n",
       "AXL_mut     0.164256 -0.129870  0.023093  0.009996 -0.000764\n",
       "CUX1_mut    0.173836 -0.123373  0.010327 -0.002339 -0.029673\n",
       "BCORL1_mut  0.158930 -0.116540  0.008321 -0.003080 -0.007190\n",
       "BRCA1_mut   0.160003 -0.114618  0.005291 -0.011872 -0.020714\n",
       "PRF1_cnv   -0.076746 -0.110522  0.100790  0.019081  0.003661\n",
       "MSH6_mut    0.177555 -0.107571  0.008402 -0.029613 -0.001194\n",
       "NTRK2_mut   0.126712 -0.107226  0.003884 -0.002059 -0.011986\n",
       "BCOR_mut    0.148634 -0.106830  0.016421 -0.020640 -0.025879"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3', 'PC4', 'PC5'], index=list(X)).sort_values('PC2').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b5c90e1e-2ff1-446f-9605-ff1fb57d0892",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CEBPA_cnv</th>\n",
       "      <td>0.101203</td>\n",
       "      <td>0.142499</td>\n",
       "      <td>0.162229</td>\n",
       "      <td>-0.027378</td>\n",
       "      <td>0.043236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRD4_cnv</th>\n",
       "      <td>0.110170</td>\n",
       "      <td>0.143232</td>\n",
       "      <td>0.147439</td>\n",
       "      <td>-0.043123</td>\n",
       "      <td>0.018426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAK3_cnv</th>\n",
       "      <td>0.106405</td>\n",
       "      <td>0.148054</td>\n",
       "      <td>0.159236</td>\n",
       "      <td>-0.030435</td>\n",
       "      <td>0.027668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNA11_cnv</th>\n",
       "      <td>0.109627</td>\n",
       "      <td>0.148060</td>\n",
       "      <td>0.139810</td>\n",
       "      <td>-0.029091</td>\n",
       "      <td>0.035643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEF2B_cnv</th>\n",
       "      <td>0.108499</td>\n",
       "      <td>0.150130</td>\n",
       "      <td>0.159112</td>\n",
       "      <td>-0.035149</td>\n",
       "      <td>0.035892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STK11_cnv</th>\n",
       "      <td>0.111508</td>\n",
       "      <td>0.151525</td>\n",
       "      <td>0.128522</td>\n",
       "      <td>-0.016531</td>\n",
       "      <td>0.043389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRTC1_cnv</th>\n",
       "      <td>0.106974</td>\n",
       "      <td>0.152625</td>\n",
       "      <td>0.155484</td>\n",
       "      <td>-0.031886</td>\n",
       "      <td>0.029999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCF3_cnv</th>\n",
       "      <td>0.111895</td>\n",
       "      <td>0.152776</td>\n",
       "      <td>0.138817</td>\n",
       "      <td>-0.019732</td>\n",
       "      <td>0.033726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMARCA4_cnv</th>\n",
       "      <td>0.115089</td>\n",
       "      <td>0.153593</td>\n",
       "      <td>0.155133</td>\n",
       "      <td>-0.017930</td>\n",
       "      <td>0.034696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR_cnv</th>\n",
       "      <td>0.145292</td>\n",
       "      <td>0.183362</td>\n",
       "      <td>-0.068747</td>\n",
       "      <td>-0.008083</td>\n",
       "      <td>-0.011265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  PC1       PC2       PC3       PC4       PC5\n",
       "CEBPA_cnv    0.101203  0.142499  0.162229 -0.027378  0.043236\n",
       "BRD4_cnv     0.110170  0.143232  0.147439 -0.043123  0.018426\n",
       "JAK3_cnv     0.106405  0.148054  0.159236 -0.030435  0.027668\n",
       "GNA11_cnv    0.109627  0.148060  0.139810 -0.029091  0.035643\n",
       "MEF2B_cnv    0.108499  0.150130  0.159112 -0.035149  0.035892\n",
       "STK11_cnv    0.111508  0.151525  0.128522 -0.016531  0.043389\n",
       "CRTC1_cnv    0.106974  0.152625  0.155484 -0.031886  0.029999\n",
       "TCF3_cnv     0.111895  0.152776  0.138817 -0.019732  0.033726\n",
       "SMARCA4_cnv  0.115089  0.153593  0.155133 -0.017930  0.034696\n",
       "EGFR_cnv     0.145292  0.183362 -0.068747 -0.008083 -0.011265"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3', 'PC4', 'PC5'], index=list(X)).sort_values('PC2').tail(10)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1e2105a6-eca9-4b77-9dd0-3b73aca0f956",
   "metadata": {},
   "source": [
    "# Prepare the ICB OncoPanel data\n",
    "\n",
    "Preprocess the data identically to GLASS... to the extent possible. Some details:\n",
    "- [x] Remove likely germline variants from the ICB OncoPanel data before classification using the gnomAD frequency filter we used for the Cox regressions.\n",
    "- [ ] Add in WES mutational and CNV data for the ~one dozen ICB patients with WES and no OncoPanel\n",
    "    - This would involve preprocessing akin to that done in `aggregate_muts_for_pathways.ipynb`, just without the filters for GISTIC peaks, OncoKB, and CADD\n",
    "- [x] Determine how to handle patients in ICB with multiple OncoPanel / WES samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69cbdfe-01b8-448e-9574-4c9e72e6c51a",
   "metadata": {},
   "source": [
    "### Load ICB cohort sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "4449ed75-ff09-47ac-baf6-b1e49a442fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_path = '../data/sample_cohort_sheet.tsv'\n",
    "pt_path = '../data/participant_cohort_sheet.tsv'\n",
    "\n",
    "icb_sample_df = pd.read_csv(clinical_path, sep=\"\\t\")\n",
    "icb_sample_df[\"sample_id_legacy\"] = icb_sample_df[\"sample_id_legacy\"].astype(str)\n",
    "\n",
    "icb_pt_df = pd.read_csv(pt_path, sep=\"\\t\")\n",
    "# icb_clinical = pd.read_csv(clinical_path, sep = '\\t')\n",
    "# icb_pt = pd.read_csv(pt_path, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c6106d6-49bf-4a52-8405-fe946a2e5ae3",
   "metadata": {},
   "source": [
    "### Load OncoPanel variants\n",
    "\n",
    "...and clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "90f2079c-98b1-475e-9838-5540929b30bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_mutations_path = '../data/oncopanel_mutations.csv' \n",
    "op_cnvs_path = '../data/oncopanel_cnvs.csv'\n",
    "op_clinical_path = '../data/oncopanel_clinical.xlsx'\n",
    "\n",
    "op_mut_df = pd.read_csv(op_mutations_path, sep = '\\t')\n",
    "op_mut_df[\"id_uniquesample\"] = op_mut_df[\"id_uniquesample\"].astype(str)\n",
    "\n",
    "op_cnv_df = pd.read_csv(op_cnvs_path, sep = '\\t')\n",
    "op_cnv_df[\"id_uniquesample\"] = op_cnv_df[\"id_uniquesample\"].astype(str)\n",
    "\n",
    "op_sample_df = pd.read_excel(op_clinical_path)\n",
    "op_sample_df[\"id_uniquesample\"] = op_sample_df[\"id_uniquesample\"].astype(str)\n",
    "op_sample_df[\"id_drspatient\"] = op_sample_df[\"id_drspatient\"].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8a5ec2-3c01-4039-8034-1cc291b5bb79",
   "metadata": {},
   "source": [
    "Filter to keep variants belonging to known ICB samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "6dcdb9af-eb8a-41d3-93a7-8b46bd3c80f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.True_"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Harmonize sample IDs in the OncoPanel data\n",
    "op_sample_df['sample_id_merge'] = np.where(op_sample_df['id_broad_sample'].notnull(),\n",
    "                                           op_sample_df['id_broad_sample'],\n",
    "                                           op_sample_df['id_uniquesample']).astype(str)\n",
    "\n",
    "op_sample_df.replace({'GBM.ICB-3-14.637.Post.FFbank': 'GBM.ICB-3-14.233.Pre'}, inplace = True)\n",
    "\n",
    "icb_op_samples = icb_sample_df[icb_sample_df['usable_oncopanel']]['sample_id_legacy'].astype(str)\n",
    "op_sample_df = op_sample_df[op_sample_df['sample_id_merge'].isin(icb_op_samples)]\n",
    "\n",
    "len(op_sample_df) == icb_sample_df['usable_oncopanel'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "64ba3adf-efa7-44f6-9e55-a8508e5d144d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "159"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kept_op_samples = op_sample_df['id_uniquesample'].tolist()\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df['id_uniquesample'].isin(kept_op_samples)]\n",
    "op_cnv_df = op_cnv_df[op_cnv_df['id_uniquesample'].isin(kept_op_samples)]\n",
    "\n",
    "len(kept_op_samples)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e3b589aa-5b41-482e-90b5-a578185b3433",
   "metadata": {},
   "source": [
    "### Filter mutations\n",
    "\n",
    "Select mutations by several criteria:\n",
    "* Filter mutations by VAF\n",
    "* Keep only variants belonging to genes with exons covered by all versions of OncoPanel\n",
    "* Keep only non-synonymous mutations\n",
    "* Filter likely germline events by gnomAD frequency\n",
    "* Keep mutations belonging to the latest pre-ICB timepoint for each participant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "e0db565c-62f3-4177-99a8-240cc7de9323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants above the VAF threshold (0.1)\n",
    "op_mut_df = op_mut_df[op_mut_df['ALLELE_FRACTION'] >= 100*VAF_THRESHOLD]\n",
    "\n",
    "# Keep only genes with exons covered by all versions of OncoPanel\n",
    "op_mut_df = op_mut_df[op_mut_df['HARMONIZED_HUGO_GENE_NAME'].isin(oncopanel_genes['Gene'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "22010afe-b10a-4857-88df-fd67930e5cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Non-synonymous mutations\n",
    "nonsyn_classifications = ['Missense',\n",
    "                          'Frame_Shift_Del',\n",
    "                          'Nonsense',\n",
    "                          'Splice_Region',\n",
    "                          'Frame_Shift_Ins',\n",
    "                          'Splice_Site',\n",
    "                          'In_Frame_Ins',\n",
    "                          'In_Frame_Del',\n",
    "                          'Translation_Start_Site']\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df['HARMONIZED_VARIANT_CLASS'].isin(nonsyn_classifications)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "6039fbe6-49f7-4eea-b8b9-9e65dd8cae76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gnomAD germline mutation filtering\n",
    "gnomad_threshold = 2e-4\n",
    "\n",
    "op_mut_df = op_mut_df[op_mut_df[\"MAX_GNOMAD_FREQUENCY\"].fillna(0.0).astype(float).lt(gnomad_threshold)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1883ad81-6a8c-4643-a754-5bef0e6cfa02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Latest pre-ICB timepoint for each participant\n",
    "def keep_latest_preicb_timepoint_variants(variant_df, sample_df, sample_id_col=\"sample_id_merge\"):\n",
    "    variant_df = variant_df.join(sample_df[[\"sample_id_legacy\", \"participant_id_legacy\", \"collection_date_dfd\", \"pre_post\"]].set_index(\"sample_id_legacy\"), on=sample_id_col)\n",
    "    variant_df = variant_df[variant_df[\"pre_post\"] == \"Pre\"]\n",
    "    latest_timepoints = variant_df[[\"participant_id_legacy\", \"collection_date_dfd\"]].groupby(\"participant_id_legacy\").max()\n",
    "    latest_timepoints.columns = [\"latest_timepoint\"]\n",
    "    variant_df = variant_df.join(latest_timepoints, on=\"participant_id_legacy\")\n",
    "\n",
    "    variant_df = variant_df[variant_df[\"collection_date_dfd\"] == variant_df[\"latest_timepoint\"]]\n",
    "    return variant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "f815699a-877d-410c-a59f-2dd36f497798",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_drspatient</th>\n",
       "      <th>id_uniquesample</th>\n",
       "      <th>id_icb</th>\n",
       "      <th>id_mj</th>\n",
       "      <th>VARIANT_CALL_ID</th>\n",
       "      <th>CANONICAL_GENE</th>\n",
       "      <th>BEST_EFF_GENE</th>\n",
       "      <th>CANONICAL_VARIANT_CLASS</th>\n",
       "      <th>BEST_EFF_VARIANT_CLASS</th>\n",
       "      <th>HARMONIZED_HUGO_GENE_NAME</th>\n",
       "      <th>...</th>\n",
       "      <th>CANONICAL_REF_SEQ_TSCP_ID</th>\n",
       "      <th>BEST_EFFECT_ENSEMBL_TSCP_ID</th>\n",
       "      <th>BEST_EFFECT_REF_SEQ_TSCP_ID</th>\n",
       "      <th>HARMONIZED_TRANSCRIPT_ID</th>\n",
       "      <th>MAX_GNOMAD_FREQUENCY</th>\n",
       "      <th>sample_id_merge</th>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th>collection_date_dfd</th>\n",
       "      <th>pre_post</th>\n",
       "      <th>latest_timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>101954597</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>221.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>210153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001126112</td>\n",
       "      <td>ENST00000269305</td>\n",
       "      <td>8.80300000000e-06</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>GBM.ICB-221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>101954597</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>221.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>210001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RB1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_000321</td>\n",
       "      <td>ENST00000267163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543883876</td>\n",
       "      <td>GBM.ICB-221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>656799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAF</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001654</td>\n",
       "      <td>ENST00000377045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>GBM.ICB-105</td>\n",
       "      <td>19.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1019542469</td>\n",
       "      <td>1019543884287</td>\n",
       "      <td>1.0</td>\n",
       "      <td>720.0</td>\n",
       "      <td>798723</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MPL</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_005373</td>\n",
       "      <td>ENST00000372470</td>\n",
       "      <td>.000028908</td>\n",
       "      <td>GBM.ICB-1-10.533.A2-Pre</td>\n",
       "      <td>GBM.ICB-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1019542469</td>\n",
       "      <td>1019543884287</td>\n",
       "      <td>1.0</td>\n",
       "      <td>720.0</td>\n",
       "      <td>798828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SETBP1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_015559</td>\n",
       "      <td>ENST00000282030</td>\n",
       "      <td>.000086861</td>\n",
       "      <td>GBM.ICB-1-10.533.A2-Pre</td>\n",
       "      <td>GBM.ICB-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4193</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264657.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264657</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925965</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Missense</td>\n",
       "      <td>GNAS</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371100.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371100</td>\n",
       "      <td>.000086133</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4195</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925895</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense</td>\n",
       "      <td>CBLB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264122.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000264122</td>\n",
       "      <td>.000115821</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4196</th>\n",
       "      <td>1019543398</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>244.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>10925856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Frameshift</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371953.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000371953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1019543916503</td>\n",
       "      <td>GBM.ICB-244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4255</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>1480352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>ATM</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>GBM.ICB-108</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pre</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>378 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id_drspatient id_uniquesample  id_icb   id_mj  VARIANT_CALL_ID  \\\n",
       "2         101954597   1019543883876   221.0   192.0           210153   \n",
       "3         101954597   1019543883876   221.0   192.0           210001   \n",
       "14        101954688   1019543884068   105.0   221.0           656799   \n",
       "24       1019542469   1019543884287     1.0   720.0           798723   \n",
       "25       1019542469   1019543884287     1.0   720.0           798828   \n",
       "...             ...             ...     ...     ...              ...   \n",
       "4193     1019543398   1019543916503   244.0  1027.0         10925958   \n",
       "4194     1019543398   1019543916503   244.0  1027.0         10925965   \n",
       "4195     1019543398   1019543916503   244.0  1027.0         10925895   \n",
       "4196     1019543398   1019543916503   244.0  1027.0         10925856   \n",
       "4255       92361207    923612315805   108.0   255.0          1480352   \n",
       "\n",
       "     CANONICAL_GENE BEST_EFF_GENE CANONICAL_VARIANT_CLASS  \\\n",
       "2               NaN           NaN                     NaN   \n",
       "3               NaN           NaN                     NaN   \n",
       "14              NaN           NaN                     NaN   \n",
       "24              NaN           NaN                     NaN   \n",
       "25              NaN           NaN                     NaN   \n",
       "...             ...           ...                     ...   \n",
       "4193            NaN           NaN                     NaN   \n",
       "4194            NaN           NaN      intergenic_variant   \n",
       "4195            NaN           NaN                     NaN   \n",
       "4196            NaN           NaN                     NaN   \n",
       "4255            NaN           ATM                     NaN   \n",
       "\n",
       "     BEST_EFF_VARIANT_CLASS HARMONIZED_HUGO_GENE_NAME  ...  \\\n",
       "2                       NaN                      TP53  ...   \n",
       "3                       NaN                       RB1  ...   \n",
       "14                      NaN                      ARAF  ...   \n",
       "24                      NaN                       MPL  ...   \n",
       "25                      NaN                    SETBP1  ...   \n",
       "...                     ...                       ...  ...   \n",
       "4193                    NaN                     STAT3  ...   \n",
       "4194               Missense                      GNAS  ...   \n",
       "4195               Missense                      CBLB  ...   \n",
       "4196             Frameshift                      PTEN  ...   \n",
       "4255            Splice_Site                       ATM  ...   \n",
       "\n",
       "     CANONICAL_REF_SEQ_TSCP_ID BEST_EFFECT_ENSEMBL_TSCP_ID  \\\n",
       "2                          NaN                         NaN   \n",
       "3                          NaN                         NaN   \n",
       "14                         NaN                         NaN   \n",
       "24                         NaN                         NaN   \n",
       "25                         NaN                         NaN   \n",
       "...                        ...                         ...   \n",
       "4193                       NaN           ENST00000264657.5   \n",
       "4194                       NaN           ENST00000371100.4   \n",
       "4195                       NaN           ENST00000264122.4   \n",
       "4196                       NaN           ENST00000371953.3   \n",
       "4255                       NaN                         NaN   \n",
       "\n",
       "     BEST_EFFECT_REF_SEQ_TSCP_ID HARMONIZED_TRANSCRIPT_ID  \\\n",
       "2                   NM_001126112          ENST00000269305   \n",
       "3                      NM_000321          ENST00000267163   \n",
       "14                     NM_001654          ENST00000377045   \n",
       "24                     NM_005373          ENST00000372470   \n",
       "25                     NM_015559          ENST00000282030   \n",
       "...                          ...                      ...   \n",
       "4193                         NaN          ENST00000264657   \n",
       "4194                         NaN          ENST00000371100   \n",
       "4195                         NaN          ENST00000264122   \n",
       "4196                         NaN          ENST00000371953   \n",
       "4255                         NaN                      NaN   \n",
       "\n",
       "     MAX_GNOMAD_FREQUENCY          sample_id_merge participant_id_legacy  \\\n",
       "2       8.80300000000e-06            1019543883876           GBM.ICB-221   \n",
       "3                     NaN            1019543883876           GBM.ICB-221   \n",
       "14                    NaN            1019543884068           GBM.ICB-105   \n",
       "24             .000028908  GBM.ICB-1-10.533.A2-Pre             GBM.ICB-1   \n",
       "25             .000086861  GBM.ICB-1-10.533.A2-Pre             GBM.ICB-1   \n",
       "...                   ...                      ...                   ...   \n",
       "4193                  NaN            1019543916503           GBM.ICB-244   \n",
       "4194           .000086133            1019543916503           GBM.ICB-244   \n",
       "4195           .000115821            1019543916503           GBM.ICB-244   \n",
       "4196                  NaN            1019543916503           GBM.ICB-244   \n",
       "4255                  NaN             923612315805           GBM.ICB-108   \n",
       "\n",
       "      collection_date_dfd  pre_post latest_timepoint  \n",
       "2                     0.0       Pre              0.0  \n",
       "3                     0.0       Pre              0.0  \n",
       "14                   19.0       Pre             19.0  \n",
       "24                    0.0       Pre              0.0  \n",
       "25                    0.0       Pre              0.0  \n",
       "...                   ...       ...              ...  \n",
       "4193                  0.0       Pre              0.0  \n",
       "4194                  0.0       Pre              0.0  \n",
       "4195                  0.0       Pre              0.0  \n",
       "4196                  0.0       Pre              0.0  \n",
       "4255                  0.0       Pre              0.0  \n",
       "\n",
       "[378 rows x 44 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_mut_df = op_mut_df.join(op_sample_df[[\"id_uniquesample\", \"sample_id_merge\"]].set_index(\"id_uniquesample\"), on=\"id_uniquesample\")\n",
    "op_mut_df = keep_latest_preicb_timepoint_variants(op_mut_df, icb_sample_df)\n",
    "\n",
    "op_mut_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3b770f8-e853-4051-b016-5e872952707e",
   "metadata": {},
   "source": [
    "### Construct pt-by-variant matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ec429b49-5e0e-42d5-8980-32520dc2b1e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_mut</th>\n",
       "      <th>AKT1_mut</th>\n",
       "      <th>ALK_mut</th>\n",
       "      <th>APC_mut</th>\n",
       "      <th>ARAF_mut</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1B_mut</th>\n",
       "      <th>ARID2_mut</th>\n",
       "      <th>ASXL1_mut</th>\n",
       "      <th>ATM_mut</th>\n",
       "      <th>ATRX_mut</th>\n",
       "      <th>...</th>\n",
       "      <th>TCF3_mut</th>\n",
       "      <th>TCF7L2_mut</th>\n",
       "      <th>TERT_mut</th>\n",
       "      <th>TET2_mut</th>\n",
       "      <th>TNFAIP3_mut</th>\n",
       "      <th>TP53_mut</th>\n",
       "      <th>TSC1_mut</th>\n",
       "      <th>TSC2_mut</th>\n",
       "      <th>WRN_mut</th>\n",
       "      <th>WT1_mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-102</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-103</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-104</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-87</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-88</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-89</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-90</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-94</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows Ã— 127 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_mut               AKT1_mut  ALK_mut  APC_mut  ARAF_mut  ARID1A_mut  \\\n",
       "participant_id_legacy                                                     \n",
       "GBM.ICB-1                     0        0        0         0           0   \n",
       "GBM.ICB-10                    0        0        0         0           0   \n",
       "GBM.ICB-102                   0        0        0         0           0   \n",
       "GBM.ICB-103                   0        0        2         0           0   \n",
       "GBM.ICB-104                   0        0        0         0           0   \n",
       "...                         ...      ...      ...       ...         ...   \n",
       "GBM.ICB-87                    0        0        0         0           0   \n",
       "GBM.ICB-88                    0        0        0         0           0   \n",
       "GBM.ICB-89                    0        0        0         0           0   \n",
       "GBM.ICB-90                    0        0        0         0           0   \n",
       "GBM.ICB-94                    0        0        0         0           0   \n",
       "\n",
       "gene_mut               ARID1B_mut  ARID2_mut  ASXL1_mut  ATM_mut  ATRX_mut  \\\n",
       "participant_id_legacy                                                        \n",
       "GBM.ICB-1                       0          0          0        0         0   \n",
       "GBM.ICB-10                      0          0          0        0         0   \n",
       "GBM.ICB-102                     0          0          0        0         0   \n",
       "GBM.ICB-103                     0          0          0        0         0   \n",
       "GBM.ICB-104                     0          0          0        0         0   \n",
       "...                           ...        ...        ...      ...       ...   \n",
       "GBM.ICB-87                      0          0          0        0         0   \n",
       "GBM.ICB-88                      0          0          0        0         0   \n",
       "GBM.ICB-89                      0          0          0        0         0   \n",
       "GBM.ICB-90                      0          0          0        0         0   \n",
       "GBM.ICB-94                      0          0          0        0         2   \n",
       "\n",
       "gene_mut               ...  TCF3_mut  TCF7L2_mut  TERT_mut  TET2_mut  \\\n",
       "participant_id_legacy  ...                                             \n",
       "GBM.ICB-1              ...         0           0         0         0   \n",
       "GBM.ICB-10             ...         0           0         0         0   \n",
       "GBM.ICB-102            ...         0           0         0         0   \n",
       "GBM.ICB-103            ...         0           0         0         0   \n",
       "GBM.ICB-104            ...         0           0         0         0   \n",
       "...                    ...       ...         ...       ...       ...   \n",
       "GBM.ICB-87             ...         0           0         0         0   \n",
       "GBM.ICB-88             ...         0           0         0         0   \n",
       "GBM.ICB-89             ...         0           0         0         0   \n",
       "GBM.ICB-90             ...         0           0         0         0   \n",
       "GBM.ICB-94             ...         0           0         0         0   \n",
       "\n",
       "gene_mut               TNFAIP3_mut  TP53_mut  TSC1_mut  TSC2_mut  WRN_mut  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-1                        0         0         0         0        0   \n",
       "GBM.ICB-10                       0         0         0         0        0   \n",
       "GBM.ICB-102                      0         0         0         0        0   \n",
       "GBM.ICB-103                      0         2         0         0        0   \n",
       "GBM.ICB-104                      0         0         0         0        0   \n",
       "...                            ...       ...       ...       ...      ...   \n",
       "GBM.ICB-87                       0         0         0         0        0   \n",
       "GBM.ICB-88                       0         0         0         0        0   \n",
       "GBM.ICB-89                       0         0         0         0        0   \n",
       "GBM.ICB-90                       0         0         0         0        0   \n",
       "GBM.ICB-94                       0         2         0         0        0   \n",
       "\n",
       "gene_mut               WT1_mut  \n",
       "participant_id_legacy           \n",
       "GBM.ICB-1                    0  \n",
       "GBM.ICB-10                   0  \n",
       "GBM.ICB-102                  0  \n",
       "GBM.ICB-103                  0  \n",
       "GBM.ICB-104                  0  \n",
       "...                        ...  \n",
       "GBM.ICB-87                   0  \n",
       "GBM.ICB-88                   0  \n",
       "GBM.ICB-89                   0  \n",
       "GBM.ICB-90                   0  \n",
       "GBM.ICB-94                   0  \n",
       "\n",
       "[137 rows x 127 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_mut_df['gene_mut'] = op_mut_df['HARMONIZED_HUGO_GENE_NAME'] + '_mut'\n",
    "\n",
    "op_variants_matrix = op_mut_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# scoring: sequence variants = 2\n",
    "op_variants_matrix *= 2\n",
    "\n",
    "op_variants_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad545274-6ca3-4204-8bd5-ad7d6590f6e4",
   "metadata": {},
   "source": [
    "### Filter CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "8b601d63-6bd2-46b9-8a81-65cf910a0ee5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_drspatient</th>\n",
       "      <th>id_uniquesample</th>\n",
       "      <th>id_icb</th>\n",
       "      <th>id_mj</th>\n",
       "      <th>VARIANT_CALL_ID</th>\n",
       "      <th>CNV_COPY_COUNT</th>\n",
       "      <th>CNV_COPY_COUNT_IN_REPORT</th>\n",
       "      <th>TUMOR_PURITY</th>\n",
       "      <th>PANEL_VERSION</th>\n",
       "      <th>CANONICAL_GENE</th>\n",
       "      <th>HARMONIZED_HUGO_GENE_NAME</th>\n",
       "      <th>HARMONIZED_VARIANT_CLASS</th>\n",
       "      <th>CNV_GROUP_ID</th>\n",
       "      <th>CNV_GROUP_CYTOBAND</th>\n",
       "      <th>CNV_GROUP_TIER</th>\n",
       "      <th>CNV_GROUP_EVENT</th>\n",
       "      <th>hlvl_call</th>\n",
       "      <th>gene_cnv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129364723</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>EGFR_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>CDKN2A_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129332074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRF1</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>PRF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IKZF1</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>IKZF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129333379</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RET</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>RET_cnv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     id_drspatient id_uniquesample  id_icb  id_mj  VARIANT_CALL_ID  \\\n",
       "124      101954688   1019543884068   105.0  221.0        129364723   \n",
       "125      101954688   1019543884068   105.0  221.0        129354263   \n",
       "126      101954688   1019543884068   105.0  221.0        129332074   \n",
       "128      101954688   1019543884068   105.0  221.0        129354974   \n",
       "129      101954688   1019543884068   105.0  221.0        129333379   \n",
       "\n",
       "     CNV_COPY_COUNT  CNV_COPY_COUNT_IN_REPORT  TUMOR_PURITY  PANEL_VERSION  \\\n",
       "124             NaN                       NaN          70.0            1.0   \n",
       "125             NaN                       NaN          70.0            1.0   \n",
       "126             NaN                       NaN          70.0            1.0   \n",
       "128             NaN                       NaN          70.0            1.0   \n",
       "129             NaN                       NaN          70.0            1.0   \n",
       "\n",
       "    CANONICAL_GENE HARMONIZED_HUGO_GENE_NAME HARMONIZED_VARIANT_CLASS  \\\n",
       "124            NaN                      EGFR                       LA   \n",
       "125            NaN                    CDKN2A                     1DEL   \n",
       "126            NaN                      PRF1                     1DEL   \n",
       "128            NaN                     IKZF1                       LA   \n",
       "129            NaN                       RET                     1DEL   \n",
       "\n",
       "     CNV_GROUP_ID CNV_GROUP_CYTOBAND CNV_GROUP_TIER CNV_GROUP_EVENT  \\\n",
       "124           NaN                NaN            NaN             NaN   \n",
       "125           NaN                NaN            NaN             NaN   \n",
       "126           NaN                NaN            NaN             NaN   \n",
       "128           NaN                NaN            NaN             NaN   \n",
       "129           NaN                NaN            NaN             NaN   \n",
       "\n",
       "     hlvl_call    gene_cnv  \n",
       "124          1    EGFR_cnv  \n",
       "125         -1  CDKN2A_cnv  \n",
       "126         -1    PRF1_cnv  \n",
       "128          1   IKZF1_cnv  \n",
       "129         -1     RET_cnv  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnv_df = op_cnv_df[op_cnv_df['HARMONIZED_HUGO_GENE_NAME'].isin(oncopanel_genes['Gene'])]\n",
    "\n",
    "op_cnv_df['hlvl_call'] = op_cnv_df['HARMONIZED_VARIANT_CLASS'].map(\n",
    "    {'2DEL': -2, '1DEL': -1, 'LA': 1, 'HA': 2}\n",
    ")\n",
    "\n",
    "op_cnv_df['gene_cnv'] = op_cnv_df['HARMONIZED_HUGO_GENE_NAME'] + '_cnv'\n",
    "\n",
    "op_cnv_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49f77769-172e-4d3f-85d3-159db637f5fe",
   "metadata": {},
   "source": [
    "Ensure no duplicate CNVs in any sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "c25b8108-9e92-46a2-97d8-9bf8dcefdc7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(1)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnv_df.groupby(['id_uniquesample', 'HARMONIZED_HUGO_GENE_NAME']).size().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "9bd3f2f5-1af0-4c44-8a26-055b272f5ae1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_drspatient</th>\n",
       "      <th>id_uniquesample</th>\n",
       "      <th>id_icb</th>\n",
       "      <th>id_mj</th>\n",
       "      <th>VARIANT_CALL_ID</th>\n",
       "      <th>CNV_COPY_COUNT</th>\n",
       "      <th>CNV_COPY_COUNT_IN_REPORT</th>\n",
       "      <th>TUMOR_PURITY</th>\n",
       "      <th>PANEL_VERSION</th>\n",
       "      <th>CANONICAL_GENE</th>\n",
       "      <th>HARMONIZED_HUGO_GENE_NAME</th>\n",
       "      <th>HARMONIZED_VARIANT_CLASS</th>\n",
       "      <th>CNV_GROUP_ID</th>\n",
       "      <th>CNV_GROUP_CYTOBAND</th>\n",
       "      <th>CNV_GROUP_TIER</th>\n",
       "      <th>CNV_GROUP_EVENT</th>\n",
       "      <th>hlvl_call</th>\n",
       "      <th>gene_cnv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129364723</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>EGFR_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>CDKN2A_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129332074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRF1</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>PRF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129354974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IKZF1</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>IKZF1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>101954688</td>\n",
       "      <td>1019543884068</td>\n",
       "      <td>105.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>129333379</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RET</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>RET_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21926</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>77678303</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARD11</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>CARD11_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21928</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>77404498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MET</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>MET_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21930</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>77336483</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>LA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ETV1_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21931</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>76593171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BUB1B</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>BUB1B_cnv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21933</th>\n",
       "      <td>92361207</td>\n",
       "      <td>923612315805</td>\n",
       "      <td>108.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>76385523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>1DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>SMARCB1_cnv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3286 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id_drspatient id_uniquesample  id_icb  id_mj  VARIANT_CALL_ID  \\\n",
       "124        101954688   1019543884068   105.0  221.0        129364723   \n",
       "125        101954688   1019543884068   105.0  221.0        129354263   \n",
       "126        101954688   1019543884068   105.0  221.0        129332074   \n",
       "128        101954688   1019543884068   105.0  221.0        129354974   \n",
       "129        101954688   1019543884068   105.0  221.0        129333379   \n",
       "...              ...             ...     ...    ...              ...   \n",
       "21926       92361207    923612315805   108.0  255.0         77678303   \n",
       "21928       92361207    923612315805   108.0  255.0         77404498   \n",
       "21930       92361207    923612315805   108.0  255.0         77336483   \n",
       "21931       92361207    923612315805   108.0  255.0         76593171   \n",
       "21933       92361207    923612315805   108.0  255.0         76385523   \n",
       "\n",
       "       CNV_COPY_COUNT  CNV_COPY_COUNT_IN_REPORT  TUMOR_PURITY  PANEL_VERSION  \\\n",
       "124               NaN                       NaN          70.0            1.0   \n",
       "125               NaN                       NaN          70.0            1.0   \n",
       "126               NaN                       NaN          70.0            1.0   \n",
       "128               NaN                       NaN          70.0            1.0   \n",
       "129               NaN                       NaN          70.0            1.0   \n",
       "...               ...                       ...           ...            ...   \n",
       "21926             NaN                       NaN          80.0            1.0   \n",
       "21928             NaN                       NaN          80.0            1.0   \n",
       "21930             NaN                       NaN          80.0            1.0   \n",
       "21931             NaN                       NaN          80.0            1.0   \n",
       "21933             NaN                       NaN          80.0            1.0   \n",
       "\n",
       "      CANONICAL_GENE HARMONIZED_HUGO_GENE_NAME HARMONIZED_VARIANT_CLASS  \\\n",
       "124              NaN                      EGFR                       LA   \n",
       "125              NaN                    CDKN2A                     1DEL   \n",
       "126              NaN                      PRF1                     1DEL   \n",
       "128              NaN                     IKZF1                       LA   \n",
       "129              NaN                       RET                     1DEL   \n",
       "...              ...                       ...                      ...   \n",
       "21926            NaN                    CARD11                       LA   \n",
       "21928            NaN                       MET                       LA   \n",
       "21930            NaN                      ETV1                       LA   \n",
       "21931            NaN                     BUB1B                     1DEL   \n",
       "21933            NaN                   SMARCB1                     1DEL   \n",
       "\n",
       "       CNV_GROUP_ID CNV_GROUP_CYTOBAND CNV_GROUP_TIER CNV_GROUP_EVENT  \\\n",
       "124             NaN                NaN            NaN             NaN   \n",
       "125             NaN                NaN            NaN             NaN   \n",
       "126             NaN                NaN            NaN             NaN   \n",
       "128             NaN                NaN            NaN             NaN   \n",
       "129             NaN                NaN            NaN             NaN   \n",
       "...             ...                ...            ...             ...   \n",
       "21926           NaN                NaN            NaN             NaN   \n",
       "21928           NaN                NaN            NaN             NaN   \n",
       "21930           NaN                NaN            NaN             NaN   \n",
       "21931           NaN                NaN            NaN             NaN   \n",
       "21933           NaN                NaN            NaN             NaN   \n",
       "\n",
       "       hlvl_call     gene_cnv  \n",
       "124            1     EGFR_cnv  \n",
       "125           -1   CDKN2A_cnv  \n",
       "126           -1     PRF1_cnv  \n",
       "128            1    IKZF1_cnv  \n",
       "129           -1      RET_cnv  \n",
       "...          ...          ...  \n",
       "21926          1   CARD11_cnv  \n",
       "21928          1      MET_cnv  \n",
       "21930          1     ETV1_cnv  \n",
       "21931         -1    BUB1B_cnv  \n",
       "21933         -1  SMARCB1_cnv  \n",
       "\n",
       "[3286 rows x 18 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "55c63f1b-a354-4555-9fe1-ee9096c348b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "op_cnv_df = op_cnv_df.join(op_sample_df[[\"id_uniquesample\", \"sample_id_merge\"]].set_index(\"id_uniquesample\"), on=\"id_uniquesample\")\n",
    "op_cnv_df = keep_latest_preicb_timepoint_variants(op_cnv_df, icb_sample_df, sample_id_col=\"sample_id_merge\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "29912f4c-a459-4201-b192-8695aea1bb45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_cnv</th>\n",
       "      <th>ABL1_cnv</th>\n",
       "      <th>AKT1_cnv</th>\n",
       "      <th>AKT2_cnv</th>\n",
       "      <th>AKT3_cnv</th>\n",
       "      <th>ALK_cnv</th>\n",
       "      <th>APC_cnv</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>ARID1B_cnv</th>\n",
       "      <th>ARID2_cnv</th>\n",
       "      <th>AR_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>TP53_cnv</th>\n",
       "      <th>TSC1_cnv</th>\n",
       "      <th>TSC2_cnv</th>\n",
       "      <th>U2AF1_cnv</th>\n",
       "      <th>VHL_cnv</th>\n",
       "      <th>WRN_cnv</th>\n",
       "      <th>WT1_cnv</th>\n",
       "      <th>XPA_cnv</th>\n",
       "      <th>XPC_cnv</th>\n",
       "      <th>ZNF217_cnv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-102</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-103</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-104</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 229 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_cnv               ABL1_cnv  AKT1_cnv  AKT2_cnv  AKT3_cnv  ALK_cnv  \\\n",
       "participant_id_legacy                                                    \n",
       "GBM.ICB-1                     0         0         0         0        0   \n",
       "GBM.ICB-10                    0         0         0         0        0   \n",
       "GBM.ICB-102                   0         0         0         0        0   \n",
       "GBM.ICB-103                   0         0        -1         0        0   \n",
       "GBM.ICB-104                   0         0         0         0        0   \n",
       "\n",
       "gene_cnv               APC_cnv  ARID1A_cnv  ARID1B_cnv  ARID2_cnv  AR_cnv  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-1                    0           0           0          0       0   \n",
       "GBM.ICB-10                   0           0          -1          0       0   \n",
       "GBM.ICB-102                  0           0          -1          0       0   \n",
       "GBM.ICB-103                  0           0           0         -1       0   \n",
       "GBM.ICB-104                  0           0           0          0       0   \n",
       "\n",
       "gene_cnv               ...  TP53_cnv  TSC1_cnv  TSC2_cnv  U2AF1_cnv  VHL_cnv  \\\n",
       "participant_id_legacy  ...                                                     \n",
       "GBM.ICB-1              ...         0         0         0          1        0   \n",
       "GBM.ICB-10             ...         0         0         0          0        0   \n",
       "GBM.ICB-102            ...         0         0         0          0        0   \n",
       "GBM.ICB-103            ...         0         0         0          0        0   \n",
       "GBM.ICB-104            ...         0         0         0          0        0   \n",
       "\n",
       "gene_cnv               WRN_cnv  WT1_cnv  XPA_cnv  XPC_cnv  ZNF217_cnv  \n",
       "participant_id_legacy                                                  \n",
       "GBM.ICB-1                    0        0        0        0           0  \n",
       "GBM.ICB-10                   0        0        0        0           0  \n",
       "GBM.ICB-102                  0        0        0        0           0  \n",
       "GBM.ICB-103                  0        0        0        0           0  \n",
       "GBM.ICB-104                  0        0        0        0           0  \n",
       "\n",
       "[5 rows x 229 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_cnvs_matrix = op_cnv_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_cnv', values = 'hlvl_call', aggfunc = 'sum', fill_value = 0)\n",
    "\n",
    "op_cnvs_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "d4b2a6aa-02e2-43ae-9770-eaae83e33f57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGzCAYAAACPa3XZAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAKCtJREFUeJzt3Qu8ndOdP/6VRHNBEhJCVCJhdFyC1iWpS1XaVKqa0mmVGW0jTChxCUYrWkyKxq2kDU0wI+gIeplgqBiTEZkWRYJB65K6pQyhSCQqNNm/1/fpf5//PjvnnJywzzq39/v12jk5+zx7P2s/ez3P/qz1rPXsLqVSqZQAADLpmmtFAABB+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4oM6QIUPSEUccUff7vHnzUpcuXYqf0JGp6+3r2LQunn/++eK9veaaa2peLj444aONiR0kdpS4/frXv17j73E1/EGDBhV//+IXv5g6utmzZ6cDDjggbbLJJql79+5piy22SF/72tfSf//3f6/xwRG3BQsWrPEccdDacMMNi/8vWbIkrbfeeunrX/96o+t8++23U69evdLf/d3fpfbsBz/4Qbr55ptTZ/TOO++kf/7nf271MPH++++nH//4x2mPPfZIvXv3Luph/D/ui7+1BeUP5/KtW7duafDgwenLX/5yeuSRR1Jndu+99xb16K233mrtonQ4wkcb1bNnzzRr1qw17r/nnnvSH//4x9SjR48WL8O+++6b/vznPxc/c4uQNW7cuCIAvPrqq+mUU05JM2bMSBMmTEjPPvts+uxnP1scGKrFgaIpAwYMSJ/73OfSLbfcUnxANeTf//3f07vvvttkQGkPOnv4mDx5crPDR0vU9RUrVhR17aSTTkqbb755Ov/889NFF11UBOi4L/4Wy7QVf//3f59++tOfpquvvjr9wz/8QxHwP/nJT3bqABLHmKhHwkftCR9t1Be+8IX085//PP3lL3+pd38Ekt122604mLW0rl27FiEofub2wx/+sOgFmjhxYtGbccYZZ6Qjjzwyffe7300PPfRQuu6664oejEof//jH02233ZYWLlzY5HMffvjhafny5enWW29t8O+xjfv27ZsOPPDAmr4mmqexUNje6noE5mgsTJs2Lf3Hf/xHEZyPPfbYIvhedtllxd/+6Z/+KbUVu+66axG4x44dWwSlf/u3f0srV65M06dPb+2i0RHFt9rSdsycOTO+Zbj085//vNSlS5fSr371q7q/rVy5srTxxhuXfvjDH5a22mqr0oEHHljvsatWrSpdeumlpR122KHUo0eP0oABA0pHH3106Y033qi33OrVq0vnnHNO6aMf/WipV69epf3226/0+OOPF885duzYuuXuvvvuoizxs2z+/Pmlr371q6VBgwaVunfvXtpyyy1LEydOLL3zzjv11hHPs8EGG5T++Mc/lg466KDi/5tssknp1FNPLf3lL39pchvEc/Xr16+03XbbrXXZynL+67/+a7F9xowZ02BZypYvX178Xr1cePXVV0vdunUrHXXUUWtd75///OfS2WefXdp2222L7b355puXvvzlL5cWLVpUb12nnHJKsZ1ie33sYx8rXXTRRcV7UPbcc88V5Y/3vlrcH+soi//Hfc8880zxuvr27Vvq06dP6YgjjiitWLGi3uOqb+X3dtmyZaWTTjqpeL+jTJtuumlp1KhRpQULFjT5esvr/v3vf1865JBDSr179y7epxNPPLHYFtV++tOflnbddddSz549i/fl0EMPLb344ov1lvn0pz9d2nHHHUsPPfRQ6VOf+lRRH6NsjSm/l3/4wx9K+++/f2n99dcvDRw4sDR58uS6bVrentW3yu1YraG6Xi7bE088UewjUbYtttiidMEFF5TWZvHixUU9+sxnPtPoMiNHjiytt956xbJlUYYJEyaUZs+eXaw73p/Yn++44441Hh/71pFHHlm8/lhuyJAhpW9961vFcaIstlPsr7H9o/wjRowo3XbbbfWep7y9ol5Wirob93/uc58rfr/55ptLX/jCF+rWt/XWW5e+//3vr7GPrst2e/fdd0tnnXVWaZtttqk7npx22mnF/ZWqj02NefPNN4vlYp+IfeOb3/xm6eGHH15j/3r00UeL5YYOHVrsu5tttllp3Lhxpddff32N+l59i+0Vrr766uI9jP0nyr799tuXfvKTn6y1jPxV/aYjbWqA1Z577pluuOGGYsxDuOOOO9LSpUvTYYcdVpwzrnbMMccUvQVxuuLEE09Mzz33XNHCevjhh9NvfvOb9JGPfKRY7qyzzkrnnntu0bsSt+gp2H///dN777231nJFb0y0TKMF179///TAAw8ULbs4FRR/q7Rq1ao0evToNGLEiHTxxRen//qv/yp6NLbZZpvi8Y2JsS5vvPFG0esR55+bq0+fPunkk08uXl+8pmjJNWSDDTZIBx10UPrFL35RrKdfv351f7vpppuKckfvSFNimRhzM3fu3OL9iG70GCty1113pccff7x4jfFZ8qUvfSndfffd6aijjip6Zu6888502mmnpZdeeildeuml6YOKcS9Dhw5NU6ZMKV7rv/zLvxSnlC644ILi79F9/o//+I9p+PDh6eijjy7uizKFb33rW8VrP/7449MOO+yQ/vSnPxXb/Pe//32j26x63VE/Y933339/URfffPPNojeq7LzzzktnnnlmsWyU47XXXivqSZzWiPq40UYb1S0b6486HtsxWt6bbbbZWrf95z//+eKUwIUXXpjmzJmTzj777KKX8Pvf/37adNNNi9Z61LEYt1Aeu7Pzzjuv83aO1xXriueI1xLb7Tvf+U7aaaed6vbLhsS+GuX85je/2egy8beoG1H+2EZl8V7Eqb/jjjuuGCcS2/crX/lKevHFF4t9Lrz88svFexunA+L93W677Yo6FeWL/TPGR8Xpyr322qv4PY4H8dhrr722qJOxXGybpvzhD38ofpbXGceWGLMSPTrxM07LxL62bNmy4nTSum631atXF2WJ1xuvYfvtt0+PPfZYsV88/fTT63zKMPa32K/j+aKOx/PFmLHoyakW+2mcvo1jZfQiP/HEE+nKK68sfkadjrEvUfYoRxyDo0wx7ixE/QpRx3bcccfiNUQvbPRuxXsWryt6uViL/y+E0MZ6Ph588MHSZZddVrQuy70K0dqMpB2qez7+53/+p3jc9ddfX+/55syZU+/+JUuWFCk9HlvZ+j7jjDPqtY4baw1W93CEKVOmFL00L7zwQt198Tzx2GgZVfrEJz5R2m233ZrcBj/60Y+Kx0brrznK5Yzeorfeeqto5X3pS19qtOcj3H777cVjrrjiinr3f/KTnyx6hKIXqSnR6onHX3LJJWv8rbxdo6UYy5x77rn1/h4t0dhe5R6SD9LzES3eStHj0r9//3r3xWtuqLUYLcJoXa+r8rort2047rjjivujNRmef/75otV/3nnn1VvuscceK1r6lfdHKzkeO2PGjGaVoVyvTjjhhHrbO+pz1OvXXnutuC9+rq23ozk9H3HfddddV3df9CpED9dXvvKVJp8vegPjsdHqbszChQuLZaJnrCx+j9dR2XsW2zXunzZtWt190aLv2rVrcZxorP6VyxDHhrK33367aO1HL0m5jpfrX/QexXZ75ZVXSvPmzSv21bj/l7/8ZaP7/jHHHFP0PlX2VDR3u0XPWLyGyvKFqAvx+N/85jfr1PNR3t8uvPDCuvuiVyZ61Kr3r4Zeyw033FAsF727ZdEbVNnbUamh5xg9enTRI8TaGfPRhkWLIQbBxTiGaFXHzxgI1pDodYhxCjGI7fXXX6+7xfiQaKVECytE70P0cJxwwglFui+LXobmiFkgZTFYLtYRras4bkaLtlq0QCp96lOfKlocTYmWVIhW37qKbRCvJcZzNFSesujpiRZM5aDe6CmKVk8MvFvbuf9f/vKXRUsotmO18nb91a9+VfTcRKuz0qmnnlpsr2gdf1ANbdfoQShvu6ZEr8Nvf/vbovX8QVS36srbIF5viFZ7tP6i/lbWxWhhbrvttnV1sSwGT0cLdF1Er03l9o7fo15H/a6l2HcqBx5Hj0L0OKytDsf+urY6XP5b9Xs2atSoul6qco9N9OqV1xnbNnoFxowZk3bfffcm61+UdZ999qn3eqKXIWa4/O53v6v3uOg9in0i3qf99tuv6PmInrRyz1Hlvh+vL97TqHfRs/Lkk0+u83aLY1b0TkSvTWU9+cxnPlP8vbqerE283uiBqOxVjf2voX208rXE4PJYb/SkhbWNGWvoOaJHOp7j05/+dPEa43ea5rRLGxYHgjgQxQdk7ODRjfvVr361wWWfeeaZosJH13tDYoppeOGFF4qf8SFQva6NN954rWWKrt/oao0P9+harVS9w8UAvnIXZVmso/px1eJAW3kAX1dxCiS6SWPmSwzua0gcpA499ND0k5/8pOiu/uhHP1oXRNZ2yiXEgflv//Zv1xj0Wim2dcxsqP4AigNu+e8fVEyFrFR+72LblrdfY+JURXRFx5TtCKdx6i1OAWy99dbNWnd13YkPyghr8YFWrosRrqqXKyuf/iuLbR8fTs0V66ou68c+9rHiZ7kMtbLlllvWC+nlbf2///u/TT6u/J43VYcbCyjV7231fhOnsCKwDBs2rMkyRP2KU57VKutf5XNEKDnkkEOK7RsBNU4pVM6qi1MS3/ve94rTLdWBqXrfb852i3oSp/qqjxHVx6zmitczcODAumn1ZbGfVovTrTGL5cYbb1xjPc0NDnEqOwLbfffdt8Yg6XiOaAjROOGjjYuejvHjx6dXXnmlOFdaea68UrSGInhcf/31Df69sR18XUT4iZ6V2HHj/G20WGL8RHx4x7U0ogyV1mW8RqV43hDnfw8++OAP3PsR4aOp3o9omcWYmDinG7MO4meMgYixGTlVH6Qrt3djGtu2f+25b1r0SESLNc6H/+d//mdxvj5auNFj0dQ4huaWP+pB3Bc9Ow2Vs/rDobIF2dZ80O1c/oCPD9vG6lP5gzjqXC3W+WFFWIzGTkNibEm06iPYxriaCJzRuIhegjgWNHffr3wN8ZgYA3LJJZc0uGyE45YS+0BMo43xV/H+RJ2M8sQ4lerX0ljjI6b7x7Eqyh9ljQAdvS/R8GnOc3R2wkcbF4PCYiBpnA6IwZCNiYNBdDnvvffeTR7Mt9pqq7pWR2XrMVpTa+uRiDAQA7Bi0FrlQLoYvFVL0U0craQIAzHF9oOEmAgfU6dOLVo3jQW2aBXGdosejwhV0bKLgZLNEY+LUxdxoajqlnzlto73JFq4la3bchd1+b0o91pUX0vgw/SMNBVqQrQQY3Bc3KLlFwNN47U3J3xE3YnBrmWLFi0qDrYxCDWUB9vGMuUeiVqKdUXXduVzR70M5TI09dpziO0Y9TYG/jY26LQ8XTw+8NZFNCQiBMTA5qZE/XrqqafWuL+6/jVHXC8lTutFQK28Fkqcqvygop48+uijxYd4Ld6veD0xADym0VcG3OptEMe5WC6ODdGLW1mvqzVWrhhcGtOQowe4sqdqXU8VdWbGfLRxsRPFqOpoxcc53qaSfLSUzznnnDX+FrMAyh9s0bKJD8uYeVDZCokP6rUph4DKx8X/f/SjH6VaWn/99YvWVHTJxs+GWnxxDYKYabO23o847dLURZLiFEv0jkT3aRxoGhtTUy1mH8Q53ug5qVYub5zOiPekeploGcW6yh/08UES40fmz59fb7k4JfRhRK9UdaCJ8lR3K0ePWZweioNpc1x++eX1fo+6FMqvJ8YIRF2Jg3v1exe/x4fYh1W5TeM54/eo1/FBVq5DobUuDhUt4RjHEuGzoetkxAXz4vRFzIKKUxTrIk6LRI9gfADGNW+aqn+xj8RpgcpxWjGrI0JadY/Luu77Mcbmw9TROGZFr+lVV121xt9irNu6XoAtXm8c6yq3d9T3cv1s6rU0dgyMfaihetTQc8R+NXPmzHUqc2em56MdaGiqWLXoEo0ekpj+GB+2MaAyDsaR5mNgVwSEGC8SraY4xRDLxVTR2GHjwze6yMtTyRoTXYzRWonHx0EjPjRj4OXaekw+iOgOjZ6ImJobrYkoewyEi9NPMdguDqoNXeG0obEf0boqH0QaOvUS3cgRUqLXqNxyXptozUbLNaYdRlniNEYcLOPDJnoTYspfhMWRI0cWF0aLsQi77LJLcZoj1hXBqHJQYUy1jAs7xc8YRBhBpNya/6BiPEeUJ7qFI1xET0Sc/44Pu9ieUZ4It7HMgw8+WGzr5ojWbkwvjBZ7fLBFEIzQFs8X4nXFVO5JkyYVrzs+KKPnJx4Xp3pibMGHubhWdPfH9NTYL6L3Kuru7bffXvSSlU8vRu9ffLhGb2H0kMR06hjfsLZxErUUdS96GaI+RHnLPRwx3TrqQOyzzd3mDV29NupSPEd5mur//d//Fft6TDWN3r7TTz+9bqp+DHqObRC9lvE+xH67LhdUi0Hl0UMX2zyeK8Jz9Op8mFNB3/jGN9LPfvazYvB07OOx/0VYiG0W98d2amhAbWNif4vniNcd9S7e/+ipqQ7bcdyK3psY+xQ9lzHmKLZlQ704sQ+F2IdjKngcU2M9cXyN0yzx/zjuRm9LhKgI8vE+0AzNmBFDK021bUpDFxkLV155ZTGVNS7sE9N0d9ppp9K3v/3t0ssvv1y3TEyxi2l1cbGgdb3I2O9+97viglQbbrhhcdGw8ePH100FrJzK1tD01srpms31i1/8oriYVFzMKqZpRpnjYlUxFbChqbaNra+hspTtsccexTLreoGgmGr33e9+t5i6+JGPfKSYShjTaOPCTpVTG08++eTiIkuxTFyQrPoiY+XnigubxTTYeN++9rWvFdOiG5tqW55SWl1vKqcEPvnkk6V99923eI/L06hjymNcxGmXXXYp1hPbJf7fnNdeXnfUgXid8fiY1nz88cc3eJGxmKK5zz77FOuIW1w0Lqb4PvXUU2tckKq5GrrIWFwgKspWPT363nvvLfaFmLr6YS4y1lAZYl9pjtjeceG/KEeUO8obF16bOnVq6b333ltj+fJFxqo1NNU0prbHlNu4yFVcKCumeMZjG7rI2EYbbVRc7G348OHNvshYtZj6GlPRyxcNi+PKnXfe+aG2W2yDuPhYLB+vIepTbKs4Pi1durTJ19+QP/3pT6VvfOMbdRcZi/83dJGxuEBbTE+P7RLLxWUM4hjZUD0pX5AxpgVX7mO33npraeeddy62a0xdjtdRnoLf0NRc6usS/zQnpACdW5z6i1MpMT5obb1kLSUGNscFq6KlCbRfxnwAAFkJHwBAVsIHAJCVMR8AQFZ6PgCArIQPAKBzX2QsLp0c37YZFyVq7UskAwDNE6M44usk4qKGa7uIXZsLHxE8WvILhQCAlrN48eK1fm1Amwsf5S/gisKv7avBAYC2YdmyZUXnQeUXabab8FE+1RLBQ/gAgPalOUMmDDgFALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AIC2HT7mz5+fxowZU3xrXVxC9eabb17jW+3OOuusNHDgwNSrV680atSo9Mwzz9SyzABAZwofK1asSLvssku6/PLLG/z7hRdemH784x+nGTNmpN/+9rdpgw02SKNHj07vvvtuLcoLALRz6/zFcgcccEBxa0j0ekydOjV973vfSwcddFBx33XXXZc222yzoofksMMO+/AlBgDatZqO+XjuuefSK6+8UpxqKevbt28aMWJEuu+++xp8zMqVK4uv4a28AQAd1zr3fDQlgkeIno5K8Xv5b9WmTJmSJk+eXMti0AYNOf321i5Cu/H8+Qe2dhEAOvZsl0mTJqWlS5fW3RYvXtzaRQIA2kv42HzzzYufr776ar374/fy36r16NEj9enTp94NAOi4aho+hg4dWoSMuXPn1t0XYzhi1suee+5Zy1UBAJ1lzMfy5cvTokWL6g0yfeSRR1K/fv3S4MGD08SJE9O5556btt122yKMnHnmmcU1QQ4++OBalx0A6Azh46GHHkojR46s+/2UU04pfo4dOzZdc8016dvf/nZxLZCjjz46vfXWW2mfffZJc+bMST179qxtyQGAdqlLKS7O0YbEaZqYnhuDT43/6DjMdmk+s12A9mhdPr9bfbYLANC5CB8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAtO/wsWrVqnTmmWemoUOHpl69eqVtttkmnXPOOalUKtV6VQBAO7RerZ/wggsuSNOnT0/XXntt2nHHHdNDDz2Uxo0bl/r27ZtOPPHEWq8OAOjs4ePee+9NBx10UDrwwAOL34cMGZJuuOGG9MADD9R6VQBAO1Tz0y577bVXmjt3bnr66aeL3x999NH061//Oh1wwAENLr9y5cq0bNmyejcAoOOqec/H6aefXgSI7bbbLnXr1q0YA3Leeeelww8/vMHlp0yZkiZPnlzrYgC0miGn397aRWg3nj//r73kdC417/n42c9+lq6//vo0a9astHDhwmLsx8UXX1z8bMikSZPS0qVL626LFy+udZEAgI7c83HaaacVvR+HHXZY8ftOO+2UXnjhhaKHY+zYsWss36NHj+IGAHQONe/5eOedd1LXrvWfNk6/rF69utarAgDaoZr3fIwZM6YY4zF48OBiqu3DDz+cLrnkknTkkUfWelUAQDtU8/Axbdq04iJjxx13XFqyZEnaYost0jHHHJPOOuusWq8KAGiHah4+evfunaZOnVrcAACq+W4XACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAANp/+HjppZfS17/+9dS/f//Uq1evtNNOO6WHHnqoJVYFALQz69X6Cd9888209957p5EjR6Y77rgjbbrppumZZ55JG2+8ca1XBQC0QzUPHxdccEEaNGhQmjlzZt19Q4cOrfVqAIB2quanXW699da0++67p0MOOSQNGDAgfeITn0hXXXVVo8uvXLkyLVu2rN4NAOi4ah4+nn322TR9+vS07bbbpjvvvDMde+yx6cQTT0zXXnttg8tPmTIl9e3bt+4WvSYAQMdV8/CxevXqtOuuu6Yf/OAHRa/H0UcfncaPH59mzJjR4PKTJk1KS5curbstXry41kUCADpy+Bg4cGDaYYcd6t23/fbbpxdffLHB5Xv06JH69OlT7wYAdFw1Dx8x0+Wpp56qd9/TTz+dttpqq1qvCgBoh2oePk4++eR0//33F6ddFi1alGbNmpWuvPLKNGHChFqvCgBoh2oePvbYY480e/bsdMMNN6Rhw4alc845J02dOjUdfvjhtV4VANAO1fw6H+GLX/xicQMAqOa7XQCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAALISPgCArIQPACAr4QMAyEr4AACyEj4AgKyEDwAgK+EDAMhK+AAAshI+AICshA8AICvhAwDISvgAADpW+Dj//PNTly5d0sSJE1t6VQBAZw8fDz74YLriiivSzjvv3JKrAQDakRYLH8uXL0+HH354uuqqq9LGG2/c6HIrV65My5Ytq3cDADquFgsfEyZMSAceeGAaNWpUk8tNmTIl9e3bt+42aNCglioSANBRw8eNN96YFi5cWASLtZk0aVJaunRp3W3x4sUtUSQAoI1Yr9ZPGOHhpJNOSnfddVfq2bPnWpfv0aNHcQMAOoeah48FCxakJUuWpF133bXuvlWrVqX58+enyy67rBjj0a1bt1qvFgDorOHjs5/9bHrsscfq3Tdu3Li03Xbbpe985zuCBwB0cjUPH717907Dhg2rd98GG2yQ+vfvv8b9AEDn4wqnAED77vloyLx583KsBgBoB/R8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHANC+w8eUKVPSHnvskXr37p0GDBiQDj744PTUU0/VejUAQDtV8/Bxzz33pAkTJqT7778/3XXXXen9999P+++/f1qxYkWtVwUAtEPr1foJ58yZU+/3a665pugBWbBgQdp3331rvToAoLOHj2pLly4tfvbr16/Bv69cubK4lS1btqyliwQAdNQBp6tXr04TJ05Me++9dxo2bFijY0T69u1bdxs0aFBLFgkA6MjhI8Z+PP744+nGG29sdJlJkyYVvSPl2+LFi1uySABARz3tcvzxx6fbbrstzZ8/P2255ZaNLtejR4/iBgB0DjUPH6VSKZ1wwglp9uzZad68eWno0KG1XgUA0I6t1xKnWmbNmpVuueWW4lofr7zySnF/jOfo1atXrVcHAHT2MR/Tp08vxm7st99+aeDAgXW3m266qdarAgDaoRY57QIA0Bjf7QIAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAVsIHAJCV8AEAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAxwgfl19+eRoyZEjq2bNnGjFiRHrggQdaalUAQGcPHzfddFM65ZRT0tlnn50WLlyYdtlllzR69Oi0ZMmSllgdANDZw8cll1ySxo8fn8aNG5d22GGHNGPGjLT++uunq6++uiVWBwC0I+vV+gnfe++9tGDBgjRp0qS6+7p27ZpGjRqV7rvvvjWWX7lyZXErW7p0afFz2bJltS4arWj1yndauwjthrrf/qnvzae+d7z3slQq5Q8fr7/+elq1alXabLPN6t0fvz/55JNrLD9lypQ0efLkNe4fNGhQrYsG7ULfqa1dAshHfe943n777dS3b9+84WNdRQ9JjA8pW716dXrjjTdS//79U5cuXVJnSIoRtBYvXpz69OnT2sXpNGz31mG7tw7bvXV0tu1eKpWK4LHFFlusddmah49NNtkkdevWLb366qv17o/fN9988zWW79GjR3GrtNFGG6XOJipmZ6icbY3t3jps99Zhu7eOzrTd+66lx6PFBpx279497bbbbmnu3Ln1ejPi9z333LPWqwMA2pkWOe0Sp1HGjh2bdt999zR8+PA0derUtGLFimL2CwDQubVI+Dj00EPTa6+9ls4666z0yiuvpI9//ONpzpw5awxC5a+nneJ6KNWnnmhZtnvrsN1bh+3eOmz3xnUpNWdODABAjfhuFwAgK+EDAMhK+AAAshI+AICshA8AICvhow15/vnn01FHHZWGDh2aevXqlbbZZptimlZ8WR8t67zzzkt77bVX8e3LnfEKu7lcfvnlaciQIalnz55pxIgR6YEHHmjtInVo8+fPT2PGjCkudx1fV3HzzTe3dpE6hfjOsj322CP17t07DRgwIB188MHpqaeeau1itSnCRxsSX7wXV4O94oor0hNPPJEuvfTSNGPGjHTGGWe0dtE6vAh4hxxySDr22GNbuygd1k033VRcgDAC9cKFC9Muu+ySRo8enZYsWdLaReuw4uKOsZ0j9JHPPffckyZMmJDuv//+dNddd6X3338/7b///sX7wV+5zkcbd9FFF6Xp06enZ599trWL0ilcc801aeLEiemtt95q7aJ0ONHTEa3Byy67rPg9gnZ86dYJJ5yQTj/99NYuXocXPR+zZ88uWuHkFRfdjB6QCCX77rtvaxenTdDz0cYtXbo09evXr7WLAR+6Z2nBggVp1KhRdfd17dq1+P2+++5r1bJBjuN4cCz//wkfbdiiRYvStGnT0jHHHNPaRYEP5fXXX0+rVq1a4ysW4vf4CgboqKKHL3pT99577zRs2LDWLk6bIXxkEF3K0eXZ1C3Ge1R66aWX0uc///liHML48eNbreydbbsD1FKM/Xj88cfTjTfe2NpF6fhfLEd9p556ajriiCOaXGbrrbeu+//LL7+cRo4cWcy+uPLKKzOUsGNa1+1Oy9lkk01St27d0quvvlrv/vh98803b7VyQUs6/vjj02233VbMOtpyyy1buzhtivCRwaabblrcmiN6PCJ47LbbbmnmzJnFeXFafrvTsrp3717U6blz59YNeIzu6Pg9DtDQkcQ8jhhIHQN8582bV1w+gfqEjzYkgsd+++2Xttpqq3TxxRcXI6TLtA5b1osvvpjeeOON4meMTXjkkUeK+//mb/4mbbjhhq1dvA4hptmOHTs27b777mn48OFp6tSpxdTDcePGtXbROqzly5cXY8fKnnvuuaJux8DHwYMHt2rZOvqpllmzZqVbbrmluNZHeVxT3759i2s48deERhsxc+bMmPbc4I2WNXbs2Aa3+913393aRetQpk2bVho8eHCpe/fupeHDh5fuv//+1i5Shxb1t6F6HfWdltPYcTyO8fyV63wAAFkZUAAAZCV8AABZCR8AQFbCBwCQlfABAGQlfAAAWQkfAEBWwgcAkJXwAQBkJXwAAFkJHwBAyun/AYoSDSYzpSuwAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_variant_median_rates(op_cnv_df, pt_col=\"participant_id_legacy\", variant_col=\"hlvl_call\")\n",
    "plt.title(\"Median CNV counts per pt in OncoPanel data\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b2800fe-6bdb-4a47-b7e1-09ef645fc540",
   "metadata": {},
   "source": [
    "# Prepare the ICB WES data\n",
    "\n",
    "Get the participants who have pre-ICB WES tumor data but NOT OncoPanel call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "cdb7439e-01da-47ca-b9cd-13850d589104",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_sample_df = icb_sample_df[(icb_sample_df[\"pre_post\"] ==  \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\")]\n",
    "pt_usable_data = pre_sample_df[[\"participant_id_legacy\", \"usable_wes\", \"usable_oncopanel\"]].groupby(\"participant_id_legacy\").any()\n",
    "pt_usable_data = pt_usable_data[pt_usable_data[\"usable_wes\"] & (~pt_usable_data[\"usable_oncopanel\"])]\n",
    "\n",
    "wes_pt_df = icb_pt_df[icb_pt_df[\"participant_id_legacy\"].isin(pt_usable_data.index)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596f1605-7297-426a-98c7-dd44a854acf2",
   "metadata": {},
   "source": [
    "### Load and process ICB WES sequence variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "9688a22b-1975-4bd6-9403-4d9f364831c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "wes_mut_df = pd.read_pickle('../../../results/2023-10-11_process_mafs/all_mut_ccfs_maf.pickle')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f23ca0c2-dcd7-4548-bc90-54e95f592660",
   "metadata": {},
   "source": [
    "Filter WES sequence variants by various criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "13c81242-ebf2-4787-8527-655c14f08d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants for pts without OncoPanel data\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"Patient_ID\"].isin(wes_pt_df[\"participant_id_legacy\"])]\n",
    "\n",
    "# Correct some sample IDs that erroneously end with \"_pair\"\n",
    "wes_mut_df[\"Sample_ID\"] = wes_mut_df[\"Sample_ID\"].map(lambda x: x[:-5] if x.endswith(\"_pair\") else x)\n",
    "\n",
    "# Filter by CCF\n",
    "cluster_ccf_threshold = 0.1\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"clust_ccf_mean\"] > cluster_ccf_threshold]\n",
    "\n",
    "# Restrict to genes covered by OncoPanel\n",
    "wes_mut_df = wes_mut_df[wes_mut_df[\"Hugo_Symbol\"].isin(oncopanel_genes[\"Gene\"])]\n",
    "\n",
    "# Remove silent and non-coding mutations. \n",
    "# This is consistent with the filtering of GLASS variants\n",
    "wes_mut_df = wes_mut_df[~wes_mut_df[\"variant_type\"].isin([\"silent\", \"non_coding\"])]\n",
    "\n",
    "# Keep only variants belonging to their latest pre-ICB timepoint.\n",
    "wes_mut_df = keep_latest_preicb_timepoint_variants(wes_mut_df, pre_sample_df, sample_id_col=\"Sample_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "4200ae71-3748-4d01-87c9-150e7d6c63ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_type\n",
       "neoantigen    26\n",
       "nonsense       4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_mut_df[\"variant_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "2e45eb41-967d-4d23-a17a-e4846f3b8838",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_df[\"gene_mut\"] = wes_mut_df[\"Hugo_Symbol\"]+\"_mut\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "a6c736e7-fe32-4515-bfbf-602afc6f86c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_mut_matrix = wes_mut_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_mut', aggfunc = 'size', fill_value = 0)\n",
    "\n",
    "# In this methodology, sequence variants are given weight = 2\n",
    "wes_mut_matrix *= 2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "c760f96b-ab72-482e-8fe0-60f6c0492ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_mut</th>\n",
       "      <th>BRD4_mut</th>\n",
       "      <th>EGFR_mut</th>\n",
       "      <th>ERCC2_mut</th>\n",
       "      <th>FLT3_mut</th>\n",
       "      <th>GNAS_mut</th>\n",
       "      <th>KDR_mut</th>\n",
       "      <th>MITF_mut</th>\n",
       "      <th>MSH6_mut</th>\n",
       "      <th>NOTCH1_mut</th>\n",
       "      <th>PDGFRA_mut</th>\n",
       "      <th>PIK3CA_mut</th>\n",
       "      <th>PTEN_mut</th>\n",
       "      <th>RB1_mut</th>\n",
       "      <th>TET2_mut</th>\n",
       "      <th>TP53_mut</th>\n",
       "      <th>TSC1_mut</th>\n",
       "      <th>WT1_mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-139</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-144</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-212</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-253</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-28</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-30</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-32</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-37</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-43</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-69</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-84</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_mut               BRD4_mut  EGFR_mut  ERCC2_mut  FLT3_mut  GNAS_mut  \\\n",
       "participant_id_legacy                                                      \n",
       "GBM.ICB-139                 0.0       4.0        0.0       0.0       0.0   \n",
       "GBM.ICB-144                 0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-212                 0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-253                 2.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-28                  0.0       2.0        0.0       0.0       0.0   \n",
       "GBM.ICB-30                  0.0       2.0        0.0       0.0       0.0   \n",
       "GBM.ICB-32                  0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-37                  0.0       0.0        0.0       0.0       2.0   \n",
       "GBM.ICB-42                  0.0       0.0        0.0       0.0       0.0   \n",
       "GBM.ICB-43                  0.0       0.0        2.0       0.0       0.0   \n",
       "GBM.ICB-69                  0.0       0.0        0.0       2.0       0.0   \n",
       "GBM.ICB-84                  0.0       0.0        0.0       0.0       0.0   \n",
       "\n",
       "gene_mut               KDR_mut  MITF_mut  MSH6_mut  NOTCH1_mut  PDGFRA_mut  \\\n",
       "participant_id_legacy                                                        \n",
       "GBM.ICB-139                0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-144                0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-212                0.0       0.0       0.0         2.0         0.0   \n",
       "GBM.ICB-253                0.0       2.0       0.0         0.0         0.0   \n",
       "GBM.ICB-28                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-30                 2.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-32                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-37                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-42                 0.0       0.0       0.0         0.0         2.0   \n",
       "GBM.ICB-43                 0.0       0.0       2.0         0.0         0.0   \n",
       "GBM.ICB-69                 0.0       0.0       0.0         0.0         0.0   \n",
       "GBM.ICB-84                 0.0       0.0       0.0         0.0         0.0   \n",
       "\n",
       "gene_mut               PIK3CA_mut  PTEN_mut  RB1_mut  TET2_mut  TP53_mut  \\\n",
       "participant_id_legacy                                                      \n",
       "GBM.ICB-139                   0.0       6.0      0.0       0.0       0.0   \n",
       "GBM.ICB-144                   0.0       2.0      0.0       0.0       2.0   \n",
       "GBM.ICB-212                   2.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-253                   0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-28                    0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-30                    0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-32                    0.0       4.0      0.0       0.0       0.0   \n",
       "GBM.ICB-37                    0.0       2.0      0.0       0.0       2.0   \n",
       "GBM.ICB-42                    0.0       2.0      0.0       0.0       0.0   \n",
       "GBM.ICB-43                    0.0       0.0      0.0       2.0       0.0   \n",
       "GBM.ICB-69                    0.0       0.0      0.0       0.0       0.0   \n",
       "GBM.ICB-84                    0.0       2.0      2.0       0.0       2.0   \n",
       "\n",
       "gene_mut               TSC1_mut  WT1_mut  \n",
       "participant_id_legacy                     \n",
       "GBM.ICB-139                 0.0      0.0  \n",
       "GBM.ICB-144                 0.0      0.0  \n",
       "GBM.ICB-212                 0.0      0.0  \n",
       "GBM.ICB-253                 0.0      0.0  \n",
       "GBM.ICB-28                  0.0      0.0  \n",
       "GBM.ICB-30                  0.0      0.0  \n",
       "GBM.ICB-32                  0.0      0.0  \n",
       "GBM.ICB-37                  0.0      2.0  \n",
       "GBM.ICB-42                  0.0      0.0  \n",
       "GBM.ICB-43                  2.0      0.0  \n",
       "GBM.ICB-69                  0.0      0.0  \n",
       "GBM.ICB-84                  0.0      0.0  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_mut_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3800996-3675-4244-9470-a9560a59295b",
   "metadata": {},
   "source": [
    "### Load and process ICB WES CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "982f90c4-3727-404f-aa85-b9088d269910",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df = pd.concat([\n",
    "    pd.read_csv('../data/icb_wes_tnpair_gene_allelic_cn.maf', sep = '\\t'),\n",
    "    pd.read_csv('../data/icb_wes_tonly_gene_allelic_cn.maf', sep = '\\t')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "1cad9964-a6ef-4c82-b590-718b2a56cd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some helper functions\n",
    "def call_hlvl(cn_col, thresholds=[0.8, 1.8, 2.2, 3.2], values=[-2,-1,0,1,2]): \n",
    "    dig = np.digitize(cn_col.values, thresholds)\n",
    "    hlvl = np.array(values)[dig]\n",
    "    return pd.Series(data=hlvl, index=cn_col.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "644b4b32-63a1-4fe2-8eef-5bda3e4e1c9e",
   "metadata": {},
   "source": [
    "Filter the variants by various criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "0c9037c7-cf2b-403a-9658-85842481f8f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only variants belonging to the latest pre-ICB timepoint for each pt\n",
    "wes_cnv_df = keep_latest_preicb_timepoint_variants(wes_cnv_df, icb_sample_df, sample_id_col=\"sample\")\n",
    "\n",
    "# Restrict to ICB pts that only have WES data (and not OncoPanel)\n",
    "wes_cnv_df = wes_cnv_df[wes_cnv_df[\"participant_id_legacy\"].isin(wes_pt_df[\"participant_id_legacy\"])]\n",
    "\n",
    "# Restrict to genes covered by OncoPanel\n",
    "wes_cnv_df = wes_cnv_df[wes_cnv_df[\"gene\"].isin(oncopanel_genes[\"Gene\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e6916e7-1cc0-411d-8230-d91d507e4dc8",
   "metadata": {},
   "source": [
    "Annotate gene-level copy number gains, losses, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "8ad22c0b-1572-4cbd-8bfc-66730b4bb605",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df[\"hlvl_call\"] = call_hlvl(wes_cnv_df[\"corrected_total_cn\"])\n",
    "\n",
    "# Use ABSOLUTE's homozygous deletions\n",
    "wes_cnv_df.loc[wes_cnv_df[\"HZ\"] == 1, \"hlvl_call\"] = -2\n",
    "\n",
    "# Finally: keep only the actual CNVs\n",
    "wes_cnv_df = wes_cnv_df.loc[wes_cnv_df[\"hlvl_call\"] != 0,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "aa5055d1-39cf-43ad-821e-a43c65c8580d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_df[\"gene_cnv\"] = wes_cnv_df[\"gene\"] + \"_cnv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "651320b3-ae3e-4616-b214-31d0a20b99e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhYAAAGdCAYAAABO2DpVAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAEe5JREFUeJzt3W2IVQX+wPHfqDjWplP2sCaO2QNblGT0YFRLWGu1EVK9iFiCJomIsEgqqNkXu+uLGGGXcilx3ah8s6L0woKiljZWfVHTw4iQRZGVND3aA82of5rCuX/OiZltSq3R350zc+fzgcN4L2fm/jhexq/nnHtOU61WqwUAQIIJGT8EAKAgLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANJNihPX398fHH38cU6dOjaamppF+eQDgIBTX09y1a1fMnDkzJkyYMHrCooiK1tbWkX5ZACBBd3d3zJo1a/SERbGnYmCwadOmjfTLAwAHobe3t9wxMPDv+KgJi4HDH0VUCAsAGFt+7jQGJ28CAGmEBQCQRlgAAGmEBQCQRlgAAGmEBQCQRlgAAGmEBQCQRlgAAGmEBQBQTVj85S9/KS/l+cPltNNOy5sGABjThn2vkDPOOCP+85///O8HTBrx240AAKPUsKugCIkZM2bUZxoAYHydY/HOO+/EzJkz46STToobbrghPvjggwOu39fXV95q9YcLANCYmmq1Wu2Xrvzss8/G7t2749RTT41PPvkkli1bFh999FFs27Ztv/dnL87LKNb7sZ6eHrdNB8acOfc9U/UIY8aO5VdVPQKJih0DLS0tP/vv97DC4se+/vrrOOGEE+KBBx6Im2++eb97LIrlh4O1trYKC2BMEha/nLAYn2FxSGdeHnnkkfGb3/wmtm/fvt91mpubywUAaHyHdB2L4rDIu+++G8cff3zeRADA+AiLe+65JzZt2hQ7duyIF198Ma699tqYOHFi/OEPf6jfhADAmDGsQyEffvhhGRFffvllHHvssfHb3/42Ojs7yz8DAAwrLNatW1e/SQCAMc+9QgCANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICABgdYbF8+fJoamqKpUuX5k0EAIy/sHj11Vdj9erVceaZZ+ZOBACMr7DYvXt33HDDDfHII4/EUUcdlT8VADB+wmLJkiVx1VVXxcKFC3923b6+vujt7R2yAACNadJwv2HdunWxZcuW8lDIL9HR0RHLli07mNkAgEbeY9Hd3R133nln/Otf/4opU6b8ou9pb2+Pnp6ewaX4GQBAYxrWHouurq7YuXNnnH322YPP7d27NzZv3hwPP/xwedhj4sSJQ76nubm5XACAxjessPjd734Xr7/++pDnFi9eHKeddlrce++9P4kKAGB8GVZYTJ06NebOnTvkuV/96ldx9NFH/+R5AGD8ceVNAKC6T4X82MaNG3MmAQDGPHssAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIBqwmLVqlVx5plnxrRp08rlggsuiGeffTZvGgBg/ITFrFmzYvny5dHV1RWvvfZaXHrppXH11VfHG2+8Ub8JAYAxY9JwVl60aNGQx/fff3+5F6OzszPOOOOM7NkAgEYOix/au3dvPPHEE7Fnz57ykMj+9PX1lcuA3t7eg31JAKDRwuL1118vQ+Kbb76JI444IjZs2BCnn376ftfv6OiIZcuWHeqcjGJz7num6hHGjB3Lr6p6BIDR9amQU089NbZu3Rovv/xy3HbbbdHW1hZvvvnmftdvb2+Pnp6ewaW7u/tQZwYAGmWPxeTJk+OUU04p/3zOOefEq6++Gn//+99j9erV+1y/ubm5XACAxnfI17Ho7+8fcg4FADB+DWuPRXFY48orr4zZs2fHrl27Yu3atbFx48b497//Xb8JAYDGDIudO3fGjTfeGJ988km0tLSUF8sqouKyyy6r34QAQGOGxaOPPlq/SQCAMc+9QgCANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACAasKio6MjzjvvvJg6dWocd9xxcc0118Tbb7+dNw0AMH7CYtOmTbFkyZLo7OyM559/Pr777ru4/PLLY8+ePfWbEAAYMyYNZ+XnnntuyOM1a9aUey66urri4osvzp4NAGjksPixnp6e8uv06dP3u05fX1+5DOjt7T2UlwQAGvHkzf7+/li6dGlcdNFFMXfu3AOel9HS0jK4tLa2HuxLAgCNGhbFuRbbtm2LdevWHXC99vb2cs/GwNLd3X2wLwkANOKhkNtvvz2efvrp2Lx5c8yaNeuA6zY3N5cLAND4hhUWtVot7rjjjtiwYUNs3LgxTjzxxPpNBgA0dlgUhz/Wrl0bTz31VHkti08//bR8vjh34rDDDqvXjABAI55jsWrVqvI8iQULFsTxxx8/uKxfv75+EwIAjXsoBABgf9wrBABIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgOrCYvPmzbFo0aKYOXNmNDU1xZNPPpk3DQAwvsJiz549MW/evFi5cmV9JgIAxqxJw/2GK6+8slwAAA45LIarr6+vXAb09vbW+yUBgEY9ebOjoyNaWloGl9bW1nq/JADQqGHR3t4ePT09g0t3d3e9XxIAaNRDIc3NzeUCADQ+17EAAKrbY7F79+7Yvn374OP3338/tm7dGtOnT4/Zs2fnTQYANH5YvPbaa3HJJZcMPr7rrrvKr21tbbFmzZrc6QCAxg6LBQsWRK1Wq880AMCY5hwLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0kzK+1EAUB9z7num6hHGjB3Lr6r09e2xAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAoNqwWLlyZcyZMyemTJkS559/frzyyit5EwEA4ycs1q9fH3fddVf8+c9/ji1btsS8efPiiiuuiJ07d9ZnQgCgccPigQceiFtuuSUWL14cp59+evzjH/+Iww8/PB577LH6TAgAjBmThrPyt99+G11dXdHe3j743IQJE2LhwoXx0ksv7fN7+vr6ymVAT09P+bW3t/fgp2ZU6e/7v6pHGDO878c+7/dq3u+2e/W/ZwZ+bq1WywuLL774Ivbu3Ru//vWvhzxfPH7rrbf2+T0dHR2xbNmynzzf2to6nJeGhtCyouoJYOR4vzfmdt+1a1e0tLTkhMXBKPZuFOdkDOjv74+vvvoqjj766GhqaopGVxReEVHd3d0xbdq0qscZN2z3atju1bDdR9543Oa1Wq2MipkzZx5wvWGFxTHHHBMTJ06Mzz77bMjzxeMZM2bs83uam5vL5YeOPPLIGG+KN954efONJrZ7NWz3atjuI2+8bfOWA+ypOKiTNydPnhznnHNOvPDCC0P2QBSPL7jggoObEgBoGMM+FFIc1mhra4tzzz035s+fHytWrIg9e/aUnxIBAMa3YYfF9ddfH59//nn86U9/ik8//TTOOuuseO65535yQiffKw4DFdf8+PHhIOrLdq+G7V4N233k2eb711T7uc+NAAD8Qu4VAgCkERYAQBphAQCkERYAQBphMUJ27NgRN998c5x44olx2GGHxcknn1yeUVzcf4X6uv/+++PCCy8sb5Y3Hi/ONlJWrlwZc+bMiSlTpsT5558fr7zyStUjNbzNmzfHokWLyishFlcyfvLJJ6seqeEVt6k477zzYurUqXHcccfFNddcE2+//XbVY40qwmKEFPdSKS4mtnr16njjjTfiwQcfLO8M+8c//rHq0RpeEW/XXXdd3HbbbVWP0rDWr19fXuOmiOUtW7bEvHnz4oorroidO3dWPVpDK64hVGzrIuoYGZs2bYolS5ZEZ2dnPP/88/Hdd9/F5ZdfXv5d8D0fN63QX//611i1alW89957VY8yLqxZsyaWLl0aX3/9ddWjNJxiD0Xxv7iHH364fFxEdHEfhTvuuCPuu+++qscbF4o9Fhs2bCj/B83IKa7rVOy5KILj4osvrnqcUcEeiwoVt5CfPn161WPAIe8R6urqioULFw4+N2HChPLxSy+9VOlsMBK/xwt+l/+PsKjI9u3b46GHHopbb7216lHgkHzxxRexd+/en1x9t3hcXJ0XGlWxZ67YC3rRRRfF3Llzqx5n1BAWh6jYzVvsgjzQUpxf8UMfffRR/P73vy+P+99yyy2VzT7etjtApuJci23btsW6deuqHmVs3yuEoe6+++646aabDrjOSSedNPjnjz/+OC655JLyUwr//Oc/R2DCxjTc7U79HHPMMTFx4sT47LPPhjxfPJ4xY0Zlc0E93X777fH000+Xn8yZNWtW1eOMKsLiEB177LHl8ksUeyqKqChuPf/444+Xx6Gp/3anviZPnly+p1944YXBEweLXcTF4+KXLzSS4vMOxUnJxYmyGzduLC8hwFDCYoQUUbFgwYI44YQT4m9/+1t5JvEA/6urrw8++CC++uqr8mtxLsDWrVvL50855ZQ44ogjqh6vIRQfNW1ra4tzzz035s+fHytWrCg/frd48eKqR2tou3fvLs/XGvD++++X7+/iRMLZs2dXOlsjH/5Yu3ZtPPXUU+W1LAbOI2ppaSmvUcT39cUIePzxx4uP9e5zob7a2tr2ud3/+9//Vj1aQ3nooYdqs2fPrk2ePLk2f/78WmdnZ9UjNbziPbyv93bxnqc+9vd7vPgdz/dcxwIASOMgPwCQRlgAAGmEBQCQRlgAAGmEBQCQRlgAAGmEBQCQRlgAAGmEBQCQRlgAAGmEBQCQRlgAAJHl/wEUIclITdx6nwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_variant_median_rates(wes_cnv_df, pt_col=\"participant_id_legacy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "204cc9d2-fd78-45f7-800b-ac352bde7987",
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_cnv_matrix = wes_cnv_df.pivot_table(index = 'participant_id_legacy', columns = 'gene_cnv', aggfunc = 'sum', values = \"hlvl_call\", fill_value = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "812517ac-9778-4704-afd7-7f002f722f15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_cnv</th>\n",
       "      <th>AKT1_cnv</th>\n",
       "      <th>AKT3_cnv</th>\n",
       "      <th>ARID2_cnv</th>\n",
       "      <th>BAP1_cnv</th>\n",
       "      <th>BUB1B_cnv</th>\n",
       "      <th>CARD11_cnv</th>\n",
       "      <th>CCND1_cnv</th>\n",
       "      <th>CCND2_cnv</th>\n",
       "      <th>CCNE1_cnv</th>\n",
       "      <th>CDK4_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>PTEN_cnv</th>\n",
       "      <th>RAD21_cnv</th>\n",
       "      <th>RARA_cnv</th>\n",
       "      <th>RB1_cnv</th>\n",
       "      <th>RET_cnv</th>\n",
       "      <th>SETD2_cnv</th>\n",
       "      <th>SMARCB1_cnv</th>\n",
       "      <th>SOX2_cnv</th>\n",
       "      <th>TNFAIP3_cnv</th>\n",
       "      <th>TP53_cnv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-139</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-144</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-212</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-253</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-28</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-30</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-32</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-37</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-42</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-43</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-69</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-84</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_cnv               AKT1_cnv  AKT3_cnv  ARID2_cnv  BAP1_cnv  BUB1B_cnv  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-139                  -1         0          0         0         -1   \n",
       "GBM.ICB-144                   0         1          0         1          0   \n",
       "GBM.ICB-212                  -1         0          0         1          0   \n",
       "GBM.ICB-253                   2         1          2         1          2   \n",
       "GBM.ICB-28                    0         0          0         0          0   \n",
       "GBM.ICB-30                    0         1          0         0         -1   \n",
       "GBM.ICB-32                    0         0          0         0          0   \n",
       "GBM.ICB-37                    2         2          1         1          0   \n",
       "GBM.ICB-42                    0         0          0         0         -1   \n",
       "GBM.ICB-43                   -1         0          0         0         -1   \n",
       "GBM.ICB-69                   -1         0          0         0          0   \n",
       "GBM.ICB-84                    0         0          0         0          0   \n",
       "\n",
       "gene_cnv               CARD11_cnv  CCND1_cnv  CCND2_cnv  CCNE1_cnv  CDK4_cnv  \\\n",
       "participant_id_legacy                                                          \n",
       "GBM.ICB-139                     3          0          0          3         0   \n",
       "GBM.ICB-144                     1          0          2          1         2   \n",
       "GBM.ICB-212                     1          0          0          0         0   \n",
       "GBM.ICB-253                     4          1          2          2         2   \n",
       "GBM.ICB-28                      1          0          0          0         0   \n",
       "GBM.ICB-30                      1          0          0          1         0   \n",
       "GBM.ICB-32                      0         -2          0          0         0   \n",
       "GBM.ICB-37                      2          1          1          1         1   \n",
       "GBM.ICB-42                      1          0          0          0         0   \n",
       "GBM.ICB-43                      0          0          0         -1         0   \n",
       "GBM.ICB-69                      1          0          0          0         0   \n",
       "GBM.ICB-84                      1          0          0          1         0   \n",
       "\n",
       "gene_cnv               ...  PTEN_cnv  RAD21_cnv  RARA_cnv  RB1_cnv  RET_cnv  \\\n",
       "participant_id_legacy  ...                                                    \n",
       "GBM.ICB-139            ...        -3          0         0        0       -3   \n",
       "GBM.ICB-144            ...         0          1         1        0        1   \n",
       "GBM.ICB-212            ...         0          0         0        0        0   \n",
       "GBM.ICB-253            ...         0          2         0        2        0   \n",
       "GBM.ICB-28             ...        -1          0         0        0        0   \n",
       "GBM.ICB-30             ...        -1          1         0        0       -1   \n",
       "GBM.ICB-32             ...         0          0         0        0        0   \n",
       "GBM.ICB-37             ...         0          0         1        0        0   \n",
       "GBM.ICB-42             ...        -1          0         0       -1       -1   \n",
       "GBM.ICB-43             ...        -1          0         0       -1       -1   \n",
       "GBM.ICB-69             ...        -1          0         0        0       -1   \n",
       "GBM.ICB-84             ...        -1          0         0       -1       -1   \n",
       "\n",
       "gene_cnv               SETD2_cnv  SMARCB1_cnv  SOX2_cnv  TNFAIP3_cnv  TP53_cnv  \n",
       "participant_id_legacy                                                           \n",
       "GBM.ICB-139                    0            0         0            0         0  \n",
       "GBM.ICB-144                    1            0         1            1         0  \n",
       "GBM.ICB-212                    1            0         0            0         0  \n",
       "GBM.ICB-253                    1            2         2            1         2  \n",
       "GBM.ICB-28                     0            2         0            0         0  \n",
       "GBM.ICB-30                     0            0         0            0         0  \n",
       "GBM.ICB-32                     0            0         0            0         0  \n",
       "GBM.ICB-37                     1            0         1            2         0  \n",
       "GBM.ICB-42                     0            0         0            0         0  \n",
       "GBM.ICB-43                     0           -1         0           -1         0  \n",
       "GBM.ICB-69                     0            0         0            0         0  \n",
       "GBM.ICB-84                     0            0         0            0         0  \n",
       "\n",
       "[12 rows x 45 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_cnv_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55bc7889-e530-4e53-bfd7-b73ad29fa5c4",
   "metadata": {},
   "source": [
    "# Combine all filtered variants for ICB cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "dd3e94fd-0650-4588-b270-cc14aefbf8f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(154, 356)\n"
     ]
    }
   ],
   "source": [
    "op_matrix = op_variants_matrix.join(op_cnvs_matrix, how = 'outer').fillna(0)\n",
    "print(op_matrix.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "9dd30925-560e-406a-b46d-08d4cf816dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_wes_matrix = wes_mut_matrix.join(wes_cnv_matrix, how=\"outer\").fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "5ec53d92-cbe0-49b6-b1f6-49dce8809ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_matrix = pd.concat((op_matrix, icb_wes_matrix)).fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f947d998-d575-42b9-9751-5c365267133a",
   "metadata": {},
   "source": [
    "Keep only alteration types significantly associated with clusters in GLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "19554dd1-5331-435e-bd2f-0b28f9477586",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glass_alterations = list(matrix_sig)\n",
    "missing = set(glass_alterations) - set(icb_matrix)\n",
    "len(missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "3a0e9f94-351d-4d14-812b-0ac2f5bf2a67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(166, 373) (365, 373)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n",
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/3716084494.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  icb_matrix[c] = 0.0\n"
     ]
    }
   ],
   "source": [
    "for c in missing:\n",
    "    icb_matrix[c] = 0.0\n",
    "\n",
    "icb_matrix = icb_matrix[list(matrix_sig)]\n",
    "\n",
    "print(icb_matrix.shape, matrix_sig.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "c73c84e0-95a3-4d35-8c11-939c937bf136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "166"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_matrix.index.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "6c55572a-290b-4107-a513-9a97072cf82d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "167"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_sample_df[(icb_sample_df[\"pre_post\"] == \"Pre\") & (icb_sample_df[\"tumor_normal\"] == \"tumor\") & (icb_sample_df[\"usable_oncopanel\"] | icb_sample_df[\"usable_wes\"])][\"participant_id_legacy\"].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0da39d67-4654-46bf-b549-ca43a864b683",
   "metadata": {},
   "source": [
    "# Classify the ICB participants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08d29fc6-9055-4af7-b8f5-cea53f0f1e92",
   "metadata": {},
   "source": [
    "## Apply the best classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "d4bbd46d-4153-43c8-9d79-a9164e12b9f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#best_model.fit(X, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "a509105d-b3b4-4e64-904d-1c850ca9006b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cluster\n",
       "cluster_3    92\n",
       "cluster_2    50\n",
       "cluster_4    12\n",
       "cluster_6     6\n",
       "cluster_1     6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_labels = [f\"cluster_{c}\" for c in best_model.classes_]\n",
    "\n",
    "icb_pred = best_model.predict_proba(icb_matrix)\n",
    "icb_pred_df = pd.DataFrame(data=icb_pred, index = icb_matrix.index, columns=cluster_labels)\n",
    "icb_pred_df[\"cluster\"] = icb_pred_df.apply(lambda row: cluster_labels[np.argmax(row)], axis=1)\n",
    "\n",
    "icb_pred_df['cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "a5d09203-1758-48d3-b383-0546cd4869e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "156\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/37/f0kgv16d10q0_dy3d9lsxqbr0000gq/T/ipykernel_5044/2732675569.py:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  icb_results['mgmt_methylated'] = icb_results['MGMT methylated Y/N'].replace({'Unmethylated': 0.0, 'Methylated': 1.0, 'Partially': 1.0})\n"
     ]
    }
   ],
   "source": [
    "icb_results = icb_pred_df\n",
    "\n",
    "icb_results = icb_results.merge(icb_pt_df, left_on='participant_id_legacy', right_on=\"participant_id_legacy\")\n",
    "\n",
    "icb_results = icb_results[icb_results['MGMT methylated Y/N'].notnull()]\n",
    "\n",
    "icb_results['setting'] = icb_results['ICB for Newly diagnosed']\n",
    "icb_results['vital_status'] = icb_results['Deceased'].map({'Deceased': 1, 'Alive': 0})\n",
    "icb_results['mgmt_methylated'] = icb_results['MGMT methylated Y/N'].replace({'Unmethylated': 0.0, 'Methylated': 1.0, 'Partially': 1.0})\n",
    "\n",
    "print(icb_results['participant_id'].nunique()) #, icb_results['sample_id_legacy'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c0a3d5-cfaa-439f-b68a-752844941f08",
   "metadata": {},
   "source": [
    "## Output predicted cluster assignments to TSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "59c4c4b6-d749-4554-9514-6c0c72a7a2e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_pt_df = icb_pt_df.merge(icb_pred_df, left_on=\"participant_id_legacy\", right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "3752fd1f-b28a-4c60-b1a2-1273a60fccaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_pt_df.to_csv(\"../results/icb_pt_w_cluster_assignments.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8574c80-74d9-4722-b809-5bb683c6bb3b",
   "metadata": {},
   "source": [
    "## Assess prognostic value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "a4f1e062-3882-4410-b6da-b189a6eb110b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 156  58\n",
      "[1] 37 58\n",
      "[1] 119  58\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ setting + mgmt_methylated + \n",
      "    cluster, data = combined)\n",
      "\n",
      "  n= 156, number of events= 144 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)   \n",
      "settingRecurrent  0.6657    1.9458   0.2047  3.252  0.00115 **\n",
      "mgmt_methylated  -0.3323    0.7172   0.1798 -1.849  0.06447 . \n",
      "clustercluster_2  0.9998    2.7178   0.5366  1.863  0.06244 . \n",
      "clustercluster_3  1.0022    2.7243   0.5217  1.921  0.05472 . \n",
      "clustercluster_4  0.8127    2.2541   0.5982  1.359  0.17427   \n",
      "clustercluster_6  0.3399    1.4047   0.6772  0.502  0.61577   \n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "settingRecurrent    1.9458     0.5139    1.3028     2.906\n",
      "mgmt_methylated     0.7172     1.3942    0.5043     1.020\n",
      "clustercluster_2    2.7178     0.3679    0.9494     7.780\n",
      "clustercluster_3    2.7243     0.3671    0.9799     7.574\n",
      "clustercluster_4    2.2541     0.4436    0.6979     7.281\n",
      "clustercluster_6    1.4047     0.7119    0.3725     5.297\n",
      "\n",
      "Concordance= 0.602  (se = 0.026 )\n",
      "Likelihood ratio test= 23.04  on 6 df,   p=8e-04\n",
      "Wald test            = 20.22  on 6 df,   p=0.003\n",
      "Score (logrank) test = 21.12  on 6 df,   p=0.002\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ mgmt_methylated + \n",
      "    cluster, data = nd)\n",
      "\n",
      "  n= 37, number of events= 32 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)   \n",
      "mgmt_methylated  -1.1774    0.3081   0.4257 -2.766  0.00568 **\n",
      "clustercluster_2  2.4127   11.1637   1.1003  2.193  0.02832 * \n",
      "clustercluster_3  1.9577    7.0827   1.0593  1.848  0.06460 . \n",
      "clustercluster_4  3.0070   20.2270   1.2649  2.377  0.01745 * \n",
      "clustercluster_6  2.2857    9.8323   1.4791  1.545  0.12227   \n",
      "---\n",
      "Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "mgmt_methylated     0.3081    3.24579    0.1338    0.7096\n",
      "clustercluster_2   11.1637    0.08958    1.2920   96.4626\n",
      "clustercluster_3    7.0827    0.14119    0.8881   56.4830\n",
      "clustercluster_4   20.2270    0.04944    1.6951  241.3561\n",
      "clustercluster_6    9.8323    0.10171    0.5415  178.5130\n",
      "\n",
      "Concordance= 0.706  (se = 0.044 )\n",
      "Likelihood ratio test= 13.33  on 5 df,   p=0.02\n",
      "Wald test            = 11.51  on 5 df,   p=0.04\n",
      "Score (logrank) test = 13.02  on 5 df,   p=0.02\n",
      "\n",
      "Call:\n",
      "coxph(formula = Surv(osicb, vital_status) ~ mgmt_methylated + \n",
      "    cluster, data = rc)\n",
      "\n",
      "  n= 119, number of events= 112 \n",
      "\n",
      "                    coef exp(coef) se(coef)      z Pr(>|z|)\n",
      "mgmt_methylated  -0.1738    0.8405   0.2032 -0.855    0.392\n",
      "clustercluster_2  0.6473    1.9104   0.6201  1.044    0.296\n",
      "clustercluster_3  0.7236    2.0619   0.6039  1.198    0.231\n",
      "clustercluster_4  0.3425    1.4084   0.6842  0.501    0.617\n",
      "clustercluster_6  0.1493    1.1610   0.7822  0.191    0.849\n",
      "\n",
      "                 exp(coef) exp(-coef) lower .95 upper .95\n",
      "mgmt_methylated     0.8405     1.1898    0.5644     1.252\n",
      "clustercluster_2    1.9104     0.5234    0.5667     6.441\n",
      "clustercluster_3    2.0619     0.4850    0.6313     6.735\n",
      "clustercluster_4    1.4084     0.7100    0.3684     5.384\n",
      "clustercluster_6    1.1610     0.8613    0.2506     5.378\n",
      "\n",
      "Concordance= 0.521  (se = 0.033 )\n",
      "Likelihood ratio test= 4.71  on 5 df,   p=0.5\n",
      "Wald test            = 4.21  on 5 df,   p=0.5\n",
      "Score (logrank) test = 4.31  on 5 df,   p=0.5\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/merrell/miniconda3/envs/dalmatian/lib/python3.11/site-packages/rpy2/robjects/pandas2ri.py:65: UserWarning: Error while trying to convert the column \"AE\". Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have <class 'float'> and <class 'str'>). If happening with a pandas DataFrame the method infer_objects() will normalize data types before conversion.\n",
      "  warnings.warn('Error while trying to convert '\n"
     ]
    }
   ],
   "source": [
    "%%R -i icb_results\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "\n",
    "combined <- icb_results %>%\n",
    "    mutate(cluster = as.factor(cluster))\n",
    "nd <- combined %>% filter(str_equal(setting, 'Newly-diagnosed'))\n",
    "rc <- combined %>% filter(str_equal(setting, 'Recurrent'))\n",
    "\n",
    "print(dim(combined))\n",
    "print(dim(nd))\n",
    "print(dim(rc))\n",
    "\n",
    "model_6_combined <- coxph(Surv(osicb, vital_status) ~ setting + mgmt_methylated + cluster,\n",
    "                          data = combined)\n",
    "model_6_nd <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "                          data = nd)\n",
    "model_6_rc <- coxph(Surv(osicb, vital_status) ~ mgmt_methylated + cluster,\n",
    "                          data = rc)\n",
    "\n",
    "print(summary(model_6_combined))\n",
    "print(summary(model_6_nd))\n",
    "print(summary(model_6_rc))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a1e240e-3b0f-4f23-a1bd-8f1b446156fc",
   "metadata": {},
   "source": [
    "## Create a heatmap of ICB cohort, grouped by predicted cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "509b1b70-e24a-41ba-a6a5-03989816c074",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTOR_mut</th>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <th>ID3_cnv</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>MYCL_mut</th>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <th>MPL_cnv</th>\n",
       "      <th>MUTYH_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>AR_mut</th>\n",
       "      <th>AR_cnv</th>\n",
       "      <th>ATRX_mut</th>\n",
       "      <th>ATRX_cnv</th>\n",
       "      <th>STAG2_cnv</th>\n",
       "      <th>BCORL1_mut</th>\n",
       "      <th>BCORL1_cnv</th>\n",
       "      <th>GPC3_mut</th>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <th>PHF6_cnv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-102</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-103</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-104</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-37</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-43</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-69</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-84</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>166 rows Ã— 373 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       MTOR_mut  MTOR_cnv  SDHB_cnv  ID3_cnv  ARID1A_mut  \\\n",
       "participant_id_legacy                                                      \n",
       "GBM.ICB-1                   0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-10                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-102                 0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-103                 0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-104                 0.0       0.0       0.0      0.0         0.0   \n",
       "...                         ...       ...       ...      ...         ...   \n",
       "GBM.ICB-37                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-42                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-43                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-69                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-84                  0.0       0.0       0.0      0.0         0.0   \n",
       "\n",
       "                       ARID1A_cnv  MYCL_mut  MYCL_cnv  MPL_cnv  MUTYH_cnv  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-1                     0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-10                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-102                   0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-103                   0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-104                   0.0       0.0       0.0      0.0        0.0   \n",
       "...                           ...       ...       ...      ...        ...   \n",
       "GBM.ICB-37                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-42                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-43                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-69                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-84                    0.0       0.0       0.0      0.0        0.0   \n",
       "\n",
       "                       ...  AR_mut  AR_cnv  ATRX_mut  ATRX_cnv  STAG2_cnv  \\\n",
       "participant_id_legacy  ...                                                  \n",
       "GBM.ICB-1              ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-10             ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-102            ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-103            ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-104            ...     0.0     0.0       0.0       0.0        0.0   \n",
       "...                    ...     ...     ...       ...       ...        ...   \n",
       "GBM.ICB-37             ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-42             ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-43             ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-69             ...     0.0     0.0       0.0       0.0        0.0   \n",
       "GBM.ICB-84             ...     0.0     0.0       0.0       0.0        0.0   \n",
       "\n",
       "                       BCORL1_mut  BCORL1_cnv  GPC3_mut  GPC3_cnv  PHF6_cnv  \n",
       "participant_id_legacy                                                        \n",
       "GBM.ICB-1                     0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-10                    0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-102                   0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-103                   0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-104                   0.0         0.0       0.0       0.0       0.0  \n",
       "...                           ...         ...       ...       ...       ...  \n",
       "GBM.ICB-37                    0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-42                    0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-43                    0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-69                    0.0         0.0       0.0       0.0       0.0  \n",
       "GBM.ICB-84                    0.0         0.0       0.0       0.0       0.0  \n",
       "\n",
       "[166 rows x 373 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "0869cc10-268f-4e42-b3c5-88734a497bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_pred_df[\"k = 6\"] = icb_pred_df[\"cluster\"].map(lambda x: x.split(\"_\")[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "88b27ac7-07c9-4304-a3e0-d25e8758af46",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_matrix = icb_matrix.join(icb_pred_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "fa9e8ebc-fda6-4d68-8e91-12e1da5e6450",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTOR_mut</th>\n",
       "      <th>MTOR_cnv</th>\n",
       "      <th>SDHB_cnv</th>\n",
       "      <th>ID3_cnv</th>\n",
       "      <th>ARID1A_mut</th>\n",
       "      <th>ARID1A_cnv</th>\n",
       "      <th>MYCL_mut</th>\n",
       "      <th>MYCL_cnv</th>\n",
       "      <th>MPL_cnv</th>\n",
       "      <th>MUTYH_cnv</th>\n",
       "      <th>...</th>\n",
       "      <th>GPC3_cnv</th>\n",
       "      <th>PHF6_cnv</th>\n",
       "      <th>cluster_1</th>\n",
       "      <th>cluster_2</th>\n",
       "      <th>cluster_3</th>\n",
       "      <th>cluster_4</th>\n",
       "      <th>cluster_5</th>\n",
       "      <th>cluster_6</th>\n",
       "      <th>cluster</th>\n",
       "      <th>k = 6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id_legacy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.074491e-118</td>\n",
       "      <td>1.264753e-189</td>\n",
       "      <td>3.715958e-72</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.015244e-90</td>\n",
       "      <td>9.048269e-168</td>\n",
       "      <td>cluster_4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.456832e-62</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.074327e-131</td>\n",
       "      <td>5.435159e-226</td>\n",
       "      <td>5.788198e-50</td>\n",
       "      <td>cluster_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-102</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>2.283606e-98</td>\n",
       "      <td>1.400666e-83</td>\n",
       "      <td>5.526794e-126</td>\n",
       "      <td>cluster_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-103</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.360496e-237</td>\n",
       "      <td>9.397800e-140</td>\n",
       "      <td>4.574012e-79</td>\n",
       "      <td>cluster_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-104</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.267862e-82</td>\n",
       "      <td>3.320733e-186</td>\n",
       "      <td>4.783794e-14</td>\n",
       "      <td>2.679068e-87</td>\n",
       "      <td>1.404272e-96</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>cluster_6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-37</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.757771e-276</td>\n",
       "      <td>1.545544e-77</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.385083e-255</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>cluster_6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.606283e-194</td>\n",
       "      <td>2.480102e-284</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>7.371171e-267</td>\n",
       "      <td>3.594879e-141</td>\n",
       "      <td>9.024257e-239</td>\n",
       "      <td>cluster_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-43</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.729017e-183</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.245601e-47</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.965613e-240</td>\n",
       "      <td>6.162254e-163</td>\n",
       "      <td>cluster_4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-69</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.617431e-32</td>\n",
       "      <td>4.383214e-135</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.905204e-74</td>\n",
       "      <td>3.223264e-107</td>\n",
       "      <td>1.049275e-152</td>\n",
       "      <td>cluster_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM.ICB-84</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.222179e-36</td>\n",
       "      <td>4.041276e-214</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.814140e-159</td>\n",
       "      <td>2.037490e-135</td>\n",
       "      <td>3.039556e-247</td>\n",
       "      <td>cluster_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>166 rows Ã— 381 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       MTOR_mut  MTOR_cnv  SDHB_cnv  ID3_cnv  ARID1A_mut  \\\n",
       "participant_id_legacy                                                      \n",
       "GBM.ICB-1                   0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-10                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-102                 0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-103                 0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-104                 0.0       0.0       0.0      0.0         0.0   \n",
       "...                         ...       ...       ...      ...         ...   \n",
       "GBM.ICB-37                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-42                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-43                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-69                  0.0       0.0       0.0      0.0         0.0   \n",
       "GBM.ICB-84                  0.0       0.0       0.0      0.0         0.0   \n",
       "\n",
       "                       ARID1A_cnv  MYCL_mut  MYCL_cnv  MPL_cnv  MUTYH_cnv  \\\n",
       "participant_id_legacy                                                       \n",
       "GBM.ICB-1                     0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-10                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-102                   0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-103                   0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-104                   0.0       0.0       0.0      0.0        0.0   \n",
       "...                           ...       ...       ...      ...        ...   \n",
       "GBM.ICB-37                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-42                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-43                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-69                    0.0       0.0       0.0      0.0        0.0   \n",
       "GBM.ICB-84                    0.0       0.0       0.0      0.0        0.0   \n",
       "\n",
       "                       ...  GPC3_cnv  PHF6_cnv      cluster_1      cluster_2  \\\n",
       "participant_id_legacy  ...                                                     \n",
       "GBM.ICB-1              ...       0.0       0.0  2.074491e-118  1.264753e-189   \n",
       "GBM.ICB-10             ...       0.0       0.0   0.000000e+00   2.456832e-62   \n",
       "GBM.ICB-102            ...       0.0       0.0   0.000000e+00   5.000000e-01   \n",
       "GBM.ICB-103            ...       0.0       0.0   0.000000e+00   0.000000e+00   \n",
       "GBM.ICB-104            ...       0.0       0.0   2.267862e-82  3.320733e-186   \n",
       "...                    ...       ...       ...            ...            ...   \n",
       "GBM.ICB-37             ...       0.0       0.0  6.757771e-276   1.545544e-77   \n",
       "GBM.ICB-42             ...       0.0       0.0  6.606283e-194  2.480102e-284   \n",
       "GBM.ICB-43             ...       0.0       0.0  8.729017e-183   0.000000e+00   \n",
       "GBM.ICB-69             ...       0.0       0.0   4.617431e-32  4.383214e-135   \n",
       "GBM.ICB-84             ...       0.0       0.0   5.222179e-36  4.041276e-214   \n",
       "\n",
       "                          cluster_3      cluster_4      cluster_5  \\\n",
       "participant_id_legacy                                               \n",
       "GBM.ICB-1              3.715958e-72   1.000000e+00   9.015244e-90   \n",
       "GBM.ICB-10             1.000000e+00  1.074327e-131  5.435159e-226   \n",
       "GBM.ICB-102            5.000000e-01   2.283606e-98   1.400666e-83   \n",
       "GBM.ICB-103            1.000000e+00  1.360496e-237  9.397800e-140   \n",
       "GBM.ICB-104            4.783794e-14   2.679068e-87   1.404272e-96   \n",
       "...                             ...            ...            ...   \n",
       "GBM.ICB-37             0.000000e+00  5.385083e-255   0.000000e+00   \n",
       "GBM.ICB-42             1.000000e+00  7.371171e-267  3.594879e-141   \n",
       "GBM.ICB-43             6.245601e-47   1.000000e+00  2.965613e-240   \n",
       "GBM.ICB-69             1.000000e+00   1.905204e-74  3.223264e-107   \n",
       "GBM.ICB-84             1.000000e+00  2.814140e-159  2.037490e-135   \n",
       "\n",
       "                           cluster_6    cluster  k = 6  \n",
       "participant_id_legacy                                   \n",
       "GBM.ICB-1              9.048269e-168  cluster_4      4  \n",
       "GBM.ICB-10              5.788198e-50  cluster_3      3  \n",
       "GBM.ICB-102            5.526794e-126  cluster_2      2  \n",
       "GBM.ICB-103             4.574012e-79  cluster_3      3  \n",
       "GBM.ICB-104             1.000000e+00  cluster_6      6  \n",
       "...                              ...        ...    ...  \n",
       "GBM.ICB-37              1.000000e+00  cluster_6      6  \n",
       "GBM.ICB-42             9.024257e-239  cluster_3      3  \n",
       "GBM.ICB-43             6.162254e-163  cluster_4      4  \n",
       "GBM.ICB-69             1.049275e-152  cluster_3      3  \n",
       "GBM.ICB-84             3.039556e-247  cluster_3      3  \n",
       "\n",
       "[166 rows x 381 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icb_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "1f30e5a0-67fa-4dd9-a293-c20b1532093c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quartz_off_screen \n",
      "                2 \n"
     ]
    }
   ],
   "source": [
    "%%R -i icb_matrix -i cnv_chrom_df -w 3000 -h 1200\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(RColorBrewer)\n",
    "\n",
    "icb_matrix <- icb_matrix[order(icb_matrix$`k = 6`),]\n",
    "\n",
    "muts <- as.matrix(icb_matrix[, grep('_mut', names(icb_matrix))])\n",
    "cnvs <- as.matrix(icb_matrix[, grep('_cnv', names(icb_matrix))])\n",
    "\n",
    "clusters <- icb_matrix[, grep('=', names(icb_matrix))]\n",
    "cluster_colors <- brewer.pal(n = 8, name = 'Dark2')\n",
    "\n",
    "chrom_colors <- rep(c(cluster_colors[6], cluster_colors[8]), length.out = 23)\n",
    "chrom_colors <- setNames(chrom_colors, 1:23)\n",
    "\n",
    "row_ha <- rowAnnotation(#Cluster_2 = clusters$`k = 2`,\n",
    "                        #Cluster_3 = clusters$`k = 3`,\n",
    "                        #Cluster_4 = clusters$`k = 4`,\n",
    "                        #Cluster_5 = clusters, #clusters$`k = 5`,\n",
    "                        Cluster_6 = clusters, #clusters$`k = 5`,\n",
    "                        col = list(#Cluster_2 = c(`1` = cluster_colors[1], `2` = cluster_colors[2]),\n",
    "                                   #Cluster_3 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3]),\n",
    "                                   #Cluster_4 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4]),\n",
    "                                   #Cluster_5 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5])),\n",
    "                                   Cluster_6 = c(`1` = cluster_colors[1], `2` = cluster_colors[2], `3` = cluster_colors[3], `4` = cluster_colors[4], `5` = cluster_colors[5], `6` = cluster_colors[6])),\n",
    "                        simple_anno_size = unit(2, 'cm'))\n",
    "\n",
    "col_ha <- columnAnnotation(df = cnv_chrom_df,\n",
    "                          simple_anno_size = unit(2, 'cm'),\n",
    "                          col = list(chrom = chrom_colors))\n",
    "\n",
    "heatmap_muts <- Heatmap(muts, name = 'SNV/indels',\n",
    "                        col = colorRamp2(c(0, 2, 4), c('white', 'grey', 'black')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "heatmap_cnvs <- Heatmap(cnvs, name = 'CNVs', right_annotation = row_ha, top_annotation = col_ha,\n",
    "                        col = colorRamp2(c(-2, 0, 2), c('blue', 'white', 'red')),\n",
    "                        cluster_rows = FALSE, cluster_columns = FALSE,\n",
    "                        show_row_names = FALSE)\n",
    "\n",
    "full_heatmap <- heatmap_muts + heatmap_cnvs\n",
    "pdf(\"icb_heatmap.pdf\", width=50, height=10)\n",
    "draw(full_heatmap)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f72e60-f3f5-4d29-8e6d-3e86df5ad15d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
